--- plugins/libnotify/gyachi-libnotify.c.orig	2011-05-19 18:17:27.000000000 -0600
+++ plugins/libnotify/gyachi-libnotify.c	2011-05-19 18:20:41.000000000 -0600
@@ -26,6 +26,7 @@
 #include <string.h>
 #include <libnotify/notification.h>
 #include <libnotify/notify.h>
+#include <gtk/gtk.h>
 
 #include "plugin_api.h"
 #include "trayicon.h"
@@ -72,7 +73,8 @@
 	win = gtk_window_new(GTK_WINDOW_TOPLEVEL);
         gtk_widget_realize(win);
 
-	my_popup=notify_notification_new_with_status_icon(mytitle, mymsg, NULL, popup_attach_systray?systray_get_status_icon():NULL);
+#if 0
+	my_popup=notify_notification_new(mytitle, mymsg, popup_attach_systray?systray_get_status_icon():NULL);
 	if (stock_id) {
 		icon_pixbuf = gtk_widget_render_icon(win, stock_id, (strcmp(stock_id, GYACHI_MAIL) == 0) ? GYACHI_ICON_SIZE_MAIL: GTK_ICON_SIZE_LARGE_TOOLBAR, "gvwindow");
 		if (icon_pixbuf) {
@@ -80,6 +82,9 @@
 		}
 		g_object_unref(icon_pixbuf);
 	}
+#else
+	my_popup=notify_notification_new(mytitle, mymsg, NULL);
+#endif
 	notify_notification_set_timeout(my_popup, timeout);
 	my_error = 0;
 	rv = notify_notification_show(my_popup, &my_error);
