--- configure.ac.tmpl.orig	2011-05-19 17:43:46.000000000 -0600
+++ configure.ac.tmpl	2011-05-19 17:44:39.000000000 -0600
@@ -15,6 +15,11 @@
 AM_GNU_GETTEXT
 AM_PROG_AS
 
+# gpgme requires LFS, see http://www.gnupg.org/documentation/manuals/gpgme/Largefile-Support-_0028LFS_0029.html
+AC_SYS_LARGEFILE
+AC_FUNC_FSEEKO
+AC_TYPE_OFF_T
+
 AC_DEFINE_UNQUOTED(GETTEXT_PACKAGE, "gyachi", [ Define package name for gettext support, in case it differs from main package ])
 AC_SUBST(GETTEXT_PACKAGE)
 
@@ -205,7 +210,7 @@
 
 dnl checks for plugin-gpgme
 if test "$enable_plugin_gpgme" != "no"; then
-	AC_CHECK_HEADER([gpgme.h], ,[AC_MSG_ERROR(cannot find include file gpgme.h. Perhaps you need to install the gpgme development package?)])
+	AC_CHECK_HEADER([gpgme/gpgme.h], ,[AC_MSG_ERROR(cannot find include file gpgme/gpgme.h. Perhaps you need to install the gpgme development package?)])
 	PLUGIN_DIRS="$PLUGIN_DIRS gpgme"
 else
 	AC_MSG_NOTICE(gpgme plugin is disabled)
--- plugins/gpgme/gyachi-gpgme.c.orig	2011-05-19 17:46:53.000000000 -0600
+++ plugins/gpgme/gyachi-gpgme.c	2011-05-19 17:47:03.000000000 -0600
@@ -24,7 +24,7 @@
  *****************************************************************************/
 
 #include "plugin_api.h"
-#include <gpgme.h>
+#include <gpgme/gpgme.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
