--- dbus/dbus-arch-deps.h.in.orig	2006-02-20 01:09:05.000000000 -0600
+++ dbus/dbus-arch-deps.h.in	2006-02-20 01:09:23.000000000 -0600
@@ -29,7 +29,7 @@
 
 #include <dbus/dbus-macros.h>
 
-DBUS_BEGIN_DECLS;
+DBUS_BEGIN_DECLS
 
 #if @DBUS_HAVE_INT64@
 #define DBUS_HAVE_INT64 1
@@ -51,6 +51,6 @@
 typedef @DBUS_INT16_TYPE@ dbus_int16_t;
 typedef unsigned @DBUS_INT16_TYPE@ dbus_uint16_t;
 
-DBUS_END_DECLS;
+DBUS_END_DECLS
 
 #endif /* DBUS_ARCH_DEPS_H */
