--- qt/Makefile.am	2006-02-16 19:04:38.000000000 -0500
+++ qt/Makefile.am.new	2006-03-15 23:39:10.000000000 -0500
@@ -62,7 +62,7 @@
 $(top_srcdir)/qt/qdbusserver.lo: qdbusserver.moc
 $(top_srcdir)/qt/qdbusconnection.lo: qdbusconnection_p.moc
 
-CLEANFILES=qdbusabstractadaptor.moc qdbusserver.moc qdbusconnection.moc
+CLEANFILES=qdbusabstractadaptor.moc qdbusserver.moc qdbusconnection_p.moc
 
 libdbus_qt4_1_la_LIBADD= $(DBUS_QT_LIBS) $(top_builddir)/dbus/libdbus-1.la
 libdbus_qt4_1_la_LDFLAGS= -version-info 1:0 -no-undefined
