diff -Nur dbus-0.60/Makefile.am dbus-0.60-new/Makefile.am
--- dbus-0.60/Makefile.am	2005-09-30 11:20:08.000000000 -0400
+++ dbus-0.60-new/Makefile.am	2006-01-02 03:15:23.000000000 -0500
@@ -6,10 +6,12 @@
 
 if HAVE_QT
    QT_SUBDIR=qt
+   QT_PC=dbus-qt-1.pc
 endif
 
 if HAVE_QT3
    QT3_SUBDIR=qt3
+   QT_PC=dbus-qt-1.pc
 endif
 
 if DBUS_USE_GCJ
@@ -36,12 +38,13 @@
 DIST_SUBDIRS=dbus bus doc glib qt qt3 gcj mono python tools test
 
 pkgconfigdir = $(libdir)/pkgconfig
-pkgconfig_DATA = dbus-1.pc $(GLIB_PC) $(MONO_PC)
+pkgconfig_DATA = dbus-1.pc $(GLIB_PC) $(MONO_PC) $(QT_PC)
 
 DISTCLEANFILES = 		\
 	dbus-1.pc		\
 	$(GLIB_PC)		\
-	$(MONO_PC)
+	$(MONO_PC)		\
+	$(QT_PC)
 
 EXTRA_DIST =			\
 	HACKING			\
diff -Nur dbus-0.60/configure.in dbus-0.60-new/configure.in
--- dbus-0.60/configure.in	2005-11-22 15:42:49.000000000 -0500
+++ dbus-0.60-new/configure.in	2006-01-02 03:10:40.000000000 -0500
@@ -1353,6 +1353,7 @@
 doc/Makefile
 dbus-1.pc
 dbus-glib-1.pc
+dbus-qt-1.pc
 dbus-sharp.pc
 test/data/valid-config-files/debug-allow-all.conf
 test/data/valid-config-files/debug-allow-all-sha1.conf
diff -Nur dbus-0.60/dbus-qt-1.pc.in dbus-0.60-new/dbus-qt-1.pc.in
--- dbus-0.60/dbus-qt-1.pc.in	1969-12-31 19:00:00.000000000 -0500
+++ dbus-0.60-new/dbus-qt-1.pc.in	2006-01-02 03:10:40.000000000 -0500
@@ -0,0 +1,12 @@
+prefix=@prefix@
+exec_prefix=@exec_prefix@
+libdir=@libdir@
+includedir=@includedir@
+
+Name: dbus-qt
+Description: QT integration for the free desktop message bus
+Version: @VERSION@
+Requires: dbus-1 qt-3.3
+Libs: -L${libdir} -ldbus-qt-1
+
+
