--- dbus-0.61/configure.in.disableqt4	2006-02-28 13:33:47.000000000 +0100
+++ dbus-0.61/configure.in	2006-02-28 13:33:29.000000000 +0100
@@ -1033,6 +1033,10 @@
     fi
     have_qt=no
 else
+ if test x$enable_qt = xno; then
+     have_qt=no;
+     have_qtest=no;
+ else
     dnl Qt4 flags
     AC_SUBST(DBUS_QT_CFLAGS)
     AC_SUBST(DBUS_QT_LIBS)
@@ -1055,6 +1059,7 @@
 
     AC_SUBST(DBUS_QTESTLIB_CFLAGS)
     AC_SUBST(DBUS_QTESTLIB_LIBS)
+  fi
 fi
 
 AM_CONDITIONAL(HAVE_QT, test x$have_qt = xyes)
