--- configure.in	2006-03-14 00:56:19.000000000 -0500
+++ configure.in.new	2006-03-14 01:02:24.000000000 -0500
@@ -290,7 +290,7 @@
 AM_CONDITIONAL(DBUS_USE_CSC, test x$enable_mono = xyes)
 
 #### Look for monodoc
-MONODOC_REQUIRED_VERSION=0.16
+MONODOC_REQUIRED_VERSION=1.1.10
 AC_SUBST(MONODOC_REQUIRED_VERSION)
 PKG_CHECK_MODULES(MONODOC_DEPENDENCY, monodoc >= $MONODOC_REQUIRED_VERSION, have_monodoc=yes, have_monodoc=no)
 
@@ -298,8 +298,9 @@
     if test x$have_monodoc = xno ; then
 	AC_MSG_ERROR([Building Mono docs explicitly required, but monodoc not found])
     else
-        AC_PATH_PROG(MONODOC, monodoc, no)
-        if test x$MONODOC = xno; then
+        AC_PATH_PROG(MONODOCER, monodocer, no)
+	AC_PATH_PROG(MDASSEMBLER, mdassembler, no)
+        if test x$MONODOCER = xno -o x$MDASSEMBLER = xno; then
            AC_MSG_ERROR([Building Mono docs explicitly required, but monodoc not found])	
 	fi
     fi
@@ -308,19 +309,26 @@
 if test x$enable_mono_docs = xauto ; then
     if test x$have_monodoc = xno ; then
         enable_mono_docs=no
-	MONODOC=
+	MONODOCER=
+	MDASSEMBLER=
     else
-        AC_PATH_PROG(MONODOC, monodoc, no)
-        if test x$MONODOC = xno; then
+        AC_PATH_PROG(MONODOCER, monodocer, no)
+	AC_PATH_PROG(MDASSEMBLER, mdassembler, no)
+        if test x$MONODOCER = xno -o x$MDASSEMBLER = xno; then
            enable_mono_docs=no
-           MONODOC=
+           MONODOCER=
+	   MDASSEMBLER=
 	else
 	   enable_mono_docs=yes
         fi
     fi
 fi
 
-AC_SUBST(MONODOC)
+MONODOC_SOURCESDIR=`pkg-config --variable=sourcesdir monodoc`
+
+AC_SUBST(MONODOCER)
+AC_SUBST(MDASSEMBLER)
+AC_SUBST(MONODOC_SOURCESDIR)
 
 AM_CONDITIONAL(ENABLE_MONODOC, test "x$enable_mono_docs" = "xyes")
 
--- mono/doc/Makefile.am	2006-02-10 16:47:10.000000000 -0500
+++ mono/doc/Makefile.am.new	2006-03-14 01:08:13.000000000 -0500
@@ -1,34 +1,30 @@
 if ENABLE_MONODOC
 TARGETS = dbus-sharp-docs.zip dbus-sharp-docs.tree dbus-sharp-docs.source
-DOCDIR = `$(MONODOC) --get-sourcesdir`
 
 install-data-local:
-	install -d -m 755 $(DESTDIR)/`monodoc --get-sourcesdir`
-	install -m 644 $(srcdir)/dbus-sharp-docs.source $(DESTDIR)/`monodoc --get-sourcesdir`
-	install -m 644 dbus-sharp-docs.tree $(DESTDIR)/`monodoc --get-sourcesdir`
-	install -m 644 dbus-sharp-docs.zip $(DESTDIR)/`monodoc --get-sourcesdir`
+	install -d -m 755 $(DESTDIR)/$(MONODOC_SOURCESDIR)
+	install -m 644 $(srcdir)/dbus-sharp-docs.source $(DESTDIR)/$(MONODOC_SOURCESDIR)
+	install -m 644 dbus-sharp-docs.tree $(DESTDIR)/$(MONODOC_SOURCESDIR)
+	install -m 644 dbus-sharp-docs.zip $(DESTDIR)/$(MONODOC_SOURCESDIR)
 
 uninstall-local:
-	rm -f $(DESTDIR)/`monodoc --get-sourcesdir`/dbus-sharp-docs.source
-	rm -f $(DESTDIR)/`monodoc --get-sourcesdir`/dbus-sharp-docs.tree
-	rm -f $(DESTDIR)/`monodoc --get-sourcesdir`/dbus-sharp-docs.zip
+	rm -f $(DESTDIR)/$(MONODOC_SOURCESDIR)/dbus-sharp-docs.source
+	rm -f $(DESTDIR)/$(MONODOC_SOURCESDIR)/dbus-sharp-docs.tree
+	rm -f $(DESTDIR)/$(MONODOC_SOURCESDIR)/dbus-sharp-docs.zip
 else
 TARGETS = 
-DOCDIR = $(datadir)
 endif
 
 noinst_DATA = $(TARGETS)
 
-ASSEMBLER = $(MONODOC) --assemble
-UPDATER = $(MONODOC) --update
-
 dbus-sharp-docs.zip: dbus-sharp-docs.tree
 
 dbus-sharp-docs.tree: $(srcdir)/en/*/*.xml $(srcdir)/en/*.xml
-	$(ASSEMBLER) --ecma $(srcdir)/en -o dbus-sharp-docs
+	$(MDASSEMBLER) --ecma $(srcdir)/en -o dbus-sharp-docs
 
-update:
-		$(UPDATER) ../dbus-sharp.dll -o ./en -f
+update: $(top_builddir)/mono/dbus-sharp.dll
+	$(MONODOCER) -assembly:$(top_builddir)/mono/dbus-sharp.dll \
+		-path:en -updateto:en -delete
 
 CLEANFILES = dbus-sharp-docs.zip dbus-sharp-docs.tree
 
