diff -ur bleachbit-0.9.0.orig/bleachbit.py bleachbit-0.9.0/bleachbit.py
--- bleachbit-0.9.0.orig/bleachbit.py	2011-09-22 01:30:18.757387142 +0200
+++ bleachbit-0.9.0/bleachbit.py	2011-09-22 01:30:52.130628710 +0200
@@ -26,14 +26,8 @@
 
 
 
-import os
 import sys
 
-if 'posix' == os.name and os.path.isdir('/usr/share/bleachbit'):
-    # This path contains bleachbit/{C,G}LI.py .  This section is
-    # unnecessary if installing BleachBit in site-packages.
-    sys.path.append('/usr/share/')
-
 
 
 
