diff -Naurp abiword-plugins-2.6.8-orig/tools/abicommand/xp/AbiCommand.cpp abiword-plugins-2.6.8/tools/abicommand/xp/AbiCommand.cpp
--- abiword-plugins-2.6.8-orig/tools/abicommand/xp/AbiCommand.cpp	2007-10-11 14:05:55.000000000 -0600
+++ abiword-plugins-2.6.8/tools/abicommand/xp/AbiCommand.cpp	2009-07-29 16:37:31.022705617 -0600
@@ -551,7 +551,7 @@ AbiCommand::parseTokens (UT_Vector * pTo
 				UT_UTF8String *pCom1 =
 					const_cast < UT_UTF8String * >(static_cast < const UT_UTF8String * >(pToks->getNthItem (1)));
 				printf(" Filename %s \n",pCom1->utf8_str());
-				char *suffix = rindex (pCom1->utf8_str (), '.');
+				const char *suffix = const_cast<const char*>(rindex (pCom1->utf8_str (), '.'));
 
 				if (suffix != NULL)
 				{
