From af8cef461c4d107f7a03645568a635d0458da9b8 Mon Sep 17 00:00:00 2001
From: Alan Hourihane <alanh@tungstengraphics.com>
Date: Wed, 24 Sep 2008 14:24:36 +0100
Subject: [PATCH] only build dri2 when DRI2 is enabled

---
 glx/Makefile.am |    6 +++++-
 1 files changed, 5 insertions(+), 1 deletions(-)

diff --git a/glx/Makefile.am b/glx/Makefile.am
index 39b96d7..0214295 100644
--- a/glx/Makefile.am
+++ b/glx/Makefile.am
@@ -45,9 +45,13 @@ glapi_sources =					\
 	glthread.h				\
 	glprocs.h
 
+if DRI2
+GLXDRI_SOURCE = glxdri2.c
+endif
+
 libglxdri_la_SOURCES = \
         glxdri.c \
-        glxdri2.c \
+	$(GLXDRI2_SOURCE) \
 	glxdricommon.h \
         extension_string.c \
         extension_string.h
-- 
1.6.0.1

