--- work.orig/common/usr/include/GL/glATI.h
+++ work/common/usr/include/GL/glATI.h
@@ -6006,6 +6006,7 @@ typedef GLvoid (APIENTRY * PFNGLDEBUGMES
 typedef GLvoid (APIENTRY * PFNGLDEBUGMESSAGECALLBACKARB)(GLvoid* callback, GLvoid* userParam);
 typedef GLuint (APIENTRY * PFNGLGETDEBUGMESSAGELOGARB)(GLuint count, GLsizei bufsize, GLenum* sources, GLenum* types, GLuint* ids, GLenum* severities, GLsizei* lengths, GLchar* messageLog);
 
+#if 0
 typedef void (APIENTRY *GLDEBUGPROCARB)(GLenum source,
                                         GLenum type,
                                         GLuint id,
@@ -6014,6 +6015,7 @@ typedef void (APIENTRY *GLDEBUGPROCARB)(
                                         const GLchar* message,
                                         GLvoid* userParam);
 // end ARB_debug_output
+#endif
 
 
 /* ARB_debug_group (KHR_debug) */
