--- work.orig/common/lib/modules/fglrx/build_mod/kcl_ioctl.c	2010-09-19 11:45:04.922760268 +0200
+++ work/common/lib/modules/fglrx/build_mod/kcl_ioctl.c	2010-09-19 12:04:29.092812959 +0200
@@ -193,7 +193,7 @@ void ATI_API_CALL KCL_IOCTL_UnregisterCo
  */
 void* ATI_API_CALL KCL_IOCTL_AllocUserSpace32(long size)
 {
-    return compat_alloc_user_space(size);
+    return arch_compat_alloc_user_space(size);
 }
 
 #endif // __x86_64__
