diff --git a/tools/gnome-session-check-accelerated-helper.c b/tools/gnome-session-check-accelerated-helper.c
index 3f83f76..b2f4f05 100644
--- a/tools/gnome-session-check-accelerated-helper.c
+++ b/tools/gnome-session-check-accelerated-helper.c
@@ -107,6 +107,7 @@ _has_hardware_gl (Display *display)
         XSetWindowAttributes cwa = { 0 };
         Window window = None;
         const char *renderer;
+        const char *vendor;
         int ret = 1;
 
         int attrlist[] = {
@@ -157,6 +158,13 @@ _has_hardware_gl (Display *display)
         if (glGetError() != GL_NO_ERROR)
                 max_texture_size = -1;
 
+        /* fglrx is currently garbage, this is Sabayon workaround
+           to avoid having fglrx and GNOME Shell enabled */
+        vendor = (const char *) glGetString (GL_VENDOR);
+        if ((strstr(vendor, "ATI ") != NULL) ||
+            (strstr(vendor, "AMD ") != NULL))
+          goto out;
+
         ret = 0;
 
 out:
