--- data/gnome-session.schemas.in	2007-01-24 15:12:52.000000000 -0600
+++ data/gnome-session.schemas.in	2007-01-24 15:13:10.000000000 -0600
@@ -50,7 +50,7 @@
          <applyto>/apps/gnome-session/options/splash_image</applyto>
          <owner>gnome</owner>
          <type>string</type>
-         <default>splash/gnome-splash.png</default>
+         <default>splash/gentoo-splash.png</default>
          <locale name="C">
             <short>Preferred Image to use for login splash screen</short>
             <long>This is a relative path value based off the $datadir/pixmaps/ directory.  Sub-directories and image names are valid values. Changing this value will effect the next session login.</long>
