From e927e44753c820450d4e2eee39ea8af2c960fb47 Mon Sep 17 00:00:00 2001
From: Dan Williams <dcbw@redhat.com>
Date: Tue, 21 Apr 2009 22:29:44 +0000
Subject: core: don't shadow basename (2)

---
diff --git a/libnm-util/tests/test-settings-defaults.c b/libnm-util/tests/test-settings-defaults.c
index 2f82038..7f0adc5 100644
--- a/libnm-util/tests/test-settings-defaults.c
+++ b/libnm-util/tests/test-settings-defaults.c
@@ -104,7 +104,7 @@ int main (int argc, char **argv)
 {
 	GError *error = NULL;
 	DBusGConnection *bus;
-	char *basename;
+	char *base;
 
 	g_type_init ();
 	bus = dbus_g_bus_get (DBUS_BUS_SESSION, NULL);
@@ -127,9 +127,9 @@ int main (int argc, char **argv)
 	test_defaults (NM_TYPE_SETTING_WIRELESS, NM_SETTING_WIRELESS_SETTING_NAME);
 	test_defaults (NM_TYPE_SETTING_WIRELESS_SECURITY, NM_SETTING_WIRELESS_SECURITY_SETTING_NAME);
 
-	basename = g_path_get_basename (argv[0]);
-	fprintf (stdout, "%s: SUCCESS\n", basename);
-	g_free (basename);
+	base = g_path_get_basename (argv[0]);
+	fprintf (stdout, "%s: SUCCESS\n", base);
+	g_free (base);
 	return 0;
 }
 
diff --git a/src/tests/test-dhcp-options.c b/src/tests/test-dhcp-options.c
index 60d3fbe..90828d9 100644
--- a/src/tests/test-dhcp-options.c
+++ b/src/tests/test-dhcp-options.c
@@ -583,7 +583,7 @@ int main (int argc, char **argv)
 {
 	GError *error = NULL;
 	DBusGConnection *bus;
-	char *basename;
+	char *base;
 
 	g_type_init ();
 	bus = dbus_g_bus_get (DBUS_BUS_SESSION, NULL);
@@ -602,9 +602,9 @@ int main (int argc, char **argv)
 	test_escaped_domain_searches ();
 	test_invalid_escaped_domain_searches ();
 
-	basename = g_path_get_basename (argv[0]);
-	fprintf (stdout, "%s: SUCCESS\n", basename);
-	g_free (basename);
+	base = g_path_get_basename (argv[0]);
+	fprintf (stdout, "%s: SUCCESS\n", base);
+	g_free (base);
 	return 0;
 }
 
diff --git a/system-settings/plugins/keyfile/tests/test-keyfile.c b/system-settings/plugins/keyfile/tests/test-keyfile.c
index 6f482b3..0f8c2cc 100644
--- a/system-settings/plugins/keyfile/tests/test-keyfile.c
+++ b/system-settings/plugins/keyfile/tests/test-keyfile.c
@@ -670,7 +670,7 @@ int main (int argc, char **argv)
 {
 	GError *error = NULL;
 	DBusGConnection *bus;
-	char *basename;
+	char *base;
 
 	g_type_init ();
 	bus = dbus_g_bus_get (DBUS_BUS_SESSION, NULL);
@@ -685,9 +685,9 @@ int main (int argc, char **argv)
 	test_read_valid_wireless_connection ();
 	test_write_wireless_connection ();
 
-	basename = g_path_get_basename (argv[0]);
-	fprintf (stdout, "%s: SUCCESS\n", basename);
-	g_free (basename);
+	base = g_path_get_basename (argv[0]);
+	fprintf (stdout, "%s: SUCCESS\n", base);
+	g_free (base);
 	return 0;
 }
 
--
cgit v0.8.2
