--- nucleo-0.6/nucleo/gl/scenegraph/sgNode.cxx.32bit	2007-02-07 15:34:05.000000000 +0100
+++ nucleo-0.6/nucleo/gl/scenegraph/sgNode.cxx	2007-02-07 15:36:11.000000000 +0100
@@ -195,7 +195,7 @@
       std::cout << "sgNode::createId: node " << std::hex << node << ", id " << id << std::endl;
     return id;
 #else
-    if ((((uintptr_t)node) >> 32) != 0)
+    if ((sizeof (void *) > 4) && (((uintptr_t)node) >> 32) != 0)
       std::cout << "sgNode::createId: got a 64-bit addressed node " << std::hex << node << std::endl;
     return (uintptr_t)node;
 #endif
