Index: nucleo/network/Makefile.am
===================================================================
RCS file: /users/asspro/roussel/cvsroot/nucleo/nucleo/network/Makefile.am,v
retrieving revision 1.9
retrieving revision 1.10
diff -u -r1.9 -r1.10
--- nucleo/nucleo/network/Makefile.am	15 Jan 2007 15:33:27 -0000	1.9
+++ nucleo/nucleo/network/Makefile.am	5 Feb 2007 13:44:22 -0000	1.10
@@ -9,7 +9,7 @@
 # this file, and for a DISCLAIMER OF ALL WARRANTIES.
 #
 
-SUBDIRS = http udp tcp
+SUBDIRS = http udp tcp dnssd
 
 AM_CPPFLAGS = -I${top_srcdir} -I${top_builddir}
 
@@ -17,12 +17,7 @@
 
 libnetwork_la_SOURCES = NetworkUtils.cxx
 
-libnetwork_la_LIBADD = udp/libudp.la tcp/libtcp.la http/libhttp.la
-
-if HAVE_DNS_SD
-SUBDIRS += dnssd
-libnetwork_la_LIBADD += dnssd/libdnssd.la 
-endif
+libnetwork_la_LIBADD = udp/libudp.la tcp/libtcp.la http/libhttp.la dnssd/libdnssd.la 
 
 # install headers
 
