--- cinepaint-0.22-1.orig/plug-ins/collect/collect.cpp  2007-03-30 09:11:33.000000000 +0000
+++ cinepaint-0.22-1/plug-ins/collect/collect.cpp       2009-08-05 22:43:14.000000000 +0000
@@ -301,7 +301,7 @@ do_collect ()
             return -1;

           // renaming the layer to the original filename
-          gimp_layer_set_name (layers[0], strrchr(fc->value(1),'/')+1);
+          gimp_layer_set_name (layers[0], strrchr(const_cast<char *>(fc->value(1)),'/')+1);
           GPrecisionType image_base_prec = gimp_drawable_precision (layers[0]);
           int base_gray = gimp_drawable_gray (layers[0]);

@@ -341,7 +341,7 @@ do_collect ()
                     #endif
                     gimp_image_add_layer (image_ID, layers[0], 0); DBG
                     // set layer name to filename
-                    gimp_layer_set_name(layers[0], strrchr(fc->value(i),'/')+1);
+                    gimp_layer_set_name(layers[0], strrchr(const_cast<char *>(fc->value(i)),'/')+1);

                     // searching max dimensions
                     if (gimp_image_width(load_image_ID) > max_w)
--- cinepaint-0.22-1.orig/plug-ins/icc_examin/icc_examin/icc_helfer_fltk.cpp    2007-04-21 09:24:09.000000000 +0000
+++ cinepaint-0.22-1/plug-ins/icc_examin/icc_examin/icc_helfer_fltk.cpp 2009-08-05 22:47:20.000000000 +0000
@@ -226,7 +226,7 @@ namespace icc_examin_ns {

   int  awake(void)
   {
-    Fl::awake(0);
+    Fl::awake();
     return 0;
   }
   int  leerWait(void) { return 0; }
