Index: cinepaint-0.22-1/app/base_frame_manager.c
===================================================================
--- cinepaint-0.22-1.orig/app/base_frame_manager.c
+++ cinepaint-0.22-1/app/base_frame_manager.c
@@ -13,6 +13,9 @@
    - bfm_check() has become public
 */ 
 
+
+#include <sys/stat.h>
+#include <sys/types.h>
 #include "base_frame_manager.h"
 #include "store_frame_manager.h"
 #include "clone.h"
Index: cinepaint-0.22-1/app/object.c
===================================================================
--- cinepaint-0.22-1.orig/app/object.c
+++ cinepaint-0.22-1/app/object.c
@@ -1,3 +1,4 @@
+#include <string.h>
 #include "object.h"
 #include "objectF.h"
 #include "objectP.h"
Index: cinepaint-0.22-1/libgimp/stdplugins-intl.h
===================================================================
--- cinepaint-0.22-1.orig/libgimp/stdplugins-intl.h
+++ cinepaint-0.22-1/libgimp/stdplugins-intl.h
@@ -23,6 +23,7 @@
 #define __STDPLUGINS_INTL_H__
 
 #include "gimpintl.h"
+#include <stdlib.h>
 
 #if defined (HAVE_BIND_TEXTDOMAIN_CODESET) && defined (GDK_WINDOWING_WIN32)
 #define BTDCS(d) bind_textdomain_codeset (d, "UTF-8")
Index: cinepaint-0.22-1/lib/drawable.c
===================================================================
--- cinepaint-0.22-1.orig/lib/drawable.c
+++ cinepaint-0.22-1/lib/drawable.c
@@ -16,6 +16,7 @@
  * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
  * Boston, MA 02111-1307, USA.
  */                                                                             
+#include <string.h>
 #include "plugin_main.h"
 #include "../lib/wire/libtile.h"
 
Index: cinepaint-0.22-1/lib/wire/wirebuffer.c
===================================================================
--- cinepaint-0.22-1.orig/lib/wire/wirebuffer.c
+++ cinepaint-0.22-1/lib/wire/wirebuffer.c
@@ -5,6 +5,7 @@
 */
 
 #include <stdio.h>
+#include <string.h>
 #include "wire.h"
 #include "wirebuffer.h"
 #include "iodebug.h"
