--- cairo-dock-1.6.3.1-old/src/cairo-dock-flying-container.c	2008-12-20 13:15:30.000000000 -0100
+++ cairo-dock-1.6.3.1/src/cairo-dock-flying-container.c	2008-12-20 13:15:53.000000000 -0100
@@ -26,6 +26,7 @@
 #include "cairo-dock-desklet.h"
 #include "cairo-dock-dock-factory.h"
 #include "cairo-dock-surface-factory.h"
+#include "cairo-dock-callbacks.h"
 #include "cairo-dock-flying-container.h"
 
 #define HAND_WIDTH 76
