--- kdelibs-3.5.5/kdeui/kiconview.cpp.orig	2005-10-10 11:06:38.000000000 -0400
+++ kdelibs-3.5.5/kdeui/kiconview.cpp	2006-10-17 04:58:58.321272299 -0400
@@ -703,7 +703,23 @@
     int textY = textRect( false ).y();
 
     if ( isSelected() ) {
-        p->fillRect( textRect( false ), cg.highlight() );
+        QRect rect = textRect( false );
+        int hRound = rect.height();
+        int wRound = rect.width();
+        QPixmap buffer(wRound * 2, hRound * 2);
+        buffer.fill( ( qGray( cg.highlight().rgb() ) > 127 ) ? black : white );
+        QPainter pBuffer(&buffer);
+        pBuffer.setBrush( cg.highlight() );
+        pBuffer.setPen( cg.highlight() );
+        pBuffer.drawRoundRect(0, 0, wRound * 2, hRound * 2,
+            1000/wRound, 1000/hRound);
+        pBuffer.end();
+        buffer.setMask( buffer.createHeuristicMask(true) );
+        QImage imageToScale = buffer.convertToImage();
+        p->drawImage(rect.x(), rect.y(),
+            imageToScale.smoothScale(wRound, hRound),
+            0, 0, -1, -1,
+            Qt::DiffuseAlphaDither | Qt::ColorOnly | Qt::AvoidDither);
         p->setPen( QPen( cg.highlightedText() ) );
     } else {
         if ( iconView()->itemTextBackground() != NoBrush )
