http://bugs.kde.org/show_bug.cgi?id=213223

--- kdelibs-4.3.3/kdecore/fakes.c
+++ kdelibs-4.3.3/kdecore/fakes.c
@@ -313,6 +313,8 @@
 #endif /* !HAVE_MKDTEMP */
 
 #ifndef HAVE_STRLCPY
+#include <string.h>
+
 KDECORE_EXPORT unsigned long strlcpy(char* d, const char* s, unsigned long bufsize)
 {
     unsigned long len, ret = strlen(s);
@@ -331,6 +333,8 @@
 #endif
 
 #ifndef HAVE_STRLCAT
+#include <string.h>
+
 KDECORE_EXPORT unsigned long strlcat(char* d, const char* s, unsigned long bufsize)
 {
     char *cp;
