--- a/src/int10/i10_v86.c
+++ b/src/int10/i10_v86.c
@@ -27,6 +27,14 @@
 #include <string.h>
 #ifdef __i386__
 #include <sys/vm86.h>
+/* 2.6.26+ kernels don't define the legacy masks. */
+#if defined(__linux__) && !defined(TF_MASK)
+      #define TF_MASK X86_EFLAGS_TF
+      #define IF_MASK X86_EFLAGS_IF
+      #define NT_MASK X86_EFLAGS_NT
+      #define VIF_MASK X86_EFLAGS_VIF
+      #define VIP_MASK X86_EFLAGS_VIP
+#endif
 #else
 #include "vm86_struct.h"
 #endif
