diff -urN hwinfo-13.11/Makefile.common hwinfo-13.11.new/Makefile.common
--- hwinfo-13.11/Makefile.common	2006-09-11 11:36:57.000000000 +0100
+++ hwinfo-13.11.new/Makefile.common	2007-01-22 18:43:43.000000000 +0000
@@ -22,7 +22,7 @@
 
 CC	= gcc
 LD	= ld
-CFLAGS	= $(RPM_OPT_FLAGS) -Wall -Wno-pointer-sign -pipe -g $(SHARED_FLAGS) $(EXTRA_FLAGS) -I$(TOPDIR)/src/hd
+CFLAGS	= $(RPM_OPT_FLAGS) -Wall -pipe -g $(SHARED_FLAGS) $(EXTRA_FLAGS) -I$(TOPDIR)/src/hd
 SHARED_FLAGS	= -fPIC
 
 LDFLAGS	= -Lsrc
