--- rtmp_def.h.old	2006-07-11 17:31:22.000000000 +0100
+++ rtmp_def.h	2006-07-11 17:41:17.000000000 +0100
@@ -820,6 +820,7 @@
  {USB_DEVICE(0x0db0,0x6877)}, /* MSI */         \
  {USB_DEVICE(0x0db0,0xa874)}, /* MSI */         \
  {USB_DEVICE(0x0db0,0xa861)}, /* MSI */         \
+ {USB_DEVICE(0x050d,0x705a)}, /* Belkin Wireless G Ver 3000uk  */ \
  {USB_DEVICE(0x07b8,0xb21d)}, /* AboCom */      \
  {USB_DEVICE(0x0769,0x31f3)}, /* Surecom */     \
  {USB_DEVICE(0x1472,0x0009)}, /* Huawei-3Com */     \
