Author: Chi-Thanh Christopher Nguyen <chithanh@cs.tu-berlin.de>
Subject: Use pkg-config to determine XPCOM_IDL_DIR instead of non-portable construct.
Bug: #284073

---

diff -ur a/macros/xpcom.m4 b/macros/xpcom.m4
--- a/macros/xpcom.m4	2009-03-02 16:29:48.000000000 +0100
+++ b/macros/xpcom.m4	2009-09-08 22:02:21.548228034 +0200
@@ -102,7 +102,7 @@
 
   if test x"${ac_cv_path_xpcom_incl}" != x; then
     XPCOM_CFLAGS="${ac_cv_path_xpcom_incl}"
-    XPCOM_IDL_CFLAGS=`echo $XPCOM_CFLAGS | sed -e 's:include:share/idl:'`
+    XPCOM_IDL_CFLAGS="-I`$PKG_CONFIG --variable=idldir libxul`/unstable"
     XPIDL=$(pkg-config libxul --variable=sdkdir)
     XPIDL="${XPIDL}xpidl"
     AC_DEFINE([HAVE_XPCOM], [1], [Use XPCOM in the NPAPI plugin])
