--- a/macros/qt4.m4	2009-03-02 16:29:42.000000000 +0100
+++ b/macros/qt4.m4	2009-09-08 14:53:31.355928666 +0200
@@ -165,8 +165,8 @@
       has_qt4="no"
     fi
 
-    AC_PATH_PROG(MOC4, moc-qt4, ,[${QTDIR}/bin /usr/lib/qt4/bin /usr/share/qt4/bin/ ${pathlist}])
-    AC_PATH_PROG(UIC4, uic-qt4, ,[${QTDIR}/bin /usr/lib/qt4/bin /usr/share/qt4/bin/ ${pathlist}])
+    AC_PATH_PROGS(MOC4, [moc-qt4 moc], ,[/usr/bin ${QTDIR}/bin /usr/lib/qt4/bin /usr/share/qt4/bin/ ${pathlist}])
+    AC_PATH_PROGS(UIC4, [uic-qt4 uic], ,[/usr/bin ${QTDIR}/bin /usr/lib/qt4/bin /usr/share/qt4/bin/ ${pathlist}])
 
 
     AC_SUBST([QT4_CFLAGS])
