From ff5404812aaf33b64ce8d4d2efedee3c8099ce9e Mon Sep 17 00:00:00 2001
From: Romain Perier <mrpouet@gentoo.org>
Date: Fri, 25 Sep 2009 17:12:53 +0200
Subject: [PATCH] Fix missing headers sub-directory for libcore when AS3 is enabled

---
 libcore/asClass.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/libcore/asClass.h b/libcore/asClass.h
index 0fe0e17..4f67c9c 100644
--- a/libcore/asClass.h
+++ b/libcore/asClass.h
@@ -33,7 +33,7 @@
 
 #ifdef ENABLE_AVM2
 # include "CodeStream.h"
-# include "AbcBlock.h"
+# include "parser/AbcBlock.h"
 #endif
 
 namespace gnash {
-- 
1.6.5.rc1

