--- firegl_public.c-orig	2006-01-21 10:54:12.007580904 -0600
+++ firegl_public.c	2006-01-21 10:56:28.515828496 -0600
@@ -2638,7 +2638,7 @@
 
     pMmPage = virt_to_page(kaddr);
 
-#if 0
+#if LINUX_VERSION_CODE >= 0x02060f
     // WARNING WARNINIG WARNNING WARNNING WARNNING WARNNING WARNNING WARNNING
     // Don't increment page usage count, cause ctx pages are allocated
     // with drm_alloc_pages, which marks all pages as reserved. Reserved
