diff -urN fglrx-install-old/common/lib/modules/fglrx/build_mod/agp3.c fglrx-install/common/lib/modules/fglrx/build_mod/agp3.c
--- fglrx-install-old/common/lib/modules/fglrx/build_mod/agp3.c	2006-09-19 19:30:42.000000000 -0500
+++ fglrx-install/common/lib/modules/fglrx/build_mod/agp3.c	2006-10-27 09:44:23.902749464 -0500
@@ -61,7 +61,6 @@
 #include <linux/modversions.h>
 #endif
 #endif
-#include <linux/config.h>
 #include <linux/module.h>
 #include <linux/pci.h>
 #include <linux/init.h>
diff -urN fglrx-install-old/common/lib/modules/fglrx/build_mod/agpgart_be.c fglrx-install/common/lib/modules/fglrx/build_mod/agpgart_be.c
--- fglrx-install-old/common/lib/modules/fglrx/build_mod/agpgart_be.c	2006-09-19 19:30:42.000000000 -0500
+++ fglrx-install/common/lib/modules/fglrx/build_mod/agpgart_be.c	2006-10-27 09:44:34.265174136 -0500
@@ -99,9 +99,6 @@
 #endif
 #endif
 
-#include <linux/config.h>
-
-
 #include <linux/module.h>
 #include <linux/types.h>
 #include <linux/kernel.h>
diff -urN fglrx-install-old/common/lib/modules/fglrx/build_mod/drm.h fglrx-install/common/lib/modules/fglrx/build_mod/drm.h
--- fglrx-install-old/common/lib/modules/fglrx/build_mod/drm.h	2006-09-19 19:30:42.000000000 -0500
+++ fglrx-install/common/lib/modules/fglrx/build_mod/drm.h	2006-10-27 09:44:52.502401656 -0500
@@ -38,7 +38,6 @@
 #define _DRM_H_
 
 #if defined(__linux__)
-#include <linux/config.h>
 #include <asm/ioctl.h>		/* For _IO* macros */
 #define DRM_IOCTL_NR(n)		_IOC_NR(n)
 #define DRM_IOC_VOID		_IOC_NONE
diff -urN fglrx-install-old/common/lib/modules/fglrx/build_mod/drmP.h fglrx-install/common/lib/modules/fglrx/build_mod/drmP.h
--- fglrx-install-old/common/lib/modules/fglrx/build_mod/drmP.h	2006-09-19 19:30:42.000000000 -0500
+++ fglrx-install/common/lib/modules/fglrx/build_mod/drmP.h	2006-10-27 09:45:03.912667032 -0500
@@ -42,7 +42,6 @@
  * can build the DRM (part of PI DRI). 4/21/2000 S + B */
 #include <asm/current.h>
 #endif /* __alpha__ */
-#include <linux/config.h>
 #include <linux/module.h>
 #include <linux/kernel.h>
 #include <linux/miscdevice.h>
diff -urN fglrx-install-old/common/lib/modules/fglrx/build_mod/firegl_public.c fglrx-install/common/lib/modules/fglrx/build_mod/firegl_public.c
--- fglrx-install-old/common/lib/modules/fglrx/build_mod/firegl_public.c	2006-09-19 19:30:42.000000000 -0500
+++ fglrx-install/common/lib/modules/fglrx/build_mod/firegl_public.c	2006-10-27 09:45:47.463046368 -0500
@@ -67,9 +67,6 @@
  * for example for CPU hot-plugging. I wonder why a desktop
  * distribution would even include such a kernel patch. */
 #ifdef CONFIG_MEM_MIRROR
-/* Prevent linux/config.h from being included again in subsequent
- * kernel headers as that would redefine CONFIG_MEM_MIRROR. */
-#include <linux/config.h>
 #warning "Disabling CONFIG_MEM_MIRROR because it does not work with non-GPL modules."
 #warning "This will break page tracking when the fglrx kernel module is used."
 #undef CONFIG_MEM_MIRROR
@@ -86,7 +83,6 @@
 #include <asm/unistd.h> /* for installing the patch wrapper */
 #include <linux/module.h>
 
-#include <linux/config.h>
 #include <linux/kernel.h>
 #include <linux/fs.h>
 #include <linux/proc_fs.h>
