--- common/lib/modules/fglrx/build_mod/kcl_debug.h.orig	2011-12-30 14:32:00.271391437 +0100
+++ common/lib/modules/fglrx/build_mod/kcl_debug.h	2011-12-30 15:48:21.647473696 +0100
@@ -85,8 +85,8 @@
 #ifdef ERROR
 #undef ERROR
 #endif
-#ifdef WARN
-#undef WARN
+#ifdef AMD_WARN
+#undef AMD_WARN
 #endif
 #ifdef INFO
 #undef INFO
@@ -122,7 +122,7 @@
 {
     SPECIAL = 0,
     ERROR  ,
-    WARN  ,
+    AMD_WARN  ,
     INFO ,
     INFOEX,
     TRACE,
@@ -163,14 +163,14 @@
 {
     {SPECIAL        ,   'S'},
     {ERROR          ,   'E'},
-    {WARN           ,   'W'},
+    {AMD_WARN       ,   'W'},
     {INFO           ,   'I'},
     {INFOEX         ,   'X'},
     {TRACE          ,   'T'},
     {PERFORMANCE    ,   'P'},
     {DUMP           ,   'D'},
 };
-#define DEFAULT_LOG_LEVEL ((U08)(___BIT(INFO) | ___BIT(INFOEX) |___BIT(ERROR) |___BIT(WARN) | ___BIT(TRACE)| ___BIT(SPECIAL)  ))
+#define DEFAULT_LOG_LEVEL ((U08)(___BIT(INFO) | ___BIT(INFOEX) |___BIT(ERROR) |___BIT(AMD_WARN) | ___BIT(TRACE)| ___BIT(SPECIAL)  ))
 #define INFO_LOG_LEVEL ((U08)(___BIT(INFO) | ___BIT(INFOEX)))
 static module_map module_type_map[LOG_M_MAX] =
 {
