--- build_mod.orig/agpgart_be.c	2005-06-19 15:59:55.000000000 +0200
+++ build_mod/agpgart_be.c	2005-06-19 15:58:59.000000000 +0200
 #if (LINUX_VERSION_CODE >= 0x020400)
 #define FGL_PM_PRESENT
 #else
@@ -7083,7 +7087,7 @@
 			if ( agp_bridge.capndx )
             {
 				printk(KERN_INFO PFX "Detected HP ZX1 AGP "
-				       "chipset at %s\n", dev->slot_name);
+				       "chipset at %s\n", pci_name(dev));
 				agp_bridge.type = HP_ZX1;
 				agp_bridge.dev = dev;
               	/* fill in the mode register */
@@ -7149,7 +7153,7 @@
                 if( cap_ptr )
                 {
     				printk(KERN_INFO PFX "Detected a Via Apollo KT400 chipset in AGP v3 mode "
-	    			       "at %s\n", dev->slot_name);
+	    			       "at %s\n", pci_name(dev));
 
                     agp_bridge.type = VIA_APOLLO_KT400;
     				agp_bridge.dev = dev;
@@ -7169,12 +7173,12 @@
                 if( cap_ptr )
                 {
     				printk(KERN_DEBUG PFX "Detected a Via Apollo KT400 chipset in AGP v2 compat mode "
-	    			       "at %s\n", dev->slot_name);
+	    			       "at %s\n", pci_name(dev));
                 }
                 else
                 {
     				printk(KERN_ERR PFX "Detected a Via Apollo KT400 chipset but no AGPv2/v3 caps found "
-	    			       "at %s\n", dev->slot_name);
+	    			       "at %s\n", pci_name(dev));
                 }
             }
             break;
@@ -7200,7 +7204,7 @@
                 if( cap_ptr )
                 {
                     printk(KERN_INFO PFX "Detected a Via P4X400 chipset in AGP v3 mode "
-                           "at %s\n", dev->slot_name);
+                           "at %s\n", pci_name(dev));
 
                     agp_bridge.type = VIA_APOLLO_P4X400;
                     agp_bridge.dev = dev;
@@ -7220,7 +7224,7 @@
                 if( cap_ptr )
                 {
                     printk(KERN_INFO PFX "Detected a Via P4X400 chipset with AGP v3.x caps in AGP v2 compatibility mode "
-                           "at %s\n", dev->slot_name);
+                           "at %s\n", pci_name(dev));
 
                     agp_bridge.type = VIA_APOLLO_P4X400;
                     agp_bridge.dev = dev;
@@ -7240,12 +7244,12 @@
                 if( cap_ptr )
                 {
                     printk(KERN_DEBUG PFX "Detected a Via P4X400 chipset with AGP v2 caps in true AGP v2 mode "
-                           "at %s\n", dev->slot_name);
+                           "at %s\n", pci_name(dev));
                 }
                 else
                 {
                     printk(KERN_ERR PFX "Detected a Via P4X400 chipset but no AGPv2/v3 caps found "
-                           "at %s\n", dev->slot_name);
+                           "at %s\n", pci_name(dev));
                 }
             }
             break;
