--- build_mod.orig/firegl_public.c	2005-10-22 10:35:00.000000000 +0200
+++ build_mod/firegl_public.c	2005-10-29 11:13:21.000000000 +0200

@@ -1590,7 +1595,11 @@
 
 int ATI_API_CALL __ke_verify_area(int type, const void * addr, unsigned long size)
 {
-    return verify_area(type, addr, size);
+#ifdef access_ok
+   return access_ok(type,addr,size) ? 0 : -EFAULT;
+#else
+   return verify_area(type, addr, size);
+#endif 
 }
 
 int ATI_API_CALL __ke_get_pci_device_info(__ke_pci_dev_t* dev, __ke_pci_device_info_t *pinfo)
