--- include/compat_wait.h
+++ include/compat_wait.h
@@ -34,13 +34,13 @@
  * 2.4.20-wolk4.0s.
  */
 
-#if VMW_HAVE_EPOLL // {
+#ifdef VMW_HAVE_EPOLL // {
 #define compat_poll_wqueues struct poll_wqueues
 #else // } {
 #define compat_poll_wqueues poll_table
 #endif // }
 
-#if VMW_HAVE_EPOLL // {
+#ifdef VMW_HAVE_EPOLL // {
 
 /* If prototype does not match, build will abort here */
 extern void poll_initwait(compat_poll_wqueues *);
