Index: b/util/grub-setup.c
===================================================================
--- a/util/grub-setup.c
+++ b/util/grub-setup.c
@@ -541,6 +541,15 @@
 
       grub_file_filter_disable_compression ();
       file = grub_file_open (core_path_dev);
+
+      if (grub_errno == GRUB_ERR_FILE_NOT_FOUND)
+	{
+	  /* Clean the previous grub_errno */
+	  grub_errno = GRUB_ERR_NONE;
+	  strcpy (core_path_dev, "/grub/core.img");
+	  file = grub_file_open (core_path_dev);
+	}
+
       if (file)
 	{
 	  if (grub_file_size (file) != core_size)
