--- bin/build-ooo.orig	2007-04-18 18:32:06.000000000 +0200
+++ bin/build-ooo	2007-04-18 18:32:45.000000000 +0200
@@ -21,9 +21,8 @@
 echo "configuring ...";
 autoconf || exit 1;
 
-echo "Env for configure:"
-set
-echo "Env for configure ends"
+echo "Log Env for configure to `pwd`/env.log"
+set > env.log
 
 eval ./configure "$CONFIGURE_OPTIONS" "$PROPAGATED_ARGS" || exit 1;
 
@@ -88,8 +88,8 @@
 # similar that gives child nmake's pain.
 unset MAKEFLAGS
 
-echo "Env:"
-set
+echo "Log Env to `pwd`/env.log"
+set > env.log
 
 # What extra localizations are avilable as $SRCDIR/GSI_*.sdf and are not
 # already merged?
