--- bin/package-ooo.orig	2010-11-29 20:16:08.700004578 +0100
+++ bin/package-ooo	2010-11-29 20:20:47.972004579 +0100
@@ -104,12 +104,12 @@
 ln -sf $OOINSTBASE/program/soffice$BINSUFFIX $PREFIX/bin/soffice$BINSUFFIX
 
 # no man-page so ...
-if test "z$VENDORNAME" != "zDebian" -a "z$VENDORNAME" != "zMandriva"; then
+if test "z$VENDORNAME" != "zDebian" -a "z$VENDORNAME" != "zMandriva" -a "z$VENDORNAME" != "zGentoo"; then
     install_script $TOOLSDIR/bin/ooconfig $PREFIX/bin/ooconfig
     install_script $OOBUILDDIR/ootool$BINSUFFIX $PREFIX/bin/ootool$BINSUFFIX
 fi
 
-if test "z$VENDORNAME" = "zMandriva"; then
+if test "z$VENDORNAME" = "zMandriva" -o "z$VENDORNAME" = "zGentoo"; then
     install_script $TOOLSDIR/bin/ooconfig $PREFIX/bin/ooconfig$BINSUFFIX
     install_script $OOBUILDDIR/ootool$BINSUFFIX $PREFIX/bin/ootool$BINSUFFIX
 fi
@@ -210,9 +210,9 @@
 	done
 	
 	# shared MIME info
-	mkdir -p $PREFIX/share/mime/packages
+	mkdir -p $DATADIR/mime/packages
 	cd $TOOLSDIR/desktop
-	cp openoffice.xml $PREFIX/share/mime/packages
+	cp openoffice.xml $DATADIR/mime/packages/libreoffice.xml
 	if test "z$RUN_POST_INSTALL_SCRIPTS" = "zyes" && 
 	   which update-mime-database >/dev/null 2>&1 ; then
 	    update-mime-database /usr/share/mime || :
