From 25a605f36544582bd454d4bc0e71e684749a0b80 Mon Sep 17 00:00:00 2001
From: Caolán McNamara <caolanm@redhat.com>
Date: Fri, 25 May 2012 14:51:04 +0000
Subject: this is supposed to just be the number of bytes

Change-Id: Iecd376aba13f94c5c62fab2f3278c5f3efda20b4
---
diff --git a/lingucomponent/source/languageguessing/simpleguesser.cxx b/lingucomponent/source/languageguessing/simpleguesser.cxx
index 383ab3e..3346ef2 100644
--- a/lingucomponent/source/languageguessing/simpleguesser.cxx
+++ b/lingucomponent/source/languageguessing/simpleguesser.cxx
@@ -117,8 +117,7 @@ vector<Guess> SimpleGuesser::GuessLanguage(const char* text)
     if (!h)
         return guesses;
 
-    //calculate le number of unicode charcters (symbols)
-    int len = utfstrlen(text);
+    int len = strlen(text);
 
     if (len > MAX_STRING_LENGTH_TO_ANALYSE)
         len = MAX_STRING_LENGTH_TO_ANALYSE;
--
cgit v0.9.0.2-2-gbebe
