--- a/vcl/unx/kde4/KDEXLib.cxx	2014-04-08 13:46:25.000000000 +0200
+++ b/vcl/unx/kde4/KDEXLib.cxx	2014-04-14 07:17:30.000000000 +0200
@@ -67,13 +67,9 @@
     eventLoopType( LibreOfficeEventLoop ),
     m_bYieldFrozen( false )
 {
-    // the timers created here means they belong to the main thread.
-    // As the timeoutTimer runs the LO event queue, which may block on a dialog,
-    // the timer has to use a Qt::QueuedConnection, otherwise the nested event
-    // loop will detect the blocking timer and drop it from the polling
-    // freezing LO X11 processing.
-    connect( &timeoutTimer, SIGNAL( timeout()), this, SLOT( timeoutActivated()), Qt::QueuedConnection );
-    connect( &userEventTimer, SIGNAL( timeout()), this, SLOT( userEventActivated()), Qt::QueuedConnection );
+    // the timers created here means they belong to the main thread
+    connect( &timeoutTimer, SIGNAL( timeout()), this, SLOT( timeoutActivated()));
+    connect( &userEventTimer, SIGNAL( timeout()), this, SLOT( userEventActivated()));
 
     // QTimer::start() can be called only in its (here main) thread, so this will
     // forward between threads if needed
