--- libreoffice-bootstrap-3.4.2.2/svx/AllLangResTarget_svx.mk.old	2011-07-24 21:49:06.000000000 +0200
+++ libreoffice-bootstrap-3.4.2.2/svx/AllLangResTarget_svx.mk	2011-07-24 22:02:11.000000000 +0200
@@ -106,11 +106,11 @@
 # hack !!!
 # just a temporary - globlmn.hrc about to be removed!
 ifeq ($(strip $(WITH_LANG)),)
-$(WORKDIR)/inc/svx/globlmn.hrc : $(realpath $(SRCDIR)/svx/inc/globlmn_tmpl.hrc)
+$(WORKDIR)/inc/svx/globlmn.hrc : $(SRCDIR)/svx/inc/globlmn_tmpl.hrc
 	echo copying $@
 	-mkdir -p $(WORKDIR)/inc/svx
-	cp $< $@
-	$(call gb_Deliver_deliver, $@, $(OUTDIR)/inc/svx/globlmn.hrc)
+	cp $(SRCDIR)/svx/inc/globlmn_tmpl.hrc $(WORKDIR)/inc/svx/globlmn.hrc
+	cp $(SRCDIR)/svx/inc/globlmn_tmpl.hrc $(OUTDIR)/inc/svx/globlmn.hrc
 	rm -f $(WORKDIR)/inc/svx/lastrun.mk
 else
 -include $(WORKDIR)/inc/svx/lastrun.mk
@@ -125,8 +125,8 @@
 	$(call gb_Helper_abbreviate_dirs_native, \
             $(gb_SrsPartMergeTarget_TRANSEXCOMMAND) \
             -p svx \
-            -i $< -o $@ -m $(realpath $(gb_SrsPartMergeTarget_SDFLOCATION)/svx/inc/localize.sdf) -l all)
-	$(call gb_Deliver_deliver, $@, $(OUTDIR)/inc/svx/globlmn.hrc)
+            -i $< -o $@ -m $(gb_SrsPartMergeTarget_SDFLOCATION)/svx/inc/localize.sdf -l all)
+	cp $(WORKDIR)/inc/svx/globlmn.hrc $(OUTDIR)/inc/svx/globlmn.hrc
 endif
 
 .PHONY : $(WORKDIR)/inc/svx/globlmn.hrc_clean
