--- sal/qa/osl/security/osl_Security.cxx.orig	2011-02-24 19:36:38.704013640 +0100
+++ sal/qa/osl/security/osl_Security.cxx	2011-02-24 19:38:50.580013586 +0100
@@ -342,8 +342,6 @@
 CPPUNIT_TEST_SUITE_REGISTRATION(osl_Security::logonUser);
 CPPUNIT_TEST_SUITE_REGISTRATION(osl_Security::getUserIdent);
 CPPUNIT_TEST_SUITE_REGISTRATION(osl_Security::getUserName);
-CPPUNIT_TEST_SUITE_REGISTRATION(osl_Security::getHomeDir);
-CPPUNIT_TEST_SUITE_REGISTRATION(osl_Security::getConfigDir);
 CPPUNIT_TEST_SUITE_REGISTRATION(osl_Security::isAdministrator);
 CPPUNIT_TEST_SUITE_REGISTRATION(osl_Security::getHandle);
 CPPUNIT_TEST_SUITE_REGISTRATION(osl_Security::UserProfile);
