--- ucb/source/ucp/webdav/NeonSession.cxx
+++ ucb/source/ucp/webdav/NeonSession.cxx
@@ -253,9 +253,6 @@
 
 // -------------------------------------------------------------------
 extern "C" int NeonSession_NeonAuth( void *       inUserData,
-#ifdef NE_FEATURE_SSPI
-                                     const char * inAuthProtocol,
-#endif
                                      const char * inRealm,
                                      int          attempt,
                                      char *       inoutUserName,
@@ -322,14 +319,6 @@
 
     bool bCanUseSystemCreds = false;
 
-#ifdef NE_FEATURE_SSPI
-    bCanUseSystemCreds
-        = (attempt == 0) && // avoid endless loops
-          ne_has_support( NE_FEATURE_SSPI ) && // Windows-only feature.
-          ( ( ne_strcasecmp( inAuthProtocol, "NTLM" ) == 0 ) ||
-            ( ne_strcasecmp( inAuthProtocol, "Negotiate" ) == 0 ) );
-#endif
-
     int theRetVal = pListener->authenticate(
                             rtl::OUString::createFromAscii( inRealm ),
                             theSession->getHostName(),
