From 0d7e2108425da1f81c20a605145679168486df59 Mon Sep 17 00:00:00 2001
From: Luboš Luňák <l.lunak@suse.cz>
Date: Fri, 06 Jan 2012 09:55:03 +0000
Subject: workaround for LO namespace pollution breaking KDE4 fpicker build

---
diff --git a/fpicker/source/unx/kde4/KDE4FilePicker.cxx b/fpicker/source/unx/kde4/KDE4FilePicker.cxx
index f282645..69be3b1 100644
--- a/fpicker/source/unx/kde4/KDE4FilePicker.cxx
+++ b/fpicker/source/unx/kde4/KDE4FilePicker.cxx
@@ -56,7 +56,10 @@
 #define Region QtXRegion
 
 //kde has an enum that uses this...OO does too
+#define LO_SETTINGS_MOUSE SETTINGS_MOUSE
 #undef SETTINGS_MOUSE
+#define LO_SETTINGS_LOCALE SETTINGS_LOCALE
+#undef SETTINGS_LOCALE
 
 #include <kfiledialog.h>
 #include <kwindowsystem.h>
@@ -70,6 +73,11 @@
 
 #undef Region
 
+#define SETTINGS_MOUSE LO_SETTINGS_MOUSE
+#undef LO_SETTINGS_MOUSE
+#define SETTINGS_LOCALE LO_SETTINGS_LOCALE
+#undef LO_SETTINGS_LOCALE
+
 using namespace ::com::sun::star;
 
 using namespace ::com::sun::star::ui::dialogs;
--
cgit v0.9.0.2-2-gbebe
