From 9540db43f3f1b11d3407a45aefda235d7e78f76b Mon Sep 17 00:00:00 2001
From: Caolán McNamara <caolanm@redhat.com>
Date: Fri, 02 Sep 2011 12:15:49 +0000
Subject: Resolves: rhbz#735182 libreoffice doesn't build with poppler-0.17.3

---
diff --git a/sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.cxx b/sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.cxx
index 5022dd5..bd6d721 100644
--- a/sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.cxx
+++ b/sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.cxx
@@ -150,7 +150,11 @@ void writeBinaryBuffer( const OutputBuffer& rBuffer )
 void writeJpeg_( OutputBuffer& o_rOutputBuf, Stream* str, bool bWithLinefeed )
 {
     // dump JPEG file as-is
+#if POPPLER_CHECK_VERSION(0, 17, 3)
+    str = str->getBaseStream();
+#else
     str = ((DCTStream *)str)->getRawStream();
+#endif
     str->reset();
 
     int c;
@@ -490,7 +494,7 @@ void PDFOutDev::endPage()
 }
 
 #if POPPLER_CHECK_VERSION(0, 17, 0)
-void PDFOutDev::processLink(AnnotLink *link, Catalog *catalog)
+void PDFOutDev::processLink(AnnotLink *link, Catalog *)
 #else
 void PDFOutDev::processLink(Link* link, Catalog*)
 #endif
--
cgit v0.9.0.2-2-gbebe
