--- libreoffice-libs-core-3.4.2.3.old/sfx2/Library_qstart.mk	2011-07-27 23:18:49.001531376 +0000
+++ libreoffice-libs-core-3.4.2.3/sfx2/Library_qstart.mk	2011-07-28 15:23:32.752723654 +0000
@@ -56,8 +56,11 @@
     $(filter-out -I%,$(GTK_CFLAGS)) \
 ))
 
+NO_AS_NEEDED=-Wl,--no-as-needed
+
 $(eval $(call gb_Library_set_ldflags,qstart_gtk,\
     $$(LDFLAGS) \
+	$(NO_AS_NEEDED) \
     $(GTK_LIBS) \
 ))
 