diff -urN libreoffice-bootstrap-3.4.2.3.old/configure.in libreoffice-bootstrap-3.4.2.3/configure.in
--- libreoffice-bootstrap-3.4.2.3.old/configure.in	2011-08-03 22:11:11.000000000 +0200
+++ libreoffice-bootstrap-3.4.2.3/configure.in	2011-08-03 22:13:55.000000000 +0200
@@ -6878,10 +6878,10 @@
 KDE_HAVE_GLIB=""
 if test "$test_kde4" = "yes" -a "$ENABLE_KDE4" = "TRUE" ; then
    qt_incdirs="$QT4INC /usr/include/qt4 /usr/include $x_includes"
-   qt_libdirs="$QT4LIB /usr/lib $x_libraries"
+   qt_libdirs="$QT4LIB /usr/lib/qt4 /usr/lib/qt /usr/lib $x_libraries"
 
-   kde_incdirs="/usr/include $x_includes"
-   kde_libdirs="/usr/lib $x_libraries"
+   kde_incdirs="/usr/include /usr/include/kde4 $x_includes"
+   kde_libdirs="/usr/lib /usr/lib/kde4 $x_libraries"
 
    if test "$host_cpu" = "x86_64" ; then
       qt_libdirs="$qt_libdirs /usr/lib64/qt4 /usr/lib64/qt /usr/lib64"
