diff --git a/configure.in b/configure.in
index f7ee08f..dab9503 100644
--- a/configure.in
+++ b/configure.in
@@ -9155,7 +9155,7 @@ if test "$SOLAR_JAVA" != "" && test "$with_junit" != "no"; then
         # check if either class-path entry is available for hamcrest or
         # it's bundled
         if "$JAVA_HOME/bin/jar" tf "$OOO_JUNIT_JAR" |grep -q hamcrest || \
-            "$UNZIP" -c "$OOO_JUNIT_JAR" META-INF/MANIFEST.MF |grep -q 'Class-Path: hamcrest'; then
+            "$UNZIP" -c "$OOO_JUNIT_JAR" META-INF/MANIFEST.MF |grep 'Class-Path:' | grep -q 'hamcrest'; then
             AC_MSG_RESULT([$OOO_JUNIT_JAR])
         else
             AC_MSG_ERROR([your junit jar neither sets a classpath nor includes hamcrest; please
