From 1296db324d5962ff904c0ed488a62132dd36a66f Mon Sep 17 00:00:00 2001
From: Rene Engelhard <rene@debian.org>
Date: Wed, 31 Oct 2012 17:00:18 +0000
Subject: fix sdext build with poppler >= 0.19.x

Conflicts:

	configure.ac

Change-Id: I655e8e15c7f7a8c292b3a1820ee48c29e847d05a
Signed-off-by: Tomas Chvatal <tchvatal@suse.cz>
---
diff --git a/sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.cxx b/sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.cxx
index 87c0ab2..9d38e9b 100644
--- a/sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.cxx
+++ b/sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.cxx
@@ -497,7 +497,9 @@ void PDFOutDev::endPage()
     printf("endPage\n");
 }
 
-#if POPPLER_CHECK_VERSION(0, 17, 0)
+#if POPPLER_CHECK_VERSION(0, 19, 0)
+void PDFOutDev::processLink(AnnotLink *link)
+#elif POPPLER_CHECK_VERSION(0, 17, 0)
 void PDFOutDev::processLink(AnnotLink *link, Catalog *)
 #else
 void PDFOutDev::processLink(Link* link, Catalog*)
diff --git a/sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.hxx b/sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.hxx
index bba8f58..1911d57 100644
--- a/sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.hxx
+++ b/sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.hxx
@@ -201,10 +201,12 @@ namespace pdfi
         // virtual void cvtDevToUser(double dx, double dy, double *ux, double *uy);
         // virtual void cvtUserToDev(double ux, double uy, int *dx, int *dy);
 
-        #if POPPLER_CHECK_VERSION(0, 17, 0)
-        virtual void processLink(AnnotLink *link, Catalog *catalog);
-        #else
         //----- link borders
+	#if POPPLER_CHECK_VERSION(0, 19, 0)
+        virtual void processLink(AnnotLink *link);
+        #elif POPPLER_CHECK_VERSION(0, 17, 0)
+	virtual void processLink(AnnotLink *link, Catalog *catalog);
+        #else
         virtual void processLink(Link *link, Catalog *catalog);
         #endif
 
--
cgit v0.9.0.2-2-gbebe
