diff -urN libreoffice-bootstrap-3.4.2.2.old/bin/distro-install-desktop-integration libreoffice-bootstrap-3.4.2.2/bin/distro-install-desktop-integration
--- libreoffice-bootstrap-3.4.2.2.old/bin/distro-install-desktop-integration	2011-07-26 16:52:39.000000000 +0200
+++ libreoffice-bootstrap-3.4.2.2/bin/distro-install-desktop-integration	2011-07-26 16:55:22.000000000 +0200
@@ -78,7 +78,7 @@
 }
 
 # install desktop integration from plain packages
-sysui_temp=`mktemp -d /tmp/distro-pack-desktop-integration-XXXXXX`
+sysui_temp=`mktemp --tmpdir -d distro-pack-desktop-integration-XXXXXX`
 cp -a sysui/unxlng*/misc/libreoffice/* "$sysui_temp"
 cp -a sysui/desktop/share/create_tree.sh "$sysui_temp"
 builddir=`pwd`
@@ -178,4 +178,4 @@
 chmod 755 "$DESTDIR$INSTALLDIR/basis$PRODUCTVERSION/program/java-set-classpath"
 test -f $DESTDIR/gid_Module_Root_Brand && echo "$INSTALLDIR/basis$PRODUCTVERSION/program/java-set-classpath" >>$DESTDIR/gid_Module_Root_Brand
 
-exit 0
\ Chybí znak konce řádku na konci souboru
+exit 0

--- libreoffice-bootstrap-3.4.2.2.old/sysui/desktop/share/create_tree.sh	2011-07-28 09:36:19.113673572 +0000
+++ libreoffice-bootstrap-3.4.2.2/sysui/desktop/share/create_tree.sh	2011-07-28 09:36:08.795995995 +0000
@@ -50,7 +50,7 @@
 ln -sf ${office_root}/program/soffice ${DESTDIR}/usr/bin/${PREFIX}
 
 cp printeradmin.sh ${DESTDIR}/usr/bin/${PREFIX}-printeradmin
-chmod 0755 ${DESTDIR}/usr/bin/${PREFIX} ${DESTDIR}/usr/bin/${PREFIX}-printeradmin
+chmod 0755 ${DESTDIR}/usr/bin/${PREFIX}-printeradmin
 
 if test "${PREFIX}" != libreoffice${PRODUCTVERSION} ; then
     # compat symlinks

