--- bin/package-ooo
+++ bin/package-ooo
@@ -126,8 +126,8 @@
 fi
 
 # create bash completion
-mkdir -p $OODESTDIR/etc/bash_completion.d
-$TOOLSDIR/bin/generate-bash-completion --binsuffix="$BINSUFFIX" $TOOLSDIR/bin/bash-completion.in $OODESTDIR/etc/bash_completion.d/libreoffice${BINSUFFIX}.sh
+mkdir -p $SYSCONFDIR/bash_completion.d
+$TOOLSDIR/bin/generate-bash-completion --binsuffix="$BINSUFFIX" $TOOLSDIR/bin/bash-completion.in $SYSCONFDIR/bash_completion.d/libreoffice${BINSUFFIX}.sh
 
 if test "z$VENDORNAME" != "zRedHat"; then
 	mkdir -p $MANDIR/man1

