--- bin/package-ooo 
+++ bin/package-ooo 
@@ -212,14 +212,6 @@
 	    done
 	done
 
-	# create symlinks below share/pixmaps to keep the backward compatibility
-	pixmapsdir=$PREFIX/share/pixmaps
-	mkdir -p $pixmapsdir
-	cd $OODESTDIR$icondir/48x48/apps
-	for icon in ooo-*.png ; do
-	    ln -sf $icondir/48x48/apps/$icon $pixmapsdir
-	done
-	
 	# shared MIME info
 	mkdir -p $DATADIR/mime/packages
 	cd $TOOLSDIR/desktop
--- bin/piece/desktop-support-app 
+++ bin/piese/desktop-support-app 
@@ -75,17 +75,6 @@
 	test -n "$list_icon_dirs" && echo "%dir $icon_dir/$size/$category" >>files-$piece.txt
     done
 
-    # create symlink below share/pixmaps to keep the backward compatibility
-    if test "$category" = "apps" ; then
-	file=$DESTDIR/$icon_dir/48x48/apps/$icon_name.png
-	if test -f $file ; then
-	    filename=`basename $file`
-	    pixmaps_dir=/usr/share/pixmaps
-	    mkdir -p $DESTDIR$pixmaps_dir
-	    ln -sf $icon_dir/48x48/apps/$filename $DESTDIR$pixmaps_dir
-	    test -n "$DESTDIR" && echo "$pixmaps_dir/$filename" >>files-$piece.txt
-	fi
-    fi
 }
 
 case "$app" in
