diff -Nurp lxdm-0.2.0.orig/src/lxdm.c lxdm-0.2.0/src/lxdm.c
--- lxdm-0.2.0.orig/src/lxdm.c	2010-04-18 15:05:01.000000000 +0200
+++ lxdm-0.2.0/src/lxdm.c	2010-08-21 13:51:56.635000006 +0200
@@ -912,7 +912,9 @@ gboolean lxdm_get_session_info(char *ses
 				g_key_file_free(cfg);
 				return FALSE;
 			}
-			name=g_key_file_get_string(cfg,"Desktop Entry","Name",NULL);
+			char *name_dir = g_strndup(session, p - session);
+			name=g_strdup(basename(name_dir));
+			g_free(name_dir);
 			exec=g_key_file_get_string(cfg,"Desktop Entry","Exec",NULL);
 			g_key_file_free(cfg);
 			if(!name || !exec)
