diff -Naur lxdm-0.2.0.a/data/lxdm.conf lxdm-0.2.0.b/data/lxdm.conf
--- lxdm-0.2.0.a/data/lxdm.conf	2010-04-17 07:40:55.000000000 -0500
+++ lxdm-0.2.0.b/data/lxdm.conf	2010-08-11 23:00:11.000000000 -0500
@@ -9,10 +9,10 @@
 
 [display]
 gtk_theme=Clearlooks
-bg=/usr/share/backgrounds/default.png
+bg=/usr/share/lxdm/themes/Sabayon/bg.jpg
 bottom_pane=1
 lang=1
-theme=Industrial
+theme=Sabayon
 
 [input]
 
