--- lxdm-0.4.0.orig/data/lxdm.conf.in
+++ lxdm-0.4.0/data/lxdm.conf.in
@@ -27,7 +27,7 @@ greeter=@FULL_LIBEXECDIR@/lxdm-greeter-gtk
 gtk_theme=Clearlooks
 
 ## background of the greeter
-bg=/usr/share/backgrounds/default.png
+bg=/usr/share/lxdm/themes/Sabayon/kgdm.png
 
 ## if show bottom pane
 bottom_pane=1
@@ -39,7 +39,7 @@ lang=1
 keyboard=0
 
 ## the theme of greeter
-theme=Industrial
+theme=Sabayon
 
 [input]
 
