--- kicker/configure.in.in	2008/01/28 11:03:28	1.1
+++ kicker/configure.in.in	2008/01/28 11:03:45
@@ -42,14 +42,14 @@
 AC_SUBST(GLIB_LIBADD)
 AC_SUBST(GLIB_LDFLAGS)
 
-dnl Check for libbeagle 0.2.0
+dnl Check for libbeagle 0.3.0
 # LIBBEAGLE_CFLAGS:  cflags for compiling libbeagle dependant sources
 # LIBBEAGLE_LIBADD:  libbeagle libraries (-l options)
 # LIBBEAGLE_LDFLAGS: flags containing path to libbeagle libraries (-L options)
 
-LIBBEAGLE_PACKAGES="libbeagle-0.0"
-LIBBEAGLE_VERSION="0.2.4"
-AC_MSG_CHECKING(for libbeagle-0.2.4 (at least $LIBBEAGLE_VERSION))
+LIBBEAGLE_PACKAGES="libbeagle-1.0"
+LIBBEAGLE_VERSION="0.3.0"
+AC_MSG_CHECKING(for libbeagle-0.3.0 (at least $LIBBEAGLE_VERSION))
 
 if $PKG_CONFIG --atleast-pkgconfig-version 0.15 ; then
    if $PKG_CONFIG --atleast-version $LIBBEAGLE_VERSION $LIBBEAGLE_PACKAGES >/dev/null 2>&1 ; then
