diff -Naur btg-0.9.7.orig/bcore/client/Makefile.am btg-0.9.7/bcore/client/Makefile.am
--- btg-0.9.7.orig/bcore/client/Makefile.am	2008-03-25 13:49:08.000000000 +0100
+++ btg-0.9.7/bcore/client/Makefile.am	2008-03-25 14:56:23.000000000 +0100
@@ -27,6 +27,6 @@
 
 libbtgclient_la_CFLAGS   = @DEBUG_CFLAGS@ -DBTG_DEBUG=@BTG_DEBUG@ -DBTG_SVNREV="@BTG_SVNREV@"
 libbtgclient_la_CXXFLAGS = @DEBUG_CXXFLAGS@ -DBTG_DEBUG=@BTG_DEBUG@ -DBTG_SVNREV="@BTG_SVNREV@"
-libbtgclient_la_LDFLAGS  = -l@BOOST_PROGRAM_OPTIONS_LIB@
+libbtgclient_la_LIBADD    = -l@BOOST_PROGRAM_OPTIONS_LIB@
 
 INCLUDES             = -I$(top_srcdir)
diff -Naur btg-0.9.7.orig/daemon/auth/Makefile.am btg-0.9.7/daemon/auth/Makefile.am
--- btg-0.9.7.orig/daemon/auth/Makefile.am	2008-03-25 13:49:08.000000000 +0100
+++ btg-0.9.7/daemon/auth/Makefile.am	2008-03-25 14:52:54.000000000 +0100
@@ -25,9 +25,9 @@
 
 btgpasswd_LDADD   = $(top_builddir)/bcore/libbtgcore.la \
                     libauth.la \
-                    $(top_builddir)/external/ini/libini.la
-
-btgpasswd_LDFLAGS  = @LIBS@ @LIBTORRENT_LIBS@ -l@BOOST_IOSTREAMS_LIB@ @LIBGNUTLS_LIBS@ @LIBGNUTLS_LIBS@ -l@BOOST_PROGRAM_OPTIONS_LIB@
+                    $(top_builddir)/external/ini/libini.la \
+                    @LIBS@ @LIBTORRENT_LIBS@ -l@BOOST_IOSTREAMS_LIB@ \
+                    @LIBGNUTLS_LIBS@ @LIBGNUTLS_LIBS@ -l@BOOST_PROGRAM_OPTIONS_LIB@
 
 btgpasswd_CFLAGS   = @CFLAGS@   @DEBUG_CFLAGS@ @LIBGNUTLS_CFLAGS@ \
 -DBTG_DEBUG=@BTG_DEBUG@ -DBTG_OPTION_SAVESESSIONS=@BTG_OPTION_SAVESESSIONS@ \
diff -Naur btg-0.9.7.orig/daemon/Makefile.am btg-0.9.7/daemon/Makefile.am
--- btg-0.9.7.orig/daemon/Makefile.am	2008-03-25 13:49:08.000000000 +0100
+++ btg-0.9.7/daemon/Makefile.am	2008-03-25 14:52:27.000000000 +0100
@@ -64,7 +64,9 @@
                     $(top_builddir)/daemon/ipfilter/libipfilter.la \
                     $(top_builddir)/daemon/auth/libauth.la $(EXTRA_LDADD) \
                     $(top_builddir)/external/ini/libini.la \
-                    $(top_builddir)/bcore/libbtgcore.la
+                    $(top_builddir)/bcore/libbtgcore.la \
+                    @LIBS@ @LIBTORRENT_LIBS@ -l@BOOST_IOSTREAMS_LIB@ \
+                    @LIBGNUTLS_LIBS@ -l@BOOST_PROGRAM_OPTIONS_LIB@
 
 #                    $(top_builddir)/bcore/libcore.la \
 #                    $(top_builddir)/bcore/externalization/libextern.la \
@@ -75,9 +77,6 @@
 #                    $(top_builddir)/external/ini/libini.la \
 #                    $(top_builddir)/bcore/auth/libbhash.la
 
-btgdaemon_LDFLAGS  = @LIBS@ @LIBTORRENT_LIBS@ -l@BOOST_IOSTREAMS_LIB@ @LIBGNUTLS_LIBS@ \
--l@BOOST_PROGRAM_OPTIONS_LIB@
-
 btgdaemon_CFLAGS   = @CFLAGS@   @DEBUG_CFLAGS@ @LIBGNUTLS_CFLAGS@ \
 -DBTG_DEBUG=@BTG_DEBUG@ -DBTG_OPTION_SAVESESSIONS=@BTG_OPTION_SAVESESSIONS@ \
 -DBTG_OPTION_EVENTCALLBACK=@BTG_OPTION_EVENTCALLBACK@ \
diff -Naur btg-0.9.7.orig/m4/rblibtorrent.m4 btg-0.9.7/m4/rblibtorrent.m4
--- btg-0.9.7.orig/m4/rblibtorrent.m4	2008-03-25 13:49:08.000000000 +0100
+++ btg-0.9.7/m4/rblibtorrent.m4	2008-03-25 13:52:44.000000000 +0100
@@ -20,7 +20,9 @@
     CXXFLAGS_SAVED="$CXXFLAGS"
     CXXFLAGS="$CXXFLAGS -I$RBLIBTORRENT_ROOT/include -I$RBLIBTORRENT_ROOT/include/libtorrent"
     LDFLAGS_SAVED="$LDFLAGS"
-    LDFLAGS="$LDFLAGS -L$RBLIBTORRENT_ROOT/lib -ltorrent"
+    LDFLAGS="$LDFLAGS -L$RBLIBTORRENT_ROOT/lib"
+    LIBS_SAVED="$LIBS"
+    LIBS="$LIBS -L$RBLIBTORRENT_ROOT/lib -ltorrent"
 
     AC_MSG_CHECKING([whether Rasterbar Libtorrent headers can be used])
 
@@ -54,6 +56,7 @@
    AC_LANG_RESTORE
    CXXFLAGS="$CXXFLAGS_SAVED"
    LDFLAGS="$LDFLAGS_SAVED"
+   LIBS="$LIBS_SAVED"
 
    if test x_$rblibtorrent_compiles != x_yes; then
      AC_MSG_ERROR(Cannot find Rasterbar Libtorrent headers)
@@ -65,7 +68,9 @@
 
 dnl   RBLIBTORRENT_CXXFLAGS="-I$RBLIBTORRENT_ROOT/include -I$RBLIBTORRENT_ROOT/include/libtorrent"
 dnl   AC_SUBST(RBLIBTORRENT_CXXFLAGS)
-dnl   RBLIBTORRENT_LDFLAGS="-L$RBLIBTORRENT_ROOT/lib -ltorrent"
+dnl   RBLIBTORRENT_LDFLAGS="-L$RBLIBTORRENT_ROOT/lib"
 dnl   AC_SUBST(RBLIBTORRENT_LDFLAGS)
+dnl   RBLIBTORRENT_LIBS="-ltorrent"
+dnl   AC_SUBST(RBLIBTORRENT_LIBS)
 
 ])dnl
