diff -up x70talk/x70.c x70talk.correct-month/x70.c
--- x70talk/x70.c	2004-05-08 08:49:00.000000000 +0100
+++ x70talk.correct-month/x70.c	2004-12-10 20:16:42.686644088 +0100
@@ -492,7 +492,7 @@ int x70_set_time(int fd, char *errbuf)
   time(&now);
   tm = localtime(&now);
   sprintf(cmdbuf, "AT+CCLK=%02d/%02d/%02d,%02d:%02d:%02d",
-          tm->tm_year-100, tm->tm_mon,  tm->tm_mday, 
+          tm->tm_year-100, tm->tm_mon+1,  tm->tm_mday, 
           tm->tm_hour, tm->tm_min,  tm->tm_sec);
 
   i = at_run(fd, cmdbuf, 0, &resp);

