--- a/vmmon-only/linux/driver.c
+++ b/vmmon-only/linux/driver.c
@@ -143,6 +143,9 @@
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 24)
 #define VMW_NOPAGE_2624
 #endif
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 36)
+#define VMW_HAVE_UNLOCKED_IOCTL
+#endif
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0) && \
     (defined(CONFIG_SMP) || defined(CONFIG_PREEMPT))
 #  define kernel_locked_by_current() kernel_locked()
