diff -ur vmci-only.before/linux/vmciKernelIf.c vmci-only/linux/vmciKernelIf.c
--- vmci-only.before/linux/vmciKernelIf.c	2010-01-02 14:56:31.000000000 -0500
+++ vmci-only/linux/vmciKernelIf.c	2010-01-02 14:57:01.000000000 -0500
@@ -36,6 +36,7 @@
 #include "compat_module.h"
 
 #include "compat_version.h"
+#include "compat_sched.h"
 #include "compat_wait.h"
 #include "compat_interrupt.h"
 #include "compat_spinlock.h"
diff -ur vmnet-only.before/vnetUserListener.c vmnet-only/vnetUserListener.c
--- vmnet-only.before/vnetUserListener.c	2010-01-02 14:56:38.000000000 -0500
+++ vmnet-only/vnetUserListener.c	2010-01-02 14:58:29.000000000 -0500
@@ -32,6 +32,7 @@
 #include <linux/netdevice.h>
 #include <linux/poll.h>
 #include "compat_skbuff.h"
+#include "compat_sched.h"
 #include "compat_wait.h"
 #include "vnetInt.h"
 
