--- mdadm-2.4.1/Monitor.c
+++ mdadm-2.4.1/Monitor.c
@@ -126,7 +126,7 @@
 			fprintf(stderr, Name ": Monitor using program \"%s\" from config file\n",
 			       alert_cmd);
 	}
-	if (scan && !mailaddr && !alert_cmd) {
+	if (scan && !mailaddr && !alert_cmd && !dosyslog) {
 		fprintf(stderr, Name ": No mail address or alert command - not monitoring.\n");
 		return 1;
 	}
@@ -465,7 +465,7 @@
 {
 	int priority;
 
-	if (!cmd && !mailaddr) {
+	if (!cmd && !mailaddr && !dosyslog) {
 		time_t now = time(0);
 	       
 		printf("%1.15s: %s on %s %s\n", ctime(&now)+4, event, dev, disc?disc:"unknown device");
--- mdadm-2.4.1/ReadMe.c
+++ mdadm-2.4.1/ReadMe.c
@@ -463,6 +464,7 @@
 "  --mail=       -m   : Address to mail alerts of failure to\n"
 "  --program=    -p   : Program to run when an event is detected\n"
 "  --alert=           : same as --program\n"
+"  --syslog      -y   : Report alerts via syslog\n"
 "  --delay=      -d   : seconds of delay between polling state. default=60\n"
 "  --config=     -c   : specify a different config file\n"
 "  --scan        -s   : find mail-address/program in config file\n"
