diff -Naur NVIDIA.orig/usr/src/nv/nv.c NVIDIA-Linux-x86-1.0-8776-pkg1/usr/src/nv/nv.c
--- NVIDIA.orig/usr/src/nv/nv.c	2006-10-17 01:31:37.000000000 -0400
+++ NVIDIA-Linux-x86-1.0-8776-pkg1/usr/src/nv/nv.c	2006-11-29 21:58:20.000000000 -0500
@@ -252,7 +252,7 @@
 long          nv_kern_unlocked_ioctl(struct file *, unsigned int, unsigned long);
 long          nv_kern_compat_ioctl(struct file *, unsigned int, unsigned long);
 void          nv_kern_isr_bh(unsigned long);
-irqreturn_t   nv_kern_isr(int, void *, struct pt_regs *);
+irqreturn_t   nv_kern_isr(int, void *);
 void          nv_kern_rc_timer(unsigned long);
 #if defined(NV_PM_SUPPORT_OLD_STYLE_APM)
 static int    nv_kern_apm_event(struct pm_dev *, pm_request_t, void *);
@@ -2561,8 +2561,7 @@
  */
 irqreturn_t nv_kern_isr(
     int   irq,
-    void *arg,
-    struct pt_regs *regs
+    void *arg
 )
 {
     nv_linux_state_t *nvl = (void *) arg;
diff -Naur NVIDIA.orig/usr/src/nv/nv-i2c.c NVIDIA-Linux-x86-1.0-8776-pkg1/usr/src/nv/nv-i2c.c
--- NVIDIA.orig/usr/src/nv/nv-i2c.c	2006-10-17 01:31:37.000000000 -0400
+++ NVIDIA-Linux-x86-1.0-8776-pkg1/usr/src/nv/nv-i2c.c	2006-11-29 21:57:36.000000000 -0500
@@ -23,8 +23,6 @@
 static struct i2c_algorithm nv_i2c_algo = {
     .master_xfer      = nv_i2c_algo_xfer,
     .smbus_xfer       = NULL,
-    .slave_send       = NULL,
-    .slave_recv       = NULL,
     .algo_control     = nv_i2c_algo_control,
     .functionality    = nv_i2c_algo_functionality,
 };
diff -Naur NVIDIA.orig/usr/src/nv/nv-linux.h NVIDIA-Linux-x86-1.0-8776-pkg1/usr/src/nv/nv-linux.h
--- NVIDIA.orig/usr/src/nv/nv-linux.h	2006-10-17 01:31:37.000000000 -0400
+++ NVIDIA-Linux-x86-1.0-8776-pkg1/usr/src/nv/nv-linux.h	2006-11-29 21:57:48.000000000 -0500
@@ -14,7 +14,6 @@
 
 #include "nv.h"
 
-#include <linux/config.h>
 #include <linux/version.h>
 #include <linux/utsname.h>
