--- a/kernel/common/inc/nv-linux.h
+++ b/kernel/common/inc/nv-linux.h
@@ -1209,6 +1209,7 @@ static inline NvU32 nv_alloc_init_flags(
 static inline NvBool nv_dma_maps_swiotlb(struct pci_dev *dev)
 {
     NvBool swiotlb_in_use = NV_FALSE;
+#if 0
 #if defined(CONFIG_SWIOTLB)
   #if defined(NV_DMA_OPS_PRESENT) || defined(NV_GET_DMA_OPS_PRESENT)
     /*
@@ -1251,7 +1252,7 @@ static inline NvBool nv_dma_maps_swiotlb
     swiotlb_in_use = (swiotlb == 1);
   #endif
 #endif
-
+#endif
     return swiotlb_in_use;
 }
