--- usr/src/nv/nv-linux.h.orig	2010-10-28 04:23:04.000000000 +0200
+++ usr/src/nv/nv-linux.h	2011-05-21 08:58:35.714633248 +0200
@@ -80,7 +80,6 @@
 #if !defined(KERNEL_2_4)
 #include <linux/sched.h>            /* suser(), capable() replacement   */
 #include <linux/moduleparam.h>      /* module_param()                   */
-#include <linux/smp_lock.h>         /* kernel_locked                    */
 #include <asm/tlbflush.h>           /* flush_tlb(), flush_tlb_all()     */
 #include <asm/kmap_types.h>         /* page table entry lookup          */
 #endif
