diff -Nurp NVIDIA-Linux-x86_64-173.14.37-pkg2.orig/usr/src/nv/nv-linux.h NVIDIA-Linux-x86_64-173.14.37-pkg2/usr/src/nv/nv-linux.h
--- NVIDIA-Linux-x86_64-173.14.37-pkg2.orig/usr/src/nv/nv-linux.h	2013-09-06 12:25:00.974466183 +0200
+++ NVIDIA-Linux-x86_64-173.14.37-pkg2/usr/src/nv/nv-linux.h	2013-09-06 12:28:44.978455609 +0200
@@ -1466,4 +1466,10 @@ int nv_verify_page_mappings(nv_pte_t *,
 #include "nv-linux-ext.h"
 #endif
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 11, 0)
+#define NV_NUM_PHYSPAGES		get_num_physpages()
+#else
+#define NV_NUM_PHYSPAGES		num_physpages
+#endif
+
 #endif  /* _NV_LINUX_H_ */
diff -Nurp NVIDIA-Linux-x86_64-173.14.37-pkg2.orig/usr/src/nv/os-interface.c NVIDIA-Linux-x86_64-173.14.37-pkg2/usr/src/nv/os-interface.c
--- NVIDIA-Linux-x86_64-173.14.37-pkg2.orig/usr/src/nv/os-interface.c	2013-03-07 03:58:57.000000000 +0100
+++ NVIDIA-Linux-x86_64-173.14.37-pkg2/usr/src/nv/os-interface.c	2013-09-06 12:28:55.577455109 +0200
@@ -292,7 +292,7 @@ NvU64 NV_API_CALL os_get_page_mask(void)
 
 NvU64 NV_API_CALL os_get_system_memory_size(void)
 {
-    return ((NvU64) num_physpages * PAGE_SIZE) / RM_PAGE_SIZE;
+    return ((NvU64) NV_NUM_PHYSPAGES * PAGE_SIZE) / RM_PAGE_SIZE;
 }
 
 //
