--- a/kernel/nv-drm.c
+++ b/kernel/nv-drm.c
@@ -201,12 +201,7 @@
     memset(&nv_obj->base, 0, sizeof(nv_obj->base));
     nv_obj->pages = pages;
 
-    ret = drm_gem_private_object_init(nvl->drm, &nv_obj->base, size);
-    if (ret)
-    {
-        status = RM_ERR_OPERATING_SYSTEM;
-        goto done;
-    }
+    drm_gem_private_object_init(nvl->drm, &nv_obj->base, size);
 
     ret = drm_gem_handle_create(file_priv, &nv_obj->base, handle);
     if (ret)
