--- sw/source/core/doc/docbm.cxx	2009-05-17 16:41:34.765483089 -0500
+++ sw/source/core/doc/docbm.cxx	2009-05-17 16:41:39.141355157 -0500
@@ -33,6 +33,7 @@
 #include "precompiled_sw.hxx"
 
 
+#include <cstdio>
 #include <tools/pstm.hxx>
 #include <svx/linkmgr.hxx>
 #include <fmtanchr.hxx>
--- sw/source/filter/ww8/WW8TableInfo.cxx	2009-05-17 17:06:33.556313757 -0500
+++ sw/source/filter/ww8/WW8TableInfo.cxx	2009-05-17 17:06:46.964547351 -0500
@@ -32,6 +32,7 @@
 #include "precompiled_sw.hxx"
 
 #include <iostream> 
+#include <cstdio>
 #include "WW8TableInfo.hxx"
 #include "swtable.hxx"
 #include "pam.hxx"
