--- cli_ure/source/climaker/makefile.mk 
+++ cli_ure/source/climaker/makefile.mk 
@@ -62,7 +62,7 @@
 
 $(BIN)$/climaker : climaker_csharp.cs
 	gmcs -debug -unsafe climaker_csharp.cs -reference:$(BIN)/cli_basetypes.dll -out:$(BIN)$/climaker.exe
-	mkbundle2 --static -L $(OUT)/lib --deps -o $(BIN)$/climaker $(BIN)$/climaker.exe $(BIN)/cli_basetypes.dll
+	mkbundle2 -L $(OUT)/lib --deps -o $(BIN)$/climaker $(BIN)$/climaker.exe $(BIN)/cli_basetypes.dll
  
 .INCLUDE :  target.mk
 .ENDIF # "$(ENABLE_MONO_CLIMAKER)" != "YES"

