--- bin/package-ooo 
+++ bin/package-ooo 
@@ -33,6 +33,7 @@
 {
     cat <<EOT >$1
 #!/bin/sh
+unset PYTHONPATH
 $OOINSTBASE/program/soffice $2 "\$@"
 EOT
     chmod 755 $1
--- patches/dev300/apply 
+++ patches/dev300/apply 
@@ -1458,6 +1458,10 @@
 [ GentooOnly ]
 SectionOwner => aprosky
 
+# No java build
+nojavanostax.diff
+# Fix freezes with dict|hyph-stuff in one dir
+hunspell-one-dir-nocrash.diff
 # Allow build to proceed with PaX enabled
 gentoo-pax-fix.diff, aprosky
 # support server-only versions of 64bit JDKs
@@ -1527,6 +1527,9 @@
 mono-climaker-config.diff
 mono-climaker-bridgetest.diff
 
+# Buildfix with Mono 2.2, aprosky
+buildfix-mono-2-2.diff
+
 [ VBAObjects ]
 SectionOwner => noelpwer
 

