--- bin/ooinstall	2009-09-05 23:39:18.000000000 +0300
+++ bin/ooinstall	2009-09-05 23:44:28.000000000 +0300
@@ -156,7 +156,7 @@
 	"-buildid $BUILD $destdir $strip " .
 	"-simple $path") && die "Failed to install: $!";
 
-if (($setup_vars{'VENDORNAME'} eq 'Novell' || $setup_vars{'VENDORNAME'} eq 'Debian' || $setup_vars{'VENDORNAME'} eq 'Mandriva' || $setup_vars{'VENDORNAME'} eq 'Frugalware' ) && $configure_vars{BUILD_TYPE} =~ m/ODK/) {
+if (($setup_vars{'VENDORNAME'} eq 'Novell' || $setup_vars{'VENDORNAME'} eq 'Debian' || $setup_vars{'VENDORNAME'} eq 'Mandriva' || $setup_vars{'VENDORNAME'} eq 'Frugalware' || $setup_vars{'VENDORNAME'} eq 'Gentoo') && $configure_vars{BUILD_TYPE} =~ m/ODK/) {
     print "Running SDK installer\n";
     system ("cd $setup_vars{OOBUILDDIR}/instsetoo_native/util ; " .
 	"perl -w $configure_vars{SOLARENV}/bin/make_installer.pl " .
