? unxlngx6.pro
Index: source/lingutil/lingutil.cxx
===================================================================
RCS file: /cvs/whiteboard/lingucomponent/source/lingutil/lingutil.cxx,v
retrieving revision 1.4
diff -u -r1.4 lingutil.cxx
--- lingucomponent/source/lingutil/lingutil.cxx	13 Jun 2008 14:02:52 -0000	1.4
+++ lingucomponent/source/lingutil/lingutil.cxx	24 Oct 2008 15:01:43 -0000
@@ -322,11 +322,15 @@
 		{
 			LanguageType nLang = MsLangId::convertIsoStringToLanguage( aIt2->aLocaleNames[0] );
 
+            if (nLang == LANGUAGE_DONTKNOW || nLang == LANGUAGE_NONE)
+            {
+                DBG_ERROR( "old style dictionary with invalid language found!" );
+                continue;
+            }
+
 			// language not yet added?
 			if (aNewStyleLanguages.count( nLang ) == 0)
-			{
 				rNewStyleDics.push_back( *aIt2 );
-			}
 		}
 		else
 		{
