--- config_office/configure.in 
+++ config_office/configure.in 
@@ -3452,6 +3452,7 @@
 dnl ===================================================================
 dnl Check for building stax
 dnl ===================================================================
+if test "$SOLAR_JAVA" != ""; then
 AC_MSG_CHECKING([whether to build the stax])
 if test -f "../external/StAX/jsr173_1.0_api.jar"; then
    BUILD_STAX=NO
@@ -3459,6 +3460,7 @@
 else
    BUILD_STAX=YES
    AC_MSG_RESULT([yes])
+fi
 fi
 AC_SUBST(BUILD_STAX)
 
--- stax/makefile.mk 
+++ stax/makefile.mk 
@@ -35,12 +35,13 @@
 TARGET=stax
 USE_JAVAVER=TRUE
 
+.IF "$(SOLAR_JAVA)" != ""
+
 # --- Settings -----------------------------------------------------
 
 .INCLUDE :	settings.mk
 .INCLUDE : antsettings.mk
 
-.IF "$(SOLAR_JAVA)" != ""
 .IF "$(JAVANUMVER:s/.//)" >= "000100060000" || "$(JDK)"=="gcj"
 all:
 	@echo "Your java version already contains StAX"
@@ -68,14 +69,20 @@
 
 OUT2CLASS=jsr173_1.0_api.jar
 
+.ENDIF			#  "$(BUILD_STAX)" != "YES"
+
+.ELSE			# $(SOLAR_JAVA)==""
+all:
+	@echo "Nothing to do: SOLAR_JAVA not set"
 .ENDIF			# $(SOLAR_JAVA)!= ""
 # --- Targets ------------------------------------------------------
 
+.IF "$(SOLAR_JAVA)" != ""
 .INCLUDE : set_ext.mk
+.ENDIF
 .INCLUDE : target.mk
 .IF "$(SOLAR_JAVA)" != ""
 .INCLUDE : tg_ext.mk
-.ENDIF
 
 .ENDIF			#  "$(BUILD_STAX)" != "YES"
 

