unchanged:
--- solenv/inc/_tg_shl.mk
+++ solenv/inc/_tg_shl.mk
@@ -441,10 +441,10 @@ $(SHL1TARGETN) : \
 .ENDIF
 .ELSE			# "$(OS)"=="MACOSX"
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_1.cmd
-	@echo $(SHL1LINKER) $(SHL1LINKFLAGS) $(SHL1SONAME) $(LINKFLAGSSHL) $(SHL1VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(SOLARLIB) $(STDSLO) $(SHL1OBJS:s/.obj/.o/) \
+	@echo $(SHL1LINKER) $(SHL1LINKFLAGS) $(SHL1SONAME) $(LINKFLAGSSHL) $(SHL1VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(STDSLO) $(SHL1OBJS:s/.obj/.o/) \
 	$(SHL1VERSIONOBJ) -o $@ \
 	`cat /dev/null $(SHL1LIBS) | tr -s " " "\n" | $(SED) s\#$(ROUT)\#$(PRJ)$/$(ROUT)\#g` \
-	$(SHL1STDLIBS) $(SHL1ARCHIVES) $(SHL1STDSHL) $(STDSHL1) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_1.cmd
+	$(SHL1STDLIBS) $(SHL1ARCHIVES) $(SHL1STDSHL) $(STDSHL1) $(SOLARLIB) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_1.cmd
 	@cat $(MISC)$/$(TARGET).$(@:b)_1.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_1.cmd
 .IF "$(UPDATER)"=="YES"
@@ -981,10 +981,10 @@ $(SHL2TARGETN) : \
 .ENDIF
 .ELSE			# "$(OS)"=="MACOSX"
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_2.cmd
-	@echo $(SHL2LINKER) $(SHL2LINKFLAGS) $(SHL2SONAME) $(LINKFLAGSSHL) $(SHL2VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(SOLARLIB) $(STDSLO) $(SHL2OBJS:s/.obj/.o/) \
+	@echo $(SHL2LINKER) $(SHL2LINKFLAGS) $(SHL2SONAME) $(LINKFLAGSSHL) $(SHL2VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(STDSLO) $(SHL2OBJS:s/.obj/.o/) \
 	$(SHL2VERSIONOBJ) -o $@ \
 	`cat /dev/null $(SHL2LIBS) | tr -s " " "\n" | $(SED) s\#$(ROUT)\#$(PRJ)$/$(ROUT)\#g` \
-	$(SHL2STDLIBS) $(SHL2ARCHIVES) $(SHL2STDSHL) $(STDSHL2) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_2.cmd
+	$(SHL2STDLIBS) $(SHL2ARCHIVES) $(SHL2STDSHL) $(STDSHL2) $(SOLARLIB) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_2.cmd
 	@cat $(MISC)$/$(TARGET).$(@:b)_2.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_2.cmd
 .IF "$(UPDATER)"=="YES"
@@ -1521,10 +1521,10 @@ $(SHL3TARGETN) : \
 .ENDIF
 .ELSE			# "$(OS)"=="MACOSX"
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_3.cmd
-	@echo $(SHL3LINKER) $(SHL3LINKFLAGS) $(SHL3SONAME) $(LINKFLAGSSHL) $(SHL3VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(SOLARLIB) $(STDSLO) $(SHL3OBJS:s/.obj/.o/) \
+	@echo $(SHL3LINKER) $(SHL3LINKFLAGS) $(SHL3SONAME) $(LINKFLAGSSHL) $(SHL3VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(STDSLO) $(SHL3OBJS:s/.obj/.o/) \
 	$(SHL3VERSIONOBJ) -o $@ \
 	`cat /dev/null $(SHL3LIBS) | tr -s " " "\n" | $(SED) s\#$(ROUT)\#$(PRJ)$/$(ROUT)\#g` \
-	$(SHL3STDLIBS) $(SHL3ARCHIVES) $(SHL3STDSHL) $(STDSHL3) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_3.cmd
+	$(SHL3STDLIBS) $(SHL3ARCHIVES) $(SHL3STDSHL) $(STDSHL3) $(SOLARLIB) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_3.cmd
 	@cat $(MISC)$/$(TARGET).$(@:b)_3.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_3.cmd
 .IF "$(UPDATER)"=="YES"
@@ -2061,10 +2061,10 @@ $(SHL4TARGETN) : \
 .ENDIF
 .ELSE			# "$(OS)"=="MACOSX"
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_4.cmd
-	@echo $(SHL4LINKER) $(SHL4LINKFLAGS) $(SHL4SONAME) $(LINKFLAGSSHL) $(SHL4VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(SOLARLIB) $(STDSLO) $(SHL4OBJS:s/.obj/.o/) \
+	@echo $(SHL4LINKER) $(SHL4LINKFLAGS) $(SHL4SONAME) $(LINKFLAGSSHL) $(SHL4VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(STDSLO) $(SHL4OBJS:s/.obj/.o/) \
 	$(SHL4VERSIONOBJ) -o $@ \
 	`cat /dev/null $(SHL4LIBS) | tr -s " " "\n" | $(SED) s\#$(ROUT)\#$(PRJ)$/$(ROUT)\#g` \
-	$(SHL4STDLIBS) $(SHL4ARCHIVES) $(SHL4STDSHL) $(STDSHL4) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_4.cmd
+	$(SHL4STDLIBS) $(SHL4ARCHIVES) $(SHL4STDSHL) $(STDSHL4) $(SOLARLIB) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_4.cmd
 	@cat $(MISC)$/$(TARGET).$(@:b)_4.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_4.cmd
 .IF "$(UPDATER)"=="YES"
@@ -2601,10 +2601,10 @@ $(SHL5TARGETN) : \
 .ENDIF
 .ELSE			# "$(OS)"=="MACOSX"
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_5.cmd
-	@echo $(SHL5LINKER) $(SHL5LINKFLAGS) $(SHL5SONAME) $(LINKFLAGSSHL) $(SHL5VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(SOLARLIB) $(STDSLO) $(SHL5OBJS:s/.obj/.o/) \
+	@echo $(SHL5LINKER) $(SHL5LINKFLAGS) $(SHL5SONAME) $(LINKFLAGSSHL) $(SHL5VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(STDSLO) $(SHL5OBJS:s/.obj/.o/) \
 	$(SHL5VERSIONOBJ) -o $@ \
 	`cat /dev/null $(SHL5LIBS) | tr -s " " "\n" | $(SED) s\#$(ROUT)\#$(PRJ)$/$(ROUT)\#g` \
-	$(SHL5STDLIBS) $(SHL5ARCHIVES) $(SHL5STDSHL) $(STDSHL5) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_5.cmd
+	$(SHL5STDLIBS) $(SHL5ARCHIVES) $(SHL5STDSHL) $(STDSHL5) $(SOLARLIB) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_5.cmd
 	@cat $(MISC)$/$(TARGET).$(@:b)_5.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_5.cmd
 .IF "$(UPDATER)"=="YES"
@@ -3141,10 +3141,10 @@ $(SHL6TARGETN) : \
 .ENDIF
 .ELSE			# "$(OS)"=="MACOSX"
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_6.cmd
-	@echo $(SHL6LINKER) $(SHL6LINKFLAGS) $(SHL6SONAME) $(LINKFLAGSSHL) $(SHL6VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(SOLARLIB) $(STDSLO) $(SHL6OBJS:s/.obj/.o/) \
+	@echo $(SHL6LINKER) $(SHL6LINKFLAGS) $(SHL6SONAME) $(LINKFLAGSSHL) $(SHL6VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(STDSLO) $(SHL6OBJS:s/.obj/.o/) \
 	$(SHL6VERSIONOBJ) -o $@ \
 	`cat /dev/null $(SHL6LIBS) | tr -s " " "\n" | $(SED) s\#$(ROUT)\#$(PRJ)$/$(ROUT)\#g` \
-	$(SHL6STDLIBS) $(SHL6ARCHIVES) $(SHL6STDSHL) $(STDSHL6) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_6.cmd
+	$(SHL6STDLIBS) $(SHL6ARCHIVES) $(SHL6STDSHL) $(STDSHL6) $(SOLARLIB) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_6.cmd
 	@cat $(MISC)$/$(TARGET).$(@:b)_6.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_6.cmd
 .IF "$(UPDATER)"=="YES"
@@ -3681,10 +3681,10 @@ $(SHL7TARGETN) : \
 .ENDIF
 .ELSE			# "$(OS)"=="MACOSX"
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_7.cmd
-	@echo $(SHL7LINKER) $(SHL7LINKFLAGS) $(SHL7SONAME) $(LINKFLAGSSHL) $(SHL7VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(SOLARLIB) $(STDSLO) $(SHL7OBJS:s/.obj/.o/) \
+	@echo $(SHL7LINKER) $(SHL7LINKFLAGS) $(SHL7SONAME) $(LINKFLAGSSHL) $(SHL7VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(STDSLO) $(SHL7OBJS:s/.obj/.o/) \
 	$(SHL7VERSIONOBJ) -o $@ \
 	`cat /dev/null $(SHL7LIBS) | tr -s " " "\n" | $(SED) s\#$(ROUT)\#$(PRJ)$/$(ROUT)\#g` \
-	$(SHL7STDLIBS) $(SHL7ARCHIVES) $(SHL7STDSHL) $(STDSHL7) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_7.cmd
+	$(SHL7STDLIBS) $(SHL7ARCHIVES) $(SHL7STDSHL) $(STDSHL7) $(SOLARLIB) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_7.cmd
 	@cat $(MISC)$/$(TARGET).$(@:b)_7.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_7.cmd
 .IF "$(UPDATER)"=="YES"
@@ -4221,10 +4221,10 @@ $(SHL8TARGETN) : \
 .ENDIF
 .ELSE			# "$(OS)"=="MACOSX"
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_8.cmd
-	@echo $(SHL8LINKER) $(SHL8LINKFLAGS) $(SHL8SONAME) $(LINKFLAGSSHL) $(SHL8VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(SOLARLIB) $(STDSLO) $(SHL8OBJS:s/.obj/.o/) \
+	@echo $(SHL8LINKER) $(SHL8LINKFLAGS) $(SHL8SONAME) $(LINKFLAGSSHL) $(SHL8VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(STDSLO) $(SHL8OBJS:s/.obj/.o/) \
 	$(SHL8VERSIONOBJ) -o $@ \
 	`cat /dev/null $(SHL8LIBS) | tr -s " " "\n" | $(SED) s\#$(ROUT)\#$(PRJ)$/$(ROUT)\#g` \
-	$(SHL8STDLIBS) $(SHL8ARCHIVES) $(SHL8STDSHL) $(STDSHL8) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_8.cmd
+	$(SHL8STDLIBS) $(SHL8ARCHIVES) $(SHL8STDSHL) $(STDSHL8) $(SOLARLIB) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_8.cmd
 	@cat $(MISC)$/$(TARGET).$(@:b)_8.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_8.cmd
 .IF "$(UPDATER)"=="YES"
@@ -4761,10 +4761,10 @@ $(SHL9TARGETN) : \
 .ENDIF
 .ELSE			# "$(OS)"=="MACOSX"
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_9.cmd
-	@echo $(SHL9LINKER) $(SHL9LINKFLAGS) $(SHL9SONAME) $(LINKFLAGSSHL) $(SHL9VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(SOLARLIB) $(STDSLO) $(SHL9OBJS:s/.obj/.o/) \
+	@echo $(SHL9LINKER) $(SHL9LINKFLAGS) $(SHL9SONAME) $(LINKFLAGSSHL) $(SHL9VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(STDSLO) $(SHL9OBJS:s/.obj/.o/) \
 	$(SHL9VERSIONOBJ) -o $@ \
 	`cat /dev/null $(SHL9LIBS) | tr -s " " "\n" | $(SED) s\#$(ROUT)\#$(PRJ)$/$(ROUT)\#g` \
-	$(SHL9STDLIBS) $(SHL9ARCHIVES) $(SHL9STDSHL) $(STDSHL9) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_9.cmd
+	$(SHL9STDLIBS) $(SHL9ARCHIVES) $(SHL9STDSHL) $(STDSHL9) $(SOLARLIB) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_9.cmd
 	@cat $(MISC)$/$(TARGET).$(@:b)_9.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_9.cmd
 .IF "$(UPDATER)"=="YES"
@@ -5301,10 +5301,10 @@ $(SHL10TARGETN) : \
 .ENDIF
 .ELSE			# "$(OS)"=="MACOSX"
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_10.cmd
-	@echo $(SHL10LINKER) $(SHL10LINKFLAGS) $(SHL10SONAME) $(LINKFLAGSSHL) $(SHL10VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(SOLARLIB) $(STDSLO) $(SHL10OBJS:s/.obj/.o/) \
+	@echo $(SHL10LINKER) $(SHL10LINKFLAGS) $(SHL10SONAME) $(LINKFLAGSSHL) $(SHL10VERSIONMAPPARA) -L$(PRJ)$/$(ROUT)$/lib $(STDSLO) $(SHL10OBJS:s/.obj/.o/) \
 	$(SHL10VERSIONOBJ) -o $@ \
 	`cat /dev/null $(SHL10LIBS) | tr -s " " "\n" | $(SED) s\#$(ROUT)\#$(PRJ)$/$(ROUT)\#g` \
-	$(SHL10STDLIBS) $(SHL10ARCHIVES) $(SHL10STDSHL) $(STDSHL10) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_10.cmd
+	$(SHL10STDLIBS) $(SHL10ARCHIVES) $(SHL10STDSHL) $(STDSHL10) $(SOLARLIB) $(LINKOUTPUT_FILTER) > $(MISC)$/$(TARGET).$(@:b)_10.cmd
 	@cat $(MISC)$/$(TARGET).$(@:b)_10.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_10.cmd
 .IF "$(UPDATER)"=="YES"
diff -u solenv/inc/_tg_app.mk solenv/inc/_tg_app.mk
--- solenv/inc/_tg_app.mk	(working copy)
+++ solenv/inc/_tg_app.mk
@@ -116,10 +116,10 @@
 .ELSE		# "$(OS)"=="MACOSX"
 	@echo unx
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_1.cmd
-	@echo $(APP1LINKER) $(APP1LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(SOLARLIB) $(STDSLO) \
+	@echo $(APP1LINKER) $(APP1LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(STDSLO) \
 	$(APP1OBJS:s/.obj/.o/) '\' >  $(MISC)$/$(TARGET).$(@:b)_1.cmd
 	@cat $(mktmp /dev/null $(APP1LIBS)) | xargs -n 1 cat | sed s\#$(ROUT)\#$(OUT)\#g | sed 's#$$# \\#'  >> $(MISC)$/$(TARGET).$(@:b)_1.cmd
-	@echo $(APP1LINKTYPEFLAG) $(APP1LIBSALCPPRT) $(APP1STDLIBS) $(APP1STDLIB) $(STDLIB1) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_1.cmd
+	@echo $(APP1LINKTYPEFLAG) $(APP1LIBSALCPPRT) $(APP1STDLIBS) $(APP1STDLIB) $(STDLIB1) $(SOLARLIB) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_1.cmd
 	cat $(MISC)$/$(TARGET).$(@:b)_1.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_1.cmd
 	@ls -l $@
@@ -400,10 +400,10 @@
 .ELSE		# "$(OS)"=="MACOSX"
 	@echo unx
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_2.cmd
-	@echo $(APP2LINKER) $(APP2LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(SOLARLIB) $(STDSLO) \
+	@echo $(APP2LINKER) $(APP2LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(STDSLO) \
 	$(APP2OBJS:s/.obj/.o/) '\' >  $(MISC)$/$(TARGET).$(@:b)_2.cmd
 	@cat $(mktmp /dev/null $(APP2LIBS)) | xargs -n 1 cat | sed s\#$(ROUT)\#$(OUT)\#g | sed 's#$$# \\#'  >> $(MISC)$/$(TARGET).$(@:b)_2.cmd
-	@echo $(APP2LINKTYPEFLAG) $(APP2LIBSALCPPRT) $(APP2STDLIBS) $(APP2STDLIB) $(STDLIB2) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_2.cmd
+	@echo $(APP2LINKTYPEFLAG) $(APP2LIBSALCPPRT) $(APP2STDLIBS) $(APP2STDLIB) $(STDLIB2) $(SOLARLIB) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_2.cmd
 	cat $(MISC)$/$(TARGET).$(@:b)_2.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_2.cmd
 	@ls -l $@
@@ -684,10 +684,10 @@
 .ELSE		# "$(OS)"=="MACOSX"
 	@echo unx
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_3.cmd
-	@echo $(APP3LINKER) $(APP3LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(SOLARLIB) $(STDSLO) \
+	@echo $(APP3LINKER) $(APP3LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(STDSLO) \
 	$(APP3OBJS:s/.obj/.o/) '\' >  $(MISC)$/$(TARGET).$(@:b)_3.cmd
 	@cat $(mktmp /dev/null $(APP3LIBS)) | xargs -n 1 cat | sed s\#$(ROUT)\#$(OUT)\#g | sed 's#$$# \\#'  >> $(MISC)$/$(TARGET).$(@:b)_3.cmd
-	@echo $(APP3LINKTYPEFLAG) $(APP3LIBSALCPPRT) $(APP3STDLIBS) $(APP3STDLIB) $(STDLIB3) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_3.cmd
+	@echo $(APP3LINKTYPEFLAG) $(APP3LIBSALCPPRT) $(APP3STDLIBS) $(APP3STDLIB) $(STDLIB3) $(SOLARLIB) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_3.cmd
 	cat $(MISC)$/$(TARGET).$(@:b)_3.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_3.cmd
 	@ls -l $@
@@ -968,10 +968,10 @@
 .ELSE		# "$(OS)"=="MACOSX"
 	@echo unx
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_4.cmd
-	@echo $(APP4LINKER) $(APP4LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(SOLARLIB) $(STDSLO) \
+	@echo $(APP4LINKER) $(APP4LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(STDSLO) \
 	$(APP4OBJS:s/.obj/.o/) '\' >  $(MISC)$/$(TARGET).$(@:b)_4.cmd
 	@cat $(mktmp /dev/null $(APP4LIBS)) | xargs -n 1 cat | sed s\#$(ROUT)\#$(OUT)\#g | sed 's#$$# \\#'  >> $(MISC)$/$(TARGET).$(@:b)_4.cmd
-	@echo $(APP4LINKTYPEFLAG) $(APP4LIBSALCPPRT) $(APP4STDLIBS) $(APP4STDLIB) $(STDLIB4) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_4.cmd
+	@echo $(APP4LINKTYPEFLAG) $(APP4LIBSALCPPRT) $(APP4STDLIBS) $(APP4STDLIB) $(STDLIB4) $(SOLARLIB) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_4.cmd
 	cat $(MISC)$/$(TARGET).$(@:b)_4.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_4.cmd
 	@ls -l $@
@@ -1252,10 +1252,10 @@
 .ELSE		# "$(OS)"=="MACOSX"
 	@echo unx
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_5.cmd
-	@echo $(APP5LINKER) $(APP5LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(SOLARLIB) $(STDSLO) \
+	@echo $(APP5LINKER) $(APP5LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(STDSLO) \
 	$(APP5OBJS:s/.obj/.o/) '\' >  $(MISC)$/$(TARGET).$(@:b)_5.cmd
 	@cat $(mktmp /dev/null $(APP5LIBS)) | xargs -n 1 cat | sed s\#$(ROUT)\#$(OUT)\#g | sed 's#$$# \\#'  >> $(MISC)$/$(TARGET).$(@:b)_5.cmd
-	@echo $(APP5LINKTYPEFLAG) $(APP5LIBSALCPPRT) $(APP5STDLIBS) $(APP5STDLIB) $(STDLIB5) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_5.cmd
+	@echo $(APP5LINKTYPEFLAG) $(APP5LIBSALCPPRT) $(APP5STDLIBS) $(APP5STDLIB) $(STDLIB5) $(SOLARLIB) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_5.cmd
 	cat $(MISC)$/$(TARGET).$(@:b)_5.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_5.cmd
 	@ls -l $@
@@ -1536,10 +1536,10 @@
 .ELSE		# "$(OS)"=="MACOSX"
 	@echo unx
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_6.cmd
-	@echo $(APP6LINKER) $(APP6LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(SOLARLIB) $(STDSLO) \
+	@echo $(APP6LINKER) $(APP6LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(STDSLO) \
 	$(APP6OBJS:s/.obj/.o/) '\' >  $(MISC)$/$(TARGET).$(@:b)_6.cmd
 	@cat $(mktmp /dev/null $(APP6LIBS)) | xargs -n 1 cat | sed s\#$(ROUT)\#$(OUT)\#g | sed 's#$$# \\#'  >> $(MISC)$/$(TARGET).$(@:b)_6.cmd
-	@echo $(APP6LINKTYPEFLAG) $(APP6LIBSALCPPRT) $(APP6STDLIBS) $(APP6STDLIB) $(STDLIB6) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_6.cmd
+	@echo $(APP6LINKTYPEFLAG) $(APP6LIBSALCPPRT) $(APP6STDLIBS) $(APP6STDLIB) $(STDLIB6) $(SOLARLIB) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_6.cmd
 	cat $(MISC)$/$(TARGET).$(@:b)_6.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_6.cmd
 	@ls -l $@
@@ -1820,10 +1820,10 @@
 .ELSE		# "$(OS)"=="MACOSX"
 	@echo unx
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_7.cmd
-	@echo $(APP7LINKER) $(APP7LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(SOLARLIB) $(STDSLO) \
+	@echo $(APP7LINKER) $(APP7LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(STDSLO) \
 	$(APP7OBJS:s/.obj/.o/) '\' >  $(MISC)$/$(TARGET).$(@:b)_7.cmd
 	@cat $(mktmp /dev/null $(APP7LIBS)) | xargs -n 1 cat | sed s\#$(ROUT)\#$(OUT)\#g | sed 's#$$# \\#'  >> $(MISC)$/$(TARGET).$(@:b)_7.cmd
-	@echo $(APP7LINKTYPEFLAG) $(APP7LIBSALCPPRT) $(APP7STDLIBS) $(APP7STDLIB) $(STDLIB7) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_7.cmd
+	@echo $(APP7LINKTYPEFLAG) $(APP7LIBSALCPPRT) $(APP7STDLIBS) $(APP7STDLIB) $(STDLIB7) $(SOLARLIB) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_7.cmd
 	cat $(MISC)$/$(TARGET).$(@:b)_7.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_7.cmd
 	@ls -l $@
@@ -2104,10 +2104,10 @@
 .ELSE		# "$(OS)"=="MACOSX"
 	@echo unx
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_8.cmd
-	@echo $(APP8LINKER) $(APP8LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(SOLARLIB) $(STDSLO) \
+	@echo $(APP8LINKER) $(APP8LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(STDSLO) \
 	$(APP8OBJS:s/.obj/.o/) '\' >  $(MISC)$/$(TARGET).$(@:b)_8.cmd
 	@cat $(mktmp /dev/null $(APP8LIBS)) | xargs -n 1 cat | sed s\#$(ROUT)\#$(OUT)\#g | sed 's#$$# \\#'  >> $(MISC)$/$(TARGET).$(@:b)_8.cmd
-	@echo $(APP8LINKTYPEFLAG) $(APP8LIBSALCPPRT) $(APP8STDLIBS) $(APP8STDLIB) $(STDLIB8) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_8.cmd
+	@echo $(APP8LINKTYPEFLAG) $(APP8LIBSALCPPRT) $(APP8STDLIBS) $(APP8STDLIB) $(STDLIB8) $(SOLARLIB) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_8.cmd
 	cat $(MISC)$/$(TARGET).$(@:b)_8.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_8.cmd
 	@ls -l $@
@@ -2388,10 +2388,10 @@
 .ELSE		# "$(OS)"=="MACOSX"
 	@echo unx
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_9.cmd
-	@echo $(APP9LINKER) $(APP9LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(SOLARLIB) $(STDSLO) \
+	@echo $(APP9LINKER) $(APP9LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(STDSLO) \
 	$(APP9OBJS:s/.obj/.o/) '\' >  $(MISC)$/$(TARGET).$(@:b)_9.cmd
 	@cat $(mktmp /dev/null $(APP9LIBS)) | xargs -n 1 cat | sed s\#$(ROUT)\#$(OUT)\#g | sed 's#$$# \\#'  >> $(MISC)$/$(TARGET).$(@:b)_9.cmd
-	@echo $(APP9LINKTYPEFLAG) $(APP9LIBSALCPPRT) $(APP9STDLIBS) $(APP9STDLIB) $(STDLIB9) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_9.cmd
+	@echo $(APP9LINKTYPEFLAG) $(APP9LIBSALCPPRT) $(APP9STDLIBS) $(APP9STDLIB) $(STDLIB9) $(SOLARLIB) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_9.cmd
 	cat $(MISC)$/$(TARGET).$(@:b)_9.cmd
 	@+source $(MISC)$/$(TARGET).$(@:b)_9.cmd
 	@ls -l $@
@@ -2672,7 +2672,7 @@
 .ELSE		# "$(OS)"=="MACOSX"
 	@echo unx
 	@-$(RM) $(MISC)$/$(TARGET).$(@:b)_10.cmd
-	@echo $(APP10LINKER) $(APP10LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(SOLARLIB) $(STDSLO) \
+	@echo $(APP10LINKER) $(APP10LINKFLAGS) $(LINKFLAGSAPP) -L$(PRJ)$/$(INPATH)$/lib $(STDSLO) \
 	$(APP10OBJS:s/.obj/.o/) '\' >  $(MISC)$/$(TARGET).$(@:b)_10.cmd
 	@cat $(mktmp /dev/null $(APP10LIBS)) | xargs -n 1 cat | sed s\#$(ROUT)\#$(OUT)\#g | sed 's#$$# \\#'  >> $(MISC)$/$(TARGET).$(@:b)_10.cmd
 	@echo $(APP10LINKTYPEFLAG) $(APP10LIBSALCPPRT) $(APP10STDLIBS) $(APP10STDLIB) $(STDLIB10) -o $@ >> $(MISC)$/$(TARGET).$(@:b)_10.cmd
