--- a/themes/default/theme.lua
+++ b/themes/default/theme.lua
@@ -77,7 +77,7 @@ theme.titlebar_maximized_button_focus_inactive  = "@AWESOME_THEMES_PATH@/default
 theme.titlebar_maximized_button_normal_active = "@AWESOME_THEMES_PATH@/default/titlebar/maximized_normal_active.png"
 theme.titlebar_maximized_button_focus_active  = "@AWESOME_THEMES_PATH@/default/titlebar/maximized_focus_active.png"
 
-theme.wallpaper = "@AWESOME_THEMES_PATH@/default/background.png"
+theme.wallpaper = "@CMAKE_INSTALL_PREFIX@/share/backgrounds/sabayonlinux.png"
 
 -- You can use your own layout icons like this:
 theme.layout_fairh = "@AWESOME_THEMES_PATH@/default/layouts/fairhw.png"
