--- a/gnome-mount.schemas.in
+++ b/gnome-mount.schemas.in
@@ -64,7 +64,7 @@
         <applyto>/system/storage/default_options/ntfs-3g/mount_options</applyto>
         <type>list</type>
         <list_type>string</list_type>
-        <default>[locale=]</default>
+        <default>[]</default>
         <locale name="C">
           <short>Default mount options for ntfs-3g fs</short>
           <long>A list of default mount options for volumes formatted with the ntfs file system using ntfs-3g.</long>
