From suse

--- ./knetworkmanager/src/knetworkmanager.desktop	2007/03/09 10:14:59	1.1
+++ ./knetworkmanager/src/knetworkmanager.desktop	2007/03/09 10:15:08
@@ -6,7 +6,7 @@
 GenericName[ja]=ネットワークマネージャ
 GenericName[pt]=Gestor de Redes
 Exec=knetworkmanager
-Icon=network
+Icon=knetworkmanager
 Type=Application
 Comment=A KDE frontend for NetworkManager
 Comment[da]=En KDE-grænseflade for Networkmanager
