http://bugs.gentoo.org/334077

--- usr/iscsid.c
+++ usr/iscsid.c
@@ -29,6 +29,7 @@
 #include <signal.h>
 #include <sys/mman.h>
 #include <sys/utsname.h>
+#include <sys/stat.h>
 
 #include "iscsid.h"
 #include "mgmt_ipc.h"
--- usr/iscsi_sysfs.c
+++ usr/iscsi_sysfs.c
@@ -22,6 +22,7 @@
 #include <string.h>
 #include <errno.h>
 #include <dirent.h>
+#include <sys/stat.h>
 
 #include "log.h"
 #include "initiator.h"
