--- configure.orig	2009-07-01 22:54:39.830348980 +0200
+++ configure	2009-07-01 23:01:11.346350261 +0200
@@ -1206,9 +1206,7 @@
 }
 EOF
   [ -n "$INCX11" ] && I_INCX11=`prefix_I "$INCX11"`
-  if test_compile "$LIBX11 $I_INCX11" Xlibs Xlibs; then
-    log_success "found"
-  fi
+  test_compile "$LIBX11 $I_INCX11" Xlibs Xlibs
 }
 
 
@@ -1288,19 +1286,6 @@
 #include <GL/glu.h>
 extern "C" int main(void)
 {
-  Display *dpy;
-  int major, minor;
-
-  dpy = XOpenDisplay(NULL);
-  if (dpy)
-  {
-    if (glXQueryVersion(dpy, &major, &minor))
-    {
-      printf("found version %u.%u, OK.\n", major, minor);
-      return 0;
-    }
-  }
-  printf("found (inactive), OK.\n");
   return 0;
 }
 EOF
