diff -ur dooble.d.orig/trunk/browser/Icons/everaldo/configuration.cfg dooble.d/trunk/browser/Icons/everaldo/configuration.cfg
--- dooble.d.orig/trunk/browser/Icons/everaldo/configuration.cfg	2011-05-15 22:59:47.000000000 +0200
+++ dooble.d/trunk/browser/Icons/everaldo/configuration.cfg	2011-09-16 13:53:36.998836258 +0200
@@ -4,110 +4,110 @@
 License = GPL
 
 [General]
-windowIcon = Icons/16x16/dooble.png
-okButtonIcon = Icons/everaldo/16x16/ok.png
-cancelButtonIcon = Icons/everaldo/16x16/cancel.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+okButtonIcon = /usr/share/dooble/Icons/everaldo/16x16/ok.png
+cancelButtonIcon = /usr/share/dooble/Icons/everaldo/16x16/cancel.png
 
 [downloadWindow]
-windowIcon = Icons/16x16/dooble.png
-downloadButtonIcon = Icons/everaldo/16x16/download.png
-clearListButtonIcon = Icons/everaldo/16x16/remove.png
-enterUrlButtonIcon = Icons/everaldo/16x16/ok.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+downloadButtonIcon = /usr/share/dooble/Icons/everaldo/16x16/download.png
+clearListButtonIcon = /usr/share/dooble/Icons/everaldo/16x16/remove.png
+enterUrlButtonIcon = /usr/share/dooble/Icons/everaldo/16x16/ok.png
 
 [downloadWindowItem]
-abortIcon = Icons/everaldo/16x16/cancel.png
-pauseIcon = Icons/everaldo/16x16/pause.png
-resumeIcon = Icons/everaldo/16x16/play.png
-retryIcon = Icons/everaldo/16x16/reload.png
+abortIcon = /usr/share/dooble/Icons/everaldo/16x16/cancel.png
+pauseIcon = /usr/share/dooble/Icons/everaldo/16x16/pause.png
+resumeIcon = /usr/share/dooble/Icons/everaldo/16x16/play.png
+retryIcon = /usr/share/dooble/Icons/everaldo/16x16/reload.png
 
 [mainWindow]
-backToolButton = Icons/everaldo/32x32/back.png
-forwardToolButton = Icons/everaldo/32x32/forward.png
-reloadToolButton = Icons/everaldo/32x32/reload.png
-stopToolButton = Icons/everaldo/32x32/stop.png
-homeToolButton = Icons/everaldo/32x32/home.png
-hideFindToolButton = Icons/everaldo/16x16/cancel.png
-actionNew_Tab = Icons/everaldo/16x16/add_tab.png
-actionNew_Window = Icons/everaldo/16x16/newwindow.png
-actionOpen_URL = Icons/everaldo/16x16/navigate.png
-actionClose_Tab = Icons/everaldo/16x16/remove_tab.png
-actionClose_Window = Icons/everaldo/16x16/closewindow.png
-actionSave_Page = Icons/everaldo/16x16/downloadmanager.png
-actionPrint = Icons/everaldo/16x16/print.png
-actionPrint_Preview = Icons/everaldo/16x16/printpreview.png
-actionQuit = Icons/everaldo/16x16/quit.png
-actionFind = Icons/everaldo/16x16/search.png
-actionReload = Icons/everaldo/16x16/reload.png
-actionStop = Icons/everaldo/16x16/stop.png
-actionDesktop = Icons/everaldo/16x16/desktop.png
-actionPage_Source = Icons/everaldo/16x16/pagesource.png
-actionDownloads = Icons/everaldo/16x16/downloadmanager.png
-actionSettings = Icons/everaldo/16x16/settings.png
-actionAbout_Dooble = Icons/everaldo/16x16/about.png
-desktopToolButton = Icons/everaldo/32x32/desktop.png
-tabWidget =  Icons/16x16/dooble.png
-windowIcon = Icons/16x16/dooble.png
-historyMenu = Icons/everaldo/16x16/clearhistory.png
-emptyIcon = Icons/16x16/dooble.png
-actionOpen_Directory = Icons/everaldo/16x16/folder_yellow.png
-actionMy_Retrieved_Files = Icons/everaldo/16x16/myretrievedfiles.png
-addTab = Icons/everaldo/16x16/add_tab.png
-removeTab = Icons/everaldo/16x16/remove_tab.png
-restoreToolButton = Icons/everaldo/16x16/restorenormalscreen.png
-actionCopy = Icons/everaldo/16x16/copy.png
-actionPaste = Icons/everaldo/16x16/paste.png
-actionSelect_All_Content = Icons/everaldo/16x16/selectallcontent.png
-zoomMenu = Icons/everaldo/16x16/zoom.png
-actionFull_Screen_Mode = Icons/everaldo/16x16/fullscreenmode.png
-viewHistory = Icons/everaldo/16x16/history.png
-nextToolButton = Icons/everaldo/16x16/next.png
-previousToolButton = Icons/everaldo/16x16/previous.png
-actionApplication_Cookies = Icons/everaldo/16x16/cookies.png
-actionP2P_Email = Icons/everaldo/16x16/email.png
+backToolButton = /usr/share/dooble/Icons/everaldo/32x32/back.png
+forwardToolButton = /usr/share/dooble/Icons/everaldo/32x32/forward.png
+reloadToolButton = /usr/share/dooble/Icons/everaldo/32x32/reload.png
+stopToolButton = /usr/share/dooble/Icons/everaldo/32x32/stop.png
+homeToolButton = /usr/share/dooble/Icons/everaldo/32x32/home.png
+hideFindToolButton = /usr/share/dooble/Icons/everaldo/16x16/cancel.png
+actionNew_Tab = /usr/share/dooble/Icons/everaldo/16x16/add_tab.png
+actionNew_Window = /usr/share/dooble/Icons/everaldo/16x16/newwindow.png
+actionOpen_URL = /usr/share/dooble/Icons/everaldo/16x16/navigate.png
+actionClose_Tab = /usr/share/dooble/Icons/everaldo/16x16/remove_tab.png
+actionClose_Window = /usr/share/dooble/Icons/everaldo/16x16/closewindow.png
+actionSave_Page = /usr/share/dooble/Icons/everaldo/16x16/downloadmanager.png
+actionPrint = /usr/share/dooble/Icons/everaldo/16x16/print.png
+actionPrint_Preview = /usr/share/dooble/Icons/everaldo/16x16/printpreview.png
+actionQuit = /usr/share/dooble/Icons/everaldo/16x16/quit.png
+actionFind = /usr/share/dooble/Icons/everaldo/16x16/search.png
+actionReload = /usr/share/dooble/Icons/everaldo/16x16/reload.png
+actionStop = /usr/share/dooble/Icons/everaldo/16x16/stop.png
+actionDesktop = /usr/share/dooble/Icons/everaldo/16x16/desktop.png
+actionPage_Source = /usr/share/dooble/Icons/everaldo/16x16/pagesource.png
+actionDownloads = /usr/share/dooble/Icons/everaldo/16x16/downloadmanager.png
+actionSettings = /usr/share/dooble/Icons/everaldo/16x16/settings.png
+actionAbout_Dooble = /usr/share/dooble/Icons/everaldo/16x16/about.png
+desktopToolButton = /usr/share/dooble/Icons/everaldo/32x32/desktop.png
+tabWidget =  /usr/share/dooble/Icons/16x16/dooble.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+historyMenu = /usr/share/dooble/Icons/everaldo/16x16/clearhistory.png
+emptyIcon = /usr/share/dooble/Icons/16x16/dooble.png
+actionOpen_Directory = /usr/share/dooble/Icons/everaldo/16x16/folder_yellow.png
+actionMy_Retrieved_Files = /usr/share/dooble/Icons/everaldo/16x16/myretrievedfiles.png
+addTab = /usr/share/dooble/Icons/everaldo/16x16/add_tab.png
+removeTab = /usr/share/dooble/Icons/everaldo/16x16/remove_tab.png
+restoreToolButton = /usr/share/dooble/Icons/everaldo/16x16/restorenormalscreen.png
+actionCopy = /usr/share/dooble/Icons/everaldo/16x16/copy.png
+actionPaste = /usr/share/dooble/Icons/everaldo/16x16/paste.png
+actionSelect_All_Content = /usr/share/dooble/Icons/everaldo/16x16/selectallcontent.png
+zoomMenu = /usr/share/dooble/Icons/everaldo/16x16/zoom.png
+actionFull_Screen_Mode = /usr/share/dooble/Icons/everaldo/16x16/fullscreenmode.png
+viewHistory = /usr/share/dooble/Icons/everaldo/16x16/history.png
+nextToolButton = /usr/share/dooble/Icons/everaldo/16x16/next.png
+previousToolButton = /usr/share/dooble/Icons/everaldo/16x16/previous.png
+actionApplication_Cookies = /usr/share/dooble/Icons/everaldo/16x16/cookies.png
+actionP2P_Email = /usr/share/dooble/Icons/everaldo/16x16/email.png
 
 [pageSourceWindow]
-actionPrint = Icons/everaldo/16x16/print.png
-actionPrint_Preview = Icons/everaldo/16x16/printpreview.png
-actionClose = Icons/everaldo/16x16/closewindow.png
-actionFind = Icons/everaldo/16x16/search.png
-actionSave_As = Icons/everaldo/16x16/filesaveas.png
-hideFindToolButton = Icons/everaldo/16x16/cancel.png
-windowIcon = Icons/16x16/dooble.png
-nextToolButton = Icons/everaldo/16x16/next.png
-previousToolButton = Icons/everaldo/16x16/previous.png
+actionPrint = /usr/share/dooble/Icons/everaldo/16x16/print.png
+actionPrint_Preview = /usr/share/dooble/Icons/everaldo/16x16/printpreview.png
+actionClose = /usr/share/dooble/Icons/everaldo/16x16/closewindow.png
+actionFind = /usr/share/dooble/Icons/everaldo/16x16/search.png
+actionSave_As = /usr/share/dooble/Icons/everaldo/16x16/filesaveas.png
+hideFindToolButton = /usr/share/dooble/Icons/everaldo/16x16/cancel.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+nextToolButton = /usr/share/dooble/Icons/everaldo/16x16/next.png
+previousToolButton = /usr/share/dooble/Icons/everaldo/16x16/previous.png
 
 [propertiesDialog]
-windowIcon = Icons/16x16/dooble.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
 
 [settingsWindow]
-windowIcon = Icons/16x16/dooble.png
-displayButtonIcon = Icons/everaldo/48x48/display.png
-historyButtonIcon = Icons/everaldo/48x48/history.png
-homeButtonIcon = Icons/everaldo/48x48/home.png
-proxyButtonIcon = Icons/everaldo/48x48/proxy.png
-securityButtonIcon = Icons/everaldo/48x48/security.png
-tabsButtonIcon = Icons/everaldo/48x48/tabs.png
-webkitButtonIcon = Icons/everaldo/48x48/webkit.png
-windowsButtonIcon = Icons/everaldo/48x48/windows.png
-copyAndPasteToolButtonIcon = Icons/everaldo/16x16/paste.png
-safeButtonIcon = Icons/everaldo/48x48/zerocrypt.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+displayButtonIcon = /usr/share/dooble/Icons/everaldo/48x48/display.png
+historyButtonIcon = /usr/share/dooble/Icons/everaldo/48x48/history.png
+homeButtonIcon = /usr/share/dooble/Icons/everaldo/48x48/home.png
+proxyButtonIcon = /usr/share/dooble/Icons/everaldo/48x48/proxy.png
+securityButtonIcon = /usr/share/dooble/Icons/everaldo/48x48/security.png
+tabsButtonIcon = /usr/share/dooble/Icons/everaldo/48x48/tabs.png
+webkitButtonIcon = /usr/share/dooble/Icons/everaldo/48x48/webkit.png
+windowsButtonIcon = /usr/share/dooble/Icons/everaldo/48x48/windows.png
+copyAndPasteToolButtonIcon = /usr/share/dooble/Icons/everaldo/16x16/paste.png
+safeButtonIcon = /usr/share/dooble/Icons/everaldo/48x48/zerocrypt.png
 
 [urlWidget]
-goToolButton = Icons/everaldo/16x16/navigate.png
-indexToolButton = Icons/16x16/dooble.png
-pulldownToolButton = Icons/everaldo/16x16/empty.png
-emptyIcon = Icons/everaldo/16x16/empty.png
-bookmarkToolButton = Icons/everaldo/16x16/bookmark.png
+goToolButton = /usr/share/dooble/Icons/everaldo/16x16/navigate.png
+indexToolButton = /usr/share/dooble/Icons/16x16/dooble.png
+pulldownToolButton = /usr/share/dooble/Icons/everaldo/16x16/empty.png
+emptyIcon = /usr/share/dooble/Icons/everaldo/16x16/empty.png
+bookmarkToolButton = /usr/share/dooble/Icons/everaldo/16x16/bookmark.png
 
 [searchWidget]
-findToolButton = Icons/everaldo/16x16/search.png
+findToolButton = /usr/share/dooble/Icons/everaldo/16x16/search.png
 
 [historyWindow]
-windowIcon = Icons/16x16/dooble.png
-deleteButtonIcon = Icons/everaldo/16x16/remove.png
-deleteAllButtonIcon = Icons/everaldo/16x16/removeall.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+deleteButtonIcon = /usr/share/dooble/Icons/everaldo/16x16/remove.png
+deleteAllButtonIcon = /usr/share/dooble/Icons/everaldo/16x16/removeall.png
 
 [cookiesWindow]
-windowIcon = Icons/16x16/dooble.png
-deleteButtonIcon = Icons/everaldo/16x16/remove.png
-deleteAllButtonIcon = Icons/everaldo/16x16/removeall.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+deleteButtonIcon = /usr/share/dooble/Icons/everaldo/16x16/remove.png
+deleteAllButtonIcon = /usr/share/dooble/Icons/everaldo/16x16/removeall.png
diff -ur dooble.d.orig/trunk/browser/Icons/meego/configuration.cfg dooble.d/trunk/browser/Icons/meego/configuration.cfg
--- dooble.d.orig/trunk/browser/Icons/meego/configuration.cfg	2011-05-15 23:02:11.000000000 +0200
+++ dooble.d/trunk/browser/Icons/meego/configuration.cfg	2011-09-16 13:53:37.016838545 +0200
@@ -4,110 +4,110 @@
 License = GPL
 
 [General]
-windowIcon = Icons/16x16/dooble.png
-okButtonIcon = Icons/meego/16x16/ok.png
-cancelButtonIcon = Icons/meego/16x16/cancel.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+okButtonIcon = /usr/share/dooble/Icons/meego/16x16/ok.png
+cancelButtonIcon = /usr/share/dooble/Icons/meego/16x16/cancel.png
 
 [downloadWindow]
-windowIcon = Icons/16x16/dooble.png
-downloadButtonIcon = Icons/meego/16x16/download.png
-clearListButtonIcon = Icons/meego/16x16/clearlist.png
-enterUrlButtonIcon = Icons/meego/16x16/ok.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+downloadButtonIcon = /usr/share/dooble/Icons/meego/16x16/download.png
+clearListButtonIcon = /usr/share/dooble/Icons/meego/16x16/clearlist.png
+enterUrlButtonIcon = /usr/share/dooble/Icons/meego/16x16/ok.png
 
 [downloadWindowItem]
-abortIcon = Icons/meego/16x16/cancel.png
-pauseIcon = Icons/meego/16x16/pause.png
-resumeIcon = Icons/meego/16x16/play.png
-retryIcon = Icons/meego/16x16/reload.png
+abortIcon = /usr/share/dooble/Icons/meego/16x16/cancel.png
+pauseIcon = /usr/share/dooble/Icons/meego/16x16/pause.png
+resumeIcon = /usr/share/dooble/Icons/meego/16x16/play.png
+retryIcon = /usr/share/dooble/Icons/meego/16x16/reload.png
 
 [mainWindow]
-backToolButton = Icons/meego/32x32/back.png
-forwardToolButton = Icons/meego/32x32/forward.png
-reloadToolButton = Icons/meego/32x32/reload.png
-stopToolButton = Icons/meego/32x32/stop.png
-homeToolButton = Icons/meego/32x32/home.png
-hideFindToolButton = Icons/meego/16x16/cancel.png
-actionNew_Tab = Icons/meego/16x16/add_tab.png
-actionNew_Window = Icons/meego/16x16/newwindow.png
-actionOpen_URL = Icons/meego/16x16/navigate.png
-actionClose_Tab = Icons/meego/16x16/remove_tab.png
-actionClose_Window = Icons/meego/16x16/closewindow.png
-actionSave_Page = Icons/meego/16x16/downloadmanager.png
-actionPrint = Icons/meego/16x16/print.png
-actionPrint_Preview = Icons/meego/16x16/printpreview.png
-actionQuit = Icons/meego/16x16/quit.png
-actionFind = Icons/meego/16x16/search.png
-actionReload = Icons/meego/16x16/reload.png
-actionStop = Icons/meego/16x16/stop.png
-actionDesktop = Icons/meego/16x16/desktop.png
-actionPage_Source = Icons/meego/16x16/pagesource.png
-actionDownloads = Icons/meego/16x16/downloadmanager.png
-actionSettings = Icons/meego/16x16/settings.png
-actionAbout_Dooble = Icons/meego/16x16/about.png
-desktopToolButton = Icons/meego/32x32/desktop.png
-tabWidget =  Icons/16x16/dooble.png
-windowIcon = Icons/16x16/dooble.png
-historyMenu = Icons/meego/16x16/clearhistory.png
-emptyIcon = Icons/16x16/dooble.png
-actionOpen_Directory = Icons/meego/16x16/folder_yellow.png
-actionMy_Retrieved_Files = Icons/meego/16x16/myretrievedfiles.png
-addTab = Icons/meego/16x16/add_tab.png
-removeTab = Icons/meego/16x16/remove_tab.png
-restoreToolButton = Icons/meego/16x16/restorenormalscreen.png
-actionCopy = Icons/meego/16x16/copy.png
-actionPaste = Icons/meego/16x16/paste.png
-actionSelect_All_Content = Icons/meego/16x16/selectallcontent.png
-zoomMenu = Icons/meego/16x16/zoom.png
-actionFull_Screen_Mode = Icons/meego/16x16/fullscreenmode.png
-viewHistory = Icons/meego/16x16/history.png
-nextToolButton = Icons/meego/16x16/next.png
-previousToolButton = Icons/meego/16x16/previous.png
-actionApplication_Cookies = Icons/meego/16x16/cookies.png
-actionP2P_Email = Icons/meego/16x16/email.png
+backToolButton = /usr/share/dooble/Icons/meego/32x32/back.png
+forwardToolButton = /usr/share/dooble/Icons/meego/32x32/forward.png
+reloadToolButton = /usr/share/dooble/Icons/meego/32x32/reload.png
+stopToolButton = /usr/share/dooble/Icons/meego/32x32/stop.png
+homeToolButton = /usr/share/dooble/Icons/meego/32x32/home.png
+hideFindToolButton = /usr/share/dooble/Icons/meego/16x16/cancel.png
+actionNew_Tab = /usr/share/dooble/Icons/meego/16x16/add_tab.png
+actionNew_Window = /usr/share/dooble/Icons/meego/16x16/newwindow.png
+actionOpen_URL = /usr/share/dooble/Icons/meego/16x16/navigate.png
+actionClose_Tab = /usr/share/dooble/Icons/meego/16x16/remove_tab.png
+actionClose_Window = /usr/share/dooble/Icons/meego/16x16/closewindow.png
+actionSave_Page = /usr/share/dooble/Icons/meego/16x16/downloadmanager.png
+actionPrint = /usr/share/dooble/Icons/meego/16x16/print.png
+actionPrint_Preview = /usr/share/dooble/Icons/meego/16x16/printpreview.png
+actionQuit = /usr/share/dooble/Icons/meego/16x16/quit.png
+actionFind = /usr/share/dooble/Icons/meego/16x16/search.png
+actionReload = /usr/share/dooble/Icons/meego/16x16/reload.png
+actionStop = /usr/share/dooble/Icons/meego/16x16/stop.png
+actionDesktop = /usr/share/dooble/Icons/meego/16x16/desktop.png
+actionPage_Source = /usr/share/dooble/Icons/meego/16x16/pagesource.png
+actionDownloads = /usr/share/dooble/Icons/meego/16x16/downloadmanager.png
+actionSettings = /usr/share/dooble/Icons/meego/16x16/settings.png
+actionAbout_Dooble = /usr/share/dooble/Icons/meego/16x16/about.png
+desktopToolButton = /usr/share/dooble/Icons/meego/32x32/desktop.png
+tabWidget =  /usr/share/dooble/Icons/16x16/dooble.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+historyMenu = /usr/share/dooble/Icons/meego/16x16/clearhistory.png
+emptyIcon = /usr/share/dooble/Icons/16x16/dooble.png
+actionOpen_Directory = /usr/share/dooble/Icons/meego/16x16/folder_yellow.png
+actionMy_Retrieved_Files = /usr/share/dooble/Icons/meego/16x16/myretrievedfiles.png
+addTab = /usr/share/dooble/Icons/meego/16x16/add_tab.png
+removeTab = /usr/share/dooble/Icons/meego/16x16/remove_tab.png
+restoreToolButton = /usr/share/dooble/Icons/meego/16x16/restorenormalscreen.png
+actionCopy = /usr/share/dooble/Icons/meego/16x16/copy.png
+actionPaste = /usr/share/dooble/Icons/meego/16x16/paste.png
+actionSelect_All_Content = /usr/share/dooble/Icons/meego/16x16/selectallcontent.png
+zoomMenu = /usr/share/dooble/Icons/meego/16x16/zoom.png
+actionFull_Screen_Mode = /usr/share/dooble/Icons/meego/16x16/fullscreenmode.png
+viewHistory = /usr/share/dooble/Icons/meego/16x16/history.png
+nextToolButton = /usr/share/dooble/Icons/meego/16x16/next.png
+previousToolButton = /usr/share/dooble/Icons/meego/16x16/previous.png
+actionApplication_Cookies = /usr/share/dooble/Icons/meego/16x16/cookies.png
+actionP2P_Email = /usr/share/dooble/Icons/meego/16x16/email.png
 
 [pageSourceWindow]
-actionPrint = Icons/meego/16x16/print.png
-actionPrint_Preview = Icons/meego/16x16/printpreview.png
-actionClose = Icons/meego/16x16/closewindow.png
-actionFind = Icons/meego/16x16/search.png
-actionSave_As = Icons/meego/16x16/filesaveas.png
-hideFindToolButton = Icons/meego/16x16/cancel.png
-windowIcon = Icons/16x16/dooble.png
-nextToolButton = Icons/meego/16x16/next.png
-previousToolButton = Icons/meego/16x16/previous.png
+actionPrint = /usr/share/dooble/Icons/meego/16x16/print.png
+actionPrint_Preview = /usr/share/dooble/Icons/meego/16x16/printpreview.png
+actionClose = /usr/share/dooble/Icons/meego/16x16/closewindow.png
+actionFind = /usr/share/dooble/Icons/meego/16x16/search.png
+actionSave_As = /usr/share/dooble/Icons/meego/16x16/filesaveas.png
+hideFindToolButton = /usr/share/dooble/Icons/meego/16x16/cancel.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+nextToolButton = /usr/share/dooble/Icons/meego/16x16/next.png
+previousToolButton = /usr/share/dooble/Icons/meego/16x16/previous.png
 
 [propertiesDialog]
-windowIcon = Icons/16x16/dooble.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
 
 [settingsWindow]
-windowIcon = Icons/16x16/dooble.png
-displayButtonIcon = Icons/meego/48x48/display.png
-historyButtonIcon = Icons/meego/48x48/history.png
-homeButtonIcon = Icons/meego/48x48/home.png
-proxyButtonIcon = Icons/meego/48x48/proxy.png
-securityButtonIcon = Icons/meego/48x48/security.png
-tabsButtonIcon = Icons/meego/48x48/tabs.png
-webkitButtonIcon = Icons/meego/48x48/webkit.png
-windowsButtonIcon = Icons/meego/48x48/windows.png
-copyAndPasteToolButtonIcon = Icons/meego/16x16/paste.png
-safeButtonIcon = Icons/meego/48x48/zerocrypt.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+displayButtonIcon = /usr/share/dooble/Icons/meego/48x48/display.png
+historyButtonIcon = /usr/share/dooble/Icons/meego/48x48/history.png
+homeButtonIcon = /usr/share/dooble/Icons/meego/48x48/home.png
+proxyButtonIcon = /usr/share/dooble/Icons/meego/48x48/proxy.png
+securityButtonIcon = /usr/share/dooble/Icons/meego/48x48/security.png
+tabsButtonIcon = /usr/share/dooble/Icons/meego/48x48/tabs.png
+webkitButtonIcon = /usr/share/dooble/Icons/meego/48x48/webkit.png
+windowsButtonIcon = /usr/share/dooble/Icons/meego/48x48/windows.png
+copyAndPasteToolButtonIcon = /usr/share/dooble/Icons/meego/16x16/paste.png
+safeButtonIcon = /usr/share/dooble/Icons/meego/48x48/zerocrypt.png
 
 [urlWidget]
-goToolButton = Icons/meego/16x16/navigate.png
-indexToolButton = Icons/16x16/dooble.png
-pulldownToolButton = Icons/meego/16x16/empty.png
-emptyIcon = Icons/meego/16x16/empty.png
-bookmarkToolButton = Icons/meego/16x16/bookmark.png
+goToolButton = /usr/share/dooble/Icons/meego/16x16/navigate.png
+indexToolButton = /usr/share/dooble/Icons/16x16/dooble.png
+pulldownToolButton = /usr/share/dooble/Icons/meego/16x16/empty.png
+emptyIcon = /usr/share/dooble/Icons/meego/16x16/empty.png
+bookmarkToolButton = /usr/share/dooble/Icons/meego/16x16/bookmark.png
 
 [searchWidget]
-findToolButton = Icons/meego/16x16/search.png
+findToolButton = /usr/share/dooble/Icons/meego/16x16/search.png
 
 [historyWindow]
-windowIcon = Icons/16x16/dooble.png
-deleteButtonIcon = Icons/meego/16x16/remove.png
-deleteAllButtonIcon = Icons/meego/16x16/removeall.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+deleteButtonIcon = /usr/share/dooble/Icons/meego/16x16/remove.png
+deleteAllButtonIcon = /usr/share/dooble/Icons/meego/16x16/removeall.png
 
 [cookiesWindow]
-windowIcon = Icons/16x16/dooble.png
-deleteButtonIcon = Icons/meego/16x16/remove.png
-deleteAllButtonIcon = Icons/meego/16x16/removeall.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+deleteButtonIcon = /usr/share/dooble/Icons/meego/16x16/remove.png
+deleteAllButtonIcon = /usr/share/dooble/Icons/meego/16x16/removeall.png
diff -ur dooble.d.orig/trunk/browser/Icons/nuovext/configuration.cfg dooble.d/trunk/browser/Icons/nuovext/configuration.cfg
--- dooble.d.orig/trunk/browser/Icons/nuovext/configuration.cfg	2011-05-15 23:01:22.000000000 +0200
+++ dooble.d/trunk/browser/Icons/nuovext/configuration.cfg	2011-09-16 13:53:37.021839180 +0200
@@ -4,108 +4,108 @@
 License = GLGPL (GNU Lesser General Public License)
 
 [General]
-windowIcon = Icons/16x16/dooble.png
-okButtonIcon = Icons/nuovext/16x16/ok.png
-cancelButtonIcon = Icons/nuovext/16x16/cancel.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+okButtonIcon = /usr/share/dooble/Icons/nuovext/16x16/ok.png
+cancelButtonIcon = /usr/share/dooble/Icons/nuovext/16x16/cancel.png
 
 [downloadWindow]
-windowIcon = Icons/16x16/dooble.png
-downloadButtonIcon = Icons/nuovext/16x16/download.png
-clearListButtonIcon = Icons/nuovext/16x16/remove.png
-enterUrlButtonIcon = Icons/nuovext/16x16/ok.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+downloadButtonIcon = /usr/share/dooble/Icons/nuovext/16x16/download.png
+clearListButtonIcon = /usr/share/dooble/Icons/nuovext/16x16/remove.png
+enterUrlButtonIcon = /usr/share/dooble/Icons/nuovext/16x16/ok.png
 
 [downloadWindowItem]
-abortIcon = Icons/nuovext/16x16/cancel.png
-pauseIcon = Icons/nuovext/16x16/suspend.png
-resumeIcon = Icons/nuovext/16x16/download.png
-retryIcon = Icons/nuovext/16x16/reload.png
+abortIcon = /usr/share/dooble/Icons/nuovext/16x16/cancel.png
+pauseIcon = /usr/share/dooble/Icons/nuovext/16x16/suspend.png
+resumeIcon = /usr/share/dooble/Icons/nuovext/16x16/download.png
+retryIcon = /usr/share/dooble/Icons/nuovext/16x16/reload.png
 
 [mainWindow]
-backToolButton = Icons/nuovext/32x32/back.png
-forwardToolButton = Icons/nuovext/32x32/forward.png
-reloadToolButton = Icons/nuovext/32x32/reload.png
-stopToolButton = Icons/nuovext/32x32/stop.png
-homeToolButton = Icons/nuovext/32x32/home.png
-hideFindToolButton = Icons/nuovext/16x16/cancel.png
-actionNew_Tab = Icons/nuovext/16x16/newtab.png
-actionNew_Window = Icons/nuovext/16x16/newwindow.png
-actionOpen_URL = Icons/nuovext/16x16/navigate.png
-actionClose_Tab = Icons/nuovext/16x16/closetab.png
-actionClose_Window = Icons/nuovext/16x16/closewindow.png
-actionSave_Page = Icons/nuovext/16x16/download.png
-actionPrint = Icons/nuovext/16x16/print.png
-actionPrint_Preview = Icons/nuovext/16x16/printpreview.png
-actionQuit = Icons/nuovext/16x16/quit.png
-actionFind = Icons/nuovext/16x16/search.png
-actionReload = Icons/nuovext/16x16/reload.png
-actionStop = Icons/nuovext/16x16/stop.png
-actionDesktop = Icons/nuovext/16x16/desktop.png
-actionPage_Source = Icons/nuovext/16x16/pagesource.png
-actionDownloads = Icons/nuovext/16x16/download.png
-actionSettings = Icons/nuovext/16x16/settings.png
-actionAbout_Dooble = Icons/nuovext/16x16/about.png
-desktopToolButton = Icons/nuovext/32x32/desktop.png
-tabWidget = Icons/nuovext/16x16/desktop.png
-windowIcon = Icons/16x16/dooble.png
-historyMenu = Icons/nuovext/16x16/clearhistory.png
-emptyIcon = Icons/16x16/dooble.png
-actionOpen_Directory = Icons/nuovext/16x16/folderopen.png
-actionMy_Retrieved_Files = Icons/nuovext/16x16/folderopen.png
-addTab = Icons/nuovext/16x16/newtab.png
-removeTab = Icons/nuovext/16x16/closetab.png
-restoreToolButton = Icons/nuovext/16x16/restorenormalscreen.png
-actionCopy = Icons/nuovext/16x16/copy.png
-actionPaste = Icons/nuovext/16x16/paste.png
-zoomMenu = Icons/nuovext/16x16/zoom.png
-actionFull_Screen_Mode = Icons/nuovext/16x16/fullscreenmode.png
-viewHistory = Icons/nuovext/16x16/history.png
-nextToolButton = Icons/nuovext/16x16/next.png
-previousToolButton = Icons/nuovext/16x16/previous.png
-actionApplication_Cookies = Icons/nuovext/16x16/cookies.png
-actionP2P_Email = Icons/nuovext/16x16/email.png
+backToolButton = /usr/share/dooble/Icons/nuovext/32x32/back.png
+forwardToolButton = /usr/share/dooble/Icons/nuovext/32x32/forward.png
+reloadToolButton = /usr/share/dooble/Icons/nuovext/32x32/reload.png
+stopToolButton = /usr/share/dooble/Icons/nuovext/32x32/stop.png
+homeToolButton = /usr/share/dooble/Icons/nuovext/32x32/home.png
+hideFindToolButton = /usr/share/dooble/Icons/nuovext/16x16/cancel.png
+actionNew_Tab = /usr/share/dooble/Icons/nuovext/16x16/newtab.png
+actionNew_Window = /usr/share/dooble/Icons/nuovext/16x16/newwindow.png
+actionOpen_URL = /usr/share/dooble/Icons/nuovext/16x16/navigate.png
+actionClose_Tab = /usr/share/dooble/Icons/nuovext/16x16/closetab.png
+actionClose_Window = /usr/share/dooble/Icons/nuovext/16x16/closewindow.png
+actionSave_Page = /usr/share/dooble/Icons/nuovext/16x16/download.png
+actionPrint = /usr/share/dooble/Icons/nuovext/16x16/print.png
+actionPrint_Preview = /usr/share/dooble/Icons/nuovext/16x16/printpreview.png
+actionQuit = /usr/share/dooble/Icons/nuovext/16x16/quit.png
+actionFind = /usr/share/dooble/Icons/nuovext/16x16/search.png
+actionReload = /usr/share/dooble/Icons/nuovext/16x16/reload.png
+actionStop = /usr/share/dooble/Icons/nuovext/16x16/stop.png
+actionDesktop = /usr/share/dooble/Icons/nuovext/16x16/desktop.png
+actionPage_Source = /usr/share/dooble/Icons/nuovext/16x16/pagesource.png
+actionDownloads = /usr/share/dooble/Icons/nuovext/16x16/download.png
+actionSettings = /usr/share/dooble/Icons/nuovext/16x16/settings.png
+actionAbout_Dooble = /usr/share/dooble/Icons/nuovext/16x16/about.png
+desktopToolButton = /usr/share/dooble/Icons/nuovext/32x32/desktop.png
+tabWidget = /usr/share/dooble/Icons/nuovext/16x16/desktop.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+historyMenu = /usr/share/dooble/Icons/nuovext/16x16/clearhistory.png
+emptyIcon = /usr/share/dooble/Icons/16x16/dooble.png
+actionOpen_Directory = /usr/share/dooble/Icons/nuovext/16x16/folderopen.png
+actionMy_Retrieved_Files = /usr/share/dooble/Icons/nuovext/16x16/folderopen.png
+addTab = /usr/share/dooble/Icons/nuovext/16x16/newtab.png
+removeTab = /usr/share/dooble/Icons/nuovext/16x16/closetab.png
+restoreToolButton = /usr/share/dooble/Icons/nuovext/16x16/restorenormalscreen.png
+actionCopy = /usr/share/dooble/Icons/nuovext/16x16/copy.png
+actionPaste = /usr/share/dooble/Icons/nuovext/16x16/paste.png
+zoomMenu = /usr/share/dooble/Icons/nuovext/16x16/zoom.png
+actionFull_Screen_Mode = /usr/share/dooble/Icons/nuovext/16x16/fullscreenmode.png
+viewHistory = /usr/share/dooble/Icons/nuovext/16x16/history.png
+nextToolButton = /usr/share/dooble/Icons/nuovext/16x16/next.png
+previousToolButton = /usr/share/dooble/Icons/nuovext/16x16/previous.png
+actionApplication_Cookies = /usr/share/dooble/Icons/nuovext/16x16/cookies.png
+actionP2P_Email = /usr/share/dooble/Icons/nuovext/16x16/email.png
 
 [pageSourceWindow]
-actionPrint = Icons/nuovext/16x16/print.png
-actionPrint_Preview = Icons/nuovext/16x16/printpreview.png
-actionClose = Icons/nuovext/16x16/closewindow.png
-actionFind = Icons/nuovext/16x16/search.png
-actionSave_As = Icons/nuovext/16x16/filesaveas.png
-hideFindToolButton = Icons/nuovext/16x16/cancel.png
-windowIcon = Icons/16x16/dooble.png
-nextToolButton = Icons/nuovext/16x16/next.png
-previousToolButton = Icons/nuovext/16x16/previous.png
+actionPrint = /usr/share/dooble/Icons/nuovext/16x16/print.png
+actionPrint_Preview = /usr/share/dooble/Icons/nuovext/16x16/printpreview.png
+actionClose = /usr/share/dooble/Icons/nuovext/16x16/closewindow.png
+actionFind = /usr/share/dooble/Icons/nuovext/16x16/search.png
+actionSave_As = /usr/share/dooble/Icons/nuovext/16x16/filesaveas.png
+hideFindToolButton = /usr/share/dooble/Icons/nuovext/16x16/cancel.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+nextToolButton = /usr/share/dooble/Icons/nuovext/16x16/next.png
+previousToolButton = /usr/share/dooble/Icons/nuovext/16x16/previous.png
 
 [propertiesDialog]
-windowIcon = Icons/16x16/dooble.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
 
 [settingsWindow]
-windowIcon = Icons/16x16/dooble.png
-displayButtonIcon = Icons/nuovext/48x48/display.png
-historyButtonIcon = Icons/nuovext/48x48/history.png
-homeButtonIcon = Icons/nuovext/48x48/home.png
-proxyButtonIcon = Icons/nuovext/48x48/proxy.png
-securityButtonIcon = Icons/nuovext/48x48/security.png
-tabsButtonIcon = Icons/nuovext/48x48/tabs.png
-webkitButtonIcon = Icons/nuovext/48x48/webkit.png
-windowsButtonIcon = Icons/nuovext/48x48/windows.png
-copyAndPasteToolButtonIcon = Icons/nuovext/16x16/paste.png
-safeButtonIcon = Icons/nuovext/48x48/zerocrypt.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+displayButtonIcon = /usr/share/dooble/Icons/nuovext/48x48/display.png
+historyButtonIcon = /usr/share/dooble/Icons/nuovext/48x48/history.png
+homeButtonIcon = /usr/share/dooble/Icons/nuovext/48x48/home.png
+proxyButtonIcon = /usr/share/dooble/Icons/nuovext/48x48/proxy.png
+securityButtonIcon = /usr/share/dooble/Icons/nuovext/48x48/security.png
+tabsButtonIcon = /usr/share/dooble/Icons/nuovext/48x48/tabs.png
+webkitButtonIcon = /usr/share/dooble/Icons/nuovext/48x48/webkit.png
+windowsButtonIcon = /usr/share/dooble/Icons/nuovext/48x48/windows.png
+copyAndPasteToolButtonIcon = /usr/share/dooble/Icons/nuovext/16x16/paste.png
+safeButtonIcon = /usr/share/dooble/Icons/nuovext/48x48/zerocrypt.png
 
 [urlWidget]
-goToolButton = Icons/nuovext/16x16/navigate.png
-pulldownToolButton = Icons/nuovext/16x16/empty.png
-emptyIcon = Icons/nuovext/16x16/empty.png
-bookmarkToolButton = Icons/nuovext/16x16/bookmark.png
+goToolButton = /usr/share/dooble/Icons/nuovext/16x16/navigate.png
+pulldownToolButton = /usr/share/dooble/Icons/nuovext/16x16/empty.png
+emptyIcon = /usr/share/dooble/Icons/nuovext/16x16/empty.png
+bookmarkToolButton = /usr/share/dooble/Icons/nuovext/16x16/bookmark.png
 
 [searchWidget]
-findToolButton = Icons/nuovext/16x16/search.png
+findToolButton = /usr/share/dooble/Icons/nuovext/16x16/search.png
 
 [historyWindow]
-windowIcon = Icons/16x16/dooble.png
-deleteButtonIcon = Icons/nuovext/16x16/remove.png
-deleteAllButtonIcon = Icons/nuovext/16x16/removeall.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+deleteButtonIcon = /usr/share/dooble/Icons/nuovext/16x16/remove.png
+deleteAllButtonIcon = /usr/share/dooble/Icons/nuovext/16x16/removeall.png
 
 [cookiesWindow]
-windowIcon = Icons/16x16/dooble.png
-deleteButtonIcon = Icons/nuovext/16x16/remove.png
-deleteAllButtonIcon = Icons/nuovext/16x16/removeall.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+deleteButtonIcon = /usr/share/dooble/Icons/nuovext/16x16/remove.png
+deleteAllButtonIcon = /usr/share/dooble/Icons/nuovext/16x16/removeall.png
diff -ur dooble.d.orig/trunk/browser/Icons/nuvola/configuration.cfg dooble.d/trunk/browser/Icons/nuvola/configuration.cfg
--- dooble.d.orig/trunk/browser/Icons/nuvola/configuration.cfg	2011-05-15 22:59:35.000000000 +0200
+++ dooble.d/trunk/browser/Icons/nuvola/configuration.cfg	2011-09-16 13:53:37.025839690 +0200
@@ -4,108 +4,108 @@
 License = GLGPL (GNU Lesser General Public License)
 
 [General]
-windowIcon = Icons/16x16/dooble.png
-okButtonIcon = Icons/nuvola/16x16/ok.png
-cancelButtonIcon = Icons/nuvola/16x16/cancel.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+okButtonIcon = /usr/share/dooble/Icons/nuvola/16x16/ok.png
+cancelButtonIcon = /usr/share/dooble/Icons/nuvola/16x16/cancel.png
 
 [downloadWindow]
-windowIcon = Icons/16x16/dooble.png
-downloadButtonIcon = Icons/nuvola/16x16/download.png
-clearListButtonIcon = Icons/nuvola/16x16/remove.png
-enterUrlButtonIcon = Icons/nuvola/16x16/ok.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+downloadButtonIcon = /usr/share/dooble/Icons/nuvola/16x16/download.png
+clearListButtonIcon = /usr/share/dooble/Icons/nuvola/16x16/remove.png
+enterUrlButtonIcon = /usr/share/dooble/Icons/nuvola/16x16/ok.png
 
 [downloadWindowItem]
-abortIcon = Icons/nuvola/16x16/cancel.png
-pauseIcon = Icons/nuvola/16x16/suspend.png
-resumeIcon = Icons/nuvola/16x16/download.png
-retryIcon = Icons/nuvola/16x16/reload.png
+abortIcon = /usr/share/dooble/Icons/nuvola/16x16/cancel.png
+pauseIcon = /usr/share/dooble/Icons/nuvola/16x16/suspend.png
+resumeIcon = /usr/share/dooble/Icons/nuvola/16x16/download.png
+retryIcon = /usr/share/dooble/Icons/nuvola/16x16/reload.png
 
 [mainWindow]
-backToolButton = Icons/nuvola/32x32/back.png
-forwardToolButton = Icons/nuvola/32x32/forward.png
-reloadToolButton = Icons/nuvola/32x32/reload.png
-stopToolButton = Icons/nuvola/32x32/stop.png
-homeToolButton = Icons/nuvola/32x32/home.png
-hideFindToolButton = Icons/nuvola/16x16/cancel.png
-actionNew_Tab = Icons/nuvola/16x16/newtab.png
-actionNew_Window = Icons/nuvola/16x16/newwindow.png
-actionOpen_URL = Icons/nuvola/16x16/navigate.png
-actionClose_Tab = Icons/nuvola/16x16/closetab.png
-actionClose_Window = Icons/nuvola/16x16/closewindow.png
-actionSave_Page = Icons/nuvola/16x16/download.png
-actionPrint = Icons/nuvola/16x16/print.png
-actionPrint_Preview = Icons/nuvola/16x16/printpreview.png
-actionQuit = Icons/nuvola/16x16/quit.png
-actionFind = Icons/nuvola/16x16/search.png
-actionReload = Icons/nuvola/16x16/reload.png
-actionStop = Icons/nuvola/16x16/stop.png
-actionDesktop = Icons/nuvola/16x16/desktop.png
-actionPage_Source = Icons/nuvola/16x16/pagesource.png
-actionDownloads = Icons/nuvola/16x16/download.png
-actionSettings = Icons/nuvola/16x16/settings.png
-actionAbout_Dooble = Icons/nuvola/16x16/about.png
-desktopToolButton = Icons/nuvola/32x32/desktop.png
-tabWidget = Icons/nuvola/16x16/desktop.png
-windowIcon = Icons/16x16/dooble.png
-historyMenu = Icons/nuvola/16x16/clearhistory.png
-emptyIcon = Icons/16x16/dooble.png
-actionOpen_Directory = Icons/nuvola/16x16/folderopen.png
-actionMy_Retrieved_Files = Icons/nuvola/16x16/folderopen.png
-addTab = Icons/nuvola/16x16/newtab.png
-removeTab = Icons/nuvola/16x16/closetab.png
-restoreToolButton = Icons/nuvola/16x16/restorenormalscreen.png
-actionCopy = Icons/nuvola/16x16/copy.png
-actionPaste = Icons/nuvola/16x16/paste.png
-zoomMenu = Icons/nuvola/16x16/zoom.png
-actionFull_Screen_Mode = Icons/nuvola/16x16/fullscreenmode.png
-viewHistory = Icons/nuvola/16x16/history.png
-nextToolButton = Icons/nuvola/16x16/next.png
-previousToolButton = Icons/nuvola/16x16/previous.png
-actionApplication_Cookies = Icons/nuvola/16x16/cookies.png
-actionP2P_Email = Icons/nuvola/16x16/email.png
+backToolButton = /usr/share/dooble/Icons/nuvola/32x32/back.png
+forwardToolButton = /usr/share/dooble/Icons/nuvola/32x32/forward.png
+reloadToolButton = /usr/share/dooble/Icons/nuvola/32x32/reload.png
+stopToolButton = /usr/share/dooble/Icons/nuvola/32x32/stop.png
+homeToolButton = /usr/share/dooble/Icons/nuvola/32x32/home.png
+hideFindToolButton = /usr/share/dooble/Icons/nuvola/16x16/cancel.png
+actionNew_Tab = /usr/share/dooble/Icons/nuvola/16x16/newtab.png
+actionNew_Window = /usr/share/dooble/Icons/nuvola/16x16/newwindow.png
+actionOpen_URL = /usr/share/dooble/Icons/nuvola/16x16/navigate.png
+actionClose_Tab = /usr/share/dooble/Icons/nuvola/16x16/closetab.png
+actionClose_Window = /usr/share/dooble/Icons/nuvola/16x16/closewindow.png
+actionSave_Page = /usr/share/dooble/Icons/nuvola/16x16/download.png
+actionPrint = /usr/share/dooble/Icons/nuvola/16x16/print.png
+actionPrint_Preview = /usr/share/dooble/Icons/nuvola/16x16/printpreview.png
+actionQuit = /usr/share/dooble/Icons/nuvola/16x16/quit.png
+actionFind = /usr/share/dooble/Icons/nuvola/16x16/search.png
+actionReload = /usr/share/dooble/Icons/nuvola/16x16/reload.png
+actionStop = /usr/share/dooble/Icons/nuvola/16x16/stop.png
+actionDesktop = /usr/share/dooble/Icons/nuvola/16x16/desktop.png
+actionPage_Source = /usr/share/dooble/Icons/nuvola/16x16/pagesource.png
+actionDownloads = /usr/share/dooble/Icons/nuvola/16x16/download.png
+actionSettings = /usr/share/dooble/Icons/nuvola/16x16/settings.png
+actionAbout_Dooble = /usr/share/dooble/Icons/nuvola/16x16/about.png
+desktopToolButton = /usr/share/dooble/Icons/nuvola/32x32/desktop.png
+tabWidget = /usr/share/dooble/Icons/nuvola/16x16/desktop.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+historyMenu = /usr/share/dooble/Icons/nuvola/16x16/clearhistory.png
+emptyIcon = /usr/share/dooble/Icons/16x16/dooble.png
+actionOpen_Directory = /usr/share/dooble/Icons/nuvola/16x16/folderopen.png
+actionMy_Retrieved_Files = /usr/share/dooble/Icons/nuvola/16x16/folderopen.png
+addTab = /usr/share/dooble/Icons/nuvola/16x16/newtab.png
+removeTab = /usr/share/dooble/Icons/nuvola/16x16/closetab.png
+restoreToolButton = /usr/share/dooble/Icons/nuvola/16x16/restorenormalscreen.png
+actionCopy = /usr/share/dooble/Icons/nuvola/16x16/copy.png
+actionPaste = /usr/share/dooble/Icons/nuvola/16x16/paste.png
+zoomMenu = /usr/share/dooble/Icons/nuvola/16x16/zoom.png
+actionFull_Screen_Mode = /usr/share/dooble/Icons/nuvola/16x16/fullscreenmode.png
+viewHistory = /usr/share/dooble/Icons/nuvola/16x16/history.png
+nextToolButton = /usr/share/dooble/Icons/nuvola/16x16/next.png
+previousToolButton = /usr/share/dooble/Icons/nuvola/16x16/previous.png
+actionApplication_Cookies = /usr/share/dooble/Icons/nuvola/16x16/cookies.png
+actionP2P_Email = /usr/share/dooble/Icons/nuvola/16x16/email.png
 
 [pageSourceWindow]
-actionPrint = Icons/nuvola/16x16/print.png
-actionPrint_Preview = Icons/nuvola/16x16/printpreview.png
-actionClose = Icons/nuvola/16x16/closewindow.png
-actionFind = Icons/nuvola/16x16/search.png
-actionSave_As = Icons/nuvola/16x16/filesaveas.png
-hideFindToolButton = Icons/nuvola/16x16/cancel.png
-windowIcon = Icons/16x16/dooble.png
-nextToolButton = Icons/nuvola/16x16/next.png
-previousToolButton = Icons/nuvola/16x16/previous.png
+actionPrint = /usr/share/dooble/Icons/nuvola/16x16/print.png
+actionPrint_Preview = /usr/share/dooble/Icons/nuvola/16x16/printpreview.png
+actionClose = /usr/share/dooble/Icons/nuvola/16x16/closewindow.png
+actionFind = /usr/share/dooble/Icons/nuvola/16x16/search.png
+actionSave_As = /usr/share/dooble/Icons/nuvola/16x16/filesaveas.png
+hideFindToolButton = /usr/share/dooble/Icons/nuvola/16x16/cancel.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+nextToolButton = /usr/share/dooble/Icons/nuvola/16x16/next.png
+previousToolButton = /usr/share/dooble/Icons/nuvola/16x16/previous.png
 
 [propertiesDialog]
-windowIcon = Icons/16x16/dooble.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
 
 [settingsWindow]
-windowIcon = Icons/16x16/dooble.png
-displayButtonIcon = Icons/nuvola/48x48/display.png
-historyButtonIcon = Icons/nuvola/48x48/history.png
-homeButtonIcon = Icons/nuvola/48x48/home.png
-proxyButtonIcon = Icons/nuvola/48x48/proxy.png
-securityButtonIcon = Icons/nuvola/48x48/security.png
-tabsButtonIcon = Icons/nuvola/48x48/tabs.png
-webkitButtonIcon = Icons/nuvola/48x48/webkit.png
-windowsButtonIcon = Icons/nuvola/48x48/windows.png
-copyAndPasteToolButtonIcon = Icons/nuvola/16x16/paste.png
-safeButtonIcon = Icons/nuvola/48x48/zerocrypt.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+displayButtonIcon = /usr/share/dooble/Icons/nuvola/48x48/display.png
+historyButtonIcon = /usr/share/dooble/Icons/nuvola/48x48/history.png
+homeButtonIcon = /usr/share/dooble/Icons/nuvola/48x48/home.png
+proxyButtonIcon = /usr/share/dooble/Icons/nuvola/48x48/proxy.png
+securityButtonIcon = /usr/share/dooble/Icons/nuvola/48x48/security.png
+tabsButtonIcon = /usr/share/dooble/Icons/nuvola/48x48/tabs.png
+webkitButtonIcon = /usr/share/dooble/Icons/nuvola/48x48/webkit.png
+windowsButtonIcon = /usr/share/dooble/Icons/nuvola/48x48/windows.png
+copyAndPasteToolButtonIcon = /usr/share/dooble/Icons/nuvola/16x16/paste.png
+safeButtonIcon = /usr/share/dooble/Icons/nuvola/48x48/zerocrypt.png
 
 [urlWidget]
-goToolButton = Icons/nuvola/16x16/navigate.png
-pulldownToolButton = Icons/nuvola/16x16/empty.png
-emptyIcon = Icons/nuvola/16x16/empty.png
-bookmarkToolButton = Icons/nuvola/16x16/bookmark.png
+goToolButton = /usr/share/dooble/Icons/nuvola/16x16/navigate.png
+pulldownToolButton = /usr/share/dooble/Icons/nuvola/16x16/empty.png
+emptyIcon = /usr/share/dooble/Icons/nuvola/16x16/empty.png
+bookmarkToolButton = /usr/share/dooble/Icons/nuvola/16x16/bookmark.png
 
 [searchWidget]
-findToolButton = Icons/nuvola/16x16/search.png
+findToolButton = /usr/share/dooble/Icons/nuvola/16x16/search.png
 
 [historyWindow]
-windowIcon = Icons/16x16/dooble.png
-deleteButtonIcon = Icons/nuvola/16x16/remove.png
-deleteAllButtonIcon = Icons/nuvola/16x16/removeall.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+deleteButtonIcon = /usr/share/dooble/Icons/nuvola/16x16/remove.png
+deleteAllButtonIcon = /usr/share/dooble/Icons/nuvola/16x16/removeall.png
 
 [cookiesWindow]
-windowIcon = Icons/16x16/dooble.png
-deleteButtonIcon = Icons/nuvola/16x16/remove.png
-deleteAllButtonIcon = Icons/nuvola/16x16/removeall.png
+windowIcon = /usr/share/dooble/Icons/16x16/dooble.png
+deleteButtonIcon = /usr/share/dooble/Icons/nuvola/16x16/remove.png
+deleteAllButtonIcon = /usr/share/dooble/Icons/nuvola/16x16/removeall.png
diff -ur dooble.d.orig/trunk/browser/Source/dooble.cc dooble.d/trunk/browser/Source/dooble.cc
--- dooble.d.orig/trunk/browser/Source/dooble.cc	2011-08-19 01:08:42.000000000 +0200
+++ dooble.d/trunk/browser/Source/dooble.cc	2011-09-16 13:53:37.033840707 +0200
@@ -327,21 +327,18 @@
 
   if(!settings.contains("iconSet"))
     settings.setValue("iconSet",
-		      QString("%1/%2").arg(QDir::currentPath()).
-		      arg("Icons/nuovext/configuration.cfg"));
+		      QString("/usr/share/dooble/Icons/nuovext/configuration.cfg"));
 
   if(!QFileInfo(settings.value("iconSet").toString()).exists())
     settings.setValue("iconSet",
-		      QString("%1/%2").arg(QDir::currentPath()).
-		      arg("Icons/nuovext/configuration.cfg"));
+		      QString("/usr/share/dooble/Icons/nuovext/configuration.cfg"));
 
   if(!settings.contains("settingsWindow/iconSet1"))
     settings.setValue("settingsWindow/iconSet1", settings.value("iconSet"));
 
   if(!QFileInfo(settings.value("settingsWindow/iconSet1").toString()).exists())
     settings.setValue("settingsWindow/iconSet1",
-		      QString("%1/%2").arg(QDir::currentPath()).
-		      arg("Icons/nuovext/configuration.cfg"));
+		      QString("/usr/share/dooble/Icons/nuovext/configuration.cfg"));
 
   dooble::s_settings.clear();
 
diff -ur dooble.d.orig/trunk/browser/Source/dsettings.cc dooble.d/trunk/browser/Source/dsettings.cc
--- dooble.d.orig/trunk/browser/Source/dsettings.cc	2011-08-18 00:04:05.000000000 +0200
+++ dooble.d/trunk/browser/Source/dsettings.cc	2011-09-16 13:54:09.782002829 +0200
@@ -80,11 +80,10 @@
   if(!dooble::s_settings.contains("iconSet"))
     {
       settings.setValue("iconSet",
-			QString("%1/%2").arg(QDir::currentPath()).
-			arg("Icons/nuovext/configuration.cfg"));
+			QString("/usr/share/dooble/Icons/nuovext/configuration.cfg"));
+
       dooble::s_settings["iconSet"] =
-	QString("%1/%2").arg(QDir::currentPath()).
-	arg("Icons/nuovext/configuration.cfg");
+	QString("/usr/share/dooble/Icons/nuovext/configuration.cfg");
     }
 
   if(!dooble::s_settings.contains("settingsWindow/iconSet1"))
