diff -ur dooble.d.orig/trunk/browser/Source/ddesktopwidget.cc dooble.d/trunk/browser/Source/ddesktopwidget.cc
--- dooble.d.orig/trunk/browser/Source/ddesktopwidget.cc	2011-08-11 00:04:01.000000000 +0200
+++ dooble.d/trunk/browser/Source/ddesktopwidget.cc	2011-09-16 13:55:49.742707322 +0200
@@ -85,9 +85,7 @@
   else
     {
       QDir currentDir(QDir::currentPath());
-      QString fileName(QString
-		       ("%1/Images/doobledesktop_background_ublue.png").arg
-		       (currentDir.path()));
+      QString fileName("/usr/share/dooble/Images/doobledesktop_background_ublue.png");
       QSettings settings;
 
       settings.setValue("desktopBackground", fileName);
