Index: dooble-0.0+svn874/Source/dooble.cc
===================================================================
--- dooble-0.0+svn874.orig/Source/dooble.cc	2009-09-13 14:37:57.527010599 +0200
+++ dooble-0.0+svn874/Source/dooble.cc	2009-09-13 14:37:57.559011026 +0200
@@ -150,8 +150,7 @@
 
   if(!settings.contains("iconSet"))
     settings.setValue("iconSet",
-		      QString("%1/%2").arg(QDir::currentPath()).
-		      arg("Icons/nuvola/configuration.cfg"));
+		      QString("/usr/share/dooble/Icons/nuvola/configuration.cfg"));
 
   if(!settings.contains("settingsWindow/iconSet1"))
     settings.setValue("settingsWindow/iconSet1", settings.value("iconSet"));
