Index: dooble-0.0+svn874/Source/dooble.cc
===================================================================
--- dooble-0.0+svn874.orig/Source/dooble.cc	2009-09-13 13:58:35.899009142 +0200
+++ dooble-0.0+svn874/Source/dooble.cc	2009-09-13 14:01:36.435493405 +0200
@@ -174,7 +174,7 @@
 		   SLOT(slotBackgroundImageChanged(void)));
 
   dooble *dbl = 0;
-  QString urlText = QString("%1/Tab/Default/search.html").arg
+  QString urlText = QString("/usr/share/dooble/Tab/Default/search.html").arg
     (QDir::currentPath());
 
   if(argc > 1)
@@ -643,7 +643,7 @@
 void dooble::slotNewTab(void)
 {
   QUrl url = QUrl::fromLocalFile
-    (QString("%1/Tab/Default/search.html").arg(QDir::currentPath()));
+    (QString("/usr/share/dooble/Tab/Default/search.html").arg(QDir::currentPath()));
 
   newTab(url);
 }
@@ -651,7 +651,7 @@
 void dooble::slotNewWindow(void)
 {
   QUrl url = QUrl::fromLocalFile
-    (QString("%1/Tab/Default/search.html").arg(QDir::currentPath()));
+    (QString("/usr/share/dooble/Tab/Default/search.html").arg(QDir::currentPath()));
 
   Q_UNUSED(new dooble(ui.historyMenu->actions(), url));
 }
Index: dooble-0.0+svn874/Source/dsettings.cc
===================================================================
--- dooble-0.0+svn874.orig/Source/dsettings.cc	2009-09-13 14:02:29.047014500 +0200
+++ dooble-0.0+svn874/Source/dsettings.cc	2009-09-13 14:02:54.852717422 +0200
@@ -86,7 +86,7 @@
   connect(ui.chooseBackgroundPushButton, SIGNAL(clicked(void)),
 	  this, SLOT(slotChooseBackgroundImage(void)));
   ui.doobleSearchLineEdit->setText
-    (QUrl::fromLocalFile(QString("%1/Tab/Default/search.html").arg
+    (QUrl::fromLocalFile(QString("/usr/share/dooble/Tab/Default/search.html").arg
 			 (QDir::currentPath())).toString());
   settings.setValue("settingsWindow/url12",
 		    ui.doobleSearchLineEdit->text().trimmed());
