diff -Nurp python-nss-0.8.orig/setup.py python-nss-0.8/setup.py
--- python-nss-0.8.orig/setup.py	2009-09-21 21:18:37.000000000 +0200
+++ python-nss-0.8/setup.py	2010-04-01 19:46:44.172982047 +0200
@@ -310,32 +310,36 @@ extra_compile_args = []
 nss_error_extension = \
     Extension('nss.error',
               sources            = ['src/py_nspr_error.c'],
-              include_dirs       = ['/usr/include/nss3', '/usr/include/nspr4'],
+              include_dirs       = ['/usr/include/nss', '/usr/include/nspr'],
               libraries          = ['nspr4'],
+              library_dirs       = ['/usr/lib/nspr', '/usr/lib/nss'],
               extra_compile_args = extra_compile_args,
               )
 
 nss_io_extension = \
     Extension('nss.io',
               sources            = ['src/py_nspr_io.c'],
-              include_dirs       = ['/usr/include/nss3', '/usr/include/nspr4'],
+              include_dirs       = ['/usr/include/nss', '/usr/include/nspr'],
               libraries          = ['nspr4'],
+              library_dirs       = ['/usr/lib/nspr', '/usr/lib/nss'],
               extra_compile_args = extra_compile_args,
               )
 
 nss_nss_extension = \
     Extension('nss.nss',
               sources            = ['src/py_nss.c'],
-              include_dirs       = ['src', '/usr/include/nss3', '/usr/include/nspr4'],
+              include_dirs       = ['src', '/usr/include/nss', '/usr/include/nspr'],
               libraries          = ['nspr4', 'ssl3'],
+              library_dirs       = ['/usr/lib/nspr', '/usr/lib/nss'],
               extra_compile_args = extra_compile_args,
               )
 
 nss_ssl_extension = \
     Extension('nss.ssl',
               sources            = ['src/py_ssl.c'],
-              include_dirs       = ['src', '/usr/include/nss3', '/usr/include/nspr4'],
+              include_dirs       = ['src', '/usr/include/nss', '/usr/include/nspr'],
               libraries          = ['nspr4', 'ssl3'],
+              library_dirs       = ['/usr/lib/nspr', '/usr/lib/nss'],
               extra_compile_args = extra_compile_args,
               )
 
