diff -Nurp xserver-xorg-video-intel-2.2.1-patched/src/i830_quirks.c xserver-xorg-video-intel-2.2.1-working/src/i830_quirks.c
--- xserver-xorg-video-intel-2.2.1-patched/src/i830_quirks.c	2008-05-01 22:05:31.000000000 -0700
+++ xserver-xorg-video-intel-2.2.1-working/src/i830_quirks.c	2008-05-02 15:23:54.000000000 -0700
@@ -226,6 +226,8 @@ static i830_quirk i830_quirk_list[] = {
     { PCI_CHIP_I965_GM, 0x1028, 0x0254, quirk_ignore_tv },
     /* Dell Inspiron 1735 */
     { PCI_CHIP_I965_GM, 0x1028, 0x0256, quirk_ignore_tv },
+    /* Dell Inspiron 1318 */
+    { PCI_CHIP_I965_GM, 0x1028, 0x0286, quirk_ignore_tv },
 
     /* Lenovo Napa TV (use dmi)*/
     { PCI_CHIP_I945_GM, 0x17aa, SUBSYS_ANY, quirk_lenovo_tv_dmi },
@@ -255,9 +257,6 @@ static i830_quirk i830_quirk_list[] = {
     /* HP Compaq 6730s has no TV output */
     { PCI_CHIP_IGD_GM, 0x103c, 0x30e8, quirk_ignore_tv },
 
-    /* Dell Inspiron 510m needs pipe A force quirk */
-    { PCI_CHIP_I855_GM, 0x1028, 0x0164, quirk_pipea_force },
-
     /* Thinkpad R31 needs pipe A force quirk */
     { PCI_CHIP_I830_M, 0x1014, 0x0505, quirk_pipea_force },
     /* Dell Latitude D500 needs pipe A force quirk */
@@ -268,18 +267,27 @@ static i830_quirk i830_quirk_list[] = {
     { PCI_CHIP_I855_GM, 0x1028, 0x0164, quirk_pipea_force },
     /* Toshiba Protege R-205, S-209 needs pipe A force quirk */
     { PCI_CHIP_I915_GM, 0x1179, 0x0001, quirk_pipea_force },
+    /* Fujitsu Lifebook P7010 needs pipe A force quirk (partial fix 15310, LP: #204603) */
+    { PCI_CHIP_I855_GM, 0x10cf, 0x1282, quirk_pipea_force },
+    /* Dell Latitude C400 (See LP: #216490) */
+    { PCI_CHIP_I855_GM, 0x1028, 0x00c8, quirk_pipea_force },
+    /* Fujitsu Siemens Amilo M1405 (see LP: #224102) */
+    { PCI_CHIP_I855_GM, 0x1734, 0x106a, quirk_pipea_force },
 
     /* ThinkPad X40 needs pipe A force quirk */
     { PCI_CHIP_I855_GM, 0x1014, 0x0557, quirk_pipea_force },
 
     /* Sony vaio PCG-r600HFP (fix bug 13722) */
     { PCI_CHIP_I830_M, 0x104d, 0x8100, quirk_ivch_dvob },
-    /* Sony Vaio VGN-SZ4MN (LP: #212163* */
+    /* Sony Vaio VGN-SZ4MN (LP: #212163) */
     { PCI_CHIP_I830_M, 0x104d, 0x81e6, quirk_pipea_force },
 
-    /* Intel 945GM hardware (See LP: #152416) */
+    /* Ordi Enduro UW31 (LP: #152416) */
     { PCI_CHIP_I945_GM, 0x1584, 0x9900, quirk_ignore_tv },
 
+    /* Littlebit Sepia X35 (rebranded Asus Z37E) has no tv output (LP: #201257) */
+    { PCI_CHIP_I965_GM, 0x1043, 0x8265, quirk_ignore_tv },
+
     { 0, 0, 0, NULL },
 };
 
