--- Jamroot.orig	2009-02-10 17:36:25.000000000 +0100
+++ Jamroot	2009-02-10 17:37:09.000000000 +0100
@@ -442,15 +442,15 @@
             local noversion-file ;
             if $(nt)
             {
-                noversion-file = [ MATCH "(.*)-[0-9_]+([.]lib)" : $(name) ] ;
+                noversion-file = [ MATCH "(.*)-[0-9_]+(.*[.]lib)" : $(name) ] ;
             }
             else
             {
                 noversion-file = 
-                  [ MATCH "(.*)-[0-9_]+([.]so)[.0-9]*" : $(name) ] 
-                  [ MATCH "(.*)-[0-9_]+([.]dylib)" : $(name) ] 
-                  [ MATCH "(.*)-[0-9_]+([.]a)" : $(name) ] 
-                  [ MATCH "(.*)-[0-9_]+([.]dll[.]a)" : $(name) ] ;
+                  [ MATCH "(.*)-[0-9_]+(.*[.]so)[.0-9]*" : $(name) ] 
+                  [ MATCH "(.*)-[0-9_]+(.*[.]dylib)" : $(name) ] 
+                  [ MATCH "(.*)-[0-9_]+(.*[.]a)" : $(name) ] 
+                  [ MATCH "(.*)-[0-9_]+(.*[.]dll[.]a)" : $(name) ] ;
             }
                         
             local new-name = 
