--- boost_1_36_0/Jamroot.orig	2008-11-14 16:36:10.000000000 +0100
+++ boost_1_36_0/Jamroot	2008-11-14 16:36:54.000000000 +0100
@@ -311,7 +311,7 @@
         if $(layout) = versioned
         {
             local result = [ common.format-name
-                <base> <toolset> <threading> <runtime> -$(BOOST_VERSION_TAG)
+                <base> <threading> <runtime> -$(BOOST_VERSION_TAG)
                 -$(BUILD_ID)
                 : $(name) : $(type) : $(property-set) ] ;
             
