diff -ru libgphoto2-2.4.0-orig/packaging/generic/print-camera-list.c libgphoto2-2.4.0/packaging/generic/print-camera-list.c
--- libgphoto2-2.4.0-orig/packaging/generic/print-camera-list.c	2007-09-26 15:04:29.133025885 +0200
+++ libgphoto2-2.4.0/packaging/generic/print-camera-list.c	2007-09-26 15:06:49.639775971 +0200
@@ -444,7 +444,7 @@
 		"BUS!=\"usb_device\", GOTO=\"libgphoto2_rules_end\"\n"
 		"ACTION!=\"add\", GOTO=\"libgphoto2_rules_end\"\n\n",
 		/* UDEV_0_98 */
-		"SUBSYSTEM!=\"usb_device\", GOTO=\"libgphoto2_rules_end\"\n"
+		"SUBSYSTEM!=\"usb|usb_device\", GOTO=\"libgphoto2_rules_end\"\n"
 		"ACTION!=\"add\", GOTO=\"libgphoto2_rules_end\"\n\n"
 	};
 	static char *usbcam_strings[] = {
