Index: knetworkmanager/src/knetworkmanager-nminfo_dbus.cpp
===================================================================
--- knetworkmanager/src/knetworkmanager-nminfo_dbus.cpp	(Revision 601012)
+++ knetworkmanager/src/knetworkmanager-nminfo_dbus.cpp	(Revision 602671)
@@ -321,10 +321,6 @@
 		active_network->insertHardwareAddress (bssid);
 		encryption->setWeCipher (we_cipher);
 		
-		if (!encryption->deserialize (&iter, we_cipher)) {
-			kdWarning() << k_funcinfo << "Failed to deserialize security. Probably we_ciphers are different." << endl;
-		}
-
 		// pass to NMI to have it stored
 		NetworkManagerInfo* nmi = _ctx->getNetworkManagerInfo ();
 		nmi->emitNetworkUpdated (active_network, automatic);
