Index: knetworkmanager/src/knetworkmanager-state_dbus.cpp
===================================================================
--- knetworkmanager/src/knetworkmanager-state_dbus.cpp	(Revision 621628)
+++ knetworkmanager/src/knetworkmanager-state_dbus.cpp	(Revision 621629)
@@ -82,7 +82,7 @@
 {
 	DBusConnection* con   = _ctx->getDBus ()->getConnection ();
 	DBusMessage*    reply = NULL;
-	bool            state = false;
+	dbus_bool_t	state = false;
 
 	if (!con || !pcall) {
 		return;
