diff -Nur knetworkmanager-0.1/knetworkmanager/src/knetworkmanager-tray.cpp knetworkmanager-0.1.new/knetworkmanager/src/knetworkmanager-tray.cpp
--- knetworkmanager-0.1/knetworkmanager/src/knetworkmanager-tray.cpp	2007-02-19 12:29:55.000000000 +0000
+++ knetworkmanager-0.1.new/knetworkmanager/src/knetworkmanager-tray.cpp	2007-02-19 12:32:56.000000000 +0000
@@ -257,6 +257,14 @@
 }
 
 void
+Tray::staticConfiguration ()
+{
+  KProcess *proc = new KProcess;
+  *proc << "kdesu" << "kcmshell kcm_knetworkconfmodule";
+  proc->start();
+}
+
+void
 Tray::configureVPN ()
 {
 	emit configureVPNConnections ();
@@ -782,6 +790,9 @@
 		}
 	}
 
+	KAction* staticAction = actionCollection()->action("static_configuration");
+	staticAction->plug(menu);
+
 	/* build-up the ``Options'' sub menu if NM is running */
 	if (state->isNetworkManagerRunning ()) {
 		if (!separated)
@@ -1095,6 +1106,10 @@
 					     SmallIcon ("wireless",  QIconSet::Automatic), 0,
 					     this, SLOT (connectHiddenNetwork ()), actionCollection (), "connect_hidden");
 	
+	staticConfigurationAction   = new KAction (i18n ("Static Connection..."),
+					     SmallIcon ("knetworkconf",  QIconSet::Automatic), 0,
+					     this, SLOT (staticConfiguration ()), actionCollection (), "static_configuration");
+	
 	configureVPNAction = new KAction (i18n ("Configure VPN..."),
 					  SmallIcon ("configure",  QIconSet::Automatic), 0,
 					  this, SLOT (configureVPN ()), actionCollection (), "configure_vpn");
diff -Nur knetworkmanager-0.1/knetworkmanager/src/knetworkmanager-tray.h knetworkmanager-0.1.new/knetworkmanager/src/knetworkmanager-tray.h
--- knetworkmanager-0.1/knetworkmanager/src/knetworkmanager-tray.h	2007-02-19 12:29:55.000000000 +0000
+++ knetworkmanager-0.1.new/knetworkmanager/src/knetworkmanager-tray.h	2007-02-19 12:29:56.000000000 +0000
@@ -110,6 +110,7 @@
 	  ~Tray ();
 
 	  KAction* connectHiddenAction;
+	  KAction* staticConfigurationAction;
 	  KAction* configureVPNAction;
 	  KAction* disconnectVPNAction;
 	  KAction* configureDialUpAction;
@@ -152,6 +153,7 @@
 	public slots:
 	  void updateTooltip          (void);
 	  void connectHiddenNetwork   (void);
+	  void staticConfiguration    (void);
 	  void configureVPN           (void);
 	  void disconnectVPN          (void);
 	  void configureDialUp        (void);
