diff -ruN poppler-0.18.4.orig/poppler/TextOutputDev.cc poppler-0.18.4/poppler/TextOutputDev.cc
--- poppler-0.18.4.orig/poppler/TextOutputDev.cc	2011-11-22 22:33:16.000000000 +0100
+++ poppler-0.18.4/poppler/TextOutputDev.cc	2012-04-05 21:57:45.000000000 +0200
@@ -4009,7 +4009,9 @@
       // not a table
       else {
         page->dumpFragment (frag->line->text + frag->start, frag->len, uMap, s);
-        s->append (eol, eolLen);
+        if (i < nFrags -1) {
+          s->append (eol, eolLen);
+        }
       }
     }
 
