--- a/gir/Makefile.am	2009-02-14 01:58:57.000000000 +0100
+++ b/gir/Makefile.am	2009-02-14 01:41:35.000000000 +0100
@@ -141,7 +141,7 @@
 	$(top_srcdir)/girepository/girepository.h
 
 GIRepository-2.0.gir: GObject-2.0.gir $(SCANNER_BIN) $(SCANNER_LIBS) $(GIREPOSITORY_FILES)
-	$(SCANNER) \
+	LD_LIBRARY_PATH=../girepository/.libs/ $(SCANNER) \
 	    --namespace GIRepository --nsversion=2.0 \
 	    --noclosure \
 	    --output $@ \
