--- a/opengl.eselect
+++ b/opengl.eselect
@@ -148,7 +148,7 @@ set_new_implementation() {
 				mkdir -p "${DST_PREFIX}/${libdir}/${moduledir}/extensions" || die "Failed to create ${DST_PREFIX}/${libdir}/${moduledir}/extensions"
 				pushd "${DST_PREFIX}/${libdir}/${moduledir}/extensions" &> /dev/null
 					# First remove old symlinks
-					for file in lib{wfb,glx,dri}.{a,so,la}; do
+					for file in lib{wfb,glx,dri,dri2}.{a,so,la}; do
 						rm -f "${file}" || die -q "Failed to delete ${DST_PREFIX}/${libdir}/${moduledir}/extensions/${file}"
 					done
 
