diff -Nurp wine-0.9.32.orig/dlls/wined3d/device.c wine-0.9.32/dlls/wined3d/device.c
--- wine-0.9.32.orig/dlls/wined3d/device.c	2007-03-02 17:12:43.000000000 +0100
+++ wine-0.9.32/dlls/wined3d/device.c	2007-03-15 15:28:37.000000000 +0100
@@ -4269,7 +4269,10 @@ static HRESULT  WINAPI  IWineD3DDeviceIm
     GLuint vbo;
 
     pIB = This->stateBlock->pIndexData;
+    if (!pIB) return WINED3D_OK;    
+    
     This->stateBlock->streamIsUP = FALSE;
+
     vbo = ((IWineD3DIndexBufferImpl *) pIB)->vbo;
 
     TRACE("(%p) : Type=(%d,%s), min=%d, CountV=%d, startIdx=%d, countP=%d\n", This,
