--- Makefile.orig	2009-03-16 23:22:39.000000000 +0100
+++ Makefile	2009-03-16 23:27:30.791754757 +0100
@@ -407,9 +407,9 @@
 	  $(SED) -e "s*DATADIR*$(DATADIR)*g" < ./inst/$(PROJECT).ini.template > "$(GLOBAL_QMC2_INI)" ; \
 	fi
 ifneq '$(ARCH)' 'Darwin'
-	@echo "Installing $(TARGET_NAME).desktop to $(GLOBAL_DATADIR)/applications"
-	@$(MKDIR) $(GLOBAL_DATADIR)/applications
-	@$(SED) -e "s*DATADIR*$(DATADIR)*g; s*EMULATOR*$(EMULATOR)*g; s*TARGET*$(TARGET_NAME)*g; s*EMUICO*$(EMUICO)*g; s*GENERICNAME*$(GENERICNAME)*g" < ./inst/$(PROJECT).desktop.template > $(GLOBAL_DATADIR)/applications/$(TARGET_NAME).desktop
+	@echo "Installing $(TARGET_NAME).desktop to $(DESTDIR)/usr/share/applications"
+	@$(MKDIR) $(DESTDIR)/usr/share/applications
+	@$(SED) -e "s*DATADIR*$(DATADIR)*g; s*EMULATOR*$(EMULATOR)*g; s*TARGET*$(TARGET_NAME)*g; s*EMUICO*$(EMUICO)*g; s*GENERICNAME*$(GENERICNAME)*g" < ./inst/$(PROJECT).desktop.template > $(DESTDIR)/usr/share/applications/$(TARGET_NAME).desktop
 endif
 	@echo "Installation complete"
 
