Index: daemon/gvfsbackendhttp.c
===================================================================
--- daemon/gvfsbackendhttp.c	(revision 2129)
+++ daemon/gvfsbackendhttp.c	(revision 2130)
@@ -346,6 +346,7 @@
                                 error->message);
 
       g_error_free (error);
+      g_object_unref (stream);
       return;
     }
 
@@ -371,6 +372,8 @@
   msg = soup_message_new_from_uri (SOUP_METHOD_GET, uri);
   soup_uri_free (uri);
 
+  soup_message_body_set_accumulate (msg->response_body, FALSE);
+
   stream = soup_input_stream_new (op_backend->session_async, msg);
   g_object_unref (msg);
 
Index: daemon/soup-input-stream.c
===================================================================
--- daemon/soup-input-stream.c	(revision 2129)
+++ daemon/soup-input-stream.c	(revision 2130)
@@ -714,6 +714,7 @@
   soup_input_stream_done_io (stream);
 
   g_simple_async_result_complete (result);
+  g_object_unref (result);
 }
 
 static void
@@ -748,6 +749,7 @@
     {
       g_simple_async_result_set_op_res_gssize (result, 0);
       g_simple_async_result_complete_in_idle (result);
+      g_object_unref (result);
       return;
     }
 
@@ -756,6 +758,7 @@
       gsize nread = read_from_leftover (priv, buffer, count);
       g_simple_async_result_set_op_res_gssize (result, nread);
       g_simple_async_result_complete_in_idle (result);
+      g_object_unref (result);
       return;
     }
 
@@ -804,6 +807,7 @@
     }
 
   g_simple_async_result_complete_in_idle (result);
+  g_object_unref (result);
 }
 
 static gboolean
