--- gvfs-0.2.4_orig/client/gvfsfusedaemon.c	2008-05-27 07:54:41.000000000 +1000
+++ gvfs-0.2.4/client/gvfsfusedaemon.c	2008-07-29 16:04:33.000000000 +1000
@@ -2033,6 +2033,7 @@
   g_signal_handlers_disconnect_by_func (volume_monitor, mount_tracker_unmounted_cb, NULL);
 
   g_main_loop_unref (subthread_main_loop);
+  subthread_main_loop = NULL;
 
   g_object_unref (volume_monitor);
   volume_monitor = NULL;
@@ -2144,7 +2145,8 @@
 vfs_destroy (gpointer param)
 {
   mount_list_free ();
-  g_main_loop_quit (subthread_main_loop);
+  if (subthread_main_loop != NULL) 
+    g_main_loop_quit (subthread_main_loop);
   g_mutex_free (mount_list_mutex);
   g_object_unref (gvfs);
 }
