--- ../gvfs-1.0.3_orig/common/gmountspec.c	2008-09-14 14:05:05.000000000 +0200
+++ common/gmountspec.c	2009-03-05 20:58:57.486331890 +0100
@@ -26,6 +26,7 @@
 #include <stdlib.h>
 
 #include <glib/gi18n-lib.h>
+#include <glib/gmessages.h>
 
 #include "gdbusutils.h"
 #include "gmountspec.h"
@@ -478,6 +479,8 @@
 {
   char *canon, *start, *p, *q;
 
+  g_return_val_if_fail(path != NULL, path);
+
   if (*path != '/')
     canon = g_strconcat ("/", path, NULL);
   else
