--- grub-2.00.orig/grub-core/loader/i386/linux.c
+++ grub-2.00/grub-core/loader/i386/linux.c
@@ -946,10 +946,6 @@ grub_cmd_linux (grub_command_t cmd __att
 	    if (! buf)
 	      goto fail;
 
-	    grub_printf_ (N_("%s is deprecated. "
-			     "Use set gfxpayload=%s before "
-			     "linux command instead.\n"),
-			 argv[i], buf);
 	    err = grub_env_set ("gfxpayload", buf);
 	    grub_free (buf);
 	    if (err)
