--- grub-1.97.2.orig/util/grub.d/10_linux.in
+++ grub-1.97.2/util/grub.d/10_linux.in
@@ -49,7 +49,7 @@ menuentry "$1" {
 EOF
   prepare_grub_to_access_device ${GRUB_DEVICE_BOOT} | sed -e "s/^/\t/"
   cat << EOF
-	linux	${rel_dirname}/${basename} root=${linux_root_device_thisversion} ro $2
+	linux	${rel_dirname}/${basename} ro $2
 EOF
   if test -n "${initrd}" ; then
     cat << EOF
@@ -61,7 +61,7 @@ EOF
 EOF
 }
 
-list=`for i in /boot/vmlinu[xz]-* /vmlinu[xz]-* ; do
+list=`for i in /boot/kernel-* /boot/vmlinu[xz]-* /vmlinu[xz]-* ; do
         if grub_file_is_not_garbage "$i" ; then echo -n "$i " ; fi
       done`
 
@@ -78,6 +78,7 @@ while [ "x$list" != "x" ] ; do
   initrd=
   for i in "initrd.img-${version}" "initrd-${version}.img" \
 	   "initrd-${version}" "initrd.img-${alt_version}" \
+	   "initramfs-genkernel-${version}" "initramfs-genkernel-${alt_version}" \
 	   "initrd-${alt_version}.img" "initrd-${alt_version}"; do
     if test -e "${dirname}/${i}" ; then
       initrd="$i"
