--- /dev/null	2004-10-20 22:28:31.361557664 +0200
+++ kcontrol/configure.in.in	2004-10-10 16:37:14.000000000 +0200
@@ -0,0 +1,11 @@
+dnl Taken from kdebase/kdm/configure.in.in to create kdebase/kcontrol/configure.in.in
+dnl when kcontrol is compiled without kdm being in the source tree
+
+KDE_EXPAND_MAKEVAR(KDE_CONFDIR, kde_confdir)
+AC_DEFINE_UNQUOTED(KDE_CONFDIR, "$KDE_CONFDIR", [KDE's configuration directory])
+KDE_EXPAND_MAKEVAR(KDE_DATADIR, kde_datadir)
+AC_DEFINE_UNQUOTED(KDE_DATADIR, "$KDE_DATADIR", [KDE's static data directory])
+KDE_EXPAND_MAKEVAR(KDE_BINDIR, kde_bindir)
+AC_DEFINE_UNQUOTED(KDE_BINDIR, "$KDE_BINDIR", [KDE's binaries directory])
+AC_SUBST(KDE_BINDIR)
+
