--- src/gf_util.c.orig
+++ src/gf_util.c
@@ -612,7 +612,13 @@
 	if(!gfire_notify_init() || !p_title)
 		return;
 
+	/* the fourth argument was removed in libnotify 0.7.0 */
+#if !defined(NOTIFY_VERSION_MINOR) || (NOTIFY_VERSION_MAJOR == 0 && NOTIFY_VERSION_MINOR < 7)
 	NotifyNotification *notification = notify_notification_new(p_title, p_msg, NULL, NULL);
+#else
+	NotifyNotification *notification = notify_notification_new (p_title, p_msg, NULL);
+#endif
+
 	notify_notification_set_urgency(notification, NOTIFY_URGENCY_NORMAL);
 	notify_notification_set_timeout(notification, NOTIFY_EXPIRES_DEFAULT);
 	g_signal_connect(notification, "closed", G_CALLBACK(gfire_notify_closed_cb), NULL);
@@ -637,7 +643,11 @@
 		return;
 	}
 
+#if !defined(NOTIFY_VERSION_MINOR) || (NOTIFY_VERSION_MAJOR == 0 && NOTIFY_VERSION_MINOR < 7)
 	notification = notify_notification_new(p_title, p_msg, NULL, NULL);
+#else
+	notification = notify_notification_new (p_title, p_msg, NULL);
+#endif
 
 	// Get Buddy Icon
 	PurpleBuddyIcon *icon = purple_buddy_get_icon(p_buddy);
