--- fs/squashfs/inode.c.orig	2007-10-10 09:28:25.000000000 +0200
+++ fs/squashfs/inode.c	2007-10-10 09:29:05.000000000 +0200
@@ -25,6 +25,7 @@
 #include <linux/module.h>
 #include <linux/zlib.h>
 #include <linux/fs.h>
+#include <linux/exportfs.h>
 #include <linux/squashfs_fs_sb.h>
 #include <linux/squashfs_fs_i.h>
 #include <linux/buffer_head.h>
@@ -2309,7 +2310,7 @@ static int __init init_inodecache(void)
 	squashfs_inode_cachep = kmem_cache_create("squashfs_inode_cache",
 	     sizeof(struct squashfs_inode_info),
 	     0, SLAB_HWCACHE_ALIGN|SLAB_RECLAIM_ACCOUNT,
-	     init_once, NULL);
+	     init_once);
 	if (squashfs_inode_cachep == NULL)
 		return -ENOMEM;
 	return 0;
