diff -NrU5 wvstreams-4.5.orig/include/wvserialize.h wvstreams-4.5/include/wvserialize.h
--- wvstreams-4.5.orig/include/wvserialize.h	2008-12-07 20:03:29.000000000 +0100
+++ wvstreams-4.5/include/wvserialize.h	2008-12-07 18:15:36.000000000 +0100
@@ -346,21 +346,21 @@
 /**
  * Deserialize a WvString.  Stops at (and includes) the terminating nul
  * (zero) character.  Serialized WvStrings are guaranteed not to contain nul
  * except as the last character.
  */
-template <>
-extern WvString _wv_deserialize<WvString>(WvBuf &buf);
+//template <>
+//extern WvString _wv_deserialize<WvString>(WvBuf &buf); 
 
 
 /** Deserialize a WvBuf. */
 // FIXME: it should be possible to do this without using a class!
 template <>
 class WvDeserialize<WvBuf *>
 {
 public:
-    static inline WvBuf *go(WvBuf &buf)
+    static WvBuf *go(WvBuf &buf)
     {
 	size_t len = wv_deserialize<size_t>(buf);
 	WvBuf *outbuf = new WvInPlaceBuf(new char[len], 0, len, true);
 	outbuf->merge(buf, len);
 	return outbuf;
diff -NrU5 wvstreams-4.5.orig/uniconf/uniconf.cc wvstreams-4.5/uniconf/uniconf.cc
--- wvstreams-4.5.orig/uniconf/uniconf.cc	2008-12-07 20:03:29.000000000 +0100
+++ wvstreams-4.5/uniconf/uniconf.cc	2008-12-07 18:15:36.000000000 +0100
@@ -6,10 +6,12 @@
  */
 #include "uniconf.h"
 #include "uniconfroot.h"
 #include "uniconfgen.h"
 #include "wvstream.h"
+#include <climits>
+#include <algorithm>
 #include <assert.h>
 
 
 UniConf::UniConf(UniConfRoot *root, const UniConfKey &fullkey)
     : xroot(root), xfullkey(fullkey)
diff -NrU5 wvstreams-4.5.orig/utils/t/wvstringmask.t.cc wvstreams-4.5/utils/t/wvstringmask.t.cc
--- wvstreams-4.5.orig/utils/t/wvstringmask.t.cc	2008-12-07 18:17:16.000000000 +0100
+++ wvstreams-4.5/utils/t/wvstringmask.t.cc	2008-12-07 20:04:30.000000000 +0100
@@ -1,9 +1,10 @@
 #include "wvtest.h"
 #include "wvstringmask.h"
 
 #include <stdio.h>
+#include <climits>
 
 WVTEST_MAIN("wvstringmask")
 {
     WvStringMask a, b(""), c(' '), d("cab");
 
diff -NrU5 wvstreams-4.5.orig/utils/wvtclstring.cc wvstreams-4.5/utils/wvtclstring.cc
--- wvstreams-4.5.orig/utils/wvtclstring.cc	2008-12-07 20:03:29.000000000 +0100
+++ wvstreams-4.5/utils/wvtclstring.cc	2008-12-07 18:15:36.000000000 +0100
@@ -6,10 +6,11 @@
 #include "wvbuf.h"
 #include "wvstream.h"
 #include "wvstring.h"
 #include "wvstringmask.h"
 #include "wvtclstring.h"
+#include <climits>
 
 const WvStringMask WVTCL_NASTY_SPACES(WVTCL_NASTY_SPACES_STR);
 const WvStringMask WVTCL_NASTY_NEWLINES(WVTCL_NASTY_NEWLINES_STR);
 const WvStringMask WVTCL_SPLITCHARS(WVTCL_SPLITCHARS_STR);
 
