diff -pruN wvstreams-4.4.1.orig/utils/wvcrash.cc wvstreams-4.4.1/utils/wvcrash.cc
--- wvstreams-4.4.1.orig/utils/wvcrash.cc	2007-08-17 07:15:07.000000000 +0100
+++ wvstreams-4.4.1/utils/wvcrash.cc	2007-11-28 15:20:09.000000000 +0100
@@ -346,10 +346,10 @@ void wvcrash_add_signal(int sig)
 #if WVCRASH_USE_SIGALTSTACK
     struct sigaction act;
     
+    memset(&act,0,sizeof(act));
     act.sa_handler = wvcrash;
     sigfillset(&act.sa_mask);
     act.sa_flags = SA_ONSTACK | SA_RESTART;
-    act.sa_restorer = NULL;
     
     if (sigaction(sig, &act, NULL))
         fprintf(stderr, "Failed to setup wvcrash handler for signal %d: %s\n",
