diff -Nru wvstreams-4.4.1.orig/wvrules.mk wvstreams-4.4.1/wvrules.mk
--- wvstreams-4.4.1.orig/wvrules.mk	2008-01-20 13:00:48.000000000 +0200
+++ wvstreams-4.4.1/wvrules.mk	2008-01-20 13:25:51.000000000 +0200
@@ -248,7 +248,7 @@
 %.E: %.cc;	$(call wvcxx,$@,$<,$*,,-E)
 %.E: %.cpp;	$(call wvcxx,$@,$<,$*,,-E)
 
-%.moc: %.h;	moc -o $@ $<
+%.moc: %.h;	$(MOC) -o $@ $<
 
 %: %.o;		$(call wvlink,$@,$^) 
 %.t: %.t.o;	$(call wvlink,$@,$(call reverse,$(filter %.o,$^)) $(filter-out %.o,$^) $(LIBWVTEST))
