diff -NrU5 wvstreams-4.5.orig/configure.ac wvstreams-4.5/configure.ac
--- wvstreams-4.5.orig/configure.ac	2008-12-08 20:47:44.000000000 +0100
+++ wvstreams-4.5/configure.ac	2008-12-08 21:12:55.000000000 +0100
@@ -466,11 +466,11 @@
     AC_CACHE_CHECK([for Qt], [wv_cv_with_qt], [
         wv_cv_with_qt=no
         CPPFLAGS_save="$CPPFLAGS"
         LDFLAGS_save="$LDFLAGS"
         LIBS_save="$LIBS"
-        for wv_qtdir in $with_qt $QTDIR $QT_SEARCH_PATH; do
+        for wv_qtdir in $with_qt $QTDIR $QT_SEARCH_PATH $(pkg-config --variable=prefix qt-mt); do
             eval wv_qtdir="$wv_qtdir"
             CPPFLAGS="$CPPFLAGS_save -I$wv_qtdir/include -I$wv_qtdir/include/qt3"
             LDFLAGS="$LDFLAGS_save -L$wv_qtdir/lib"
             LIBS="$LIBS_save -lqt-mt"
             AC_TRY_LINK([#include <qstring.h>],
