diff -NrU5 wvstreams-4.5.orig/configure.ac wvstreams-4.5/configure.ac
--- wvstreams-4.5.orig/configure.ac	2008-12-07 23:22:50.000000000 +0100
+++ wvstreams-4.5/configure.ac	2008-12-08 00:52:38.000000000 +0100
@@ -346,12 +346,12 @@
 fi
 
 # dbus
 if test "$with_dbus" != "no"; then
     WV_APPEND(CPPFLAGS, -DDBUS_API_SUBJECT_TO_CHANGE -Iinclude/dbus-upstream)
-    if test "$with_dbus" = ""; then
-        DBUS_LIBDIR="$(pkg-config --libs-only-L dbus-1 | sed 's/^-L//' | awk '{print $1}')"
+    if test "$with_dbus" = "" -o "$with_dbus" = "yes"; then
+        DBUS_LIBDIR="$(pkg-config --variable=libdir dbus-1)"
 	if test -z "$DBUS_LIBDIR"; then
 	    DBUS_LIBDIR=/usr/lib  # Default path for static linking
 	fi
         WV_APPEND(CPPFLAGS, [`pkg-config --cflags dbus-1`])
         WV_APPEND(LDFLAGS, [`pkg-config --libs-only-L dbus-1`])
