--- gnome-shell-3.8.0.1.orig/data/org.gnome.shell.gschema.xml.in.in
+++ gnome-shell-3.8.0.1/data/org.gnome.shell.gschema.xml.in.in
@@ -228,7 +228,7 @@ value here is from the GsmPresenceStatus
     </key>
 
     <key name="button-layout" type="s">
-      <default>":close"</default>
+      <default>":minimize,maximize,close"</default>
       <_summary>Arrangement of buttons on the titlebar</_summary>
       <_description>
         This key overrides the key in org.gnome.desktop.wm.preferences when
