diff -Nurp gnome-shell-3.0.2.orig/data/gnome-shell.schemas gnome-shell-3.0.2/data/gnome-shell.schemas
--- gnome-shell-3.0.2.orig/data/gnome-shell.schemas	2011-03-17 16:45:49.000000000 +0100
+++ gnome-shell-3.0.2/data/gnome-shell.schemas	2011-08-05 10:43:17.747964830 +0200
@@ -22,7 +22,7 @@
         <applyto>/desktop/gnome/shell/windows/button_layout</applyto>
         <owner>gnome-shell</owner>
         <type>string</type>
-        <default>:close</default>
+        <default>:minimize,maximize,close</default>
         <locale name="C">
            <short>Arrangement of buttons on the titlebar</short>
            <long>
