--- 389-admin-1.1.11.rc1.orig/admserv/cfgstuff/httpd-2.2.conf.in
+++ 389-admin-1.1.11.rc1/admserv/cfgstuff/httpd-2.2.conf.in
@@ -141,7 +141,7 @@ LoadModule cache_module @moddir@/mod_cac
 LoadModule disk_cache_module @moddir@/mod_disk_cache.so
 LoadModule file_cache_module @moddir@/mod_file_cache.so
 LoadModule mem_cache_module @moddir@/mod_mem_cache.so
-LoadModule cgi_module @moddir@/mod_cgi.so
+LoadModule cgid_module @moddir@/mod_cgid.so
 LoadModule restartd_module     @admmoddir@/mod_restartd.so
 LoadModule nss_module         @nssmoddir@/libmodnss.so
 LoadModule admserv_module     @admmoddir@/mod_admserv.so
--- 389-admin-1.1.11.rc1.orig/admserv/cfgstuff/httpd.conf.in
+++ 389-admin-1.1.11.rc1/admserv/cfgstuff/httpd.conf.in
@@ -140,7 +140,7 @@ LoadModule cache_module @moddir@/mod_cac
 LoadModule disk_cache_module @moddir@/mod_disk_cache.so
 LoadModule file_cache_module @moddir@/mod_file_cache.so
 LoadModule mem_cache_module @moddir@/mod_mem_cache.so
-LoadModule cgi_module @moddir@/mod_cgi.so
+LoadModule cgid_module @moddir@/mod_cgid.so
 LoadModule restartd_module     @admmoddir@/mod_restartd.so
 LoadModule nss_module         @nssmoddir@/libmodnss.so
 LoadModule admserv_module     @admmoddir@/mod_admserv.so
