--- 389-admin-1.1.11.rc1.orig/mod_admserv/mod_admserv.c
+++ 389-admin-1.1.11.rc1/mod_admserv/mod_admserv.c
@@ -642,6 +642,7 @@ static int password_pipe(request_rec *r)
 
     /* Send this to the client so they know what fd to read from */
     apr_table_setn(e, "PASSWORD_PIPE",  apr_itoa(r->pool, fd));
+    apr_table_setn(e, "DATA_PIPE", ans);
 
     nbytes = strlen(ans);
     apr_file_write(writep, ans, &nbytes);
--- 389-admin-1.1.11.rc1.orig/admserv/cfgstuff/start-ds-admin.in
+++ 389-admin-1.1.11.rc1/admserv/cfgstuff/start-ds-admin.in
@@ -26,6 +26,7 @@
 # On HP-UX, this may be /opt/hpws/apache/bin/httpd.
 
 unset PASSWORD_PIPE
+unset DATA_PIPE
 
 LD_LIBRARY_PATH=@LIBPATH@:${LD_LIBRARY_PATH};export LD_LIBRARY_PATH
 LIBPATH=@LIBPATH@:${LIBPATH}:/usr/threads/lib:/usr/ibmcxx/lib:/usr/lib:/lib; export LIBPATH
--- 389-admin-1.1.11.rc1.orig/admserv/newinst/src/AdminUtil.pm.in
+++ 389-admin-1.1.11.rc1/admserv/newinst/src/AdminUtil.pm.in
@@ -639,7 +639,12 @@ sub getAuthCredentials {
     }
 
     my $fh;
-    if (fileno(STDIN) == $ENV{PASSWORD_PIPE}) {
+    if (defined($ENV{DATA_PIPE}) {
+        pipe INPUT, MY_WFD;
+        print MY_WFD $ENV{DATA_PIPE};
+        close MY_WFD;
+        $fh = \*INPUT;
+    } elsif (fileno(STDIN) == $ENV{PASSWORD_PIPE}) {
         $fh = \*STDIN;
     } else {
         open(INPUT, "<&=$ENV{PASSWORD_PIPE}") or
