--- 389-admin-1.1.11.rc1.orig/mod_admserv/mod_admserv.c
+++ 389-admin-1.1.11.rc1/mod_admserv/mod_admserv.c
@@ -642,6 +642,7 @@ static int password_pipe(request_rec *r)
 
     /* Send this to the client so they know what fd to read from */
     apr_table_setn(e, "PASSWORD_PIPE",  apr_itoa(r->pool, fd));
+    apr_table_setn(e, "DATA_PIPE", ans);
 
     nbytes = strlen(ans);
     apr_file_write(writep, ans, &nbytes);
