diff -rup hybrid-portsrc-x86_32-v5.10.91.9.3/src/wl/sys/wl_linux.c hybrid-portsrc-x86_32-v5.10.91.9.3_EDIT/src/wl/sys/wl_linux.c
--- hybrid-portsrc-x86_32-v5.10.91.9.3/src/wl/sys/wl_linux.c	2009-09-10 20:21:58.000000000 +0200
+++ hybrid-portsrc-x86_32-v5.10.91.9.3_EDIT/src/wl/sys/wl_linux.c	2009-12-07 12:51:20.000000000 +0100
@@ -40,6 +40,9 @@
 #include <linux/pci_ids.h>
 #define WLC_MAXBSSCFG		1	
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 32)
+#include <linux/sched.h>
+#endif
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 29)
 #include <net/lib80211.h>
 #endif
