diff -Naur source/src/wl/sys/wl_iw.c mod/src/wl/sys/wl_iw.c
--- source/src/wl/sys/wl_iw.c	2009-01-19 23:02:56.000000000 +0100
+++ mod/src/wl/sys/wl_iw.c	2009-02-02 11:39:05.977254751 +0100
@@ -580,6 +580,7 @@
 	range->enc_capa |= IW_ENC_CAPA_CIPHER_TKIP;
 	range->enc_capa |= IW_ENC_CAPA_CIPHER_CCMP;
 	range->enc_capa |= IW_ENC_CAPA_WPA2;
+	range->scan_capa = IW_SCAN_CAPA_ESSID;
 #endif 
 
 	return 0;

