diff -urN Retroshare-0.5.1-old/libbitdht/src/libbitdht.pro Retroshare-0.5.1-new/libbitdht/src/libbitdht.pro
--- Retroshare-0.5.1-old/libbitdht/src/libbitdht.pro	2011-02-26 21:32:01.899776679 +0100
+++ Retroshare-0.5.1-new/libbitdht/src/libbitdht.pro	2011-02-27 05:46:56.511552869 +0100
@@ -2,7 +2,7 @@
 CONFIG += staticlib
 CONFIG -= qt
 TARGET = bitdht
-QMAKE_CXXFLAGS *= -Wall -DBE_DEBUG
+QMAKE_CXXFLAGS *= -Wall -DBE_DEBUG -Wno-unused-parameter
 
 profiling {
 	QMAKE_CXXFLAGS -= -fomit-frame-pointer
diff -urN Retroshare-0.5.1-old/libretroshare/src/libretroshare.pro Retroshare-0.5.1-new/libretroshare/src/libretroshare.pro
--- Retroshare-0.5.1-old/libretroshare/src/libretroshare.pro	2011-02-26 21:32:10.514562657 +0100
+++ Retroshare-0.5.1-new/libretroshare/src/libretroshare.pro	2011-02-27 05:52:59.901555841 +0100
@@ -149,7 +149,7 @@
 	isEmpty(LIB_DIR) { LIB_DIR = $${PREFIX}/lib/ }
 
 	DESTDIR = lib
-	QMAKE_CXXFLAGS *= -Wall -D_FILE_OFFSET_BITS=64
+	QMAKE_CXXFLAGS *= -Wall -D_FILE_OFFSET_BITS=64 -Wno-unused-parameter
 	QMAKE_CC = g++
 
 	SSL_DIR = /usr/include/openssl
diff -urN Retroshare-0.5.1-old/retroshare-gui/src/RetroShare.pro Retroshare-0.5.1-new/retroshare-gui/src/RetroShare.pro
--- Retroshare-0.5.1-old/retroshare-gui/src/RetroShare.pro	2011-02-26 21:30:55.637113747 +0100
+++ Retroshare-0.5.1-new/retroshare-gui/src/RetroShare.pro	2011-02-27 06:20:42.905888012 +0100
@@ -27,7 +27,11 @@
 # Put lib dir in QMAKE_LFLAGS so it appears before -L/usr/lib
 linux-* {
 	#CONFIG += version_detail_bash_script
-	QMAKE_CXXFLAGS *= -D_FILE_OFFSET_BITS=64
+	QMAKE_CXXFLAGS *= -D_FILE_OFFSET_BITS=64 -Wno-unused-parameter
+
+	isEmpty(PREFIX) {
+		PREFIX = /usr
+	}
 
 	system(which gpgme-config >/dev/null 2>&1) {
 		INCLUDEPATH += $$system(gpgme-config --cflags | sed -e "s/-I//g")
@@ -41,6 +45,9 @@
 	LIBS += -lssl -lgpgme -lupnp -lXss -lgnome-keyring
 	DEFINES *= HAVE_XSS # for idle time, libx screensaver extensions
 	DEFINES *= UBUNTU
+
+	target.path = $${PREFIX}/bin
+	INSTALLS *= target
 }
 
 linux-g++ {
diff -urN Retroshare-0.5.1-old/retroshare-nogui/src/retroshare-nogui.pro Retroshare-0.5.1-new/retroshare-nogui/src/retroshare-nogui.pro
--- Retroshare-0.5.1-old/retroshare-nogui/src/retroshare-nogui.pro	2011-02-26 21:32:11.668400055 +0100
+++ Retroshare-0.5.1-new/retroshare-nogui/src/retroshare-nogui.pro	2011-03-14 23:57:43.084048926 +0100
@@ -1,12 +1,17 @@
 TEMPLATE = app
 TARGET = retroshare-nogui
 CONFIG += bitdht
+CONFIG -= qt
 
 ################################# Linux ##########################################
 linux-* {
 	#CONFIG += version_detail_bash_script
 	QMAKE_CXXFLAGS *= -D_FILE_OFFSET_BITS=64
 
+	isEmpty(PREFIX) {
+		PREFIX = /usr
+	}
+
 	system(which gpgme-config >/dev/null 2>&1) {
 		INCLUDEPATH += $$system(gpgme-config --cflags | sed -e "s/-I//g")
 	} else {
@@ -15,6 +20,9 @@
 
 	LIBS += ../../libretroshare/src/lib/libretroshare.a
 	LIBS += -lssl -lgpgme -lupnp -lgnome-keyring
+
+	target.path = $${PREFIX}/bin
+	INSTALLS *= target
 }
 
 linux-g++ {
