--- Makefile	1970-01-01 01:00:00.000000000 +0100
+++ Makefile	2008-02-28 19:17:15.000000000 +0100
@@ -0,0 +1,29 @@
+PREFIX=/usr
+PYTHON_SITEPACKAGES=`python -c 'from distutils.sysconfig import get_python_lib; import os; print os.path.normcase(os.path.realpath(get_python_lib()))'`
+
+all: compile
+
+clean:
+	rm -f *.pyo
+	rm -f *.pyc
+
+compile:
+	python -c "import compileall; compileall.compile_dir('.')" || die "Compile failed."
+	python -O -c "import compileall; compileall.compile_dir('.')" || die "Optimization failed."
+
+make-install-dirs: 
+	mkdir -p $(DESTDIR)$(PREFIX)/share/xfce4/pyNetworkManager/icons
+	mkdir -p $(DESTDIR)$(PREFIX)/share/xfce4/pyNetworkManager/lib
+	mkdir -p $(DESTDIR)$(PREFIX)/share/xfce4/panel-plugins
+	mkdir -p $(DESTDIR)$(PYTHON_SITEPACKAGES)/pyNM/gtk2
+	mkdir -p $(DESTDIR)$(PYTHON_SITEPACKAGES)/pyNM/sysNetworkSpeed
+
+install: make-install-dirs
+	install -m 755 isconnected.py panelxfce.py traygtk.py windowgtk.py $(DESTDIR)$(PREFIX)/share/xfce4/pyNetworkManager/lib
+	install -m 755 pyNM/*.py $(DESTDIR)$(PYTHON_SITEPACKAGES)/pyNM
+	install -m 755 pyNM/gtk2/*.py pyNM/gtk2/setup_net.glade $(DESTDIR)$(PYTHON_SITEPACKAGES)/pyNM/gtk2
+	install -m 755 pyNM/sysNetworkSpeed/*.py $(DESTDIR)$(PYTHON_SITEPACKAGES)/pyNM/sysNetworkSpeed
+	install -m 644 icons/*.png $(DESTDIR)$(PREFIX)/share/xfce4/pyNetworkManager/icons
+	install -m 644 pyNetworkManager.desktop $(DESTDIR)$(PREFIX)/share/xfce4/panel-plugins
+	echo "ICON_PATH = '$(PREFIX)/share/xfce4/pyNetworkManager/icons/'" > pyNetworkManagerPaths.py
+	install -m 644 pyNetworkManagerPaths.py $(DESTDIR)$(PREFIX)/share/xfce4/pyNetworkManager/lib
+
