diff -Nurp wis-go7007-linux-0.9.8.orig/kernel/go7007-i2c.c wis-go7007-linux-0.9.8/kernel/go7007-i2c.c
--- wis-go7007-linux-0.9.8.orig/kernel/go7007-i2c.c	2008-02-25 13:34:38.000000000 +0000
+++ wis-go7007-linux-0.9.8/kernel/go7007-i2c.c	2008-02-25 13:33:18.000000000 +0000
@@ -296,7 +296,9 @@ static struct i2c_algorithm go7007_algo 
 #endif
 	.smbus_xfer	= go7007_smbus_xfer,
 	.master_xfer	= go7007_i2c_master_xfer,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,24)
 	.algo_control	= go7007_algo_control,
+#endif
 	.functionality	= go7007_functionality,
 };
 
diff -Nurp wis-go7007-linux-0.9.8.orig/kernel/go7007-usb.c wis-go7007-linux-0.9.8/kernel/go7007-usb.c
--- wis-go7007-linux-0.9.8.orig/kernel/go7007-usb.c	2008-02-25 13:33:18.000000000 +0000
+++ wis-go7007-linux-0.9.8/kernel/go7007-usb.c	2008-02-25 13:35:29.000000000 +0000
@@ -925,7 +925,9 @@ static struct i2c_algorithm go7007_usb_a
 	.id		= I2C_ALGO_GO7007,
 #endif
 	.master_xfer	= go7007_usb_i2c_master_xfer,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,24)
 	.algo_control	= go7007_usb_algo_control,
+#endif
 	.functionality	= go7007_usb_functionality,
 };
 
