--- genkernel-3.4.41.orig/defaults/initrd.defaults
+++ genkernel-3.4.41/defaults/initrd.defaults
@@ -69,5 +69,5 @@ DEFAULT_NFSOPTIONS="ro,nolock,rsize=1024
 
 # Only sections that are in by default or those that
 # are not module groups need to be defined here...
-HWOPTS='keymap cache modules pata sata scsi usb firewire waitscan slowusb lvm dmraid mdadm fs net virtio'
-MY_HWOPTS='modules pata sata scsi slowusb usb firewire waitscan dmraid mdadm fs net iscsi crypto virtio'
+HWOPTS='keymap cache modules pata sata scsi usb firewire waitscan lvm dmraid mdadm fs net virtio'
+MY_HWOPTS='modules pata sata scsi usb firewire waitscan dmraid mdadm fs net iscsi crypto virtio'
--- genkernel-3.4.41.orig/defaults/initrd.scripts
+++ genkernel-3.4.41/defaults/initrd.scripts
@@ -416,27 +416,6 @@ prompt_user(){
 	esac
 }
 
-setup_slowusb() {
-	# This function removes unset DO_slowusb if there is no usb-storage attached.
-	# If noslowusb is set, skip this function
-	[ "${DO_slowusb}" ] || return
-
-	# Unset DO_slowusb, so we can set it again if usb-storage has something attached
-	unset DO_slowusb
-
-	local usb_storage_dir="/sys/bus/usb/drivers/usb-storage"
-	if [ ! -d "${usb_storage_dir}" ]
-	then
-		# no automated slowusb required. no usb-storage devices attached.
-		return
-	fi
-	for x in "${usb_storage_dir}"/*
-	do
-		[ -d "${x}" ]  && [ "${x}" != "${usb_storage_dir}/module" ] \
-			&& { DO_slowusb="1" ; break ; }
-	done
-}
-
 cmdline_hwopts() {
 	# Scan CMDLINE for any "doscsi" or "noscsi"-type arguments
 	local FOUND
--- genkernel-3.4.41.orig/defaults/linuxrc
+++ genkernel-3.4.41/defaults/linuxrc
@@ -326,14 +326,8 @@ mdev -s || bad_msg "mdev -s failed"
 # Apply scan delay if specified
 sdelay
 
-# Setup slow USB bits
-setup_slowusb
-
 cd /
 
-# if doslowusb is passed, pause other 10 seconds here, after mdev load
-[ "${DO_slowusb}" ] && sleep 10
-
 # Start iSCSI
 if [ -e /bin/iscsistart ]
 then
--- genkernel-3.4.41.orig/doc/genkernel.8.txt
+++ genkernel-3.4.41/doc/genkernel.8.txt
@@ -458,11 +458,6 @@ which the ramdisk scripts would recogniz
     Activate SCSI devices on bootup, necessary when SCSI support is
     compiled as modules and you're using SCSI or SATA devices.
 
-*noslowusb*::
-    By default genkernel pause for 10 seconds if it finds a attached
-    usb-storage device to give them time to initiate.
-    This option skips that pause.
-
 *keymap*='MAP'::
     Set keymap to 'MAP', e.g. *keymap*=de.
     For valid values of 'MAP' please see
