--- a/nm-applet.conf
+++ b/nm-applet.conf
@@ -8,17 +8,17 @@
 		<allow send_destination="org.freedesktop.NetworkManagerInfo"/>
                 <allow send_interface="org.freedesktop.NetworkManagerInfo"/>
 	</policy>
-	<policy at_console="true">
+	<policy group="plugdev">
 		<allow own="org.freedesktop.NetworkManagerInfo"/>
 
 		<allow send_destination="org.freedesktop.NetworkManagerInfo"/>
                 <allow send_interface="org.freedesktop.NetworkManagerInfo"/>
 	</policy>
 	<policy context="default">
-		<deny own="org.freedesktop.NetworkManagerInfo"/>
+		<allow own="org.freedesktop.NetworkManagerInfo"/>
 
-		<deny send_destination="org.freedesktop.NetworkManagerInfo"/>
-		<deny send_interface="org.freedesktop.NetworkManagerInfo"/>
+		<allow send_destination="org.freedesktop.NetworkManagerInfo"/>
+		<allow send_interface="org.freedesktop.NetworkManagerInfo"/>
 	</policy>
 </busconfig>
 
