Index: xserver-xorg-video-intel/src/Makefile.am
===================================================================
--- xserver-xorg-video-intel.orig/src/Makefile.am
+++ xserver-xorg-video-intel/src/Makefile.am
@@ -210,3 +210,9 @@
 
 uninstall-local:
 	(cd $(DESTDIR)$(intel_drv_ladir) && rm -f i810_drv.so)
+
+pcidatadir = $(datadir)/xserver-xorg/pci
+pcidata_DATA = intel.ids
+
+intel.ids : common.h
+	awk '/^#define PCI_CHIP/ {print $$3 }' $(srcdir)/common.h | sed 's/^0x/8086/' > intel.ids
