Index: kdebindings/python/pykde4/sip/plasma/applet.sip
===================================================================
--- kdebindings/python/pykde4/sip/plasma/applet.sip	(revision 1089998)
+++ kdebindings/python/pykde4/sip/plasma/applet.sip	(working copy)
@@ -214,6 +214,14 @@
     void                    unpublish ();
     bool                    isPublished () const;
     void                    runAssociatedApplication ();
+
+%ConvertToSubClassCode
+    sipType = NULL;
+
+    if (dynamic_cast<Plasma::Containment*>(sipCpp))
+        sipType = sipType_Plasma_Containment;
+%End
+
 };
 // Applet
 
