--- trunk/KDE/kdebindings/python/pykde4/tools/pykdeuic4/pykdeuic4.py	2009/12/21 19:21:54	1064783
+++ trunk/KDE/kdebindings/python/pykde4/tools/pykdeuic4/pykdeuic4.py	2009/12/21 19:22:32	1064784
@@ -39,7 +39,7 @@
     def __init__(self,string):
         original_i18n_string.__init__(self,string)
     def __str__(self):
-        return "kdecore.i18n(\"%s\")" % (self.escape(self.string),)
+        return "kdecore.i18n(\"%s\")" % (qtproxies.escape(self.string),)
 qtproxies.i18n_string = kde_i18n_string
 
 def kdeFilter():
