--- NetworkManager-0.9.0.orig/initscript/Gentoo/NetworkManager.in
+++ NetworkManager-0.9.0/initscript/Gentoo/NetworkManager.in
@@ -8,6 +8,12 @@ depend() {
 }
 
 start() {
+   if [ -f /etc/.nonm ]; then
+      einfo "Skipping NetworkManager, /etc/.nonm found or \"nonm\" bootparam passed."
+      rm /etc/.nonm
+      return 0
+   fi
+
    ebegin "Starting NetworkManager"
    start-stop-daemon --start --quiet --pidfile /var/run/NetworkManager.pid \
       --exec /usr/sbin/NetworkManager -- --pid-file /var/run/NetworkManager.pid
