--- NetworkManager-0.6.4/src/dhcp-manager/nm-dhcp-manager.c	2006-03-29 14:03:02.000000000 -0500
+++ fixed-nm-dhcp-manager.c	2007-01-23 18:33:14.000000000 -0500
@@ -604,7 +604,7 @@
 		case DHCDBD_REBIND:
 			return "rebind";
 		case DHCDBD_TIMEOUT:
-			return "timeout";
+			return "bound";
 		case DHCDBD_FAIL:
 			return "fail";
 		case DHCDBD_START:
@@ -681,14 +681,14 @@
 						}
 						break;
 
-					case DHCDBD_TIMEOUT:		/* timed out contacting DHCP server */
-						if (nm_act_request_get_stage (req) == NM_ACT_STAGE_IP_CONFIG_START)
-						{
-							nm_device_activate_schedule_stage4_ip_config_timeout (req);
-							remove_timeout (manager, req);
+					case DHCDBD_TIMEOUT:		/* timed out contacting DHCP server */		
+                        			nm_info ("Static IP override"); 
+                        			state = DHCDBD_BOUND; 
+                        			if (nm_act_request_get_stage (req) == NM_ACT_STAGE_IP_CONFIG_START) { 
+							nm_device_activate_schedule_stage4_ip_config_get (req); 
+							remove_timeout (manager, req); 
 						}
-						break;					
-
+						break;
 					case DHCDBD_FAIL:		/* all attempts to contact server timed out, sleeping */
 					case DHCDBD_ABEND:		/* dhclient exited abnormally */
 //					case DHCDBD_END:		/* dhclient exited normally */
