--- NetworkManager.orig/src/nm-device-802-11-wireless.c
+++ NetworkManager/src/nm-device-802-11-wireless.c
@@ -2924,7 +2924,7 @@ supplicant_interface_init (NMDevice80211
 	 */
 	kernel_driver = nm_device_get_driver (NM_DEVICE (self));
 	if (!strcmp (kernel_driver, "ath_pci"))
-		wpa_driver = "madwifi";
+		wpa_driver = "wext";
 	else if (!strcmp (kernel_driver, "prism54"))
 		wpa_driver = "prism54";
 	else
