--- src/NetworkManager.conf~	2010-11-03 13:04:05.000000000 +0000
+++ src/NetworkManager.conf	2010-12-17 12:12:48.391177001 +0000
@@ -57,6 +57,9 @@
                 <allow send_destination="org.freedesktop.NetworkManager"
                        send_interface="org.freedesktop.NetworkManager.VPN.Connection"/>
 
+                <allow send_destination="org.freedesktop.NetworkManager"
+                       send_interface="org.freedesktop.NetworkManager.VPN.Plugin"/>
+
                 <deny send_destination="org.freedesktop.NetworkManager"
                        send_interface="org.freedesktop.NetworkManager"
                        send_member="SetLogging"/>
@@ -73,6 +76,12 @@
                        send_interface="org.freedesktop.NetworkManager"
                        send_member="wake"/>
         </policy>
+        <policy group="plugdev">
+                <allow send_destination="org.freedesktop.NetworkManager"/>
+
+                <deny send_destination="org.freedesktop.NetworkManager"
+                      send_interface="org.freedesktop.NetworkManager.PPP"/>
+        </policy>
         <policy context="default">
                 <deny own="org.freedesktop.NetworkManager"/>
                 <deny own="org.freedesktop.NetworkManagerSystemSettings"/>
