diff -urN NetworkManager.cvs/src/backends/NetworkManagerGentoo.c NetworkManager/src/backends/NetworkManagerGentoo.c
--- NetworkManager.cvs/src/backends/NetworkManagerGentoo.c	2005-08-06 22:26:40.000000000 -0500
+++ NetworkManager/src/backends/NetworkManagerGentoo.c	2005-08-07 02:28:15.000000000 -0500
@@ -168,7 +168,9 @@
  */
 void nm_system_enable_loopback (void)
 {
-	nm_spawn_process("/etc/init.d/net.lo start");
+  /* No need to run net.lo if it is already running */
+	if (nm_spawn_process ("/etc/init.d/net.lo status") != 0)
+		nm_spawn_process("/etc/init.d/net.lo start");
 }
 
