diff -Nurp NetworkManager-0.6.5_p20070823.orig/initscript/Gentoo/NetworkManager NetworkManager-0.6.5_p20070823/initscript/Gentoo/NetworkManager
--- NetworkManager-0.6.5_p20070823.orig/initscript/Gentoo/NetworkManager.in
+++ NetworkManager-0.6.5_p20070823/initscript/Gentoo/NetworkManager.in
@@ -31,6 +31,7 @@ pidfile=/var/run/NetworkManager.pid
 
 depend() {
 	need dhcdbd hald
+	before nfsmount
 }
 
 start()
