--- src/named-manager/nm-named-manager.c.orig	2006-03-25 00:38:24.000000000 -0500
+++ src/named-manager/nm-named-manager.c	2006-03-25 00:44:15.000000000 -0500
@@ -31,6 +31,7 @@
 #include <resolv.h>
 #include <sys/socket.h>
 #include <sys/types.h>
+#include <sys/stat.h>
 #include <arpa/inet.h>
 #include <syslog.h>
 #include <glib.h>
@@ -396,6 +397,8 @@
 	if (rename (tmp_resolv_conf, RESOLV_CONF) < 0)
 		goto lose;
 	nm_system_update_dns ();
+
+	chmod(RESOLV_CONF, S_IREAD | S_IWRITE | S_IRGRP | S_IROTH);
 	return TRUE;
 
 lose:
