--- src/nm-device.c.old	2006-12-13 18:37:34.000000000 +0100
+++ src/nm-device.c	2006-12-13 18:46:13.000000000 +0100
@@ -1828,7 +1828,7 @@ nm_device_bring_up_wait (NMDevice *self,
 	args[0] = self;
 	args[1] = &err;
 	args[2] = GINT_TO_POINTER (cancelable);
-	nm_wait_for_completion (400, G_USEC_PER_SEC / 200, NULL, nm_completion_device_is_up_test, args);
+	nm_wait_for_completion (200, G_USEC_PER_SEC / 50, NULL, nm_completion_device_is_up_test, args);
 	if (err)
 		nm_info ("failed to bring up device %s", self->priv->iface);
 	return err;
