Index: src/NetworkManager.conf
===================================================================
--- src/NetworkManager.conf.orig
+++ src/NetworkManager.conf
@@ -11,7 +11,7 @@
                 <allow send_destination="org.freedesktop.NetworkManager.PPP"/>
                 <allow send_interface="org.freedesktop.NetworkManager.PPP"/>
         </policy>
-        <policy at_console="true">
+        <policy group="plugdev">
                 <allow send_destination="org.freedesktop.NetworkManager"/>
                 <allow send_interface="org.freedesktop.NetworkManager"/>
         </policy>
