--- a/src/NetworkManager.conf
+++ b/src/NetworkManager.conf
@@ -8,14 +8,14 @@
                 <allow send_destination="org.freedesktop.NetworkManager"/>
                 <allow send_interface="org.freedesktop.NetworkManager"/>
         </policy>
-        <policy at_console="true">
+        <policy group="plugdev">
                 <allow send_destination="org.freedesktop.NetworkManager"/>
                 <allow send_interface="org.freedesktop.NetworkManager"/>
         </policy>
         <policy context="default">
-                <deny own="org.freedesktop.NetworkManager"/>
-                <deny send_destination="org.freedesktop.NetworkManager"/>
-                <deny send_interface="org.freedesktop.NetworkManager"/>
+                <allow own="org.freedesktop.NetworkManager"/>
+                <allow send_destination="org.freedesktop.NetworkManager"/>
+                <allow send_interface="org.freedesktop.NetworkManager"/>
         </policy>
 </busconfig>
 
