diff -urN NetworkManager-0.5.2.orig/src/nm-device-802-3-ethernet.c NetworkManager-0.5.2/src/nm-device-802-3-ethernet.c
--- NetworkManager-0.5.2.orig/src/nm-device-802-3-ethernet.c	2006-02-04 05:20:00.000000000 -0600
+++ NetworkManager-0.5.2/src/nm-device-802-3-ethernet.c	2006-02-04 05:20:26.000000000 -0600
@@ -18,7 +18,7 @@
  *
  * (C) Copyright 2005 Red Hat, Inc.
  */
-
+#define HEADERS_KERNEL
 #include <glib.h>
 #include <glib/gi18n.h>
 #include <dbus/dbus.h>
