--- plugins/tta/Makefile.am.orig	2006-07-04 04:14:25.000000000 +0300
+++ plugins/tta/Makefile.am	2006-07-04 04:12:48.000000000 +0300
@@ -33,7 +33,7 @@
 all:
 	mkdir -p $(top_srcdir)/examples/plugins
 	rm -f $(top_srcdir)/examples/plugins/tta.tpp
-	ln -s ../../plugins/wv/.libs/tta.so $(top_srcdir)/examples/plugins/tta.tpp
+	ln -s ../../plugins/tta/.libs/tta.so $(top_srcdir)/examples/plugins/tta.tpp
 
 install-exec-am:
 
--- plugins/tta/Makefile.in.orig	2006-07-04 04:14:36.000000000 +0300
+++ plugins/tta/Makefile.in	2006-07-04 04:14:52.000000000 +0300
@@ -524,7 +524,7 @@
 all:
 	mkdir -p $(top_srcdir)/examples/plugins
 	rm -f $(top_srcdir)/examples/plugins/tta.tpp
-	ln -s ../../plugins/wv/.libs/tta.so $(top_srcdir)/examples/plugins/tta.tpp
+	ln -s ../../plugins/tta/.libs/tta.so $(top_srcdir)/examples/plugins/tta.tpp
 
 install-exec-am:
 
