diff -Nurp PyQt-x11-gpl-4.5.orig/sip/QtGui/qgraphicslinearlayout.sip PyQt-x11-gpl-4.5/sip/QtGui/qgraphicslinearlayout.sip
--- PyQt-x11-gpl-4.5.orig/sip/QtGui/qgraphicslinearlayout.sip	2009-06-05 10:50:38.000000000 +0200
+++ PyQt-x11-gpl-4.5/sip/QtGui/qgraphicslinearlayout.sip	2009-06-25 15:22:56.297020424 +0200
@@ -74,7 +74,7 @@ public:
         {
             PyObject *itmo = sipGetPyObject(itm, sipType_QGraphicsLayoutItem);
         
-            if (itmo)
+            if (itmo && ! itm->parentLayoutItem())
                 sipTransferBack(itmo);
         }
 %End
