diff -ru PyQt-x11-gpl-4.4_old/designer/pluginloader.cpp PyQt-x11-gpl-4.4/designer/pluginloader.cpp
--- PyQt-x11-gpl-4.4_old/designer/pluginloader.cpp	2008-05-20 20:12:41.000000000 +0200
+++ PyQt-x11-gpl-4.4/designer/pluginloader.cpp	2008-05-20 20:13:37.000000000 +0200
@@ -280,7 +280,7 @@
 // Return the named attribute object from the named module.
 PyObject *PyCustomWidgets::getModuleAttr(const char *module, const char *attr)
 {
-#if PY_VERSION_HEX >= 0x020500
+#if PY_VERSION_HEX >= 0x02050000
     PyObject *mod = PyImport_ImportModule(module);
 #else
     PyObject *mod = PyImport_ImportModule(const_cast<char *>(module));
@@ -292,7 +292,7 @@
         return 0;
     }
 
-#if PY_VERSION_HEX >= 0x020500
+#if PY_VERSION_HEX >= 0x02050000
     PyObject *obj = PyObject_GetAttrString(mod, attr);
 #else
     PyObject *obj = PyObject_GetAttrString(mod, const_cast<char *>(attr));
