--- a/etc.Linux/os-release	2018-08-23 14:14:09.589195025 +0200
+++ b/etc.Linux/os-release	2018-08-23 14:15:03.337358583 +0200
@@ -1,7 +1,7 @@
-NAME=Gentoo
-ID=gentoo
-PRETTY_NAME="Gentoo/Linux"
+NAME=Sabayon
+ID=sabayon
+PRETTY_NAME="Sabayon/Linux"
 ANSI_COLOR="1;32"
-HOME_URL="https://www.gentoo.org/"
-SUPPORT_URL="https://www.gentoo.org/support/"
-BUG_REPORT_URL="https://bugs.gentoo.org/"
+HOME_URL="https://www.sabayon.org/"
+SUPPORT_URL="https://forum.sabayon.org/"
+BUG_REPORT_URL="https://bugs.sabayon.org/"
