diff -NurdB evolution-data-server-1.3.7/camel/Makefile.am evolution-data-server-1.3.7-patched/camel/Makefile.am
--- evolution-data-server-1.3.7/camel/Makefile.am	2005-07-08 11:57:02.000000000 -0500
+++ evolution-data-server-1.3.7-patched/camel/Makefile.am	2005-08-17 15:40:24.000000000 -0500
@@ -271,6 +271,8 @@
 	camel-lock-helper.c 			\
 	camel-lock-helper.h
 
+camel_lock_helper_1_2_LDFLAGS = -Wl,-z,now
+
 # NOTE: This needs to be cleaned up, it shouldn't link with all this crap,
 # because it might be setuid.
 camel_lock_helper_1_2_LDADD =			\
