diff -u camel.orig/providers/imap/camel-imap-store.c camel/providers/imap/camel-imap-store.c
--- camel.orig/providers/imap/camel-imap-store.c	2005-03-03 10:30:32.785549256 -0600
+++ camel/providers/imap/camel-imap-store.c	2005-03-03 10:33:18.970285336 -0600
@@ -924,7 +924,7 @@
 		port = ssl_options[i].port;
 	} else {
 		mode = MODE_CLEAR;
-		serv = "imap";
+		serv = "imap2";
 		port = IMAP_PORT;
 	}
 	
diff -u camel.orig/providers/pop3/camel-pop3-store.c camel/providers/pop3/camel-pop3-store.c
--- camel.orig/providers/pop3/camel-pop3-store.c	2005-03-03 10:30:32.779550168 -0600
+++ camel/providers/pop3/camel-pop3-store.c	2005-03-03 10:34:40.812843376 -0600
@@ -308,7 +308,7 @@
 		port = ssl_options[i].port;
 	} else {
 		mode = MODE_CLEAR;
-		serv = "pop3";
+		serv = "pop-3";
 		port = POP3S_PORT;
 	}
 	
