diff --exclude-from=/home/dang/.diffrc -u -ruN evolution-data-server-1.7.4.orig/Makefile.am evolution-data-server-1.7.4/Makefile.am
--- evolution-data-server-1.7.4.orig/Makefile.am	2006-04-24 06:39:59.000000000 -0400
+++ evolution-data-server-1.7.4/Makefile.am	2006-07-20 18:15:50.000000000 -0400
@@ -5,8 +5,8 @@
 LIBDB = libdb
 endif
 
-SUBDIRS = $(LIBDB) win32 libedataserver servers camel addressbook calendar libedataserverui src docs art po
-DIST_SUBDIRS= win32 libedataserver servers camel addressbook calendar libedataserverui src docs art po
+SUBDIRS = $(LIBDB) win32 libedataserver camel addressbook calendar libedataserverui servers server.deps src docs art po
+DIST_SUBDIRS= win32 libedataserver camel addressbook calendar libedataserverui servers server.deps src docs art po
 
 changelogs =			\
 	ChangeLog
