diff --exclude-from=/home/dang/.diffrc -u -ruN evolution-data-server-1.7.90.1.orig/camel/camel-disco-diary.c evolution-data-server-1.7.90.1/camel/camel-disco-diary.c
--- evolution-data-server-1.7.90.1.orig/camel/camel-disco-diary.c	2006-02-16 04:36:28.000000000 -0500
+++ evolution-data-server-1.7.90.1/camel/camel-disco-diary.c	2006-08-01 22:55:43.000000000 -0400
@@ -404,6 +404,7 @@
 
 	/* Truncate the log */
 	ftruncate (fileno (diary->file), 0);
+	rewind (diary->file);
 }
 
 CamelDiscoDiary *
