--- servers/exchange/storage/exchange-account.c.orig	2006-10-13 08:07:21.000000000 -0500
+++ servers/exchange/storage/exchange-account.c	2006-10-13 08:07:57.000000000 -0500
@@ -1244,8 +1244,7 @@
 					      account->priv->identity_name,
 					      account->priv->identity_email,
 					      account->priv->source_uri,
-					      account->priv->offline_sync);
-	d(g_print ("exchange-account.c:setup_account_hierarchies:offline_sync=%d\n", account->priv->offline_sync));
+					      TRUE);
 
 	setup_hierarchy (account, hier);
 	g_free (phys_uri_prefix);
@@ -1277,7 +1276,7 @@
 					      account->priv->identity_name,
 					      account->priv->identity_email,
 					      account->priv->source_uri,
-					      account->priv->offline_sync);
+					      TRUE);
 	setup_hierarchy (account, hier);
 	g_free (phys_uri_prefix);
 
