--- system-config-date-1.9.55.orig/src/date_gui.py
+++ system-config-date-1.9.55/src/date_gui.py
@@ -36,7 +36,7 @@ import scdMainWindow
 import time
 
 import gettext
-_ = lambda x: gettext.ldgettext ("system-config-date", x)
+_ = lambda x: gettext.ldgettext ("system-config-date", x).decode("raw_unicode_escape")
 
 ##
 ## compat
--- system-config-date-1.9.55.orig/src/scdMainWindow.py
+++ system-config-date-1.9.55/src/scdMainWindow.py
@@ -41,7 +41,7 @@ from scdate.core import timezoneBackend
 import signal
 
 import gettext
-_ = lambda x: gettext.ldgettext ("system-config-date", x)
+_ = lambda x: gettext.ldgettext ("system-config-date", x).decode("raw_unicode_escape")
 
 # Initialize date and time zone backend
 dateBackend = dateBackend.dateBackend()
--- system-config-date-1.9.55.orig/src/system-config-date.py
+++ system-config-date-1.9.55/src/system-config-date.py
@@ -33,7 +33,7 @@ import signal
 import getopt
 
 import gettext
-_ = lambda x: gettext.ldgettext ("system-config-date", x)
+_ = lambda x: gettext.ldgettext ("system-config-date", x).decode("raw_unicode_escape")
 
 def showhelp ():
     sys.stderr.write (_("""Usage: system-config-date [options]
--- system-config-date-1.9.55.orig/src/timezone_map_gui.py
+++ system-config-date-1.9.55/src/timezone_map_gui.py
@@ -38,7 +38,7 @@ import random
 import sys
 
 import gettext
-_ = lambda x: gettext.ldgettext ("system-config-date", x)
+_ = lambda x: gettext.ldgettext ("system-config-date", x).decode("raw_unicode_escape")
 
 class Enum(object):
     def __init__ (self, *args):
--- system-config-date-1.9.57.orig/src/scdate/core/zonetab.py
+++ system-config-date-1.9.57/src/scdate/core/zonetab.py
@@ -29,7 +29,7 @@ import locale
 import warnings
 
 from gettext import ldgettext
-ugettext = lambda x: unicode(ldgettext("system-config-date", x), "utf-8")
+ugettext = lambda x: unicode(ldgettext("system-config-date", x), "raw_unicode_escape")
 
 __all__ = ("ZoneTabEntry", "ZoneTab")
 
