--- gsettings-desktop-schemas-3.0.1.orig/schemas/org.gnome.desktop.background.gschema.xml.in.in
+++ gsettings-desktop-schemas-3.0.1/schemas/org.gnome.desktop.background.gschema.xml.in.in
@@ -53,7 +53,7 @@
       </_description>
     </key>
     <key name="show-desktop-icons" type="b">
-      <default>false</default>
+      <default>true</default>
       <_summary>Have file manager handle the desktop</_summary>
       <_description>
         If set to true, then file manager will draw the icons on the desktop.
