--- udev-021/udev-add.c.orig	2004-03-06 18:38:37.000000000 -0500
+++ udev-021/udev-add.c	2004-03-06 18:39:05.000000000 -0500
@@ -226,10 +226,12 @@
 		if (endptr[0] == '\0')
 			uid = (uid_t) id;
 		else {
+			struct passwd *pw;
+
 			if (strncmp(dev->owner, LOCAL_USER, sizeof(LOCAL_USER)) == 0)
 				set_to_local_user(dev->owner);
 
-			struct passwd *pw = getpwnam(dev->owner);
+			pw = getpwnam(dev->owner);
 			if (pw == NULL)
 				dbg("specified user unknown '%s'", dev->owner);
 			else
