From f026c2980c7c82599427645f7cd36e18bc02536f Mon Sep 17 00:00:00 2001
From: Fabio Erculiani <lxnay@sabayon.org>
Date: Sun, 11 May 2014 21:18:37 +0100
Subject: [PATCH] devices: enable UUID= for dm-based devices in fstab

---
 blivet/devices.py | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/blivet/devices.py b/blivet/devices.py
index 137a61b..714c849 100644
--- a/blivet/devices.py
+++ b/blivet/devices.py
@@ -1820,11 +1820,6 @@ class DMDevice(StorageDevice):
         return d
 
     @property
-    def fstabSpec(self):
-        """ Return the device specifier for use in /etc/fstab. """
-        return self.path
-
-    @property
     def mapName(self):
         """ This device's device-mapper map name """
         return self.name
-- 
1.9.2

