--- a/schemas/desktop_gnome_background.schemas.in.in
+++ b/schemas/desktop_gnome_background.schemas.in.in
@@ -29,7 +29,7 @@
         <applyto>/desktop/gnome/background/picture_filename</applyto>
         <owner>gnome</owner>
         <type>string</type>
-	<default>@datadir@/pixmaps/backgrounds/gnome/background-default.jpg</default>
+	<default>@datadir@/backgrounds/kgdm.jpg</default>
         <locale name="C">
         <short>Picture Filename</short>
         <long>File to use for the background image.</long>
