diff -ur xscreensaver-5.06.orig/driver/screensaver-properties.desktop.in xscreensaver-5.06/driver/screensaver-properties.desktop.in
--- xscreensaver-5.06.orig/driver/screensaver-properties.desktop.in	2008-01-27 01:16:59.000000000 +0200
+++ xscreensaver-5.06/driver/screensaver-properties.desktop.in	2008-08-05 00:07:25.000000000 +0300
@@ -1,6 +1,6 @@
 [Desktop Entry]
 Exec=xscreensaver-demo
-Icon=xscreensaver
+Icon=/usr/share/pixmaps/xscreensaver.xpm
 Terminal=false
 _Name=Screensaver
 _Comment=Change screensaver properties
diff -ur xscreensaver-5.06.orig/driver/XScreenSaver.ad.in xscreensaver-5.06/driver/XScreenSaver.ad.in
--- xscreensaver-5.06.orig/driver/XScreenSaver.ad.in	2008-07-17 00:03:56.000000000 +0300
+++ xscreensaver-5.06/driver/XScreenSaver.ad.in	2008-08-05 00:00:15.000000000 +0300
@@ -31,36 +31,36 @@
 *mode:			random
 *timeout:		0:10:00
 *cycle:			0:10:00
-*lockTimeout:		0:00:00
+*lockTimeout:		0:20:00
 *passwdTimeout:		0:00:30
-*dpmsEnabled:		False
-*dpmsStandby:		2:00:00
-*dpmsSuspend:		2:00:00
-*dpmsOff:		4:00:00
-*grabDesktopImages:	True
+*dpmsEnabled:		True
+*dpmsStandby:		1:00:00
+*dpmsSuspend:		1:00:00
+*dpmsOff:		2:00:00
+*grabDesktopImages:	False
 *grabVideoFrames:	False
 *chooseRandomImages:	@DEFAULT_IMAGES_P@
 *imageDirectory:	@DEFAULT_IMAGE_DIRECTORY@
 *nice:			10
 *memoryLimit:		0
-*lock:			False
+*lock:			True
 *verbose:		False
 *timestamp:		True
 *fade:			True
 *unfade:		False
 *fadeSeconds:		0:00:03
 *fadeTicks:		20
-*splash:		True
+*splash:		False
 *splashDuration:	0:00:05
 *visualID:		default
 *captureStderr: 	True
 *ignoreUninstalledPrograms: False
 
-*textMode:		file
-*textLiteral:		XScreenSaver
-*textFile:		@DEFAULT_TEXT_FILE@
-*textProgram:		fortune
-*textURL:		http://www.livejournal.com/stats/latest-rss.bml
+*textMode:		literal
+*textLiteral:		Gentoo Linux
+*textFile:		/etc/gentoo-release
+*textProgram:		/usr/bin/fortune
+*textURL:		http://planet.gentoo.org/rss20.xml
 
 *overlayTextForeground:	#FFFF00
 *overlayTextBackground:	#000000
@@ -117,7 +117,10 @@
 
 ! The format used for printing the date and time in the password dialog box
 ! (see the strftime(3) manual page for details.)
-*dateFormat:		%d-%b-%y (%a); %I:%M %p
+! Preferred time representation for the current locale
+*dateFormat:    %x %X
+! Xscreensaver's default
+! *dateFormat:		%d-%b-%y (%a); %I:%M %p
 ! To show the time only:
 ! *dateFormat:		%I:%M %p
 ! For 24 hour time:
diff -ur xscreensaver-5.06.orig/hacks/barcode.c xscreensaver-5.06/hacks/barcode.c
--- xscreensaver-5.06.orig/hacks/barcode.c	2006-04-06 09:40:32.000000000 +0300
+++ xscreensaver-5.06/hacks/barcode.c	2008-08-05 00:00:21.000000000 +0300
@@ -131,7 +131,6 @@
   "children",
   "chocolate",
   "CLONE",
-  "cock",
   "constriction",
   "contrition",
   "cop",
@@ -265,7 +264,6 @@
   "punishment",
   "punk rock",
   "punk",
-  "pussy",
   "quagmire",
   "quarantine",
   "quartz",
@@ -333,8 +331,6 @@
   "vegetarian",
   "venom",
   "verifiability",
-  "viagra",
-  "vibrator",
   "victim",
   "vignette",
   "villainy",
diff -ur xscreensaver-5.06.orig/hacks/glx/glsnake.c xscreensaver-5.06/hacks/glx/glsnake.c
--- xscreensaver-5.06.orig/hacks/glx/glsnake.c	2006-05-03 22:39:21.000000000 +0300
+++ xscreensaver-5.06/hacks/glx/glsnake.c	2008-08-05 00:00:21.000000000 +0300
@@ -565,17 +565,17 @@
 	ZERO, PIN, ZERO, ZERO, ZERO, ZERO, PIN, ZERO, ZERO, ZERO, ZERO,
 	ZERO, PIN, ZERO }
     },
-    { "erect penis",     /* thanks benno */
+    { "shuffle board",     /* thanks benno */
       { PIN, ZERO, PIN, PIN, ZERO, ZERO, PIN, ZERO, ZERO, ZERO, PIN,
 	PIN, ZERO, ZERO, ZERO, RIGHT, ZERO, ZERO, ZERO, ZERO, ZERO, ZERO,
 	ZERO, ZERO }
     },
-    { "flaccid penis",
+    { "flaccid anchor",
       { PIN, ZERO, PIN, PIN, ZERO, ZERO, PIN, ZERO, ZERO, ZERO, PIN,
 	PIN, ZERO, ZERO, ZERO, RIGHT, PIN, ZERO, ZERO, ZERO, ZERO, ZERO,
 	ZERO, ZERO }
     },
-    { "vagina",
+    { "engagement ring",
       { RIGHT, ZERO, ZERO, ZERO, RIGHT, ZERO, ZERO, PIN, ZERO, ZERO,
 	LEFT, ZERO, ZERO, ZERO, LEFT, ZERO, LEFT, PIN, LEFT, PIN, RIGHT,
 	PIN, RIGHT, ZERO }
@@ -995,7 +995,7 @@
     { "Parrot",
         { ZERO, ZERO, ZERO, ZERO, RIGHT, RIGHT, ZERO, LEFT, PIN, RIGHT, ZERO, RIGHT, ZERO, RIGHT, ZERO, RIGHT, PIN, LEFT, ZERO, RIGHT, LEFT, ZERO, PIN, ZERO }
     },
-    { "Penis",
+    { "Shuttle",
         { PIN, PIN, RIGHT, ZERO, PIN, PIN, ZERO, PIN, ZERO, ZERO, RIGHT, PIN, LEFT, ZERO, ZERO, PIN, ZERO, PIN, PIN, ZERO, LEFT, PIN, PIN, ZERO }
     },
     { "PictureComingSoon",
