--- configure.ac.old	2011-06-22 12:41:19.140037770 +0200
+++ configure.ac	2011-06-22 12:43:46.939252073 +0200
@@ -245,7 +245,7 @@
 # ---- Check these flags do work by building a small irrLicht sample
 echo "checking for irrlicht libraries..."
 save_LIBS=$LIBS
-LIBS="$LIBS $irrlicht_LIBS $opengl_LIBS"
+LIBS="$LIBS $irrlicht_LIBS $opengl_LIBS -lm -lz -lX11"
 
 irrlicht_sample_built_fine="false"
 
--- src/Makefile.am.old	2011-06-22 14:15:09.992552914 +0200
+++ src/Makefile.am	2011-06-22 14:16:38.820196346 +0200
@@ -396,5 +396,6 @@
 supertuxkart_LDADD = \
         $(irrlicht_LIBS) $(fribidi_LIBS) $(bullet_LIBS) $(enet_LIBS) \
         $(opengl_LIBS) $(openal_LIBS) $(oggvorbis_LIBS) \
-        $(INTLLIBS) $(LIBCURL_LIBS) $(LIBCURL_CFLAGS)
+        $(INTLLIBS) $(LIBCURL_LIBS) $(LIBCURL_CFLAGS) \
+	-lm -lX11 -lz
 
