--- CMakeLists.txt.1	2008-08-01 07:28:01.487913000 +0100
+++ CMakeLists.txt	2009-09-06 19:52:31.850265813 +0100
@@ -1,12 +1,4 @@
-if (WITH_FEDORA)
-    add_subdirectory(about/fedora)
-    set_source_files_properties(
-        sysinfo.cpp
-        COMPILE_FLAGS -DWITH_FEDORA
-    )
-else (WITH_FEDORA)
-    add_subdirectory(about/suse)
-endif (WITH_FEDORA)
+add_subdirectory(about/sabayon)
 
 set(kio_sysinfo_SRCS
    sysinfo.cpp
