--- libproxy-0.2.3.orig/configure.ac	2008-06-06 02:26:16.000000000 +0530
+++ libproxy-0.2.3/configure.ac	2009-01-25 15:12:45.720319398 +0530
@@ -21,6 +22,7 @@
 		[PKG_CHECK_MODULES(mozjs, mozilla-js, have_mozjs=yes, have_mozjs=no)])])
 PKG_CHECK_MODULES(NetworkManager, NetworkManager, 
 					have_networkmanager=yes, have_networkmanager=no)
+PKG_CHECK_MODULES(dbus, dbus-1, have_dbus=yes, have_dbus=no)
 
 # Environmental Variable
 AC_ARG_WITH([envvar],
@@ -133,15 +135,15 @@
             [AS_HELP_STRING([--with-networkmanager],
             [build NetworkManager plugin @<:@automatic@:>@])],
             [],
-            [test x$have_networkmanager == xyes && with_networkmanager=yes])
+            [test x$have_networkmanager == xyes -a x$have_dbus == xyes && with_networkmanager=yes])
 if test x$with_networkmanager = xyes; then
-	if test x$have_networkmanager == xyes; then
-		NETWORKMANAGER_CFLAGS="$NetworkManager_CFLAGS"
-		NETWORKMANAGER_LIBS="$NetworkManager_LIBS"
+	if test x$have_networkmanager == xyes -a x$have_dbus == xyes; then
+		NETWORKMANAGER_CFLAGS="$NetworkManager_CFLAGS $dbus_CFLAGS"
+		NETWORKMANAGER_LIBS="$NetworkManager_LIBS $dbus_LIBS"
 		AC_SUBST(NETWORKMANAGER_CFLAGS)
 		AC_SUBST(NETWORKMANAGER_LIBS)
 	else
-		echo "NetworkManager plugin requires: NetworkManager!"
+		echo "NetworkManager plugin requires: NetworkManager and dbus!"
 		exit 1
 	fi
 else
