diff -Naur libproxy-0.2.3.orig/configure.ac libproxy-0.2.3/configure.ac
--- libproxy-0.2.3.orig/configure.ac	2008-06-06 02:26:16.000000000 +0530
+++ libproxy-0.2.3/configure.ac	2009-01-25 13:56:15.335319374 +0530
@@ -10,6 +10,7 @@
 AC_PROG_LIBTOOL
 AC_PROG_LN_S
 AC_PROG_MAKE_SET
+AM_PROG_CC_C_O
 
 ### Checks for libraries for plugins.
 PKG_CHECK_MODULES(x11, x11, have_x11=yes, have_x11=no)
@@ -154,8 +155,8 @@
 AC_ARG_WITH([python],
             [AS_HELP_STRING([--with-python],
             [build Python bindings @<:@automatic@:>@])],
-            [AM_PATH_PYTHON([2.5], with_python=yes)],
-            [AM_PATH_PYTHON([2.5], with_python=yes, with_python=no)])
+            [AM_PATH_PYTHON([2.5], have_python=yes, have_python=no)],
+            [have_python=yes && with_python=yes])
 AM_CONDITIONAL([WITH_PYTHON], [test x$with_python = xyes])
 
 # Java
