--- work/common/lib/modules/fglrx/build_mod/firegl_public.c
+++ work.new/common/lib/modules/fglrx/build_mod/firegl_public.c
@@ -1441,7 +1441,9 @@
 todo !!!
 #endif
 
-unsigned long ATI_API_CALL __ke__cmpxchg(volatile void *ptr, unsigned long old,
+void __cmpxchg_wrong_size(void) {}
+
+unsigned long ATI_API_CALL __ke__cmpxchg(volatile unsigned long *ptr, unsigned long old,
          unsigned long new, int size)
 {
 #ifndef __HAVE_ARCH_CMPXCHG
--- work/common/lib/modules/fglrx/build_mod/firegl_public.h
+++ work.new/common/lib/modules/fglrx/build_mod/firegl_public.h
@@ -441,7 +441,8 @@
 extern void ATI_API_CALL KCL_SIGNAL_BlockAll(int (*notifier)(void *priv), void *pPriv, __ke_sigset_t *pSigMask);
 extern void ATI_API_CALL KCL_SIGNAL_UnblockAll(void);
 
-extern unsigned long ATI_API_CALL __ke__cmpxchg(volatile void *ptr, unsigned long old,                      
+extern void __cmpxchg_wrong_size(void);
+extern unsigned long ATI_API_CALL __ke__cmpxchg(volatile unsigned long *ptr, unsigned long old,                      
                       unsigned long new, int size);
 
 #define __ke_cmpxchg(ptr,o,n)                        \
--- work/common/lib/modules/fglrx/build_mod/kcl_io.c
+++ work.new/common/lib/modules/fglrx/build_mod/kcl_io.c
@@ -37,8 +37,9 @@
  */
 
 #include <linux/version.h>
 #include <linux/autoconf.h>
 #include <linux/poll.h>
+#include <linux/signal.h>
 #include <asm/io.h>
 
 #include "kcl_config.h"
