diff -ur kupfer-206.original/work//kupfer-v206/kupfer/plugin/core/alternatives.py kupfer-206/work//kupfer-v206/kupfer/plugin/core/alternatives.py
--- kupfer-206.original/work//kupfer-v206/kupfer/plugin/core/alternatives.py	2011-04-14 16:00:04.000000000 -0400
+++ kupfer-206/work//kupfer-v206/kupfer/plugin/core/alternatives.py	2011-05-24 11:32:45.173788075 -0400
@@ -20,9 +20,9 @@
 	plugin_support.register_alternative(__name__, 'terminal', 'xfce4-terminal',
 		**{
 			'name': _("XFCE Terminal"),
-			'argv': ['xfce4-terminal'],
+			'argv': ['Terminal'],
 			'exearg': '-x',
-			'desktopid': "xfce4-terminal.desktop",
+			'desktopid': "Terminal.desktop",
 			'startup_notify': True,
 		})
 
