--- amarok/src/mediadevice/daap/mongrel/http11/Makefile.am.orig	2006-11-02 23:22:08.000000000 -0800
+++ amarok/src/mediadevice/daap/mongrel/http11/Makefile.am	2006-11-02 23:22:22.000000000 -0800
@@ -4,7 +4,7 @@
 
 amarokrubylib_DATA = http11.rb
 
-INCLUDES = -I$(ruby_includes) -Dinline=__inline__
+INCLUDES = -I$(ruby_includes) -Dinline=__inline__ -Dasm=__asm__
 
 libhttp11_la_LDFLAGS = \
 	-shared \
