diff -urN amarok-2.3.1/src/core-impl/collections/audiocd/FormatSelectionDialog.ui amarok-2.3.1-new/src/core-impl/collections/audiocd/FormatSelectionDialog.ui
--- amarok-2.3.1/src/core-impl/collections/audiocd/FormatSelectionDialog.ui	2010-05-27 18:37:21.000000000 +0000
+++ amarok-2.3.1-new/src/core-impl/collections/audiocd/FormatSelectionDialog.ui	2010-06-11 18:23:00.157305026 +0000
@@ -13,9 +13,6 @@
   <property name="windowTitle">
    <string>Amarok</string>
   </property>
-  <property name="accessibleName">
-   <string/>
-  </property>
   <layout class="QGridLayout" name="gridLayout_2">
    <item row="0" column="0" colspan="2">
     <widget class="QGroupBox" name="groupBox">
