Index: amarok-1.4.5/amarok/src/magnatunebrowser/magnatunealbumdownloader.cpp
===================================================================
--- amarok-1.4.5/amarok/src/magnatunebrowser/magnatunealbumdownloader.cpp	(revision 633568)
+++ amarok-1.4.5/amarok/src/magnatunebrowser/magnatunealbumdownloader.cpp	(revision 633677)
@@ -89,8 +89,8 @@
 
     //ok, now we have the .zip file downloaded. All we need is to unpack it to the desired location and add it to the collection.
 
-    QString unzipString = "unzip \""+m_tempDir.name() + m_currentAlbumFileName + "\" -d \"" + m_currentAlbumUnpackLocation + "\" &";
+    QString unzipString = "unzip "+ KProcess::quote( m_tempDir.name() + m_currentAlbumFileName) + " -d " +KProcess::quote( m_currentAlbumUnpackLocation ) + " &";
 
     debug() << "unpacking: " << unzipString << endl;
 
     system( unzipString.ascii() );
