diff -ru linux-wlan-ng-0.2.8/src/p80211/p80211conv.c linux-wlan-ng-0.2.8.devel/src/p80211/p80211conv.c
--- linux-wlan-ng-0.2.8/src/p80211/p80211conv.c	2007-03-19 16:37:00.000000000 +0100
+++ linux-wlan-ng-0.2.8.devel/src/p80211/p80211conv.c	2007-06-05 18:57:49.234754000 +0200
@@ -498,7 +498,9 @@
 	}
 
 	skb->protocol = eth_type_trans(skb, netdev);
-	skb->mac.raw = (unsigned char *) e_hdr; /* new MAC header */
+	/* XXX: I *think* this is right. -jeffm */
+	skb_reset_mac_header(skb);
+//	skb->mac.raw = (unsigned char *) e_hdr; /* new MAC header */
 
         /* jkriegl: process signal and noise as set in hfa384x_int_rx() */
 	/* jkriegl: only process signal/noise if requested by iwspy */
diff -ru linux-wlan-ng-0.2.8/src/p80211/p80211netdev.c linux-wlan-ng-0.2.8.devel/src/p80211/p80211netdev.c
--- linux-wlan-ng-0.2.8/src/p80211/p80211netdev.c	2007-03-19 16:37:00.000000000 +0100
+++ linux-wlan-ng-0.2.8.devel/src/p80211/p80211netdev.c	2007-06-05 19:03:35.952167000 +0200
@@ -389,7 +389,7 @@
 
 				/* set up various data fields */
 				skb->dev = dev;
-				skb->mac.raw = skb->data ;
+				skb_reset_mac_header(skb);
 				skb->ip_summed = CHECKSUM_NONE;
 				skb->pkt_type = PACKET_OTHERHOST;
 				skb->protocol = htons(ETH_P_80211_RAW); 
diff -ru linux-wlan-ng-0.2.8/src/prism2/driver/hfa384x.c linux-wlan-ng-0.2.8.devel/src/prism2/driver/hfa384x.c
--- linux-wlan-ng-0.2.8/src/prism2/driver/hfa384x.c	2007-03-19 16:37:00.000000000 +0100
+++ linux-wlan-ng-0.2.8.devel/src/prism2/driver/hfa384x.c	2007-06-05 19:04:30.845216000 +0200
@@ -3645,7 +3645,7 @@
 		/* the prism2 cards don't return the FCS */
 		datap = skb_put(skb, WLAN_CRC_LEN);
 		memset (datap, 0xff, WLAN_CRC_LEN);
-		skb->mac.raw = skb->data;
+		skb_reset_mac_header(skb);
 
 		/* Attach the rxmeta, set some stuff */
 		p80211skb_rxmeta_attach(wlandev, skb);
diff -ru linux-wlan-ng-0.2.8/src/prism2/driver/hfa384x_usb.c linux-wlan-ng-0.2.8.devel/src/prism2/driver/hfa384x_usb.c
--- linux-wlan-ng-0.2.8/src/prism2/driver/hfa384x_usb.c	2007-03-19 16:37:00.000000000 +0100
+++ linux-wlan-ng-0.2.8.devel/src/prism2/driver/hfa384x_usb.c	2007-06-05 19:04:46.641407000 +0200
@@ -4255,7 +4255,7 @@
 		/* The prism2 series does not return the CRC */
 		memset(skb_put(skb, WLAN_CRC_LEN), 0xff, WLAN_CRC_LEN);
 
-		skb->mac.raw = skb->data;
+		skb_reset_mac_header(skb);
 
 		/* Attach the rxmeta, set some stuff */
 		p80211skb_rxmeta_attach(wlandev, skb);
diff -ru linux-wlan-ng-0.2.8/src/prism2/driver/prism2_pci.c linux-wlan-ng-0.2.8.devel/src/prism2/driver/prism2_pci.c
--- linux-wlan-ng-0.2.8/src/prism2/driver/prism2_pci.c	2007-03-19 16:37:00.000000000 +0100
+++ linux-wlan-ng-0.2.8.devel/src/prism2/driver/prism2_pci.c	2007-06-05 19:33:15.406928000 +0200
@@ -275,7 +275,7 @@
 static int __init prism2pci_init(void)
 {
         WLAN_LOG_NOTICE("%s Loaded\n", version);
-	return pci_module_init(&prism2_pci_drv_id);
+	return pci_register_driver(&prism2_pci_drv_id);
 };
 
 static void __exit prism2pci_cleanup(void)
Only in linux-wlan-ng-0.2.8.devel/src/prism2/driver: prism2_pci.c~
diff -ru linux-wlan-ng-0.2.8/src/prism2/driver/prism2_plx.c linux-wlan-ng-0.2.8.devel/src/prism2/driver/prism2_plx.c
--- linux-wlan-ng-0.2.8/src/prism2/driver/prism2_plx.c	2007-03-19 16:37:00.000000000 +0100
+++ linux-wlan-ng-0.2.8.devel/src/prism2/driver/prism2_plx.c	2007-06-05 19:33:12.758581000 +0200
@@ -420,7 +420,7 @@
 static int __init prism2plx_init(void)
 {
         WLAN_LOG_NOTICE("%s Loaded\n", version);
-	return pci_module_init(&prism2_plx_drv_id);
+	return pci_register_driver(&prism2_plx_drv_id);
 };
 
 static void __exit prism2plx_cleanup(void)
Only in linux-wlan-ng-0.2.8.devel/src/prism2/driver: prism2_plx.c~
