Fix lirc building in sandbox

From: Daniel Drake <dsd@gentoo.org>

Index: linux-wlan-ng-0.2.5/scripts/Makefile
===================================================================
--- linux-wlan-ng-0.2.5.orig/scripts/Makefile
+++ linux-wlan-ng-0.2.5/scripts/Makefile
@@ -27,7 +27,7 @@ get_version:
 	$(CC) -I$(KERNEL_SOURCE)/include get_version.c -o get_version
 
 get_options:
-	$(MAKE) -f $(CURR_DIR)/Makefile.get_options -C $(KERNEL_SOURCE) get_version_target
+	$(MAKE) -f $(CURR_DIR)/Makefile.get_options -C $(KERNEL_SOURCE) get_version_target M=$(shell pwd)
 
 clean:
 	$(RM) get_version make.opts 
