diff -Nurp hydrogen-20070428.orig/configure hydrogen-20070428/configure
--- hydrogen-20070428.orig/configure	2006-10-12 12:02:21.000000000 +0200
+++ hydrogen-20070428/configure	2007-05-06 13:27:19.000000000 +0200
@@ -33,7 +33,7 @@ if test "x$QTDIR" = "x"; then
 			/usr \
 			; do
 		if test "x$ac_qtlib" = "x"; then
-			if test -f $X/lib/libQtCore.a -o -f $X/lib/libQtCore.so -o -d $X/lib/QtCore.framework -o -f $X/lib/libQtCore.dylib; then
+			if test -f $X/libQtCore.a -o -f $X/libQtCore.so -o -d $X/QtCore.framework -o -f $X/libQtCore.dylib; then
 				ac_qtlib=$X
 				QTDIR=$X
 				echo "   |-> searching in $X [ OK ]"
@@ -47,7 +47,7 @@ else
 	echo ""
 	ac_qtlib=""
 	if test "x$ac_qtlib" = "x"; then
-		if test -f $QTDIR/lib/libQtCore.so -o -f $QTDIR/lib/libQtCore.a -o -f $QTDIR/lib/libQtCore_debug.so -o -f $QTDIR/lib/libQtCore_debug.a -o -d $QTDIR/lib/QtCore.framework -o -f $QTDIR/lib/libQtCore.dylib; then
+		if test -f $QTDIR/libQtCore.so -o -f $QTDIR/libQtCore.a -o -f $QTDIR/libQtCore_debug.so -o -f $QTDIR/libQtCore_debug.a -o -d $QTDIR/QtCore.framework -o -f $QTDIR/libQtCore.dylib; then
 			ac_qtlib=$QTDIR
 			echo " * Searching for QT library in $QTDIR [ OK ]"
 		else
@@ -67,9 +67,9 @@ prefix=/usr/local
 
 echo ""
 echo " * Creating all.mak..."
-#$QTDIR/bin/qmake -o all.mak all.pro
+#qmake -o all.mak all.pro
 
-$QTDIR/bin/qmake all.pro \
+qmake all.pro \
 "prefix=$prefix"
 
 rm -f config.h
