Index: qt/qtr.pro
===================================================================
--- qt/qtr.pro	(revision 12449)
+++ qt/qtr.pro	(working copy)
@@ -19,7 +19,9 @@
 INCLUDEPATH = $${EVENT_TOP}/include $${INCLUDEPATH}
 INCLUDEPATH += $${TRANSMISSION_TOP}
 LIBS += $${TRANSMISSION_TOP}/libtransmission/libtransmission.a
-LIBS += $${TRANSMISSION_TOP}/third-party/libutp/libutp.a
+exists( $${TRANSMISSION_TOP}/third-party/libutp/libutp.a ) { 
+    LIBS += $${TRANSMISSION_TOP}/third-party/libutp/libutp.a
+}
 LIBS += $${TRANSMISSION_TOP}/third-party/dht/libdht.a
 LIBS += $${TRANSMISSION_TOP}/third-party/miniupnp/libminiupnp.a
 LIBS += $${TRANSMISSION_TOP}/third-party/libnatpmp/libnatpmp.a
