diff -Nura oregano-0.60.0/SConstruct oregano-0.60.0-nomime/SConstruct
--- oregano-0.60.0/SConstruct	2006-09-30 01:47:23.000000000 -0300
+++ oregano-0.60.0-nomime/SConstruct	2006-10-05 16:02:50.000000000 -0300
@@ -120,7 +120,3 @@
 CEnv.Alias ('install', CEnv.Install (os.path.join (CEnv['INSTALL_DIR'], 'share/mime/packages'), Split('oregano.xml')))
 CEnv.Alias ('install', CEnv.Install (os.path.join (CEnv['INSTALL_DIR'], 'share/mime-info'), Split('oregano.mime oregano.keys')))
 CEnv.Alias ('install', CEnv.Install (os.path.join (CEnv['INSTALL_DIR'], 'share/applications'), Split('oregano.desktop')))
-
-# Update mime database #
-if CEnv['RunUpdateMimeDatabase'] and ('install' in COMMAND_LINE_TARGETS):
-	CEnv.Alias ('install', CEnv.Command ('update-mime-database', 'oregano.xml', "update-mime-database "+os.path.join (CEnv['INSTALL_DIR'], 'share/mime')))
