--- src/auth.c	21 May 2007 15:19:41 -0000	1.49
+++ src/auth.c	12 Jun 2007 18:54:15 -0000
@@ -449,12 +449,12 @@
    */
   m = pr_module_get("mod_auth_pam.c");
   if (m) {
-    pr_trace_msg(trace_channel, 4,
-      "using module 'mod_auth_pam.c' to authenticate user '%s'", name);
-
     mr = dispatch_auth(cmd, "auth", &m);
 
     if (MODRET_ISHANDLED(mr)) {
+      pr_trace_msg(trace_channel, 4,
+        "module 'mod_auth_pam.c' used for authenticating user '%s'", name);
+
       res = MODRET_HASDATA(mr) ? PR_AUTH_RFC2228_OK : PR_AUTH_OK;
 
       if (cmd->tmp_pool) {
@@ -512,12 +512,12 @@
    */
   m = pr_module_get("mod_auth_pam.c");
   if (m) {
-    pr_trace_msg(trace_channel, 4,
-      "using module 'mod_auth_pam.c' to authenticate user '%s'", name);
-
-    mr = dispatch_auth(cmd, "auth", &m);
+    mr = dispatch_auth(cmd, "check", &m);
 
     if (MODRET_ISHANDLED(mr)) {
+      pr_trace_msg(trace_channel, 4,
+        "module 'mod_auth_pam.c' used for authenticating user '%s'", name);
+
       res = MODRET_HASDATA(mr) ? PR_AUTH_RFC2228_OK : PR_AUTH_OK;
 
       if (cmd->tmp_pool) {
