--- configure.in.orig	2009-02-24 21:58:08.000000000 +0100
+++ configure.in	2009-02-24 22:21:20.000000000 +0100
@@ -54,8 +54,8 @@
 
 dnl AC_PROG_LIBTOOL relies on this variable
 top_builddir=.
-AC_LIBTOOL_DLOPEN([lib/libltdl])
-AC_LIBLTDL_CONVENIENCE([lib/libltdl])
+AC_LIBTOOL_DLOPEN
+AC_WITH_LTDL
 
 dnl Checks for programs.
 AC_PROG_CC
@@ -485,18 +485,12 @@
       ac_core_modules="$ac_core_modules mod_dso.o"
       ac_build_core_modules="$ac_build_core_modules modules/mod_dso.o"
 
-      MAIN_LDFLAGS="-L\$(top_srcdir)/lib/libltdl -dlopen self -export-dynamic"
+      MAIN_LDFLAGS="-dlopen self -export-dynamic"
       MAIN_LIBS="\$(LIBLTDL)"
 
       MODULE_LDFLAGS="-avoid-version -export-dynamic -module"
 
-      INSTALL_DEPS="install-libltdl"
-      LIB_DEPS="libltdl"
-      MODULE_DEPS="libltdl"
       AC_DEFINE(PR_USE_DSO, 1, [Define if using DSO support.])
-
-      dnl Run configure scripts in subdirectories
-      AC_CONFIG_SUBDIRS(lib/libltdl)
     fi
   ])
 
--- modules/Makefile.in.orig	2009-02-24 22:34:00.000000000 +0100
+++ modules/Makefile.in	2009-02-24 22:34:13.000000000 +0100
@@ -229,7 +229,6 @@
 mod_dso.o: ../include/trace.h ../include/encode.h ../include/compat.h
 mod_dso.o: ../include/proctitle.h ../include/pidfile.h ../include/env.h
 mod_dso.o: ../include/pr-syslog.h ../include/mod_ctrls.h
-mod_dso.o: ../lib/libltdl/ltdl.h
 mod_facl.o: ../include/conf.h ../include/version.h ../config.h
 mod_facl.o: ../include/default_paths.h ../include/options.h ../include/pool.h
 mod_facl.o: ../include/str.h ../include/regexp.h ../include/table.h
--- modules/mod_dso.c.orig	2009-02-24 22:32:50.000000000 +0100
+++ modules/mod_dso.c	2009-02-24 22:33:03.000000000 +0100
@@ -31,8 +31,7 @@
 #include "conf.h"
 #include "mod_ctrls.h"
 
-/* Make sure we use the libltdl shipped with proftpd, not the system libltdl. */
-#include "lib/libltdl/ltdl.h"
+#include <ltdl.h>
 
 #define MOD_DSO_VERSION		"mod_dso/0.4"
 
