diff -Nurp trunk.orig/configure.ac trunk/configure.ac
--- trunk.orig/configure.ac	2007-05-15 23:44:01.000000000 +0200
+++ trunk/configure.ac	2007-05-15 23:41:58.000000000 +0200
@@ -148,7 +148,7 @@ AC_SUBST([DRI_CFLAGS])
 AC_SUBST([XORG_CFLAGS])
 AC_SUBST([moduledir])
 
-DRIVER_NAME=via
+DRIVER_NAME=openchrome
 AC_SUBST([DRIVER_NAME])
 
 AC_DEFINE(X_USE_LINEARFB,1,[Compatibility define for older Xen])
diff -Nurp trunk.orig/man/openchrome.man trunk/man/openchrome.man
--- trunk.orig/man/openchrome.man	1970-01-01 01:00:00.000000000 +0100
+++ trunk/man/openchrome.man	2007-05-15 23:41:58.000000000 +0200
@@ -0,0 +1,230 @@
+.\" Shorthand for double quote that works everywhere,
+.\" also within other double quotes:
+.ds q \N'34'
+.TH VIA __drivermansuffix__ __vendorversion__
+.SH NAME
+via \- video driver for VIA Unichromes
+.SH SYNOPSIS
+.nf
+.B "Section \*qDevice\*q"
+.BI "  Identifier \*q" string \*q
+.B  "  Driver \*qvia\*q"
+\ \ ...
+.B EndSection
+.fi
+
+.SH DESCRIPTION
+.B via
+is an __xservername__ driver for VIA chipsets that have  an integrated
+Unichrome graphics engine.
+.PP
+The
+.B via
+driver supports the CLE266, KM/N400, K8M/N800, PM/N800 and CN400 chipsets
+from VIA, including 2D acceleration and the Xv video overlay extensions.
+Flat panel, TV, and VGA outputs are supported, depending on the hardware
+configuration.
+.PP
+3D direct rendering is available using experimental drivers from Mesa
+(www.mesa3d.org).  There is also an XvMC client library for hardware
+acceleration of MPEG1/MPEG2 decoding (available on the CLE266,
+PM/N800, K8M/N800, and CN400 chipsets) that uses the Direct Rendering
+Infrastructure (DRI).  The XvMC client library implements a non-standard
+"VLD" extension to the XvMC standard.  The current Direct Rendering
+Manager (DRM) kernel module is available at dri.sourceforge.net.
+.PP
+The driver supports free modes for Unichrome Pros (K8M/N800, PM/N800, and
+CN400).  For plain Unichromes (CLE266, KM/N400), it currently supports
+only a limited number of dotclocks, so if you are using X modelines you
+must make sure that the dotclock is one of those supported.  Supported
+dotclocks on plain Unichromes are currently (in MHz): 25.2, 25.312,
+26.591, 31.5, 31.704, 32.663, 33.750, 35.5, 36.0, 39.822, 40.0, 41.164,
+46.981, 49.5, 50.0, 56.3, 57.284, 64.995, 65.0, 65.028, 74.480,
+75.0, 78.8, 81.613, 94.5, 108.0, 108.28, 122.0, 122.726, 135.0,
+148.5, 155.8, 157.5, 161.793, 162.0, 175.5, 189.0, 202.5, 204.8,
+218.3, 229.5.  On top of this, bandwidth restrictions apply for both
+Unichromes and Unichrome Pros.
+.PP
+.SH CONFIGURATION DETAILS
+Please refer to __xconfigfile__(__filemansuffix__) for general configuration
+details.  This section only covers configuration details specific to this
+driver.
+.PP
+The following driver
+.B options
+are supported:
+.TP
+.BI "Option \*qAccelMethod\*q  \*q" string \*q
+The driver supports "XAA" and "EXA" acceleration methods.  The default
+method is XAA, since EXA is still experimental.  Contrary to XAA, EXA
+implements acceleration for screen uploads and downlads (if DRI is
+enabled) and for the Render/Composite extension.
+.TP
+.BI "Option \*qActiveDevice\*q  \*q" string \*q
+Specifies the active device combination.  Any string containing "CRT",
+"LCD", "DFP", "TV" should be possible. "CRT" represents anything that
+is connected to the VGA port, "LCD" and "DFP" are for laptop panels
+(not TFT screens attached to the VGA port), "TV" is self-explanatory.
+The default is to use what is detected.  The driver is currently unable
+to use LCD and TV simultaneously, and will favour the LCD.
+.TP
+.BI "Option \*qAGPMem\*q  \*q" integer \*q
+Sets the amount of AGP memory that is allocated at X server startup.
+The allocated memory will be "integer" kB.  This AGP memory is used for
+the AGP command buffer (if option "EnableAGPDMA" is set to "true"), for
+DRI textures, and for the EXA scratch area.  The driver will allocate at
+least one system page of AGP memory and, if the AGP command buffer is
+used, at least 2MB + one system page.  If there is no room for the EXA
+scratch area in AGP space, it will be allocated from VRAM.  If there is
+no room for DRI textures, they will be allocated from the DRI part of
+VRAM (see the option "MaxDRIMem").  The default amount of AGP is
+32768kB.  Note that the AGP aperture set in the BIOS must be able
+to accomodate the amount of AGP memory specified here.  Otherwise no
+AGP memory will be available.  It is safe to set a very large AGP
+aperture in the BIOS.
+.TP
+.BI "Option \*qCenter\*q  \*q" boolean \*q
+Enables or disables image centering on DVI displays.
+.TP
+.BI "Option \*qDisableIRQ\*q  \*q" boolean \*q
+Disables the Vblank IRQ.  This is a workaround for some mainboards that
+have problems with IRQs from the Unichrome engine.  With IRQs disabled,
+DRI clients have no way to synchronize drawing to Vblank.
+.TP
+.BI "Option \*qDisableVQ\*q  \*q" boolean \*q
+Disables the use of VQ.  VQ is enabled by default.
+.TP
+.BI "Option \*qEnableAGPDMA\*q  \*q" boolean \*q
+Enables the AGP DMA functionality in DRM.  This requires that DRI is enabled
+and will force 2D and 3D acceleration to use AGP DMA.  The XvMC DRI client
+will also make use of this on the CLE266 to consume much less CPU. 
+.TP
+.BI "Option \*qExaNoComposite\*q  \*q" boolean \*q
+If EXA is enabled (using the option "AccelMethod"), this option enables
+or disables acceleration of compositing.  Since EXA, and in particular its
+composite acceleration, is still experimental, this is a way to disable
+a misbehaving composite acceleration.
+.TP
+.BI "Option \*qExaScratchSize\*q  \*q" integer \*q
+Sets the size of the EXA scratch area to "integer" kB.  This area is
+used by EXA as a last place to look for available space for pixmaps.
+Too little space will slow compositing down.  This option should be set
+to the size of the largest pixmap used.  If you have a screen width of
+over 1024 pixels and use 24 bpp, set this to 8192.  Otherwise you can
+leave this at the default 4096.  The space will be allocated from AGP
+memory if available, otherwise from VRAM.
+.TP
+.BI "Option \*qHWCursor\*q  \*q" boolean \*q
+Enables or disables the use of hardware cursors.  The default is enabled.
+.TP
+.BI "Option \*qLCDDualEdge\*q  \*q" boolean \*q
+Enables or disables the use of dual-edge mode to set the LCD.
+.TP
+.BI "Option \*qMaxDRIMem\*q  \*q" integer \*q
+Sets the maximum amount of VRAM memory allocated for DRI clients to
+"integer" kB.  Normally DRI clients  get half the available VRAM size,
+but in some cases it may make sense to limit this amount.  For example,
+if you are using a composite manager and you want to give as much memory
+as possible to the EXA pixmap storage area.
+.TP
+.BI "Option \*qMigrationHeuristic\*q  \*q" string \*q
+Sets the heuristic for EXA pixmap migration.  This is an EXA core
+option, and on Xorg xserver versions after 1.1.0 this defaults to
+"smart".  The Openchrome driver performs best with "greedy", so you
+should really add this option to your configuration file.  The third
+possibility is "always", which might become more useful in the future.
+.TP
+.BI "Option \*qNoAccel\*q  \*q" boolean \*q
+Disables or enables acceleration.  Default: acceleration is enabled.
+.TP
+.BI "Option \*qNoAGPFor2D\*q  \*q" boolean \*q
+With this option set, 2D acceleration will not use AGP DMA even if
+it is enabled.
+.TP
+.BI "Option \*qNoXVDMA\*q  \*q" boolean \*q
+If DRI is enabled, Xv normally uses PCI DMA to transfer video images
+from system to frame-buffer memory.  This is somewhat slower than
+direct copies due to the limitations of the PCI bus, but on the other
+hand it decreases CPU usage significantly, particularly on computers
+with fast processors.  Some video players are buggy and will display
+rendering artifacts when PCI DMA is used.  If you experience this,
+or don't want your PCI bus to be stressed with Xv images, set this
+option to "true".  This option has no effect when DRI is not enabled.
+.TP
+.BI "Option \*qPanelSize\*q  \*q" string \*q
+Specifies the size (width x height) of the LCD panel attached to the
+system.  The sizes 640x480, 800x600, 1024x768, 1280x1024, and 1400x1050
+are supported.
+.TP
+.BI "Option \*qRotate\*q  \*q" string \*q
+Rotates the display either clockwise ("CW") or counterclockwise ("CCW").
+Rotation is only supported unaccelerated.
+.TP
+.BI "Option \*qShadowFB\*q  \*q" boolean \*q
+Uses a shadow frame buffer.  This is required when rotating the display,
+but otherwise defaults to disabled.
+.TP
+.BI "Option \*qSWCursor\*q  \*q" boolean \*q
+Enables or disables the use of a software cursor.  The default is disabled.
+.TP
+.BI "Option \*qTVDeflicker\*q  \*q" integer \*q
+Specifies the deflicker setting for TV output.  Valid values are "0", "1",
+and "2".  0) No deflicker, 1) 1:1:1 deflicker, 2) 1:2:1 deflicker.
+.TP
+.BI "Option \*qTVDotCrawl\*q  \*q" boolean \*q
+Enables or disables dotcrawl.
+.TP
+.BI "Option \*qTVOutput\*q  \*q" string \*q
+Specifies which TV output to use.  The driver supports "S-Video",
+"Composite", "SC", "RGB" and "YCbCr" outputs.  Note that on some
+EPIA boards the compositer-video port is shared with audio-out and
+is selected via a jumper.
+.TP
+.BI "Option \*qTVType\*q  \*q" string \*q
+Specifies TV output format.  The driver currently supports "NTSC" and
+"PAL" timings only.
+.TP
+.BI "Option \*qVBEModes\*q  \*q" boolean \*q
+Uses the VBE BIOS calls to set the display mode.  This mimics the
+behaviour of the vesa video driver but still provides acceleration and
+other features.  This option may be used if your hardware works with
+the vesa driver but not with the Openchrome driver.  It may not work
+on 64-bit systems.  Using "VBEModes" may speed up driver acceleration
+significantly due to a more aggressive hardware setting, particularly
+on systems with low memory bandwidth.  Your refresh rate may be limited
+to 60 Hz on some systems.
+.TP
+.BI "Option \*qVBESaveRestore\*q  \*q" boolean \*q
+Uses the VBE BIOS calls to save and restore the display state when the
+X server is launched.  This can be extremely slow on some hardware, and
+the system may appear to have locked for 10 seconds or so.  The default
+is to use the driver builtin function.  This option only works if option
+"VBEModes" is enabled.
+.TP
+.BI "Option \*qVideoRAM\*q  \*q" integer \*q
+Overrides the VideoRAM autodetection.  This should never be needed.
+.PP 
+.SH "TV ENCODERS"
+Unichromes tend to be paired with several different TV encoders.
+.TP
+.BI "VIA Technologies VT1621"
+Still untested, as no combination with a Unichrome is known or available.
+Supports the following normal modes: "640x480" and "800x600".  Use
+"640x480Over" and "800x600Over" for vertical overscan.  These modes
+are made available by the driver; modelines provided in __xconfigfile__
+will be ignored.
+.TP
+.BI "VIA Technologies VT1622, VT1622A, VT1623"
+Supports the following modes: "640x480", "800x600", "1024x768",
+"848x480", "720x480" (NTSC only) and "720x576" (PAL only).  Use
+"640x480Over", "800x600Over", "1024x768Over", "848x480Over",
+"720x480Over" (NTSC) and "720x576Over" (PAL) for vertical overscan.
+The modes "720x480Noscale" (NTSC) and "720x576Noscale" (PAL) (available
+on VT1622 only) provide cleaner TV output (unscaled with only minimal
+overscan).  These modes are made available by the driver; modelines
+provided in __xconfigfile__ will be ignored.
+
+.SH "SEE ALSO"
+__xservername__(__appmansuffix__), __xconfigfile__(__filemansuffix__), xorgconfig(__appmansuffix__), Xserver(__appmansuffix__), X(__miscmansuffix__)
+.SH AUTHORS
+Authors include: ...
diff -Nurp trunk.orig/unichrome/Imakefile trunk/unichrome/Imakefile
--- trunk.orig/unichrome/Imakefile	2007-05-15 23:41:58.000000000 +0200
+++ trunk/unichrome/Imakefile	2007-05-15 23:47:39.000000000 +0200
@@ -90,7 +90,7 @@ DEFXNEEDDRMLOCK = -DX_NEED_DRMLOCK
 
 #if BuildXF86DRI
 DRISRCS = via_dri.c via_xvmc.c
-DRIOBJS = via_dri.o via_xvmc.o
+DRIOBJS = openchrome_dri.o openchrome_xvmc.o
 DRIINCLUDES = -I$(SERVERSRC)/GL/dri -I$(LIBSRC)/GL/dri \
        $(X_DRM_H_LOCATION) -I$(TOP)/include
 DRIDEFINES = $(GLX_DEFINES)
@@ -121,24 +121,24 @@ SRCS = via_driver.c \
     via_vt162x.c \
     $(DRISRCS)
 
-OBJS = via_driver.o \
-    via_3d.o \
-    via_accel.o \
-    via_bandwidth.o \
-    via_ch7xxx.o \
-    via_mode.o \
-    via_cursor.o \
-    via_shadow.o \
-    via_dga.o \
-    via_video.o \
-    via_i2c.o \
-    via_id.o \
-    via_swov.o \
-    via_memory.o \
-    via_memcpy.o \
-    via_vbe.o \
-    via_vgahw.o \
-    via_vt162x.o \
+OBJS = openchrome_driver.o \
+    openchrome_3d.o \
+    openchrome_accel.o \
+    openchrome_bandwidth.o \
+    openchrome_ch7xxx.o \
+    openchrome_mode.o \
+    openchrome_cursor.o \
+    openchrome_shadow.o \
+    openchrome_dga.o \
+    openchrome_video.o \
+    openchrome_i2c.o \
+    openchrome_id.o \
+    openchrome_swov.o \
+    openchrome_memory.o \
+    openchrome_memcpy.o \
+    openchrome_vbe.o \
+    openchrome_vgahw.o \
+    openchrome_vt162x.o \
     $(DRIOBJS)
 
 #if defined(XF86DriverSDK)
@@ -168,64 +168,64 @@ NormalAsmObjectRule()
 
 ModuleObjectRule()
 #ifdef X_NEED_NEWMODULETARGET
-ObjectModuleTarget(via, $(OBJS),drivers)
+ObjectModuleTarget(openchrome, $(OBJS),drivers)
 #else
-ObjectModuleTarget(via, $(OBJS))
+ObjectModuleTarget(openchrome, $(OBJS))
 #endif
 
 #ifdef InstallVideoObjectModule
-InstallVideoObjectModule(via,$(MODULEDIR))
+InstallVideoObjectModule(openchrome,$(MODULEDIR))
 #else
-InstallObjectModule(via,$(MODULEDIR),drivers)
+InstallObjectModule(openchrome,$(MODULEDIR),drivers)
 #endif
 
 #if !defined(XF86DriverSDK)
-CppManTarget(via,)
-InstallModuleManPage(via)
+CppManTarget(openchrome,)
+InstallModuleManPage(openchrome)
 #endif
 
 DependTarget()
 
-InstallDriverSDKNonExecFile(Imakefile,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_3d.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_3d.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_accel.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_bandwidth.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_bios.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_ch7xxx.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_ch7xxx.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_cursor.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_dga.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_dmabuffer.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_dri.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_dri.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_driver.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_driver.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_drm.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_drmclient.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_i2c.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_id.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_id.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_memcpy.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_memcpy.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_memory.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_mode.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_mode.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_priv.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_regs.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_shadow.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_swov.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_swov.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_vbe.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_vgahw.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_vgahw.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_video.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_video.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_vt162x.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_vt162x.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_xvmc.c,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_xvmc.h,$(DRIVERSDKDIR)/drivers/via)
-InstallDriverSDKNonExecFile(via_xvpriv.h,$(DRIVERSDKDIR)/drivers/via)
+InstallDriverSDKNonExecFile(Imakefile,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_3d.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_3d.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_accel.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_bandwidth.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_bios.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_ch7xxx.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_ch7xxx.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_cursor.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_dga.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_dmabuffer.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_dri.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_dri.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_driver.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_driver.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_drm.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_drmclient.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_i2c.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_id.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_id.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_memcpy.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_memcpy.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_memory.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_mode.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_mode.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_priv.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_regs.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_shadow.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_swov.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_swov.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_vbe.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_vgahw.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_vgahw.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_video.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_video.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_vt162x.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_vt162x.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_xvmc.c,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_xvmc.h,$(DRIVERSDKDIR)/drivers/openchrome)
+InstallDriverSDKNonExecFile(via_xvpriv.h,$(DRIVERSDKDIR)/drivers/openchrome)
 
 InstallDriverSDKObjectModule(via,$(DRIVERSDKMODULEDIR),drivers)
diff -Nurp trunk.orig/unichrome/Makefile.am trunk/unichrome/Makefile.am
--- trunk.orig/unichrome/Makefile.am	2007-05-15 23:44:01.000000000 +0200
+++ trunk/unichrome/Makefile.am	2007-05-15 23:45:18.000000000 +0200
@@ -26,11 +26,11 @@
 EXTRA_DIST = svnversion.h
 CONFIG_CLEAN_FILES= svnversion.h
 AM_CFLAGS = @XORG_CFLAGS@ @DRI_CFLAGS@
-via_drv_la_LTLIBRARIES = via_drv.la
-via_drv_la_LDFLAGS = -module -avoid-version
-via_drv_ladir = @moduledir@/drivers
+openchrome_drv_la_LTLIBRARIES = openchrome_drv.la
+openchrome_drv_la_LDFLAGS = -module -avoid-version
+openchrome_drv_ladir = @moduledir@/drivers
 
-via_drv_la_SOURCES = \
+openchrome_drv_la_SOURCES = \
 	 via_3d_reg.h \
 	 via_3d.c \
 	 via_3d.h \
@@ -68,7 +68,7 @@ via_drv_la_SOURCES = \
          via_xvpriv.h
 
 if DRI
-via_drv_la_SOURCES += \
+openchrome_drv_la_SOURCES += \
          via_dri.c \
          via_dri.h \
          via_drmclient.h \
@@ -84,7 +84,7 @@ EXTRA_DIST += \
          via_xvmc.h 
 endif
 
-via_driver.lo: svnversion.h
+openchrome_driver.lo: svnversion.h
 svnversion.h:
 	@if [ -f svnrelease.h ]; then \
 	echo '#include "svnrelease.h"' > $@.tmp; \
diff -Nurp trunk.orig/unichrome/openchrome.man trunk/unichrome/openchrome.man
--- trunk.orig/unichrome/openchrome.man	1970-01-01 01:00:00.000000000 +0100
+++ trunk/unichrome/openchrome.man	2007-05-15 23:41:58.000000000 +0200
@@ -0,0 +1,230 @@
+.\" Shorthand for double quote that works everywhere,
+.\" also within other double quotes:
+.ds q \N'34'
+.TH VIA __drivermansuffix__ __vendorversion__
+.SH NAME
+via \- video driver for VIA Unichromes
+.SH SYNOPSIS
+.nf
+.B "Section \*qDevice\*q"
+.BI "  Identifier \*q" string \*q
+.B  "  Driver \*qvia\*q"
+\ \ ...
+.B EndSection
+.fi
+
+.SH DESCRIPTION
+.B via
+is an __xservername__ driver for VIA chipsets that have  an integrated
+Unichrome graphics engine.
+.PP
+The
+.B via
+driver supports the CLE266, KM/N400, K8M/N800, PM/N800 and CN400 chipsets
+from VIA, including 2D acceleration and the Xv video overlay extensions.
+Flat panel, TV, and VGA outputs are supported, depending on the hardware
+configuration.
+.PP
+3D direct rendering is available using experimental drivers from Mesa
+(www.mesa3d.org).  There is also an XvMC client library for hardware
+acceleration of MPEG1/MPEG2 decoding (available on the CLE266,
+PM/N800, K8M/N800, and CN400 chipsets) that uses the Direct Rendering
+Infrastructure (DRI).  The XvMC client library implements a non-standard
+"VLD" extension to the XvMC standard.  The current Direct Rendering
+Manager (DRM) kernel module is available at dri.sourceforge.net.
+.PP
+The driver supports free modes for Unichrome Pros (K8M/N800, PM/N800, and
+CN400).  For plain Unichromes (CLE266, KM/N400), it currently supports
+only a limited number of dotclocks, so if you are using X modelines you
+must make sure that the dotclock is one of those supported.  Supported
+dotclocks on plain Unichromes are currently (in MHz): 25.2, 25.312,
+26.591, 31.5, 31.704, 32.663, 33.750, 35.5, 36.0, 39.822, 40.0, 41.164,
+46.981, 49.5, 50.0, 56.3, 57.284, 64.995, 65.0, 65.028, 74.480,
+75.0, 78.8, 81.613, 94.5, 108.0, 108.28, 122.0, 122.726, 135.0,
+148.5, 155.8, 157.5, 161.793, 162.0, 175.5, 189.0, 202.5, 204.8,
+218.3, 229.5.  On top of this, bandwidth restrictions apply for both
+Unichromes and Unichrome Pros.
+.PP
+.SH CONFIGURATION DETAILS
+Please refer to __xconfigfile__(__filemansuffix__) for general configuration
+details.  This section only covers configuration details specific to this
+driver.
+.PP
+The following driver
+.B options
+are supported:
+.TP
+.BI "Option \*qAccelMethod\*q  \*q" string \*q
+The driver supports "XAA" and "EXA" acceleration methods.  The default
+method is XAA, since EXA is still experimental.  Contrary to XAA, EXA
+implements acceleration for screen uploads and downlads (if DRI is
+enabled) and for the Render/Composite extension.
+.TP
+.BI "Option \*qActiveDevice\*q  \*q" string \*q
+Specifies the active device combination.  Any string containing "CRT",
+"LCD", "DFP", "TV" should be possible. "CRT" represents anything that
+is connected to the VGA port, "LCD" and "DFP" are for laptop panels
+(not TFT screens attached to the VGA port), "TV" is self-explanatory.
+The default is to use what is detected.  The driver is currently unable
+to use LCD and TV simultaneously, and will favour the LCD.
+.TP
+.BI "Option \*qAGPMem\*q  \*q" integer \*q
+Sets the amount of AGP memory that is allocated at X server startup.
+The allocated memory will be "integer" kB.  This AGP memory is used for
+the AGP command buffer (if option "EnableAGPDMA" is set to "true"), for
+DRI textures, and for the EXA scratch area.  The driver will allocate at
+least one system page of AGP memory and, if the AGP command buffer is
+used, at least 2MB + one system page.  If there is no room for the EXA
+scratch area in AGP space, it will be allocated from VRAM.  If there is
+no room for DRI textures, they will be allocated from the DRI part of
+VRAM (see the option "MaxDRIMem").  The default amount of AGP is
+32768kB.  Note that the AGP aperture set in the BIOS must be able
+to accomodate the amount of AGP memory specified here.  Otherwise no
+AGP memory will be available.  It is safe to set a very large AGP
+aperture in the BIOS.
+.TP
+.BI "Option \*qCenter\*q  \*q" boolean \*q
+Enables or disables image centering on DVI displays.
+.TP
+.BI "Option \*qDisableIRQ\*q  \*q" boolean \*q
+Disables the Vblank IRQ.  This is a workaround for some mainboards that
+have problems with IRQs from the Unichrome engine.  With IRQs disabled,
+DRI clients have no way to synchronize drawing to Vblank.
+.TP
+.BI "Option \*qDisableVQ\*q  \*q" boolean \*q
+Disables the use of VQ.  VQ is enabled by default.
+.TP
+.BI "Option \*qEnableAGPDMA\*q  \*q" boolean \*q
+Enables the AGP DMA functionality in DRM.  This requires that DRI is enabled
+and will force 2D and 3D acceleration to use AGP DMA.  The XvMC DRI client
+will also make use of this on the CLE266 to consume much less CPU. 
+.TP
+.BI "Option \*qExaNoComposite\*q  \*q" boolean \*q
+If EXA is enabled (using the option "AccelMethod"), this option enables
+or disables acceleration of compositing.  Since EXA, and in particular its
+composite acceleration, is still experimental, this is a way to disable
+a misbehaving composite acceleration.
+.TP
+.BI "Option \*qExaScratchSize\*q  \*q" integer \*q
+Sets the size of the EXA scratch area to "integer" kB.  This area is
+used by EXA as a last place to look for available space for pixmaps.
+Too little space will slow compositing down.  This option should be set
+to the size of the largest pixmap used.  If you have a screen width of
+over 1024 pixels and use 24 bpp, set this to 8192.  Otherwise you can
+leave this at the default 4096.  The space will be allocated from AGP
+memory if available, otherwise from VRAM.
+.TP
+.BI "Option \*qHWCursor\*q  \*q" boolean \*q
+Enables or disables the use of hardware cursors.  The default is enabled.
+.TP
+.BI "Option \*qLCDDualEdge\*q  \*q" boolean \*q
+Enables or disables the use of dual-edge mode to set the LCD.
+.TP
+.BI "Option \*qMaxDRIMem\*q  \*q" integer \*q
+Sets the maximum amount of VRAM memory allocated for DRI clients to
+"integer" kB.  Normally DRI clients  get half the available VRAM size,
+but in some cases it may make sense to limit this amount.  For example,
+if you are using a composite manager and you want to give as much memory
+as possible to the EXA pixmap storage area.
+.TP
+.BI "Option \*qMigrationHeuristic\*q  \*q" string \*q
+Sets the heuristic for EXA pixmap migration.  This is an EXA core
+option, and on Xorg xserver versions after 1.1.0 this defaults to
+"smart".  The Openchrome driver performs best with "greedy", so you
+should really add this option to your configuration file.  The third
+possibility is "always", which might become more useful in the future.
+.TP
+.BI "Option \*qNoAccel\*q  \*q" boolean \*q
+Disables or enables acceleration.  Default: acceleration is enabled.
+.TP
+.BI "Option \*qNoAGPFor2D\*q  \*q" boolean \*q
+With this option set, 2D acceleration will not use AGP DMA even if
+it is enabled.
+.TP
+.BI "Option \*qNoXVDMA\*q  \*q" boolean \*q
+If DRI is enabled, Xv normally uses PCI DMA to transfer video images
+from system to frame-buffer memory.  This is somewhat slower than
+direct copies due to the limitations of the PCI bus, but on the other
+hand it decreases CPU usage significantly, particularly on computers
+with fast processors.  Some video players are buggy and will display
+rendering artifacts when PCI DMA is used.  If you experience this,
+or don't want your PCI bus to be stressed with Xv images, set this
+option to "true".  This option has no effect when DRI is not enabled.
+.TP
+.BI "Option \*qPanelSize\*q  \*q" string \*q
+Specifies the size (width x height) of the LCD panel attached to the
+system.  The sizes 640x480, 800x600, 1024x768, 1280x1024, and 1400x1050
+are supported.
+.TP
+.BI "Option \*qRotate\*q  \*q" string \*q
+Rotates the display either clockwise ("CW") or counterclockwise ("CCW").
+Rotation is only supported unaccelerated.
+.TP
+.BI "Option \*qShadowFB\*q  \*q" boolean \*q
+Uses a shadow frame buffer.  This is required when rotating the display,
+but otherwise defaults to disabled.
+.TP
+.BI "Option \*qSWCursor\*q  \*q" boolean \*q
+Enables or disables the use of a software cursor.  The default is disabled.
+.TP
+.BI "Option \*qTVDeflicker\*q  \*q" integer \*q
+Specifies the deflicker setting for TV output.  Valid values are "0", "1",
+and "2".  0) No deflicker, 1) 1:1:1 deflicker, 2) 1:2:1 deflicker.
+.TP
+.BI "Option \*qTVDotCrawl\*q  \*q" boolean \*q
+Enables or disables dotcrawl.
+.TP
+.BI "Option \*qTVOutput\*q  \*q" string \*q
+Specifies which TV output to use.  The driver supports "S-Video",
+"Composite", "SC", "RGB" and "YCbCr" outputs.  Note that on some
+EPIA boards the compositer-video port is shared with audio-out and
+is selected via a jumper.
+.TP
+.BI "Option \*qTVType\*q  \*q" string \*q
+Specifies TV output format.  The driver currently supports "NTSC" and
+"PAL" timings only.
+.TP
+.BI "Option \*qVBEModes\*q  \*q" boolean \*q
+Uses the VBE BIOS calls to set the display mode.  This mimics the
+behaviour of the vesa video driver but still provides acceleration and
+other features.  This option may be used if your hardware works with
+the vesa driver but not with the Openchrome driver.  It may not work
+on 64-bit systems.  Using "VBEModes" may speed up driver acceleration
+significantly due to a more aggressive hardware setting, particularly
+on systems with low memory bandwidth.  Your refresh rate may be limited
+to 60 Hz on some systems.
+.TP
+.BI "Option \*qVBESaveRestore\*q  \*q" boolean \*q
+Uses the VBE BIOS calls to save and restore the display state when the
+X server is launched.  This can be extremely slow on some hardware, and
+the system may appear to have locked for 10 seconds or so.  The default
+is to use the driver builtin function.  This option only works if option
+"VBEModes" is enabled.
+.TP
+.BI "Option \*qVideoRAM\*q  \*q" integer \*q
+Overrides the VideoRAM autodetection.  This should never be needed.
+.PP 
+.SH "TV ENCODERS"
+Unichromes tend to be paired with several different TV encoders.
+.TP
+.BI "VIA Technologies VT1621"
+Still untested, as no combination with a Unichrome is known or available.
+Supports the following normal modes: "640x480" and "800x600".  Use
+"640x480Over" and "800x600Over" for vertical overscan.  These modes
+are made available by the driver; modelines provided in __xconfigfile__
+will be ignored.
+.TP
+.BI "VIA Technologies VT1622, VT1622A, VT1623"
+Supports the following modes: "640x480", "800x600", "1024x768",
+"848x480", "720x480" (NTSC only) and "720x576" (PAL only).  Use
+"640x480Over", "800x600Over", "1024x768Over", "848x480Over",
+"720x480Over" (NTSC) and "720x576Over" (PAL) for vertical overscan.
+The modes "720x480Noscale" (NTSC) and "720x576Noscale" (PAL) (available
+on VT1622 only) provide cleaner TV output (unscaled with only minimal
+overscan).  These modes are made available by the driver; modelines
+provided in __xconfigfile__ will be ignored.
+
+.SH "SEE ALSO"
+__xservername__(__appmansuffix__), __xconfigfile__(__filemansuffix__), xorgconfig(__appmansuffix__), Xserver(__appmansuffix__), X(__miscmansuffix__)
+.SH AUTHORS
+Authors include: ...
diff -Nurp trunk.orig/unichrome/via_driver.c trunk/unichrome/via_driver.c
--- trunk.orig/unichrome/via_driver.c	2007-05-15 23:41:58.000000000 +0200
+++ trunk/unichrome/via_driver.c	2007-05-16 00:04:24.000000000 +0200
@@ -43,8 +43,6 @@
 #define DPMS_SERVER
 #include <X11/extensions/dpms.h>
 
-#include "svnversion.h"
-
 #include "via_driver.h"
 #include "via_video.h"
 
@@ -393,7 +391,7 @@ static const char *driSymbols[] = {
 static MODULESETUPPROTO(VIASetup);
 
 static XF86ModuleVersionInfo VIAVersRec = {
-    "via",
+    "openchrome",
     MODULEVENDORSTRING,
     MODINFOSTRING1,
     MODINFOSTRING2,
@@ -409,7 +407,7 @@ static XF86ModuleVersionInfo VIAVersRec 
     {0, 0, 0, 0}
 };
 
-XF86ModuleData viaModuleData = {&VIAVersRec, VIASetup, NULL};
+XF86ModuleData openchromeModuleData = {&VIAVersRec, VIASetup, NULL};
 
 static pointer VIASetup(pointer module, pointer opts, int *errmaj, int *errmin)
 {
diff -Nurp trunk.orig/unichrome/via.man trunk/unichrome/via.man
--- trunk.orig/unichrome/via.man	2007-05-15 23:44:01.000000000 +0200
+++ trunk/unichrome/via.man	1970-01-01 01:00:00.000000000 +0100
@@ -1,230 +0,0 @@
-.\" Shorthand for double quote that works everywhere,
-.\" also within other double quotes:
-.ds q \N'34'
-.TH VIA __drivermansuffix__ __vendorversion__
-.SH NAME
-via \- video driver for VIA Unichromes
-.SH SYNOPSIS
-.nf
-.B "Section \*qDevice\*q"
-.BI "  Identifier \*q" string \*q
-.B  "  Driver \*qvia\*q"
-\ \ ...
-.B EndSection
-.fi
-
-.SH DESCRIPTION
-.B via
-is an __xservername__ driver for VIA chipsets that have  an integrated
-Unichrome graphics engine.
-.PP
-The
-.B via
-driver supports the CLE266, KM/N400, K8M/N800, PM/N800 and CN400 chipsets
-from VIA, including 2D acceleration and the Xv video overlay extensions.
-Flat panel, TV, and VGA outputs are supported, depending on the hardware
-configuration.
-.PP
-3D direct rendering is available using experimental drivers from Mesa
-(www.mesa3d.org).  There is also an XvMC client library for hardware
-acceleration of MPEG1/MPEG2 decoding (available on the CLE266,
-PM/N800, K8M/N800, and CN400 chipsets) that uses the Direct Rendering
-Infrastructure (DRI).  The XvMC client library implements a non-standard
-"VLD" extension to the XvMC standard.  The current Direct Rendering
-Manager (DRM) kernel module is available at dri.sourceforge.net.
-.PP
-The driver supports free modes for Unichrome Pros (K8M/N800, PM/N800, and
-CN400).  For plain Unichromes (CLE266, KM/N400), it currently supports
-only a limited number of dotclocks, so if you are using X modelines you
-must make sure that the dotclock is one of those supported.  Supported
-dotclocks on plain Unichromes are currently (in MHz): 25.2, 25.312,
-26.591, 31.5, 31.704, 32.663, 33.750, 35.5, 36.0, 39.822, 40.0, 41.164,
-46.981, 49.5, 50.0, 56.3, 57.284, 64.995, 65.0, 65.028, 74.480,
-75.0, 78.8, 81.613, 94.5, 108.0, 108.28, 122.0, 122.726, 135.0,
-148.5, 155.8, 157.5, 161.793, 162.0, 175.5, 189.0, 202.5, 204.8,
-218.3, 229.5.  On top of this, bandwidth restrictions apply for both
-Unichromes and Unichrome Pros.
-.PP
-.SH CONFIGURATION DETAILS
-Please refer to __xconfigfile__(__filemansuffix__) for general configuration
-details.  This section only covers configuration details specific to this
-driver.
-.PP
-The following driver
-.B options
-are supported:
-.TP
-.BI "Option \*qAccelMethod\*q  \*q" string \*q
-The driver supports "XAA" and "EXA" acceleration methods.  The default
-method is XAA, since EXA is still experimental.  Contrary to XAA, EXA
-implements acceleration for screen uploads and downlads (if DRI is
-enabled) and for the Render/Composite extension.
-.TP
-.BI "Option \*qActiveDevice\*q  \*q" string \*q
-Specifies the active device combination.  Any string containing "CRT",
-"LCD", "DFP", "TV" should be possible. "CRT" represents anything that
-is connected to the VGA port, "LCD" and "DFP" are for laptop panels
-(not TFT screens attached to the VGA port), "TV" is self-explanatory.
-The default is to use what is detected.  The driver is currently unable
-to use LCD and TV simultaneously, and will favour the LCD.
-.TP
-.BI "Option \*qAGPMem\*q  \*q" integer \*q
-Sets the amount of AGP memory that is allocated at X server startup.
-The allocated memory will be "integer" kB.  This AGP memory is used for
-the AGP command buffer (if option "EnableAGPDMA" is set to "true"), for
-DRI textures, and for the EXA scratch area.  The driver will allocate at
-least one system page of AGP memory and, if the AGP command buffer is
-used, at least 2MB + one system page.  If there is no room for the EXA
-scratch area in AGP space, it will be allocated from VRAM.  If there is
-no room for DRI textures, they will be allocated from the DRI part of
-VRAM (see the option "MaxDRIMem").  The default amount of AGP is
-32768kB.  Note that the AGP aperture set in the BIOS must be able
-to accomodate the amount of AGP memory specified here.  Otherwise no
-AGP memory will be available.  It is safe to set a very large AGP
-aperture in the BIOS.
-.TP
-.BI "Option \*qCenter\*q  \*q" boolean \*q
-Enables or disables image centering on DVI displays.
-.TP
-.BI "Option \*qDisableIRQ\*q  \*q" boolean \*q
-Disables the Vblank IRQ.  This is a workaround for some mainboards that
-have problems with IRQs from the Unichrome engine.  With IRQs disabled,
-DRI clients have no way to synchronize drawing to Vblank.
-.TP
-.BI "Option \*qDisableVQ\*q  \*q" boolean \*q
-Disables the use of VQ.  VQ is enabled by default.
-.TP
-.BI "Option \*qEnableAGPDMA\*q  \*q" boolean \*q
-Enables the AGP DMA functionality in DRM.  This requires that DRI is enabled
-and will force 2D and 3D acceleration to use AGP DMA.  The XvMC DRI client
-will also make use of this on the CLE266 to consume much less CPU. 
-.TP
-.BI "Option \*qExaNoComposite\*q  \*q" boolean \*q
-If EXA is enabled (using the option "AccelMethod"), this option enables
-or disables acceleration of compositing.  Since EXA, and in particular its
-composite acceleration, is still experimental, this is a way to disable
-a misbehaving composite acceleration.
-.TP
-.BI "Option \*qExaScratchSize\*q  \*q" integer \*q
-Sets the size of the EXA scratch area to "integer" kB.  This area is
-used by EXA as a last place to look for available space for pixmaps.
-Too little space will slow compositing down.  This option should be set
-to the size of the largest pixmap used.  If you have a screen width of
-over 1024 pixels and use 24 bpp, set this to 8192.  Otherwise you can
-leave this at the default 4096.  The space will be allocated from AGP
-memory if available, otherwise from VRAM.
-.TP
-.BI "Option \*qHWCursor\*q  \*q" boolean \*q
-Enables or disables the use of hardware cursors.  The default is enabled.
-.TP
-.BI "Option \*qLCDDualEdge\*q  \*q" boolean \*q
-Enables or disables the use of dual-edge mode to set the LCD.
-.TP
-.BI "Option \*qMaxDRIMem\*q  \*q" integer \*q
-Sets the maximum amount of VRAM memory allocated for DRI clients to
-"integer" kB.  Normally DRI clients  get half the available VRAM size,
-but in some cases it may make sense to limit this amount.  For example,
-if you are using a composite manager and you want to give as much memory
-as possible to the EXA pixmap storage area.
-.TP
-.BI "Option \*qMigrationHeuristic\*q  \*q" string \*q
-Sets the heuristic for EXA pixmap migration.  This is an EXA core
-option, and on Xorg xserver versions after 1.1.0 this defaults to
-"smart".  The Openchrome driver performs best with "greedy", so you
-should really add this option to your configuration file.  The third
-possibility is "always", which might become more useful in the future.
-.TP
-.BI "Option \*qNoAccel\*q  \*q" boolean \*q
-Disables or enables acceleration.  Default: acceleration is enabled.
-.TP
-.BI "Option \*qNoAGPFor2D\*q  \*q" boolean \*q
-With this option set, 2D acceleration will not use AGP DMA even if
-it is enabled.
-.TP
-.BI "Option \*qNoXVDMA\*q  \*q" boolean \*q
-If DRI is enabled, Xv normally uses PCI DMA to transfer video images
-from system to frame-buffer memory.  This is somewhat slower than
-direct copies due to the limitations of the PCI bus, but on the other
-hand it decreases CPU usage significantly, particularly on computers
-with fast processors.  Some video players are buggy and will display
-rendering artifacts when PCI DMA is used.  If you experience this,
-or don't want your PCI bus to be stressed with Xv images, set this
-option to "true".  This option has no effect when DRI is not enabled.
-.TP
-.BI "Option \*qPanelSize\*q  \*q" string \*q
-Specifies the size (width x height) of the LCD panel attached to the
-system.  The sizes 640x480, 800x600, 1024x768, 1280x1024, and 1400x1050
-are supported.
-.TP
-.BI "Option \*qRotate\*q  \*q" string \*q
-Rotates the display either clockwise ("CW") or counterclockwise ("CCW").
-Rotation is only supported unaccelerated.
-.TP
-.BI "Option \*qShadowFB\*q  \*q" boolean \*q
-Uses a shadow frame buffer.  This is required when rotating the display,
-but otherwise defaults to disabled.
-.TP
-.BI "Option \*qSWCursor\*q  \*q" boolean \*q
-Enables or disables the use of a software cursor.  The default is disabled.
-.TP
-.BI "Option \*qTVDeflicker\*q  \*q" integer \*q
-Specifies the deflicker setting for TV output.  Valid values are "0", "1",
-and "2".  0) No deflicker, 1) 1:1:1 deflicker, 2) 1:2:1 deflicker.
-.TP
-.BI "Option \*qTVDotCrawl\*q  \*q" boolean \*q
-Enables or disables dotcrawl.
-.TP
-.BI "Option \*qTVOutput\*q  \*q" string \*q
-Specifies which TV output to use.  The driver supports "S-Video",
-"Composite", "SC", "RGB" and "YCbCr" outputs.  Note that on some
-EPIA boards the compositer-video port is shared with audio-out and
-is selected via a jumper.
-.TP
-.BI "Option \*qTVType\*q  \*q" string \*q
-Specifies TV output format.  The driver currently supports "NTSC" and
-"PAL" timings only.
-.TP
-.BI "Option \*qVBEModes\*q  \*q" boolean \*q
-Uses the VBE BIOS calls to set the display mode.  This mimics the
-behaviour of the vesa video driver but still provides acceleration and
-other features.  This option may be used if your hardware works with
-the vesa driver but not with the Openchrome driver.  It may not work
-on 64-bit systems.  Using "VBEModes" may speed up driver acceleration
-significantly due to a more aggressive hardware setting, particularly
-on systems with low memory bandwidth.  Your refresh rate may be limited
-to 60 Hz on some systems.
-.TP
-.BI "Option \*qVBESaveRestore\*q  \*q" boolean \*q
-Uses the VBE BIOS calls to save and restore the display state when the
-X server is launched.  This can be extremely slow on some hardware, and
-the system may appear to have locked for 10 seconds or so.  The default
-is to use the driver builtin function.  This option only works if option
-"VBEModes" is enabled.
-.TP
-.BI "Option \*qVideoRAM\*q  \*q" integer \*q
-Overrides the VideoRAM autodetection.  This should never be needed.
-.PP 
-.SH "TV ENCODERS"
-Unichromes tend to be paired with several different TV encoders.
-.TP
-.BI "VIA Technologies VT1621"
-Still untested, as no combination with a Unichrome is known or available.
-Supports the following normal modes: "640x480" and "800x600".  Use
-"640x480Over" and "800x600Over" for vertical overscan.  These modes
-are made available by the driver; modelines provided in __xconfigfile__
-will be ignored.
-.TP
-.BI "VIA Technologies VT1622, VT1622A, VT1623"
-Supports the following modes: "640x480", "800x600", "1024x768",
-"848x480", "720x480" (NTSC only) and "720x576" (PAL only).  Use
-"640x480Over", "800x600Over", "1024x768Over", "848x480Over",
-"720x480Over" (NTSC) and "720x576Over" (PAL) for vertical overscan.
-The modes "720x480Noscale" (NTSC) and "720x576Noscale" (PAL) (available
-on VT1622 only) provide cleaner TV output (unscaled with only minimal
-overscan).  These modes are made available by the driver; modelines
-provided in __xconfigfile__ will be ignored.
-
-.SH "SEE ALSO"
-__xservername__(__appmansuffix__), __xconfigfile__(__filemansuffix__), xorgconfig(__appmansuffix__), Xserver(__appmansuffix__), X(__miscmansuffix__)
-.SH AUTHORS
-Authors include: ...
diff -Nurp trunk.orig/libxvmc/Imakefile trunk/libxvmc/Imakefile
--- trunk.orig/libxvmc/Imakefile	2007-05-16 00:37:57.000000000 +0200
+++ trunk/libxvmc/Imakefile	2007-05-16 00:46:58.000000000 +0200
@@ -2,7 +2,7 @@
 #define DoSharedLib SharedLibXvMC
 #define DoDebugLib DebugLibXvMC
 #define DoProfileLib ProfileLibXvMC
-/*#define LibName viaXvMC*/
+/*#define LibName openchromeXvMC*/
 #define SoRev SOXVMCREV
 #define LibHeaders NO
 
@@ -41,10 +41,10 @@ ALLOC_DEFINES = -DMALLOC_0_RETURNS_NULL
 
 #include <Library.tmpl>
 
-SharedLibraryTarget(viaXvMC,$(SoRev),$(OBJS) $(EXTRASHAREDOBJS),.,.)
-SharedLibraryTarget(viaXvMCPro,$(SoRev),$(OBJSPRO) $(EXTRASHAREDOBJS),.,.)
-InstallSharedLibrary(viaXvMC,$(SoRev),$(SHLIBDIR))
-InstallSharedLibrary(viaXvMCPro,$(SoRev),$(SHLIBDIR))
+SharedLibraryTarget(openchromeXvMC,$(SoRev),$(OBJS) $(EXTRASHAREDOBJS),.,.)
+SharedLibraryTarget(openchromeXvMCPro,$(SoRev),$(OBJSPRO) $(EXTRASHAREDOBJS),.,.)
+InstallSharedLibrary(openchromeXvMC,$(SoRev),$(SHLIBDIR))
+InstallSharedLibrary(openchromeXvMCPro,$(SoRev),$(SHLIBDIR))
 
 #if defined(LinuxArchitecture)
 OS_SUBDIR = linux
diff -Nurp trunk.orig/libxvmc/Makefile.am trunk/libxvmc/Makefile.am
--- trunk.orig/libxvmc/Makefile.am	2007-05-16 00:37:57.000000000 +0200
+++ trunk/libxvmc/Makefile.am	2007-05-16 00:51:46.000000000 +0200
@@ -1,7 +1,7 @@
 if XVMC
-lib_LTLIBRARIES=libviaXvMC.la libviaXvMCPro.la
+lib_LTLIBRARIES=libopenchromeXvMC.la libopenchromeXvMCPro.la
 
-libviaXvMC_la_SOURCES = \
+libopenchromeXvMC_la_SOURCES = \
 	viaLowLevel.c \
 	driDrawable.c \
 	viaXvMC.c \
@@ -12,7 +12,7 @@ libviaXvMC_la_SOURCES = \
 	xf86dri.h \
 	xf86dristr.h \
 	vldXvMC.h
-libviaXvMCPro_la_SOURCES = \
+libopenchromeXvMCPro_la_SOURCES = \
 	viaLowLevelPro.c \
 	driDrawable.c \
 	viaXvMC.c \
@@ -24,11 +24,11 @@ libviaXvMCPro_la_SOURCES = \
 	xf86dristr.h \
 	vldXvMC.h
 
-libviaXvMC_la_CFLAGS = @XORG_CFLAGS@ @DRI_CFLAGS@ -I$(top_srcdir)/unichrome -I$(top_srcdir)/libxvmc -DTRUE=1 -DFALSE=0
-libviaXvMC_la_LDFLAGS = @DRI_LIBS@ -version-number 1:0:0
+libopenchromeXvMC_la_CFLAGS = @XORG_CFLAGS@ @DRI_CFLAGS@ -I$(top_srcdir)/unichrome -I$(top_srcdir)/libxvmc -DTRUE=1 -DFALSE=0
+libopenchromeXvMC_la_LDFLAGS = @DRI_LIBS@ -version-number 1:0:0
 
-libviaXvMCPro_la_CFLAGS = @XORG_CFLAGS@ @DRI_CFLAGS@ -I$(top_srcdir)/unichrome -I$(top_srcdir)/libxvmc -DTRUE=1 -DFALSE=0
-libviaXvMCPro_la_LDFLAGS = @DRI_LIBS@ -version-number 1:0:0
+libopenchromeXvMCPro_la_CFLAGS = @XORG_CFLAGS@ @DRI_CFLAGS@ -I$(top_srcdir)/unichrome -I$(top_srcdir)/libxvmc -DTRUE=1 -DFALSE=0
+libopenchromeXvMCPro_la_LDFLAGS = @DRI_LIBS@ -version-number 1:0:0
 else
 EXTRA_DIST = \
 	driDrawable.c \
