diff -Nurp et131x-1.2.3-1.orig/et131x_initpci.c et131x-1.2.3-1/et131x_initpci.c
--- et131x-1.2.3-1.orig/et131x_initpci.c	2007-11-08 15:44:22.000000000 +0000
+++ et131x-1.2.3-1/et131x_initpci.c	2008-02-25 13:17:01.000000000 +0000
@@ -1452,7 +1452,6 @@ int et131x_pci_setup( struct pci_dev *pd
      *************************************************************************/
     DBG_TRACE( et131x_dbginfo, "Setting fundamental net_device info...\n" );
 
-    SET_MODULE_OWNER( netdev );
     SET_NETDEV_DEV( netdev, &pdev->dev );
 
     if( pci_using_dac )
