--- et131x-1.2.3/debian/rules.orig	2007-10-12 11:26:25.000000000 +0000
+++ et131x-1.2.3/debian/rules	2007-10-12 11:27:04.000000000 +0000
@@ -97,7 +97,7 @@ binary-modules: kdist_config
 	dh_clean -k
 	dh_installdirs lib/modules/$(KVERS)/extra
 	# Build the module
-	$(MAKE) KERNEL_SRC=$(KSRC) KVERS=$(KVERS)
+	$(MAKE) KERNEL_SRC=$(KSRC) KERNEL_VER=$(KVERS) KERNEL_DIR=$(KSRC)
 	# Install the module
 	cp et131x.$ko debian/$(PKGNAME)/lib/modules/$(KVERS)/extra
 	
