diff -Nurp et131x-1.2.3.orig/et131x_netdev.c et131x-1.2.3/et131x_netdev.c
--- et131x-1.2.3.orig/et131x_netdev.c	2007-01-25 22:31:35.000000000 +0100
+++ et131x-1.2.3/et131x_netdev.c	2007-06-13 09:42:42.000000000 +0200
@@ -1602,7 +1602,7 @@ void et131x_vlan_rx_kill_vid( struct net
 
     if( pAdapter->vlgrp )
     {
-        pAdapter->vlgrp->vlan_devices[vid] = NULL;
+	vlan_group_set_device(pAdapter->vlgrp, vid, NULL);
     }
 
 
