--- SConstruct.orig	2009-07-20 15:55:10.000000000 +0000
+++ SConstruct	2009-08-07 19:03:22.420541992 +0000
@@ -550,6 +550,9 @@
 if not env['WIIMOTE']:
 	print 'WIIMOTE not enabled. Build with \'scons WIIMOTE=1\' to enable support.'
 
+libraries['sndfile-ardour'] = LibraryInfo()
+libraries['sndfile-ardour'].ParseConfig('pkg-config --cflags --libs sndfile')
+
 libraries['jack'] = LibraryInfo()
 libraries['jack'].ParseConfig('pkg-config --cflags --libs jack')
 
@@ -1082,9 +1085,9 @@
 # cannot use system one for the time being
 #
     
-    libraries['sndfile-ardour'] = LibraryInfo(LIBS='libsndfile-ardour',
-                                    LIBPATH='#libs/libsndfile',
-                                    CPPPATH=['#libs/libsndfile/src'])
+#    libraries['sndfile-ardour'] = LibraryInfo(LIBS='libsndfile-ardour',
+#                                    LIBPATH='#libs/libsndfile',
+#                                    CPPPATH=['#libs/libsndfile/src'])
 
 #    libraries['libglademm'] = LibraryInfo()
 #    libraries['libglademm'].ParseConfig ('pkg-config --cflags --libs libglademm-2.4')
@@ -1105,7 +1108,6 @@
     ]
     
     subdirs = [
-        'libs/libsndfile',
         'libs/pbd',
         'libs/midi++2',
         'libs/ardour',
