diff -ur xrdp-0.3.1.orig/libxrdp/Makefile xrdp-0.3.1/libxrdp/Makefile
--- xrdp-0.3.1.orig/libxrdp/Makefile	2006-03-12 19:58:12.000000000 -0500
+++ xrdp-0.3.1/libxrdp/Makefile	2006-08-01 16:44:51.000000000 -0400
@@ -13,7 +13,7 @@
 DEFINES = -DXRDP_CFG_FILE=\"$(CFGDIR)/xrdp.ini\" \
           -DXRDP_KEY_FILE=\"$(CFGDIR)/rsakeys.ini\"
 
-CFLAGS = -Wall -O2 -I../common -fPIC $(DEFINES)
+CFLAGS = $(MYCFLAGS) -I../common -fPIC $(DEFINES)
 #CFLAGS += -DXRDP_DEBUG
 C_OS_FLAGS = $(CFLAGS) -c
 LDFLAGS = -shared
diff -ur xrdp-0.3.1.orig/rdp/Makefile xrdp-0.3.1/rdp/Makefile
--- xrdp-0.3.1.orig/rdp/Makefile	2006-03-03 20:02:04.000000000 -0500
+++ xrdp-0.3.1/rdp/Makefile	2006-08-01 16:44:51.000000000 -0400
@@ -10,7 +10,7 @@
 
 DEFINES =
 
-CFLAGS = -Wall -O2 -I../common -fPIC $(DEFINES)
+CFLAGS = $(MYCFLAGS) -I../common -fPIC $(DEFINES)
 #CFLAGS += -DXRDP_DEBUG
 C_OS_FLAGS = $(CFLAGS) -c
 LDFLAGS = -shared
diff -ur xrdp-0.3.1.orig/sesman/Makefile xrdp-0.3.1/sesman/Makefile
--- xrdp-0.3.1.orig/sesman/Makefile	2006-05-26 09:10:14.000000000 -0400
+++ xrdp-0.3.1/sesman/Makefile	2006-08-01 16:44:51.000000000 -0400
@@ -15,7 +15,7 @@
 DEFINES = -DSESMAN_CFG_FILE=\"$(CFGDIR)/sesman.ini\" \
           -DSESMAN_PID_FILE=\"$(PIDDIR)/sesman.pid\"
 
-CFLAGS = -Wall -O2 -I../common $(DEFINES)
+CFLAGS = $(MYCFLAGS) -I../common $(DEFINES)
 LDFLAGS = -L /usr/gnu/lib $(DEFINES)
 C_OS_FLAGS = $(CFLAGS) -c
 CC = gcc
diff -ur xrdp-0.3.1.orig/vnc/Makefile xrdp-0.3.1/vnc/Makefile
--- xrdp-0.3.1.orig/vnc/Makefile	2006-03-03 20:02:06.000000000 -0500
+++ xrdp-0.3.1/vnc/Makefile	2006-08-01 16:44:51.000000000 -0400
@@ -9,7 +9,7 @@
 
 DEFINES =
 
-CFLAGS = -Wall -O2 -I../common -fPIC $(DEFINES)
+CFLAGS = $(MYCFLAGS) -I../common -fPIC $(DEFINES)
 C_OS_FLAGS = $(CFLAGS) -c
 LDFLAGS = -shared
 LIBS = -ldl
diff -ur xrdp-0.3.1.orig/xrdp/Makefile xrdp-0.3.1/xrdp/Makefile
--- xrdp-0.3.1.orig/xrdp/Makefile	2006-03-03 20:02:08.000000000 -0500
+++ xrdp-0.3.1/xrdp/Makefile	2006-08-01 16:44:51.000000000 -0400
@@ -14,7 +14,7 @@
 DEFINES = -DXRDP_CFG_FILE=\"$(CFGDIR)/xrdp.ini\" \
           -DXRDP_PID_FILE=\"$(PIDDIR)/xrdp.pid\"
 
-CFLAGS = -Wall -O2 -I../common -I../libxrdp $(DEFINES)
+CFLAGS = $(MYCFLAGS) -I../common -I../libxrdp $(DEFINES)
 #CFLAGS += -DXRDP_DEBUG
 C_OS_FLAGS = $(CFLAGS) -c
 LDFLAGS = -L/usr/gnu/lib -L../libxrdp -Wl,-rpath,.
diff -ur xrdp-0.3.1.orig/xup/Makefile xrdp-0.3.1/xup/Makefile
--- xrdp-0.3.1.orig/xup/Makefile	2006-03-03 20:02:09.000000000 -0500
+++ xrdp-0.3.1/xup/Makefile	2006-08-01 16:44:51.000000000 -0400
@@ -9,7 +9,7 @@
 
 DEFINES =
 
-CFLAGS = -Wall -O2 -I../common -fPIC $(DEFINES)
+CFLAGS = $(MYCFLAGS) -I../common -fPIC $(DEFINES)
 #CFLAGS += -DXRDP_DEBUG
 C_OS_FLAGS = $(CFLAGS) -c
 LDFLAGS = -shared
