*** acx-20080112.orig/ioctl.c	2008-11-18 15:32:38.000000000 +0100
--- acx-20080112/ioctl.c	2008-11-18 15:34:58.000000000 +0100
*************** end_unlock:
*** 481,486 ****
--- 481,494 ----
  	return result;
  }
  
+ #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 27) && !defined(IW_REQUEST_FLAG_COMPAT)
+ #define        iwe_stream_add_event(a, b, c, d, e)     iwe_stream_add_event(b, c, d, e)
+ #define        iwe_stream_add_point(a, b, c, d, e)     iwe_stream_add_point(b, c, d, e)
+ #define        iwe_stream_add_value(a, b, c, d, e, f)  \
+        iwe_stream_add_value(b, c, d, e, f)
+ #define        iwe_stream_lcp_len(a)                   IW_EV_LCP_LEN
+ #endif
+ 
  
  /***********************************************************************
  ** acx_s_scan_add_station
*************** end_unlock:
*** 489,494 ****
--- 497,503 ----
  static char*
  acx_s_scan_add_station(
  	acx_device_t *adev,
+ 	struct iw_request_info *info,
  	char *ptr,
  	char *end_buf,
  	struct client *bss)
*************** acx_s_scan_add_station(
*** 503,516 ****
  	iwe.u.ap_addr.sa_family = ARPHRD_ETHER;
  	MAC_COPY(iwe.u.ap_addr.sa_data, bss->bssid);
  	acxlog_mac(L_IOCTL, "scan, station address: ", bss->bssid, "\n");
! 	ptr = iwe_stream_add_event(ptr, end_buf, &iwe, IW_EV_ADDR_LEN);
  
  	/* Add ESSID */
  	iwe.cmd = SIOCGIWESSID;
  	iwe.u.data.length = bss->essid_len;
  	iwe.u.data.flags = 1;
  	log(L_IOCTL, "scan, essid: %s\n", bss->essid);
! 	ptr = iwe_stream_add_point(ptr, end_buf, &iwe, bss->essid);
  
  	/* Add mode */
  	iwe.cmd = SIOCGIWMODE;
--- 512,526 ----
  	iwe.u.ap_addr.sa_family = ARPHRD_ETHER;
  	MAC_COPY(iwe.u.ap_addr.sa_data, bss->bssid);
  	acxlog_mac(L_IOCTL, "scan, station address: ", bss->bssid, "\n");
! 	ptr = iwe_stream_add_event(info, ptr, end_buf, &iwe, IW_EV_ADDR_LEN);
! 
  
  	/* Add ESSID */
  	iwe.cmd = SIOCGIWESSID;
  	iwe.u.data.length = bss->essid_len;
  	iwe.u.data.flags = 1;
  	log(L_IOCTL, "scan, essid: %s\n", bss->essid);
! 	ptr = iwe_stream_add_point(info, ptr, end_buf, &iwe, bss->essid);
  
  	/* Add mode */
  	iwe.cmd = SIOCGIWMODE;
*************** acx_s_scan_add_station(
*** 520,526 ****
  		else
  			iwe.u.mode = IW_MODE_ADHOC;
  		log(L_IOCTL, "scan, mode: %d\n", iwe.u.mode);
! 		ptr = iwe_stream_add_event(ptr, end_buf, &iwe, IW_EV_UINT_LEN);
  	}
  
  	/* Add frequency */
--- 530,536 ----
  		else
  			iwe.u.mode = IW_MODE_ADHOC;
  		log(L_IOCTL, "scan, mode: %d\n", iwe.u.mode);
! 		ptr = iwe_stream_add_event(info, ptr, end_buf, &iwe, IW_EV_UINT_LEN);
  	}
  
  	/* Add frequency */
*************** acx_s_scan_add_station(
*** 528,534 ****
  	iwe.u.freq.m = acx_channel_freq[bss->channel - 1] * 100000;
  	iwe.u.freq.e = 1;
  	log(L_IOCTL, "scan, frequency: %d\n", iwe.u.freq.m);
! 	ptr = iwe_stream_add_event(ptr, end_buf, &iwe, IW_EV_FREQ_LEN);
  
  	/* Add link quality */
  	iwe.cmd = IWEVQUAL;
--- 538,544 ----
  	iwe.u.freq.m = acx_channel_freq[bss->channel - 1] * 100000;
  	iwe.u.freq.e = 1;
  	log(L_IOCTL, "scan, frequency: %d\n", iwe.u.freq.m);
! 	ptr = iwe_stream_add_event(info, ptr, end_buf, &iwe, IW_EV_FREQ_LEN);
  
  	/* Add link quality */
  	iwe.cmd = IWEVQUAL;
*************** acx_s_scan_add_station(
*** 546,552 ****
  	iwe.u.qual.updated = 7;
  	log(L_IOCTL, "scan, link quality: %d/%d/%d\n",
  			iwe.u.qual.level, iwe.u.qual.noise, iwe.u.qual.qual);
! 	ptr = iwe_stream_add_event(ptr, end_buf, &iwe, IW_EV_QUAL_LEN);
  
  	/* Add encryption */
  	iwe.cmd = SIOCGIWENCODE;
--- 556,562 ----
  	iwe.u.qual.updated = 7;
  	log(L_IOCTL, "scan, link quality: %d/%d/%d\n",
  			iwe.u.qual.level, iwe.u.qual.noise, iwe.u.qual.qual);
! 	ptr = iwe_stream_add_event(info, ptr, end_buf, &iwe, IW_EV_QUAL_LEN);
  
  	/* Add encryption */
  	iwe.cmd = SIOCGIWENCODE;
*************** acx_s_scan_add_station(
*** 556,562 ****
  		iwe.u.data.flags = IW_ENCODE_DISABLED;
  	iwe.u.data.length = 0;
  	log(L_IOCTL, "scan, encryption flags: %X\n", iwe.u.data.flags);
! 	ptr = iwe_stream_add_point(ptr, end_buf, &iwe, bss->essid);
  
  	/* add rates */
  	iwe.cmd = SIOCGIWRATE;
--- 566,572 ----
  		iwe.u.data.flags = IW_ENCODE_DISABLED;
  	iwe.u.data.length = 0;
  	log(L_IOCTL, "scan, encryption flags: %X\n", iwe.u.data.flags);
! 	ptr = iwe_stream_add_point(info, ptr, end_buf, &iwe, bss->essid);
  
  	/* add rates */
  	iwe.cmd = SIOCGIWRATE;
*************** acx_s_scan_add_station(
*** 570,576 ****
  		if (rate & 1) {
  			iwe.u.bitrate.value = *p * 500000; /* units of 500kb/s */
  			log(L_IOCTL, "scan, rate: %d\n", iwe.u.bitrate.value);
! 			ptr_rate = iwe_stream_add_value(ptr, ptr_rate, end_buf,
  						&iwe, IW_EV_PARAM_LEN);
  		}
  		rate >>= 1;
--- 580,586 ----
  		if (rate & 1) {
  			iwe.u.bitrate.value = *p * 500000; /* units of 500kb/s */
  			log(L_IOCTL, "scan, rate: %d\n", iwe.u.bitrate.value);
! 			ptr_rate = iwe_stream_add_value(info, ptr, ptr_rate, end_buf,
  						&iwe, IW_EV_PARAM_LEN);
  		}
  		rate >>= 1;
*************** acx_ioctl_get_scan(
*** 625,631 ****
  	for (i = 0; i < VEC_SIZE(adev->sta_list); i++) {
  		struct client *bss = &adev->sta_list[i];
  		if (!bss->used) continue;
! 		ptr = acx_s_scan_add_station(adev, ptr,
  			extra + IW_SCAN_MAX_DATA, bss);
  	}
  	dwrq->length = ptr - extra;
--- 635,641 ----
  	for (i = 0; i < VEC_SIZE(adev->sta_list); i++) {
  		struct client *bss = &adev->sta_list[i];
  		if (!bss->used) continue;
! 		ptr = acx_s_scan_add_station(adev, info, ptr,
  			extra + IW_SCAN_MAX_DATA, bss);
  	}
  	dwrq->length = ptr - extra;
