--- acx-20080112/pci.c.orig 
+++ acx-20080112/pci.c 
@@ -1632,8 +1632,10 @@
 	acx_show_card_eeprom_id(adev);
 #endif /* NONESSENTIAL_FEATURES */
 
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,24))
 #ifdef SET_MODULE_OWNER
-	SET_MODULE_OWNER(ndev);
+ 	SET_MODULE_OWNER(ndev);
+#endif
 #endif
 	SET_NETDEV_DEV(ndev, &pdev->dev);
 
@@ -2128,7 +2130,11 @@
 /* TODO: pci_set_power_state(pdev, PCI_D0); ? */
 
 	/* request shared IRQ handler */
+	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,21))	
+		if (request_irq(ndev->irq, acxpci_i_interrupt, IRQF_SHARED, ndev->name, ndev)) {
+	#else
 	if (request_irq(ndev->irq, acxpci_i_interrupt, SA_SHIRQ, ndev->name, ndev)) {
+	#endif		
 		printk("%s: request_irq FAILED\n", ndev->name);
 		result = -EAGAIN;
 		goto done;
--- acx-20080112/usb.c.orig
+++ acx-20080112/usb.c
@@ -859,7 +859,9 @@
 	ndev->watchdog_timeo = 4 * HZ;
 #endif
 	ndev->change_mtu = &acx_e_change_mtu;
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,24))
 	SET_MODULE_OWNER(ndev);
+#endif
 
 	/* Setup private driver context */
 
