diff -Naur acx-20060212.orig/common.c acx-20060212/common.c
--- acx-20060212.orig/common.c	2006-02-04 12:18:37.000000000 +0100
+++ acx-20060212/common.c	2006-11-30 23:10:05.000000000 +0100
@@ -30,7 +30,6 @@
 ** ---------------------------------------------------------------------
 */
 
-#include <linux/config.h>
 #include <linux/version.h>
 #include <linux/module.h>
 #include <linux/kernel.h>
diff -Naur acx-20060212.orig/conv.c acx-20060212/conv.c
--- acx-20060212.orig/conv.c	2006-02-01 10:49:35.000000000 +0100
+++ acx-20060212/conv.c	2006-11-30 23:09:55.000000000 +0100
@@ -30,7 +30,6 @@
 ** ---------------------------------------------------------------------
 */
 
-#include <linux/config.h>
 #include <linux/version.h>
 #include <linux/skbuff.h>
 #include <linux/if_arp.h>
diff -Naur acx-20060212.orig/ioctl.c acx-20060212/ioctl.c
--- acx-20060212.orig/ioctl.c	2006-02-01 10:49:39.000000000 +0100
+++ acx-20060212/ioctl.c	2006-11-30 23:10:00.000000000 +0100
@@ -30,7 +30,6 @@
 ** ---------------------------------------------------------------------
 */
 
-#include <linux/config.h>
 #include <linux/version.h>
 #include <linux/kernel.h>
 #include <linux/types.h>
diff -Naur acx-20060212.orig/pci.c acx-20060212/pci.c
--- acx-20060212.orig/pci.c	2006-02-11 13:28:43.000000000 +0100
+++ acx-20060212/pci.c	2006-11-30 23:10:15.000000000 +0100
@@ -31,8 +31,10 @@
 */
 #define ACX_PCI 1
 
-#include <linux/config.h>
 #include <linux/version.h>
+#ifndef UTS_RELEASE /* required for Lx 2.6.18 and above */
+#include <linux/utsrelease.h>
+#endif
 #include <linux/compiler.h> /* required for Lx 2.6.8 ?? */
 #include <linux/kernel.h>
 #include <linux/module.h>
diff -Naur acx-20060212.orig/usb.c acx-20060212/usb.c
--- acx-20060212.orig/usb.c	2006-02-11 13:28:08.000000000 +0100
+++ acx-20060212/usb.c	2006-11-30 23:10:19.000000000 +0100
@@ -45,7 +45,6 @@
 #define ACX_USB 1
 
 #include <linux/version.h>
-#include <linux/config.h>
 #include <linux/types.h>
 #include <linux/module.h>
 #include <linux/moduleparam.h>
diff -Naur acx-20060212.orig/wlan.c acx-20060212/wlan.c
--- acx-20060212.orig/wlan.c	2006-02-01 10:49:39.000000000 +0100
+++ acx-20060212/wlan.c	2006-11-30 23:09:37.000000000 +0100
@@ -37,7 +37,6 @@
 ** http://www.linux-wlan.com
 */
 
-#include <linux/config.h>
 #include <linux/version.h>
 #include <linux/types.h>
 #include <linux/if_arp.h>
