diff -Naur mozilla-1.9.1.orig/config/autoconf.mk.in mozilla-1.9.1/config/autoconf.mk.in
--- mozilla-1.9.1.orig/config/autoconf.mk.in	2009-03-06 03:01:44.000000000 +0530
+++ mozilla-1.9.1/config/autoconf.mk.in	2009-04-26 18:59:19.992997721 +0530
@@ -57,14 +57,14 @@
 prefix		= @prefix@
 exec_prefix	= @exec_prefix@
 bindir		= @bindir@
-includedir	= @includedir@/$(MOZ_APP_NAME)-$(MOZ_APP_VERSION)
+includedir	= @includedir@/$(MOZ_APP_NAME)-@PV@
 libdir		= @libdir@
 datadir		= @datadir@
 mandir		= @mandir@
-idldir		= $(datadir)/idl/$(MOZ_APP_NAME)-$(MOZ_APP_VERSION)
+idldir		= $(datadir)/idl/$(MOZ_APP_NAME)-@PV@
 
-installdir	= $(libdir)/$(MOZ_APP_NAME)-$(MOZ_APP_VERSION)
-sdkdir		= $(libdir)/$(MOZ_APP_NAME)-devel-$(MOZ_APP_VERSION)
+installdir	= $(libdir)/$(MOZ_APP_NAME)-@PV@
+sdkdir		= $(libdir)/$(MOZ_APP_NAME)-devel-@PV@
 
 DIST		= $(DEPTH)/dist
 LIBXUL_SDK      = @LIBXUL_SDK@
