---
 toolkit/mozapps/installer/packager.mk |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: mozilla/toolkit/mozapps/installer/packager.mk
===================================================================
--- mozilla.orig/toolkit/mozapps/installer/packager.mk
+++ mozilla/toolkit/mozapps/installer/packager.mk
@@ -419,17 +419,18 @@
 	  (cd $(DESTDIR)$(idldir)/stable && tar -xf -)
 	(cd $(DIST)/idl && tar $(TAR_CREATE_FLAGS) - .) | \
 	  (cd $(DESTDIR)$(idldir)/unstable && tar -xf -)
 # SDK directory is the libs + a bunch of symlinks
 	$(NSINSTALL) -D $(DESTDIR)$(sdkdir)/sdk/lib
 	if test -f $(DIST)/sdk/include/xpcom-config.h; then \
 	  $(SYSINSTALL) $(IFLAGS1) $(DIST)/sdk/include/xpcom-config.h $(DESTDIR)$(sdkdir); \
 	fi
-	(cd $(DIST)/sdk/lib && tar $(TAR_CREATE_FLAGS) - .) | (cd $(DESTDIR)$(sdkdir)/sdk/lib && tar -xf -)
+	(cd $(DIST)/sdk/lib && tar --exclude='*.so' $(TAR_CREATE_FLAGS) - .) | (cd $(DESTDIR)$(sdkdir)/sdk/lib && tar -xf -)
+	ln -s $(addprefix $(installdir)/,$(notdir $(wildcard $(DIST)/sdk/lib/*.so))) $(DESTDIR)$(sdkdir)/sdk/lib
 	$(RM) -f $(DESTDIR)$(sdkdir)/lib $(DESTDIR)$(sdkdir)/bin $(DESTDIR)$(sdkdir)/sdk/include $(DESTDIR)$(sdkdir)/include $(DESTDIR)$(sdkdir)/sdk/idl $(DESTDIR)$(sdkdir)/idl
 	ln -s $(sdkdir)/sdk/lib $(DESTDIR)$(sdkdir)/lib
 	ln -s $(installdir) $(DESTDIR)$(sdkdir)/bin
 	ln -s $(includedir)/stable $(DESTDIR)$(sdkdir)/sdk/include
 	ln -s $(includedir)/unstable $(DESTDIR)$(sdkdir)/include
 	ln -s $(idldir)/stable $(DESTDIR)$(sdkdir)/sdk/idl
 	ln -s $(idldir)/unstable $(DESTDIR)$(sdkdir)/idl
 endif # INSTALL_SDK
