Index: modules/zlib/standalone/Makefile.in
===================================================================
RCS file: /cvsroot/mozilla/modules/zlib/standalone/Makefile.in,v
retrieving revision 1.20
diff -u -p -8 -r1.20 Makefile.in
--- mozilla/modules/zlib/standalone/Makefile.in	31 Dec 2005 16:02:07 -0000	1.20
+++ mozilla/modules/zlib/standalone/Makefile.in	1 Jan 2008 23:27:42 -0000
@@ -43,17 +43,19 @@ VPATH		= @srcdir@
 
 include $(DEPTH)/config/autoconf.mk
 include $(srcdir)/../src/objs.mk
 
 MODULE		= zlib
 LIBRARY_NAME	= mozz_s
 
 CSRCS		= $(MODULES_ZLIB_SRC_LCSRCS)
+ifndef MOZ_NATIVE_ZLIB
 EXPORTS		= $(MODULES_ZLIB_SRC_EXPORTS)
+endif
 
 FORCE_STATIC_LIB = 1
 SRCS_IN_OBJDIR	= 1
 
 ifeq ($(OS_ARCH),WINNT)
 USE_STATIC_LIBS=1
 endif
 
