commit a1edaa30b3df51ad3c5fc52548aef6b8aed8d693
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Wed Mar 9 23:12:56 2011 +0100

    entropy: drop entropy.tools deprecated functions usage

diff --git a/backends/entropy/entropyBackend.py b/backends/entropy/entropyBackend.py
index 61d6e82..0487b12 100755
--- a/backends/entropy/entropyBackend.py
+++ b/backends/entropy/entropyBackend.py
@@ -72,6 +72,7 @@ except ImportError:
 from entropy.fetchers import UrlFetcher
 
 import entropy.tools
+import entropy.dep
 
 PK_DEBUG = False
 
@@ -627,7 +628,7 @@ class PackageKitEntropyMixin(object):
             package.prepare(match, "fetch", metaopts)
             myrepo = package.pkgmeta['repository']
             obj = down_data.setdefault(myrepo, set())
-            obj.add(entropy.tools.dep_getkey(package.pkgmeta['atom']))
+            obj.add(entropy.dep.dep_getkey(package.pkgmeta['atom']))
 
             x_rc = package.run()
             if x_rc != 0:
