commit 37b95f785a1f5e342238a6295738622d624e7950
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Wed Dec 29 09:50:12 2010 +0100

    entropy: fix compatibility with newer Entropy releases

diff --git a/backends/entropy/entropyBackend.py b/backends/entropy/entropyBackend.py
index b06b550..2ce2489 100755
--- a/backends/entropy/entropyBackend.py
+++ b/backends/entropy/entropyBackend.py
@@ -1168,8 +1168,8 @@ class PackageKitEntropyBackend(PackageKitBaseBackend, PackageKitEntropyMixin):
 
             updates = []
             keyslot = c_repo.retrieveKeySlotAggregated(pkg_id)
-            matches, m_rc = self._entropy.atom_match(keyslot, multiMatch = True,
-                multiRepo = True)
+            matches, m_rc = self._entropy.atom_match(keyslot, multi_match = True,
+                multi_repo = True)
             for m_pkg_id, m_repo_id in matches:
                 if (m_pkg_id, m_repo_id) == (pkg_id, repo_name):
                     continue # fliter myself
