--- a/src/linux.c.old	2012-01-12 08:18:58.525502553 -0500
+++ b/src/linux.c	2012-01-12 08:19:29.786069277 -0500
@@ -1671,6 +1671,8 @@ void get_battery_stuff(char *buffer, uns
  			the tradition! */
  			else if (strncmp(buf, "POWER_SUPPLY_CURRENT_NOW=", 25) == 0)
  				sscanf(buf, "POWER_SUPPLY_CURRENT_NOW=%d", &present_rate);
+			else if (strncmp(buf, "POWER_SUPPLY_POWER_NOW=", 23) == 0)
+				sscanf(buf, "POWER_SUPPLY_POWER_NOW=%d", &present_rate);
  			else if (strncmp(buf, "POWER_SUPPLY_ENERGY_NOW=", 24) == 0)
  				sscanf(buf, "POWER_SUPPLY_ENERGY_NOW=%d", &remaining_capacity);
  			else if (strncmp(buf, "POWER_SUPPLY_ENERGY_FULL=", 25) == 0)
