--- openswan-2.6.35.orig/lib/libcrypto/libmd5/Makefile	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/lib/libcrypto/libmd5/Makefile	2011-08-18 15:57:04.324257675 +0200
@@ -12,7 +12,7 @@ LIB=libmd5.a
 
 ifeq ($(USE_LIBNSS),true)
 CFLAGS+=-DHAVE_LIBNSS
-CFLAGS+=-I/usr/include/nspr4 -I/usr/include/nss3
+CFLAGS+=-I/usr/include/nspr -I/usr/include/nss
 endif
 
 include ${srcdir}../../Makefile.library
--- openswan-2.6.35.orig/lib/libcrypto/libsha1/Makefile	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/lib/libcrypto/libsha1/Makefile	2011-08-18 15:57:04.328257678 +0200
@@ -12,7 +12,7 @@ LIB=libsha1.a
 
 ifeq ($(USE_LIBNSS),true)
 CFLAGS+=-DHAVE_LIBNSS  
-CFLAGS+=-I/usr/include/nspr4 -I/usr/include/nss3
+CFLAGS+=-I/usr/include/nspr -I/usr/include/nss
 endif
 
 
--- openswan-2.6.35.orig/lib/libcrypto/libsha2/Makefile	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/lib/libcrypto/libsha2/Makefile	2011-08-18 15:57:04.328257678 +0200
@@ -15,5 +15,5 @@ include ${srcdir}../../Makefile.library
 
 ifeq ($(USE_LIBNSS),true)
 CFLAGS+=-DHAVE_LIBNSS
-INCLUDES+=-I/usr/include/nspr4 -I/usr/include/nss3
+INCLUDES+=-I/usr/include/nspr -I/usr/include/nss
 endif
--- openswan-2.6.35.orig/lib/libdns/Makefile	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/lib/libdns/Makefile	2011-08-18 15:57:04.328257678 +0200
@@ -31,7 +31,7 @@ CINCLUDES =	-I${srcdir}${ISCARCH}/includ
 		-I. -I./include -I${srcdir}include \
 		-I${srcdir}../libisc/include
 CDEFINES = ${USERCOMPILE} ${PORTINCLUDE}
-CWARNINGS = -Werror
+CWARNINGS =
 
 CFLAGS+=${CINCLUDES} ${CDEFINES} ${CWARNINGS}
 
--- openswan-2.6.35.orig/lib/libipsecconf/Makefile	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/lib/libipsecconf/Makefile	2011-08-18 15:57:04.328257678 +0200
@@ -33,7 +33,7 @@ SRCS+=interfaces.c
 #CFLAGS+=-DFIPS_CHECK
 ifeq ($(USE_LIBNSS),true)
 CFLAGS+=-DHAVE_LIBNSS
-CFLAGS+=-I/usr/include/nspr4 -I/usr/include/nss3
+CFLAGS+=-I/usr/include/nspr -I/usr/include/nss
 endif
 
 ifeq ($(USE_NM),true)
--- openswan-2.6.35.orig/lib/libisc/Makefile	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/lib/libisc/Makefile	2011-08-18 15:57:04.329257678 +0200
@@ -23,7 +23,7 @@ CINCLUDES =	-I${srcdir}${ISCARCH}/includ
 		-I. -I./include -I${srcdir}include \
 		$(PORTINCLUDE) -I${OPENSWANSRCDIR}/include
 CDEFINES = -DHAVE_STRERROR ${USERCOMPILE} ${PORTINCLUDE}
-# CWARNINGS = -Werror
+CWARNINGS =
 
 CFLAGS+=${CINCLUDES} ${CDEFINES} ${CWARNINGS}
 
--- openswan-2.6.35.orig/lib/liblwres/Makefile	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/lib/liblwres/Makefile	2011-08-18 15:57:04.329257678 +0200
@@ -23,7 +23,7 @@ CINCLUDES =	-I${srcdir}${ISCARCH}/includ
 		-I${srcdir}../libisc/include \
 		-I. -I./include -I${srcdir}include 
 CDEFINES =
-CWARNINGS = -Werror
+CWARNINGS =
 
 CFLAGS+=${USERCOMPILE} ${PORTINCLUDE} ${CINCLUDES} ${CDEFINES} ${CWARNINGS} 
 
--- openswan-2.6.35.orig/lib/libopenswan/Makefile	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/lib/libopenswan/Makefile	2011-08-18 15:57:04.329257678 +0200
@@ -105,7 +105,7 @@ endif
 
 ifeq ($(USE_LIBNSS),true)
 CFLAGS+=-DHAVE_LIBNSS
-CFLAGS+=-I/usr/include/nspr4 -I/usr/include/nss3
+CFLAGS+=-I/usr/include/nspr -I/usr/include/nss
 endif
 
 ifeq ($(USE_MODP_RFC5114),true)
--- openswan-2.6.35.orig/Makefile.inc	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/Makefile.inc	2011-08-18 15:57:04.330257678 +0200
@@ -49,7 +49,7 @@ SHELL=/bin/sh
 DESTDIR?=
 
 # "local" part of tree, used in building other pathnames
-INC_USRLOCAL=/usr/local
+INC_USRLOCAL?=/usr
 
 # PUBDIR is where the "ipsec" command goes; beware, many things define PATH
 # settings which are assumed to include it (or at least, to include *some*
--- openswan-2.6.35.orig/packaging/makefiles/module.make	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/packaging/makefiles/module.make	2011-08-18 15:57:04.330257678 +0200
@@ -25,7 +25,6 @@ EXTRA_CFLAGS += -DDISABLE_UDP_CHECKSUM
 
 EXTRA_CFLAGS += $(KLIPSCOMPILE)
 EXTRA_CFLAGS += -Wall -DIPCOMP_PREFIX
-#EXTRA_CFLAGS += -Werror
 #EXTRA_CFLAGS += -Wconversion 
 #EXTRA_CFLAGS += -Wmissing-prototypes 
 # 'override CFLAGS' should really be 'EXTRA_CFLAGS'
--- openswan-2.6.35.orig/programs/Makefile.program	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/programs/Makefile.program	2011-08-18 15:57:04.331257678 +0200
@@ -48,12 +48,7 @@ CFLAGS+=-DFINALCONFDDIR=\"${FINALCONFDDI
 CFLAGS+=-DFINALCONFFILE=\"${FINALCONFFILE}\"
 CFLAGS+=-DFINALVARDIR=\"${FINALVARDIR}\"
 
-# die if there are any warnings
-ifndef WERROR
-WERROR:= -Werror
-endif
 
-CFLAGS+= ${WERROR}
 
 ifneq ($(LD_LIBRARY_PATH),)
 LDFLAGS=-L$(LD_LIBRARY_PATH)
@@ -104,8 +99,6 @@ endif
 ifneq ($(NOINSTALL),true)
 
 doinstall:: $(PROGRAM) $(CONFFILES) $(EXTRA8MAN) $(EXTRA5MAN) $(EXTRA5PROC) $(LIBFILES) $(CONFDFILES)
-	# remove any old vendor file installed previously
-	@rm -f $(FINALLIBEXECDIR)/vendor
 	@mkdir -p $(PROGRAMDIR) $(MANDIR8) $(MANDIR5) $(LIBDIR) $(CONFDIR) $(CONFDDIR) $(CONFDDIR)/$(CONFDSUBDIR) $(EXAMPLECONFDIR)
 	@if [ -n "$(PROGRAM)" ]; then $(INSTALL) $(INSTBINFLAGS) $(PROGRAM) $(PROGRAMDIR); fi
 	@$(foreach f, $(addsuffix .8, $(PROGRAM)), \
@@ -180,6 +173,7 @@ ifeq ($(USE_LIBNSS),false)
 LDLIBS=${LIBS} ${USERLINK} ${LIBS} ${EXTRALIBS} -lgmp
 else
 LDLIBS=${LIBS} ${USERLINK} ${LIBS} ${EXTRALIBS} -lgmp -lnspr4 -lnss3
+LDFLAGS+=-L/usr/lib/nspr -L/usr/lib/nss
 endif
 
 
--- openswan-2.6.35.orig/programs/pluto/Makefile.options	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/programs/pluto/Makefile.options	2011-08-18 15:57:04.331257678 +0200
@@ -44,7 +44,7 @@ endif
 
 # -O on Linux makes gcc coredump when compiling sha1.c
 # -Wundef is nice but RHL5.2 compiler doesn't support it
-CFLAGS +=-g -Wall -W -Wmissing-prototypes -Wpointer-arith -Wbad-function-cast \
+CFLAGS +=-Wall -W -Wmissing-prototypes -Wpointer-arith -Wbad-function-cast \
 	-Wcast-qual -Wmissing-declarations -Wwrite-strings \
 	-Wstrict-prototypes # -Wundef
 
@@ -58,7 +58,7 @@ HDRDIRS = -I${OPENSWANSRCDIR}/programs/p
 
 # Use LIBNSS
 ifeq ($(USE_LIBNSS),true)
-HDRDIRS+= -I/usr/include/nspr4 -I/usr/include/nss3	
+HDRDIRS+= -I/usr/include/nspr -I/usr/include/nss	
 endif
 
 # BYTE_ORDER = -DBIG_ENDIAN=4321 -DLITTLE_ENDIAN=1234 -DBYTE_ORDER=BIG_ENDIAN
@@ -315,6 +315,7 @@ HAVE_LIBNSS=1
 DEFINES+=-DHAVE_LIBNSS
 LIBSPLUTO+= -lnspr4
 LIBSPLUTO+= -lnss3
+LDFLAGS+=-L/usr/lib/nspr -L/usr/lib/nss
 # For avoiding implicit DSO linking
 LIBSPLUTO+= -lpthread
 ifeq ($(USE_FIPSCHECK),true)
--- openswan-2.6.35.orig/programs/rsasigkey/Makefile	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/programs/rsasigkey/Makefile	2011-08-18 15:57:04.332257678 +0200
@@ -21,13 +21,14 @@ ifeq ($(USE_FIPSCHECK),true)
 FIPS_CHECK=1
 CFLAGS+=-DFIPS_CHECK
 endif
-INCLUDES+=-I/usr/include/nspr4 -I/usr/include/nss3
+INCLUDES+=-I/usr/include/nspr -I/usr/include/nss
 endif
 
 PROGRAM=rsasigkey
 LIBS=${OPENSWANLIB} -lgmp
 ifeq ($(USE_LIBNSS),true)
 LIBS+= -lnspr4 -lnss3 -lplc4
+LDFLAGS+=-L/usr/lib/nspr -L/usr/lib/nss
 ifeq ($(USE_FIPSCHECK),true)
 LIBS+=  -lfipscheck
 endif
--- openswan-2.6.35.orig/programs/setup/Makefile	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/programs/setup/Makefile	2011-08-18 15:57:04.332257678 +0200
@@ -16,11 +16,7 @@
 OPENSWANSRCDIR?=$(shell cd ../..; pwd)
 include ${OPENSWANSRCDIR}/Makefile.inc
 
-# this dance is because setup has to get installed as /etc/rc.d/init.d/ipsec
-# not as /etc/rc.d/init.d/setup.
-PROGRAMDIR=$(RCDIR)
 PROGRAM=setup
-NOINSTALL=true
 EXTRA8MAN=setup.8
 
 include ${srcdir}../Makefile.program
@@ -34,13 +30,11 @@ doinstall:: $(PROGRAM) $(CONFFILES) $(EX
 	# remove any old vendor file installed previously
 	@rm -f $(BINDIR)/setup
 	# check the directories we need exist (ie., make install DESTDIR=/tmp/IPsec)
-	@mkdir -p $(RCDIR) $(BINDIR)
+	@mkdir -p $(BINDIR)
 	# install and link everything
-	@$(INSTALL) $(INSTBINFLAGS) setup $(RCDIR)/ipsec
-	@ln -s $(FINALRCDIR)/ipsec $(BINDIR)/setup
+	@$(INSTALL) $(INSTBINFLAGS) setup $(BINDIR)/setup
 
 install_file_list::
-	@echo $(RCDIR)/ipsec
 	@echo $(BINDIR)/setup
 
 cleanall::
--- openswan-2.6.35.orig/programs/showhostkey/Makefile	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/programs/showhostkey/Makefile	2011-08-18 15:57:04.332257678 +0200
@@ -16,7 +16,7 @@ OPENSWANSRCDIR?=$(shell cd ../..; pwd)
 include ${OPENSWANSRCDIR}/Makefile.inc
 ifeq ($(USE_LIBNSS),true)
 CFLAGS+=-DHAVE_LIBNSS
-INCLUDES+=-I/usr/include/nspr4 -I/usr/include/nss3
+INCLUDES+=-I/usr/include/nspr -I/usr/include/nss
 endif
 
 PROGRAM=showhostkey
--- openswan-2.6.35.orig/testing/utils/make-uml.sh	2011-07-23 22:53:13.000000000 +0200
+++ openswan-2.6.35/testing/utils/make-uml.sh	2011-08-18 15:57:43.132257653 +0200
@@ -254,7 +254,7 @@ fi
 
 cd $OPENSWANSRCDIR || exit 1
 
-make ${WERROR:-WERROR=-Werror} USE_OBJDIR=true programs
+make USE_OBJDIR=true programs
 
 # now, execute the Makefile that we have created!
 cd $POOLSPACE && make $OPENSWANHOSTS 
