diff -Nru openswan-2.6.23.orig/programs/pluto/Makefile openswan-2.6.23/programs/pluto/Makefile
--- openswan-2.6.23.orig/programs/pluto/Makefile	2009-09-09 02:42:54.000000000 +0200
+++ openswan-2.6.23/programs/pluto/Makefile	2009-09-11 19:27:47.000000000 +0200
@@ -43,6 +43,7 @@
 # End of configuration coping options.
 
 CPPFLAGS = $(HDRDIRS) $(DEFINES) \
+	-DALLOW_MICROSOFT_BAD_PROPOSAL \
 	-DSHARED_SECRETS_FILE=\"${FINALCONFDIR}/ipsec.secrets\" \
 	-DPOLICYGROUPSDIR=\"${FINALCONFDDIR}/policies\" \
 	-DPERPEERLOGDIR=\"${FINALLOGDIR}/pluto/peer\"
