--- nm-applet-dbus-126.conf	2009-03-31 11:10:19.000000000 -0500
+++ nm-applet-dbus-126.conf.new	2009-04-15 05:33:15.000000000 -0500
@@ -23,7 +23,7 @@
                 <allow send_destination="org.freedesktop.NetworkManagerUserSettings"
                        send_interface="org.freedesktop.NetworkManagerSettings.Connection.Secrets"/>
         </policy>
-        <policy at_console="true">
+        <policy group="plugdev">
                 <allow own="org.freedesktop.NetworkManagerUserSettings"/>
 
                 <allow send_destination="org.freedesktop.NetworkManagerUserSettings"
