From 1ae0b66f870e2c1bb97d8034544c4f1f46aeb2dd Mon Sep 17 00:00:00 2001
From: Fabio Erculiani <lxnay@sabayon.org>
Date: Sun, 12 Jun 2016 14:47:46 +0200
Subject: [PATCH 3/3] Call udev.settle() when committing to disk.

---
 blivet/formats/disklabel.py | 1 +
 1 file changed, 1 insertion(+)

diff --git a/blivet/formats/disklabel.py b/blivet/formats/disklabel.py
index ed42bef..676de82 100644
--- a/blivet/formats/disklabel.py
+++ b/blivet/formats/disklabel.py
@@ -283,6 +283,7 @@ class DiskLabel(DeviceFormat):
             raise DiskLabelCommitError(msg)
         else:
             self.updateOrigPartedDisk()
+            udev.settle()
 
     def addPartition(self, start, end, ptype=None):
         """ Add a partition to the disklabel.
-- 
2.7.4

