diff -up nautilus-share-0.7.2/src/nautilus-share.c.bak nautilus-share-0.7.2/src/nautilus-share.c
--- nautilus-share-0.7.2/src/nautilus-share.c.bak	2008-11-12 15:56:37.000000000 -0800
+++ nautilus-share-0.7.2/src/nautilus-share.c	2008-11-12 15:57:27.000000000 -0800
@@ -1128,7 +1128,7 @@ nautilus_share_get_file_items (NautilusM
   item = nautilus_menu_item_new ("NautilusShare::share",
 				 _("Sharing Options"),
 				 _("Share this Folder"),
-				 "stock_shared-by-me");
+				 "folder-remote");
   g_signal_connect (item, "activate",
 		    G_CALLBACK (share_this_folder_callback),
 		    fileinfo);
