diff -up nautilus-share-0.7.2/src/Makefile.am.bak nautilus-share-0.7.2/src/Makefile.am
--- nautilus-share-0.7.2/src/Makefile.am.bak	2008-11-12 16:00:55.000000000 -0800
+++ nautilus-share-0.7.2/src/Makefile.am	2008-11-12 16:01:38.000000000 -0800
@@ -11,7 +11,7 @@ libnautilus_share_la_CFLAGS =				\
 	$(GLIB_CFLAGS)
 
 # FIXME: get this from the .pc file
-nautilus_extensiondir=$(libdir)/nautilus/extensions-1.0
+nautilus_extensiondir=$(libdir)/nautilus/extensions-2.0
 
 nautilus_extension_LTLIBRARIES=libnautilus-share.la
 
diff -up nautilus-share-0.7.2/src/Makefile.in.bak nautilus-share-0.7.2/src/Makefile.in
--- nautilus-share-0.7.2/src/Makefile.in.bak	2008-11-12 16:01:03.000000000 -0800
+++ nautilus-share-0.7.2/src/Makefile.in	2008-11-12 16:01:55.000000000 -0800
@@ -245,7 +245,7 @@ libnautilus_share_la_CFLAGS = \
 
 
 # FIXME: get this from the .pc file
-nautilus_extensiondir = $(libdir)/nautilus/extensions-1.0
+nautilus_extensiondir = $(libdir)/nautilus/extensions-2.0
 nautilus_extension_LTLIBRARIES = libnautilus-share.la
 libnautilus_share_la_SOURCES = \
 	nautilus-share.c        \
