--- screenfetch-dev~	2015-05-21 10:38:27.187030904 +0200
+++ screenfetch-dev	2015-05-21 10:39:50.374805454 +0200
@@ -435,8 +435,8 @@
 				distro_codename=null
 			elif [[ "${distro_detect}" == "CentOS" || "${distro_detect}" =~ "RedHatEnterprise" ]]; then
 				distro="Red Hat Enterprise Linux"
-			elif [[ "${distro_detect}" == "Sabayon" ]]; then
-				distro="Sabayon"
+			elif [[ "${distro_detect}" == "sabayon" ]]; then
+				distro="sabayon"
 			elif [[ "${distro_detect}" == "SolusOS" ]]; then
 				distro="SolusOS"
 			elif [[ "${distro_detect}" == "Trisquel" ]]; then
@@ -540,7 +540,7 @@
 					elif [ -f /etc/redhat-release ] && grep -q "CentOS" /etc/redhat-release; then distro="CentOS"
 					elif [ -f /etc/slackware-version ]; then distro="Slackware"
 					elif [ -f /usr/share/doc/tc/release.txt ]; then distro="TinyCore"
-					elif [ -f /etc/sabayon-edition ]; then distro="Sabayon"; fi
+					elif [ -f /etc/sabayon-edition ]; then distro="sabayon"; fi
 				else
 					if [[ -x /usr/bin/sw_vers ]] && /usr/bin/sw_vers | grep -i "Mac OS X" >/dev/null; then
 						distro="Mac OS X"
@@ -725,13 +725,13 @@
 
 # Package Count - Begin
 detectpkgs () {
-	pkgs="Unknown"
+	pkgs=""
 	case "${distro}" in
 		'Arch Linux'|'Parabola GNU/Linux-libre'|'Chakra'|'Manjaro'|'Antergos'|'KaOS') pkgs=$(pacman -Qq | wc -l) ;;
 		'Frugalware') pkgs=$(pacman-g2 -Q | wc -l) ;;
 		'Fuduntu'|'Ubuntu'|'Mint'|'SolusOS'|'Debian'|'Raspbian'|'LMDE'|'CrunchBang'|'Peppermint'|'LinuxDeepin'|'Deepin'|'Kali Linux'|'Trisquel'|'elementary OS') pkgs=$(dpkg --get-selections | wc -l) ;;
 		'Slackware') pkgs=$(ls -1 /var/log/packages | wc -l) ;;
-		'Gentoo'|'Sabayon'|'Funtoo') pkgs=$(ls -d /var/db/pkg/*/* | wc -l) ;;
+		'Gentoo'|'Funtoo'|'sabayon') pkgs=$(ls -d /var/db/pkg/*/* | wc -l) ;;
 		'Fedora'|'Korora'|'openSUSE'|'Red Hat Enterprise Linux'|'CentOS'|'Mandriva'|'Mandrake'|'Mageia'|'Viperr') pkgs=$(rpm -qa | wc -l) ;;
 		'Void') pkgs=$(xbps-query -l|wc -l);;
 		'Evolve OS') pkgs=$(pisi list-installed | wc -l);;
@@ -3077,28 +3077,29 @@
 "${c1}                                             .")
 		;;
 
-                "Sabayon")
+                "sabayon")
                         if [[ "$no_color" != "1" ]]; then
                                 c1=$(getColor 'white') # White
-				c2=$(getColor 'light blue') # Blue
+				c2=$(getColor 'light grey') # Red
                         fi
                         if [ -n "${my_lcolor}" ]; then c1="${my_lcolor}"; fi
                         startline="0"
-                        fulloutput=("${c2}            ...........                %s"
+                        fulloutput=(
+"${c2}            ...........                %s"
 "${c2}         ..             ..            %s"
 "${c2}      ..                   ..         %s"
-"${c2}    ..           ${c1}o           ${c2}..       %s"
-"${c2}  ..            ${c1}:W'            ${c2}..     %s"
-"${c2} ..             ${c1}.d.             ${c2}..    %s"
-"${c2}:.             ${c1}.KNO              ${c2}.:   %s"
-"${c2}:.             ${c1}cNNN.             ${c2}.:   %s"
-"${c2}:              ${c1}dXXX,              ${c2}:   %s"
-"${c2}:   ${c1}.          dXXX,       .cd,   ${c2}:   %s"
-"${c2}:   ${c1}'kc ..     dKKK.    ,ll;:'    ${c2}:   %s"
-"${c2}:     ${c1}.xkkxc;..dkkkc',cxkkl       ${c2}:   %s"
-"${c2}:.     ${c1}.,cdddddddddddddo:.       ${c2}.:   %s"
-"${c2} ..         ${c1}:lllllll:           ${c2}..    %s"
-"${c2}   ..         ${c1}',,,,,          ${c2}..      %s"
+"${c2}    ..           ${c1}:           ${c2}..       %s"
+"${c2}  ..            ${c1}.:.            ${c2}..     %s"
+"${c2} ..             ${c1}.:.             ${c2}..    %s"
+"${c2}:.             ${c1}.:::.             ${c2}.:   %s"
+"${c2}:.             ${c1}.:::.             ${c2}.:   %s"
+"${c2}:              ${c1}.:::.          :   ${c2}:   %s"
+"${c2}:   ${c1}:          .:::.        ...   ${c2}:   %s"
+"${c2}:   ${c1}....       .:::.    ......    ${c2}:   %s"
+"${c2}:     ${c1}.....................       ${c2}:   %s"
+"${c2}:.     ${c1}...................       ${c2}.:   %s"
+"${c2} ..         ${c1}..........           ${c2}..   %s"
+"${c2}   ..         ${c1}......          ${c2}..      %s"
 "${c2}     ..                     ..        %s"
 "${c2}        ..               ..           %s"
 "${c2}          ...............             %s")
@@ -3365,7 +3366,7 @@
 	myascii="${distro}"
 	[[ "${asc_distro}" ]] && myascii="${asc_distro}"
 	case ${myascii} in
-		"Arch Linux - Old"|"Fedora"|"Korora"|"Mandriva"|"Mandrake"|"Chakra"|"Sabayon"|"Slackware"|"Mac OS X"|"Trisquel"|"Kali Linux"|"Jiyuu Linux"|"Antergos"|"KaOS") labelcolor=$(getColor 'light blue');;
+		"Arch Linux - Old"|"Fedora"|"Korora"|"Mandriva"|"Mandrake"|"Chakra"|"sabayon"|"Slackware"|"Mac OS X"|"Trisquel"|"Kali Linux"|"Jiyuu Linux"|"Antergos"|"KaOS") labelcolor=$(getColor 'light blue');;
 		"Arch Linux"|"Frugalware"|"Mageia"|"Deepin") labelcolor=$(getColor 'light cyan');;
 		"Mint"|"LMDE"|"openSUSE"|"LinuxDeepin"|"DragonflyBSD"|"Manjaro"|"Manjaro-tree"|"Android"|"Void") labelcolor=$(getColor 'light green');;
 		"Ubuntu-3color"|"FreeBSD"|"FreeBSD - Old"|"Debian"|"Raspbian"|"BSD"|"Red Hat Enterprise Linux"|"Peppermint"|"Cygwin"|"Fuduntu"|"NetBSD"|"Scientific Linux"|"DragonFlyBSD"|"BackTrack Linux") labelcolor=$(getColor 'light red');;
