diff --git a/screenfetch-dev b/screenfetch-dev
index 94c412b..d1267ad 100755
--- a/screenfetch-dev
+++ b/screenfetch-dev
@@ -467,6 +467,8 @@ detectdistro () {
 				"Gentoo")
 					if [[ "$(lsb_release -sd)" =~ "Funtoo" ]]; then
 						distro="Funtoo"
+					elif [[ "$(lsb_release -sd)" =~ "Sabayon" ]]; then
+						distro="Sabayon"
 					else
 						distro="Gentoo"
 					fi
@@ -4319,27 +4321,28 @@ asciiText () {
 		"Sabayon")
 			if [[ "$no_color" != "1" ]]; then
 				c1=$(getColor 'white') # White
-				c2=$(getColor 'light blue') # Blue
+				c2=$(getColor 'light grey') # Red
 			fi
 			if [ -n "${my_lcolor}" ]; then c1="${my_lcolor}"; c2="${my_lcolor}"; fi
 			startline="0"
 			logowidth="38"
+
 			fulloutput=(
 "${c2}            ...........               %s"
 "${c2}         ..             ..            %s"
 "${c2}      ..                   ..         %s"
-"${c2}    ..           ${c1}o           ${c2}..       %s"
-"${c2}  ..            ${c1}:W'            ${c2}..     %s"
-"${c2} ..             ${c1}.d.             ${c2}..    %s"
-"${c2}:.             ${c1}.KNO              ${c2}.:   %s"
-"${c2}:.             ${c1}cNNN.             ${c2}.:   %s"
-"${c2}:              ${c1}dXXX,              ${c2}:   %s"
-"${c2}:   ${c1}.          dXXX,       .cd,   ${c2}:   %s"
-"${c2}:   ${c1}'kc ..     dKKK.    ,ll;:'    ${c2}:   %s"
-"${c2}:     ${c1}.xkkxc;..dkkkc',cxkkl       ${c2}:   %s"
-"${c2}:.     ${c1}.,cdddddddddddddo:.       ${c2}.:   %s"
-"${c2} ..         ${c1}:lllllll:           ${c2}..    %s"
-"${c2}   ..         ${c1}',,,,,          ${c2}..      %s"
+"${c2}    ..           ${c1}:           ${c2}..       %s"
+"${c2}  ..            ${c1}.:.            ${c2}..     %s"
+"${c2} ..             ${c1}.:.             ${c2}..    %s"
+"${c2}:.             ${c1}.:::.             ${c2}.:   %s"
+"${c2}:.             ${c1}.:::.             ${c2}.:   %s"
+"${c2}:              ${c1}.:::.          :   ${c2}:   %s"
+"${c2}:   ${c1}:          .:::.        ...   ${c2}:   %s"
+"${c2}:   ${c1}....       .:::.    ......    ${c2}:   %s"
+"${c2}:     ${c1}.....................       ${c2}:   %s"
+"${c2}:.     ${c1}...................       ${c2}.:   %s"
+"${c2} ..         ${c1}..........           ${c2}..   %s"
+"${c2}   ..         ${c1}......          ${c2}..      %s"
 "${c2}     ..                     ..        %s"
 "${c2}        ..               ..           %s"
 "${c2}          ...............             %s")
