--- a/configure.ac	2008-09-26 14:46:36.000000000 +0000
+++ b/configure.ac	2008-09-26 14:48:50.000000000 +0000
@@ -127,7 +127,7 @@
 AM_CONDITIONAL(GROUP_PLUGIN, test "x$use_group" = "xyes")
 
 PKG_CHECK_MODULES(SCALEFILTER, compiz-text compiz-scale, [use_scalefilter=yes], [use_scalefilter=no])
-AM_CONDITIONAL(SCALEFILTER_PLUGIN, text "x$use_scalefilter" = "xyes")
+AM_CONDITIONAL(SCALEFILTER_PLUGIN, test "x$use_scalefilter" = "xyes")
 
 PKG_CHECK_MODULES(COMPIZCUBE, compiz-cube, [have_compiz_cube=yes], [have_compiz_cube=no])
 AM_CONDITIONAL(CUBEADDON_PLUGIN, test "x$have_compiz_cube" = "xyes")
