From 19f8761365b825ae9949559048afd90c2f662de8 Mon Sep 17 00:00:00 2001
From: Roy Marples <roy@uberpc.marples.name>
Date: Fri, 3 Oct 2008 18:50:30 +0100
Subject: [PATCH] kenv is not always available.

---
 init.d/hostid.in |    8 +++++---
 1 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/init.d/hostid.in b/init.d/hostid.in
index a058884..584fd2d 100644
--- a/init.d/hostid.in
+++ b/init.d/hostid.in
@@ -38,9 +38,11 @@ _set()
 # otherwise we generate a random UUID.
 reset()
 {
-	local uuid=$(kenv smbios.system.uuid 2>/dev/null)
-	local x="[0-9a-f]"
-	local y="${x}${x}${x}${x}"
+	local uuid= x="[0-9a-f]" y="${x}${x}${x}${x}"
+
+	if type kenv >/dev/null 2>&1; then
+		uuid=$(kenv smbios.system.uuid 2>/dev/null)
+	fi
 	case "${uuid}" in
 		${y}${y}-${y}-${y}-${y}-${y}${y}${y});;
 		*) uuid=;;
-- 
1.6.0.2

