From a327a7d1e795fb6d6c49a459f5863d9d1fdc0b9c Mon Sep 17 00:00:00 2001
From: Roy Marples <roy@marples.name>
Date: Wed, 8 Oct 2008 14:58:17 +0100
Subject: [PATCH] Fix utmp permissions, Gentoo #240437.

---
 init.d/bootmisc.in |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/init.d/bootmisc.in b/init.d/bootmisc.in
index 50e83a0..529097e 100644
--- a/init.d/bootmisc.in
+++ b/init.d/bootmisc.in
@@ -77,10 +77,13 @@ start()
 	if dir_writeable /var/run; then
 		ebegin "Creating user login records"
 		cp /dev/null /var/run/utmp
+		chgrp utmp /var/run/utmp
+		chmod 0664 /var/run/utmp
 		if dir_writeable /var/log; then
 			logw=true
 			[ -e /var/log/wtmp ] || cp /dev/null /var/log/wtmp
-			chmod 0644 /var/run/utmp /var/log/wtmp
+			chgrp utmp /var/log/wtmp
+			chmod 0664 /var/log/wtmp
 		fi
 		eend 0
 
-- 
1.6.0.2

