From 70999050b4786f620f9a52443348b87411f3c738 Mon Sep 17 00:00:00 2001
From: Fabio Erculiani <lxnay@sabayon.org>
Date: Wed, 17 Jul 2013 13:31:17 +0200
Subject: [PATCH] sabayon: g_type_init has been deprecated in glib-2.35+

---
 isys/iface.c        | 6 ++++++
 loader/net.c        | 2 ++
 loader/nfsinstall.c | 2 ++
 3 files changed, 10 insertions(+)

diff --git a/isys/iface.c b/isys/iface.c
index 5897286..d740665 100644
--- a/isys/iface.c
+++ b/isys/iface.c
@@ -176,7 +176,9 @@ char *iface_ip2str(char *ifname, int family) {
         return NULL;
     }
 
+#if !defined(GLIB_VERSION_2_36)
     g_type_init();
+#endif
 
     client = nm_client_new();
     if (!client) {
@@ -412,7 +414,9 @@ gboolean is_nm_connected(void) {
     NMState state;
     NMClient *client = NULL;
 
+#if !defined(GLIB_VERSION_2_36)
     g_type_init();
+#endif
 
     client = nm_client_new();
     if (!client)
@@ -432,7 +436,9 @@ gboolean is_nm_running(void) {
     gboolean running;
     NMClient *client = NULL;
 
+#if !defined(GLIB_VERSION_2_36)
     g_type_init();
+#endif
 
     client = nm_client_new();
     if (!client)
diff --git a/loader/net.c b/loader/net.c
index eff4782..d4dbe5c 100644
--- a/loader/net.c
+++ b/loader/net.c
@@ -2066,7 +2066,9 @@ int get_connection(iface_t *iface) {
                   iface->device, 0);
     }
 
+#if !defined(GLIB_VERSION_2_36)
     g_type_init();
+#endif
 
     client = nm_client_new();
     if (!client) {
diff --git a/loader/nfsinstall.c b/loader/nfsinstall.c
index ae13874..11643af 100644
--- a/loader/nfsinstall.c
+++ b/loader/nfsinstall.c
@@ -419,7 +419,9 @@ int getFileFromNfs(char * url, char * dest, struct loaderData_s * loaderData) {
      * the dhcp/bootp information
      */
     if (!url) {
+#if !defined(GLIB_VERSION_2_36)
         g_type_init();
+#endif
 
         client = nm_client_new();
         if (!client) {
-- 
1.8.2.1

