From 9cc9c55488edd82cadc5250cbbdd82f5b0ed8c38 Mon Sep 17 00:00:00 2001
From: mudler <mudler@sabayon.org>
Date: Tue, 13 Sep 2016 10:15:41 +0200
Subject: [PATCH] bootloader.py: keep language and keyboard mapping settings

---
 pyanaconda/bootloader.py | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/pyanaconda/bootloader.py b/pyanaconda/bootloader.py
index 2ba2e6d..31eb8b4 100644
--- a/pyanaconda/bootloader.py
+++ b/pyanaconda/bootloader.py
@@ -256,7 +256,8 @@ class BootLoader(object):
                             "noht", "acpi", "video", "pci", "nodmraid",
                             "nompath", "nomodeset", "noiswmd", "fips",
                             "selinux", "biosdevname", "ipv6.disable",
-                            "net.ifnames"]
+                            "net.ifnames", "vconsole.keymap", "locale",
+                            "keymap"]
     preserve_args = []
 
     _trusted_boot = False
-- 
2.9.2

