From 63dcd22d43e713e484adb628492c2bfea80af672 Mon Sep 17 00:00:00 2001
From: mudler <mudler@sabayon.org>
Date: Fri, 23 Sep 2016 14:15:30 +0200
Subject: [PATCH] bootloader: preserve rd.vconsole.keymap

---
 pyanaconda/bootloader.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pyanaconda/bootloader.py b/pyanaconda/bootloader.py
index 981e09c..eb7b618 100644
--- a/pyanaconda/bootloader.py
+++ b/pyanaconda/bootloader.py
@@ -257,7 +257,7 @@ class BootLoader(object):
                             "nompath", "nomodeset", "noiswmd", "fips",
                             "selinux", "biosdevname", "ipv6.disable",
                             "net.ifnames", "vconsole.keymap", "locale",
-                            "keymap"]
+                            "rd.vconsole.keymap", "keymap"]
     preserve_args = []
 
     _trusted_boot = False
-- 
2.10.0

