From c1374619daf0c9fe43908f8aa58a162d677e89a3 Mon Sep 17 00:00:00 2001
From: mudler <mudler@sabayon.org>
Date: Sat, 15 Oct 2016 00:22:46 +0200
Subject: [PATCH] pyanaconda/nm: fixing 'No such property PermHwAddress'

---
 pyanaconda/nm.py | 26 ++++++++++++++++++++++++--
 1 file changed, 24 insertions(+), 2 deletions(-)

diff --git a/pyanaconda/nm.py b/pyanaconda/nm.py
index 26a198f..4a04ca6 100644
--- a/pyanaconda/nm.py
+++ b/pyanaconda/nm.py
@@ -382,6 +382,28 @@ def nm_device_perm_hwaddress(name):
     """
     return nm_device_property(name, "PermHwAddress")
 
+def nm_device_valid_hwaddress(name):
+    """Return valid hardware address of device depending on type of the device
+       ('PermHwAddress' property for wired and wireless or 'HwAddress' property for others)
+       :param name: name of device
+       :type name: str
+       :return: active hardware address of device
+                ('HwAddress' or 'PermHwAddress' property)
+       :rtype: str
+       :raise UnknownDeviceError: if device is not found
+       :raise PropertyNotFoundError: if property is not found
+    """
+    if nm_device_type_is_ethernet(name) or nm_device_type_is_wifi(name):
+        try:
+            return nm_device_perm_hwaddress(name)
+        except PropertyNotFoundError:
+            # TODO: Remove this if everything will work well
+            # fallback solution
+            log.warning("Device %s don't have property PermHwAddress", name)
+            return nm_device_hwaddress(name)
+    else:
+        return nm_device_hwaddress(name)
+
 def nm_device_active_con_uuid(name):
     """Return uuid of device's active connection
 
@@ -562,7 +584,7 @@ def nm_hwaddr_to_device_name(hwaddr):
         :rtype: str
     """
     for device in nm_devices():
-        if nm_device_perm_hwaddress(device).upper() == hwaddr.upper():
+        if nm_device_valid_hwaddress(device).upper() == hwaddr.upper():
             return device
     return None
 
@@ -614,7 +636,7 @@ def _device_settings(name):
         settings = _find_settings(name, 'connection', 'interface-name')
         if not settings:
             try:
-                hwaddr_str = nm_device_perm_hwaddress(name)
+                hwaddr_str = nm_device_valid_hwaddress(name)
             except PropertyNotFoundError:
                 settings = []
             else:
-- 
2.10.0

