From 1b6894137c19052918b0d88608639fedf7158de6 Mon Sep 17 00:00:00 2001
From: mudler <mudler@sabayon.org>
Date: Sat, 15 Oct 2016 15:18:09 +0200
Subject: [PATCH] pyanaconda/nm: handle org.freedesktop.DBus.Error.InvalidArgs

---
 pyanaconda/nm.py | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/pyanaconda/nm.py b/pyanaconda/nm.py
index 4a04ca6..c8e0813 100644
--- a/pyanaconda/nm.py
+++ b/pyanaconda/nm.py
@@ -119,7 +119,8 @@ def _get_property(object_path, prop, interface_name_suffix=""):
     try:
         prop = proxy.Get('(ss)', interface_name, prop)
     except GLib.GError as e:
-        if "org.freedesktop.DBus.Error.AccessDenied" in e.message:
+        if ("org.freedesktop.DBus.Error.AccessDenied" in e.message or
+            "org.freedesktop.DBus.Error.InvalidArgs" in e.message):
             return None
         elif "org.freedesktop.DBus.Error.UnknownMethod" in e.message:
             raise UnknownMethodGetError
-- 
2.10.0

