From d1775a967b1b560198e1f672ee72fb28163b0ebd Mon Sep 17 00:00:00 2001
From: Francesco Ferro <skullbocks@sabayon.org>
Date: Fri, 8 Dec 2017 20:48:46 +0100
Subject: [PATCH] Add some packages to the delete list

---
 pyanaconda/sabayon/utils.py | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/pyanaconda/sabayon/utils.py b/pyanaconda/sabayon/utils.py
index e764690fb..26246670a 100644
--- a/pyanaconda/sabayon/utils.py
+++ b/pyanaconda/sabayon/utils.py
@@ -651,10 +651,22 @@ blacklist nouveau
             "app-admin/calamares-sabayon-branding",
             "app-admin/calamares-sabayon-base-modules",
             "app-admin/calamares",
+            "app-admin/setools",
+            "app-emulation/spice-vdagent",
+            "app-i18n/langtable",
             "dev-libs/libreport",
+            "dev-libs/libtimezonemap",
             "dev-libs/satyr",
+            "dev-python/ipy",
+            "dev-python/pyblock",
+            "dev-python/python-bugzilla",
             "dev-python/python-blivet",
             "dev-python/python-meh",
+            "dev-python/python-nss",
+            "dev-python/pyparted",
+            "dev-python/sepolgen",
+            "net-misc/fcoe-utils",
+            "net-misc/tightvnc",
             "dev-util/glade",
             "dev-util/pykickstart",
             "sys-apps/policycoreutils",
-- 
2.15.1

