diff -Nurp gcc-config-1.8.orig/gcc-config gcc-config-1.8/gcc-config
--- gcc-config-1.8.orig/gcc-config	2012-11-19 04:11:11.000000000 +0100
+++ gcc-config-1.8/gcc-config	2013-05-22 22:22:06.187665419 +0200
@@ -209,6 +209,14 @@ update_wrappers() {
 		)
 	) )
 
+	# Sabayon: base-gcc does not bring any gcc executables. Return 1
+	# if ${ROOT}${GCC_PATH} does not exist without even trying to
+	# `cd` it.
+	if [[ ! -d "${ROOT}${GCC_PATH}" ]]; then
+		ewarn "The GCC compiler for ${CTARGET} is not installed."
+		return 1
+	fi
+
 	# See what new stuff we need to wrap up.
 	local new_wrappers=( $(
 		uniq_wrapper_list "${CC_COMP_VERSION}" $(
