diff -puN kernel/profile.c~kvm-add-vm-exit-profiling-fix kernel/profile.c
--- a/kernel/profile.c~kvm-add-vm-exit-profiling-fix
+++ a/kernel/profile.c
@@ -331,7 +331,6 @@ out:
 	local_irq_restore(flags);
 	put_cpu();
 }
-EXPORT_SYMBOL_GPL(profile_hits);
 
 static int __devinit profile_cpu_callback(struct notifier_block *info,
 					unsigned long action, void *__cpu)
@@ -401,6 +400,8 @@ void profile_hits(int type, void *__pc, 
 }
 #endif /* !CONFIG_SMP */
 
+EXPORT_SYMBOL_GPL(profile_hits);
+
 void profile_tick(int type)
 {
 	struct pt_regs *regs = get_irq_regs();
