diff --git a/sys-apps/gpu-detector/files/gpu-configuration b/sys-apps/gpu-detector/files/gpu-configuration
index a6e8afa..7075263 100755
--- a/sys-apps/gpu-detector/files/gpu-configuration
+++ b/sys-apps/gpu-detector/files/gpu-configuration
@@ -245,6 +245,7 @@ def deploy_nvidia_xxxxxx_drivers(ver):
 			return False
 
 		from entropy.client.interfaces import Client
+		from entropy.const import const_kill_threads
 		Equo = Client()
 
 		# prepare system
@@ -262,6 +263,8 @@ def deploy_nvidia_xxxxxx_drivers(ver):
 		if hasattr(Equo,'destroy'):
 			Equo.destroy()
 		del Equo
+		# Entropy Cacher still running
+                const_kill_threads()
 
 		# remove old garbage - copy over - create module
 		os.system("tar xjf %s -C / &> /dev/null" % (package_file,))
