--- arts-1.5.0/mcop/mcoputils.cc.orig	2005-12-19 20:19:50.000000000 +0100
+++ arts-1.5.0/mcop/mcoputils.cc	2005-12-19 20:18:18.000000000 +0100
@@ -46,7 +46,13 @@
   result = lstat(tmp_dir, &stat_buf);
   if ((result == -1) && (errno == ENOENT))
   {
-    return 1;
+    result = mkdir(tmp_dir, 0700);
+    if (result == -1)
+    {
+       arts_warning("Error: Can not create directory \"%s\".\n", tmp_dir);
+       return 1;
+    }
+    result = stat(tmp_dir, &stat_buf);
   }
   if ((result == -1) || (!S_ISDIR(stat_buf.st_mode)))
   {
