diff -Nurp kvm-52.orig/libkvm/Makefile kvm-52/libkvm/Makefile
--- kvm-52.orig/libkvm/Makefile	2007-11-12 10:26:13.000000000 +0100
+++ kvm-52/libkvm/Makefile	2007-11-16 14:33:48.000000000 +0100
@@ -26,10 +26,6 @@ libkvm.a: libkvm.o $(libkvm-$(ARCH)-objs
 
 install:
 	install -D libkvm.h $(DESTDIR)/$(PREFIX)/include/libkvm.h
-	install -D $(KERNELDIR)/include/linux/kvm.h \
-		$(DESTDIR)/$(PREFIX)/include/linux/kvm.h
-	install -D $(KERNELDIR)/include/linux/kvm_para.h \
-		$(DESTDIR)/$(PREFIX)/include/linux/kvm_para.h
 	install -D libkvm.a $(DESTDIR)/$(PREFIX)/$(LIBDIR)/libkvm.a
 
 -include .*.d
