Index: image.h
===================================================================
--- a/engine/core/video/image.h	(revision 3239)
+++ b/engine/core/video/image.h	(working copy)
@@ -27,7 +27,6 @@
 
 // 3rd party library includes
 #include <SDL.h>
-#include <png.h>
 
 // FIFE includes
 // These includes are split up in two parts, separated by one empty line
Index: image.cpp
===================================================================
--- a/engine/core/video/image.cpp	(revision 3239)
+++ b/engine/core/video/image.cpp	(working copy)
@@ -25,6 +25,7 @@
 
 // 3rd party library includes
 #include <SDL.h>
+#include <png.h>
 
 // FIFE includes
 // These includes are split up in two parts, separated by one empty line
