diff -Nurp fltk-1.1.9.orig/src/filename_list.cxx fltk-1.1.9/src/filename_list.cxx
--- fltk-1.1.9.orig/src/filename_list.cxx	2006-06-09 18:16:34.000000000 +0200
+++ fltk-1.1.9/src/filename_list.cxx	2009-06-18 18:00:30.062366850 +0200
@@ -67,7 +67,7 @@ int fl_filename_list(const char *d, dire
   // The vast majority of UNIX systems want the sort function to have this
   // prototype, most likely so that it can be passed to qsort without any
   // changes:
-  int n = scandir(d, list, 0, (int(*)(const void*,const void*))sort);
+  int n = scandir(d, list, 0, (int(*)(const dirent**,const dirent**))sort);
 #else
   // This version is when we define our own scandir (WIN32 and perhaps
   // some Unix systems) and apparently on IRIX:
