diff -Nur fltk-1.1.9.orig/configure.in fltk-1.1.9/configure.in
--- fltk-1.1.9.orig/configure.in	2009-01-08 17:53:52.000000000 +0000
+++ fltk-1.1.9/configure.in	2009-01-09 17:23:38.000000000 +0000
@@ -919,7 +916,7 @@
         if test "x$with_optim" != x; then
 	    OPTIM="$with_optim $OPTIM"
 	else
-            OPTIM="-Os $OPTIM"
+            OPTIM="$OPTIM"
 	fi
     fi
 
diff -Nur fltk-1.1.9.orig/documentation/Makefile fltk-1.1.9/documentation/Makefile
--- fltk-1.1.9.orig/documentation/Makefile	2009-01-08 17:53:51.000000000 +0000
+++ fltk-1.1.9/documentation/Makefile	2009-01-09 17:22:19.000000000 +0000
@@ -237,11 +237,6 @@
 		$(INSTALL_DATA) $$file $(DESTDIR)$(docdir); \
 	done
 	echo "Installing man pages in $(DESTDIR)$(mandir)..."
-	-$(INSTALL_DIR) $(DESTDIR)$(mandir)/cat1
-	$(INSTALL_MAN) fluid.$(CAT1EXT) $(DESTDIR)$(mandir)/cat1
-	$(INSTALL_MAN) fltk-config.$(CAT1EXT) $(DESTDIR)$(mandir)/cat1
-	-$(INSTALL_DIR) $(DESTDIR)$(mandir)/cat3
-	$(INSTALL_MAN) fltk.$(CAT3EXT) $(DESTDIR)$(mandir)/cat3
 	-$(INSTALL_DIR) $(DESTDIR)$(mandir)/man1
 	$(INSTALL_MAN) fluid.man $(DESTDIR)$(mandir)/man1/fluid.1
 	$(INSTALL_MAN) fltk-config.man $(DESTDIR)$(mandir)/man1/fltk-config.1
@@ -249,10 +244,6 @@
 	$(INSTALL_MAN) fltk.man $(DESTDIR)$(mandir)/man3/fltk.3
 
 install-linux install-osx:
-	-$(INSTALL_DIR) $(DESTDIR)$(mandir)/cat6
-	$(INSTALL_MAN) blocks.$(CAT6EXT) $(DESTDIR)$(mandir)/cat6
-	$(INSTALL_MAN) checkers.$(CAT6EXT) $(DESTDIR)$(mandir)/cat6
-	$(INSTALL_MAN) sudoku.$(CAT6EXT) $(DESTDIR)$(mandir)/cat6
 	-$(INSTALL_DIR) $(DESTDIR)$(mandir)/man6
 	$(INSTALL_MAN) blocks.man $(DESTDIR)$(mandir)/man6/blocks.6
 	$(INSTALL_MAN) checkers.man $(DESTDIR)$(mandir)/man6/checkers.6
@@ -261,17 +252,11 @@
 
 uninstall:
 	$(RMDIR) $(DESTDIR)$(docdir)
-	$(RM) $(DESTDIR)$(mandir)/cat1/fluid.$(CAT1EXT)
 	$(RM) $(DESTDIR)$(mandir)/man1/fluid.1
-	$(RM) $(DESTDIR)$(mandir)/cat1/fltk-config.$(CAT1EXT)
 	$(RM) $(DESTDIR)$(mandir)/man1/fltk-config.1
-	$(RM) $(DESTDIR)$(mandir)/cat3/fltk.$(CAT3EXT)
 	$(RM) $(DESTDIR)$(mandir)/man3/fltk.3
 
 uninstall-linux uninstall-osx:
-	$(RM) $(DESTDIR)$(mandir)/cat6/blocks.$(CAT6EXT)
-	$(RM) $(DESTDIR)$(mandir)/cat6/checkers.$(CAT6EXT)
-	$(RM) $(DESTDIR)$(mandir)/cat6/sudoku.$(CAT6EXT)
 	$(RM) $(DESTDIR)$(mandir)/man6/blocks.6
 	$(RM) $(DESTDIR)$(mandir)/man6/checkers.6
 	$(RM) $(DESTDIR)$(mandir)/man6/sudoku.6
diff -Nur fltk-1.1.9.orig/Makefile fltk-1.1.9/Makefile
--- fltk-1.1.9.orig/Makefile	2009-01-08 17:53:52.000000000 +0000
+++ fltk-1.1.9/Makefile	2009-01-09 17:20:49.000000000 +0000
@@ -27,7 +27,7 @@
 
 include makeinclude
 
-DIRS	=	$(IMAGEDIRS) src fluid test documentation
+DIRS	=	$(IMAGEDIRS) src fluid
 
 all: makeinclude fltk-config
 	for dir in $(DIRS); do\
@@ -35,6 +35,10 @@
 		(cd $$dir; $(MAKE) $(MFLAGS)) || exit 1;\
 	done
 
+check: all	
+	$(MAKE) $(MFLAGS) -C test || exit 1
+
+
 install: makeinclude
 	-mkdir -p $(DESTDIR)$(bindir)
 	$(RM) $(DESTDIR)$(bindir)/fltk-config
@@ -110,9 +114,6 @@
 	touch config.h
 	chmod +x fltk-config
 
-configure: configure.in
-	autoconf
-
 portable-dist:
 	epm -v -s fltk.xpm fltk
 
