diff -Nurp kuser-4.3.0.orig/kuser/ku_filessettings.ui kuser-4.3.0/kuser/ku_filessettings.ui
--- kuser-4.3.0.orig/kuser/ku_filessettings.ui	2009-04-28 15:47:13.000000000 +0200
+++ kuser-4.3.0/kuser/ku_filessettings.ui	2009-08-04 22:39:22.181152557 +0200
@@ -41,6 +41,12 @@
         <property name="text" >
          <string>MD5 shadow passwords</string>
         </property>
+        <property name="checkable">
+         <bool>false</bool>
+	</property>
+        <property name="enabled">
+         <bool>false</bool>
+        </property>
        </widget>
       </item>
       <item row="3" column="1" >
diff -Nurp kuser-4.3.0.orig/kuser/kuser.kcfg kuser-4.3.0/kuser/kuser.kcfg
--- kuser-4.3.0.orig/kuser/kuser.kcfg	2009-04-15 12:26:26.000000000 +0200
+++ kuser-4.3.0/kuser/kuser.kcfg	2009-08-04 22:36:01.911397376 +0200
@@ -112,6 +112,7 @@
     <entry name="md5shadow" type="Bool">
       <label>MD5 Shadow passwords</label>
       <whatsthis>Check this if you want the passwords in the shadow file MD5 hashed. Leave this unchecked if DES encryption should be used.</whatsthis>
+      <default>false</default>
     </entry>
     <entry name="shadowsrc" type="String">
       <label>Shadow password file</label>
diff -Nurp kuser-4.3.0.orig/kuser/ku_userfiles.cpp kuser-4.3.0/kuser/ku_userfiles.cpp
--- kuser-4.3.0.orig/kuser/ku_userfiles.cpp	2008-12-04 09:35:50.000000000 +0100
+++ kuser-4.3.0/kuser/ku_userfiles.cpp	2009-08-04 22:36:01.917282844 +0200
@@ -453,7 +453,7 @@ bool KU_UserFiles::savepwd()
 void KU_UserFiles::createPassword( KU_User &user, const QString &password )
 {
   if ( caps & Cap_Shadow ) {
-    user.setSPwd( encryptPass( password, mCfg->md5shadow() ) );
+    user.setSPwd( encryptPass( password, false /* md5 always false, was: mCfg->md5shadow() */ ) );
     user.setPwd( QString::fromLatin1("x") );
   } else
     user.setPwd( encryptPass( password, false ) );
