--- sabayon-functions.sh~	2015-05-20 23:17:15.773361057 +0200
+++ sabayon-functions.sh	2015-05-20 23:18:10.116595018 +0200
@@ -140,10 +140,6 @@
     return 1
 }
 
-sabayon_setup_motd() {
-    echo -e "\n\tWelcome to `cat /etc/sabayon-edition`\n\t`uname -p`\n\t`uname -o` `uname -r`\n" > /etc/motd
-}
-
 sabayon_setup_vt_autologin() {
     if openrc_running; then
         . /sbin/livecd-functions.sh
--- sabayonlive.sh~	2015-05-20 23:17:30.746874466 +0200
+++ sabayonlive.sh	2015-05-20 23:17:51.497200127 +0200
@@ -150,7 +150,6 @@
     # /etc/profile.env variables
     setup_locale
     sabayon_setup_autologin
-    sabayon_setup_motd
     sabayon_setup_vt_autologin
     sabayon_setup_oem_livecd
 }
