From 7e8cbd63bbd59470e0991a93a0f4a829b23660a6 Mon Sep 17 00:00:00 2001
From: mudler <mudler@dark-lab.net>
Date: Sun, 30 Aug 2015 16:01:19 +0200
Subject: [PATCH] sabayon-functions.sh: adding SDDM support for autologin

---
 sabayon-functions.sh | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/sabayon-functions.sh b/sabayon-functions.sh
index 3531a0b..6fa17c6 100644
--- a/sabayon-functions.sh
+++ b/sabayon-functions.sh
@@ -5,6 +5,7 @@ CUSTOM_GDM_FILE="/etc/gdm/custom.conf"
 KDM_FILE="/usr/share/config/kdm/kdmrc"
 LXDM_FILE="/etc/lxdm/lxdm.conf"
 LIGHTDM_FILE="/etc/lightdm/lightdm.conf"
+SDDM_FILE="/etc/sddm.conf"
 OEM_FILE="/etc/oemlive.sh"
 OEM_FILE_NEW="/etc/oem/liveboot.sh"
 LIVE_USER_GROUPS="audio bumblebee cdrom cdrw clamav console entropy games \
@@ -59,6 +60,13 @@ sabayon_setup_autologin() {
         sed -i "/^#.*autologin=/ s/^#//" $LXDM_FILE
     fi
 
+    # SDDM
+    if [ -f "$SDDM_FILE" ]; then
+        sed -i "s/Relogin=.*/Relogin=true/" $SDDM_FILE
+        sed -i "s/^User=.*/User=${LIVE_USER}/" $SDDM_FILE
+        sed -i "s/^Session=.*/Session=default/" $SDDM_FILE
+    fi
+
     # LightDM
     if [ -f "$LIGHTDM_FILE" ]; then
         sed -i "s/autologin-user=.*/autologin-user=${LIVE_USER}/" $LIGHTDM_FILE
@@ -95,6 +103,14 @@ sabayon_disable_autologin() {
     if [ -f "$LIGHTDM_FILE" ]; then
         sed -i "s/^autologin-user=.*/#autologin-user=/" $LIGHTDM_FILE
     fi
+
+    # SDDM
+    if [ -f "$SDDM_FILE" ]; then
+        sed -i "s/Relogin=.*/Relogin=false/" $SDDM_FILE
+        sed -i "s/^Session=.*/Session=/" $SDDM_FILE
+        sed -i "s/^User=.*/User=/" $SDDM_FILE
+    fi
+
 }
 
 sabayon_setup_home_mount() {
