--- ~sabayon-functions.sh	2015-07-12 14:59:45.333928709 +0200
+++ sabayon-functions.sh	2015-07-12 15:02:19.640585105 +0200
@@ -249,8 +249,8 @@
 }
 
 sabayon_setup_text_installer_motd() {
-    echo "Welcome to Sabayon Linux Text installation." >> /etc/motd
-    echo "to run the installation type: installer <and PRESS ENTER>" >> /etc/motd
+    echo "Welcome to Sabayon Linux installation. Text installation unfortunately it's not available:" >> /etc/motd
+    echo "to run the installation start the X server with vesa and run 'installer'" >> /etc/motd
 }
 
 sabayon_is_text_install() {
