diff --git a/Makefile.in b/Makefile.in
index 5b4cd77..4d1882a 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -306,10 +306,10 @@ EXTRA_DIST = autogen.sh \
 
 
 #DISTCLEANFILES=gnome-doc-utils.make
-Graphicsdir = $(datadir)/applications
+Graphicsdir = /usr/share/applications
 Graphics_in_files = gmameui.desktop.in
 Graphics_DATA = $(Graphics_in_files:.desktop.in=.desktop)
-pixmapdir = $(datadir)/pixmaps
+pixmapdir = /usr/share/pixmaps
 pixmap_DATA = gmameui.png 
 icondir = $(datadir)/gmameui
 icon_DATA = gmameui.png \
