Index: control-center/src/control-center.c
===================================================================
--- control-center/src/control-center.c	(revision 106)
+++ control-center/src/control-center.c	(working copy)
@@ -144,7 +144,7 @@
 	}
 
 	AppShellData *app_data = appshelldata_new (
-		"preferences.menu", NULL, CONTROL_CENTER_PREFIX, GTK_ICON_SIZE_DIALOG, TRUE);
+		"settings.menu", NULL, CONTROL_CENTER_PREFIX, GTK_ICON_SIZE_DIALOG, TRUE);
 	generate_categories (app_data);
 
 	GSList *actions = get_actions_list ();
