diff -Nur -x '*.orig' -x '*~' gnome-main-menu-0.9.10/main-menu/etc/slab.schemas.in.in gnome-main-menu-0.9.10.new/main-menu/etc/slab.schemas.in.in
--- gnome-main-menu-0.9.10/main-menu/etc/slab.schemas.in.in	2007-11-22 03:06:50.000000000 +1100
+++ gnome-main-menu-0.9.10.new/main-menu/etc/slab.schemas.in.in	2008-08-17 18:38:07.000000000 +1000
@@ -113,7 +113,7 @@
 			<applyto>/desktop/gnome/applications/main-menu/file-area/file_mgr_open_cmd</applyto>
 			<owner>gnome-main-menu</owner>
 			<type>string</type>
-			<default>nautilus --browser FILE_URI</default>
+			<default>nautilus FILE_URI</default>
 			<locale name="C">
 				<short>This is the command to execute when the "Open in File Manager" menu item is activated.</short>
 				<long>This is the command to execute when the "Open in File Manager" menu item is activated.  FILE_URI is replaced with a uri corresponding to the dirname of the activated file.</long>
