Index: kdebluetooth/kdebluetooth-common/Makefile.am
================================================================================
--- kdebluetooth/kbtobexclient/kbtobexclient.desktop
+++ kdebluetooth/kbtobexclient/kbtobexclient.desktop
@@ -4,10 +4,7 @@
 Icon=kdebluetooth
 Type=Application
 MimeType=bluetooth/obex-object-push-profile
-Name=kbtobexclient
-Name[ta]=kbtobex கிளைஞ்சன்
-Name[tr]=kbtobexistemcisi
-Name[xx]=xxkbtobexclientxx
+Name=Send File via Bluetooth
 GenericName=Bluetooth OBEX Client
 GenericName[bg]=OBEX клиент за Bluetooth
 GenericName[br]=Arlun OBEX Bluetooth
--- kdebluetooth/kbtserialchat/kbtserialchat.desktop
+++ kdebluetooth/kbtserialchat/kbtserialchat.desktop
@@ -5,9 +5,7 @@
 Type=Application
 MimeType=bluetooth/serial-port-profile
 Name=kbtserialchat
-Name[sv]=Kbtserialchat
-Name[tr]=kbtserisohbet
-Name[xx]=xxkbtserialchatxx
+Name=Bluetooth Terminal
 GenericName=Bluetooth Chat
 GenericName[bg]=Разговор с Bluetooth
 GenericName[br]=Flapañ gant Bluetooth
@@ -57,3 +55,4 @@
 Comment[ta]=புலுடுத்துக்கான சிறிய முடிவு பயன்பாடு.
 Comment[tr]=Bluetooth için küçük seri terminal uygulaması
 Comment[xx]=xxSmall serial terminal application for Bluetoothxx
+X-SuSE-Unimportant=true
--- kdebluetooth/kdebluetooth-common/Makefile.am
+++ kdebluetooth/kdebluetooth-common/Makefile.am
@@ -7,18 +7,6 @@
 
 EXTRA_DIST = $(kde_servicetypes_DATA)
 
-# Bluetooth directory entry in the control center
-xdg_directory_DATA = kde-settings-network-bluetooth.directory
-
-# from kdebase/applnk/Makefile.am
-install-data-local:
-	$(mkinstalldirs) $(DESTDIR)$(kde_appsdir)/Settings/Network/Bluetooth
-	$(INSTALL_DATA) $(srcdir)/kde-settings-network-bluetooth.directory \
-	  $(DESTDIR)$(kde_appsdir)/Settings/Network/Bluetooth/.directory
-
-uninstall-local:
-	-rm -f $(DESTDIR)$(kde_appsdir)/Settings/Network/Bluetooth/.directory
-
 # Update scripts
 #update_DATA = kdebluetooth.upd kdebluetooth.sh
 #updatedir = $(kde_datadir)/kconf_update
