Index: kbluetoothd/kcm/servicetab.cpp
================================================================================
--- kdebluetooth/kbluetoothd/kcm/servicetab.cpp
+++ kdebluetooth/kbluetoothd/kcm/servicetab.cpp
@@ -389,7 +389,7 @@
 {
     DCOPCall dcop(client, "kbluetoothd", "MetaServer");
     if (dcop.call("services()")!="QStringList") {
-        disableInterface(i18n("DCOP error when calling services()"));
+        disableInterface(i18n("No Bluetooth device found, disabling interface."));
         return QStringList();
     }
     else {
--- kdebluetooth/kbluetoothd/libkbluetoothd/checkdaemon.cpp
+++ kdebluetooth/kbluetoothd/libkbluetoothd/checkdaemon.cpp
@@ -36,8 +36,8 @@
                     KMessageBox::information(mainWidget,
                         i18n("<p>The KDE Bluetooth Daemon (<i>kbluetoothd</i>) is \
 needed and was started.</p>"),
-                        i18n("KDE Bluetooth Framework"),
-                    "kbluetoothd_autostart_msgbox");
+                        i18n("KDE Bluetooth Framework")/*,
+                    "kbluetoothd_autostart_msgbox"*/);
                 }
                 return true;
             }
