--- kdebluetooth/libkbluetooth/sdpdevice.cpp~	2008-03-31 17:47:49.000000000 +0200
+++ kdebluetooth/libkbluetooth/sdpdevice.cpp	2008-03-31 17:48:10.000000000 +0200
@@ -9,6 +9,7 @@
  *   (at your option) any later version.                                   *
  ***************************************************************************/
 
+#include <cstdlib>
 #include <assert.h>
 #include "sdpdevice.h"
 #include <kdebug.h>
--- kdebluetooth/kcm_btpaired/pairedtab.cpp~	2008-03-31 17:52:44.000000000 +0200
+++ kdebluetooth/kcm_btpaired/pairedtab.cpp	2008-03-31 17:52:56.000000000 +0200
@@ -55,8 +55,8 @@
 using namespace KBluetooth;
 
 struct HcidLinkKeyStruct {
-        uint8_t  key[16];
-        uint8_t  type;
+        unsigned char  key[16];
+        unsigned char  type;
         time_t   time;
 };
 
--- kdebluetooth/kcm_btpaired/pairedtab.h~	2008-03-31 17:51:32.000000000 +0200
+++ kdebluetooth/kcm_btpaired/pairedtab.h	2008-03-31 17:52:28.000000000 +0200
@@ -65,8 +65,8 @@
         QString remoteAddr;
         QString remoteName;
         QString remoteClass;
-        uint8_t linkKey[16];
-        uint8_t type;
+        unsigned char linkKey[16];
+        unsigned char type;
         QString lastUsedTime;
 	QString lastSeenTime;
         QListViewItem *listViewItem;
--- kdebluetooth/kbluetooth/sdprecord.cpp~	2008-03-31 17:49:08.000000000 +0200
+++ kdebluetooth/kbluetooth/sdprecord.cpp	2008-03-31 17:49:16.000000000 +0200
@@ -11,6 +11,7 @@
 
 #include "sdprecord.h"
 
+#include <cstdlib>
 #include <qdom.h>
 #include <kdebug.h>
 #include <bluetooth/sdp_lib.h>
