diff --exclude=CVS --exclude=.svn -up --new-file --recursive kdebluetooth-1.0_beta1.orig/kdebluetooth/irmcsync/irmcsynckonnector/clientmanager.h kdebluetooth-1.0_beta1/kdebluetooth/irmcsync/irmcsynckonnector/clientmanager.h
--- kdebluetooth-1.0_beta1.orig/kdebluetooth/irmcsync/irmcsynckonnector/clientmanager.h	2005-03-20 17:46:48.000000000 +0100
+++ kdebluetooth-1.0_beta1/kdebluetooth/irmcsync/irmcsynckonnector/clientmanager.h	2006-03-23 21:14:52.000000000 +0100
@@ -58,7 +58,7 @@ namespace KSync
 		~ClientManager();
 
 
-		void ClientManager::setObexClient(Client*);
+		void setObexClient(Client*);
 
 		/**
 		   Returns if the worker is idel at the monment.
