--- kdebluetooth/libkbluetooth/sdpattribute.cpp
+++ kdebluetooth/libkbluetooth/sdpattribute.cpp
@@ -121,23 +121,23 @@
         }
     }
     else {
-        ret.value.uuid128.data[0] = (lo >> 0) && 0xFF;
-        ret.value.uuid128.data[1] = (lo >> 8) && 0xFF;
-        ret.value.uuid128.data[2] = (lo >> 16) && 0xFF;
-        ret.value.uuid128.data[3] = (lo >> 24) && 0xFF;
-        ret.value.uuid128.data[4] = (lo >> 32) && 0xFF;
-        ret.value.uuid128.data[5] = (lo >> 40) && 0xFF;
-        ret.value.uuid128.data[6] = (lo >> 48) && 0xFF;
-        ret.value.uuid128.data[7] = (lo >> 56) && 0xFF;
+        ret.value.uuid128.data[0] = (lo >> 0) & 0xFF;
+        ret.value.uuid128.data[1] = (lo >> 8) & 0xFF;
+        ret.value.uuid128.data[2] = (lo >> 16) & 0xFF;
+        ret.value.uuid128.data[3] = (lo >> 24) & 0xFF;
+        ret.value.uuid128.data[4] = (lo >> 32) & 0xFF;
+        ret.value.uuid128.data[5] = (lo >> 40) & 0xFF;
+        ret.value.uuid128.data[6] = (lo >> 48) & 0xFF;
+        ret.value.uuid128.data[7] = (lo >> 56) & 0xFF;
     
-        ret.value.uuid128.data[8] = (hi >> 0) && 0xFF;
-        ret.value.uuid128.data[9] = (hi >> 8) && 0xFF;
-        ret.value.uuid128.data[10] = (hi >> 16) && 0xFF;
-        ret.value.uuid128.data[11] = (hi >> 24) && 0xFF;
-        ret.value.uuid128.data[12] = (hi >> 32) && 0xFF;
-        ret.value.uuid128.data[13] = (hi >> 40) && 0xFF;
-        ret.value.uuid128.data[14] = (hi >> 48) && 0xFF;
-        ret.value.uuid128.data[15] = (hi >> 56) && 0xFF;
+        ret.value.uuid128.data[8] = (hi >> 0) & 0xFF;
+        ret.value.uuid128.data[9] = (hi >> 8) & 0xFF;
+        ret.value.uuid128.data[10] = (hi >> 16) & 0xFF;
+        ret.value.uuid128.data[11] = (hi >> 24) & 0xFF;
+        ret.value.uuid128.data[12] = (hi >> 32) & 0xFF;
+        ret.value.uuid128.data[13] = (hi >> 40) & 0xFF;
+        ret.value.uuid128.data[14] = (hi >> 48) & 0xFF;
+        ret.value.uuid128.data[15] = (hi >> 56) & 0xFF;
     
         ret.type = SDP_UUID128;
     }
