diff -rNur kdebluetooth-1.0_beta1.old/kdebluetooth/kdebluetooth-common/mimetypes/computer-device-class.desktop kdebluetooth-1.0_beta1/kdebluetooth/kdebluetooth-common/mimetypes/computer-device-class.desktop
--- kdebluetooth-1.0_beta1.old/kdebluetooth/kdebluetooth-common/mimetypes/computer-device-class.desktop	2005-03-21 03:16:48.000000000 +1030
+++ kdebluetooth-1.0_beta1/kdebluetooth/kdebluetooth-common/mimetypes/computer-device-class.desktop	2006-04-08 22:55:57.000000000 +0930
@@ -3,6 +3,7 @@
 Type=MimeType
 MimeType=bluetooth/computer-device-class
 Icon=konsole
+X-KDE-IsAlso=inode/directory
 Comment=Computer/PDA Bluetooth Device
 Comment[bg]=Компютърно/PDA устройство Bluetooth
 Comment[bs]=Računar/PDA Bluetooth uređaj
diff -rNur kdebluetooth-1.0_beta1.old/kdebluetooth/kdebluetooth-common/mimetypes/misc-device-class.desktop kdebluetooth-1.0_beta1/kdebluetooth/kdebluetooth-common/mimetypes/misc-device-class.desktop
--- kdebluetooth-1.0_beta1.old/kdebluetooth/kdebluetooth-common/mimetypes/misc-device-class.desktop	2005-03-21 03:16:48.000000000 +1030
+++ kdebluetooth-1.0_beta1/kdebluetooth/kdebluetooth-common/mimetypes/misc-device-class.desktop	2006-04-08 22:55:31.000000000 +0930
@@ -3,6 +3,7 @@
 Type=MimeType
 MimeType=bluetooth/misc-device-class
 Icon=kdebluetooth
+X-KDE-IsAlso=inode/directory
 Comment=Miscellaneous Bluetooth Device
 Comment[bg]=Разни устройства Bluetooth
 Comment[da]=Diverse Bluetooth-enheder
diff -rNur kdebluetooth-1.0_beta1.old/kdebluetooth/kdebluetooth-common/mimetypes/phone-device-class.desktop kdebluetooth-1.0_beta1/kdebluetooth/kdebluetooth-common/mimetypes/phone-device-class.desktop
--- kdebluetooth-1.0_beta1.old/kdebluetooth/kdebluetooth-common/mimetypes/phone-device-class.desktop	2005-03-21 03:16:48.000000000 +1030
+++ kdebluetooth-1.0_beta1/kdebluetooth/kdebluetooth-common/mimetypes/phone-device-class.desktop	2006-04-08 22:54:46.000000000 +0930
@@ -3,6 +3,7 @@
 Type=MimeType
 MimeType=bluetooth/phone-device-class
 Icon=kaddressbook
+X-KDE-IsAlso=inode/directory
 Comment=Phone Bluetooth Device
 Comment[bg]=Телефонно устройство Bluetooth
 Comment[da]=Telefon Bluetooth-enhed
diff -rNur kdebluetooth-1.0_beta1.old/kdebluetooth/kioslave/bluetooth/bluetooth.protocol kdebluetooth-1.0_beta1/kdebluetooth/kioslave/bluetooth/bluetooth.protocol
--- kdebluetooth-1.0_beta1.old/kdebluetooth/kioslave/bluetooth/bluetooth.protocol	2005-03-21 03:16:48.000000000 +1030
+++ kdebluetooth-1.0_beta1/kdebluetooth/kioslave/bluetooth/bluetooth.protocol	2006-04-08 22:56:36.000000000 +0930
@@ -4,6 +4,7 @@
 input=none
 output=filesystem
 reading=true
+listing=Name,Type,Size,Date,AccessDate,Access,Owner,Group,Link
 Icon=kdebluetooth
 Description=Bluetooth inquiry protocol
 Description[da]=Bluetooth forespørgselsprotokol
diff -rNur kdebluetooth-1.0_beta1.old/kdebluetooth/kioslave/bluetooth/kiobluetooth.cpp kdebluetooth-1.0_beta1/kdebluetooth/kioslave/bluetooth/kiobluetooth.cpp
--- kdebluetooth-1.0_beta1.old/kdebluetooth/kioslave/bluetooth/kiobluetooth.cpp	2005-03-21 03:16:48.000000000 +1030
+++ kdebluetooth-1.0_beta1/kdebluetooth/kioslave/bluetooth/kiobluetooth.cpp	2006-04-08 22:58:56.000000000 +0930
@@ -64,6 +64,7 @@
     if (url.hasHost() == false && path == "/")
     {
         createDirEntry(entry, "Bluetooth neighbourhood");
+        statEntry(entry);
         finished();
     }
     else {
@@ -284,7 +285,8 @@
 
     addAtom(entry, UDS_MIME_TYPE, mimeType);
     addAtom(entry, UDS_FILE_TYPE, S_IFDIR);
-    addAtom(entry, UDS_GUESSED_MIME_TYPE, "inode/folder");
+    // No more necessary
+    // addAtom(entry, UDS_GUESSED_MIME_TYPE, "inode/folder");
 
     return true;
 }
diff -rNur kdebluetooth-1.0_beta1.old/kdebluetooth/kioslave/sdp/btsdp.cpp kdebluetooth-1.0_beta1/kdebluetooth/kioslave/sdp/btsdp.cpp
--- kdebluetooth-1.0_beta1.old/kdebluetooth/kioslave/sdp/btsdp.cpp	2005-03-21 03:16:48.000000000 +1030
+++ kdebluetooth-1.0_beta1/kdebluetooth/kioslave/sdp/btsdp.cpp	2006-04-08 23:06:53.000000000 +0930
@@ -131,12 +131,21 @@
         if (path == "/")
         {
             createDirEntry(entry, "Bluetooth neighbourhood");
+            statEntry(entry);
             finished();
         }
         else
         {
-            error(KIO::ERR_SLAVE_DEFINED,
+            QRegExp reg = QRegExp("^/uuid-(0x[a-f,A-F,0-9:]+)/");
+            if (reg.search(path) >= 0) {
+                createDirEntry(entry, "More services");
+                statEntry(entry);
+                finished();
+            }
+            else {
+            	error(KIO::ERR_SLAVE_DEFINED,
                   i18n("Could not stat %1. Unknown device").arg(url.url()));
+            }
         }
     }
     else {
diff -rNur kdebluetooth-1.0_beta1.old/kdebluetooth/kioslave/sdp/btsdp.protocol kdebluetooth-1.0_beta1/kdebluetooth/kioslave/sdp/btsdp.protocol
--- kdebluetooth-1.0_beta1.old/kdebluetooth/kioslave/sdp/btsdp.protocol	2005-03-21 03:16:48.000000000 +1030
+++ kdebluetooth-1.0_beta1/kdebluetooth/kioslave/sdp/btsdp.protocol	2006-04-08 22:59:38.000000000 +0930
@@ -5,6 +5,7 @@
 output=filesystem
 reading=true
 Icon=kdebluetooth
+listing=Name,Type,Size,Date,AccessDate,Access,Owner,Group,Link
 Description=Bluetooth service discovery protocol
 Description[da]=Bluetooth-protokol til opdagelse af serviceydelser
 Description[et]=Bluetoothi teenuste avastamise protokoll
