--- kdebluetooth-1.0_beta8.orig/kdebluetooth/kdebluetooth-common/dunhandler/dunhandler
+++ kdebluetooth-1.0_beta8/kdebluetooth/kdebluetooth-common/dunhandler/dunhandler
@@ -45,7 +45,7 @@
 # Find the first unbinded /dev/rfcommX
 DEV_NUMBER=0
 while test "x"$(echo $RFCOMM_SHOW_OUT | grep "rfcomm"$DEV_NUMBER":"| cut -d' ' -f1 ) != "x"; do
-	DEV_NUMBER=$[$DEV_NUMBER+1]
+	DEV_NUMBER=$(($DEV_NUMBER+1))
 	if test $DEV_NUMBER -gt 255; then break; fi;
 done;
 
--- kdebluetooth-1.0_beta8.orig/kdebluetooth/kioslave/obex/kio_obex/kioobex_start
+++ kdebluetooth-1.0_beta8/kdebluetooth/kioslave/obex/kio_obex/kioobex_start
@@ -3,7 +3,7 @@
 # This shell script is used as a mimetype handler
 # for the bluetooth/obex-ftp-protocol mimetype. 
 
-if (( $# == 0 )) ; then
+if [ $# = 0 ] ; then
     konqueror "sdp:/"
 else
     host=$(echo $@ | sed 's/sdp:\/\/\[\(..:..:..:..:..:..\).*/\1/')
