diff -Nurp kdebluetooth-1.0_beta2.orig/kdebluetooth/kbluetoothd/kbluetoothd/application.cpp kdebluetooth-1.0_beta2/kdebluetooth/kbluetoothd/kbluetoothd/application.cpp
--- kdebluetooth-1.0_beta2.orig/kdebluetooth/kbluetoothd/kbluetoothd/application.cpp	2006-10-16 12:23:56.000000000 +0200
+++ kdebluetooth-1.0_beta2/kdebluetooth/kbluetoothd/kbluetoothd/application.cpp	2007-02-16 19:26:44.000000000 +0100
@@ -58,15 +58,6 @@ KBluetoothdApp::KBluetoothdApp() :
     bFirstNewInstance = true;
     trayIcon->slotMruMenuUpdate();
 
-    // Warn the user if the location of kbluepin has changed
-    // Don't show if bluez-libs is 3.x
-#ifndef HCI_VIRTUAL
-    QString dontShowKey = QString("kbluepinLocInfo%1").arg(KDEBluetoothConfig::libDir);
-    dontShowKey = dontShowKey.replace("/", "-").replace(" ","-");
-    KMessageBox::information(trayIcon, i18n("<p>You may replace bluez's pin helper program \
-with kbluepin; it is located in <b>%1</b> now.</p>").arg(QString(KDEBluetoothConfig::libDir)),
-        "KBluetoothD", dontShowKey);
-#endif
 }
 
 KBluetoothdApp::~KBluetoothdApp()
