--- kdebluetooth/kbluetoothd/kbluetoothd/kbluetoothd.autostart.desktop
+++ kdebluetooth/kbluetoothd/kbluetoothd/kbluetoothd.autostart.desktop
@@ -40,4 +40,4 @@
 X-DCOP-ServiceType=Unique
 #X-KDE-Library=klipper_panelapplet
 #X-KDE-UniqueApplet=true
-X-KDE-autostart-condition=kbluetoothdrc:General:AutoStart:true
+X-KDE-autostart-condition=kbluetoothdrc:General:AutoStart:false
--- kdebluetooth/libkbluetooth/hcisocket.cpp
+++ kdebluetooth/libkbluetooth/hcisocket.cpp
@@ -56,7 +56,7 @@
     s = ::socket(AF_BLUETOOTH, SOCK_RAW, BTPROTO_HCI);
     if (s < 0)
     {
-        kdWarning() << "HciSocket::open(): Error creating socket" << endl;
+        kdDebug() << "HciSocket::open(): Error creating socket" << endl;
         emit error(errno, i18n("Error creating socket"));
         return false;
     }
