diff -Nurp gdm-2.20.10.orig/daemon/slave.c gdm-2.20.10/daemon/slave.c
--- gdm-2.20.10.orig/daemon/slave.c	2009-03-19 05:59:27.000000000 +0100
+++ gdm-2.20.10/daemon/slave.c	2009-08-13 00:53:09.854163205 +0200
@@ -4638,10 +4638,6 @@ gdm_slave_session_start (void)
 		language = g_strdup (usrlang);
 	}
 
-	tmp = gdm_strip_extension (session, ".desktop");
-	g_free (session);
-	session = tmp;
-
 	if (ve_string_empty (session)) {
 		g_free (session);
 		session = find_a_session ();
@@ -4651,6 +4647,10 @@ gdm_slave_session_start (void)
 		}
 	}
 
+	tmp = gdm_strip_extension (session, ".desktop");
+	g_free (session);
+	session = tmp;
+
 	if G_LIKELY (ve_string_empty (language)) {
 		g_free (language);
 		language = NULL;
Files gdm-2.20.10.orig/daemon/.slave.c.swp and gdm-2.20.10/daemon/.slave.c.swp differ
