diff --git a/data/session-setup.entries b/data/session-setup.entries
index b84e16f..a47d761 100644
--- a/data/session-setup.entries
+++ b/data/session-setup.entries
@@ -406,14 +406,14 @@
     <entry>
       <key>/desktop/gnome/interface/accessibility</key>
       <value>
-        <bool>true</bool>
+        <bool>false</bool>
       </value>
     </entry>
 
     <entry>
       <key>/desktop/gnome/accessibility/keyboard/enable</key>
       <value>
-        <bool>true</bool>
+        <bool>false</bool>
       </value>
     </entry>
 
