diff -Nurp gdm-2.20.10.orig/config/gdm.conf.in gdm-2.20.10/config/gdm.conf.in
--- gdm-2.20.10.orig/config/gdm.conf.in	2009-03-19 05:59:27.000000000 +0100
+++ gdm-2.20.10/config/gdm.conf.in	2009-06-26 15:43:46.906889864 +0200
@@ -236,8 +236,8 @@ XnestUnscaledFontPath=@X_XNEST_UNSCALED_
 #ConsoleNotify=true
 
 # How long gdm should wait before it assumes a started Xserver is defunct and
-# kills it.  10 seconds should be long enough for X, but Xgl may need 20 or 25. 
-GdmXserverTimeout=10
+# kills it.  35 seconds should be long enough. 
+GdmXserverTimeout=35
 
 [security]
 # Allow root to login.  It makes sense to turn this off for kiosk use, when
diff -Nurp gdm-2.20.10.orig/daemon/gdm-daemon-config-entries.h gdm-2.20.10/daemon/gdm-daemon-config-entries.h
--- gdm-2.20.10.orig/daemon/gdm-daemon-config-entries.h	2009-03-19 05:59:27.000000000 +0100
+++ gdm-2.20.10/daemon/gdm-daemon-config-entries.h	2009-06-26 15:44:18.794901623 +0200
@@ -350,7 +350,7 @@ static const GdmConfigEntry gdm_daemon_c
 	{ GDM_CONFIG_GROUP_DAEMON, "ConsoleCannotHandle", GDM_CONFIG_VALUE_STRING, "am,ar,az,bn,el,fa,gu,hi,ja,ko,ml,mr,pa,ta,zh", GDM_ID_CONSOLE_CANNOT_HANDLE },
 
 	/* How long to wait before assuming an Xserver has timed out */
-	{ GDM_CONFIG_GROUP_DAEMON, "GdmXserverTimeout", GDM_CONFIG_VALUE_INT, "10", GDM_ID_XSERVER_TIMEOUT },
+	{ GDM_CONFIG_GROUP_DAEMON, "GdmXserverTimeout", GDM_CONFIG_VALUE_INT, "35", GDM_ID_XSERVER_TIMEOUT },
 
 	{ GDM_CONFIG_GROUP_DAEMON, "SystemCommandsInMenu", GDM_CONFIG_VALUE_STRING_ARRAY, "HALT;REBOOT;SUSPEND;CUSTOM_CMD", GDM_ID_SYSTEM_COMMANDS_IN_MENU },
 	{ GDM_CONFIG_GROUP_DAEMON, "AllowLogoutActions", GDM_CONFIG_VALUE_STRING_ARRAY, "HALT;REBOOT;SUSPEND;CUSTOM_CMD", GDM_ID_ALLOW_LOGOUT_ACTIONS },
diff -Nurp gdm-2.20.10.orig/daemon/gdm-daemon-config-keys.h gdm-2.20.10/daemon/gdm-daemon-config-keys.h
--- gdm-2.20.10.orig/daemon/gdm-daemon-config-keys.h	2009-03-19 05:59:27.000000000 +0100
+++ gdm-2.20.10/daemon/gdm-daemon-config-keys.h	2009-06-26 15:44:05.410150893 +0200
@@ -98,7 +98,7 @@ G_BEGIN_DECLS
 #define GDM_KEY_FIRST_VT "daemon/FirstVT=7"
 #define GDM_KEY_VT_ALLOCATION "daemon/VTAllocation=true"
 #define GDM_KEY_CONSOLE_CANNOT_HANDLE "daemon/ConsoleCannotHandle=am,ar,az,bn,el,fa,gu,hi,ja,ko,ml,mr,pa,ta,zh"
-#define GDM_KEY_XSERVER_TIMEOUT "daemon/GdmXserverTimeout=10"
+#define GDM_KEY_XSERVER_TIMEOUT "daemon/GdmXserverTimeout=35"
 #define GDM_KEY_SYSTEM_COMMANDS_IN_MENU "daemon/SystemCommandsInMenu=HALT;REBOOT;SUSPEND;CUSTOM_CMD"
 #define GDM_KEY_ALLOW_LOGOUT_ACTIONS "daemon/AllowLogoutActions=HALT;REBOOT;SUSPEND;CUSTOM_CMD"
 #define GDM_KEY_RBAC_SYSTEM_COMMAND_KEYS "daemon/RBACSystemCommandKeys=" GDM_RBAC_SYSCMD_KEYS
