From 07fb1b31d818f308beb1c3800c4b90830b57d01b Mon Sep 17 00:00:00 2001
From: Sobhan Mohammadpour <sobhanmohammadpour1@yahoo.fr>
Date: Fri, 25 Jan 2013 10:03:31 +0330
Subject: [PATCH] don't load accessbility

---
 data/00-upstream-settings | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/data/00-upstream-settings b/data/00-upstream-settings
index 660a295..3993786 100644
--- a/data/00-upstream-settings
+++ b/data/00-upstream-settings
@@ -7,7 +7,7 @@
 #
 
 [org/gnome/desktop/a11y/keyboard]
-enable=true
+enable=false
 
 [org/gnome/desktop/background]
 show-desktop-icons=false
@@ -16,7 +16,7 @@ show-desktop-icons=false
 exec='/bin/true'
 
 [org/gnome/desktop/interface]
-toolkit-accessibility=true
+toolkit-accessibility=false
 
 [org/gnome/desktop/lockdown]
 disable-application-handlers=true
-- 
1.8.1

