--- gdm-2.32.0.orig/gui/simple-greeter/gdm-simple-greeter.schemas.in
+++ gdm-2.32.0/gui/simple-greeter/gdm-simple-greeter.schemas.in
@@ -38,7 +38,7 @@
       <applyto>/apps/gdm/simple-greeter/logo_icon_name</applyto>
       <owner>gdm-simple-greeter</owner>
       <type>string</type>
-      <default>computer</default>
+      <default>sabayon-gdm</default>
       <locale name="C">
         <short>Icon name to use for greeter logo</short>
         <long>Set to the themed icon name to use for the greeter logo.</long>
