diff -Nurp gdm-2.20.10.orig/config/gdm.conf-custom.in gdm-2.20.10/config/gdm.conf-custom.in
--- gdm-2.20.10.orig/config/gdm.conf-custom.in	2009-03-19 05:59:27.000000000 +0100
+++ gdm-2.20.10/config/gdm.conf-custom.in	2009-06-26 15:47:12.313899931 +0200
@@ -59,8 +59,13 @@
 [xdmcp]
 
 [gui]
+GtkRC=/usr/share/themes/Clearlooks/gtk-2.0/gtkrc
 
 [greeter]
+GlobalFaceDir=/usr/share/
+GraphicalThemes=Sabayon_Linux
+UseCirclesInEntry=true
+GraphicalThemeRand=true
 
 [chooser]
 
