Index: alsa-lib-1.0.22/src/alisp/alisp.c
===================================================================
--- alsa-lib-1.0.22.orig/src/alisp/alisp.c	2010-02-01 20:09:51.000000000 -0500
+++ alsa-lib-1.0.22/src/alisp/alisp.c	2010-02-01 20:13:32.000000000 -0500
@@ -1712,7 +1712,7 @@
 			delete_tree(instance, p1);
 		p1 = eval(instance, car(p));
 		if (alisp_compare_type(p1, ALISP_OBJ_STRING))
-			snd_output_printf(instance->out, p1->value.s);
+			snd_output_printf(instance->out, "%s", p1->value.s);
 		else
 			princ_object(instance->out, p1);
 		n = cdr(p);
