--- lib/ss/Makefile.in
+++ lib/ss/Makefile.in
@@ -5,6 +5,7 @@
 srcdir = @srcdir@
 top_srcdir = @top_srcdir@
 VPATH = @srcdir@
+builddir = @builddir@
 top_builddir = ../..
 my_dir = lib/ss
 INSTALL = @INSTALL@
@@ -163,14 +163,14 @@
 		$(DESTDIR)$(man1dir)/mk_cmds.1 
 	$(RM) -rf $(DESTDIR)$(includedir)/ss $(DESTDIR)$(datadir)/ss
 
-test_ss: test_ss.o test_cmd.o $(LIBEXT2FS) $(LIBSS) $(LIBCOM_ERR)
+test_ss: test_ss.o test_cmd.o $(LIBSS) $(LIBCOM_ERR)
 	@echo "	LD $@"
 	@$(CC) -o $@ test_ss.o test_cmd.o $(ALL_CFLAGS) \
 		$(LIBSS) $(LIBCOM_ERR)
 
 check:: all test_ss
 	@echo "	RUN TEST test_ss"
-	@(test_ss -f $(srcdir)/test_script >& test_out; exit 0)
+	@($(builddir)/test_ss -f $(srcdir)/test_script >& test_out; exit 0)
 	@if ! diff test_out $(srcdir)/test_script_expected > test.diff; then \
 		echo "Regression test for ss library failed!"; exit 1 ; fi
 
