Index: pulseaudio-0.9.14/src/utils/padsp
===================================================================
--- pulseaudio-0.9.14.orig/src/utils/padsp	2009-01-21 18:14:59.000000000 -0500
+++ pulseaudio-0.9.14/src/utils/padsp	2009-01-21 18:24:06.000000000 -0500
@@ -75,10 +75,12 @@
 
 shift $(( $OPTIND - 1 ))
 
-if [ x"$LD_PRELOAD" = x ] ; then
-   LD_PRELOAD="libpulsedsp.so"
-else
-   LD_PRELOAD="$LD_PRELOAD libpulsedsp.so"
+if `pgrep -f /usr/bin/pulseaudio 1>/dev/null` ; then
+   if [ x"$LD_PRELOAD" = x ] ; then
+	LD_PRELOAD="libpulsedsp.so"
+   else
+	LD_PRELOAD="$LD_PRELOAD libpulsedsp.so"
+   fi
 fi
 
 export LD_PRELOAD
