Index: pulseaudio-0.9.22~0.9.21+stable-queue-24-gfa64/src/modules/alsa/mixer/paths/analog-output.conf.common
===================================================================
--- pulseaudio-0.9.22~0.9.21+stable-queue-24-gfa64.orig/src/modules/alsa/mixer/paths/analog-output.conf.common	2010-01-15 18:21:18.000000000 -0500
+++ pulseaudio-0.9.22~0.9.21+stable-queue-24-gfa64/src/modules/alsa/mixer/paths/analog-output.conf.common	2010-01-15 19:17:12.000000000 -0500
@@ -121,6 +121,9 @@
 name = output-bass-boost-off
 priority = 10
 
+[Element IEC958 Optical Raw]
+switch = off
+
 ;;; 'Analog Output'
 
 [Element Analog Output]
