Index: pulseaudio-0.9.22~0.9.21+stable-queue-24-gfa64/src/daemon/daemon.conf.in
===================================================================
--- pulseaudio-0.9.22~0.9.21+stable-queue-24-gfa64.orig/src/daemon/daemon.conf.in	2010-01-15 19:16:30.000000000 -0500
+++ pulseaudio-0.9.22~0.9.21+stable-queue-24-gfa64/src/daemon/daemon.conf.in	2010-01-15 19:17:37.000000000 -0500
@@ -54,7 +54,7 @@
 ; enable-remixing = yes
 ; enable-lfe-remixing = no
 
-; flat-volumes = yes
+flat-volumes = no
 
 ; rlimit-fsize = -1
 ; rlimit-data = -1
