Index: pulseaudio-0.9.22~0.9.21+stable-queue-32-g8478/src/daemon/default.pa.in
===================================================================
--- pulseaudio-0.9.22~0.9.21+stable-queue-32-g8478.orig/src/daemon/default.pa.in	2010-01-28 19:41:35.000000000 -0500
+++ pulseaudio-0.9.22~0.9.21+stable-queue-32-g8478/src/daemon/default.pa.in	2010-01-28 19:41:43.000000000 -0500
@@ -114,7 +114,7 @@
 load-module module-position-event-sounds
 
 ### Cork music streams when a phone stream is active
-load-module module-cork-music-on-phone
+#load-module module-cork-music-on-phone
 
 # X11 modules should not be started from default.pa so that one daemon
 # can be shared by multiple sessions.
