From 33fbdc45ae2f8ce17c77b5c5f39c1e9d44dfc455 Mon Sep 17 00:00:00 2001
From: Fabio Erculiani <lxnay@sabayon.org>
Date: Fri, 26 Apr 2013 01:16:09 +0200
Subject: [PATCH] linuxrc: fix QUIET variable handling

---
 defaults/initrd.defaults | 1 -
 defaults/linuxrc         | 5 +++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/defaults/initrd.defaults b/defaults/initrd.defaults
index cbf18f3..bc07dd0 100755
--- a/defaults/initrd.defaults
+++ b/defaults/initrd.defaults
@@ -50,7 +50,6 @@ KMINOR=`echo $KV | cut -f2 -d.`
 KVER="${KMAJOR}.${KMINOR}"
 MISCOPTS='debug detect'
 
-QUIET='1'
 ROOT_LINKS='bin sbin lib lib32 lib64 boot usr opt emul'
 ROOT_TREES='etc root home var'
 KSUFF='.ko'
diff --git a/defaults/linuxrc b/defaults/linuxrc
index ba9c5d5..d17fd05 100755
--- a/defaults/linuxrc
+++ b/defaults/linuxrc
@@ -31,8 +31,6 @@ then
 	[ -e /linuxrc ] && rm /linuxrc
 fi
 
-quiet_kmsg
-
 CMDLINE=$(cat /proc/cmdline)
 # Scan CMDLINE for any specified real_root= or cdroot arguments
 FAKE_ROOT=''
@@ -41,6 +39,7 @@ FAKE_ROOTFLAGS=''
 INIT_OPTS=''
 ROOTFSTYPE='auto'
 CRYPT_SILENT=0
+QUIET=''
 
 mkdir -p /etc/cmdline /etc/modprobe.d
 for x in ${CMDLINE}
@@ -270,6 +269,8 @@ do
 	esac
 done
 
+quiet_kmsg
+
 if [ "${CDROOT}" = '0' ]
 then
 	if [ -z "${REAL_ROOT}" -a "${FAKE_ROOT}" != "/dev/ram0" ]
-- 
1.8.1.5

