https://bugs.gentoo.org/show_bug.cgi?id=411453#attach_309805
--- gentoo-init.lisp.orig  2012-04-22 21:38:24.768658710 +0200
+++ gentoo-init.lisp       2012-04-22 21:38:37.996868213 +0200
@@ -2,6 +2,4 @@
 #+(or sbcl ecl) (require :asdf)
 #-(or sbcl ecl) (load #p"/usr/share/common-lisp/source/asdf/asdf.lisp")
 (push #p"/usr/share/common-lisp/systems/" asdf:*central-registry*)
-(asdf:oos 'asdf:load-op :asdf-binary-locations)
-(setf asdf:*centralize-lisp-binaries* t)
-(setf asdf:*source-to-target-mappings* '((#p"/usr/lib/sbcl/" nil) (#p"/usr/lib64/sbcl/" nil)))
+(asdf:enable-asdf-binary-locations-compatibility :centralize-lisp-binaries t :source-to-target-mappings '((#p"/usr/lib/sbcl/" nil) (#p"/usr/lib64/sbcl/" nil)))
