diff -Nru xen-4.0.1/tools/blktap/drivers/blk_linux.c xen-4.0.1-patched/tools/blktap/drivers/blk_linux.c
--- xen-4.0.1/tools/blktap/drivers/blk_linux.c    2010-08-25 07:22:07.000000000 -0300
+++ xen-4.0.1-patched/tools/blktap/drivers/blk_linux.c    2010-12-16 17:42:14.694504013 -0200
@@ -1,6 +1,6 @@
 #include <inttypes.h>
 #include <sys/ioctl.h>
-#include <linux/fs.h>
+#include <sys/mount.h>
 #include "tapdisk.h"
 #include "blk.h"
 
diff -Nru xen-4.0.1/tools/blktap/drivers/Makefile xen-4.0.1-patched/tools/blktap/drivers/Makefile
--- xen-4.0.1/tools/blktap/drivers/Makefile    2010-08-25 07:22:07.000000000 -0300
+++ xen-4.0.1-patched/tools/blktap/drivers/Makefile    2010-12-16 17:45:06.214503843 -0200
@@ -29,8 +29,9 @@
 MEMSHRLIBS += $(MEMSHR_DIR)/libmemshr.a
 endif
 
-LDFLAGS_blktapctrl := $(LDFLAGS_libxenctrl) $(LDFLAGS_libxenstore) $(MEMSHRLIBS) -L../lib -lblktap -lrt -lm -lpthread
-LDFLAGS_img := $(LIBAIO_DIR)/libaio.a $(CRYPT_LIB) -lpthread -lz
+LDFLAGS_xen := $(LDFLAGS_libxenctrl) $(LDFLAGS_libxenstore)
+LDFLAGS_blktapctrl := $(LDFLAGS_xen) $(MEMSHRLIBS) -L../lib -lblktap -lrt -lm -lpthread $(LDFLAGS_xen)
+LDFLAGS_img := $(LIBAIO_DIR)/libaio.a $(CRYPT_LIB) -lpthread -lz $(LDFLAGS_xen)
 
 BLK-OBJS-y  := block-aio.o
 BLK-OBJS-y  += block-sync.o
diff -Nru xen-4.0.1/tools/blktap2/drivers/blk_linux.c xen-4.0.1-patched/tools/blktap2/drivers/blk_linux.c
--- xen-4.0.1/tools/blktap2/drivers/blk_linux.c    2010-08-25 07:22:07.000000000 -0300
+++ xen-4.0.1-patched/tools/blktap2/drivers/blk_linux.c    2010-12-16 17:43:12.624504046 -0200
@@ -1,7 +1,7 @@
 #include <inttypes.h>
+#include <errno.h>
 #include <sys/ioctl.h>
-#include <linux/fs.h>
-#include <linux/errno.h>
+#include <sys/mount.h>
 #include "tapdisk.h"
 #include "blk.h"
