diff -Naur soundkonverter-0.2.90/src/metadata/Makefile.am soundkonverter-0.2.90_new/src/metadata/Makefile.am
--- soundkonverter-0.2.90/src/metadata/Makefile.am	2006-08-31 16:28:04.000000000 +0200
+++ soundkonverter-0.2.90_new/src/metadata/Makefile.am	2006-11-20 05:56:45.000000000 +0100
@@ -4,7 +4,7 @@
 
 
 noinst_LTLIBRARIES = libmetadata.la
-libmetadata_la_SOURCES = tagengine.cpp tagengine.cpp tplugins.cpp
+libmetadata_la_SOURCES = tagengine.cpp tplugins.cpp
 SUBDIRS = wma audible rmff aac m4a ape
 libmetadata_la_LIBADD = $(taglib_libs) wma/libtagwma.la \
 	audible/libtagaudible.la rmff/libtagrealmedia.la aac/libtagaac.la $(MP4_LDADD) \
diff -Naur soundkonverter-0.2.90/src/pluginloader/Makefile.am soundkonverter-0.2.90_new/src/pluginloader/Makefile.am
--- soundkonverter-0.2.90/src/pluginloader/Makefile.am	2006-08-31 16:27:59.000000000 +0200
+++ soundkonverter-0.2.90_new/src/pluginloader/Makefile.am	2006-11-20 05:57:22.000000000 +0100
@@ -5,5 +5,5 @@
 
 noinst_LTLIBRARIES = libpluginloader.la
 libpluginloader_la_SOURCES = pluginloaderbase.cpp \
-			convertpluginloader.cpp	convertpluginloader.cpp pluginloaderbase.cpp replaygainpluginloader.cpp formatinfoloader.cpp \
+			convertpluginloader.cpp	replaygainpluginloader.cpp formatinfoloader.cpp \
 	ripperpluginloader.cpp
