--- ./src/wm-tester/test-resizing.c	2009-08-05 23:53:32.000000000 +0200
+++ ./src/wm-tester/test-resizing.c.new	2009-12-06 19:59:29.000000000 +0100
@@ -91,7 +91,7 @@
   XSizeHints hints;
   int screen;
   XEvent ev;
-  int x, y, width, height;
+  int x = 0, y = 0, width = 0, height = 0;
   Pixmap pix;
   GC gc;  
   XGCValues gc_vals;
--- ./src/core/boxes.c	2009-09-01 00:03:52.000000000 +0200
+++ ./src/core/boxes.c.new	2009-12-06 20:02:19.000000000 +0100
@@ -915,7 +915,10 @@
       MetaRectangle *compare_rect = temp->data;
       MetaRectangle  overlap;
       int            maximal_overlap_amount_for_compare;
-     
+    
+      overlap.width = 0;
+      overlap.height = 0;
+
       /* If x is fixed and the entire width of rect doesn't fit in compare,
        * skip the rectangle.
        */
