--- configure.ac.old	2006-11-04 15:27:24.000000000 -0700
+++ configure.ac	2006-11-04 15:28:32.000000000 -0700
@@ -1342,7 +1342,7 @@
 
 	dnl these only go in xorg-config.h
 	XF86CONFIGFILE="xorg.conf"
-	CONFIGFILE="$sysconfdir/$XF86CONFIGFILE"
+	CONFIGFILE="$sysconfdir/X11/$XF86CONFIGFILE"
 	LOGPREFIX="$logdir/Xorg."
 	AC_DEFINE(XORG_SERVER, 1, [Building Xorg server])
 	AC_DEFINE(XORGSERVER, 1, [Building Xorg server])
