diff --git a/hw/xprint/Makefile.am b/hw/xprint/Makefile.am
index dc8764a..728e69d 100644
--- a/hw/xprint/Makefile.am
+++ b/hw/xprint/Makefile.am
@@ -12,7 +12,7 @@ Xprt_LDFLAGS = -L$(top_srcdir)
 Xprt_LDADD = @XPRINT_LIBS@ ps/libps.la raster/libraster.la  \
 	pcl/libpcl.la pcl-mono/libpcl.la $(top_builddir)/fb/libfb.la \
 	$(top_builddir)/render/librender.la $(top_builddir)/mi/libmi.la \
-	$(top_builddir)/Xext/libXext.la  @FREETYPE_LIBS@
+	$(top_builddir)/Xext/libXext.la  @FREETYPE_LIBS@ @XSERVER_LIBS@
 
 miinitext-wrapper.c:
 	echo "#include \"$(top_srcdir)/mi/miinitext.c\"" >> $@
