--- configure.ac	2010-03-16 22:24:14.000000000 +0100
+++ configure.ac.new	2010-03-16 22:33:10.000000000 +0100
@@ -1915,8 +1915,14 @@
         AC_DEFINE(KDRIVEFBDEV, 1, [Build fbdev-based kdrive server])
     fi
 
-
-    PKG_CHECK_MODULES([TSLIB], [tslib-0.0], [HAVE_TSLIB="yes"], [HAVE_TSLIB="no"])
+dnl Do tripple check for tslib in the transition phase to unversioned pkgconfig files...
+    PKG_CHECK_MODULES([TSLIB], [tslib], [HAVE_TSLIB="yes"], [
+        PKG_CHECK_MODULES([TSLIB], [tslib-1.0], [HAVE_TSLIB="yes"], [
+           	PKG_CHECK_MODULES([TSLIB], [tslib-0.0], [HAVE_TSLIB="yes"], [
+		    HAVE_TSLIB="no"
+		])
+	])
+    ])
     if test "x$HAVE_TSLIB" = xno; then
         AC_CHECK_LIB(ts, ts_open, [HAVE_TSLIB="yes"])
     fi
