--- include/inputstr.h-orig	2008-02-19 10:57:32.000000000 +0000
+++ include/inputstr.h	2008-02-19 10:58:00.000000000 +0000
@@ -159,7 +159,7 @@
     int			  *axisVal;
     int                   lastx, lasty; /* last event recorded, not posted to
                                          * client; see dix/devices.c */
-    int                   dxremaind, dyremaind; /* for acceleration */
+    float                 dxremaind, dyremaind; /* for acceleration */
     CARD8	 	  mode;
 } ValuatorClassRec, *ValuatorClassPtr;
 
