--- ../Mesa-6.4.2/src/mesa/drivers/dri/common/glcontextmodes.c	2005-11-30 02:02:16.000000000 +0300
+++ ../Mesa-6.4.2/src/mesa/drivers/dri/common/glcontextmodes.c	2006-02-07 10:38:46.000000000 +0300
@@ -39,6 +39,9 @@
 # include "imports.h"
 # define __glXMemset  memset
 #else
+# if defined (HAVE_DIX_CONFIG_H)
+#  include <dix-config.h>
+# endif
 # include <X11/X.h>
 # include <GL/glx.h>
 # include "GL/glxint.h"
