diff --git a/GL/mesa/main/Makefile.am b/GL/mesa/main/Makefile.am
index 867fe57..4d5b0cd 100644
--- a/GL/mesa/main/Makefile.am
+++ b/GL/mesa/main/Makefile.am
@@ -29,6 +27,7 @@ nodist_libmain_la_SOURCES = accum.c \
                     api_loopback.c \
                     api_noop.c \
                     api_validate.c \
+                    arrayobj.c \
                     attrib.c \
                     blend.c \
                     bufferobj.c \
@@ -67,6 +66,7 @@ nodist_libmain_la_SOURCES = accum.c \
                     points.c \
                     polygon.c \
                     rastpos.c \
+                    rbadaptors.c \
                     renderbuffer.c \
                     state.c \
                     stencil.c \
diff --git a/GL/symlink-mesa.sh b/GL/symlink-mesa.sh
index ccaa2a9..32f839a 100755
--- a/GL/symlink-mesa.sh
+++ b/GL/symlink-mesa.sh
@@ -88,8 +88,11 @@ symlink_mesa_main() {
     action api_noop.h
     action api_validate.c
     action api_validate.h
+    action arrayobj.c
+    action arrayobj.h
     action attrib.c
     action attrib.h
+    action bitset.h
     action blend.c
     action blend.h
     action bufferobj.c
@@ -168,6 +171,8 @@ symlink_mesa_main() {
     action polygon.h
     action rastpos.c
     action rastpos.h
+    action rbadaptors.c
+    action rbadaptors.h
     action renderbuffer.c
     action renderbuffer.h
     action simple_list.h
@@ -468,7 +468,7 @@ symlink_mesa_shader_slang_library() {
     action slang_core_gc.h
     action slang_fragment_builtin_gc.h
     action slang_shader_syn.h
-    action slang_version_syn.h
+    action slang_pp_version_syn.h
     action slang_vertex_builtin_gc.h
 }        
 
