diff -Nurp 0.36.1.orig/build/configurator/source/configurator.cpp 0.36.1/build/configurator/source/configurator.cpp
--- 0.36.1.orig/build/configurator/source/configurator.cpp	2009-06-09 17:25:38.181996449 +0200
+++ 0.36.1/build/configurator/source/configurator.cpp	2009-06-09 17:35:52.849767990 +0200
@@ -2955,13 +2955,14 @@ void MyDialog::OnTestNet(wxCommandEvent&
 	btnUpdate->Enable(false);
 	timer1->Start(10000);
 	std::string lshort = conv(language->CanonicalName).substr(0, 2);
+	/*
 	networkhtmw->LoadPage(wxString(_(REPO_HTML_SERVERLIST))+
 						  wxString(conv("?version="))+
 						  wxString(_(RORNET_VERSION))+
 						  wxString(conv("&lang="))+
 						  conv(lshort));	
-
-//	networkhtmw->LoadPage("http://rigsofrods.blogspot.com/");
+	*/
+	networkhtmw->LoadPage(conv("http://rigsofrods.blogspot.com/"));
 	/*
 	SWInetSocket mySocket;
 	SWBaseSocket::SWBaseError error;
