diff -Nurp 0.36.1.orig/build/main/CMakeLists.txt 0.36.1/build/main/CMakeLists.txt
--- 0.36.1.orig/build/main/CMakeLists.txt	2009-06-09 15:59:53.907871379 +0200
+++ 0.36.1/build/main/CMakeLists.txt	2009-06-09 16:25:59.255882462 +0200
@@ -152,7 +152,7 @@ IF(WIN32)
   
   target_link_libraries(RoR openal32)
 ELSEIF(UNIX)
-  SET_TARGET_PROPERTIES(RoR PROPERTIES COMPILE_FLAGS "-msse2 -msse2 -mmmx -msse -msse3 -mssse3 -m3dnow")
+  SET_TARGET_PROPERTIES(RoR PROPERTIES COMPILE_FLAGS "-mmmx")
   SET_TARGET_PROPERTIES(RoR PROPERTIES LINK_FLAGS    "")
   target_link_libraries(RoR openal)
 ELSEIF(APPLE)
