diff -Nurp 0.36.1.orig/build/configurator/source/configurator.cpp 0.36.1/build/configurator/source/configurator.cpp
--- 0.36.1.orig/build/configurator/source/configurator.cpp	2009-06-09 16:46:04.860996399 +0200
+++ 0.36.1/build/configurator/source/configurator.cpp	2009-06-09 16:51:33.334871695 +0200
@@ -133,6 +133,16 @@ inline std::string conv(const wxString& 
 	return std::string(s.mb_str(wxConvUTF8));
 }
 
+inline const char *conv2(const wxString& s)
+{
+	return std::string(s.mb_str(wxConvUTF8)).c_str();
+}
+
+inline const char *conv2(const std::string& s)
+{
+	return s.c_str();
+}
+
 // Define a new application type, each program should derive a class from wxApp
 class MyApp : public wxApp
 {
@@ -392,7 +402,7 @@ public:
 		{
 			if(counter>MAX_EVENTS)
 				break;
-			eventChoices[counter] = wxString(it->first.c_str(), wxConvUTF8);;
+			eventChoices[counter] = conv(it->first);
 		}
 		cbe = new wxComboBox(this, 1, _("Keyboard"), wxPoint(80,5), wxSize(300, 25), counter, eventChoices, wxCB_READONLY);
 		cbe->SetSelection(0);
@@ -422,7 +432,7 @@ public:
 	void onChangeEventComboBox(wxCommandEvent& event)
 	{
 		std::string s = inputevents[conv(cbe->GetValue())];
-		desctext->SetLabel(_("Event Description: ") + conv(s.c_str()));
+		desctext->SetLabel(_("Event Description: ") + conv(s));
 		selectedEvent = conv(cbe->GetValue());
 	}
 
@@ -2218,7 +2228,7 @@ bool MyDialog::LoadConfig()
 
 	wxString rorcfg=app->UserPath + wxFileName::GetPathSeparator() + _T("config") + wxFileName::GetPathSeparator() + _T("RoR.cfg");
 
-	printf("reading from Config file: %s\n", rorcfg.c_str());
+	printf("reading from Config file: %s\n", conv2(rorcfg));
 	// Don't trim whitespace
 	cfg.load(rorcfg.ToUTF8().data(), "=:\t", false);
 
