Index: scorched3d-41.3dfsg/src/common/main.h
===================================================================
--- scorched3d-41.3dfsg.orig/src/common/main.h	2008-01-26 20:41:37.000000000 +0100
+++ scorched3d-41.3dfsg/src/common/main.h	2008-01-26 20:45:15.000000000 +0100
@@ -21,6 +21,8 @@
 #define WIN32_LEAN_AND_MEAN
 
 #include <windows.h>
+#include <cstdlib>
+#include <vector>
 
 #ifdef HAVE_UNISTD_H
 #include <unistd.h>
Index: scorched3d-41.3dfsg/src/common/sha2.h
===================================================================
--- scorched3d-41.3dfsg.orig/src/common/sha2.h	2008-01-26 20:41:37.000000000 +0100
+++ scorched3d-41.3dfsg/src/common/sha2.h	2008-01-26 20:45:15.000000000 +0100
@@ -46,6 +46,7 @@
 
 #include <string>
 #include <stdexcept>
+#include <cstring>
 
 // NOTE: You may need to define things by hand for your system: 
 typedef unsigned char  sha_byte;           // Exactly 1 byte 
Index: scorched3d-41.3dfsg/src/common/DefinesFile.cpp
===================================================================
--- scorched3d-41.3dfsg.orig/src/common/DefinesFile.cpp	2008-01-26 20:41:37.000000000 +0100
+++ scorched3d-41.3dfsg/src/common/DefinesFile.cpp	2008-01-26 20:45:15.000000000 +0100
@@ -21,6 +21,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string>
+#include <cstring>
 #include <common/DefinesFile.h>
 #include <sys/types.h>
 #include <sys/stat.h>
Index: scorched3d-41.3dfsg/src/common/LoggerI.cpp
===================================================================
--- scorched3d-41.3dfsg.orig/src/common/LoggerI.cpp	2008-01-26 20:45:33.000000000 +0100
+++ scorched3d-41.3dfsg/src/common/LoggerI.cpp	2008-01-26 20:46:03.000000000 +0100
@@ -20,6 +20,7 @@
 
 #include <common/LoggerI.h>
 #include <time.h>
+#include <cstring>
 
 LoggerInfo::LoggerInfo(
 	const char *message, 
