diff -Nurp virtinst-0.101.0.orig/virt-install virtinst-0.101.0/virt-install
--- virtinst-0.101.0.orig/virt-install	2007-02-20 21:26:23.000000000 +0100
+++ virtinst-0.101.0/virt-install	2007-02-25 13:09:48.000000000 +0100
@@ -390,9 +390,9 @@ def main():
         hvm = True
         type = "qemu"
         if options.accelerate:
-            if util.is_kvm_capable():
+            if virtinst.util.is_kvm_capable():
                 type = "kvm"
-            elif util.is_kqemu_capable():
+            elif virtinst.util.is_kqemu_capable():
                 type = "kqemu"
 
     if hvm:
