diff -Nurp virtinst-0.101.0.orig/virtinst/FullVirtGuest.py virtinst-0.101.0/virtinst/FullVirtGuest.py
--- virtinst-0.101.0.orig/virtinst/FullVirtGuest.py
+++ virtinst-0.101.0/virtinst/FullVirtGuest.py
@@ -19,7 +19,8 @@ import util
 import DistroManager
 
 class FullVirtGuest(Guest.XenGuest):
-    OS_TYPES = { "Linux" : { "Red Hat Enterprise Linux AS 2.1/3" : { "acpi" : True, "apic": True }, \
+    OS_TYPES = { "Linux" : { "Sabayon Linux" : { "acpi" : True, "apic": True } }, \
+                             "Red Hat Enterprise Linux AS 2.1/3" : { "acpi" : True, "apic": True }, \
                              "Red Hat Enterprise Linux 4" : { "acpi" : True, "apic": True }, \
                              "Red Hat Enterprise Linux 5" : { "acpi" : True, "apic": True }, \
                              "Fedora Core 4-6" : { "acpi" : True, "apic": True }, \
