--- common/lib/modules/fglrx/build_mod/firegl_public.c.orig	2009-09-10 20:14:10.302938534 +0200
+++ common/lib/modules/fglrx/build_mod/firegl_public.c	2009-09-10 20:16:21.828943780 +0200
@@ -183,6 +183,7 @@
 #include <linux/string.h>
 #include <linux/gfp.h>
 #include <linux/swap.h>
+#include <linux/pid.h>
 
 #include "firegl_public.h"
 #include "kcl_osconfig.h"
@@ -1315,7 +1316,7 @@
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,26)  
    p = find_task_by_pid( pid );
 #else   
-   p = find_task_by_vpid( pid );
+   p = pid_task( pid, PIDTYPE_PID );
 #endif   
    if (p) 
    {
