--- common/lib/modules/fglrx/build_mod/firegl_public.c
+++ common/lib/modules/fglrx/build_mod/firegl_public.c
@@ -41,6 +41,7 @@
 #if !defined(CONFIG_X86_BIGSMP)
 #if !defined(CONFIG_X86_VISWS)
 #if !defined(CONFIG_X86_GENERICARCH)
+#if !defined(CONFIG_X86_32)
 #error unknown or undefined architecture configured
 #endif
 #endif
@@ -53,6 +54,7 @@
 #endif
 #endif
 #endif
+#endif
 
 /* The dirty-page-tracking patch included in NLD 9 SMP kernels defines
  * a static inline function that uses a GPL-only symbol in a header
