diff -uNr vmnet-only.old/bridge.c vmnet-only/bridge.c
--- vmnet-only.old/bridge.c	2008-01-06 21:25:04.000000000 +0000
+++ vmnet-only/bridge.c	2008-01-06 21:27:42.000000000 +0000
@@ -32,7 +32,7 @@
 #include <linux/tcp.h>
 #include <net/tcp.h>
 
-#ifdef CONFIG_NET_RADIO
+#if defined CONFIG_NET_RADIO || defined CONFIG_WLAN_80211
 #   include <linux/wireless.h>
 #endif
 #include "vmnetInt.h"
@@ -764,7 +764,7 @@
 {
 #ifdef CONFIG_WIRELESS_EXT
    return dev->wireless_handlers != NULL;
-#elif !defined(CONFIG_NET_RADIO)
+#elif !defined CONFIG_NET_RADIO && !defined CONFIG_WLAN_80211
    return FALSE;
 #elif WIRELESS_EXT > 19
    return dev->wireless_handlers != NULL;
