commit 344f0b42e473c216301923df71627834560cf149
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Tue Oct 25 11:45:41 2011 +0200

    fix random build failures during e2fsprogs, force -j1

diff --git a/gen_compile.sh b/gen_compile.sh
index fce0750..4f4ed25 100755
--- a/gen_compile.sh
+++ b/gen_compile.sh
@@ -600,7 +600,7 @@ compile_e2fsprogs() {
 		LDFLAGS=-static ./configure >> ${LOGFILE} 2>&1 ||
 			gen_die 'Configuring e2fsprogs failed!'
 		print_info 1 'e2fsprogs: >> Compiling...'
-		MAKE=${UTILS_MAKE} compile_generic "" ""
+		MAKE=${UTILS_MAKE} MAKEOPTS="${MAKEOPTS} -j1" compile_generic "" ""
 		print_info 1 'blkid: >> Copying to cache...'
 		[ -f "${TEMP}/${E2FSPROGS_DIR}/misc/blkid" ] ||
 			gen_die 'Blkid executable does not exist!'
