diff -Nurp python-report-0.10.orig/python/report/templates/bugzilla-template/__init__.py python-report-0.10/python/report/templates/bugzilla-template/__init__.py
--- python-report-0.10.orig/python/report/templates/bugzilla-template/__init__.py	2010-04-02 19:32:19.729109477 +0200
+++ python-report-0.10/python/report/templates/bugzilla-template/__init__.py	2010-04-02 19:32:54.640917830 +0200
@@ -106,7 +106,6 @@ import filer
 def sendToBugzilla(component, signature, io, bzfiler,
                    optionsDict, fileName, fileDescription):
 
-    import rpmUtils.arch
 
     class BugzillaCommunicationException (Exception):
         pass
@@ -204,7 +203,7 @@ def sendToBugzilla(component, signature,
             bug = withBugzillaDo(bzfiler, lambda b: b.createbug(product=bzfiler.getproduct(),
                                            component=component,
                                            version=bzfiler.getversion(),
-                                           platform=rpmUtils.arch.getBaseArch(),
+                                           # platform=rpmUtils.arch.getBaseArch(),
                                            bug_severity="medium",
                                            priority="medium",
                                            op_sys="Linux",
