diff -Naur fslint-2.40/fslint-gui fslint-2.40-2/fslint-gui
--- fslint-2.40/fslint-gui	2009-07-21 00:10:23.000000000 +0100
+++ fslint-2.40-2/fslint-gui	2010-04-27 23:56:38.063576229 +0100
@@ -34,8 +34,8 @@
 def puts(string=""): #print is problematic in python 3
     sys.stdout.write(string+'\n')
 
-liblocation=os.path.dirname(os.path.abspath(sys.argv[0]))
-locale_base=liblocation+'/po/locale'
+liblocation = '/usr/share/fslint/'
+locale_base = None
 try:
     import fslint
     if sys.argv[0][0] != '/':
