--- ./dix/window.c	2006/05/29 18:12:27	1.1
+++ ./dix/window.c	2006/05/29 18:15:15
@@ -1265,8 +1265,12 @@
 	    result = EventSelectForWindow(pWin, client, (Mask )*pVlist);
 	    if (result)
 	    {
-		error = result;
-		goto PatchUp;
+		if (vmask != CWEventMask ||
+		    (Mask)*pVlist != SubstructureRedirectMask)
+		{
+		   error = result;
+		   goto PatchUp;
+		}
 	    }
 	    pVlist++;
 	    break;

