--- ./composite/compalloc.c
+++ ./composite/compalloc.c
@@ -486,12 +486,22 @@ compNewPixmap (WindowPtr pWin, int x, int y, int w, int h)
 	
 	ValidateGC(&pPixmap->drawable, pGC);
 	dixChangeGC (serverClient, pGC, GCSubwindowMode, &val, NULL);
+#if 1
+    val = 0;
+    dixChangeGC (serverClient, pGC, GCForeground, &val, NULL);
+ 
+    {
+        xRectangle rect = { 0, 0, w, h };
+        (*pGC->ops->PolyFillRect) (&pPixmap->drawable, pGC, 1, &rect);
+    }
+#else
 	(*pGC->ops->CopyArea) (&pParent->drawable,
 			       &pPixmap->drawable,
 			       pGC,
 			       x - pParent->drawable.x,
 			       y - pParent->drawable.y,
 			       w, h, 0, 0);
+#endif
 	FreeScratchGC (pGC);
     }
     return pPixmap;
