diff -NrU5 boinc-6.4.5.orig/clientgui/common/wxFlatNotebook.cpp boinc-6.4.5/clientgui/common/wxFlatNotebook.cpp
--- boinc-6.4.5.orig/clientgui/common/wxFlatNotebook.cpp	2009-05-12 15:56:03.000000000 -0600
+++ boinc-6.4.5/clientgui/common/wxFlatNotebook.cpp	2009-05-12 17:19:57.000000000 -0600
@@ -605,11 +605,11 @@
 // 
 //	wxPageContainerBase
 //
 ///////////////////////////////////////////////////////////////////////////////////////////
 
-BEGIN_EVENT_TABLE(wxPageContainerBase, wxControl)
+BEGIN_EVENT_TABLE(wxPageContainerBase, wxPanel)
 EVT_PAINT(wxPageContainerBase::OnPaint)
 EVT_SIZE(wxPageContainerBase::OnSize)
 EVT_LEFT_DOWN(wxPageContainerBase::OnLeftDown)
 EVT_LEFT_UP(wxPageContainerBase::OnLeftUp)
 EVT_RIGHT_DOWN(wxPageContainerBase::OnRightDown)
diff -NrU5 boinc-6.4.5.orig/sched/sched_driver.cpp boinc-6.4.5/sched/sched_driver.cpp
--- boinc-6.4.5.orig/sched/sched_driver.cpp	2009-05-12 15:56:03.000000000 -0600
+++ boinc-6.4.5/sched/sched_driver.cpp	2009-05-12 16:01:00.000000000 -0600
@@ -38,10 +38,11 @@
 #define AUTHENTICATOR    "49bcae97f1788385b0f41123acdf5694"
     // authenticator of a user record
 #define HOSTID "7"
     // ID of a host belonging to that user
 
+#include <cstdio>
 #include <vector>
 #include "util.h"
 
 using std::vector;
 
diff -NrU5 boinc-6.4.5.orig/sched/sched_util.cpp boinc-6.4.5/sched/sched_util.cpp
--- boinc-6.4.5.orig/sched/sched_util.cpp	2009-05-12 15:56:03.000000000 -0600
+++ boinc-6.4.5/sched/sched_util.cpp	2009-05-12 17:58:59.000000000 -0600
@@ -102,11 +102,11 @@
 #ifndef _USING_FCGI_
 int try_fopen(const char* path, FILE*& f, const char* mode) {
 #else
 int try_fopen(const char* path, FCGI_FILE*& f, const char *mode) {
 #endif
-    char* p;
+    const char* p;
     DIR* d;
     char dirpath[256];
 
 #ifndef _USING_FCGI_
     f = fopen(path, mode);
