diff -NrU5 boinc-6.4.5.orig/client/Makefile.in boinc-6.4.5/client/Makefile.in
--- boinc-6.4.5.orig/client/Makefile.in	2009-05-14 00:59:41.000000000 -0600
+++ boinc-6.4.5/client/Makefile.in	2009-05-14 01:01:10.000000000 -0600
@@ -1330,11 +1330,10 @@
 @CLIENT_BIN_FILENAME@: boinc_client
 	rm -f $@
 	rm -f $@.unmodified
 	cp $? $@.unmodified
 	@LN@ $? $@
-	@STRIP@ $@
 
 clean-local:
 	rm -f @CLIENT_BIN_FILENAME@
 	rm -f @CLIENT_BIN_FILENAME@.unmodified
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
diff -NrU5 boinc-6.4.5.orig/lib/Makefile.in boinc-6.4.5/lib/Makefile.in
--- boinc-6.4.5.orig/lib/Makefile.in	2009-05-14 00:59:41.000000000 -0600
+++ boinc-6.4.5/lib/Makefile.in	2009-05-14 01:00:31.000000000 -0600
@@ -950,11 +950,10 @@
 @CLIENT_CMD_BIN_FILENAME@: boinc_cmd$(EXEEXT)
 	rm -f $@
 	rm -f $@.unmodified
 	cp $? $@.unmodified
 	@LN@ $? $@
-	@STRIP@ $@
 
 clean-local:
 	rm -f @CLIENT_CMD_BIN_FILENAME@
 	rm -f @CLIENT_CMD_BIN_FILENAME@.unmodified
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
