--- linux-2.6.32-openvz-039.10.orig/fs/ntfs/super.c
+++ linux-2.6.32-openvz-039.10/fs/ntfs/super.c
@@ -3047,7 +3047,7 @@ iput_tmp_ino_err_out_now:
 	 * method again... FIXME: Do we need to do this twice now because of
 	 * attribute inodes? I think not, so leave as is for now... (AIA)
 	 */
-	if (invalidate_inodes(sb)) {
+	if (invalidate_inodes(sb, true)) {
 		ntfs_error(sb, "Busy inodes left. This is most likely a NTFS "
 				"driver bug.");
 		/* Copied from fs/super.c. I just love this message. (-; */
--- linux-2.6.32-openvz-039.10.orig/fs/smbfs/inode.c
+++ linux-2.6.32-openvz-039.10/fs/smbfs/inode.c
@@ -229,7 +229,7 @@ smb_invalidate_inodes(struct smb_sb_info
 {
 	VERBOSE("\n");
 	shrink_dcache_sb(SB_of(server));
-	invalidate_inodes(SB_of(server));
+	invalidate_inodes(SB_of(server), true);
 }
 
 /*
