From 0bb1a5f19e09dc553761ddd90bf6319eab94a597 Mon Sep 17 00:00:00 2001
From: Chris Wilson <chris@chris-wilson.co.uk>
Date: Wed, 09 Mar 2011 08:26:44 +0000
Subject: Update priv->stride after bo reallocation

Signed-off-by: Chris Wilson <chris@chris-wilson.co.uk>
---
diff --git a/src/intel_uxa.c b/src/intel_uxa.c
index 754bb4e..13d8cf9 100644
--- a/src/intel_uxa.c
+++ b/src/intel_uxa.c
@@ -668,9 +668,7 @@ void intel_set_pixmap_bo(PixmapPtr pixmap, dri_bo * bo)
 		priv->bo = bo;
 		priv->stride = intel_pixmap_pitch(pixmap);
 
-		ret = drm_intel_bo_get_tiling(bo,
-					      &tiling,
-					      &swizzle_mode);
+		ret = drm_intel_bo_get_tiling(bo, &tiling, &swizzle_mode);
 		if (ret != 0) {
 			FatalError("Couldn't get tiling on bo %p: %s\n",
 				   bo, strerror(-ret));
@@ -796,6 +794,8 @@ static Bool intel_uxa_put_image(PixmapPtr pixmap,
 
 			if (tiling != I915_TILING_NONE)
 				drm_intel_bo_set_tiling(bo, &tiling, stride);
+			priv->stride = stride;
+			priv->tiling = tiling;
 
 			screen->ModifyPixmapHeader(pixmap,
 						   w, h,
--
cgit v0.8.3-6-g21f6
