diff -ru pommed-1.21~/pommed/Makefile pommed-1.21/pommed/Makefile
--- pommed-1.21~/pommed/Makefile	2008-09-28 14:25:56.000000000 +0200
+++ pommed-1.21/pommed/Makefile	2008-09-28 14:26:17.000000000 +0200
@@ -16,7 +16,7 @@
 
 INOTIFY_CFLAGS = $(shell test -e /usr/include/sys/inotify.h || echo -DNO_SYS_INOTIFY_H)
 
-CFLAGS = -g -O2 -Wall $(DBUS_CFLAGS) $(ALSA_CFLAGS) $(AUDIOFILE_CFLAGS) $(CONFUSE_CFLAGS) $(INOTIFY_CFLAGS)
+CFLAGS += -g -Wall $(DBUS_CFLAGS) $(ALSA_CFLAGS) $(AUDIOFILE_CFLAGS) $(CONFUSE_CFLAGS) $(INOTIFY_CFLAGS)
 
 LDFLAGS = -pthread -lrt $(DBUS_LIBS) $(ALSA_LIBS) $(AUDIOFILE_LIBS) $(CONFUSE_LIBS)
 
@@ -50,6 +50,7 @@
 
 	CFLAGS += $(LIBPCI_CFLAGS)
 	LDFLAGS += $(LIBPCI_LIBS)
+	LIBS = $(LIBPCI_LIBS)
 else
 	LIBS = /usr/lib/libpci.a
 	LDFLAGS += -lz	
