--- webkit/glue/user_agent.cc.old	2010-09-28 03:56:01.000000000 -0500
+++ webkit/glue/user_agent.cc	2010-10-04 11:25:29.000000000 -0500
@@ -103,7 +103,7 @@
   StringAppendF(
       result,
       "Mozilla/5.0 (%s; %c; %s; %s) AppleWebKit/%d.%d"
-      " (KHTML, like Gecko) %s Safari/%d.%d",
+      " (KHTML, like Gecko) Sabayon %s Safari/%d.%d",
       mimic_windows ? "Windows" : kUserAgentPlatform,
       kUserAgentSecurity,
       ((mimic_windows ? "Windows " : "") + BuildOSCpuInfo()).c_str(),
