--- webkit/glue/user_agent.cc.orig	2011-03-19 03:28:25.000000000 -0500
+++ webkit/glue/user_agent.cc	2011-03-24 15:02:31.172764189 -0500
@@ -125,7 +125,7 @@
   base::StringAppendF(
       result,
       "Mozilla/5.0 (%s%s) AppleWebKit/%d.%d"
-      " (KHTML, like Gecko) %s Safari/%d.%d",
+      " (KHTML, like Gecko) Sabayon %s Safari/%d.%d",
       mimic_windows ? "Windows " : kUserAgentPlatform,
       BuildOSCpuInfo().c_str(),
       WEBKIT_VERSION_MAJOR,
