--- webkit/glue/plugins/plugin_list.cc_posix.orig	2010-10-09 11:05:44.000000000 +0200
+++ webkit/plugins/npapi/plugin_list_posix.cc	2010-10-09 11:06:35.000000000 +0200
@@ -89,7 +89,9 @@
 bool IsBlacklistedPlugin(const FilePath& path) {
   const char* kBlackListedPlugins[] = {
     "nppdf.so",           // Adobe PDF
+#ifndef GENTOO_CHROMIUM_ENABLE_GECKO_MEDIAPLAYER
     "gecko-mediaplayer",  // Gecko Media Player
+#endif  // GENTOO_CHROMIUM_ENABLE_GECKO_MEDIAPLAYER
   };
   std::string filename = path.BaseName().value();
   for (size_t i = 0; i < arraysize(kBlackListedPlugins); i++) {
