--- base/i18n/word_iterator.cc.orig	2010-12-17 07:27:02.000000000 +0000
+++ base/i18n/word_iterator.cc	2010-12-17 07:29:03.000000000 +0000
@@ -21,7 +21,7 @@
 
 WordIterator::~WordIterator() {
   if (iter_)
-    ubrk_close(iter_);
+    ubrk_close(static_cast<UBreakIterator*>(iter_));
 }
 
 bool WordIterator::Init() {
@@ -45,13 +45,13 @@
     NOTREACHED() << "ubrk_open failed";
     return false;
   }
-  ubrk_first(iter_);  // Move the iterator to the beginning of the string.
+  ubrk_first(static_cast<UBreakIterator*>(iter_));  // Move the iterator to the beginning of the string.
   return true;
 }
 
 bool WordIterator::Advance() {
   prev_ = pos_;
-  const int32_t pos = ubrk_next(iter_);
+  const int32_t pos = ubrk_next(static_cast<UBreakIterator*>(iter_));
   if (pos == UBRK_DONE) {
     pos_ = npos;
     return false;
@@ -62,7 +62,7 @@
 }
 
 bool WordIterator::IsWord() const {
-  return (ubrk_getRuleStatus(iter_) != UBRK_WORD_NONE);
+  return (ubrk_getRuleStatus(static_cast<UBreakIterator*>(iter_)) != UBRK_WORD_NONE);
 }
 
 string16 WordIterator::GetWord() const {
