--- build/linux/system.gyp.orig	2011-06-28 07:58:51.000000000 +0000
+++ build/linux/system.gyp	2011-06-28 08:00:05.000000000 +0000
@@ -314,7 +314,7 @@
       'target_name': 'libgcrypt',
       'type': 'settings',
       'conditions': [
-        ['_toolset=="target"', {
+        ['_toolset=="target" and use_cups==1', {
           'direct_dependent_settings': {
             'cflags': [
               '<!@(libgcrypt-config --cflags)',
