Index: third_party/WebKit/Source/WebCore/platform/audio/AudioArray.h
===================================================================
--- third_party/WebKit/Source/WebCore/platform/audio/AudioArray.h	(revision 105431)
+++ third_party/WebKit/Source/WebCore/platform/audio/AudioArray.h	(revision 118455)
@@ -61,6 +61,9 @@
         unsigned initialSize = sizeof(T) * n;
 
-        // 16-byte alignment for 128bit SIMD.
+#if USE(WEBAUDIO_FFMPEG)
+        const size_t alignment = 32;
+#else
         const size_t alignment = 16;
+#endif
 
         if (m_allocation)
