--- build/linux/system.gyp.orig	2010-09-07 01:22:40.000000000 +0200
+++ build/linux/system.gyp	2010-09-07 01:31:23.000000000 +0200
@@ -3,6 +3,10 @@
 # found in the LICENSE file.
 
 {
+  'variables': {
+    'use_gnome_keyring%': 1,
+  },
+
   'conditions': [
     ['sysroot!=""', {
       'variables': {
@@ -259,11 +263,14 @@
       'target_name': 'gnome-keyring',
       'type': 'settings',
       'conditions': [
-        ['chromeos==0', {
+        ['chromeos==0 and use_gnome_keyring==1', {
           'direct_dependent_settings': {
             'cflags': [
               '<!@(<(pkg-config) --cflags gnome-keyring-1)',
             ],
+            'defines': [
+              'USE_GNOME_KEYRING',
+            ],
             'conditions': [
               ['linux_link_gnome_keyring==0', {
                 'defines': ['DLOPEN_GNOME_KEYRING'],
--- chrome/chrome_browser.gypi.orig	2010-09-07 02:57:21.000000000 +0200
+++ chrome/chrome_browser.gypi	2010-09-07 02:58:23.000000000 +0200
@@ -3062,6 +3062,12 @@
             ],
           },
         }],
+        ['chromeos==1 or use_gnome_keyring==0', {
+          'sources!': [
+            'browser/password_manager/native_backend_gnome_x.h',
+            'browser/password_manager/native_backend_gnome_x.cc',
+          ],
+        }],
         ['touchui==0', {
           'sources!': [
             # Nothing yet.
@@ -3096,7 +3102,7 @@
             'third_party/mozilla_security_manager/nsUsageArrayHelper.h',
           ],
           'conditions': [
-            ['chromeos==0', {
+            ['chromeos==0 and use_gnome_keyring==1', {
               'dependencies': [
                 '../build/linux/system.gyp:gnome-keyring',
               ],
--- chrome/browser/profile_impl.cc.orig	2010-09-07 01:35:40.000000000 +0200
+++ chrome/browser/profile_impl.cc	2010-09-07 01:36:26.000000000 +0200
@@ -89,7 +89,9 @@
 #include "chrome/browser/password_manager/password_store_mac.h"
 #elif defined(OS_POSIX) && !defined(OS_CHROMEOS)
 #include "base/xdg_util.h"
+#if defined(USE_GNOME_KEYRING)
 #include "chrome/browser/password_manager/native_backend_gnome_x.h"
+#endif
 #include "chrome/browser/password_manager/native_backend_kwallet_x.h"
 #include "chrome/browser/password_manager/password_store_x.h"
 #endif
@@ -923,12 +925,14 @@
       backend.reset();
   } else if (desktop_env == base::DESKTOP_ENVIRONMENT_GNOME ||
              desktop_env == base::DESKTOP_ENVIRONMENT_XFCE) {
+#if defined(USE_GNOME_KEYRING)
     LOG(INFO) << "Trying GNOME keyring for password storage.";
     backend.reset(new NativeBackendGnome());
     if (backend->Init())
       LOG(INFO) << "Using GNOME keyring for password storage.";
     else
       backend.reset();
+#endif  // defined(USE_GNOME_KEYRING)
   }
   // TODO(mdm): this can change to a WARNING when we detect by default.
   if (!backend.get())
