--- remoting/base/encoder_vp8.cc.orig	2010-11-10 09:52:56.000000000 +0100
+++ remoting/base/encoder_vp8.cc	2010-11-10 09:53:32.000000000 +0100
@@ -11,9 +11,9 @@
 
 extern "C" {
 #define VPX_CODEC_DISABLE_COMPAT 1
-#include "third_party/libvpx/include/vpx/vpx_codec.h"
-#include "third_party/libvpx/include/vpx/vpx_encoder.h"
-#include "third_party/libvpx/include/vpx/vp8cx.h"
+#include "vpx/vpx_codec.h"
+#include "vpx/vpx_encoder.h"
+#include "vpx/vp8cx.h"
 }
 
 namespace remoting {
--- remoting/base/decoder_vp8.cc.orig	2010-11-10 09:54:14.000000000 +0100
+++ remoting/base/decoder_vp8.cc	2010-11-10 09:54:37.000000000 +0100
@@ -10,9 +10,9 @@
 
 extern "C" {
 #define VPX_CODEC_DISABLE_COMPAT 1
-#include "third_party/libvpx/include/vpx/vpx_codec.h"
-#include "third_party/libvpx/include/vpx/vpx_decoder.h"
-#include "third_party/libvpx/include/vpx/vp8dx.h"
+#include "vpx/vpx_codec.h"
+#include "vpx/vpx_decoder.h"
+#include "vpx/vp8dx.h"
 }
 
 namespace remoting {
--- third_party/ffmpeg/ffmpeg_stub_headers.fragment.orig	2010-11-10 09:55:43.000000000 +0100
+++ third_party/ffmpeg/ffmpeg_stub_headers.fragment	2010-11-10 09:56:09.000000000 +0100
@@ -7,9 +7,9 @@
 #include <libavformat/avformat.h>
 
 #define VPX_CODEC_DISABLE_COMPAT 1
-#include "third_party/libvpx/include/vpx/vpx_codec.h"
-#include "third_party/libvpx/include/vpx/vpx_encoder.h"
-#include "third_party/libvpx/include/vpx/vpx_decoder.h"
+#include "vpx/vpx_codec.h"
+#include "vpx/vpx_encoder.h"
+#include "vpx/vpx_decoder.h"
 
 typedef int (*AVLockMgrOperation)(void**, enum AVLockOp);
 
--- third_party/ffmpeg/ffmpeg.gyp.orig	2010-11-10 11:53:27.000000000 +0100
+++ third_party/ffmpeg/ffmpeg.gyp	2010-11-10 11:55:07.000000000 +0100
@@ -902,10 +902,20 @@
               'cflags': [
                 '<!@(pkg-config --cflags libavcodec libavformat libavutil)',
               ],
+              'link_settings': {
+                'libraries': [
+                  '-lvpx',
+                ],
+              },
               'direct_dependent_settings': {
                 'cflags': [
                   '<!@(pkg-config --cflags libavcodec libavformat libavutil)',
                 ],
+                'link_settings': {
+                  'libraries': [
+                    '-lvpx',
+                  ],
+                },
               },
             }, {  # else use_system_ffmpeg==0, add local copy to include path
               'include_dirs': [
