--- webkit/glue/plugins/pepper_private.cc.orig	2010-10-06 01:20:05.000000000 +0200
+++ webkit/glue/plugins/pepper_private.cc	2010-10-06 01:20:39.000000000 +0200
@@ -6,13 +6,14 @@
 
 #include "webkit/glue/plugins/pepper_private.h"
 
+#include "unicode/usearch.h"
+
 #include "app/resource_bundle.h"
 #include "base/utf_string_conversions.h"
 #include "grit/webkit_resources.h"
 #include "grit/webkit_strings.h"
 #include "skia/ext/platform_canvas.h"
 #include "third_party/skia/include/core/SkBitmap.h"
-#include "third_party/icu/public/i18n/unicode/usearch.h"
 #include "webkit/glue/webkit_glue.h"
 #include "webkit/glue/plugins/pepper_image_data.h"
 #include "webkit/glue/plugins/pepper_plugin_module.h"
