--- tools/gyp/pylib/gyp/generator/make.py.orig	2010-08-10 23:58:52.000000000 +0200
+++ tools/gyp/pylib/gyp/generator/make.py	2010-08-11 00:00:15.000000000 +0200
@@ -656,7 +656,7 @@
     part_of_all: flag indicating this target is part of 'all'
     """
     for action in actions:
-      name = self.target + '_' + action['action_name']
+      name = self.target + '_' + action['action_name'].replace(' ', '_')
       self.WriteLn('### Rules for action "%s":' % action['action_name'])
       inputs = action['inputs']
       outputs = action['outputs']
@@ -725,7 +725,7 @@
     part_of_all: flag indicating this target is part of 'all'
     """
     for rule in rules:
-      name = self.target + '_' + rule['rule_name']
+      name = self.target + '_' + rule['rule_name'].replace(' ', '_')
       count = 0
       self.WriteLn('### Generated for rule %s:' % name)
 
