diff -Nurp chromium-16.0.912.4.orig/webkit/glue/user_agent.cc chromium-16.0.912.4/webkit/glue/user_agent.cc
--- chromium-16.0.912.4.orig/webkit/glue/user_agent.cc	2011-10-21 19:16:42.000000000 +0200
+++ chromium-16.0.912.4/webkit/glue/user_agent.cc	2011-10-23 22:39:09.000000000 +0200
@@ -134,7 +134,7 @@ std::string BuildUserAgentFromProduct(co
   base::StringAppendF(
       &user_agent,
       "Mozilla/5.0 (%s%s) AppleWebKit/%d.%d"
-      " (KHTML, like Gecko) %s Safari/%d.%d",
+      " (KHTML, like Gecko) Sabayon %s Safari/%d.%d",
       kUserAgentPlatform,
       webkit_glue::BuildOSCpuInfo().c_str(),
       WEBKIT_VERSION_MAJOR,
