--- webkit/glue/plugins/plugin_list_posix.cc.orig	2010-08-11 02:50:36.000000000 +0200
+++ webkit/glue/plugins/plugin_list_posix.cc	2010-08-11 02:51:37.000000000 +0200
@@ -146,6 +146,7 @@
   // This varies across different browsers and versions, so check 'em all.
   plugin_dirs->push_back(FilePath("/usr/lib/browser-plugins"));
   plugin_dirs->push_back(FilePath("/usr/lib/mozilla/plugins"));
+  plugin_dirs->push_back(FilePath("/usr/lib/nsbrowser/plugins"));
   plugin_dirs->push_back(FilePath("/usr/lib/firefox/plugins"));
   plugin_dirs->push_back(FilePath("/usr/lib/xulrunner-addons/plugins"));
 
@@ -154,6 +155,7 @@
   // But a user reported on their Fedora system they are separate.
   plugin_dirs->push_back(FilePath("/usr/lib64/browser-plugins"));
   plugin_dirs->push_back(FilePath("/usr/lib64/mozilla/plugins"));
+  plugin_dirs->push_back(FilePath("/usr/lib64/nsbrowser/plugins"));
   plugin_dirs->push_back(FilePath("/usr/lib64/firefox/plugins"));
   plugin_dirs->push_back(FilePath("/usr/lib64/xulrunner-addons/plugins"));
 #endif
