--- content/browser/renderer_host/clipboard_message_filter.cc.orig	2011-04-21 12:29:49.000000000 +0200
+++ content/browser/renderer_host/clipboard_message_filter.cc	2011-04-21 12:31:14.000000000 +0200
@@ -4,6 +4,12 @@
 
 #include "content/browser/renderer_host/clipboard_message_filter.h"
 
+#if defined(USE_SYSTEM_ZLIB)
+#include <zlib.h>
+#else
+#include "third_party/zlib/zlib.h"
+#endif
+
 #include "base/stl_util-inl.h"
 #include "chrome/browser/browser_process.h"
 #include "chrome/browser/clipboard_dispatcher.h"
@@ -11,7 +17,6 @@
 #include "googleurl/src/gurl.h"
 #include "ipc/ipc_message_macros.h"
 #include "third_party/skia/include/core/SkBitmap.h"
-#include "third_party/zlib/zlib.h"
 #include "ui/gfx/codec/png_codec.h"
 #include "ui/gfx/size.h"
 
--- content/content_browser.gypi.orig	2011-04-21 12:32:18.000000000 +0200
+++ content/content_browser.gypi	2011-04-21 12:32:56.000000000 +0200
@@ -13,7 +13,8 @@
         '../ppapi/ppapi.gyp:ppapi_proxy',
         '../skia/skia.gyp:skia',
         '../third_party/flac/flac.gyp:libflac',
-        '../third_party/speex/speex.gyp:libspeex',
+        '../third_party/speex/speex.gyp:libspeex',	
+        '../third_party/zlib/zlib.gyp:zlib',
         '../third_party/WebKit/Source/WebKit/chromium/WebKit.gyp:webkit',
         '../ui/ui.gyp:ui_base',
       ],
