--- chromium-17.0.963.46.orig/breakpad/src/client/windows/build/common.gypi
+++ chromium-17.0.963.46/breakpad/src/client/windows/build/common.gypi
@@ -992,8 +992,6 @@
                   ['armv7==1', {
                     'cflags': [
                       '-march=armv7-a',
-                      '-mtune=cortex-a8',
-                      '-mfloat-abi=softfp',
                     ],
                     'conditions': [
                       ['arm_neon==1', {
--- chromium-17.0.963.46.orig/build/common.gypi
+++ chromium-17.0.963.46/build/common.gypi
@@ -1816,8 +1816,6 @@
                   ['armv7==1', {
                     'cflags': [
                       '-march=armv7-a',
-                      '-mtune=cortex-a8',
-                      '-mfloat-abi=softfp',
                     ],
                     'conditions': [
                       ['arm_neon==1', {
--- chromium-17.0.963.46.orig/native_client/build/common.gypi
+++ chromium-17.0.963.46/native_client/build/common.gypi
@@ -287,9 +287,7 @@
               'cflags': [
                 '-Wno-abi',
                 '-march=armv7-a',
-                '-mtune=cortex-a8',
                 '-mfpu=neon',
-                '-mfloat-abi=softfp',
                 '-fno-exceptions',
                 '-Wall',
                 '-fPIC',
--- chromium-17.0.963.46.orig/third_party/ffmpeg/ffmpeg.gyp
+++ chromium-17.0.963.46/third_party/ffmpeg/ffmpeg.gyp
@@ -169,8 +169,6 @@
                 '-fPIC',
                 '-mthumb',
                 '-march=armv7-a',
-                '-mtune=cortex-a8',
-                '-mfloat-abi=softfp',
               ],
               'conditions': [
                 ['arm_neon == 0', {
--- chromium-17.0.963.46.orig/third_party/libvpx/source/libvpx/build/make/configure.sh
+++ chromium-17.0.963.46/third_party/libvpx/source/libvpx/build/make/configure.sh
@@ -670,8 +670,8 @@ process_common_toolchain() {
                 check_add_asflags -mcpu=${tgt_isa}
             elif enabled armv7
             then
-                check_add_cflags -march=armv7-a -mcpu=cortex-a8 -mfpu=neon -mfloat-abi=softfp  #-ftree-vectorize
-                check_add_asflags -mcpu=cortex-a8 -mfpu=neon -mfloat-abi=softfp  #-march=armv7-a
+                check_add_cflags -march=armv7-a -mcpu=cortex-a8 -mfpu=neon #-ftree-vectorize
+                check_add_asflags -mcpu=cortex-a8 -mfpu=neon #-march=armv7-a
             else
                 check_add_cflags -march=${tgt_isa}
                 check_add_asflags -march=${tgt_isa}
