--- webkit/glue/plugins/pepper_private.cc.orig	2010-10-13 13:22:28.000000000 +0200
+++ webkit/glue/plugins/pepper_private.cc	2010-10-13 13:22:56.000000000 +0200
@@ -6,6 +6,8 @@
 
 #include "webkit/glue/plugins/pepper_private.h"
 
+#include "unicode/usearch.h"
+
 #include "app/resource_bundle.h"
 #include "base/utf_string_conversions.h"
 #include "grit/webkit_resources.h"
@@ -13,7 +15,6 @@
 #include "skia/ext/platform_canvas.h"
 #include "third_party/ppapi/c/pp_resource.h"
 #include "third_party/skia/include/core/SkBitmap.h"
-#include "third_party/icu/public/i18n/unicode/usearch.h"
 #include "webkit/glue/webkit_glue.h"
 #include "webkit/glue/plugins/pepper_image_data.h"
 #include "webkit/glue/plugins/pepper_plugin_delegate.h"
