diff -ru ../chromium-5.0.322.2/tools/gyp/pylib/gyp/generator/make.py ./tools/gyp/pylib/gyp/generator/make.py
--- ../chromium-5.0.322.2/tools/gyp/pylib/gyp/generator/make.py	2010-01-25 23:04:11.000000000 +0100
+++ ./tools/gyp/pylib/gyp/generator/make.py	2010-02-19 00:48:45.381303499 +0100
@@ -148,9 +148,9 @@
 define fixup_dep
 sed -i -e "s|^$(notdir $@)|$@|" $(depfile).tmp
 sed -e "s|^[^:]*: *||" -e "s| *\\\\$$||" -e 's|^ *||' \
-    -e "/./s|$$|:|" $(depfile).tmp >> $(depfile).tmp
-cat $(depfile).tmp >> $(depfile)
-rm -f $(depfile).tmp
+    -e "/./s|$$|:|" $(depfile).tmp > $(depfile).tmpXXX
+cat $(depfile).tmpXXX >> $(depfile)
+rm -f $(depfile).tmpXXX
 endef
 """
 """
