--- os-prober-1.53/os-probes/mounted/common/90linux-distro	2011-07-23 04:01:00.000000000 +0300
+++ os-prober-1.53.new/os-probes/mounted/common/90linux-distro	2015-04-21 23:22:57.835424199 +0300
@@ -85,6 +85,12 @@
 	elif [ -e "$dir/etc/gentoo-release" ]; then
 		short="Gentoo"
 		long="$(cat "$dir/etc/gentoo-release")"
+	elif [ -e "$dir/etc/kogaion-release" ]; then
+		short="Kogaion"
+		long="$(cat "$dir/etc/kogaion-release")"
+       elif [ -e "$dir/etc/sabayon-release" ]; then
+               short="Sabayon"
+               long="$(cat "$dir/etc/sabayon-release")"
+       elif [ -e "$dir/etc/spike-release" ]; then
+               short="Spike"
+               long="$(cat "$dir/etc/spike-release")"
+	elif [ -e "$dir/etc/argent-release" ]; then
+		short="Argent"
+		long="$(cat "$dir/etc/argent-release")"
 	elif [ -e "$dir/etc/cobalt-release" ]; then
 		short="Cobalt"
 		long="$(cat "$dir/etc/cobalt-release")"
