diff -ur Singular-3-0-4.old/kernel/febase.cc Singular-3-0-4/kernel/febase.cc
--- Singular-3-0-4.old/kernel/febase.cc	2009-07-07 01:15:49.088761926 +0300
+++ Singular-3-0-4/kernel/febase.cc	2009-07-07 01:16:51.157012952 +0300
@@ -743,13 +743,13 @@
       char* dir_sep;
       struct passwd *pw_entry;
       strcpy (longpath, path);
-      dir_sep = strchr(longpath, DIR_SEP);
+      dir_sep = const_cast<char*> ( strchr(longpath, DIR_SEP) );  // TODO: fix the ugly casting away of const
       *dir_sep = '\0';
       pw_entry = getpwnam(&longpath[1]);
       if (pw_entry != NULL)
       {
         strcpy(longpath, pw_entry->pw_dir);
-        dir_sep = strchr(path, DIR_SEP);
+        dir_sep = const_cast<char*> ( strchr(path, DIR_SEP) );  // TODO: fix the ugly casting away of const
         strcat(longpath, dir_sep);
         path = longpath;
       }
diff -ur Singular-3-0-4.old/kernel/mpr_complex.cc Singular-3-0-4/kernel/mpr_complex.cc
--- Singular-3-0-4.old/kernel/mpr_complex.cc	2009-07-07 01:34:11.290760532 +0300
+++ Singular-3-0-4/kernel/mpr_complex.cc	2009-07-07 01:55:24.209129950 +0300
@@ -82,7 +82,7 @@
   BOOLEAN neg=false;
   if (*in == '-') { in++; neg=TRUE; }
   char *s;
-  if ((s=strchr(in,'E')) !=NULL)
+  if ((s = const_cast<char*> ( strchr(in,'E') )) !=NULL)  /* TODO: fix the ugly casting away of const */
   {
     *s='e';
   }
