diff -Naur Singular-3-0-4/emacs/.emacs-singular Singular-3-0-4.new/emacs/.emacs-singular
--- Singular-3-0-4/emacs/.emacs-singular	2004-12-09 11:27:54.000000000 -0500
+++ Singular-3-0-4.new/emacs/.emacs-singular	2008-11-07 18:06:50.000000000 -0500
@@ -96,10 +96,10 @@
        ;; unix stuff 
 
        ;; lazy-lock seems to be broken under mswindows:
-       ;; causes delay of input (only reacts after mouse click), or even total 
+       ;; causes delay of input (only reacts after mouse click), or even total
        ;; hang
-       (add-hook 'font-lock-mode-hook 'lazy-lock-mode)
-       (setq font-lock-support-mode 'lazy-lock-mode)
+       ;(add-hook 'font-lock-mode-hook 'lazy-lock-mode)
+       ;(setq font-lock-support-mode 'lazy-lock-mode)
        
        ;; somewhat nicer scrolling This causes a core dump with 21.1.9
        ;; under cygwin, when a file is opened with a button 
diff -Naur Singular-3-0-4/emacs/singular.el Singular-3-0-4.new/emacs/singular.el
--- Singular-3-0-4/emacs/singular.el	2000-12-18 10:44:32.000000000 -0500
+++ Singular-3-0-4.new/emacs/singular.el	2008-11-07 18:06:02.000000000 -0500
@@ -865,7 +865,9 @@
   (add-hook 'singular-post-output-filter-functions 'singular-prompt-remove-filter nil t)
 
   ;; some relict from Comint mode
-  (setq comint-prompt-regexp singular-comint-prompt-regexp))
+  (setq comint-prompt-regexp singular-comint-prompt-regexp)
+  (setq comint-use-prompt-regexp-instead-of-fields 1)
+  )
 ;;}}}
 
 ;;{{{ Miscellaneous
