--- ximian-artwork-0.2.32/configure.ac	2003-08-29 19:02:58.000000000 +0100
+++ ximian-artwork-0.2.32/configure.ac.new	2005-06-25 14:16:51.110708216 +0100
@@ -19,30 +19,11 @@
 
 AC_PROG_INTLTOOL([0.23])
 
-AM_PATH_GTK(1.2.9, , 
-        AC_MSG_ERROR([*** GTK+-1.2.9 not installed - please install first ***]))
-AC_CHECK_LIB(gtk, gtk_style_set_prop_experimental, :,
-             AC_MSG_ERROR([*** gtk_style_set_prop_experimental() not found; perhaps you have an old CVS version of GTK+ ***]),
-             $GTK_LIBS)
-
-PKG_CHECK_MODULES(XIMART, gtk+-2.0)
-AC_SUBST(XIMART_CFLAGS)
-AC_SUBST(XIMART_LIBS)
-
-GTK2_VERSION=`$PKG_CONFIG --variable=gtk_binary_version gtk+-2.0`
-AC_SUBST(GTK2_VERSION)
-
 AC_CONFIG_FILES([
 Makefile
 galeon/Makefile
 gdm/Makefile
 gnome-session/Makefile
-gtk/Makefile
-gtk/gtk1-engine/Makefile
-gtk/gtk1-engine/gtk/Makefile
-gtk/gtk2-engine/Makefile
-gtk/gtk2-engine/gtk-2.0/Makefile
-gtk/gtk2-engine/gtk-2.0/gtkrc
 icons/Makefile
 icons/Industrial/Makefile
 icons/Industrial/24x24/Makefile
--- ximian-artwork-0.2.32/Makefile.am	2003-02-21 19:53:29.000000000 +0000
+++ ximian-artwork-0.2.32/Makefile.am.new	2005-06-25 14:17:31.967099421 +0100
@@ -1,4 +1,4 @@
-SUBDIRS = po gdm gnome-session gtk icons galeon metacity nautilus wallpapers xmms
+SUBDIRS = po gdm gnome-session icons galeon metacity nautilus xmms
 
 THEME_IN_FILES=index.theme.in
 
