diff -uNr ximian-artwork-0.2.26/gtk/gtk2-engine/industrial_style.c ximian-artwork-0.2.26.patched/gtk/gtk2-engine/industrial_style.c
--- ximian-artwork-0.2.26/gtk/gtk2-engine/industrial_style.c	2003-05-19 23:34:47.000000000 +0200
+++ ximian-artwork-0.2.26.patched/gtk/gtk2-engine/industrial_style.c	2003-08-11 02:23:35.000000000 +0200
@@ -1242,9 +1242,11 @@
 	       gint           height,
 	       gboolean       fill)
 {
-  if (shadow_type == GTK_SHADOW_NONE)
+  if (shadow_type == GTK_SHADOW_NONE) {
+	parent_class->draw_flat_box (style, window, state_type, shadow_type, area, widget, detail, x, y, width, height);
     return;
-  
+}
+
   g_return_if_fail (style != NULL);
   g_return_if_fail (window != NULL);
 
