--- js/src/xpconnect/src/Makefile.in~	2009-09-20 13:13:15.508713367 -0500
+++ js/src/xpconnect/src/Makefile.in	2009-09-20 13:14:30.994713367 -0500
@@ -77,6 +77,10 @@
 		   widget \
 		   $(NULL)
 
+ifdef MOZ_XPCTOOLS
+DEFINES         +=  -DXPC_TOOLS_SUPPORT
+REQUIRES	+= xpctools
+endif
 
 CPPSRCS		= \
 		nsScriptError.cpp \
@@ -174,11 +178,6 @@
 
 endif # ENABLE_JIT
 
-ifdef MOZ_XPCTOOLS
-DEFINES         +=  -DXPC_TOOLS_SUPPORT
-REQUIRES	+= xpctools
-endif
-
 ifdef XPC_IDISPATCH_SUPPORT
 DEFINES		+= -DXPC_IDISPATCH_SUPPORT
 ifdef XPC_COMOBJECT
