diff -r 4e4c7457e8f7 js/src/ctypes/libffi/src/arm/sysv.S
--- a/js/src/ctypes/libffi/src/arm/sysv.S	Sat Apr 02 11:48:22 2011 -0400
+++ b/js/src/ctypes/libffi/src/arm/sysv.S	Sun Apr 03 04:55:15 2011 +0300
@@ -229,6 +229,10 @@
         .size    CNAME(ffi_call_SYSV),.ffi_call_SYSV_end-CNAME(ffi_call_SYSV)
 
 
+/* Below are VFP hard-float ABI call and closure implementations.
+   Add VFP FPU directive here. */
+	.fpu	vfp
+
 	@ r0:   fn
 	@ r1:   &ecif
 	@ r2:   cif->bytes
