http://marc.info/?l=opensolaris-desktop-discuss&m=124625949400704&w=2

--- firefox/js/src/nanojit/Nativei386.h.orig	2008-12-19 16:11:09.743712966 +0800
+++ firefox/js/src/nanojit/Nativei386.h	2008-12-19 16:12:32.156744648 +0800
@@ -40,6 +40,14 @@
 #ifndef __nanojit_Nativei386__
 #define __nanojit_Nativei386__
 
+#undef EAX
+#undef EBX
+#undef ECX
+#undef EDX
+#undef ESP
+#undef EBP
+#undef ESI
+#undef EDI
 
 namespace nanojit
 {
