diff -Nurp mozilla-2.0.orig/netwerk/protocol/http/nsHttpHandler.cpp mozilla-2.0/netwerk/protocol/http/nsHttpHandler.cpp
--- mozilla-2.0.orig/netwerk/protocol/http/nsHttpHandler.cpp	2011-03-19 00:34:03.000000000 +0100
+++ mozilla-2.0/netwerk/protocol/http/nsHttpHandler.cpp	2011-03-30 08:40:39.000000000 +0200
@@ -651,6 +651,9 @@ nsHttpHandler::BuildUserAgent()
     mUserAgent += '/';
     mUserAgent += mProductSub;
 
+    // Sabayon portion
+    mUserAgent.AppendLiteral(" Sabayon");
+
     // "Firefox/x.y.z" compatibility token
     if (!mCompatFirefox.IsEmpty()) {
         mUserAgent += ' ';
