diff -urN mozilla-1.9.2-orig/content/base/src/nsDocument.cpp mozilla-1.9.2/content/base/src/nsDocument.cpp
--- mozilla-1.9.2-orig/content/base/src/nsDocument.cpp	2010-01-28 22:16:08.683135912 -0600
+++ mozilla-1.9.2/content/base/src/nsDocument.cpp	2010-01-28 22:17:57.605126833 -0600
@@ -169,7 +169,9 @@
 #include "nsIPropertyBag2.h"
 #include "nsIDOMPageTransitionEvent.h"
 #include "nsFrameLoader.h"
+#ifdef MOZ_MEDIA
 #include "nsHTMLMediaElement.h"
+#endif
 
 #include "mozAutoDocUpdate.h"
 
diff -urN mozilla-1.9.2-orig/content/base/src/nsNodeUtils.cpp mozilla-1.9.2/content/base/src/nsNodeUtils.cpp
--- mozilla-1.9.2-orig/content/base/src/nsNodeUtils.cpp	2010-01-28 22:16:08.686138846 -0600
+++ mozilla-1.9.2/content/base/src/nsNodeUtils.cpp	2010-01-28 22:17:14.618150370 -0600
@@ -55,7 +55,9 @@
 #endif
 #include "nsBindingManager.h"
 #include "nsGenericHTMLElement.h"
+#ifdef MOZ_MEDIA
 #include "nsHTMLMediaElement.h"
+#endif
 
 // This macro expects the ownerDocument of content_ to be in scope as
 // |nsIDocument* doc|
