diff -urN mozilla.orig/xpcom/glue/standalone/nsGlueLinkingDlopen.cpp mozilla.dwokfur/xpcom/glue/standalone/nsGlueLinkingDlopen.cpp
--- mozilla.orig/xpcom/glue/standalone/nsGlueLinkingDlopen.cpp	2009-08-23 08:56:52.000000000 +0200
+++ mozilla.dwokfur/xpcom/glue/standalone/nsGlueLinkingDlopen.cpp	2009-08-23 10:00:46.000000000 +0200
@@ -87,7 +87,8 @@
 XPCOMGlueLoad(const char *xpcomFile)
 {
     char xulFile[MAXPATHLEN];
-    char *lastSlash = strrchr(xpcomFile, '/');
+    char *xpcomFileb = const_cast<char*>(xpcomFile);
+    char *lastSlash = strrchr(xpcomFileb, '/');
     if (!lastSlash) {
         snprintf(xulFile, MAXPATHLEN, "./%s", xpcomFile);
         return XPCOMGlueLoad(xulFile);

