Index: src/bin/e_start_main.c
===================================================================
--- src/bin/e_start_main.c	(revisione 62886)
+++ src/bin/e_start_main.c	(copia locale)
@@ -60,14 +60,11 @@
 {
    char *p;
 
-   if (!_prefix_try_proc())
-     {
-	if (!_prefix_try_argv(argv0))
-	  {
-	     _prefix_fallbacks();
-	     return 0;
-	  }
-     }
+   if (!_prefix_try_argv(argv0))
+   {
+     _prefix_fallbacks();
+     return 0;
+   }
    /* _exe_path is now a full absolute path TO this exe - figure out rest */
    /*   if
     * exe        = /blah/whatever/bin/exe
