--- rss-glx_0.8.1/configure.in.orig	2006-02-25 16:10:54.000000000 -0500
+++ rss-glx_0.8.1/configure.in	2008-05-11 00:32:26.000000000 -0400
@@ -7,14 +7,12 @@
 AC_PROG_AWK
 
 AC_PROG_CC
-CFLAGS=-g
 AC_ARG_ENABLE([debug], [AC_HELP_STRING([--enable-debug],
     [build with debugging information default=yes])],
     [test "$enableval" = "no" && CFLAGS="-O2"])
 CFLAGS="$CFLAGS -Wall"
 
 AC_PROG_CXX
-CXXFLAGS=-g
 AC_ARG_ENABLE([debug], [AC_HELP_STRING([--enable-debug],
     [build with debugging information default=yes])],
     [test "$enableval" = "no" && CXXFLAGS="-O2"])
--- rss-glx_0.8.1/reallyslick/cpp_src/Makefile.am.orig	2006-02-25 15:21:23.000000000 -0500
+++ rss-glx_0.8.1/reallyslick/cpp_src/Makefile.am	2008-05-11 00:35:33.000000000 -0400
@@ -19,7 +19,7 @@
 dist_config_DATA = cyclone.xml fieldlines.xml flocks.xml flux.xml helios.xml hyperspace.xml plasma.xml skyrocket.xml solarwinds.xml
 endif
 
-if INSTALL_XSCREENSAVER_CONFIG_FILES
+if INSTALL_KDE_CONFIG_FILES
 dist_kdessconfig_DATA = cyclone.desktop fieldlines.desktop flocks.desktop flux.desktop helios.desktop hyperspace.desktop plasma.desktop skyrocket.desktop solarwinds.desktop
 endif
 
--- rss-glx_0.8.1/reallyslick/cpp_src/skyrocket.cpp.orig	2006-02-25 17:01:10.000000000 -0500
+++ rss-glx_0.8.1/reallyslick/cpp_src/skyrocket.cpp	2008-05-11 00:39:04.000000000 -0400
@@ -23,6 +23,7 @@
 #include <GL/gl.h>
 #include <GL/glu.h>
 
+#include <cassert>
 #include <list>
 
 #include "driver.h"
