diff -Nurp Democracy-0.9.5.1.orig/platform/gtk-x11/democracyplayer Democracy-0.9.5.1/platform/gtk-x11/democracyplayer
--- Democracy-0.9.5.1.orig/platform/gtk-x11/democracyplayer	2007-02-20 20:56:49.000000000 +0100
+++ Democracy-0.9.5.1/platform/gtk-x11/democracyplayer	2007-02-25 18:17:50.000000000 +0100
@@ -75,28 +75,4 @@ elif options.profile:
     eventloop.profile_file = options.profile
     profile.run ('main()', options.profile)
 else:
-    try:
-	import dbus
-	try:
-	    import dbus_bindings
-	except ImportError, main_error:
-	    try:
-		dbus_bindings = dbus.dbus_bindings
-	    except:
-		raise main_error
-        import onetime
-        try:
-            onetime.OneTime()
-            startapp()
-        except onetime.NameExistsException:
-            bus = dbus.SessionBus()
-            proxy_obj = bus.get_object('org.participatoryculture.dtv.onetime', '/org/participatoryculture/dtv/OneTime')
-            iface = dbus.Interface(proxy_obj, 'org.participatoryculture.dtv.OneTimeIface')
-    
-            dbusargs = dbus.Array(args, signature="s")
-    
-            iface.HandleArgs(dbusargs)
-        except dbus_bindings.DBusException:
-            startapp()
-    except ImportError:
-        startapp()
+    startapp()
