From 13facad4fa3be24d07768892135caebd88e95fcc Mon Sep 17 00:00:00 2001
From: Tambet Ingo <tambet@gmail.com>
Date: Fri, 13 Feb 2009 08:54:21 +0000
Subject: Close the serial device on HUP.

---
diff --git a/src/mm-serial.c b/src/mm-serial.c
index fb3cad1..c81ca70 100644
--- a/src/mm-serial.c
+++ b/src/mm-serial.c
@@ -476,7 +476,13 @@ data_available (GIOChannel *source,
     gsize bytes_read;
     GIOStatus status;
 
-    if (condition & G_IO_HUP || condition & G_IO_ERR) {
+    if (condition & G_IO_HUP) {
+        g_string_truncate (priv->response, 0);
+        mm_serial_close (self);
+        return FALSE;
+    }
+
+    if (condition & G_IO_ERR) {
         g_string_truncate (priv->response, 0);
         return TRUE;
     }
--
cgit v0.8.2
