--- libs/libmythtv/dvbtypes.h.orig	2009-05-29 14:05:20.000000000 -0600
+++ libs/libmythtv/dvbtypes.h	2009-05-29 14:06:22.000000000 -0600
@@ -19,15 +19,15 @@
 #include <linux/dvb/frontend.h>
 #include <linux/dvb/dmx.h>
 
-#if (DVB_API_VERSION != 3)
-#    error "DVB driver includes with API version 3 not found!"
+#if (DVB_API_VERSION != 3 && DVB_API_VERSION != 5)
+#    error "DVB driver includes with API version 3 or 5 not found!"
 #endif
 
 #ifndef DVB_API_VERSION_MINOR
 #    define DVB_API_VERSION_MINOR 0
 #endif
 
-#if (DVB_API_VERSION >= 3 && DVB_API_VERSION_MINOR >= 1)
+#if ((DVB_API_VERSION == 3 && DVB_API_VERSION_MINOR >= 1) || (DVB_API_VERSION > 3))
 #    define USE_ATSC
 #else
 #warning DVB API version < 3.1
