Index: libs/libswscale/libswscale.pro
===================================================================
--- libs/libswscale/libswscale.pro	(revision 22859)
+++ libs/libswscale/libswscale.pro	(working copy)
@@ -14,6 +14,11 @@
         QMAKE_CFLAGS_DEBUG += -fomit-frame-pointer
 }
 
+# gcc-4.2 and newer can not compile with PIC on x86
+contains(ARCH_X86_32, yes) {
+        QMAKE_CFLAGS_SHLIB =
+}
+
 !profile:QMAKE_CFLAGS_DEBUG += -O
 
 INCLUDEPATH = .. ../..
Index: libs/libavcodec/libavcodec.pro
===================================================================
--- libs/libavcodec/libavcodec.pro	(revision 22859)
+++ libs/libavcodec/libavcodec.pro	(working copy)
@@ -25,7 +25,8 @@
         QMAKE_CFLAGS_DEBUG += -fomit-frame-pointer
 }
 # "-Os" can not compiled with PIC
-contains(CONFIG_SMALL, yes):contains(ARCH_X86_32, yes) {
+# gcc-4.2 and newer can not compile with PIC on x86
+contains(ARCH_X86_32, yes) {
 	QMAKE_CFLAGS_SHLIB =
 }
 
