diff -ruN a/programs/mythfrontend/mythfrontend.pro b/mythtv.new/programs/mythfrontend/mythfrontend.pro
--- a/programs/mythfrontend/mythfrontend.pro	2009-06-07 12:32:17.000000000 -0700
+++ b/programs/mythfrontend/mythfrontend.pro	2009-06-07 12:34:45.000000000 -0700
@@ -16,6 +16,9 @@
 setting.files += MFEXML_scpd.xml
 setting.extra = -ldconfig
 
+# Gentoo sandbox-2.0:
+setting.extra -= -ldconfig
+
 INSTALLS += setting
 
 QMAKE_CLEAN += $(TARGET)
