Index: libs/libmythtv/NuppelVideoRecorder.cpp
===================================================================
--- a/mythtv/libs/libmythtv/NuppelVideoRecorder.cpp	(revision 18814)
+++ b/mythtv/libs/libmythtv/NuppelVideoRecorder.cpp	(working copy)
@@ -981,7 +981,7 @@
             correct_bttv = true;
 
         QString driver = (char *)vcap.driver;
-        if (driver == "cx8800" || driver == "go7007" || driver == "em28xx")
+        if (driver == "cx8800" || driver == "go7007" || driver == "em28xx" || driver == "bttv")
         {
             channelfd = open(videodevice.ascii(), O_RDWR);
             if (channelfd < 0)
