--- android-sdk-linux_86/tools/android.old      2010-08-16 19:56:17.522813374 -0400
+++ android-sdk-linux_86/tools/android  2010-08-16 19:56:57.876813929 -0400
@@ -93,13 +93,7 @@
 if [ -n "$ANDROID_SWT" ]; then
     swtpath="$ANDROID_SWT"
 else
-    vmarch=`"${java_cmd}" -jar "${frameworkdir}/archquery.jar"`
-    if [ -n "$ANDROID_BUILD_TOP" ]; then
-        osname=`uname -s | tr A-Z a-z`
-        swtpath="${ANDROID_BUILD_TOP}/prebuilt/${osname}-${vmarch}/swt"
-    else
-        swtpath="${frameworkdir}/${vmarch}"
-    fi
+    swtpath=`dirname \`java-config -p swt-3.5\``
 fi

 # Combine the swtpath and the framework dir path.
--- android-sdk-linux_86/tools/ddms.old         2010-08-21 18:25:12.000000000 +0200
+++ android-sdk-linux_86/tools/ddms             2010-08-26 14:44:36.000000000 +0200
@@ -85,13 +85,7 @@
 if [ -n "$ANDROID_SWT" ]; then
     swtpath="$ANDROID_SWT"
 else
-    vmarch=`${javaCmd} -jar "${frameworkdir}"/archquery.jar`
-    if [ -n "$ANDROID_BUILD_TOP" ]; then
-        osname=`uname -s | tr A-Z a-z`
-        swtpath="${ANDROID_BUILD_TOP}/prebuilt/${osname}-${vmarch}/swt"
-    else
-        swtpath="${frameworkdir}/${vmarch}"
-    fi
+    swtpath=`dirname \`java-config -p swt-3.5\``
 fi
 
 # Combine the swtpath and the framework dir path.
