diff --git a/libraries/entropy/spm.py b/libraries/entropy/spm.py
index ba1d0fc..957849f 100644
--- a/libraries/entropy/spm.py
+++ b/libraries/entropy/spm.py
@@ -1686,6 +1686,29 @@ class PortagePlugin(SpmPlugin):
             mysettings['PORTAGE_TMPDIR'] = str(portage_tmpdir)
             mysettings.backup_changes("PORTAGE_TMPDIR")
 
+        # create FAKE ${PORTDIR} directory and licenses subdir
+        portdir = os.path.join(portage_tmpdir, "portdir")
+        portdir_lic = os.path.join(portdir, "licenses")
+        if not os.path.isdir(portdir):
+            os.mkdir(portdir) # portage_tmpdir must be available!
+        # create licenses subdir
+        if not os.path.isdir(portdir_lic):
+            os.mkdir(portdir_lic)
+
+        # set fake PORTDIR
+        old_portdir = mysettings["PORTDIR"][:]
+        mysettings["PORTDIR"] = portdir
+        mysettings.backup_changes("PORTDIR")
+
+        ### WORKAROUND for buggy check_license() in eutils.eclass
+        ### that looks for file availability before considering
+        ### ACCEPT_LICENSE
+        for lic in licenses:
+            lic_path = os.path.join(portdir_lic, lic)
+            if not os.access(lic_path, os.F_OK):
+                lic_f = open(lic_path, "w")
+                lic_f.close()
+
         mydbapi = self.portage.fakedbapi(settings=mysettings)
         mydbapi.cpv_inject(cpv, metadata = metadata)
 
@@ -1717,6 +1740,11 @@ class PortagePlugin(SpmPlugin):
         if portage_tmpdir_created:
             shutil.rmtree(portage_tmpdir,True)
 
+        # reset PORTDIR back to its old path
+        # for security !
+        mysettings["PORTDIR"] = old_portdir
+        mysettings.backup_changes("PORTDIR")
+
         del mydbapi
         del metadata
         del keys

