commit f225a7dab5eb49bf2a2cc71f35919f3dd09029a3
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Mon Aug 16 19:58:09 2010 +0200

    [entropy.fetchers] fix MultipleUrlFetcher output update function

diff --git a/libraries/entropy/fetchers.py b/libraries/entropy/fetchers.py
index ea0dbc2..2985b55 100644
--- a/libraries/entropy/fetchers.py
+++ b/libraries/entropy/fetchers.py
@@ -647,7 +647,7 @@ class MultipleUrlFetcher(TextInterface):
                 self.__multiple_fetcher = multiple
 
             def update(self):
-                return self.__multiple_fetcher.output()
+                return self.__multiple_fetcher.update()
 
             def _push_progress_to_output(self):
                 return
