diff -Nurp entropy-0.99.49.21.orig/libraries/entropy/db/__init__.py entropy-0.99.49.21/libraries/entropy/db/__init__.py
--- entropy-0.99.49.21.orig/libraries/entropy/db/__init__.py	2010-08-18 14:40:50.000000000 +0200
+++ entropy-0.99.49.21/libraries/entropy/db/__init__.py	2010-08-18 21:52:22.000000000 +0200
@@ -2612,7 +2612,7 @@ class EntropyRepository(EntropyRepositor
                 obj = cached.setdefault(pkg_id, set())
                 obj.add(flag)
             self.__setLiveCache("retrieveUseflags", cached)
-        return cached.get(package_id)
+        return cached.get(package_id, set())
 
     def retrieveEclasses(self, package_id):
         """
