diff -Nurp genkernel-3.4.9.orig/gen_initramfs.sh genkernel-3.4.9/gen_initramfs.sh
--- genkernel-3.4.9.orig/gen_initramfs.sh	2008-03-16 18:21:58.000000000 +0000
+++ genkernel-3.4.9/gen_initramfs.sh	2008-03-16 18:24:51.000000000 +0000
@@ -227,7 +227,7 @@ append_mdadm(){
 }
 
 append_splash(){
-	if [ -x /usr/bin/splash_geninitramfs ] || [ -x /sbin/splash_geninitramfs ]
+	if [ -x /usr/bin/splash_geninitramfs ] || [ -x /sbin/splash_geninitramfs ] || [ -x /usr/sbin/splash_geninitramfs ]
 	then
 		[ -z "${SPLASH_THEME}" ] && [ -e /etc/conf.d/splash ] && source /etc/conf.d/splash
 		[ -z "${SPLASH_THEME}" ] && SPLASH_THEME=default
