--- genkernel-3.4.6.orig/generic/linuxrc
+++ genkernel-3.4.6/generic/linuxrc
@@ -46,6 +46,7 @@ CMDLINE="`cat /proc/cmdline`"
 REAL_ROOT=''
 FAKE_ROOT=''
 REAL_ROOTFLAGS=''
+REAL_INITD="/sbin/init"
 for x in ${CMDLINE}
 do
 	case "${x}" in
@@ -61,6 +62,9 @@ do
 		real_init\=*)
 			REAL_INIT=`parse_opt "${x}"`
 		;;
+		accel_init\=*)
+			REAL_INITD=`parse_opt "${x}"`
+		;;
 		init_opts\=*)
 			INIT_OPTS=`parse_opt "${x}"`
 		;;
@@ -719,7 +723,7 @@ then
 	exec chroot . /bin/sh <<- EOF
 		umount /tmp/.initrd || echo "*: Failed to unmount the initrd!"
 		/sbin/blockdev --flushbufs /dev/ram0 >/dev/null 2>&1
-		exec ${REAL_INIT:-/sbin/init} ${INIT_OPTS}
+		exec ${REAL_INIT:-${REAL_INITD}} ${INIT_OPTS}
 EOF
 elif [ "$0" = '/init' ]
 then
@@ -734,7 +738,7 @@ then
 		umount /proc || echo '*: Failed to unmount the initrd /proc!'
 	echo -n '.'
 
-	exec switch_root -c "/dev/console" "${CHROOT}" ${REAL_INIT:-/sbin/init} ${INIT_OPTS}
+	exec switch_root -c "/dev/console" "${CHROOT}" ${REAL_INIT:-${REAL_INITD}} ${INIT_OPTS}
 fi
 
 splash 'verbose'
