diff -Nurp genkernel-3.4.9_pre3.orig/gen_bootloader_grub.awk genkernel-3.4.9_pre3/gen_bootloader_grub.awk
--- genkernel-3.4.9_pre3.orig/gen_bootloader_grub.awk	2007-09-17 21:52:20.000000000 +0200
+++ genkernel-3.4.9_pre3/gen_bootloader_grub.awk	2007-10-10 02:22:36.000000000 +0200
@@ -67,7 +67,7 @@ if(state == "1")
 	if(have_k == "1" && ((match($0, /^title=/) || match($0, /^title */)) || NR == LIMIT))
 	{
 		state = "2";
-		print extraA "title=Gentoo Linux (" KV ")\n";
+		print extraA "title=Sabayon Linux (" KV ")\n";
 		print extraB my_kernel;
 		if(extraC)
 			print "\n" extraC;
diff -Nurp genkernel-3.4.9_pre3.orig/gen_bootloader.sh genkernel-3.4.9_pre3/gen_bootloader.sh
--- genkernel-3.4.9_pre3.orig/gen_bootloader.sh	2007-09-17 21:52:18.000000000 +0200
+++ genkernel-3.4.9_pre3/gen_bootloader.sh	2007-10-10 02:23:36.000000000 +0200
@@ -54,7 +54,7 @@ EOF
 
 			# Add grub configuration to grub.conf	
 			echo "# Genkernel generated entry, see GRUB documentation for details" >> ${GRUB_CONF}
-			echo "title=Gentoo Linux ($KV)" >> ${GRUB_CONF}
+			echo "title=Sabayon Linux ($KV)" >> ${GRUB_CONF}
 			echo -e "\troot (${GRUB_BOOT_DISK},${GRUB_BOOT_PARTITION})" >> ${GRUB_CONF}
 			if [ "${BUILD_INITRD}" -eq '0' ]
 			then
