diff -Nurp genkernel-3.4.5.orig/generic/linuxrc genkernel-3.4.5/generic/linuxrc
--- genkernel-3.4.5.orig/generic/linuxrc	2006-12-29 14:07:12.000000000 +0100
+++ genkernel-3.4.5/generic/linuxrc	2006-12-29 17:57:21.000000000 +0100
@@ -46,6 +46,7 @@ CMDLINE="`cat /proc/cmdline`"
 REAL_ROOT=''
 FAKE_ROOT=''
 REAL_ROOTFLAGS=''
+REAL_INITD="/sbin/init"
 for x in ${CMDLINE}
 do
 	case "${x}" in
@@ -61,6 +62,9 @@ do
 		real_init\=*)
 			REAL_INIT=`parse_opt "${x}"`
 		;;
+		accel_init\=*)
+			REAL_INITD=`parse_opt "${x}"`
+		;;
 		init_opts\=*)
 			INIT_OPTS=`parse_opt "${x}"`
 		;;
@@ -718,7 +722,7 @@ then
 	exec chroot . /bin/sh <<- EOF
 		umount /tmp/.initrd || echo "*: Failed to unmount the initrd!"
 		/sbin/blockdev --flushbufs /dev/ram0 >/dev/null 2>&1
-		exec ${REAL_INIT:-/sbin/init} ${INIT_OPTS}
+		exec ${REAL_INIT:-${REAL_INITD}} ${INIT_OPTS}
 EOF
 elif [ "$0" = '/init' ]
 then
@@ -773,7 +777,7 @@ then
 	# Start the real root init
 	(/bin/busybox sleep 3; /bin/busybox rm -r /bin/busybox /dev) & 
 	exec /bin/busybox chroot . /bin/sh <<- EOF
-	exec /sbin/init ${REAL_INIT}
+	exec ${REAL_INITD} ${REAL_INIT}
 EOF
 
 fi
