diff -Nurp genkernel-3.4.9.orig/generic/initrd.scripts genkernel-3.4.9/generic/initrd.scripts
--- genkernel-3.4.9.orig/generic/initrd.scripts	2008-02-27 14:04:01.000000000 +0000
+++ genkernel-3.4.9/generic/initrd.scripts	2008-02-27 14:12:57.000000000 +0000
@@ -99,7 +99,7 @@ findmediamount() {
 #				else
 #					mount -r -t auto ${x} ${mntdir} &>/dev/null
 #				fi
-				mount -r -t ${CDROOT_TYPE} ${x} ${mntdir} &>/dev/null
+				mount -r -t ${CDROOT_TYPE} ${x} ${mntdir} >/dev/null 2>&1
 				if [ "$?" = '0' ]
 				then
 					# Check for the media
