diff --git a/defaults/linuxrc b/defaults/linuxrc
index f434339..4a12c21 100755
--- a/defaults/linuxrc
+++ b/defaults/linuxrc
@@ -822,6 +822,11 @@ then
 	mkdir initramfs proc tmp sys run 2>/dev/null
 	chmod 1777 tmp
 
+	# have handy /mnt/cdrom (CDROOT_PATH) as well
+	_new_cdroot="${NEW_ROOT}${CDROOT_PATH}"
+	[ ! -d "${_new_cdroot}" ] && mkdir -p "${_new_cdroot}"
+	mount --bind "${CDROOT_PATH}" "${_new_cdroot}"
+
 	fi
 
 	#UML=$(cat /proc/cpuinfo|grep UML|sed -e 's|model name.*: ||')
