diff -Nurp genkernel-3.4.10.908.orig//arch/x86/modules_load genkernel-3.4.10.908/arch/x86/modules_load
--- genkernel-3.4.10.908.orig//arch/x86/modules_load	2010-12-24 23:59:54.000000000 +0100
+++ genkernel-3.4.10.908/arch/x86/modules_load	2011-01-03 15:58:19.482896059 +0100
@@ -28,3 +28,5 @@ MODULES_USB="ehci-hcd uhci usb-ohci hid
 # Filesystems
 MODULES_FS="ext2 ext3 ext4 btrfs reiserfs jfs nfs xfs fuse"
 
+# Crypto stuff, required for Sabayon crypt installs
+MODULES_CRYPT="aes-i586 sha1_generic sha256_generic sha512_generic xts"
diff -Nurp genkernel-3.4.10.908.orig//arch/x86_64/modules_load genkernel-3.4.10.908/arch/x86_64/modules_load
--- genkernel-3.4.10.908.orig//arch/x86_64/modules_load	2010-12-24 23:59:54.000000000 +0100
+++ genkernel-3.4.10.908/arch/x86_64/modules_load	2011-01-03 15:58:42.596356030 +0100
@@ -26,3 +26,6 @@ MODULES_USB="ehci-hcd uhci usb-ohci hid
 
 # Filesystems
 MODULES_FS="ext2 ext3 ext4 btrfs reiserfs jfs nfs xfs fuse"
+
+# Crypto stuff, required for Sabayon crypt installs
+MODULES_CRYPT="aes-x86_64 sha1_generic sha256_generic sha512_generic xts"
