--- genkernel-3.4.24.orig/arch/x86/modules_load
+++ genkernel-3.4.24/arch/x86/modules_load
@@ -29,3 +29,6 @@ MODULES_FS="ext2 ext3 ext4 btrfs reiserf
 
 # Crypto stuff, required for Sabayon crypt installs
 MODULES_CRYPT="aes-i586 sha1_generic sha256_generic sha512_generic xts"
+
+# Virtio support (KVM mainly)
+MODULES_VIRTIO="virtio virtio_balloon virtio_console virtio_net virtio_blk virtio_pci virtio_ring"
--- genkernel-3.4.24.orig/arch/x86_64/modules_load
+++ genkernel-3.4.24/arch/x86_64/modules_load
@@ -28,3 +28,6 @@ MODULES_FS="ext2 ext3 ext4 btrfs reiserf
 
 # Crypto stuff, required for Sabayon crypt installs
 MODULES_CRYPT="aes-x86_64 sha1_generic sha256_generic sha512_generic xts"
+
+# Virtio support (KVM mainly)
+MODULES_VIRTIO="virtio virtio_balloon virtio_console virtio_net virtio_blk virtio_pci virtio_ring"
--- genkernel-3.4.24.orig/defaults/initrd.defaults
+++ genkernel-3.4.24/defaults/initrd.defaults
@@ -80,5 +80,5 @@ DEFAULT_NFSOPTIONS="ro,nolock,rsize=1024
 
 # Only sections that are in by default or those that
 # are not module groups need to be defined here...
-HWOPTS='keymap cache modules pata sata scsi usb firewire waitscan slowusb lvm dmraid mdadm fs net'
-MY_HWOPTS='modules pata sata scsi slowusb usb firewire waitscan dmraid mdadm fs net iscsi'
+HWOPTS='keymap cache modules pata sata scsi usb firewire waitscan slowusb lvm dmraid mdadm fs net virtio'
+MY_HWOPTS='modules pata sata scsi slowusb usb firewire waitscan dmraid mdadm fs net iscsi virtio'
