diff -Nurp genkernel-3.4.8.orig/generic/linuxrc genkernel-3.4.8/generic/linuxrc
--- genkernel-3.4.8.orig/generic/linuxrc	2007-06-15 09:32:36.000000000 +0200
+++ genkernel-3.4.8/generic/linuxrc	2007-06-15 09:37:19.000000000 +0200
@@ -554,6 +554,9 @@ then
 		FS_LOCATION='mnt/livecd'
 	fi
 
+	# bind inside the newly created root tree
+	mount --bind ${NEW_ROOT}/mnt/cdrom ${NEW_ROOT}/mnt/cdrom
+
 #
 # End cdrom looptype determination and mounting if necessary
 #
