diff -Nurp genkernel-3.4.9_pre11.orig/generic/initrd.defaults genkernel-3.4.9_pre11/generic/initrd.defaults
--- genkernel-3.4.9_pre11.orig/generic/initrd.defaults	2007-12-04 00:18:23.000000000 +0100
+++ genkernel-3.4.9_pre11/generic/initrd.defaults	2007-12-11 11:58:13.000000000 +0100
@@ -78,5 +78,5 @@ CONSOLE="/dev/console"
 
 # Only sections that are in by default or those that
 # are not module groups need to be defined here...
-HWOPTS='usb firewire keymap cache pata sata evms lvm dmraid mdadm slowusb fs'
+HWOPTS='usb firewire keymap cache pata sata evms lvm dmraid mdadm fs'
 MY_HWOPTS='usb firewire pata sata dmraid mdadm fs net'
diff -Nurp genkernel-3.4.9_pre11.orig/generic/initrd.scripts genkernel-3.4.9_pre11/generic/initrd.scripts
--- genkernel-3.4.9_pre11.orig/generic/initrd.scripts	2007-12-11 11:36:54.000000000 +0100
+++ genkernel-3.4.9_pre11/generic/initrd.scripts	2007-12-11 12:03:10.000000000 +0100
@@ -361,7 +361,6 @@ start_dev_mgr() {
 	if [ "${USE_UDEV_NORMAL}" -eq '1' ]
 	then
 		cd /sys
-		[ "${DO_slowusb}" ] && sleep 10
 
 	        # Smarter control
         	[ "${DO_slowusb}" ] || \
@@ -371,17 +370,14 @@ start_dev_mgr() {
                         	FORCE_slowusb="1"
                 	fi
         	done
-        	[ "${FORCE_slowusb}" ] && sleep 10
 
 		kill_devfsd
 		good_msg 'Activating mdev'
 #		runUdev
 		runmdev
-		[ "${DO_slowusb}" ] && sleep 20
 
                 # Smarter control
-                [ "${DO_slowusb}" ] || \
-                [ "${FORCE_slowusb}" ] && sleep 20
+                [ "${FORCE_slowusb}" ] && sleep 10
 
 		cd /
 	else
