commit 226b9f9a9e713829c32ee5d11e70efe44f22f111
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Mon Jan 31 14:08:35 2011 +0100

    [sulfur] improve reliability of sulfur startup script

diff --git a/sulfur/misc/sulfur b/sulfur/misc/sulfur
index cb0a018..90dc476 100755
--- a/sulfur/misc/sulfur
+++ b/sulfur/misc/sulfur
@@ -4,8 +4,11 @@ if "--debug" in sys.argv:
     os.environ['SULFUR_DEBUG'] = "1"
     sys.argv.remove("--debug")
 
-sulfur_path = " /usr/lib/entropy/sulfur/sulfur_client.py " + ' '.join(sys.argv[1:])
-su_cmd = "xterm -e \"su -c '"+sulfur_path+"'\""
+args = sys.argv[1:]
+sulfur_path = "/usr/lib/entropy/sulfur/sulfur_client.py"
+if args:
+    sulfur_path += " " + " ".join(args)
+su_cmd = "xterm -e \"su -c '" + sulfur_path + "'\""
 
 def setup_env():
     global sulfur_path, su_cmd
@@ -31,11 +34,11 @@ def setup_env():
         for item in path:
             itempath = os.path.join(item, 'kdesu')
             if os.access(itempath,os.X_OK):
-                su_cmd = itempath+' "'+sulfur_path+'"'
+                su_cmd = itempath+' "' + sulfur_path + '"'
                 break
     else:
-        if os.access('/usr/bin/gksu',os.X_OK):
-            su_cmd = '/usr/bin/gksu "'+sulfur_path+'"'
+        if os.access('/usr/bin/gksu', os.X_OK):
+            su_cmd = '/usr/bin/gksu "' + sulfur_path + '"'
 
 
 
