diff -Nurp entropy-0.98.4.2.orig/sulfur/src/sulfur/package.py entropy-0.98.4.2/sulfur/src/sulfur/package.py
--- entropy-0.98.4.2.orig/sulfur/src/sulfur/package.py	2009-08-18 23:02:46.000000000 +0200
+++ entropy-0.98.4.2/sulfur/src/sulfur/package.py	2009-08-19 14:44:11.038163254 +0200
@@ -242,7 +242,11 @@ class EntropyPackage:
                 self.matched_id)
             if repoid is None:
                 repoid = _("Not available")
-        else: return self.matched_repo
+            return repoid
+        elif self.matched_repo:
+            return self.matched_repo
+
+        return _("Not available")
 
     def get_idpackage(self):
         return self.matched_id
