Index: FvwmCompositor/main/compositor.cxx
===================================================================
RCS file: /users/asspro/ihm/metisse-cvs/metisse/FvwmCompositor/main/compositor.cxx,v
retrieving revision 1.5
retrieving revision 1.6
diff -u -r1.5 -r1.6
--- metisse/FvwmCompositor/main/compositor.cxx	23 Jan 2007 22:50:47 -0000	1.5
+++ metisse/FvwmCompositor/main/compositor.cxx	7 Feb 2007 14:44:01 -0000	1.6
@@ -206,7 +206,7 @@
   FvwmModule fvwmModule(
 	  AScr, metisseDesktop, fvwm_fds[0], fvwm_fds[1], myName, restartString);
 
-  glClearColor(0.1,0.5,0.4,1) ;
+  glClearColor(0.7,0.71,0.7,1) ;
   glEnable(GL_BLEND);
   glBlendFunc (GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
 
