--- metisse-20061227/FvwmCompositor/config/fonts.defaults	2006-12-13 00:07:28.000000000 +0100
+++ metisse-20061227/FvwmCompositor/config/fonts	2007-01-02 18:38:46.000000000 +0100
@@ -12,52 +12,52 @@
 
 DestroyFunc FuncFontsSet-WinFont
 AddToFunc   FuncFontsSet-WinFont
-+ I Style * InactiveFont "Shadow=$0:xft:sans:size=12:$1:$2;-*-helvetica-$3-$4-*-120-*"
++ I Style * InactiveFont "Shadow=$0:xft:sans:size=10:$1:$2;-*-helvetica-$3-$4-*-120-*"
 
 DestroyFunc FuncFontsSet-ActiveWinFont
 AddToFunc   FuncFontsSet-ActiveWinFont
-+ I Style * Font "Shadow=$0:xft:sans:size=12:$1:$2;-*-helvetica-$3-$4-*-120-*"
++ I Style * Font "Shadow=$0:xft:sans:size=10:$1:$2;-*-helvetica-$3-$4-*-120-*"
 
 # Menu fonts add minspace=true?
 
 DestroyFunc FuncFontsSet-MenuFont
 AddToFunc   FuncFontsSet-MenuFont
-+ I MenuStyle $0 Font "Shadow=$1:xft:sans:size=13:$2:$3;-*-helvetica-$4-$5-*-120-*"
++ I MenuStyle $0 Font "Shadow=$1:xft:sans:size=10:$2:$3;-*-helvetica-$4-$5-*-120-*"
 
 DestroyFunc FuncFontsSet-WinMenuFont
 AddToFunc   FuncFontsSet-WinMenuFont
-+ I MenuStyle $0 Font "Shadow=$1:xft:sans:size=12:minspace=true:$2:$3;-*-helvetica-$4-$5-*-120-*"
++ I MenuStyle $0 Font "Shadow=$1:xft:sans:size=10:minspace=true:$2:$3;-*-helvetica-$4-$5-*-120-*"
 
 DestroyFunc FuncFontsSet-Utf8MenuFont
 AddToFunc   FuncFontsSet-Utf8MenuFont
-+ I MenuStyle $0 Font "StringEncoding=UTF-8:Shadow=$1:xft:sans:size=13:minspace=true:$2:$3;-*-helvetica-$4-$5-*-120-*"
++ I MenuStyle $0 Font "StringEncoding=UTF-8:Shadow=$1:xft:sans:size=10:minspace=true:$2:$3;-*-helvetica-$4-$5-*-120-*"
 
 
 #---------------------------------------
 # Modules fonts ... We should have Fontsets as we have Colorsets ...
 
-*FvwmIdent: Font "xft:sans:size=12;-*-helvetica-medium-r-*-120-*"
+*FvwmIdent: Font "xft:sans:size=10;-*-helvetica-medium-r-*-120-*"
 
 # See also ~/.fvwm-metisse/.FvwmForm which can be create with
 # FvwmForm-Form
-*FvwmFormDefault: Font "xft:sans:size=12;-*-helvetica-medium-r-*-120-*"
-*FvwmFormDefault: InputFont "xft:Monospace:size=11;-*-fixed-medium-r-*-120-*"
-*FvwmFormDefault: ButtonFont "xft:sans:size=12;-*-helvetica-medium-r-*-120-*"
-*FvwmFormDefault: TimeoutFont "xft:sans:size=12;-*-helvetica-medium-r-*-120-*"
+*FvwmFormDefault: Font "xft:sans:size=10;-*-helvetica-medium-r-*-120-*"
+*FvwmFormDefault: InputFont "xft:Monospace:size=10;-*-fixed-medium-r-*-120-*"
+*FvwmFormDefault: ButtonFont "xft:sans:size=10;-*-helvetica-medium-r-*-120-*"
+*FvwmFormDefault: TimeoutFont "xft:sans:size=10;-*-helvetica-medium-r-*-120-*"
 
 # default font for IconBox, IconMan, TaskBar and WinList
 DestroyFunc AFuncSetFont-ModuleWindowList
 AddToFunc   AFuncSetFont-ModuleWindowList
-+ I *$0: $1 "shadow=1:xft:sans:size=12;-*-helvetica-medium-r-*-120-*"
++ I *$0: $1 "shadow=1:xft:sans:size=10;-*-helvetica-medium-r-*-120-*"
 
 # selected font (TaskBar)
 DestroyFunc AFuncSetFont-ModuleSelectedWindowList
 AddToFunc   AFuncSetFont-ModuleSelectedWindowList
-+ I *$0: $1 "shadow=1:xft:sans:bold:size=12;-*-helvetica-bold-r-*-120-*"
++ I *$0: $1 "shadow=1:xft:sans:bold:size=10;-*-helvetica-bold-r-*-120-*"
 
 DestroyFunc AFuncSetFont-ModuleTips
 AddToFunc   AFuncSetFont-ModuleTips
-+ I *$0: $1 "shadow=0:xft:sans:size=12;-*-helvetica-medium-r-*-120-*"
++ I *$0: $1 "shadow=0:xft:sans:size=10;-*-helvetica-medium-r-*-120-*"
 
 # small font (Pager mini window font)
 DestroyFunc AFuncSetFont-ModuleSmall
--- metisse-20061227/FvwmCompositor/config/background.defaults	2006-12-04 22:13:58.000000000 +0100
+++ metisse-20061227/FvwmCompositor/config/background	2007-01-02 18:38:46.000000000 +0100
@@ -19,6 +19,6 @@
 
 DestroyFunc AFuncSetBackground
 AddToFunc   AFuncSetBackground
-+ I AFuncCheckAndSetScaledBg $[METISSE_FVWM_CONFIG_DIR]/images/backgrounds/sky-metisse.jpg
++ I AFuncCheckAndSetScaledBg /usr/share/backgrounds/sabayonlinux.jpg
 # opale background
 #+ I AFuncCheckAndSetScaledBg $[OPALE_CONFIG_DIR]/backgrounds/sky-default.jpg
--- metisse-20061227/FvwmCompositor/config/metisserc.in.defaults	2007-01-02 18:40:57.000000000 +0100
+++ metisse-20061227/FvwmCompositor/config/metisserc.in	2007-01-02 19:00:28.000000000 +0100
@@ -343,11 +343,12 @@
 # Default Config
 DestroyFunc FuncLoadDefaultConfig
 AddToFunc   FuncLoadDefaultConfig
-+ I FuncTheme Metisse
++ I FuncTheme Ia_Ora_blue
 + I AFuncConfigMoveTransparency 0.7
 + I AFuncConfigResizeTransparency 0.7
-+ I FuncPanConfigPanel Panel $[vp.width] 22 +0-0 12 22 16x16/ 176 12
++ I FuncPanConfigPanel PagerOnly 56
 + I FuncPanConfigWindowListResolution Page
++ I EWMHNumberOfDesktops 1
 + I Echo Default Config Loaded
 
 SetEnv METISSERC_DEF_CONFIG_VER 0
