--- metisse-0.4.0-rc4/xserver/GL/mesa/Makefile.am.fixglx	2004-07-14 06:57:56.000000000 +0200
+++ metisse-0.4.0-rc4/xserver/GL/mesa/Makefile.am	2007-02-07 14:25:20.000000000 +0100
@@ -144,7 +144,6 @@
 	main/varray.c  \
 	main/varray.h  \
 	main/version.h \
-	main/vsnprintf.c  \
 	main/vtxfmt.c  \
 	main/vtxfmt.h  \
 	main/vtxfmt_tmp.h  \
--- metisse-0.4.0-rc4/configure.ac.fixglx	2007-01-23 11:44:31.000000000 +0100
+++ metisse-0.4.0-rc4/configure.ac	2007-02-07 14:30:24.000000000 +0100
@@ -427,7 +427,7 @@
 
 with_glx=no
 problem_glx="disabled by default try --enable-glx or --enable-glx-x86"
-if test "$GLXX" = yes; then
+if test "$GLX" = yes; then
   GLX=yes
   with_glx=yes
   problem_glx="(without x86 optimisation)"
