--- configure.ac.old	2007-04-25 22:26:19.000000000 +0200
+++ configure.ac	2007-04-25 22:27:41.000000000 +0200
***************
*** 138,143 ****
--- 138,146 ----
    FLUID="$with_fltk/bin/fluid"
    CPPFLAGS="-I$with_fltk/include $CPPFLAGS"
    EXTRA_DIRS="${EXTRA_DIRS} $with_fltk"
+ else
+   CPPFLAGS="$(fltk-config --cxxflags) $CPPFLAGS"
+   LIBS="$(fltk-config --use-gl --ldflags) $LIBS"
  fi
  AC_SUBST(FLUID)
