--- src/wizard.h	2010/04/11 16:36:38	536
+++ src/wizard.h	2010/08/14 07:27:35	552
@@ -10,6 +10,7 @@
 #include <OpenThreads/Thread>
 class LogWindow;
 class Advanced;
+using std::string;
 #include <FL/Fl_Double_Window.H>
 #include <FL/Fl_Wizard.H>
 #include <FL/Fl_Group.H>

--- src/run_posix.cxx	2010/01/23 21:49:08	522
+++ src/run_posix.cxx	2010/08/01 18:05:08	542
@@ -49,6 +49,7 @@
 #include "logwin.h"

 using std::string;
+using std::vector;

 int
 Wizard::run_fgfs( const std::string &args )
