--- src/amule.cpp.old	2006-12-09 10:49:53.000000000 +0100
+++ src/amule.cpp	2006-12-09 10:50:49.000000000 +0100
@@ -436,7 +436,6 @@ bool CamuleApp::OnInit()
 	cmdline.AddSwitch(wxT("d"), wxT("disable-fatal"), wxT("Does not handle fatal exception."));
 	cmdline.AddSwitch(wxT("o"), wxT("log-stdout"), wxT("Print log messages to stdout."));
 	cmdline.AddSwitch(wxT("r"), wxT("reset-config"), wxT("Resets config to default values."));
-	cmdline.AddSwitch(wxT("eveniflfroencomplainsthiswillstay"), wxT("even-if-lfroen-complains-this-will-stay"), wxT("Runs aMule scary SVN development version at your own risk. Even if lfroen complains."));
 	
 	// Show help on --help or invalid commands
 	if ( cmdline.Parse() ) {
@@ -446,11 +445,6 @@ bool CamuleApp::OnInit()
 		return false;
 	}	
 
-	if ( !cmdline.Found(wxT("even-if-lfroen-complains-this-will-stay")) ) {
-		printf("This binary requires you to use the flag --even-if-lfroen-complains-this-will-stay and only if you're very sure of it. Warning: If you're lfroen, you might find this annoying. That pleases me.\n");
-		return false;
-	}
-	
 	
 #ifdef AMULE_DAEMON
 	if ( cmdline.Found(wxT("config-dir"), &ConfigDir) ) {
