diff -Nurp amule-cvs.orig/src/amule.cpp amule-cvs/src/amule.cpp
--- amule-cvs.orig/src/amule.cpp	2007-06-19 21:11:43.000000000 +0200
+++ amule-cvs/src/amule.cpp	2007-06-24 11:27:05.000000000 +0200
@@ -454,7 +454,6 @@ bool CamuleApp::OnInit()
 	cmdline.AddSwitch(wxT("d"), wxT("disable-fatal"), wxT("Does not handle fatal exception."));
 	cmdline.AddSwitch(wxT("o"), wxT("log-stdout"), wxT("Print log messages to stdout."));
 	cmdline.AddSwitch(wxT("r"), wxT("reset-config"), wxT("Resets config to default values."));
-	cmdline.AddSwitch(wxT("eveniflfroensurreptitiouslyremovesitthiswillstay"), wxT("even-if-lfroen-surreptitiously-removes-it-this-will-stay"), wxT("Runs aMule scary SVN development version at your own risk. Even if lfroen surreptitiously tried to remove it."));
 	
 	// Show help on --help or invalid commands
 	if ( cmdline.Parse() ) {
@@ -464,11 +463,6 @@ bool CamuleApp::OnInit()
 		return false;
 	}	
 
-	if ( !cmdline.Found(wxT("even-if-lfroen-surreptitiously-removes-it-this-will-stay")) ) {
-		printf("This binary requires you to use the flag --even-if-lfroen-surreptitiously-removes-it-this-will-stay and only if you're very sure of it. Warning: If you're lfroen or wuischke, you might find this annoying. That pleases me.\n");
-		return false;
-	}	
-	
 #ifdef AMULE_DAEMON
 	if ( cmdline.Found(wxT("config-dir"), &ConfigDir) ) {
 		if (ConfigDir.Last() != wxFileName::GetPathSeparator()) {
