Index: kopete/plugins/nowlistening/configure.in.in
===================================================================
--- kopete/plugins/nowlistening/configure.in.in	(revision 433553)
+++ kopete/plugins/nowlistening/configure.in.in	(working copy)
@@ -41,7 +41,19 @@
   fi
 ])
 
-AC_CHECK_XMMS
+AC_ARG_WITH(xmms,
+  [AS_HELP_STRING(--with-xmms,
+    [enable support for XMMS @<:@default=check@:>@])],
+  [], with_xmms=check)
+
+if test "x$with_xmms" != xno; then
+  AC_CHECK_XMMS
+
+  if test "x$with_xmms" != xcheck && test "x$ac_cv_have_xmms" = xno; then
+    AC_MSG_FAILURE([--with-xmms was given, but test for XMMS failed])
+  fi
+fi
+
 AC_SUBST(XMMS_LIBS)
 AC_SUBST(XMMS_LDFLAGS)
 AC_SUBST(XMMS_INCLUDES)
Index: configure.in.bot
===================================================================
--- configure.in.bot	(revision 433553)
+++ configure.in.bot	(working copy)
@@ -63,7 +63,7 @@
   all_tests=bad
 fi
 
-if test -z "$XMMS_LIBS"; then
+if test "x$with_xmms" = xcheck && test -z "$XMMS_LIBS"; then
   echo ""
   echo "You're missing the XMMS libraries, or the libxmms development package."
   echo "Without libxmms Kopete's NowListening plugin won't be able to talk to"
