--- kopete/libkopete/avdevice/videodevice.h.orig	2006-10-17 14:34:13.000000000 +0200
+++ kopete/libkopete/avdevice/videodevice.h	2006-10-17 14:37:00.000000000 +0200
@@ -35,8 +35,8 @@
 
 #if defined(__linux__) && defined(ENABLE_AV)
 
-#include <asm/types.h>
 #undef __STRICT_ANSI__
+#include <linux/types.h>
 #ifndef __u64 //required by videodev.h
 #define __u64 unsigned long long
 #endif // __u64
--- kopete/libkopete/avdevice/videoinput.h.orig	2006-10-17 14:38:43.000000000 +0200
+++ kopete/libkopete/avdevice/videoinput.h	2006-10-17 14:39:50.000000000 +0200
@@ -21,8 +21,8 @@
 #define KOPETE_AVVIDEOINPUT_H
 
 #ifdef __linux__
-#include <asm/types.h>
 #undef __STRICT_ANSI__
+#include <linux/types.h>
 #endif // __linux__
 #ifndef __u64 //required by videodev.h
 #define __u64 unsigned long long
