--- ./kopete/protocols/jabber/jingle/libjingle/talk/base/asynctcpsocket.cc.orig	2008-07-04 20:03:14.000000000 -0700
+++ ./kopete/protocols/jabber/jingle/libjingle/talk/base/asynctcpsocket.cc	2008-07-04 20:04:41.000000000 -0700
@@ -32,6 +32,7 @@
 #include "talk/base/byteorder.h"
 #include "talk/base/common.h"
 #include "talk/base/logging.h"
+#include <cstring>
 
 #if defined(_MSC_VER) && _MSC_VER < 1300
 namespace std {
--- ./kopete/protocols/jabber/jingle/libjingle/talk/base/socketadapters.cc.orig	2008-07-04 20:05:09.000000000 -0700
+++ ./kopete/protocols/jabber/jingle/libjingle/talk/base/socketadapters.cc	2008-07-04 20:05:17.000000000 -0700
@@ -42,6 +42,7 @@
 #endif
 
 #include <cassert>
+#include <cstring>
 
 #include "talk/base/base64.h"
 #include "talk/base/basicdefs.h"
--- ./kopete/protocols/jabber/jingle/libjingle/talk/base/messagequeue.h.orig	2008-07-04 19:11:03.000000000 -0700
+++ ./kopete/protocols/jabber/jingle/libjingle/talk/base/messagequeue.h	2008-07-04 19:11:14.000000000 -0700
@@ -32,6 +32,7 @@
 #include "talk/base/criticalsection.h"
 #include "talk/base/socketserver.h"
 #include "talk/base/jtime.h"
+#include <cstring>
 #include <vector>
 #include <queue>
 #include <algorithm>
--- ./kopete/protocols/jabber/jingle/libjingle/talk/base/physicalsocketserver.cc.orig	2008-07-04 19:11:50.000000000 -0700
+++ ./kopete/protocols/jabber/jingle/libjingle/talk/base/physicalsocketserver.cc	2008-07-04 20:01:59.000000000 -0700
@@ -30,11 +30,13 @@
 #endif
 
 #include <cassert>
+#include <algorithm>
 
 #ifdef POSIX
 extern "C" {
 #include <errno.h>
 #include <fcntl.h>
+#include <string.h>
 #include <sys/time.h>
 #include <unistd.h>
 }
--- ./kopete/protocols/jabber/jingle/libjingle/talk/base/host.cc.orig	2008-07-04 20:06:02.000000000 -0700
+++ ./kopete/protocols/jabber/jingle/libjingle/talk/base/host.cc	2008-07-04 20:05:58.000000000 -0700
@@ -29,7 +29,8 @@
 #include "talk/base/logging.h"
 #include "talk/base/network.h"
 #include "talk/base/socket.h"
-#include <string>
+#include <cstring>
+#include <cstdlib>
 #include <iostream>
 #include <cassert>
 #include <errno.h>
--- ./kopete/protocols/jabber/jingle/libjingle/talk/base/bytebuffer.cc.orig	2008-07-04 20:02:39.000000000 -0700
+++ ./kopete/protocols/jabber/jingle/libjingle/talk/base/bytebuffer.cc	2008-07-04 20:02:52.000000000 -0700
@@ -30,6 +30,7 @@
 #include "talk/base/byteorder.h"
 #include <algorithm>
 #include <cassert>
+#include <cstring>
 
 #if defined(_MSC_VER) && _MSC_VER < 1300
 namespace std {
--- ./kopete/protocols/jabber/jingle/libjingle/talk/xmpp/xmpppassword.h.orig	2008-07-04 20:08:44.000000000 -0700
+++ ./kopete/protocols/jabber/jingle/libjingle/talk/xmpp/xmpppassword.h	2008-07-04 20:08:59.000000000 -0700
@@ -31,6 +31,8 @@
 #include "talk/base/linked_ptr.h"
 #include "talk/base/scoped_ptr.h"
 
+#include <cstring>
+
 namespace buzz {
 
 class XmppPasswordImpl {
--- ./kopete/protocols/jabber/jingle/libjingle/talk/p2p/base/stun.cc.orig	2008-07-04 20:06:41.000000000 -0700
+++ ./kopete/protocols/jabber/jingle/libjingle/talk/p2p/base/stun.cc	2008-07-04 20:06:48.000000000 -0700
@@ -29,6 +29,7 @@
 #include "talk/p2p/base/stun.h"
 #include <iostream>
 #include <cassert>
+#include <cstring>
 
 #if defined(_MSC_VER) && _MSC_VER < 1300
 namespace std {
--- ./kopete/protocols/jabber/jingle/libjingle/talk/p2p/base/stunserver.cc.orig	2008-07-04 20:09:48.000000000 -0700
+++ ./kopete/protocols/jabber/jingle/libjingle/talk/p2p/base/stunserver.cc	2008-07-04 20:10:18.000000000 -0700
@@ -28,6 +28,7 @@
 #include "talk/base/bytebuffer.h"
 #include "talk/p2p/base/stunserver.h"
 #include <iostream>
+#include <cstring>
 
 #ifdef POSIX
 extern "C" {
