Index: kopete/protocols/oscar/liboscar/oscartypes.h
===================================================================
--- kopete/protocols/oscar/liboscar/oscartypes.h	(revision 561015)
+++ kopete/protocols/oscar/liboscar/oscartypes.h	(revision 561016)
@@ -232,13 +232,13 @@
 
 	/* ICQ Version Characteristics */
 	const unsigned char ICQ_TCP_VERSION 	= 0x0008;
-	const char ICQ_CLIENTSTRING[] 		= "ICQ Inc. - Product of ICQ (TM).2003a.5.45.1.3777.85";
+	const char ICQ_CLIENTSTRING[] 		= "ICQBasic";
 	const WORD ICQ_CLIENTID 		= 0x010A;
-	const WORD ICQ_MAJOR 			= 0x0005;
-	const WORD ICQ_MINOR 			= 0x002D;
-	const WORD ICQ_POINT 			= 0x0001;
-	const WORD ICQ_BUILD 			= 0x0EC1;
-	const char ICQ_OTHER[] 			= { 0x00, 0x00, 0x00, 0x55 };
+	const WORD ICQ_MAJOR 			= 0x0014;
+	const WORD ICQ_MINOR 			= 0x0034;
+	const WORD ICQ_POINT 			= 0x0000;
+	const WORD ICQ_BUILD 			= 0x0A0F;
+	const char ICQ_OTHER[] 			= { 0x00, 0x00, 0x04, 0x3d };
 	const char ICQ_COUNTRY[] 		= "us";
 	const char ICQ_LANG[] 			= "en";
 
