Index: kopete-0.12-alpha1/kopete/libkopete/configure.in.in
===================================================================
--- kopete-0.12-alpha1.orig/kopete/libkopete/configure.in.in
+++ kopete-0.12-alpha1/kopete/libkopete/configure.in.in
@@ -4,7 +4,11 @@ LDFLAGS="$X_LDFLAGS"
 
 LIB_XSS=
 
-KDE_CHECK_HEADER(X11/extensions/scrnsaver.h,
+AC_ARG_WITH([xscreensaver],
+  AC_HELP_STRING([--without-xscreensaver], [Disable XScreenSaver support (default: check)]) )
+
+if test "x$with_xscreensaver" != "xno"; then
+  KDE_CHECK_HEADER(X11/extensions/scrnsaver.h,
     [
     AC_CHECK_LIB(Xext,XScreenSaverQueryInfo,
 	[
@@ -27,6 +31,7 @@ KDE_CHECK_HEADER(X11/extensions/scrnsave
      [
      #include <X11/Xlib.h>
      ] )
+fi
 
 AC_SUBST(LIB_XSS)
 LDFLAGS="$xss_save_ldflags"
