http://bugs.gentoo.org/407137

--- configure.ac
+++ configure.ac
@@ -73,7 +73,6 @@
 
 AC_PROG_CC
 AC_PROG_CXX
-AC_CHECK_PROG([HAVE_CXX],[$CXX],[yes],[no])
 AC_C_INLINE
 if test "x$GCC" = "xyes" ; then
 
@@ -209,11 +208,7 @@
 
 AC_MSG_CHECKING([µTP])
 build_utp="no"
-if test "x$HAVE_CXX" = "xyes" ; then
-    have_utp="yes"
-else
-    have_utp="no"
-fi
+have_utp="yes"
 AC_ARG_ENABLE([utp],
               AS_HELP_STRING([--enable-utp],[build µTP support]),
               [want_utp=${enableval}],
