--- src/pcm/pcm_direct.c	2006-06-21 17:31:59.000000000 +0200
+++ src/pcm/pcm_direct.c	2006-07-29 21:04:11.542194680 +0200
@@ -1392,7 +1394,8 @@
 		}
 		bindings[cchannel] = schannel;
 	}
-	if (dmix->type == SND_PCM_TYPE_DSNOOP)
+	if (dmix->type == SND_PCM_TYPE_DSNOOP ||
+	    ! dmix->bindings)
 		goto __skip_same_dst;
 	for (chn = 0; chn < count; chn++) {
 		for (chn1 = 0; chn1 < count; chn1++) {
