diff -r 402a902f18a8 -r 8c310c3277b7 src/pcm/pcm_rate.c
--- a/src/pcm/pcm_rate.c	Tue May 02 17:48:47 2006 +0200
+++ b/src/pcm/pcm_rate.c	Wed May 03 19:18:28 2006 +0200
@@ -603,7 +603,7 @@ static inline void snd_pcm_rate_sync_hwp
 	 */
 	rate->hw_ptr =
 		(slave_hw_ptr / rate->gen.slave->period_size) * pcm->period_size +
-		rate->ops.output_frames(rate->obj, slave_hw_ptr % rate->gen.slave->period_size);
+		rate->ops.input_frames(rate->obj, slave_hw_ptr % rate->gen.slave->period_size);
 }
 
 static int snd_pcm_rate_hwsync(snd_pcm_t *pcm)
