--- gnome-themes-standard-3.0.2.orig/themes/Adwaita/index.theme
+++ gnome-themes-standard-3.0.2/themes/Adwaita/index.theme
@@ -103,6 +103,6 @@ Comment[zh_TW]=只有唯一一個
 Encoding=UTF-8
 GtkTheme=Adwaita
 MetacityTheme=Adwaita
-IconTheme=gnome
+IconTheme=elementary
 CursorTheme=Adwaita
 CursorSize=24
--- gnome-themes-standard-3.0.2.orig/themes/Adwaita/index.theme.in
+++ gnome-themes-standard-3.0.2/themes/Adwaita/index.theme.in
@@ -5,6 +5,6 @@ _Comment=There is only one
 Encoding=UTF-8
 GtkTheme=Adwaita
 MetacityTheme=Adwaita
-IconTheme=gnome
+IconTheme=elementary
 CursorTheme=Adwaita
 CursorSize=24
