diff -Nurp madwifi-0.9.4-r4133-20100621.orig/ath/if_ath_ahb.c madwifi-0.9.4-r4133-20100621/ath/if_ath_ahb.c
--- madwifi-0.9.4-r4133-20100621.orig/ath/if_ath_ahb.c	2009-08-13 16:45:07.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/ath/if_ath_ahb.c	2011-03-17 14:23:14.000000000 +0000
@@ -11,7 +11,12 @@
 #endif
 
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/ath/if_ath.c madwifi-0.9.4-r4133-20100621/ath/if_ath.c
--- madwifi-0.9.4-r4133-20100621.orig/ath/if_ath.c	2010-06-20 12:13:10.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/ath/if_ath.c	2011-03-17 14:14:11.000000000 +0000
@@ -45,7 +45,12 @@
 #include "opt_ah.h"
 
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/ath/if_ath_pci.c madwifi-0.9.4-r4133-20100621/ath/if_ath_pci.c
--- madwifi-0.9.4-r4133-20100621.orig/ath/if_ath_pci.c	2009-01-14 03:16:12.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/ath/if_ath_pci.c	2011-03-17 14:15:40.000000000 +0000
@@ -43,7 +43,12 @@
 #endif
 
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/ath_hal/ah_os.c madwifi-0.9.4-r4133-20100621/ath_hal/ah_os.c
--- madwifi-0.9.4-r4133-20100621.orig/ath_hal/ah_os.c	2010-01-09 08:47:02.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/ath_hal/ah_os.c	2011-03-17 14:15:23.000000000 +0000
@@ -45,7 +45,12 @@
 #define USE_REAL_TIME_DELAY
 
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/ath_rate/amrr/amrr.c madwifi-0.9.4-r4133-20100621/ath_rate/amrr/amrr.c
--- madwifi-0.9.4-r4133-20100621.orig/ath_rate/amrr/amrr.c	2010-01-09 08:47:02.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/ath_rate/amrr/amrr.c	2011-03-17 14:12:10.000000000 +0000
@@ -44,7 +44,12 @@
  *    Mathieu Lacage, Hossein Manshaei, Thierry Turletti
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/ath_rate/minstrel/minstrel.c madwifi-0.9.4-r4133-20100621/ath_rate/minstrel/minstrel.c
--- madwifi-0.9.4-r4133-20100621.orig/ath_rate/minstrel/minstrel.c	2009-09-19 04:09:09.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/ath_rate/minstrel/minstrel.c	2011-03-17 14:16:17.000000000 +0000
@@ -88,7 +88,12 @@
  * a few different packet sizes independently for each link. */
 
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 
 #include <linux/version.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/ath_rate/onoe/onoe.c madwifi-0.9.4-r4133-20100621/ath_rate/onoe/onoe.c
--- madwifi-0.9.4-r4133-20100621.orig/ath_rate/onoe/onoe.c	2010-01-09 08:47:02.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/ath_rate/onoe/onoe.c	2011-03-17 14:17:36.000000000 +0000
@@ -40,7 +40,12 @@
  * Atsushi Onoe's rate control algorithm.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/ath_rate/sample/sample.c madwifi-0.9.4-r4133-20100621/ath_rate/sample/sample.c
--- madwifi-0.9.4-r4133-20100621.orig/ath_rate/sample/sample.c	2009-09-19 04:09:09.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/ath_rate/sample/sample.c	2011-03-17 14:18:32.000000000 +0000
@@ -42,7 +42,12 @@
  */
 
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_acl.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_acl.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_acl.c	2006-09-20 08:45:13.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_acl.c	2011-03-17 14:21:03.000000000 +0000
@@ -46,7 +46,12 @@
  * or rejected (ACL_POLICY_DENT).
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_beacon.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_beacon.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_beacon.c	2007-12-13 06:21:32.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_beacon.c	2011-03-17 14:17:29.000000000 +0000
@@ -39,7 +39,12 @@
  * IEEE 802.11 beacon handling routines
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211.c	2009-06-25 22:23:39.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211.c	2011-03-17 14:15:56.000000000 +0000
@@ -39,7 +39,12 @@
  * IEEE 802.11 generic handler
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_crypto.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_crypto.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_crypto.c	2006-09-24 21:06:25.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_crypto.c	2011-03-17 14:18:26.000000000 +0000
@@ -39,7 +39,12 @@
  * IEEE 802.11 generic crypto support.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_crypto_ccmp.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_crypto_ccmp.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_crypto_ccmp.c	2010-03-23 22:50:36.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_crypto_ccmp.c	2011-03-17 14:20:56.000000000 +0000
@@ -39,7 +39,12 @@
  * it's license is included below.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_crypto_none.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_crypto_none.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_crypto_none.c	2006-09-20 08:45:13.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_crypto_none.c	2011-03-17 14:17:43.000000000 +0000
@@ -35,7 +35,12 @@
  * IEEE 802.11 NULL crypto support.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_crypto_tkip.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_crypto_tkip.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_crypto_tkip.c	2010-03-23 22:50:36.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_crypto_tkip.c	2011-03-17 14:22:00.000000000 +0000
@@ -39,7 +39,12 @@
  * it's license is included below.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_crypto_wep.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_crypto_wep.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_crypto_wep.c	2010-03-23 22:50:36.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_crypto_wep.c	2011-03-17 14:21:46.000000000 +0000
@@ -35,7 +35,12 @@
  * IEEE 802.11 WEP crypto support.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_input.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_input.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_input.c	2007-07-25 15:26:38.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_input.c	2011-03-17 14:18:20.000000000 +0000
@@ -39,7 +39,12 @@
  * IEEE 802.11 input handling.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_linux.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_linux.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_linux.c	2010-01-09 08:47:02.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_linux.c	2011-03-17 14:20:31.000000000 +0000
@@ -34,7 +34,12 @@
  * IEEE 802.11 support (Linux-specific code)
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_monitor.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_monitor.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_monitor.c	2008-01-29 02:21:47.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_monitor.c	2011-03-17 14:20:26.000000000 +0000
@@ -35,7 +35,12 @@
  * IEEE 802.11 monitor mode 
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_node.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_node.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_node.c	2007-05-23 08:43:05.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_node.c	2011-03-17 14:19:01.000000000 +0000
@@ -39,7 +39,12 @@
  * IEEE 802.11 node handling support.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_output.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_output.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_output.c	2009-01-14 03:16:12.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_output.c	2011-03-17 14:19:14.000000000 +0000
@@ -39,7 +39,12 @@
  * IEEE 802.11 output handling.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_power.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_power.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_power.c	2008-06-04 08:09:17.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_power.c	2011-03-17 14:19:08.000000000 +0000
@@ -39,7 +39,12 @@
  * IEEE 802.11 power save support.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_proto.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_proto.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_proto.c	2009-01-14 03:16:12.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_proto.c	2011-03-17 14:19:41.000000000 +0000
@@ -39,7 +39,12 @@
  * IEEE 802.11 protocol support.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/kmod.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_rate.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_rate.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_rate.c	2007-12-12 04:11:07.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_rate.c	2011-03-17 14:20:20.000000000 +0000
@@ -35,7 +35,12 @@
  * Atheros module glue for rate control algorithms.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 
 #include <linux/version.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_scan_ap.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_scan_ap.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_scan_ap.c	2006-09-20 08:45:13.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_scan_ap.c	2011-03-17 14:21:10.000000000 +0000
@@ -38,7 +38,12 @@
  * IEEE 802.11 ap scanning support.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_scan.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_scan.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_scan.c	2008-01-29 09:17:23.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_scan.c	2011-03-17 14:19:47.000000000 +0000
@@ -38,7 +38,12 @@
  * IEEE 802.11 scanning support.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_scan_sta.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_scan_sta.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_scan_sta.c	2007-10-16 08:58:14.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_scan_sta.c	2011-03-17 14:21:52.000000000 +0000
@@ -38,7 +38,12 @@
  * IEEE 802.11 station scanning support.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_wireless.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_wireless.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_wireless.c	2009-06-24 20:40:32.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_wireless.c	2011-03-17 14:19:53.000000000 +0000
@@ -40,7 +40,12 @@
  * Wireless extensions support for 802.11 common code.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 
 #include <linux/version.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_xauth.c madwifi-0.9.4-r4133-20100621/net80211/ieee80211_xauth.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/ieee80211_xauth.c	2006-09-20 08:45:13.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/ieee80211_xauth.c	2011-03-17 14:22:16.000000000 +0000
@@ -47,7 +47,12 @@
  * entirely from messages about stations joining and leaving.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/net80211/if_media.c madwifi-0.9.4-r4133-20100621/net80211/if_media.c
--- madwifi-0.9.4-r4133-20100621.orig/net80211/if_media.c	2006-09-20 08:45:13.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/net80211/if_media.c	2011-03-17 14:15:01.000000000 +0000
@@ -50,7 +50,12 @@
 #endif
 
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/regression/ccmp/test_ccmp.c madwifi-0.9.4-r4133-20100621/regression/ccmp/test_ccmp.c
--- madwifi-0.9.4-r4133-20100621.orig/regression/ccmp/test_ccmp.c	2008-08-27 16:04:04.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/regression/ccmp/test_ccmp.c	2011-03-17 14:23:47.000000000 +0000
@@ -45,7 +45,12 @@
  * 1, 2, and 3.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/regression/tkip/test_tkip.c madwifi-0.9.4-r4133-20100621/regression/tkip/test_tkip.c
--- madwifi-0.9.4-r4133-20100621.orig/regression/tkip/test_tkip.c	2008-08-27 16:04:04.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/regression/tkip/test_tkip.c	2011-03-17 14:23:53.000000000 +0000
@@ -35,7 +35,12 @@
  * TKIP test module.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
diff -Nurp madwifi-0.9.4-r4133-20100621.orig/regression/wep/test_wep.c madwifi-0.9.4-r4133-20100621/regression/wep/test_wep.c
--- madwifi-0.9.4-r4133-20100621.orig/regression/wep/test_wep.c	2008-08-27 16:04:04.000000000 +0000
+++ madwifi-0.9.4-r4133-20100621/regression/wep/test_wep.c	2011-03-17 14:23:42.000000000 +0000
@@ -45,7 +45,12 @@
  * 1, 2, and 3.
  */
 #ifndef AUTOCONF_INCLUDED
-#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 37)
+#include <generated/autoconf.h>
+#else
+#include <linux/autoconf.h>
+#endif
 #endif
 #include <linux/version.h>
 #include <linux/module.h>
