From 51c1fdb0770a63ee7cf80922517a3296076e579c Mon Sep 17 00:00:00 2001
From: Fabio Erculiani <lxnay@sabayonlinux.org>
Date: Thu, 1 Oct 2009 19:23:50 +0200
Subject: [PATCH] [entropy.db] add new method "getSystemPackages"

---
 libraries/entropy/db.py |   10 ++++++++++
 1 files changed, 10 insertions(+), 0 deletions(-)

diff --git a/libraries/entropy/db.py b/libraries/entropy/db.py
index 9f11290..808c7e0 100644
--- a/libraries/entropy/db.py
+++ b/libraries/entropy/db.py
@@ -3169,6 +3169,16 @@ class EntropyRepository:
         """
         return get_spm(self).get_package_category_description_metadata(category)
 
+    def getSystemPackages(self):
+        """
+        Obtain a list of system packages on this repository.
+
+        @return: list (set) of system package identifiers
+        @rtype: set
+        """
+        cur = self.cursor.execute("SELECT idpackage FROM systempackages")
+        return self._cur2set(cur)
+
     def getIDPackages(self, atom):
         """
         Obtain repository package identifiers from atom string.
-- 
1.6.3.3

