--- libnl-1.0-pre5/include/netlink-local.h.__u64_x86_64	2005-11-02 12:29:06.000000000 -0500
+++ libnl-1.0-pre5/include/netlink-local.h	2006-05-26 20:09:28.000000000 -0400
@@ -30,11 +30,7 @@
 #include <arpa/inet.h>
 #include <netdb.h>
 
-typedef uint8_t		__u8;
-typedef uint16_t	__u16;
-typedef uint32_t	__u32;
-typedef int32_t		__s32;
-typedef uint64_t	__u64;
+#include <linux/types.h>
 
 /* local header copies */
 #include <linux/if.h>
