diff -Nurp libnl-1.1.orig/include/linux/if_link.h libnl-1.1/include/linux/if_link.h
--- libnl-1.1.orig/include/linux/if_link.h	2008-01-30 17:47:40.000000000 +0000
+++ libnl-1.1/include/linux/if_link.h	2008-01-30 17:47:51.000000000 +0000
@@ -1,6 +1,7 @@
 #ifndef _LINUX_IF_LINK_H
 #define _LINUX_IF_LINK_H
 
+typedef unsigned long long __u64;
 #include <linux/netlink.h>
 
 /* The struct should be in sync with struct net_device_stats */
