diff -Nurp ParaView3.orig/Utilities/hdf5/H5Zdeflate.c ParaView3/Utilities/hdf5/H5Zdeflate.c
--- ParaView3.orig/Utilities/hdf5/H5Zdeflate.c	2007-01-15 16:44:26.000000000 +0100
+++ ParaView3/Utilities/hdf5/H5Zdeflate.c	2009-07-29 17:28:03.832529035 +0200
@@ -17,6 +17,7 @@
  *              Friday, August 27, 1999
  */
 
+#include <zlib.h>
 #define H5Z_PACKAGE    /*suppress error about including H5Zpkg    */
 
 
