diff -Naur ParaView3/Utilities/Xdmf2/libsrc/XdmfArray.h ParaView3.new/Utilities/Xdmf2/libsrc/XdmfArray.h
--- ParaView3/Utilities/Xdmf2/libsrc/XdmfArray.h	2008-02-21 11:55:28.000000000 -0500
+++ ParaView3.new/Utilities/Xdmf2/libsrc/XdmfArray.h	2008-10-15 18:02:17.000000000 -0400
@@ -26,6 +26,7 @@
 #define __XdmfArray_h
 
 #include "XdmfDataDesc.h"
+#include <cstdlib>
 
 #define XDMF_LONGEST_LENGTH     (XdmfInt64)~0
 #define XDMF_ARRAY_TAG_LENGTH  80
diff -Naur ParaView3/Utilities/Xdmf2/libsrc/XdmfDataDesc.h ParaView3.new/Utilities/Xdmf2/libsrc/XdmfDataDesc.h
--- ParaView3/Utilities/Xdmf2/libsrc/XdmfDataDesc.h	2008-02-21 11:55:28.000000000 -0500
+++ ParaView3.new/Utilities/Xdmf2/libsrc/XdmfDataDesc.h	2008-10-15 18:02:17.000000000 -0400
@@ -27,6 +27,7 @@
 
 #include "XdmfObject.h"
 #include "XdmfHDFSupport.h"
+#include <cstdlib>
 
 #define XDMF_SELECTALL    0
 #define XDMF_HYPERSLAB    1
diff -Naur ParaView3/Utilities/Xdmf2/libsrc/XdmfH5Driver.cxx ParaView3.new/Utilities/Xdmf2/libsrc/XdmfH5Driver.cxx
--- ParaView3/Utilities/Xdmf2/libsrc/XdmfH5Driver.cxx	2007-07-12 15:10:04.000000000 -0400
+++ ParaView3.new/Utilities/Xdmf2/libsrc/XdmfH5Driver.cxx	2008-10-15 18:02:49.000000000 -0400
@@ -34,6 +34,7 @@
 #include "XdmfDsmComm.h"
 #include "assert.h"
 #include "hdf5.h"
+#include <cstdlib>
 
 #define HDF_IO_DEBUG 1
 #undef HDF_IO_DEBUG
diff -Naur ParaView3/Utilities/Xdmf2/libsrc/XdmfObject.h ParaView3.new/Utilities/Xdmf2/libsrc/XdmfObject.h
--- ParaView3/Utilities/Xdmf2/libsrc/XdmfObject.h	2008-02-08 10:25:42.000000000 -0500
+++ ParaView3.new/Utilities/Xdmf2/libsrc/XdmfObject.h	2008-10-15 18:02:17.000000000 -0400
@@ -121,6 +121,9 @@
 #define __DEPRECATED
 #   include <fstream>
 #   include <iomanip>
+
+#include <cstring>
+
 using std::cerr;
 using std::cout;
 using std::cin;
diff -Naur ParaView3/Utilities/Xdmf2/libsrc/XdmfRoot.h ParaView3.new/Utilities/Xdmf2/libsrc/XdmfRoot.h
--- ParaView3/Utilities/Xdmf2/libsrc/XdmfRoot.h	2007-04-25 15:23:59.000000000 -0400
+++ ParaView3.new/Utilities/Xdmf2/libsrc/XdmfRoot.h	2008-10-15 18:02:17.000000000 -0400
@@ -26,7 +26,7 @@
 #define __XdmfRoot_h
 
 #include "XdmfElement.h"
-
+#include <cstdlib>
 
 /*! XdmfRoot represents the Root Element in
 Xdmf. In XML it is the Element :
