# adjust function signature to changed hdf-1.6.7 API
diff -Naur ParaView3/Utilities/Xdmf2/libsrc/XdmfDataDesc.cxx ParaView3.new/Utilities/Xdmf2/libsrc/XdmfDataDesc.cxx
--- ParaView3/Utilities/Xdmf2/libsrc/XdmfDataDesc.cxx	2008-02-21 17:20:24.000000000 -0500
+++ ParaView3.new/Utilities/Xdmf2/libsrc/XdmfDataDesc.cxx	2008-10-15 16:02:08.000000000 -0400
@@ -350,7 +350,7 @@
 status = H5Sselect_elements( this->DataSpace,
         H5S_SELECT_SET,
          NElements,
-         ( const hsize_t **)HCoordinates);
+         ( const hsize_t *)HCoordinates);
 #else
 status = H5Sselect_elements( this->DataSpace,
         H5S_SELECT_SET,
