diff --git a/INSTALL b/INSTALL
index d9d8a4c..e9469e9 100644
--- a/INSTALL
+++ b/INSTALL
@@ -12,7 +12,7 @@ DEPENDENCIES
 In order to compile and install FSter you need:
   - cmake >= 2.8.0
   - fuse >= 2.8.1
-  - tracker-client >= 0.7.11
+  - tracker-client >= 0.8.1
   - libxml2 >= 2.7.4
 
 
@@ -72,7 +72,7 @@ sure of the following steps:
      cmake invocation
 
 4.b) Adjust your PKG_CONFIG_PATH environment variable to include the
-     folder where tracker-client-0.7.pc is located. Run the command
+     folder where tracker-client-0.8.pc is located. Run the command
      `locate` to figure out where they are.
 
 4.c) adjust your LD_LIBRARY_PATH setting to your tracker installation to
diff --git a/README b/README
index 65a7921..07ea6b7 100644
--- a/README
+++ b/README
@@ -13,7 +13,7 @@ REQUIREMENTS
 FSter depends on:
 
   - fuse >= 2.8.1
-  - tracker-client >= 0.7.11
+  - tracker-client >= 0.8.1
   - libxml2 >= 2.7.4
 
 INSTALLATION
diff --git a/content-plugins/CMakeLists.txt b/content-plugins/CMakeLists.txt
index 2536471..5318a2a 100644
--- a/content-plugins/CMakeLists.txt
+++ b/content-plugins/CMakeLists.txt
@@ -1,7 +1,7 @@
 find_package(PkgConfig REQUIRED)
 pkg_check_modules(GLIB REQUIRED glib-2.0)
 pkg_check_modules(LIBXML REQUIRED libxml-2.0)
-pkg_check_modules(TRACKER REQUIRED tracker-client-0.7)
+pkg_check_modules(TRACKER REQUIRED tracker-client-0.8)
 
 ADD_DEFINITIONS(
     -Wall
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index b02ef91..690daba 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -4,7 +4,7 @@ pkg_check_modules(GIO REQUIRED gio-2.0)
 pkg_check_modules(GTHREAD2 REQUIRED gthread-2.0)
 pkg_check_modules(DBUSGLIB REQUIRED dbus-glib-1)
 pkg_check_modules(LIBXML REQUIRED libxml-2.0)
-pkg_check_modules(TRACKER_CLIENT REQUIRED tracker-client-0.7)
+pkg_check_modules(TRACKER_CLIENT REQUIRED tracker-client-0.8)
 
 SET(CMAKE_SKIP_BUILD_RPATH TRUE)
 SET(CMAKE_BUILD_WITH_INSTALL_RPATH TRUE)

