--- pdftk-1.41/pdftk/Makefile.Base.orig	2006-10-25 01:44:06.000000000 +0200
+++ pdftk-1.41/pdftk/Makefile.Base	2008-08-15 17:21:18.073275624 +0200
@@ -63,7 +63,7 @@
 	$(CXX) pdftk.cc -I$(java_libs_root) $(CPPFLAGS) -c
 
 pdftk : pdftk.o attachments.o report.o $(itext_libs) $(libgcj_local_libs) $(afm_objects)
-	$(CXX) pdftk.o attachments.o report.o $(itext_libs) $(libgcj_local_libs) $(afm_objects) -I$(java_libs_root) $(CPPFLAGS) $(CXXFLAGS) -o pdftk
+	$(CXX) pdftk.o attachments.o report.o $(itext_libs) $(libgcj_local_libs) $(afm_objects) -I$(java_libs_root) $(CPPFLAGS) $(CXXFLAGS) $(LDFLAGS) -o pdftk
 
 install:
 	/usr/bin/install pdftk /usr/local/bin 
