diff -uNr alsa-utils-1.0.14.ORIG/configure.in alsa-utils-1.0.14/configure.in
--- alsa-utils-1.0.14.ORIG/configure.in	2007-06-27 16:22:05.000000000 +0100
+++ alsa-utils-1.0.14/configure.in	2007-06-27 16:22:37.000000000 +0100
@@ -82,6 +82,18 @@
   TESTSOUND="$dir/test.wav")
 AC_SUBST(TESTSOUND)
 
+AC_ARG_ENABLE([sequencer],
+  [AC_HELP_STRING([--disable-sequencer], [Disable sequencer utilities (aconnect, aplaymidi, aseqdump, aseqnet)])] )
+
+if test "x$enable_sequencer" != "xno"; then
+   AC_CHECK_LIB([asound], [snd_seq_open], [has_seq=yes], [has_seq=no])
+fi
+if test "x$enable_sequencer" = "xyes" && test "x$has_seq" = "xno"; then
+   AC_MSG_FAILURE([cannot find snd_seq_open. ALSA was built without sequencer support])
+fi
+
+AM_CONDITIONAL([BUILD_SEQ], [test "x$has_seq" = "xyes"])
+
 AM_CONFIG_HEADER(include/aconfig.h)
 
 dnl Checks for typedefs, structures, and compiler characteristics.
diff -uNr alsa-utils-1.0.14.ORIG/Makefile.am alsa-utils-1.0.14/Makefile.am
--- alsa-utils-1.0.14.ORIG/Makefile.am	2007-06-27 16:22:05.000000000 +0100
+++ alsa-utils-1.0.14/Makefile.am	2007-06-27 16:25:06.000000000 +0100
@@ -4,7 +4,10 @@
 else
 ALSAMIXER_DIR=
 endif
-SUBDIRS= include alsactl alsaconf $(ALSAMIXER_DIR) amidi amixer aplay iecset seq speaker-test utils m4 po
+if BUILD_SEQ
+SEQ_DIR=seq
+endif
+SUBDIRS= include alsactl alsaconf $(ALSAMIXER_DIR) amidi amixer aplay iecset $(SEQ_DIR) speaker-test utils m4 po
 EXTRA_DIST= config.rpath  config.rpath mkinstalldirs ChangeLog INSTALL TODO README configure hgcompile depcomp
 AUTOMAKE_OPTIONS=foreign
 ACLOCAL_AMFLAGS = -I m4
