Index: alsa-utils-1.0.11rc5/alsaconf/alsaconf.in
===================================================================
--- alsa-utils-1.0.11rc5.orig/alsaconf/alsaconf.in
+++ alsa-utils-1.0.11rc5/alsaconf/alsaconf.in
@@ -1394,7 +1394,7 @@ if [ x$devs_found != x ]; then
          Following card(s) are found on your system.
          Choose a soundcard to configure:
 ")
-    $DIALOG --title "$title" --menu "$msg" 17 76 8 "${devs_found[@]}" 2> $FOUND || acex 0
+    $DIALOG --title "$title" --menu "$msg" 17 76 8 "${devs_found[@]}" --output-fd 3 3> $FOUND || acex 0
     card=`head -n 1 $FOUND`
     if [ "$card" = "legacy" ]; then
 	ac_config_legacy
