--- dcbd-0.9.24.orig/event_iface.c
+++ dcbd-0.9.24/event_iface.c
@@ -47,13 +47,6 @@
 #include "lldp/ports.h"
 #include "lldp/l2_packet.h"
 
-extern struct if_nameindex {
-	unsigned int if_index;
-	char         *if_name;
-};
-extern struct if_nameindex *if_nameindex(void);
-extern void if_freenameindex(struct if_nameindex *);
-
 static void event_if_decode_rta(int type, struct rtattr *rta);
 
 #define MAX_PAYLOAD 4096  /* maximum payload size*/
