diff -Nurp dcbd-0.9.7-current/dcbd.init dcbd-0.9.7-corrected/dcbd.init
--- dcbd-0.9.7-current/dcbd.init	2009-03-11 17:51:59.000000000 +0100
+++ dcbd-0.9.7-corrected/dcbd.init	2009-03-11 18:19:10.000000000 +0100
@@ -33,15 +33,16 @@
 # dcbd		This shell script takes care of starting and stopping
 #		dcbd (DCB capabilities exchange protocol)
 #
-# chkconfig: 345 20 80
+# chkconfig: - 20 80
 # description: Data Center Bridging Exchange protocol daemon
 #
 ### BEGIN INIT INFO
 # Provides: dcbd
 # Required-Start: network
 # Required-Stop:
-# Default-Start: 3 5
+# Default-Start:
 # Default-Stop:
+# Short-Description: Data Center Bridging Exchange protocol daemon
 # Description: Data Center Bridging Exchange protocol daemon
 ### END INIT INFO
 
