diff -Nurp dcbd-0.9.15/dcbd.8 dcbd-0.9.15-corrected/dcbd.8
--- dcbd-0.9.15/dcbd.8	2009-07-09 01:17:03.000000000 +0200
+++ dcbd-0.9.15-corrected/dcbd.8	2009-08-17 13:07:08.693587866 +0200
@@ -61,7 +61,7 @@ show usage information
 .TP
 .BI "-f" " filename"
 use the specified file as the configuration file instead of the default file:
-/etc/sysconfig/dcbd/dcbd.conf
+/etc/dcbd/dcbd.conf
 .B dcbd
 expects the directory of the configuration file to exist, but if the
 configuration file does not exist, then a default configuration file will
diff -Nurp dcbd-0.9.15/include/dcbd.h dcbd-0.9.15-corrected/include/dcbd.h
--- dcbd-0.9.15/include/dcbd.h	2009-07-09 01:13:57.000000000 +0200
+++ dcbd-0.9.15-corrected/include/dcbd.h	2009-08-17 13:07:33.313595906 +0200
@@ -43,7 +43,7 @@
 #define CLIF_IFACE_DIR "/var/run/dcbd"
 #define CLIF_IFACE_IFNAME "clif"
 
-#define DEFAULT_CFG_FILE "/etc/sysconfig/dcbd/dcbd.conf"
+#define DEFAULT_CFG_FILE "/etc/dcbd/dcbd.conf"
 
 #define DEFAULT_DCBX_VERSION  2
 
diff -Nurp dcbd-0.9.15/Makefile.am dcbd-0.9.15-corrected/Makefile.am
--- dcbd-0.9.15/Makefile.am	2009-07-09 01:13:57.000000000 +0200
+++ dcbd-0.9.15-corrected/Makefile.am	2009-08-17 13:08:13.101591160 +0200
@@ -58,7 +58,7 @@ nltest_SOURCES = nltest.c nltest.h
 dist_man_MANS = dcbd.8 dcbtool.8
 
 ## force the creation of an empty configuration directory at install time
-dcbdconfigdir = /etc/sysconfig/dcbd
+dcbdconfigdir = /etc/dcbd
 dcbdconfig_DATA =
 
 ## pkg-config data file, for client interface software to find headers
diff -Nurp dcbd-0.9.15/Makefile.in dcbd-0.9.15-corrected/Makefile.in
--- dcbd-0.9.15/Makefile.in	2009-07-09 01:14:09.000000000 +0200
+++ dcbd-0.9.15-corrected/Makefile.in	2009-08-17 13:08:21.275576489 +0200
@@ -299,7 +299,7 @@ $(dcbd_include_HEADERS) $(noinst_HEADERS
 
 nltest_SOURCES = nltest.c nltest.h
 dist_man_MANS = dcbd.8 dcbtool.8
-dcbdconfigdir = /etc/sysconfig/dcbd
+dcbdconfigdir = /etc/dcbd
 dcbdconfig_DATA = 
 pkgconfigdir = ${libdir}/pkgconfig
 pkgconfig_DATA = dcbd.pc
diff -Nurp dcbd-0.9.15/README dcbd-0.9.15-corrected/README
--- dcbd-0.9.15/README	2009-07-09 01:13:57.000000000 +0200
+++ dcbd-0.9.15-corrected/README	2009-08-17 13:08:37.654597087 +0200
@@ -159,7 +159,7 @@ dcbd Application Install
      ./configure --prefix=/usr; make; make install
 
    This will build and install 'dcbd' and 'dcbtool' to /usr/sbin, make
-   the '/etc/sysconfig/dcbd' directory (default location of
+   the '/etc/dcbd' directory (default location of
    the dcbd.conf file) and setup dcbd to run as a system service using the
    chkconfig program.  Verify that the dcbd service is working as expected
    with the 'service dcbd status' command. If the service is not on, issue
@@ -181,7 +181,7 @@ Options
 dcbd has the following command line options:
     -h show usage information
     -f configfile    use the specified file as the config file instead of
-                     the default file - /etc/sysconfig/dcbd/dcbd.conf
+                     the default file - /etc/dcbd/dcbd.conf
     -d run dcbd as a daemon
     -v show dcbd version
 
