diff -rup virt-manager-0.3.1-orig/src/virtManager/create.py virt-manager-0.3.1-new/src/virtManager/create.py
--- virt-manager-0.3.1-orig/src/virtManager/create.py	2007-02-20 15:21:37.000000000 -0500
+++ virt-manager-0.3.1-new/src/virtManager/create.py	2007-02-20 16:43:02.000000000 -0500
@@ -688,10 +688,11 @@ class vmmCreate(gobject.GObject):
 
 
         elif page_num == 2: # the virt method page
-            if self.get_config_method() == VM_FULLY_VIRT and not virtinst.util.is_hvm_capable():
-                self._validation_error_box(_("Hardware Support Required"), \
-                                           _("Your hardware does not appear to support full virtualization. Only paravirtualized guests will be available on this hardware."))
-                return False
+            if self.get_config_method() == VM_FULLY_VIRT:
+                if self.connection.get_type() != "QEMU" and not virtinst.util.is_hvm_capable():
+                    self._validation_error_box(_("Hardware Support Required"), \
+                                               _("Your hardware does not appear to support full virtualization. Only paravirtualized guests will be available on this hardware."))
+                    return False
 
         elif page_num == 3: # the fully virt media page
             if self.window.get_widget("media-iso-image").get_active():
Only in virt-manager-0.3.1-new/src/virtManager: create.py~
