--- openrc-0.6.8.orig/etc/rc.conf.in
+++ openrc-0.6.8/etc/rc.conf.in
@@ -13,7 +13,7 @@
 # boot so you can choose to start specific services. Set to "NO" to disable
 # this feature. This feature is automatically disabled if rc_parallel is
 # set to YES.
-#rc_interactive="YES"
+rc_interactive="YES"
 
 # If we need to drop to a shell, you can specify it here.
 # If not specified we use $SHELL, otherwise the one specified in /etc/passwd,
