diff -Nurp openrc-0.5.3.orig/sh/rc-functions.sh.in openrc-0.5.3/sh/rc-functions.sh.in
--- openrc-0.5.3.orig/sh/rc-functions.sh.in	2009-12-02 20:08:02.000000000 +0100
+++ openrc-0.5.3/sh/rc-functions.sh.in	2009-12-26 05:42:17.000000000 +0100
@@ -18,10 +18,10 @@ _addon_warn()
 import_addon()
 {
 	if [ -e /@LIB@/rc/addons/"$1".sh ]; then
-		_addon_warn
+		# _addon_warn
 		. /@LIB@/rc/addons/"$1".sh
 	elif [ -e /@LIB@/rcscripts/addons/"$1".sh ]; then
-		_addon_warn
+		# _addon_warn
 		. /@LIB@/rcscripts/addons/"$1".sh
 	else
 		return 1
