diff -Naur openrc-0.6.1.a/init.d/savecache.in openrc-0.6.1.b/init.d/savecache.in
--- openrc-0.6.1.a/init.d/savecache.in	2010-03-22 15:02:12.000000000 -0500
+++ openrc-0.6.1.b/init.d/savecache.in	2010-07-13 13:56:47.000000000 -0500
@@ -10,7 +10,7 @@
 		ewarn "WARNING: clock skew detected!"
 		if ! yesno "savecache_skewed"; then
 			eerror "Not saving deptree cache"
-			return 1
+			return 0
 		fi
 	fi
 	ebegin "Saving dependency cache"
