--- openrc-0.5.3.orig/conf.d/hwclock
+++ openrc-0.5.3/conf.d/hwclock
@@ -2,12 +2,12 @@
 # Greenwich Mean Time).  If that clock is set to the local time, then 
 # set CLOCK to "local".  Note that if you dual boot with Windows, then 
 # you should set it to "local".
-clock="UTC"
+clock="local"
 
 # If you want to set the Hardware Clock to the current System Time 
 # (software clock) during shutdown, then say "YES" here.
 # You normally don't need to do this if you run a ntp daemon.
-clock_systohc="NO"
+clock_systohc="YES"
 
 # If you wish to pass any other arguments to hwclock during bootup,
 # you may do so here. Alpha users may wish to use --arc or --srm here.
--- openrc-0.5.3.orig/etc/rc.conf.in
+++ openrc-0.5.3/etc/rc.conf.in
@@ -38,13 +38,13 @@
 # This allows net.wlan and any service not matching net.* to be plugged.
 # Example - rc_hotplug="*"
 # This allows all services to be hotplugged
-#rc_hotplug="*"
+rc_hotplug="!net.*"
 
 # rc_logger launches a logging daemon to log the entire rc process to
 # /var/log/rc.log
 # NOTE: Linux systems require the devfs service to be started before
 # logging can take place and as such cannot log the sysinit runlevel.
-#rc_logger="YES"
+rc_logger="YES"
 
 # By default we filter the environment for our running scripts. To allow other
 # variables through, add them here. Use a * to allow all variables through.
