--- src/info/qt_ui.cpp.old	2010-06-06 14:45:17.007326722 +0300
+++ src/info/qt_ui.cpp	2010-06-06 14:45:32.622321695 +0300
@@ -260,7 +260,7 @@
                 int64_t position,
                 int64_t size) {
   if (!g_use_gui)
-    console_show_element(level, text, position);
+    console_show_element(level, text, position, size);
 
   else if (0 <= position)
     gui->add_item(level, Q(create_element_text(text, position, size).c_str()));
