diff -ruN cups-1.5.2.orig/conf/cupsd.conf.in cups-1.5.2/conf/cupsd.conf.in
--- cups-1.5.2.orig/conf/cupsd.conf.in	2010-12-09 22:24:51.000000000 +0100
+++ cups-1.5.2/conf/cupsd.conf.in	2012-04-19 00:12:31.004045949 +0200
@@ -18,7 +18,8 @@
 @CUPS_LISTEN_DOMAINSOCKET@
 
 # Show shared printers on the local network.
-Browsing On
+# We switch this off by default in Gentoo, to avoid an unnecessary open port.
+Browsing Off
 BrowseOrder allow,deny
 BrowseAllow all
 BrowseLocalProtocols @CUPS_BROWSE_LOCAL_PROTOCOLS@
