diff -ruN cups-1.5.2.orig/scripting/perl/Makefile.PL cups-1.5.2/scripting/perl/Makefile.PL
--- cups-1.5.2.orig/scripting/perl/Makefile.PL	2002-08-19 22:43:03.000000000 +0200
+++ cups-1.5.2/scripting/perl/Makefile.PL	2012-04-19 22:11:45.000000000 +0200
@@ -8,7 +8,7 @@
     ($] >= 5.005 ?    ## Add these new keywords supported since 5.005
       (ABSTRACT_FROM => 'CUPS.pm', # retrieve abstract from module
        AUTHOR     => 'A. U. Thor <a.u.thor@a.galaxy.far.far.away>') : ()),
-    'LIBS'		=> ['-lcups '], # e.g., '-lm'
+    'LIBS'		=> ['-L../../cups -lcups '], # e.g., '-lm'
     'DEFINE'		=> '', # e.g., '-DHAVE_SOMETHING'
 	# Insert -I. if you add *.h files later:
     'INC'		=> '', # e.g., '-I/usr/include/other'
