--- libntfs/Makefile.am
+++ libntfs/Makefile.am
@@ -38,8 +38,8 @@
 	-DLTVERSION_LIBNTFS=\"$(LTVERSION_LIBNTFS)\"
 
 if ENABLE_CRYPTO
-libntfs_la_LDFLAGS += `libgnutls-config --libs` `pkg-config --libs libconfig`
-libntfs_la_CFLAGS += `libgnutls-config --cflags` `pkg-config --cflags libconfig`
+libntfs_la_LDFLAGS += `pkg-config --libs gnutls libconfig`
+libntfs_la_CFLAGS += `pkg-config --cflags gnutls libconfig`
 endif
 
 libntfs_la_SOURCES =	\
--- libntfs/Makefile.in
+++ libntfs/Makefile.in
@@ -61,8 +61,8 @@
 build_triplet = @build@
 host_triplet = @host@
 target_triplet = @target@
-@ENABLE_CRYPTO_TRUE@am__append_1 = `libgnutls-config --libs` `pkg-config --libs libconfig`
-@ENABLE_CRYPTO_TRUE@am__append_2 = `libgnutls-config --cflags` `pkg-config --cflags libconfig`
+@ENABLE_CRYPTO_TRUE@am__append_1 = `pkg-config --libs gnutls libconfig`
+@ENABLE_CRYPTO_TRUE@am__append_2 = `pkg-config --cflags gnutls libconfig`
 subdir = libntfs
 DIST_COMMON = $(srcdir)/Makefile.am $(srcdir)/Makefile.in \
 	$(srcdir)/libntfs-gnomevfs.8.in $(srcdir)/libntfs.8.in \
--- ntfsprogs/Makefile.am
+++ ntfsprogs/Makefile.am
@@ -125,8 +125,8 @@
 if ENABLE_CRYPTO
 ntfsdecrypt_SOURCES	= ntfsdecrypt.c utils.c utils.h
 ntfsdecrypt_LDADD	= $(AM_LIBS)
-ntfsdecrypt_LDFLAGS	= $(AM_LFLAGS) `libgnutls-config --libs`
-ntfsdecrypt_CFLAGS	= `libgnutls-config --cflags`
+ntfsdecrypt_LDFLAGS	= $(AM_LFLAGS) `pkg-config --libs gnutls`
+ntfsdecrypt_CFLAGS	= `pkg-config --cflags gnutls`
 endif
 
 # Extra targets
--- ntfsprogs/Makefile.in
+++ ntfsprogs/Makefile.in
@@ -449,8 +449,8 @@
 ntfsdump_logfile_LDFLAGS = $(AM_LFLAGS)
 @ENABLE_CRYPTO_TRUE@ntfsdecrypt_SOURCES = ntfsdecrypt.c utils.c utils.h
 @ENABLE_CRYPTO_TRUE@ntfsdecrypt_LDADD = $(AM_LIBS)
-@ENABLE_CRYPTO_TRUE@ntfsdecrypt_LDFLAGS = $(AM_LFLAGS) `libgnutls-config --libs`
-@ENABLE_CRYPTO_TRUE@ntfsdecrypt_CFLAGS = `libgnutls-config --cflags`
+@ENABLE_CRYPTO_TRUE@ntfsdecrypt_LDFLAGS = $(AM_LFLAGS) `pkg-config --libs gnutls`
+@ENABLE_CRYPTO_TRUE@ntfsdecrypt_CFLAGS = `pkg-config --cflags gnutls`
 all: all-am
 
 .SUFFIXES:
