--- ntfsprogs-2.0.0/configure.ac
+++ ntfsprogs-2.0.0/configure.ac
@@ -79,6 +79,10 @@
 	ac_configure_args="$ac_configure_args --prefix $prefix"
 fi
 
+AC_ARG_ENABLE([extras],[AS_HELP_STRING([--enable-extras],[build/install the extra programs])],,enable_extras=no)
+AM_CONDITIONAL(ENABLE_EXTRAS, test "x$enable_extras" = xyes)
+
+
 # Command-line options.
 AC_ARG_ENABLE(debug,
 	AS_HELP_STRING(--enable-debug,enable additional debugging code and
--- ntfsprogs-2.0.0/ntfsprogs/Makefile.am
+++ ntfsprogs-2.0.0/ntfsprogs/Makefile.am
@@ -14,7 +14,8 @@
 bin_PROGRAMS		= ntfsfix ntfsinfo ntfscluster ntfsls ntfscat ntfscmp
 sbin_PROGRAMS		= mkntfs ntfslabel ntfsundelete ntfsresize ntfsclone \
 			  ntfscp
-EXTRA_PROGRAMS		= ntfsdump_logfile ntfswipe ntfstruncate ntfsmove \
+EXTRA_PROGRAMS		=
+extra_progs		= ntfsdump_logfile ntfswipe ntfstruncate ntfsmove \
 			  ntfsmftalloc ntfsck
 
 man_MANS		= mkntfs.8 ntfsfix.8 ntfslabel.8 ntfsinfo.8 \
@@ -37,6 +38,12 @@
 EXTRA_PROGRAMS		+= ntfsdecrypt
 endif
 
+if ENABLE_EXTRAS
+bin_PROGRAMS		+= $(extra_progs)
+else
+EXTRA_PROGRAMS		+= $(extra_progs)
+endif
+
 # Set the include path.
 AM_CPPFLAGS		= -I$(top_srcdir)/include/ntfs $(all_includes)
 
