Disable tests requiring an unreleased version of gobject-introspection.

diff --git a/tests/test_everything.py b/tests/test_everything.py
index 19d345b..4a03890 100644
--- a/tests/test_everything.py
+++ b/tests/test_everything.py
@@ -483,22 +483,3 @@ class TestAdvancedInterfaces(unittest.TestCase):
         self.assertEquals(ret[0], 51);
         self.assertEquals(ret[1], 61);
         self.assertEquals(ret[2], 32);
-
-    def test_obj_skip_return_val_no_out(self):
-        obj = Everything.TestObj();
-        # raises an error for 0, succeeds for any other value
-        self.assertRaises(GLib.GError, obj.skip_return_val_no_out, 0)
-
-        ret = obj.skip_return_val_no_out(1)
-        self.assertEquals(ret, None)
-
-class TestSignals(unittest.TestCase):
-    def test_object_param_signal(self):
-        obj = Everything.TestObj();
-
-        def callback (obj_param):
-            self.assertEquals(obj_param.props.int, 3)
-            self.assertEquals(obj_param.__grefcount__, 2)
-
-        obj.connect('sig-with-obj', callback)
-        obj.emit_sig_with_obj()
