From: Nathan Phillip Brink <binki@gentoo.org>
Subject: Don't call LD when linking, use CC instead.

--- a/libdump/Makefile
+++ b/libdump/Makefile
@@ -26,12 +26,12 @@
 all32:	$(OBJ32)
 	$(AR) $(NAME32).a $(OBJ32)
 	$(RANLIB) $(NAME32).a
-	$(LD) -r $(OBJ32) -o $(NAME32).o $(LDFLAGS32)
+	$(CC) -Wl,-r -nostdlib $(OBJ32) -o $(NAME32).o $(LDFLAGS32)
 
 all64:	$(OBJ64)
 	$(AR) $(NAME64).a $(OBJ64)
 	$(RANLIB) $(NAME64).a
-	$(LD) -r $(OBJ64) -o $(NAME64).o $(LDFLAGS64)
+	$(CC) -Wl,-r -nostdlib $(OBJ64) -o $(NAME64).o $(LDFLAGS64)
 
 clean	:
 	@rm -f $(OBJ32) $(OBJ64) *~ *\# \#* .\#* dump_main.o

--- a/elfsh/Makefile
+++ b/elfsh/Makefile
@@ -36,7 +36,7 @@
 			 -I../libedfmt/include/  $(STATOPT2)                        \
 			 $(ASMOPT) -DERESI64 $(NCURSESLNOPT) $(READLNOPT)
 
-LDFLAGS32		+= -L../libedfmt/ -ledfmt32 -L../libelfsh/ -lelfsh32        \
+LDFLAGS32		+= -L../libedfmt/ -ledfmt32 -L../libdump/ -L../libelfsh/ -lelfsh32        \
 			-L../libetrace/ -letrace32 -L../libstderesi/ -lstderesi32   \
 			-L/usr/pkg/lib/ -L../libmjollnir/ -lmjollnir32              \
                         -L../libetrace -letrace32 $(LPTHREAD) -L../libasm/ -lasm32  \
