diff -uNr xf86-input-synaptics-0.15.0.ORIG/src/synaptics.c xf86-input-synaptics-0.15.0/src/synaptics.c
--- xf86-input-synaptics-0.15.0.ORIG/src/synaptics.c	2008-08-22 17:34:20.000000000 +0100
+++ xf86-input-synaptics-0.15.0/src/synaptics.c	2008-08-22 17:34:45.000000000 +0100
@@ -648,10 +648,10 @@
 #endif
 			    );
     /* X valuator */
-    xf86InitValuatorAxisStruct(dev, 0, 0, -1, 1, 0, 1);
+    xf86InitValuatorAxisStruct(dev, 0, -1, -1, 1, 0, 1);
     xf86InitValuatorDefaults(dev, 0);
     /* Y valuator */
-    xf86InitValuatorAxisStruct(dev, 1, 0, -1, 1, 0, 1);
+    xf86InitValuatorAxisStruct(dev, 1, -1, -1, 1, 0, 1);
     xf86InitValuatorDefaults(dev, 1);
 #if GET_ABI_MAJOR(ABI_XINPUT_VERSION) == 0
     xf86MotionHistoryAllocate(local);
