--- xf86-input-synaptics-1.2.1.orig/conf/11-x11-synaptics.fdi
+++ xf86-input-synaptics-1.2.1/conf/11-x11-synaptics.fdi
@@ -5,27 +5,85 @@
         <merge key="input.x11_driver" type="string">synaptics</merge>
         <!-- Arbitrary options can be passed to the driver using
              the input.x11_options property since xorg-server-1.5. -->
-        <!-- EXAMPLES:
-        Switch on shared memory, enables the driver to be configured at runtime
+
+	<!-- Switch on shared memory, enables the driver to be configured at runtime -->
 	<merge key="input.x11_options.SHMConfig" type="string">true</merge>
 
-	Maximum movement of the finger for detecting a tap
+	<!-- Maximum movement of the finger for detecting a tap -->
 	<merge key="input.x11_options.MaxTapMove" type="string">2000</merge>
 
-	Enable vertical scrolling when dragging along the right edge
+	<!-- Enable vertical scrolling when dragging along the right edge -->
 	<merge key="input.x11_options.VertEdgeScroll" type="string">true</merge>
 
-	Enable vertical scrolling when dragging with two fingers anywhere on the touchpad
+	<!-- Enable vertical scrolling when dragging with two fingers anywhere on the touchpad -->
 	<merge key="input.x11_options.VertTwoFingerScroll" type="string">true</merge>
 
-	Enable horizontal scrolling when dragging with two fingers anywhere on the touchpad
+	<!-- Enable horizontal scrolling when dragging with two fingers anywhere on the touchpad -->
 	<merge key="input.x11_options.HorizTwoFingerScroll" type="string">true</merge>
 
-	If on, circular scrolling is used
-	<merge key="input.x11_options.CircularScrolling" type="string">true</merge>
+	<!-- Disable circular scrolling by default, this could cause issue to some touchpads -->
+	<merge key="input.x11_options.CircularScrolling" type="string">false</merge>
+
+	<!-- Strictly define buttons used for tapping -->
+        <merge key="input.x11_options.TapButton1" type="string">1</merge>
+        <merge key="input.x11_options.TapButton2" type="string">2</merge>
+        <merge key="input.x11_options.TapButton3" type="string">3</merge>
 
-	For other possible options, check CONFIGURATION DETAILS in synaptics man page
+    </match>
+
+    <!-- MacBook 5 touchpad support -->
+    <match key="info.capabilities" contains="input.touchpad">
+      <match key="info.product" contains="Synaptics TouchPad">
+        <merge key="input.x11_driver" type="string">synaptics</merge>
+        <!-- Arbitrary options can be passed to the driver using 
+             the input.x11_options property since xorg-server-1.5. -->
+        <!-- EXAMPLE:
+        <merge key="input.x11_options.LeftEdge" type="string">120</merge>
         -->
+      </match>
+      <match key="info.product" contains="AlpsPS/2 ALPS">
+        <merge key="input.x11_driver" type="string">synaptics</merge>
+      </match>
+      <match key="info.product" contains="appletouch">
+        <merge key="input.x11_driver" type="string">synaptics</merge>
+      </match>
+      <match key="info.product" contains="bcm5974">
+        <merge key="input.x11_driver" type="string">synaptics</merge>
+        <merge key="input.x11_options.LeftEdge" type="string">0</merge>
+        <merge key="input.x11_options.RightEdge" type="string">1280</merge>
+        <merge key="input.x11_options.TopEdge" type="string">0</merge>
+        <merge key="input.x11_options.BottomEdge" type="string">800</merge>
+        <merge key="input.x11_options.ClickFinger1" type="string">1</merge>
+        <merge key="input.x11_options.ClickFinger2" type="string">3</merge>
+        <merge key="input.x11_options.ClickFinger3" type="string">2</merge>
+        <merge key="input.x11_options.HorizEdgeScroll" type="string">0</merge>
+        <merge key="input.x11_options.VertEdgeScroll" type="string">0</merge>
+        <merge key="input.x11_options.VertTwoFingerScroll" type="string">1</merge>
+        <merge key="input.x11_options.HorizTwoFingerScroll" type="string">1</merge>
+        <merge key="input.x11_options.HorizScrollDelta" type="string">0</merge>
+        <merge key="input.x11_options.VertScrollDelta" type="string">40</merge>
+        <merge key="input.x11_options.PressureMotionMinZ" type="string">10</merge>
+        <merge key="input.x11_options.FingerLow" type="string">16</merge>
+        <merge key="input.x11_options.FingerHigh" type="string">80</merge>
+        <merge key="input.x11_options.FingerPress" type="string">256</merge>
+        <merge key="input.x11_options.PalmDetect" type="string">1</merge>
+        <merge key="input.x11_options.PalmMinWidth" type="string">10</merge>
+        <merge key="input.x11_options.PalmMinZ" type="string">200</merge>
+        <merge key="input.x11_options.MinSpeed" type="string">0.8</merge>
+        <merge key="input.x11_options.MaxSpeed" type="string">1.2</merge>
+        <merge key="input.x11_options.AccelFactor" type="string">0.10</merge>
+        <merge key="input.x11_options.MaxTapMove" type="string">25</merge>
+        <merge key="input.x11_options.MaxTapTime" type="string">223</merge>
+        <merge key="input.x11_options.MaxDoubleTapTime" type="string">200</merge>
+        <merge key="input.x11_options.TapButton1" type="string">0</merge>
+        <merge key="input.x11_options.TapButton2" type="string">0</merge>
+        <merge key="input.x11_options.TapButton3" type="string">0</merge>
+        <merge key="input.x11_options.RTCornerButton" type="string">0</merge>
+        <merge key="input.x11_options.RBCornerButton" type="string">0</merge>
+        <merge key="input.x11_options.LTCornerButton" type="string">0</merge>
+        <merge key="input.x11_options.LBCornerButton" type="string">0</merge>
+      </match>
     </match>
+
   </device>
 </deviceinfo>
