diff -Nurp unicorn.orig/msw/fp_stubs.c unicorn/msw/fp_stubs.c
--- unicorn.orig/msw/fp_stubs.c	2006-12-25 19:29:05.000000000 +0100
+++ unicorn/msw/fp_stubs.c	2006-12-25 19:33:27.000000000 +0100
@@ -1,6 +1,8 @@
 #include <linux/version.h>
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,18)
 #	include <linux/config.h>
+#else
+#	include <linux/autoconf.h>
 #endif
 #if defined(CONFIG_MODVERSIONS) && !defined(MODVERSIONS)
 #define MODVERSIONS
diff -Nurp unicorn.orig/msw/linrapi.c unicorn/msw/linrapi.c
--- unicorn.orig/msw/linrapi.c	2006-12-25 19:29:05.000000000 +0100
+++ unicorn/msw/linrapi.c	2006-12-25 19:32:56.000000000 +0100
@@ -8,6 +8,8 @@
 #include <linux/version.h>
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,18)
 #	include <linux/config.h>
+#else
+#	include <linux/autoconf.h>
 #endif
 #if defined(CONFIG_MODVERSIONS) && (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,0))
 #include <linux/modversions.h>
diff -Nurp unicorn.orig/unicorn_atm/unicorn_atmdrv.c unicorn/unicorn_atm/unicorn_atmdrv.c
--- unicorn.orig/unicorn_atm/unicorn_atmdrv.c	2006-12-25 19:29:05.000000000 +0100
+++ unicorn/unicorn_atm/unicorn_atmdrv.c	2006-12-25 19:37:04.000000000 +0100
@@ -7,6 +7,8 @@
 #include <linux/version.h>
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,18)
 #	include <linux/config.h>
+#else
+#	include <linux/autoconf.h>
 #endif
 #if defined(CONFIG_MODVERSIONS) && (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,0))
 #include <linux/modversions.h>
diff -Nurp unicorn.orig/unicorn_eth/unicorn_ethdrv.c unicorn/unicorn_eth/unicorn_ethdrv.c
--- unicorn.orig/unicorn_eth/unicorn_ethdrv.c	2006-12-25 19:29:05.000000000 +0100
+++ unicorn/unicorn_eth/unicorn_ethdrv.c	2006-12-25 19:37:23.000000000 +0100
@@ -6,7 +6,9 @@
 */
 #include <linux/version.h>
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,18)
-#	include <linux/config.h>
+#       include <linux/config.h>
+#else
+#       include <linux/autoconf.h>
 #endif
 #if defined(CONFIG_MODVERSIONS) && (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,0))
 #include <linux/modversions.h>
diff -Nurp unicorn.orig/unicorn_pci/unicorn_pcidrv.c unicorn/unicorn_pci/unicorn_pcidrv.c
--- unicorn.orig/unicorn_pci/unicorn_pcidrv.c	2006-12-25 19:29:05.000000000 +0100
+++ unicorn/unicorn_pci/unicorn_pcidrv.c	2006-12-25 19:37:37.000000000 +0100
@@ -6,7 +6,9 @@
 */
 #include <linux/version.h>
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,18)
-#	include <linux/config.h>
+#       include <linux/config.h>
+#else
+#       include <linux/autoconf.h>
 #endif
 #if defined(CONFIG_MODVERSIONS) && (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,0))
 #include <linux/modversions.h>
diff -Nurp unicorn.orig/unicorn_usb/unicorn_usbdrv.c unicorn/unicorn_usb/unicorn_usbdrv.c
--- unicorn.orig/unicorn_usb/unicorn_usbdrv.c	2006-12-25 19:29:05.000000000 +0100
+++ unicorn/unicorn_usb/unicorn_usbdrv.c	2006-12-25 19:37:48.000000000 +0100
@@ -6,7 +6,9 @@
 */
 #include <linux/version.h>
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,18)
-#	include <linux/config.h>
+#       include <linux/config.h>
+#else
+#       include <linux/autoconf.h>
 #endif
 #if defined(CONFIG_MODVERSIONS) && (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,0))
 #include <linux/modversions.h>
