--- a/files/usr/share/gnome-session/sessions/cinnamon.session
+++ b/files/usr/share/gnome-session/sessions/cinnamon.session
@@ -1,5 +1,5 @@
 [GNOME Session]
 Name=Cinnamon
-RequiredComponents=cinnamon;gnome-settings-daemon;
+RequiredComponents=cinnamon;gnome-settings-daemon;nemo-autostart;cinnamon-screensaver;gnome-fallback-media-keys-helper;gnome-fallback-mount-helper;
 DesktopName=GNOME
 
--- a/files/usr/share/gnome-session/sessions/cinnamon2d.session
+++ b/files/usr/share/gnome-session/sessions/cinnamon2d.session
@@ -1,5 +1,5 @@
 [GNOME Session]
 Name=Cinnamon (Software Rendering)
-RequiredComponents=cinnamon2d;gnome-settings-daemon;
+RequiredComponents=cinnamon2d;gnome-settings-daemon;nemo-autostart;cinnamon2d-screensaver;gnome-fallback-media-keys-helper;gnome-fallback-mount-helper;
 DesktopName=GNOME
 
