--- a/files/usr/lib/cinnamon-settings/modules/cs_info.py
+++ b/files/usr/lib/cinnamon-settings/modules/cs_info.py
@@ -31,12 +31,12 @@ def getProcessOut(command):
 def getGraphicsInfos():
     cards = {}
     count = 0
-    for card in getProcessOut(("lspci")):
+    for card in getProcessOut(("/usr/sbin/lspci")):
         if not "VGA" in card:
             continue
         cardId = card.split()[0]
         cardName = None
-        for line in getProcessOut(("lspci", "-v", "-s", cardId)):
+        for line in getProcessOut(("/usr/sbin/lspci", "-v", "-s", cardId)):
             if line.startswith(cardId):
                 cardName = (line.split(":")[2].split("(rev")[0].strip())
   
