--- gspcav2/Makefile.orig	2006-09-24 18:11:12.000000000 +0200
+++ gspcav2/Makefile	2006-09-24 18:10:49.000000000 +0200
@@ -26,9 +26,9 @@
 # Setup defines
 DEFINES   += -DCONFIG_USB_GSPCA_MODULE=1 -DMODULE -D__KERNEL__
 DEFINES   += -DVID_HARDWARE_GSPCA=0xFF -DGSPCA_VERSION=\"$(VERSION)\"
+CFLAGS += $(DEFINES)
 
 ifneq ($(KERNELRELEASE),)   # We were called by kbuild
-CFLAGS += $(DEFINES) 
 obj-m += gspca.o
 gspca-objs := gspcaV2.o
 
