Index: coregrind/m_libcassert.c
===================================================================
--- coregrind/m_libcassert.c	(revision 10904)
+++ coregrind/m_libcassert.c	(working copy)
@@ -65,8 +65,8 @@
 #elif defined(VGP_ppc32_linux) || defined(VGP_ppc32_aix5)
 #  define GET_REAL_PC_SP_AND_FP(pc, sp, fp)                   \
       asm("mflr 0;"                   /* r0 = lr */           \
-          "bl m_libcassert_get_ip;"   /* lr = pc */           \
-          "m_libcassert_get_ip:\n"                            \
+          "bl 0f;"                    /* lr = pc */           \
+          "0:\n"                                              \
           "mflr %0;"                \
           "mtlr 0;"                   /* restore lr */        \
           "mr %1,1;"                \
@@ -79,8 +79,8 @@
 #elif defined(VGP_ppc64_linux) || defined(VGP_ppc64_aix5)
 #  define GET_REAL_PC_SP_AND_FP(pc, sp, fp)                   \
       asm("mflr 0;"                   /* r0 = lr */           \
-          "bl .m_libcassert_get_ip;"  /* lr = pc */           \
-          ".m_libcassert_get_ip:\n"                           \
+          "bl 0f;"                    /* lr = pc */           \
+          "0:\n"                                              \
           "mflr %0;"                \
           "mtlr 0;"                   /* restore lr */        \
           "mr %1,1;"                \
