diff -Nurp libvirt-0.3.2.orig/src/qemu_conf.c libvirt-0.3.2/src/qemu_conf.c
--- libvirt-0.3.2.orig/src/qemu_conf.c	2007-08-14 07:58:57.000000000 +0200
+++ libvirt-0.3.2/src/qemu_conf.c	2007-09-04 00:14:42.000000000 +0200
@@ -293,7 +293,7 @@ static char *qemudLocateBinaryForArch(vi
     char *path;
 
     if (virtType == QEMUD_VIRT_KVM)
-        name = "qemu-kvm";
+        name = "kvm";
     else
         name = qemudDefaultBinaryForArch(arch);
 
diff -Nurp libvirt-0.3.2.orig/src/qemu_driver.c libvirt-0.3.2/src/qemu_driver.c
--- libvirt-0.3.2.orig/src/qemu_driver.c	2007-08-21 13:42:30.000000000 +0200
+++ libvirt-0.3.2/src/qemu_driver.c	2007-09-04 00:14:56.000000000 +0200
@@ -1525,7 +1525,7 @@ static char *qemudGetCapabilities(virCon
             r = virBufferAdd (xml,
                            "\
       <domain type=\"kvm\">\n\
-        <emulator>/usr/bin/qemu-kvm</emulator>\n\
+        <emulator>/usr/bin/kvm</emulator>\n\
       </domain>\n", -1);
             if (r == -1) goto vir_buffer_failed;
         }
