diff -Nurp libvirt-0.3.2.orig/tests/Makefile libvirt-0.3.2/tests/Makefile
--- libvirt-0.3.2.orig/tests/Makefile	2007-09-04 00:15:46.000000000 +0200
+++ libvirt-0.3.2/tests/Makefile	2007-09-04 00:18:28.000000000 +0200
@@ -38,7 +38,7 @@ build_triplet = x86_64-pc-linux-gnu
 host_triplet = x86_64-pc-linux-gnu
 noinst_PROGRAMS = xmlrpctest$(EXEEXT) xml2sexprtest$(EXEEXT) \
 	sexpr2xmltest$(EXEEXT) virshtest$(EXEEXT) conftest$(EXEEXT) \
-	reconnect$(EXEEXT) xmconfigtest$(EXEEXT) xencapstest$(EXEEXT) \
+	reconnect$(EXEEXT) xmconfigtest$(EXEEXT) \
 	qemuxml2argvtest$(EXEEXT) qemuxml2xmltest$(EXEEXT) \
 	nodeinfotest$(EXEEXT)
 #am__append_1 = reconnect
@@ -112,13 +112,13 @@ LINK = $(LIBTOOL) --tag=CC --mode=link $
 SOURCES = $(conftest_SOURCES) $(nodeinfotest_SOURCES) \
 	$(qemuxml2argvtest_SOURCES) $(qemuxml2xmltest_SOURCES) \
 	$(reconnect_SOURCES) $(sexpr2xmltest_SOURCES) \
-	$(virshtest_SOURCES) $(xencapstest_SOURCES) \
+	$(virshtest_SOURCES) \
 	$(xmconfigtest_SOURCES) $(xml2sexprtest_SOURCES) \
 	$(xmlrpctest_SOURCES)
 DIST_SOURCES = $(conftest_SOURCES) $(nodeinfotest_SOURCES) \
 	$(qemuxml2argvtest_SOURCES) $(qemuxml2xmltest_SOURCES) \
 	$(reconnect_SOURCES) $(sexpr2xmltest_SOURCES) \
-	$(virshtest_SOURCES) $(xencapstest_SOURCES) \
+	$(virshtest_SOURCES) \
 	$(xmconfigtest_SOURCES) $(xml2sexprtest_SOURCES) \
 	$(xmlrpctest_SOURCES)
 RECURSIVE_TARGETS = all-recursive check-recursive dvi-recursive \
@@ -282,7 +282,7 @@ sbindir = ${exec_prefix}/sbin
 sharedstatedir = ${prefix}/com
 sysconfdir = /etc
 target_alias = 
-SUBDIRS = virshdata confdata sexpr2xmldata xml2sexprdata xmconfigdata xencapsdata
+SUBDIRS = virshdata confdata sexpr2xmldata xml2sexprdata xmconfigdata
 
 # Wierd libtool related juju...
 #
@@ -312,7 +312,7 @@ LDADDS = \
 
 EXTRA_DIST = xmlrpcserver.py test_conf.sh
 TESTS = xml2sexprtest sexpr2xmltest virshtest test_conf.sh \
-	xmconfigtest xencapstest qemuxml2argvtest qemuxml2xmltest \
+	xmconfigtest qemuxml2argvtest qemuxml2xmltest \
 	nodeinfotest $(am__append_1)
 
 # Note: xmlrpc.[c|h] is not in libvirt yet
