*** libvirt-0.2.0/qemud/Makefile.in.orig	2007-02-14 19:43:17.000000000 +0100
--- libvirt-0.2.0/qemud/Makefile.in	2007-02-14 19:43:56.000000000 +0100
*************** libvirt_qemud_SOURCES = qemud.c internal
*** 233,239 ****
  #-D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED=1 -D_POSIX_C_SOURCE=199506L
  libvirt_qemud_CFLAGS = \
          -I$(top_srcdir)/include -I$(top_builddir)/include $(LIBXML_CFLAGS) \
!         -Werror -Wall -Wextra -DLOCAL_STATE_DIR="\"$(localstatedir)\"" \
          -DSYSCONF_DIR="\"$(sysconfdir)\""
  
  libvirt_qemud_LDFLAGS = $(LIBXML_LIBS) $(SYSFS_LIBS)
--- 233,239 ----
  #-D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED=1 -D_POSIX_C_SOURCE=199506L
  libvirt_qemud_CFLAGS = \
          -I$(top_srcdir)/include -I$(top_builddir)/include $(LIBXML_CFLAGS) \
!         -Wall -Wextra -DLOCAL_STATE_DIR="\"$(localstatedir)\"" \
          -DSYSCONF_DIR="\"$(sysconfdir)\""
  
  libvirt_qemud_LDFLAGS = $(LIBXML_LIBS) $(SYSFS_LIBS)
*** libvirt-0.2.0/qemud/Makefile.am.orig	2007-02-14 19:43:09.000000000 +0100
--- libvirt-0.2.0/qemud/Makefile.am	2007-02-14 19:44:02.000000000 +0100
*************** libvirt_qemud_SOURCES = qemud.c internal
*** 13,19 ****
  #-D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED=1 -D_POSIX_C_SOURCE=199506L
  libvirt_qemud_CFLAGS = \
          -I$(top_srcdir)/include -I$(top_builddir)/include $(LIBXML_CFLAGS) \
!         -Werror -Wall -Wextra -DLOCAL_STATE_DIR="\"$(localstatedir)\"" \
          -DSYSCONF_DIR="\"$(sysconfdir)\""
  libvirt_qemud_LDFLAGS = $(LIBXML_LIBS) $(SYSFS_LIBS)
  libvirt_qemud_DEPENDENCIES =
--- 13,19 ----
  #-D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED=1 -D_POSIX_C_SOURCE=199506L
  libvirt_qemud_CFLAGS = \
          -I$(top_srcdir)/include -I$(top_builddir)/include $(LIBXML_CFLAGS) \
!         -Wall -Wextra -DLOCAL_STATE_DIR="\"$(localstatedir)\"" \
          -DSYSCONF_DIR="\"$(sysconfdir)\""
  libvirt_qemud_LDFLAGS = $(LIBXML_LIBS) $(SYSFS_LIBS)
  libvirt_qemud_DEPENDENCIES =
