--- tests/Makefile.old	2007-05-05 15:30:12.000000000 +0200
+++ tests/Makefile	2007-05-05 15:33:36.000000000 +0200
@@ -38,7 +38,7 @@ build_triplet = i686-pc-linux-gnu
 host_triplet = i686-pc-linux-gnu
 noinst_PROGRAMS = xmlrpctest$(EXEEXT) xml2sexprtest$(EXEEXT) \
 	sexpr2xmltest$(EXEEXT) virshtest$(EXEEXT) conftest$(EXEEXT) \
-	reconnect$(EXEEXT) xmconfigtest$(EXEEXT) xencapstest$(EXEEXT)
+	reconnect$(EXEEXT) xmconfigtest$(EXEEXT)
 #am__append_1 = reconnect
 subdir = tests
 DIST_COMMON = $(srcdir)/Makefile.am $(srcdir)/Makefile.in
@@ -97,11 +97,11 @@ LINK = $(LIBTOOL) --tag=CC --mode=link $
 	$(AM_LDFLAGS) $(LDFLAGS) -o $@
 SOURCES = $(conftest_SOURCES) $(reconnect_SOURCES) \
 	$(sexpr2xmltest_SOURCES) $(virshtest_SOURCES) \
-	$(xencapstest_SOURCES) $(xmconfigtest_SOURCES) \
+	$(xmconfigtest_SOURCES) \
 	$(xml2sexprtest_SOURCES) $(xmlrpctest_SOURCES)
 DIST_SOURCES = $(conftest_SOURCES) $(reconnect_SOURCES) \
 	$(sexpr2xmltest_SOURCES) $(virshtest_SOURCES) \
-	$(xencapstest_SOURCES) $(xmconfigtest_SOURCES) \
+	$(xmconfigtest_SOURCES) \
 	$(xml2sexprtest_SOURCES) $(xmlrpctest_SOURCES)
 RECURSIVE_TARGETS = all-recursive check-recursive dvi-recursive \
 	html-recursive info-recursive install-data-recursive \
@@ -265,7 +265,7 @@ sbindir = ${exec_prefix}/sbin
 sharedstatedir = ${prefix}/com
 sysconfdir = ${prefix}/etc
 target_alias = 
-SUBDIRS = virshdata confdata sexpr2xmldata xml2sexprdata xmconfigdata xencapsdata
+SUBDIRS = virshdata confdata sexpr2xmldata xml2sexprdata xmconfigdata
 
 # Wierd libtool related juju...
 #
@@ -294,7 +294,7 @@ LDADDS = \
 
 EXTRA_DIST = xmlrpcserver.py test_conf.sh
 TESTS = xml2sexprtest sexpr2xmltest virshtest test_conf.sh \
-	xmconfigtest xencapstest $(am__append_1)
+	xmconfigtest $(am__append_1)
 
 # Note: xmlrpc.[c|h] is not in libvirt yet
 xmlrpctest_SOURCES = \
