diff -rup libvirt-0.2.0-orig/qemud/conf.c libvirt-0.2.0-new/qemud/conf.c
--- libvirt-0.2.0-orig/qemud/conf.c	2007-02-16 16:54:39.000000000 -0500
+++ libvirt-0.2.0-new/qemud/conf.c	2007-02-16 16:55:14.000000000 -0500
@@ -542,8 +542,6 @@ static struct qemud_vm_def *qemudParseXM
 
     if (!strcmp((char *)prop, "qemu"))
         def->virtType = QEMUD_VIRT_QEMU;
-    else if (!strcmp((char *)prop, "kqemu"))
-        def->virtType = QEMUD_VIRT_KQEMU;
     else if (!strcmp((char *)prop, "kvm"))
         def->virtType = QEMUD_VIRT_KVM;
     else {
@@ -938,7 +936,6 @@ int qemudBuildCommandLine(struct qemud_s
 
     len = 1 + /* qemu */
         2 + /* machine type */
-        (vm->def->virtType == QEMUD_VIRT_QEMU ? 1 : 0) + /* Disable kqemu */
         2 * vm->def->ndisks + /* disks*/
         (vm->def->nnets > 0 ? (4 * vm->def->nnets) : 2) + /* networks */
         2 + /* memory*/
@@ -963,10 +960,6 @@ int qemudBuildCommandLine(struct qemud_s
         goto no_memory;
     if (!((*argv)[++n] = strdup(vm->def->os.machine)))
         goto no_memory;
-    if (vm->def->virtType == QEMUD_VIRT_QEMU) {
-        if (!((*argv)[++n] = strdup("-no-kqemu")))
-        goto no_memory;
-    }
     if (!((*argv)[++n] = strdup("-m")))
         goto no_memory;
     if (!((*argv)[++n] = strdup(memory)))
@@ -1713,9 +1706,6 @@ char *qemudGenerateXML(struct qemud_serv
     case QEMUD_VIRT_QEMU:
         type = "qemu";
         break;
-    case QEMUD_VIRT_KQEMU:
-        type = "kqemu";
-        break;
     case QEMUD_VIRT_KVM:
         type = "kvm";
         break;
Only in libvirt-0.2.0-new/qemud: conf.c~
