diff -Nurp libvirt-0.2.2.orig/qemud/conf.c libvirt-0.2.2/qemud/conf.c
--- libvirt-0.2.2.orig/qemud/conf.c	2007-04-17 10:34:42.000000000 +0200
+++ libvirt-0.2.2/qemud/conf.c	2007-05-06 14:44:57.000000000 +0200
@@ -190,7 +190,7 @@ static char *qemudLocateBinaryForArch(st
     char *path;
 
     if (virtType == QEMUD_VIRT_KVM)
-        name = "qemu-kvm";
+        name = "kvm";
     else
         name = qemudDefaultBinaryForArch(arch);
 
diff -Nurp libvirt-0.2.2.orig/qemud/dispatch.c libvirt-0.2.2/qemud/dispatch.c
--- libvirt-0.2.2.orig/qemud/dispatch.c	2007-03-16 13:52:56.000000000 +0100
+++ libvirt-0.2.2/qemud/dispatch.c	2007-05-06 14:45:07.000000000 +0200
@@ -186,7 +186,7 @@ qemudDispatchGetCapabilities (struct qem
             r = bufferAdd (xml,
                            "\
       <domain type=\"kvm\">\n\
-        <emulator>/usr/bin/qemu-kvm</emulator>\n\
+        <emulator>/usr/bin/kvm</emulator>\n\
       </domain>\n", -1);
             if (r == -1) goto vir_buffer_failed;
         }
