diff --git a/src/xcb_auth.c b/src/xcb_auth.c
index ffc0515..ae87b4a 100644
--- a/src/xcb_auth.c
+++ b/src/xcb_auth.c
@@ -80,7 +80,7 @@ static int authname_match(enum auth_prot
 #define SIN6_ADDR(s) (&((struct sockaddr_in6 *)s)->sin6_addr)
 
 static Xauth *get_authptr(struct sockaddr *sockname, unsigned int socknamelen,
-                          int display)
+                          int display, char *host)
 {
     char *addr = 0;
     int addrlen = 0;
@@ -89,6 +89,7 @@ static Xauth *get_authptr(struct sockadd
     char dispbuf[40];   /* big enough to hold more than 2^64 base 10 */
     int authnamelens[N_AUTH_PROTOS];
     int i;
+    Xauth *authptr = NULL;
 
     family = FamilyLocal; /* 256 */
     switch(sockname->sa_family)
@@ -128,10 +129,20 @@ static Xauth *get_authptr(struct sockadd
 
     for (i = 0; i < N_AUTH_PROTOS; i++)
 	authnamelens[i] = strlen(authnames[i]);
-    return XauGetBestAuthByAddr (family,
-                                 (unsigned short) addrlen, addr,
-                                 (unsigned short) strlen(dispbuf), dispbuf,
-                                 N_AUTH_PROTOS, authnames, authnamelens);
+    authptr = XauGetBestAuthByAddr (family,
+                                    (unsigned short) addrlen, addr,
+                                    (unsigned short) strlen(dispbuf), dispbuf,
+                                    N_AUTH_PROTOS, authnames, authnamelens);
+    // && !phostname
+    if ((!authptr || !authptr->data_length) && !host && family == FamilyLocal) {
+	if ( (addr = getenv("XAUTHLOCALHOSTNAME")) ) {
+	    authptr = XauGetBestAuthByAddr (family,
+                                            (unsigned short) strlen(addr), addr,
+                                            (unsigned short) strlen(dispbuf), dispbuf,
+                                            N_AUTH_PROTOS, authnames, authnamelens);
+	}
+    }
+    return authptr;
 }
 
 #ifdef HASXDMAUTH
@@ -228,7 +239,7 @@ #endif
     return 0;   /* Unknown authorization type */
 }
 
-int _xcb_get_auth_info(int fd, xcb_auth_info_t *info, int display)
+int _xcb_get_auth_info(int fd, xcb_auth_info_t *info, int display, char *host)
 {
     /* code adapted from Xlib/ConnDis.c, xtrans/Xtranssocket.c,
        xtrans/Xtransutils.c */
@@ -241,7 +252,7 @@ int _xcb_get_auth_info(int fd, xcb_auth_
     if (getpeername(fd, sockname, &socknamelen) == -1)
         return 0;  /* can only authenticate sockets */
 
-    authptr = get_authptr(sockname, socknamelen, display);
+    authptr = get_authptr(sockname, socknamelen, display, host);
     if (authptr == 0)
         return 0;   /* cannot find good auth data */
 
diff --git a/src/xcb_util.c b/src/xcb_util.c
index eeee1dd..156aa9a 100644
--- a/src/xcb_util.c
+++ b/src/xcb_util.c
@@ -241,7 +241,7 @@ xcb_connection_t *xcb_connect(const char
     if(fd == -1)
         return (xcb_connection_t *) &error_connection;
 
-    if(_xcb_get_auth_info(fd, &auth, display))
+    if(_xcb_get_auth_info(fd, &auth, display, host))
     {
         c = xcb_connect_to_fd(fd, &auth);
         free(auth.name);
diff --git a/src/xcbint.h b/src/xcbint.h
index a8e167c..c7ee375 100644
--- a/src/xcbint.h
+++ b/src/xcbint.h
@@ -187,7 +187,7 @@ int _xcb_conn_wait(xcb_connection_t *c, 
 
 /* xcb_auth.c */
 
-int _xcb_get_auth_info(int fd, xcb_auth_info_t *info, int display);
+int _xcb_get_auth_info(int fd, xcb_auth_info_t *info, int display, char *host);
 
 #ifdef GCC_HAS_VISIBILITY
 #pragma GCC visibility pop

