--- aegisub-2.1.8.orig/configure.in	2010-01-27 22:47:46.000000000 +0100
+++ aegisub-2.1.8/configure.in	2010-06-13 22:26:39.421120926 +0200
@@ -500,22 +500,7 @@
   PKG_CHECK_MODULES(PORTAUDIO, portaudio-2.0 >= portaudio_required_version, [with_portaudio="yes"], [with_portaudio="no"])
 fi
 
-if test "$with_portaudio" != "no"; then
-  AC_AGI_COMPILE([PortAudio], [portaudio], [$PORTAUDIO_CFLAGS], [$PORTAUDIO_LIBS],[
-#include <portaudio.h>
-int main(void) {
-  PaError err = Pa_Initialize();
-  if (err != paNoError) return 1;
-  return 0;
-} ])
-fi
-
-if test "$agi_cv_with_portaudio" = "no" && test "$with_portaudio" = "yes"; then
-  AC_MSG_WARN([PortAudio detected, but it doesn't work...])
-  with_portaudio="no"
-fi
-
-if test "$agi_cv_with_portaudio" = "yes" && test "$with_portaudio" = "yes"; then
+if test "$with_portaudio" = "yes"; then
   AC_DEFINE(WITH_PORTAUDIO, 1, [Enable PortAudio v19 support])
   found_audio_player="yes"
 fi
@@ -533,22 +518,7 @@
   PKG_CHECK_MODULES(LIBPULSE, libpulse >= pulseaudio_required_version, [with_pulseaudio="yes"], [with_pulseaudio="no"])
 fi
 
-if test "$with_pulseaudio" != "no"; then
-  AC_AGI_COMPILE([PuleseAudio], [pulseaudio], [$LIBPULSE_CFLAGS], [$LIBPULSE_LIBS],[
-#include <pulse/pulseaudio.h>
-int main(void) {
-pa_threaded_mainloop *mainloop = pa_threaded_mainloop_new();
-if (!mainloop) return 1;
-return 0;
-} ])
-fi
-
-if test "$agi_cv_with_pulseaudio" = "no" && test "$with_pulseaudio" = "yes"; then
-  AC_MSG_WARN([PulseAudio detected, but it doesn't work...])
-  with_pulseaudio="no"
-fi
-
-if test "$agi_cv_with_pulseaudio" = "yes" && test "$with_pulseaudio" = "yes"; then
+if test "$with_pulseaudio" = "yes"; then
   AC_DEFINE(WITH_PULSEAUDIO, 1, [Enable PulseAudio support])
   found_audio_player="yes"
 fi
@@ -565,30 +535,7 @@
   PKG_CHECK_MODULES(OPENAL, openal >= 0.0.8, [with_openal=yes], [with_openal=no])
 fi
 
-if test "$with_openal" != "no"; then
-  AC_AGI_COMPILE([OpenAL], [openal], [$OPENAL_CFLAGS], [$OPENAL_LIBS],[
-#if defined(__APPLE__)
-#include <OpenAL/AL.h>
-#include <OpenAL/ALC.h>
-#else
-#include <AL/al.h>
-#include <AL/alc.h>
-#endif
-int main(void) {
-  ALCdevice *device = alcOpenDevice(0);
-  if (!device) return 1;
-  ALCcontext *context = alcCreateContext(device, 0);
-  if (!context) return 1;
-  return 0;
-} ])
-fi
-
-if test "$agi_cv_with_openal" = "no" && test "$with_openal" = "yes"; then
-  AC_MSG_WARN([OpenAL detected, but it doesn't work...])
-  with_openal="no"
-fi
-
-if test "$agi_cv_with_openal" = "yes" && test "$with_openal" = "yes"; then
+if test "$with_openal" = "yes"; then
   found_audio_player="yes"
   AC_DEFINE(WITH_OPENAL, 1, [Enable OpenAL support])
 fi
