*** UrbanTerror/q3ut4/autoexec.cfg	Sat Jan 27 13:18:18 2007
--- UrbanTerror/q3ut4/autoexecmy.cfg	Sat Dec 19 11:34:43 2009
***************
*** 1,5 ****
! //In this file you can do settings that will be executed everytime Urban Terror is started. It will not get overwritten.
! //Lines starting with // are ignored.
! //Example1: exec yourconfig.cfg
! //Example2: set cg_fov "110"
! //Example3: bind x ut_weaptoggle knife
\ No newline at end of file
--- 1,6 ----
! //In this file you can do settings that will be executed everytime Urban Terror is started. It will not get overwritten.
! //Lines starting with // are ignored.
! //Example1: exec yourconfig.cfg
! //Example2: set cg_fov "110"
! //Example3: bind x ut_weaptoggle knife
! seta cl_cURLLib "/usr/lib/libcurl.so"
