--- openarena-engine-0.8.1/Makefile	2008-08-09 22:22:37.000000000 +0200
+++ openarena-engine-0.8.1/Makefile.new	2009-01-07 19:33:31.000000000 +0100
@@ -231,8 +231,7 @@
   endif
   endif
 
-  BASE_CFLAGS = -Wall -fno-strict-aliasing -Wimplicit -Wstrict-prototypes \
-    -pipe -DUSE_ICON $(shell sdl-config --cflags)
+  BASE_CFLAGS = %CFLAGS% -DUSE_ICON $(shell sdl-config --cflags)
 
   ifeq ($(USE_OPENAL),1)
     BASE_CFLAGS += -DUSE_OPENAL
@@ -323,7 +322,7 @@
   endif
 
   DEBUG_CFLAGS = $(BASE_CFLAGS) -g -O0
-  RELEASE_CFLAGS=$(BASE_CFLAGS) -DNDEBUG $(OPTIMIZE)
+  RELEASE_CFLAGS=$(BASE_CFLAGS) -DNDEBUG
 
 else # ifeq Linux
 
