diff --git a/core/libs/Makefile.am b/core/libs/Makefile.am
index d070b29..c0a9258 100644
--- a/core/libs/Makefile.am
+++ b/core/libs/Makefile.am
@@ -53,7 +53,7 @@ libfreetype.a: klibc_compat.h
 	if ! $(MAKE) --question $(AM_MAKEFLAGS) CFLAGS="$(K_CFLAGS)" library ; then \
 		$(call infmsg,MAKE,libfreetype.a) ; \
 	fi ; \
-	$(MAKE) $(AM_MAKEFLAGS) CFLAGS="-c $(K_CFLAGS)" CCexe="$(BUILD_CC)" library $(OUTPUT)
+	$(MAKE) $(AM_MAKEFLAGS) CFLAGS="-c $(K_CFLAGS)" CCexe="$(BUILD_CC)" ANSIFLAGS="" library $(OUTPUT)
 	$(Q)$(MKDIR_P) $(@D)
 	$(Q)cp -f libfreetype/.libs/libfreetype.a $@
 
@@ -62,18 +62,18 @@ libjpeg.a:
 	$(Q)cd libjpeg ; \
 	if test ! -e ./Makefile ; then \
 		$(call infmsg,CONF,libjpeg.a) ; \
-		CC=$(KLCC) CFLAGS="$(K_CFLAGS) -I$(abs_top_builddir)/libs/libjpeg" $(LIBJPEG_SOURCE)/configure \
+		CC=$(KLCC) CFLAGS="$(K_CFLAGS) -I$(abs_top_builddir)/libs/libjpeg -include $(abs_top_srcdir)/libs/klibc_compat.h" $(LIBJPEG_SOURCE)/configure \
 		--host=$(host) --build=$(build_triplet) $(OUTPUT) ; \
 		echo "#define NO_GETENV 1" >> jconfig.h ; \
 		$(SED) -i "s#CFLAGS=.*#CFLAGS=$(K_CFLAGS) -I$(abs_top_builddir)/libs/libjpeg#" Makefile ; \
 		$(SED) -i "s:#define JMESSAGE.*::" $(LIBJPEG_SOURCE)/jerror.c ; \
 	fi ; \
-	if ! $(MAKE) --question $(AM_MAKEFLAGS) CC="$(KLCC)" libjpeg.a ; then  \
+	if ! $(MAKE) --question $(AM_MAKEFLAGS) CC="$(KLCC)" libjpeg.la ; then  \
 		$(call infmsg,MAKE,libjpeg.a) ; \
 	fi ; \
-	$(MAKE) $(AM_MAKEFLAGS) CC="$(KLCC)" libjpeg.a $(OUTPUT)
+	$(MAKE) $(AM_MAKEFLAGS) CC="$(KLCC)" libjpeg.la $(OUTPUT)
 	$(Q)$(MKDIR_P) $(@D)
-	$(Q)cp -f libjpeg/libjpeg.a $@
+	$(Q)cp -f libjpeg/.libs/libjpeg.a $@
 
 libpng.a: libz.a
 	$(Q)cd $(LIBPNG_SOURCE) ; \
