commit 7cf5fd02bbc6422ddbc9af7185cc47657148deb0
Author: Michał Januszewski <spock@gentoo.org>
Date:   Sun Apr 4 16:19:40 2010 +0200

    gentoo: only enable the effects that are actually selected.

diff --git a/gentoo/splash.c b/gentoo/splash.c
index 4b9a5b9..5220571 100644
--- a/gentoo/splash.c
+++ b/gentoo/splash.c
@@ -660,7 +660,7 @@ static int splash_start(const char *runlevel)
 			 (config->type == fbspl_reboot) ? "reboot" : ((config->type == fbspl_shutdown) ? "shutdown" : "bootup"),
 			 (config->kdmode == KD_GRAPHICS) ? "--kdgraphics" : "",
 			 (config->textbox_visible) ? "--textbox" : "",
-			 (config->effects & (FBSPL_EFF_FADEOUT | FBSPL_EFF_FADEIN)) ? "--effects=fadeout,fadein" :
+			 ((config->effects & (FBSPL_EFF_FADEOUT | FBSPL_EFF_FADEIN)) == (FBSPL_EFF_FADEOUT | FBSPL_EFF_FADEIN)) ? "--effects=fadeout,fadein" :
 				 ((config->effects & FBSPL_EFF_FADEOUT) ? "--effects=fadeout" :
 					 ((config->effects & FBSPL_EFF_FADEIN) ? "--effects=fadein" : "")));
 
@@ -716,11 +716,7 @@ static int splash_stop(const char *runlevel)
 
 	/* If we don't get a runlevel argument, then we're being executed
 	 * because of a rc-abort event and we don't save any data. */
-	if (runlevel == NULL) {
-		return fbsplash_cache_cleanup(NULL);
-	} else {
-		return fbsplash_cache_cleanup(save);
-	}
+	return fbsplash_cache_cleanup(save);
 }
 
 int rc_plugin_hook(RC_HOOK hook, const char *name)

commit d711147fc835cf408a9a4e7432fc565a6b952183
Author: Michał Januszewski <spock@gentoo.org>
Date:   Sun Apr 4 16:57:23 2010 +0200

    gentoo: add some comments in splash.conf

diff --git a/gentoo/splash.conf b/gentoo/splash.conf
index 4cf0a32..f056382 100644
--- a/gentoo/splash.conf
+++ b/gentoo/splash.conf
@@ -21,7 +21,10 @@ SPLASH_TEXTBOX="no"
 SPLASH_AUTOVERBOSE="0"
 
 # Name of the boot service starting X.  Useful in cases when X is started
-# by a custom initscript, other than the usual 'xdm'.
+# by a custom initscript, other than the usual 'xdm'.  If you are using a
+# custom X initscript and do not set this variable correctly, do not be
+# surprised if you end up at tty1 at the end of the boot process (instead of
+# in an X session).
 # SPLASH_XSERVICE="xdm"
 
 #########################################################################
