diff --git a/core/src/Makefile.am b/core/src/Makefile.am
index 20cb03e..fa6919d 100644
--- a/core/src/Makefile.am
+++ b/core/src/Makefile.am
@@ -119,7 +119,7 @@ libfbsplashrender_la_SOURCES  = \
 	fbsplash.h
 libfbsplashrender_la_CFLAGS   = $(AM_CFLAGS)
 libfbsplashrender_la_LDFLAGS  = $(AM_LDFLAGS) -version-info $(libfbsplashrender_version)
-libfbsplashrender_la_LIBADD   = -lfbsplash
+libfbsplashrender_la_LIBADD   = libfbsplash.la
 
 libfbsplashrender_la_CFLAGS  += $(JPEG_CFLAGS)
 libfbsplashrender_la_LIBADD  += $(JPEG_LIBS)
@@ -258,10 +258,10 @@ fbcondecor_helper_LDFLAGS  += -static
 endif
 
 # Override automake's generation of target rules for building fbcondecor_helper.
-fbcondecor_helper: $(fbcondecor_helper_LDADD) $(fbcondecor_helper_OBJECTS)
+fbcondecor_helper: $(fbcondecor_helper_LDADD) $(fbcondecor_helper_OBJECTS) $(lib_LTLIBRARIES)
 	@$(call infmsg,LD,$@)
 	$(Q)$(MKDIR_P) $(@D)
-	$(Q)$(KLCC) $(fbcondecor_helper_CFLAGS) $(fbcondecor_helper_LDFLAGS) $+ $(fbcondecor_helper_LDADD) -o $@
+	$(Q)$(KLCC) $(fbcondecor_helper_CFLAGS) $(fbcondecor_helper_LDFLAGS) $(fbcondecor_helper_OBJECTS) $(fbcondecor_helper_LDADD) -o $@
 	$(Q)$(STRIP) $@
 
 # Override automake's generation of target rules for building fbcondecor_helper object files.

