--- linux-2.6.32-openvz-039.11.orig//arch/um/Makefile-i386
+++ linux-2.6.32-openvz-039.11/arch/um/Makefile-i386
@@ -4,7 +4,7 @@ TOP_ADDR := $(CONFIG_TOP_ADDR)
 
 START := 0x8048000
 
-LDFLAGS			+= -m elf_i386
+LDFLAGS			+= -m32
 ELF_ARCH		:= $(SUBARCH)
 ELF_FORMAT 		:= elf32-$(SUBARCH)
 OBJCOPYFLAGS  		:= -O binary -R .note -R .comment -S
--- linux-2.6.32-openvz-039.11.orig//arch/um/Makefile-x86_64
+++ linux-2.6.32-openvz-039.11/arch/um/Makefile-x86_64
@@ -10,7 +10,7 @@ KBUILD_CFLAGS += $(_extra_flags_)
 
 CHECKFLAGS  += -m64 -D__x86_64__
 KBUILD_AFLAGS += -m64
-LDFLAGS += -m elf_x86_64
+LDFLAGS += -m64
 KBUILD_CPPFLAGS += -m64
 
 ELF_ARCH := i386:x86-64
--- linux-2.6.32-openvz-039.11.orig//arch/x86/vdso/Makefile
+++ linux-2.6.32-openvz-039.11/arch/x86/vdso/Makefile
@@ -26,7 +26,7 @@ targets += vdso-rhel5.so vdso.so vdso-rh
 
 export CPPFLAGS_vdso.lds += -P -C
 
-VDSO_LDFLAGS_vdso.lds = -m elf_x86_64 -Wl,-soname=linux-vdso.so.1 \
+VDSO_LDFLAGS_vdso.lds = -m64 -Wl,-soname=linux-vdso.so.1 \
 		      	-Wl,-z,max-page-size=4096 -Wl,-z,common-page-size=4096
 
 VDSO_LDFLAGS_vdso-rhel5.lds = ${VDSO_LDFLAGS_vdso.lds}
@@ -77,7 +77,7 @@ vdso32.so-$(VDSO32-y)		+= sysenter
 vdso32-images			= $(vdso32.so-y:%=vdso32-%.so)
 
 CPPFLAGS_vdso32.lds = $(CPPFLAGS_vdso.lds)
-VDSO_LDFLAGS_vdso32.lds = -m elf_i386 -Wl,-soname=linux-gate.so.1
+VDSO_LDFLAGS_vdso32.lds = -m32 -Wl,-soname=linux-gate.so.1
 
 # This makes sure the $(obj) subdirectory exists even though vdso32/
 # is not a kbuild sub-make subdirectory.
--- linux-2.6.32-openvz-039.11.orig//scripts/recordmcount.pl
+++ linux-2.6.32-openvz-039.11/scripts/recordmcount.pl
@@ -170,7 +170,7 @@ if ($arch eq "x86_64") {
     $alignment = 8;
 
     # force flags for this arch
-    $ld .= " -m elf_x86_64";
+    $ld .= " -m64";
     $objdump .= " -M x86-64";
     $objcopy .= " -O elf64-x86-64";
     $cc .= " -m64";
@@ -179,7 +179,7 @@ if ($arch eq "x86_64") {
     $alignment = 4;
 
     # force flags for this arch
-    $ld .= " -m elf_i386";
+    $ld .= " -m64";
     $objdump .= " -M i386";
     $objcopy .= " -O elf32-i386";
     $cc .= " -m32";
