diff -urB a/utils/nfsd/nfssvc.c b/utils/nfsd/nfssvc.c
--- a/utils/nfsd/nfssvc.c	2009-11-04 11:13:56.000000000 +0000
+++ b/utils/nfsd/nfssvc.c	2009-12-08 23:09:35.227072145 +0000
@@ -180,7 +180,7 @@
 			goto error;
 		}
 
-		snprintf(buf, sizeof(buf), "%d\n", sockfd); 
+		snprintf(buf, sizeof(buf), "%d\n", sockfd);
 		if (write(fd, buf, strlen(buf)) != strlen(buf)) {
 			/*
 			 * this error may be common on older kernels that don't
@@ -212,7 +212,7 @@
 nfssvc_set_sockets(const int family, const unsigned int protobits,
 		   const char *host, const char *port)
 {
-	struct addrinfo hints = { .ai_flags = AI_PASSIVE | AI_ADDRCONFIG };
+	struct addrinfo hints = { .ai_flags = AI_PASSIVE };
 
 	hints.ai_family = family;
 
