diff -Nurp kdm-4.2.2.orig/kdm/config.def kdm-4.2.2/kdm/config.def
--- kdm-4.2.2.orig/kdm/config.def	2009-04-22 14:37:55.757715074 +0200
+++ kdm-4.2.2/kdm/config.def	2009-04-22 14:39:17.349814823 +0200
@@ -1243,10 +1243,10 @@ Type: enum
  None/BO_NONE: no boot manager
  Grub/BO_GRUB: Grub boot manager
  Lilo/BO_LILO: Lilo boot manager (Linux on i386 &amp; x86-64 only)
-Default: None
+Default: Grub
 User: core
 User: greeter
-Instance: #Grub
+Instance: Grub
 Merge: kdm:UseLilo(P_UseLilo)
 Comment: &
 Description:
