diff -Nurp kdm-4.2.1.orig/kdm/config.def kdm-4.2.1/kdm/config.def
--- kdm-4.2.1.orig/kdm/config.def	2009-02-26 15:16:59.000000000 +0100
+++ kdm-4.2.1/kdm/config.def	2009-03-24 11:17:03.407884095 +0100
@@ -2657,7 +2657,7 @@ Key: Theme
 Type: string
 Default: ""
 User: greeter
-Instance: */KDMDATA "/themes/oxygen"
+Instance: */KDMDATA "/themes/Glassified"
 Comment: &
 Description:
  The theme to use for the greeter. Can point to either a directory or an XML
