diff -Nurp kdm-4.2.2.orig/kdm/kfrontend/genkdmconf.c kdm-4.2.2/kdm/kfrontend/genkdmconf.c
--- kdm-4.2.2.orig/kdm/kfrontend/genkdmconf.c	2008-12-04 09:34:57.000000000 +0100
+++ kdm-4.2.2/kdm/kfrontend/genkdmconf.c	2009-04-13 21:13:59.178952600 +0200
@@ -722,7 +722,7 @@ static const char def_background[] =
 "Program=\n"
 "ReverseBlending=false\n"
 "UseSHM=false\n"
-"Wallpaper=default_blue.jpg\n"
+"Wallpaper=/usr/share/backgrounds/sabayonlinux.png\n"
 "WallpaperList=\n"
 "WallpaperMode=Scaled\n";
 
