diff -Nurp kdm-3.5.7.orig/kdm/kfrontend/genkdmconf.c kdm-3.5.7/kdm/kfrontend/genkdmconf.c
--- kdm-3.5.7.orig/kdm/kfrontend/genkdmconf.c	2007-06-09 12:39:15.000000000 +0200
+++ kdm-3.5.7/kdm/kfrontend/genkdmconf.c	2007-06-09 12:59:35.000000000 +0200
@@ -687,6 +687,15 @@ static const char def_session2[] =
 "[ -f /etc/xprofile ] && . /etc/xprofile\n"
 "[ -f $HOME/.xprofile ] && . $HOME/.xprofile\n"
 "\n"
+"# launches a session dbus instance\n"
+"\n"
+"dbuslaunch=\"`which dbus-launch 2>/dev/null`\"\n"
+"if [ -n \"$dbuslaunch\" ] && [ -x \"$dbuslaunch\" ] && [ -z \"$DBUS_SESSION_BUS_ADDRESS\" ]; then\n"
+"  eval `$dbuslaunch --sh-syntax --exit-with-session`\n"
+"  export DBUS_SESSION_BUS_ADDRESS\n"
+"  export DBUS_SESSION_BUS_PID\n"
+"fi\n"
+"\n\n"
 "case $session in\n"
 "  \"\")\n"
 "    exec xmessage -center -buttons OK:0 -default OK \"Sorry, $DESKTOP_SESSION is no valid session.\"\n"
