--- kdm-3.5.6.orig/kdm/kfrontend/genkdmconf.c
+++ kdm-3.5.6/kdm/kfrontend/genkdmconf.c
@@ -579,7 +579,8 @@ static const char def_setup[] =
 "#! /bin/sh\n"
 "# Xsetup - run as root before the login dialog appears\n"
 "\n"
-"#xconsole -geometry 480x130-0-0 -notify -verbose -fn fixed -exitOnFail -file /dev/xconsole &\n";
+"xset r rate 195 35\n"
+"kbdrate -s -r 25\n";
 
 static const char def_startup[] =
 "#! /bin/sh\n"
