diff -Nurp kdm-4.2.4.orig/kdm/config.def kdm-4.2.4/kdm/config.def
--- kdm-4.2.4.orig/kdm/config.def	2009-06-26 16:05:09.975905369 +0200
+++ kdm-4.2.4/kdm/config.def	2009-06-26 17:35:46.683009283 +0200
@@ -1332,9 +1332,9 @@ Description:
 
 Key: ServerTimeout
 Type: int
-Default: 15
+Default: 35
 User: core
-Instance: #:*/
+Instance: :*/35
 Comment:
  How long to wait for a local &X-Server; to come up.
 Description:
