--- kdm/kfrontend/kgapp.cpp	2009/01/06 17:13:38	906699
+++ kdm/kfrontend/kgapp.cpp	2009/02/07 15:59:44	922805
@@ -70,7 +70,7 @@
 
 }
 
-GreeterApp::GreeterApp( int argc, char **argv ) :
+GreeterApp::GreeterApp( int &argc, char **argv ) :
 	inherited( argc, argv ),
 	regrabPtr( false ), regrabKbd( false ),
 	dragWidget( 0 )
@@ -296,13 +296,14 @@
 	}
 
 	static char *fakeArgv[] = { (char *)"kdmgreet", 0 };
+	static int fakeArgc = as(fakeArgv) - 1;
 
 	KCrash::setFlags( KCrash::KeepFDs | KCrash::SaferDialog | KCrash::AlwaysDirectly );
 	KCrash::setApplicationName( QLatin1String( fakeArgv[0] ) );
 	KCrash::setCrashHandler( KCrash::defaultCrashHandler );
 	XSetIOErrorHandler( xIOErr );
 	KComponentData inst( fakeArgv[0] );
-	GreeterApp app( as(fakeArgv) - 1, fakeArgv );
+	GreeterApp app( fakeArgc, fakeArgv );
 	foreach (const QString &dir, KGlobal::dirs()->resourceDirs( "qtplugins" ))
 		app.addLibraryPath( dir );
 	initQAppConfig();
