diff -Nurp kdm-3.5.10.orig/kdm/config.def kdm-3.5.10/kdm/config.def
--- kdm-3.5.10.orig/kdm/config.def	2008-12-17 15:45:29.000000000 +0000
+++ kdm-3.5.10/kdm/config.def	2008-12-17 15:47:57.000000000 +0000
@@ -1863,10 +1863,10 @@ Description:
 
 Key: AutoLoginAgain
 Type: bool
-Default: false
+Default: true
 User: core(autoAgain)
 User: greeter
-Instance: #:0/true
+Instance: :0/true
 Comment: &
 Description:
  If true, auto-login after logout. If false, auto-login is performed only
