diff -Nurp kdm-4.2.2.orig/kdm/config.def kdm-4.2.2/kdm/config.def
--- kdm-4.2.2.orig/kdm/config.def	2009-04-22 14:39:17.349814823 +0200
+++ kdm-4.2.2/kdm/config.def	2009-04-22 14:42:27.453827178 +0200
@@ -1448,9 +1448,9 @@ Description:
 
 Key: TerminateServer
 Type: bool
-Default: false
+Default: true
 User: core
-Instance: #:*/!
+Instance: :*/true
 Merge: xdm
 Comment:
  Restart instead of resetting the local &X-Server; after session exit.
