--- iscsitarget-1.4.20.2.orig/Makefile
+++ iscsitarget-1.4.20.2/Makefile
@@ -37,6 +37,10 @@ KMIN := $(shell echo $(KVER) | \
 	sed -e 's/^[0-9][0-9]*\.\([0-9][0-9]*\)\.[0-9][0-9]*.*/\1/')
 KREV := $(shell echo $(KVER) | \
 	sed -e 's/^[0-9][0-9]*\.[0-9][0-9]*\.\([0-9][0-9]*\).*/\1/')
+# Linux 3.0 support
+ifeq ($(KREV),0)
+	KREV := 00
+endif
 
 kver_eq = $(shell [ $(KMAJ)$(KMIN)$(KREV) -eq $(1)$(2)$(3) ] && \
 	echo 1 || echo 0)
