Index: iscsitarget-1.4.19/kernel/target.c
===================================================================
--- iscsitarget-1.4.19.orig/kernel/target.c	2009-09-28 21:08:20.000000000 +0200
+++ iscsitarget-1.4.19/kernel/target.c	2010-12-08 17:02:59.247843222 +0100
@@ -7,11 +7,12 @@
 #include "iscsi.h"
 #include "digest.h"
 #include "iscsi_dbg.h"
+#include <linux/semaphore.h>
 
 #define	MAX_NR_TARGETS	(1UL << 30)
 
 static LIST_HEAD(target_list);
-static DECLARE_MUTEX(target_list_sem);
+static DEFINE_SEMAPHORE(target_list_sem);
 static u32 next_target_id;
 static u32 nr_targets;
 
@@ -157,7 +158,7 @@ static int iscsi_target_create(struct ta
 
 	strncpy(target->name, name, sizeof(target->name) - 1);
 
-	init_MUTEX(&target->target_sem);
+	sema_init(&target->target_sem, 0);
 	spin_lock_init(&target->session_list_lock);
 
 	INIT_LIST_HEAD(&target->session_list);
