diff -Nurp Miro-0.9.9.1.orig/platform/gtk-x11/miro.real Miro-0.9.9.1/platform/gtk-x11/miro.real
--- Miro-0.9.9.1.orig/platform/gtk-x11/miro.real	2007-09-06 00:02:44.000000000 +0200
+++ Miro-0.9.9.1/platform/gtk-x11/miro.real	2007-09-16 11:13:42.000000000 +0200
@@ -78,29 +78,4 @@ elif options.profile:
     eventloop.profile_file = options.profile
     profile.run ('main()', options.profile)
 else:
-    try:
-        import dbus
-        try:
-            import dbus_bindings
-        except ImportError, main_error:
-            try:
-                dbus_bindings = dbus.dbus_bindings
-            except:
-                raise main_error
-        import onetime
-        try:
-            onetime.OneTime()
-            startapp()
-        except onetime.NameExistsException:
-            bus = dbus.SessionBus()
-            proxy_obj = bus.get_object('org.participatoryculture.dtv.onetime', '/org/participatoryculture/dtv/OneTime')
-            iface = dbus.Interface(proxy_obj, 'org.participatoryculture.dtv.OneTimeIface')
-            for i in xrange(len(args)):
-                args[i] = args[i].decode('latin1')
-            dbusargs = dbus.Array(args, signature="s")
-    
-            iface.HandleArgs(dbusargs)
-        except dbus_bindings.DBusException:
-            startapp()
-    except ImportError:
-        startapp()
+    startapp()
