--- Miro-0.9.9.1.orig/platform/gtk-x11/setup.py	2007-09-16 00:06:13.000000000 -0300
+++ Miro-0.9.9.1/platform/gtk-x11/setup.py	2007-09-16 00:07:55.000000000 -0300
@@ -285,7 +285,7 @@
 
 #### Xine Extension ####
 xine_options = parsePkgConfig('pkg-config', 
-        'libxine pygtk-2.0 gtk+-2.0 glib-2.0 gthread-2.0')
+        'libxine pygtk-2.0 gtk+-2.0 glib-2.0 gthread-2.0 x11')
 
 # If you get XINE crashes, uncommenting this might fix it. It's
 # necessary on Debian Etch and Ubuntu Feisty right now.

