--- hal-info-20070618/fdi/preprobe/10osvendor/10-ide-drives.fdi	2007-03-01 03:32:53.000000000 +0100
+++ hal-info-20070618.new/fdi/preprobe/10osvendor/10-ide-drives.fdi	2007-08-06 08:28:54.000000000 +0200
@@ -3,6 +3,19 @@
 <deviceinfo version="0.2">
 
   <device>
+    <!-- Handle weird Dell Laptops that slow down the system if the optical 
+        drive is polled (RH Bug #138148) -->
+    <match key="storage.bus" string="ide">
+      <match key="storage.model" string="HL-DT-STCD-RW/DVD-ROM GCC-4240N">
+       <match key="block.device" string="/dev/hdb">
+          <merge key="storage.media_check_enabled" type="bool">false</merge>
+       </match>
+       <match key="block.device" string="/dev/hdd">
+          <merge key="storage.media_check_enabled" type="bool">false</merge>
+       </match>
+      </match>
+    </match>
+
     <!-- Handle weird IDE drives (Jaz and Zip) by matching on the IDE model -->
     <match key="storage.bus" string="ide">
       <match key="storage.drive_type" string="disk">
