--- root/aclocal.m4.old	2006-10-21 18:51:30.000000000 +0200
+++ aclocal.m4	2006-10-21 20:48:07.000000000 +0200
@@ -495,8 +495,8 @@
 	AC_MSG_NOTICE([     not building text-mode ... requires ncurses library])
 elif test "$have_panel" = "no"; then
 	AC_MSG_NOTICE([     not building text-mode ... requires panel library])
-elif test "$have_ha_glib" = "no"; then
-	AC_MSG_NOTICE([     not building text-mode ... requires $ha_glib library])
+#elif test "$have_ha_glib" = "no"; then
+#	AC_MSG_NOTICE([     not building text-mode ... requires $ha_glib library])
 elif test "$have_math" = "no"; then
 	AC_MSG_NOTICE([     not building text-mode ... requires math library])
 else
@@ -920,7 +920,16 @@
 	fi
 fi
 
+ha_glib=glib
+have_ha_glib=$have_glib
+HA_GLIB_CPPFLAGS=$GLIB_CPPFLAGS
+HA_GLIB_LIBS=$GLIB_LIBS
+
 if test "x$no_glib2" = x; then
+	ha_glib=glib2
+	have_ha_glib=$have_glib2
+	HA_GLIB_CPPFLAGS=$GLIB2_CPPFLAGS
+	HA_GLIB_LIBS=$GLIB2_LIBS
 	AC_MSG_RESULT([yes])
 	ifelse([$2], , :, [$2])
 else
@@ -1286,10 +1295,10 @@
 		   [ifelse([$3], , :, [$3])])
 fi
 
-ha_glib=glib
-have_ha_glib=$have_glib
-HA_GLIB_CPPFLAGS=$GLIB_CPPFLAGS
-HA_GLIB_LIBS=$GLIB_LIBS
+#ha_glib=glib
+#have_ha_glib=$have_glib
+#HA_GLIB_CPPFLAGS=$GLIB_CPPFLAGS
+#HA_GLIB_LIBS=$GLIB_LIBS
 
 if test "$have_ha_version" = "yes"; then
 	if test -d /usr/include/heartbeat -a	\
