diff -urN xpp-1.5.old/mainwindow.fl xpp-1.5/mainwindow.fl
--- xpp-1.5.old/mainwindow.fl	2011-01-20 15:01:49.000000000 +0300
+++ xpp-1.5/mainwindow.fl	2011-01-20 15:02:33.000000000 +0300
@@ -70,7 +70,7 @@
   fileList->value(print->filelist);
 }}
           private xywh {440 75 65 25} down_box DOWN_BOX shortcut 0x80062 align 16
-          code0 {\#include <FL/fl_file_chooser.H>}
+          code0 {\#include <FL/Fl_File_Chooser.H>}
           code1 {\#include "xpp.h"}
         }
         Fl_Scroll printerList {
diff -urN xpp-1.5.old/xpp.cxx xpp-1.5/xpp.cxx
--- xpp-1.5.old/xpp.cxx	2011-01-20 15:01:49.000000000 +0300
+++ xpp-1.5/xpp.cxx	2011-01-20 15:04:24.000000000 +0300
@@ -2500,7 +2500,7 @@
     // Redraw the destination menu in the main window
 
     printerPack->parent()->redraw();
-    printerPack->draw();
+    ((Fl_Widget*)printerPack)->draw();
   
   }
 
@@ -2674,7 +2674,8 @@
   // Redraw the destination menu in the main window
 
   printerPack->parent()->redraw();
-  printerPack->draw();
+  ((Fl_Widget*)printerPack)->draw();
+  //printerPack->draw();
   
   return(dest_index);
 }
diff -urN xpp-1.5.old/xppmain.cxx xpp-1.5/xppmain.cxx
--- xpp-1.5.old/xppmain.cxx	2011-01-20 15:01:49.000000000 +0300
+++ xpp-1.5/xppmain.cxx	2011-01-20 15:05:17.000000000 +0300
@@ -32,7 +32,7 @@
  */
 
 #include "mainwindow.h"
-#include <Fl/Fl_Shared_Image.h>
+#include <FL/Fl_Shared_Image.H>
 
 /*
  * 'main()' - Do all what xpp should do
