--- kicker/applets/configure.in.in	1970-01-01 02:00:00.000000000 +0200
+++ kicker/applets/configure.in.in	2004-10-10 16:37:13.000000000 +0200
@@ -0,0 +1,24 @@
+AC_MSG_CHECKING([whether devices kioslave should be compiled])
+
+case `uname -s` in
+    FreeBSD)
+        DEVICESDIR=devices
+        ;;
+    Linux)
+        DEVICESDIR=devices
+        ;;
+    *)
+        DEVICESDIR=
+        ;;
+esac
+
+AC_SUBST(DEVICESDIR)
+if test x$DEVICESDIR != x; then
+    AC_MSG_RESULT([yes])
+else
+    AC_MSG_RESULT([no])
+fi
+
+AM_CONDITIONAL(include_devicesdir, test -n "$DEVICESDIR")
+
+AC_CHECK_FUNCS(setmntent)
