--- kicker/taskbar/taskcontainer.cpp	2007-01-07 19:11:16.000000000 -0500
+++ kicker/taskbar/taskcontainer.cpp	2007-01-07 19:14:20.000000000 -0500
@@ -54,6 +54,12 @@
 #include "taskcontainer.h"
 #include "taskcontainer.moc"
 
+#ifdef Q_WS_X11
+#include <X11/Xlib.h>
+#include <X11/Xatom.h>
+#include <X11/Xutil.h>
+#endif
+
 TaskContainer::TaskContainer(Task::Ptr task, TaskBar* bar,
                              QWidget *parent, const char *name)
     : QToolButton(parent, name),
@@ -1333,16 +1339,37 @@
         return;
     }
 
+    for (Task::List::Iterator it = tasks.begin(); it != tasks.end(); ++it)
+    {
+#ifdef Q_WS_X11
+		WId w = (*it)->window();
+		long int value=1;
+		Atom a = XInternAtom(qt_xdisplay(),"_WIN_SHOW_THUMBNAIL",false);
+		XChangeProperty(qt_xdisplay(),w,a,XA_CARDINAL,32,PropModeReplace,(unsigned char *)(&value),1);
+#endif
+    }
+
+#if 0
     if (!KickerSettings::showMouseOverEffects())
     {
         QString tooltip = "<qt>" + QStyleSheet::escape(name()) + "</qt>";
         QToolTip::add(this, tooltip);
         return;
     }
+#endif
 }
 
 void TaskContainer::leaveEvent(QEvent*)
 {
+    for (Task::List::Iterator it = tasks.begin(); it != tasks.end(); ++it)
+    {
+#ifdef Q_WS_X11
+		WId w = (*it)->window();
+		Atom a = XInternAtom(qt_xdisplay(),"_WIN_SHOW_THUMBNAIL",false);
+		XDeleteProperty(qt_xdisplay(),w,a);
+#endif
+    }
+
     m_mouseOver = false;
     update();
 }
