diff -Nurp suse_kickoff_qstyle-nosvn/kicker/kicker/buttons/knewbutton.cpp suse_kickoff_qstyle-nosvn-slintegration/kicker/kicker/buttons/knewbutton.cpp
--- suse_kickoff_qstyle-nosvn/kicker/kicker/buttons/knewbutton.cpp	2007-01-28 13:43:47.000000000 +0100
+++ suse_kickoff_qstyle-nosvn-slintegration/kicker/kicker/buttons/knewbutton.cpp	2007-01-28 13:44:11.000000000 +0100
@@ -65,7 +65,7 @@ KNewButton::KNewButton( QWidget* parent 
 
     setIconAlignment((Qt::AlignmentFlags)(AlignTop|AlignRight));
     setAcceptDrops(true);
-    setIcon("kmenu-suse");
+    setIcon("about_kde");
     setDrawArrow(false);
 
     m_movie = new QMovie(locate("appdata", "pics/kmenu_basic.mng"));
diff -Nurp suse_kickoff_qstyle-nosvn/kicker/kicker/ui/appletop_mnu.cpp suse_kickoff_qstyle-nosvn-slintegration/kicker/kicker/ui/appletop_mnu.cpp
--- suse_kickoff_qstyle-nosvn/kicker/kicker/ui/appletop_mnu.cpp	2007-01-28 13:43:07.000000000 +0100
+++ suse_kickoff_qstyle-nosvn-slintegration/kicker/kicker/ui/appletop_mnu.cpp	2007-01-28 13:43:38.000000000 +0100
@@ -169,7 +169,7 @@ PanelAppletOpMenu::PanelAppletOpMenu(int
         }
 
         if (KickerSettings::legacyKMenu()) 
-          insertItem(SmallIcon("suse"), i18n("Switch to SUSE Menu Style"), this, SLOT(toggleLegacy()));
+          insertItem(SmallIcon("about_kde"), i18n("Switch to Sabayon Linux Menu Style"), this, SLOT(toggleLegacy()));
         else
           insertItem(SmallIcon("about_kde"), i18n("Switch to KDE Menu Style"), this, SLOT(toggleLegacy()));
     }
diff -Nurp suse_kickoff_qstyle-nosvn/kicker/kicker/ui/itemview.cpp suse_kickoff_qstyle-nosvn-slintegration/kicker/kicker/ui/itemview.cpp
--- suse_kickoff_qstyle-nosvn/kicker/kicker/ui/itemview.cpp	2007-01-27 18:46:31.000000000 +0100
+++ suse_kickoff_qstyle-nosvn-slintegration/kicker/kicker/ui/itemview.cpp	2007-01-28 12:29:33.000000000 +0100
@@ -152,7 +152,7 @@ void KMenuItem::setup()
 {
     // if someone configured a larger generalFont than 10pt, he might have a _real_ problem with 7pt
     // the 7pt could be read out of konquerorrc I guess
-    float min_font_size = 7. * QMAX(1., KGlobalSettings::generalFont().pointSizeFloat() / 10.);
+    float min_font_size = 9. * QMAX(1., KGlobalSettings::generalFont().pointSizeFloat() / 10.);
 
     const int expected_height = 38;
     description_font_size = QMAX( pointSize( expected_height * .3, listView() ) + KickerSettings::kickoffFontPointSizeOffset(), min_font_size ) ;
@@ -369,7 +369,7 @@ void KMenuItemSeparator::setup()
 
     QFont f = QFont();
     QFontMetrics fm(f);
-    f.setPointSize( 8 + KickerSettings::kickoffFontPointSizeOffset() );
+    f.setPointSize( 9 + KickerSettings::kickoffFontPointSizeOffset() );
     if ( itemAbove() && !text( 0 ).isEmpty() )
         setHeight( (int)QMAX( 34., fm.height() * 1.4) );
     else
@@ -421,7 +421,7 @@ void KMenuItemSeparator::paintCell(QPain
       }
       QFont f = listView()->font();
       f.setWeight( QFont::Normal );
-      f.setPointSize( 8 + KickerSettings::kickoffFontPointSizeOffset() );
+      f.setPointSize( 9 + KickerSettings::kickoffFontPointSizeOffset() );
       p->setFont( f );
       QColor myColor = cg.color( QColorGroup::Text ).light( 200 );
       if ( qGray( myColor.rgb() ) == 0 )
@@ -527,7 +527,7 @@ void KMenuItemHeader::paintCell(QPainter
 
     int r = left_margin + margin * 2;
 
-    const int min_font_size = 7;
+    const int min_font_size = 9;
     int title_font_pixelSize = qRound( pixelSize( QMAX( pointSize( 12, listView() ) + KickerSettings::kickoffFontPointSizeOffset(), min_font_size + 1 ), listView() ) );
 
     QFont f1 = p->font();
diff -Nurp suse_kickoff_qstyle-nosvn/kicker/kicker/ui/k_new_mnu.cpp suse_kickoff_qstyle-nosvn-slintegration/kicker/kicker/ui/k_new_mnu.cpp
--- suse_kickoff_qstyle-nosvn/kicker/kicker/ui/k_new_mnu.cpp	2007-01-27 18:46:31.000000000 +0100
+++ suse_kickoff_qstyle-nosvn-slintegration/kicker/kicker/ui/k_new_mnu.cpp	2007-01-28 13:38:34.000000000 +0100
@@ -393,9 +393,9 @@ KMenu::KMenu()
     m_searchPixmap->setPixmap( BarIcon( "find", 32 ) );
 
     QFont f = font();
-    f.setPointSize( kMax( 7, (f.pointSize() * 4 / 5 ) + KickerSettings::kickoffFontPointSizeOffset() ) );
+    f.setPointSize( kMax( 9, (f.pointSize() * 4 / 5 ) + KickerSettings::kickoffFontPointSizeOffset() ) );
     m_tabBar->setFont ( f );
-    f.setPointSize( kMax( 7, (f.pointSize() * 3 / 2 ) + KickerSettings::kickoffFontPointSizeOffset() ) );
+    f.setPointSize( kMax( 9, (f.pointSize() * 3 / 2 ) + KickerSettings::kickoffFontPointSizeOffset() ) );
     m_searchLabel->setFont( f );
 
     static_cast<KLineEdit*>(m_kcommand->lineEdit())->setClickMessage(i18n( "Applications, Contacts and Documents" ) );
@@ -1197,7 +1197,9 @@ void KMenu::initialize()
      */
     KConfig ksmserver("ksmserverrc", false, false);
     ksmserver.setGroup("General");
-    connect( m_lock, SIGNAL(clicked()), SLOT(slotOpenSUSE()));
+    //connect( m_lock, SIGNAL(clicked()), SLOT(slotLock()));
+    m_lock->hide();
+
     m_tabBar->setTabEnabled(LeaveTab, kapp->authorize("logout"));
 
     // load search field history
@@ -1331,8 +1333,8 @@ void KMenu::insertStaticItems()
 
     m_systemView->insertSeparator( nId++, i18n("Applications"), index++);
 
-    KService::Ptr p = KService::serviceByStorageId("/usr/share/applications/YaST.desktop");
-    m_systemView->insertMenuItem(p, nId++, index++);
+    m_systemView->insertItem( "kcontrol", i18n( "KDE Settings" ),
+                                     i18n( "KDE Control Center" ), "/usr/kde/3.5/share/applications/kde/KControl.desktop", nId++, index++ );
 
     m_systemView->insertItem( "info", i18n( "System Information" ),
                               "sysinfo:/",  "sysinfo:/", nId++, index++ );
@@ -1685,6 +1687,7 @@ void KMenu::setOrientation(MenuOrientati
     {
         QWidget *footer = m_footer->mainWidget();
         QPixmap pix( 64, footer->height() );
+
         QPainter p( &pix );
         p.fillRect( 0, 0, 64, footer->height(), m_lock->colorGroup().brush( QColorGroup::Base ) );
         p.fillRect( 0, m_orientation == BottomUp ? footer->height() - 2 : 0,
