diff -Nurp kicker-3.5.8.orig/kicker/libkicker/kickertip.h kicker-3.5.8/kicker/libkicker/kickertip.h
--- kicker-3.5.8.orig/kicker/libkicker/kickertip.h	2007-10-26 10:15:38.000000000 +0200
+++ kicker-3.5.8/kicker/libkicker/kickertip.h	2007-10-26 10:16:31.000000000 +0200
@@ -110,6 +110,8 @@ private:
     QTimer m_frameTimer;
     bool m_dirty;
 
+    bool m_toolTipsEnabled;
+
     const QWidget* m_tippingFor;
 
     static KickerTip* m_self;
