diff -aurp open-iscsi-2.0-870-rc1/etc/iscsid.conf open-iscsi-2.0-870-rc1.work/etc/iscsid.conf
--- open-iscsi-2.0-870-rc1/etc/iscsid.conf	2008-06-30 20:14:03.000000000 -0500
+++ open-iscsi-2.0-870-rc1.work/etc/iscsid.conf	2008-06-30 21:08:29.000000000 -0500
@@ -27,8 +27,8 @@
 # To request that the iscsi initd scripts startup a session set to "automatic".
 # node.startup = automatic
 #
-# To manually startup the session set to "manual". The default is manual.
-node.startup = manual
+# To manually startup the session set to "manual". The default is automatic.
+node.startup = automatic
 
 # *************
 # CHAP Settings
diff -aurp open-iscsi-2.0-870-rc1/README open-iscsi-2.0-870-rc1.work/README
--- open-iscsi-2.0-870-rc1/README	2008-06-30 20:14:03.000000000 -0500
+++ open-iscsi-2.0-870-rc1.work/README	2008-06-30 21:08:29.000000000 -0500
@@ -78,11 +78,6 @@ the cache sync command will fail.
 - iscsiadm's -P 3 option will not print out scsi devices.
 - iscsid will not automatically online devices.
 
-You need to enable "Cryptographic API" under "Cryptographic options" in the
-kernel config. And you must enable "CRC32c CRC algorithm" even if
-you do not use header or data digests. They are the kernel options,
-CONFIG_CRYPTO and CONFIG_CRYPTO_CRC32C, respectively.
-
 By default the kernel source found at
 /lib/modules/`uname -a`/build
 will be used to compile the open-iscsi modules. To specify a different
@@ -694,7 +689,7 @@ Red Hat or Fedora:
 -----------------
 To start open-iscsi in Red Hat/Fedora you can do:
 
-	service open-iscsi start
+	service iscsi start
 
 To get open-iscsi to automatically start at run time you may have to
 run:
@@ -873,6 +868,8 @@ To login to all the automated nodes, sim
 e.g /etc/init.d/open-iscsi restart. On your next startup the nodes will
 be logged into autmotically.
 
+To set the startup value, so that nodes are not logged into automatically
+use the value "manual".
 
 8. Advanced Configuration
 =========================
