diff --git a/pithos/pithosconfig.py b/pithos/pithosconfig.py
index dfaaf5d..f82b458 100644
--- a/pithos/pithosconfig.py
+++ b/pithos/pithosconfig.py
@@ -16,7 +16,10 @@
 
 # where your project will head for your data (for instance, images and ui files)
 # by default, this is ../data, relative your trunk layout
-__pithos_data_directory__ = '../data/'
+from os.path import isdir, normpath, join
+__pithos_data_directory__ = (normpath(join(__file__, '../data/'))
+                             if isdir(join(__file__, '../data/'))
+                             else '/usr/share/pithos')
 __license__ = 'GPL-3'
 
 VERSION = '0.3.17'
