--- setup.py.orig	2011-05-07 18:22:10.404999969 +0200
+++ setup.py	2011-05-07 18:38:08.286999988 +0200
@@ -78,6 +78,7 @@
     description='Pandora.com client for the GNOME desktop',
     #long_description='Here a longer description',
     url='https://launchpad.net/pithos',
-    cmdclass={'install': InstallAndUpdateDataDirectory, 'build_icons':build_icons, 'build':build_extra}
+    cmdclass={'install': InstallAndUpdateDataDirectory, 'build_icons':build_icons, 'build':build_extra},
+    requires = []
     )
 
