--- gnome-media-2.32.0.orig//configure.ac
+++ gnome-media-2.32.0/configure.ac
@@ -155,7 +155,13 @@ AM_CONDITIONAL(HAVE_GMP, [test x$have_pr
 dnl Profiles glade ui: needed to edit .ui files with glade3
 have_gladeui=no
 
-PKG_CHECK_MODULES(GLADEUI, [gladeui-1.0], have_gladeui=yes, have_gladeui=no)
+AC_ARG_ENABLE([gladeui],
+  AS_HELP_STRING([--enable-gladeui], [enable gladeui usage]), 
+  ,enable_gladeui=yes)
+
+if test "x$enable_gladeui" = "xyes"; then
+	PKG_CHECK_MODULES(GLADEUI, [gladeui-1.0], have_gladeui=yes, have_gladeui=no)
+fi
 AC_SUBST(GLADEUI_CFLAGS)
 AC_SUBST(GLADEUI_LIBS)
 AM_CONDITIONAL(HAVE_GLADEUI, [test x$have_gladeui = xyes])
