Bug-Novell: https://bugzilla.novell.com/show_bug.cgi?id=536778
diff --git a/libslab/bookmark-agent.c b/libslab/bookmark-agent.c
index bdd418a..52322de 100644
--- a/libslab/bookmark-agent.c
+++ b/libslab/bookmark-agent.c
@@ -26,6 +26,8 @@
 #	define PACKAGE "gnome-main-menu"
 #endif
 
+#define MAIN_MENU "gnome-main-menu"
+
 #include <gtk/gtk.h>
 
 #include <string.h>
@@ -531,7 +533,7 @@ bookmark_agent_new (BookmarkStoreType type)
 		priv->user_modifiable = GPOINTER_TO_INT (libslab_get_gconf_value (priv->lockdown_key));
 
 		priv->user_store_path = g_build_filename (
-			g_get_user_data_dir (), PACKAGE, priv->store_filename, NULL);
+			g_get_user_data_dir (), MAIN_MENU, priv->store_filename, NULL);
 
 		priv->update_path = update_user_spec_path;
 
@@ -928,7 +930,7 @@ find_package_data_file (const gchar *filename)
 	dirs = g_get_system_data_dirs ();
 	
 	for (i = 0; ! path && dirs && dirs [i]; ++i) {
-		path = g_build_filename (dirs [i], PACKAGE, filename, NULL);
+		path = g_build_filename (dirs [i], MAIN_MENU, filename, NULL);
 		
 		if (! g_file_test (path, G_FILE_TEST_EXISTS)) {
 			g_free (path);


