commit 664c61ae03d721608a9232086e9b96bbbd25467c
Author: Fabio Erculiani <lxnay@sabayon.org>
Date:   Wed Nov 2 11:05:58 2011 +0100

    [6/sabayon-artwork-lxde] fix GtkLabel colors (white)

diff --git a/6/sabayon-artwork-lxde/lxdm/Sabayon/gtkrc b/6/sabayon-artwork-lxde/lxdm/Sabayon/gtkrc
index 747f7a0..9d20eab 100644
--- a/6/sabayon-artwork-lxde/lxdm/Sabayon/gtkrc
+++ b/6/sabayon-artwork-lxde/lxdm/Sabayon/gtkrc
@@ -17,19 +17,19 @@ style "back"
 style "time"
 {
 	fg[NORMAL]="#ffffff"
-	font_name="Sans 12"
+	font_name="Sans 10"
 }
 
 style "prompt"
 {
 	fg[NORMAL]="#ffffff"
-	font_name="Sans Bold 14"
+	font_name="Sans Bold 10"
 }
 
 # Normally this won't paint an invisible event box, but we did some hacks in greeter.c.
 style "bottom"
 {
-	font_name="Sans 12"
+	font_name="Sans 10"
 	fg[NORMAL] = "#9E9D9B"
 	engine "pixmap"
 	{
@@ -48,3 +48,4 @@ widget "lxdm.*.time" style "time"
 widget "lxdm.*.prompt" style "prompt"
 widget "lxdm.*.bottom_pane" style "bottom"
 widget_class "GtkWindow.*.GtkEventBox.*.<GtkLabel>" style "bottom"
+widget_class "GtkWindow.*.<GtkLabel>" style "prompt"
