diff -ru usr/src/nv/conftest.sh usr/src/nv.1544019/conftest.sh
--- usr/src/nv/conftest.sh	2006-05-15 22:42:17.000000000 +0200
+++ usr/src/nv.1544019/conftest.sh	2006-06-25 14:23:25.189839750 +0200
@@ -43,7 +43,8 @@
         # Determine the number of arguments expected by remap_page_range.
         #
 
-        echo "#include <linux/mm.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/mm.h>
         void conftest_remap_page_range(void) {
            remap_page_range();
         }" > conftest$$.c
@@ -56,7 +57,8 @@
           exit 1
         fi
 
-        echo "#include <linux/mm.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/mm.h>
         int conftest_remap_page_range(void) {
            pgprot_t pgprot = __pgprot(0);
            return remap_page_range(NULL, 0L, 0L, 0L, pgprot);
@@ -71,7 +73,8 @@
           exit 0
         fi
 
-        echo "#include <linux/mm.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/mm.h>
         int conftest_remap_page_range(void) {
            pgprot_t pgprot = __pgprot(0);
            return remap_page_range(0L, 0L, 0L, pgprot);
@@ -248,7 +251,8 @@
             # tree or at headers shipped for a specific kernel.
             # Determine the kernel version using a compile check.
             #
-            echo "#include <linux/version.h>
+            echo "#include <linux/autoconf.h>
+            #include <linux/version.h>
             #include <linux/utsname.h>
             int main() {
               if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,0)) {
@@ -341,7 +345,8 @@
             # tree or at headers shipped for a specific kernel.
             # Determine the kernel version using a compile check.
             #
-            echo "#include <linux/version.h>
+            echo "#include <linux/autoconf.h>
+            #include <linux/version.h>
             int main() {
                 printf(\"%s\", UTS_RELEASE);
                 return 0;
@@ -494,7 +499,8 @@
         # Determine if change_page_attr() is present
         #
         
-        echo "#include <linux/version.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/version.h>
         #include <linux/utsname.h>
         #include <linux/mm.h>
         #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
@@ -520,7 +526,8 @@
         # Determine if pci_get_class() is present.
         #
 
-        echo "#include <linux/pci.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/pci.h>
         void conftest_pci_get_class(void) {
             pci_get_class();
         }" > conftest$$.c
@@ -541,7 +548,8 @@
         # Determine if remap_pfn_range() is present.
         #
 
-        echo "#include <linux/mm.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/mm.h>
         void conftest_remap_pfn_range(void) {
             remap_pfn_range();
         }" > conftest$$.c
@@ -562,7 +570,8 @@
         # Is 'rlim' a member of current->signal?
         #
 
-        echo "#include <linux/sched.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/sched.h>
         struct rlimit *conftest_signal_struct_rlim(void) {
             return current->signal->rlim;
         }" > conftest$$.c
@@ -584,7 +593,8 @@
         # aware.
         #
 
-        echo "#include <linux/types.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/types.h>
         #include <linux/agp_backend.h>
         typedef struct agp_bridge_data agp_bridge_data;
         agp_bridge_data *conftest_agp_backend_acquire(struct pci_dev *dev) {
@@ -608,7 +618,8 @@
         # it takes.
         #
 
-        echo "#include <linux/vmalloc.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/vmalloc.h>
         void conftest_vmap(void) {
             vmap();
         }" > conftest$$.c
@@ -621,7 +632,8 @@
             exit 1
         fi
 
-        echo "#include <linux/vmalloc.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/vmalloc.h>
         void *conftest_vmap(struct page **pg, int cnt) {
             return vmap(pg, cnt);
         }" > conftest$$.c
@@ -635,7 +647,8 @@
             exit 0
         fi
 
-        echo "#include <linux/vmalloc.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/vmalloc.h>
         #include <linux/mm.h>
         void *conftest_vmap(struct page **pg, int cnt) {
             return vmap(pg, cnt, 0, PAGE_KERNEL);
@@ -662,7 +675,8 @@
         # the inc_use() and dec_use() callback entries.
         #
 
-        echo "#include <linux/i2c.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/i2c.h>
         int conftest_i2c_adapter_inc_use() {
             return offsetof(struct i2c_adapter, inc_use);
         }" > conftest$$.c
@@ -684,7 +698,8 @@
         # kernel.
         #
 
-        echo "#include <linux/sched.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/sched.h>
         int conftest_sysctl_max_map_count(void) {
             return sysctl_max_map_count;
         }" > conftest$$.c
@@ -705,7 +720,8 @@
         # Does linux/pm.h declare the pm_message_t type?
         #
 
-        echo "#include <linux/pm.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/pm.h>
         void conftest_pm_message_t(pm_message_t state) {
             pm_message_t *p = &state;
         }" > conftest$$.c
@@ -726,7 +742,8 @@
         # Determine if pci_choose_state() is present.
         #
 
-        echo "#include <linux/pci.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/pci.h>
         void conftest_pci_choose_state(void) {
             pci_choose_state();
         }" > conftest$$.c
@@ -747,7 +764,8 @@
         # Determine if vm_insert_page() is present.
         #
 
-        echo "#include <linux/mm.h>
+        echo "#include <linux/autoconf.h>
+        #include <linux/mm.h>
         void conftest_vm_insert_page(void) {
             vm_insert_page();
         }" > conftest$$.c
