Index: b/hald/hald-cache-test.sh
===================================================================
--- a/hald/hald-cache-test.sh	2007-03-07 16:52:19.000000000 -0500
+++ b/hald/hald-cache-test.sh	2007-06-14 12:10:19.000000000 -0400
@@ -3,9 +3,9 @@
 mkdir .local-fdi-test
 make -C ../fdi install DESTDIR=`pwd`/.local-fdi-test prefix=/
 
-HAL_FDI_SOURCE_PREPROBE=.local-fdi-test/share/hal/fdi/preprobe
-HAL_FDI_SOURCE_INFORMATION=.local-fdi-test/share/hal/fdi/preprobe
-HAL_FDI_SOURCE_POLICY=.local-fdi-test/share/hal/fdi/policy
+HAL_FDI_SOURCE_PREPROBE=.local-fdi-test/usr/share/hal/fdi/preprobe
+HAL_FDI_SOURCE_INFORMATION=.local-fdi-test/usr/share/hal/fdi/preprobe
+HAL_FDI_SOURCE_POLICY=.local-fdi-test/usr/share/hal/fdi/policy
 HAL_FDI_CACHE_NAME=.local-fdi-test/hald-local-fdi-cache
 export HAL_FDI_SOURCE_PREPROBE HAL_FDI_SOURCE_INFORMATION \
        HAL_FDI_SOURCE_POLICY HAL_FDI_CACHE_NAME
