--- hal-0.5.9.1.orig/configure.in	2007-05-02 05:57:02.000000000 +0200
+++ hal-0.5.9.1/configure.in	2007-09-15 18:40:05.000000000 +0200
@@ -215,12 +215,14 @@ if test "x$use_parted" = "xyes" ; then
 	    (major == 1 && minor == 8 && micro == 0) ||
 	    (major == 1 && minor == 8 && micro == 1) ||
 	    (major == 1 && minor == 8 && micro == 2) ||
-	    (major == 1 && minor == 8 && micro == 6))
+	    (major == 1 && minor == 8 && micro == 6) ||
+	    (major == 1 && minor == 8 && micro == 7) ||
+	    (major == 1 && minor == 8 && micro == 8))
 	   return 0;
 
 	return 1;
     }
-	,AC_MSG_RESULT( ),AC_MSG_ERROR([*** Requires libparted == 1.7.1/1.8.0/1.8.1/1.8.2 or 1.8.6]) )
+	,AC_MSG_RESULT( ),AC_MSG_ERROR([*** Requires libparted == 1.7.1/1.8.0/1.8.1/1.8.2/1.8.6/1.8.7 or 1.8.8]) )
     PARTED_LIBS=-lparted
     AC_SUBST(PARTED_LIBS)
     LDFLAGS=
