diff -Nurp hal-0.5.8.1.orig/tools/hal-storage-mount.c hal-0.5.8.1/tools/hal-storage-mount.c
--- hal-0.5.8.1.orig/tools/hal-storage-mount.c
+++ hal-0.5.8.1/tools/hal-storage-mount.c
@@ -813,6 +813,17 @@ handle_mount (LibHalContext *hal_ctx, 
 	} else if (libhal_volume_get_fstype (volume) != NULL && strlen (libhal_volume_get_fstype (volume)) > 0) {
 		mount_do_fstype = (char *) map_fstype (libhal_volume_get_fstype (volume));
 	}
+
+	/* Dirty hack while we're waiting for a better KDE support for filesystems' selection */
+	/* we are going to force NTFS-FUSE instead of NTFS */
+	if (strcmp (mount_do_fstype, "ntfs") == 0) {
+		int rc = system("which ntfs-3g");
+		if ( rc == 1 )
+			mount_do_fstype = "ntfs-fuse";
+		else
+			mount_do_fstype = "ntfs-3g";
+	}
+
 	args[na++] = MOUNT_TYPE_OPT;
 	args[na++] = mount_do_fstype;
 
