--- hal-0.5.12rc1.orig/tools/hal-storage-mount.c	2009-03-16 20:23:53.744118591 +0100
+++ hal-0.5.12rc1/tools/hal-storage-mount.c	2009-03-16 20:31:37.546232405 +0100
@@ -695,6 +695,15 @@ handle_mount (LibHalContext *hal_ctx, 
 		mount_do_fstype = "auto";
 	}
 
+	/* Try to use NTFS-3G when possible */
+	if (strcmp (mount_do_fstype, "ntfs") == 0) {
+		int rc = system("which ntfs-3g &> /dev/null");
+		if ( rc == 0 )
+			mount_do_fstype = "ntfs-3g";
+	}
+
+
+
 	/* check the mount options */
 	dbus_error_init (&error);
 	
