--- hal-0.5.8.1/libhal/libhal.c.orig	2006-09-27 20:35:03.000000000 +0200
+++ hal-0.5.8.1/libhal/libhal.c	2006-09-27 20:35:35.000000000 +0200
@@ -1457,7 +1457,7 @@ libhal_device_set_property_helper (LibHa
 	reply = dbus_connection_send_with_reply_and_block (ctx->connection,
 							   message, -1,
 							   error);
-	if (dbus_error_is_set (error)) {
+	if (error != NULL && dbus_error_is_set (error)) {
 		dbus_message_unref (message);
 		return FALSE;
 	}
