Index: b/partutil/Makefile.am
===================================================================
--- a/partutil/Makefile.am	2007-04-11 15:50:44.000000000 -0400
+++ b/partutil/Makefile.am	2007-04-12 11:42:23.000000000 -0400
@@ -1,17 +1,12 @@
-if HALD_COMPILE_SOLARIS 
-# do nothing
-else if HALD_COMPILE_FREEBSD
-# do nothing
-else 
+if HALD_COMPILE_LINUX
+noinst_LTLIBRARIES = libpartutil.la
+endif
 
 INCLUDES = @GLIB_CFLAGS@
 
-noinst_LTLIBRARIES = libpartutil.la
-
 libpartutil_la_SOURCES = partutil.h partutil.c ../hald/logger.c
 
 libpartutil_la_LIBADD = @GLIB_LIBS@ @PARTED_LIBS@
 
 clean-local :
 	rm -f *~
-endif
