Index: b/hald/freebsd/hf-usb.c
===================================================================
--- a/hald/freebsd/hf-usb.c	2007-04-12 11:42:05.000000000 -0400
+++ b/hald/freebsd/hf-usb.c	2007-04-12 11:59:40.000000000 -0400
@@ -372,11 +372,11 @@
       can_wake_up = (config_desc.bmAttributes & UC_REMOTE_WAKEUP) != 0;
       num_interfaces = config_desc.bNumInterface;
 
-      if (config_desc->iConfiguration != 0)
+      if (config_desc.iConfiguration != 0)
 	{
 	  char *configuration;
 
-	  configuration = hf_usb_get_string_descriptor(controller->fd, di->udi_addr, config_desc->iConfiguration, NULL);
+	  configuration = hf_usb_get_string_descriptor(controller->fd, di->udi_addr, config_desc.iConfiguration, NULL);
 	  if (configuration)
 	    {
 	      hal_device_property_set_string(device, "usb_device.configuration", configuration);
@@ -424,6 +424,8 @@
 
 static HalDevice *
 hf_usb_interface_device_new (HalDevice *parent,
+			     Controller *controller,
+			     const struct usb_device_info *di,
 			     const usb_interface_descriptor_t *desc)
 {
   HalDevice *device;
@@ -576,7 +578,7 @@
 	      break;
 	    }
 
-	  if_device = hf_usb_interface_device_new(device, if_desc);
+	  if_device = hf_usb_interface_device_new(device, controller, device_info, if_desc);
 	  hf_device_preprobe_and_add(if_device);
 
 	  p += USB_INTERFACE_DESCRIPTOR_SIZE + if_desc->bNumEndpoints * USB_ENDPOINT_DESCRIPTOR_SIZE;
Index: b/hald/freebsd/hf-pci.c
===================================================================
--- a/hald/freebsd/hf-pci.c	2007-04-12 11:42:05.000000000 -0400
+++ b/hald/freebsd/hf-pci.c	2007-04-12 12:04:48.000000000 -0400
@@ -38,6 +38,7 @@
 
 #include "../hald.h"
 #include "../ids.h"
+#include "../device.h"
 #include "../logger.h"
 #include "../util.h"
 
@@ -210,9 +211,15 @@
 
 	  info = g_new(DeviceInfo, 1);
 	  info->device = hf_device_store_match(hald_get_gdl(),
-			                       hal_property_new_int("pci.freebsd.bus", p->pc_sel.pc_bus),
-					       hal_property_new_int("pci.freebsd.device", p->pc_sel.pc_dev),
-					       hal_property_new_int("pci.freebsd.function", p->pc_sel.pc_func),
+			  		       "pci.freebsd.bus",
+					       HAL_PROPERTY_TYPE_INT32,
+					       p->pc_sel.pc_bus,
+					       "pci.freebsd.device",
+					       HAL_PROPERTY_TYPE_INT32,
+					       p->pc_sel.pc_dev,
+					       "pci.freebsd.function",
+					       HAL_PROPERTY_TYPE_INT32,
+					       p->pc_sel.pc_func,
 					       NULL);
 	  info->p = *p;
 	  info->secondary_bus = hf_pci_get_register(p, PCIR_SECBUS_1);
