Index: b/fdi/policy/10osvendor/20-storage-methods.fdi
===================================================================
--- a/fdi/policy/10osvendor/20-storage-methods.fdi	2007-03-27 01:18:54.000000000 -0400
+++ b/fdi/policy/10osvendor/20-storage-methods.fdi	2007-04-18 10:20:54.000000000 -0400
@@ -186,6 +186,7 @@
 	  <append key="volume.mount.valid_options" type="strlist">gid=</append>
 	  <append key="volume.mount.valid_options" type="strlist">umask=</append>
 	  <append key="volume.mount.valid_options" type="strlist">locale=</append>
+	  <append key="volume.mount.valid_options" type="strlist">utf8</append>
 	</match>
 	<match key="/org/freedesktop/Hal/devices/computer:system.kernel.name" string="FreeBSD">
 	  <append key="volume.mount.valid_options" type="strlist">-u=</append>
