diff -Naur alsa-driver-1.0.6a.orig/alsa-kernel/include/asound.h alsa-driver-1.0.6a/alsa-kernel/include/asound.h
--- alsa-driver-1.0.6a.orig/alsa-kernel/include/asound.h	2004-08-03 11:55:55.000000000 -0700
+++ alsa-driver-1.0.6a/alsa-kernel/include/asound.h	2004-08-25 11:05:21.445603678 -0700
@@ -63,6 +63,10 @@
 #include <sys/types.h>
 #endif
 
+#ifndef __user
+#define __user
+#endif
+
 /*
  *  protocol version
  */
