diff -Nur ov51x-jpeg-1.5.9-old/ov51x-jpeg.h ov51x-jpeg-1.5.9/ov51x-jpeg.h
--- ov51x-jpeg-1.5.9-old/ov51x-jpeg.h	2009-05-04 19:42:00.000000000 +0300
+++ ov51x-jpeg-1.5.9/ov51x-jpeg.h	2009-05-04 19:43:01.000000000 +0300
@@ -63,6 +63,12 @@
 #include <media/v4l2-ioctl.h>
 #endif
 
+/* Add missing info and warn macros by hand for new kernels */
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 28)
+#define info(args...) printk(KERN_INFO KBUILD_MODNAME args)
+#define warn(args...) printk(KERN_WARNING KBUILD_MODNAME args)
+#endif
+
 /* --------------------------------- */
 /* DEFINES FOR OV511 AND OTHER CHIPS */
 /* --------------------------------- */

