diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/AdminServer.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/AdminServer.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/AdminServer.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/AdminServer.java	2010-06-24 19:25:07.677735214 +0200
@@ -248,7 +248,7 @@ public class AdminServer extends Abstrac
 
 
         try {
-            String adminURL = _consoleInfo.getAdminURL() + "admin-serv/tasks/operation/StatusPing";
+            String adminURL = _consoleInfo.getAdminURL() + "admin-serv/commands/tasks/operation/StatusPing";
             AdmTask task = new AdmTask(new URL(adminURL),
                     _consoleInfo.getAuthenticationDN(),
                     _consoleInfo.getAuthenticationPassword());
@@ -340,7 +340,7 @@ public class AdminServer extends Abstrac
             // Invoke CGI to configure the server with the cloned configuration information.
             // Some attributes, such as nsadminenableenduser, must be forced set.
             configURL =
-                    new String(_consoleInfo.getAdminURL() + "admin-serv/tasks/Configuration/ServerSetup?op=force_set" +
+                    new String(_consoleInfo.getAdminURL() + "admin-serv/commands/tasks/Configuration/ServerSetup?op=force_set" +
                     "&configuration.nsadminaccessaddresses=" +
                     addresses + "&configuration.nsadminaccesshosts=" +
                     hosts + "&configuration.nsadmincachelifetime=" +
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/AdminServerProduct.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/AdminServerProduct.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/AdminServerProduct.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/AdminServerProduct.java	2010-06-24 19:25:07.678734800 +0200
@@ -143,7 +143,7 @@ public class AdminServerProduct implemen
     protected boolean migrateConfig(String adminURL, String serverRoot) {
 
         String commandString =
-                new String(adminURL + "admin-serv/tasks/operation/MigrateConfig" +
+                new String(adminURL + "admin-serv/commands/tasks/operation/MigrateConfig" +
                 "?oldServerRoot=" + URLByteEncoder.encodeUTF8(serverRoot));
         Debug.println(
                 "TRACE AdminServerProduct.migrateConfig: MigrateConfig commandString = " +
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/cmdln/CommandLine.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/cmdln/CommandLine.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/cmdln/CommandLine.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/cmdln/CommandLine.java	2010-06-24 19:25:15.381922078 +0200
@@ -183,13 +183,13 @@ public class CommandLine extends Object
 
 
     protected static final String PACKAGE_DIR = "com/netscape/management/admserv/cmdln";
-    protected static final String RESTART_TASK = "admin-serv/tasks/operation/Restart";
-    protected static final String STOP_TASK = "admin-serv/tasks/operation/Stop";
-    protected static final String READ_LOG_TASK = "admin-serv/tasks/configuration/ReadLog";
-    protected static final String SERVER_SETUP_TASK = "admin-serv/tasks/configuration/ServerSetup";
-    protected static final String ACCESS_SETUP_TASK = "admin-serv/tasks/configuration/AccessSetup";
-    protected static final String DIRECTORY_SETUP_TASK = "admin-serv/tasks/configuration/DirectorySetup";
-    protected static final String UG_DIRECTORY_SETUP_TASK = "admin-serv/tasks/configuration/UGDirectorySetup";
+    protected static final String RESTART_TASK = "admin-serv/commands/tasks/operation/Restart";
+    protected static final String STOP_TASK = "admin-serv/commands/tasks/operation/Stop";
+    protected static final String READ_LOG_TASK = "admin-serv/commands/tasks/configuration/ReadLog";
+    protected static final String SERVER_SETUP_TASK = "admin-serv/commands/tasks/configuration/ServerSetup";
+    protected static final String ACCESS_SETUP_TASK = "admin-serv/commands/tasks/configuration/AccessSetup";
+    protected static final String DIRECTORY_SETUP_TASK = "admin-serv/commands/tasks/configuration/DirectorySetup";
+    protected static final String UG_DIRECTORY_SETUP_TASK = "admin-serv/commands/tasks/configuration/UGDirectorySetup";
     protected static final String DATA_FILE = "admconfig.dat";
     protected static final String VERSION_FILE = "version.dat";
 
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/logging/AccessLogViewerModel.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/logging/AccessLogViewerModel.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/logging/AccessLogViewerModel.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/logging/AccessLogViewerModel.java	2010-06-24 19:25:28.430785279 +0200
@@ -45,7 +45,7 @@ public class AccessLogViewerModel extend
     public AccessLogViewerModel(ConsoleInfo consoleInfo, String url) {
         super();
         _consoleInfo = consoleInfo;
-        //_url = "http://gumby.mcom.com:666/admin-serv/tasks/Configuration/ReadLog"; //_consoleInfo.getAdminURL() + url;
+        //_url = "http://gumby.mcom.com:666/admin-serv/commands/tasks/Configuration/ReadLog"; //_consoleInfo.getAdminURL() + url;
         //_username = "uid=admin, ou=People, o=airius.com"; //_consoleInfo.getAuthenticationDN();
         //_password = "admin"; //_consoleInfo.getAuthenticationPassword();
         _url = _consoleInfo.getAdminURL() + url;
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/LoggingNode.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/LoggingNode.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/LoggingNode.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/LoggingNode.java	2010-06-24 19:25:07.678734800 +0200
@@ -128,7 +128,7 @@ class AccessLogNode extends ResourceObje
     }
 
     public Component getCustomPanel() {
-        return new LogViewer(new AccessLogViewerModel(_ci, "admin-serv/tasks/Configuration/ReadLog"));
+        return new LogViewer(new AccessLogViewerModel(_ci, "admin-serv/commands/tasks/Configuration/ReadLog"));
     }
 }
 
@@ -142,6 +142,6 @@ class ErrorLogNode extends ResourceObjec
     }
 
     public Component getCustomPanel() {
-        return new LogViewer(new ErrorLogViewerModel(_ci, "admin-serv/tasks/Configuration/ReadLog"));
+        return new LogViewer(new ErrorLogViewerModel(_ci, "admin-serv/commands/tasks/Configuration/ReadLog"));
     }
 }
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/CGIAccessSetup.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/CGIAccessSetup.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/CGIAccessSetup.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/CGIAccessSetup.java	2010-06-24 19:24:56.927832477 +0200
@@ -32,7 +32,7 @@ import com.netscape.management.admserv.c
   * @author miodrag@netscape.com
   */
 public class CGIAccessSetup extends CGIDataModel {
-    private static String _taskURL = "admin-serv/tasks/Configuration/AccessSetup";
+    private static String _taskURL = "admin-serv/commands/tasks/Configuration/AccessSetup";
 
     private boolean _dsgwAdded = false;
 
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/CGIDirectorySetup.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/CGIDirectorySetup.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/CGIDirectorySetup.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/CGIDirectorySetup.java	2010-06-24 19:24:56.929922739 +0200
@@ -32,7 +32,7 @@ import com.netscape.management.admserv.c
   * @author miodrag@netscape.com
   */
 public class CGIDirectorySetup extends CGIDataModel {
-    private static String _taskURL = "admin-serv/tasks/Configuration/DirectorySetup";
+    private static String _taskURL = "admin-serv/commands/tasks/Configuration/DirectorySetup";
 
     public CGIDirectorySetup(ConsoleInfo consoleInfo) {
         super(consoleInfo, _taskURL);
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/CGILoggingSetup.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/CGILoggingSetup.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/CGILoggingSetup.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/CGILoggingSetup.java	2010-06-24 19:24:56.929922739 +0200
@@ -30,7 +30,7 @@ import com.netscape.management.client.ut
   * @author miodrag@netscape.com
   */
 public class CGILoggingSetup extends CGIDataModel {
-    private static String _taskURL = "admin-serv/tasks/Configuration/ServerSetup";
+    private static String _taskURL = "admin-serv/commands/tasks/Configuration/ServerSetup";
 
     public CGILoggingSetup(ConsoleInfo consoleInfo) {
         super(consoleInfo, _taskURL);
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/CGIServerSetup.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/CGIServerSetup.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/CGIServerSetup.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/CGIServerSetup.java	2010-06-24 19:24:56.930835722 +0200
@@ -32,7 +32,7 @@ import com.netscape.management.admserv.c
   * @author miodrag@netscape.com
   */
 public class CGIServerSetup extends CGIDataModel {
-    private static String _taskURL = "admin-serv/tasks/Configuration/ServerSetup";
+    private static String _taskURL = "admin-serv/commands/tasks/Configuration/ServerSetup";
 
     private String _portValue = null;
     private String _addrValue = null;
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/CGISNMPSetup.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/CGISNMPSetup.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/CGISNMPSetup.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/CGISNMPSetup.java	2010-06-24 19:24:56.930835722 +0200
@@ -32,7 +32,7 @@ import com.netscape.management.admserv.c
   * @author miodrag@netscape.com
   */
 public class CGISNMPSetup extends CGIDataModel {
-    private static String _taskURL = "admin-serv/tasks/Configuration/SNMPSetup";
+    private static String _taskURL = "admin-serv/commands/tasks/Configuration/SNMPSetup";
 
     // String used as keys in the Hashtable for a community entry
     public static final String COMMUNITY_NAME = "community";
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/CGIUGDirectorySetup.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/CGIUGDirectorySetup.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/CGIUGDirectorySetup.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/CGIUGDirectorySetup.java	2010-06-24 19:24:56.931778239 +0200
@@ -32,7 +32,7 @@ import com.netscape.management.admserv.c
   * @author miodrag@netscape.com
   */
 public class CGIUGDirectorySetup extends CGIDataModel {
-    private static String _taskURL = "admin-serv/tasks/Configuration/UGDirectorySetup";
+    private static String _taskURL = "admin-serv/commands/tasks/Configuration/UGDirectorySetup";
 
     public CGIUGDirectorySetup(ConsoleInfo consoleInfo) {
         super(consoleInfo, _taskURL);
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/RestartOperation.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/RestartOperation.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/RestartOperation.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/RestartOperation.java	2010-06-24 19:24:56.936922811 +0200
@@ -44,7 +44,7 @@ public class RestartOperation extends Ad
     String _i18nStopped = _resource.getString("restart","Stopped");
 
     public RestartOperation(ConsoleInfo consoleInfo) {
-        super(consoleInfo, "admin-serv/tasks/Operation/Restart");
+        super(consoleInfo, "admin-serv/commands/tasks/Operation/Restart");
         _restartControl =
                 (IRestartControl)_consoleInfo.get(IRestartControl.ID);
         createActionPanel(/*multiLined=*/true);
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/SNMPStatusPanel.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/SNMPStatusPanel.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/SNMPStatusPanel.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/SNMPStatusPanel.java	2010-06-24 19:24:56.939922669 +0200
@@ -253,7 +253,7 @@ public class SNMPStatusPanel extends Plu
     }
 
     protected void checkStatus(ConsoleInfo ci, boolean init) {
-        String taskURL = "admin-serv/tasks/Operation/SNMPControl?ACTION=STATUS";
+        String taskURL = "admin-serv/commands/tasks/Operation/SNMPControl?ACTION=STATUS";
         AdminOperation operation =
                 new AdminOperation(ci, taskURL);
         StatusCheckListener listener = new StatusCheckListener(operation);
@@ -273,7 +273,7 @@ public class SNMPStatusPanel extends Plu
     }
 
     protected void restart(ConsoleInfo ci) {
-        String taskURL = "admin-serv/tasks/Operation/SNMPControl?ACTION=RESTART";
+        String taskURL = "admin-serv/commands/tasks/Operation/SNMPControl?ACTION=RESTART";
         AdminOperation operation =
                 new AdminOperation(ci, taskURL);
         StopStartListener listener =
@@ -288,7 +288,7 @@ public class SNMPStatusPanel extends Plu
     }
 
     protected void stop(ConsoleInfo ci) {
-        String taskURL = "admin-serv/tasks/Operation/SNMPControl?ACTION=STOP";
+        String taskURL = "admin-serv/commands/tasks/Operation/SNMPControl?ACTION=STOP";
         AdminOperation operation =
                 new AdminOperation(ci, taskURL);
         StopStartListener listener = new StopStartListener(operation, true);
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/StopOperation.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/StopOperation.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/StopOperation.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/StopOperation.java	2010-06-24 19:24:56.939922669 +0200
@@ -37,7 +37,7 @@ public class StopOperation extends Admin
     boolean _serverRunning = true;
 
     public StopOperation(ConsoleInfo consoleInfo) {
-        super(consoleInfo, "admin-serv/tasks/Operation/Stop");
+        super(consoleInfo, "admin-serv/commands/tasks/Operation/Stop");
         createActionPanel(/*multiLined=*/true);
     }
 
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/TurnOnSSL.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/TurnOnSSL.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/panel/TurnOnSSL.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/panel/TurnOnSSL.java	2010-06-24 19:24:56.941778589 +0200
@@ -266,7 +266,7 @@ public class TurnOnSSL implements IPlugi
 
         try {
             AdmTask task = new AdmTask(
-                    new URL(_consoleInfo.getAdminURL() + "admin-serv/tasks/configuration/SSLActivate"),
+                    new URL(_consoleInfo.getAdminURL() + "admin-serv/commands/tasks/configuration/SSLActivate"),
                     _consoleInfo.getAuthenticationDN(),
                     _consoleInfo.getAuthenticationPassword());
             task.exec();
@@ -388,7 +388,7 @@ public class TurnOnSSL implements IPlugi
             try {
                 
                 if (validEntries()) {
-                    String adminURL = _consoleInfo.getAdminURL() + "admin-serv/tasks/configuration/SSLActivate";
+                    String adminURL = _consoleInfo.getAdminURL() + "admin-serv/commands/tasks/configuration/SSLActivate";
                     AdmTask task = new AdmTask(new URL(adminURL),
                                                _consoleInfo.getAuthenticationDN(),
                                                _consoleInfo.getAuthenticationPassword());
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/SNMPNode.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/SNMPNode.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/SNMPNode.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/SNMPNode.java	2010-06-24 19:25:07.680734826 +0200
@@ -112,7 +112,7 @@ public class SNMPNode extends ResourceOb
                 _configPanel = null;
                 page.setCustomPanel(getCustomPanel());
             } else if (item.getID().equals(MENU_STOP)) {
-                String taskURL = "admin-serv/tasks/Operation/SNMPControl?ACTION=STOP";
+                String taskURL = "admin-serv/commands/tasks/Operation/SNMPControl?ACTION=STOP";
                 ResourcePage page = (ResourcePage) viewInstance;
                 AdminOperation operation =
                         new AdminOperation(_consoleInfo, taskURL);
@@ -125,7 +125,7 @@ public class SNMPNode extends ResourceOb
                 dialog.dispose();
                 ModalDialogUtil.sleep();
             } else if (item.getID().equals(MENU_RESTART)) {
-                String taskURL = "admin-serv/tasks/Operation/SNMPControl?ACTION=RESTART";
+                String taskURL = "admin-serv/commands/tasks/Operation/SNMPControl?ACTION=RESTART";
                 ResourcePage page = (ResourcePage) viewInstance;
                 AdminOperation operation =
                         new AdminOperation(_consoleInfo, taskURL);
diff -Nurp 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/task/ServerSetup.java 389-admin-console-1.1.5/src/com/netscape/management/admserv/task/ServerSetup.java
--- 389-admin-console-1.1.5.orig/src/com/netscape/management/admserv/task/ServerSetup.java	2010-06-02 16:57:20.000000000 +0200
+++ 389-admin-console-1.1.5/src/com/netscape/management/admserv/task/ServerSetup.java	2010-06-24 19:25:24.866735947 +0200
@@ -62,7 +62,7 @@ public class ServerSetup extends AdminTa
     public JDialog getConfigDialog(Component parent) {
 
         DialogFrame configDialog = null;
-        String taskURL = "admin-serv/tasks/Configuration/ServerSetup";
+        String taskURL = "admin-serv/commands/tasks/Configuration/ServerSetup";
         String adminURL = _consoleInfo.getAdminURL() + taskURL;
         configDialog = new DialogFrame(parent, getName(),
                 new BaseConfigPanel(
