--- midori-0.4.1.orig/midori/midori-websettings.c
+++ midori-0.4.1/midori/midori-websettings.c
@@ -1134,6 +1134,7 @@ generate_ident_string (MidoriWebSettings* web_settings,
     #endif
     const int webcore_major = WEBKIT_USER_AGENT_MAJOR_VERSION;
     const int webcore_minor = WEBKIT_USER_AGENT_MINOR_VERSION;
+    const gchar* distro = "Sabayon";
 
     #if WEBKIT_CHECK_VERSION (1, 1, 18)
     g_object_set (web_settings, "enable-site-specific-quirks",
@@ -1143,26 +1144,26 @@ generate_ident_string (MidoriWebSettings* web_settings,
     switch (identify_as)
     {
     case MIDORI_IDENT_GENUINE:
-        return g_strdup_printf ("Mozilla/5.0 (%s %s) AppleWebKit/%d.%d+ %s",
-            platform, os, webcore_major, webcore_minor, appname);
+        return g_strdup_printf ("Mozilla/5.0 (%s %s) AppleWebKit/%d.%d+ %s %s",
+            platform, os, webcore_major, webcore_minor, distro, appname);
     case MIDORI_IDENT_MIDORI:
     case MIDORI_IDENT_SAFARI:
         g_object_set (web_settings, "enable-site-specific-quirks", TRUE, NULL);
         return g_strdup_printf ("Mozilla/5.0 (Macintosh; U; Intel Mac OS X; %s) "
-            "AppleWebKit/%d+ (KHTML, like Gecko) Version/5.0 Safari/%d.%d+ %s",
-            lang, webcore_major, webcore_major, webcore_minor, appname);
+            "AppleWebKit/%d+ (KHTML, like Gecko) Version/5.0 %s Safari/%d.%d+ %s",
+            lang, webcore_major, distro, webcore_major, webcore_minor, appname);
     case MIDORI_IDENT_IPHONE:
         return g_strdup_printf ("Mozilla/5.0 (iPhone; U; CPU like Mac OS X; %s) "
-            "AppleWebKit/532+ (KHTML, like Gecko) Version/3.0 Mobile/1A538b Safari/419.3 %s",
-                                lang, appname);
+            "AppleWebKit/532+ (KHTML, like Gecko) Version/3.0 Mobile/1A538b %s Safari/419.3 %s",
+                                lang, distro, appname);
     case MIDORI_IDENT_FIREFOX:
-        return g_strdup_printf ("Mozilla/5.0 (%s %s; rv:2.0.1) Gecko/20100101 Firefox/4.0.1 %s",
-                                platform, os, appname);
+        return g_strdup_printf ("Mozilla/5.0 (%s %s; rv:2.0.1) Gecko/20100101 %s Firefox/4.0.1 %s",
+                                platform, os, distro, appname);
     case MIDORI_IDENT_EXPLORER:
-        return g_strdup_printf ("Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; %s) %s",
-                                lang, appname);
+        return g_strdup_printf ("Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; %s) %s %s",
+                                lang, distro, appname);
     default:
-        return g_strdup_printf ("%s", appname);
+        return g_strdup_printf ("%s %s", distro, appname);
     }
 }
 
