--- midori-0.4.1.orig/midori/midori-websettings.c
+++ midori-0.4.1/midori/midori-websettings.c
@@ -1087,7 +1087,7 @@ get_sys_name (void)
         #else
         struct utsname name;
         if (uname (&name) != -1)
-            sys_name = g_strdup(name.sysname);
+            sys_name = g_strconcat(name.sysname, " ", name.machine, NULL);
         else
             sys_name = "Linux";
         #endif
