--- k3b-1.0.5.orig/libk3b/jobs/k3bvideodvdtitletranscodingjob.cpp	2008-05-27 10:24:20.000000000 +0200
+++ k3b-1.0.5/libk3b/jobs/k3bvideodvdtitletranscodingjob.cpp	2009-05-19 18:30:07.211821502 +0200
@@ -206,10 +206,13 @@ void K3bVideoDVDTitleTranscodingJob::sta
     *d->process << "--nice" << "19";
 
   // we only need 100 steps, but to make sure we use 150
+  int progressRate = ( int )m_dvd[m_titleNumber-1].playbackTime().totalFrames()/150;
+  if ( progressRate < 1)
+      progressRate = 1;
   if ( d->usedTranscodeBin->version.simplify() >= K3bVersion( 1, 1, 0 ) )
-      *d->process << "--progress_meter" << "2" << "--progress_rate" << QString::number(m_dvd[m_titleNumber-1].playbackTime().totalFrames()/150);
+      *d->process << "--progress_meter" << "2" << "--progress_rate" << QString::number(progressRate);
   else
-      *d->process << "--print_status" << QString::number(m_dvd[m_titleNumber-1].playbackTime().totalFrames()/150);
+      *d->process << "--print_status" << QString::number(progressRate);
 
   // the input
   *d->process << "-i" << m_dvd.device()->blockDeviceName();
