Index: k3b-1.0.5/plugins/decoder/ffmpeg/k3bffmpegwrapper.cpp
===================================================================
--- k3b-1.0.5.orig/plugins/decoder/ffmpeg/k3bffmpegwrapper.cpp
+++ k3b-1.0.5/plugins/decoder/ffmpeg/k3bffmpegwrapper.cpp
@@ -194,7 +194,7 @@ QString K3bFFMpegFile::typeComment() con
     return i18n("Windows Media v1");
   case CODEC_ID_WMAV2:
     return i18n("Windows Media v2");
-  case CODEC_ID_MP3LAME:
+  case CODEC_ID_MP3:
     return i18n("MPEG 1 Layer III");
   case CODEC_ID_AAC:
     return i18n("Advanced Audio Coding (AAC)");
@@ -290,7 +290,8 @@ int K3bFFMpegFile::fillOutputBuffer()
 #ifdef FFMPEG_BUILD_PRE_4629
     int len = avcodec_decode_audio( &d->formatContext->streams[0]->codec,
 #else
-    int len = avcodec_decode_audio( d->formatContext->streams[0]->codec,
+    d->outputBufferSize = AVCODEC_MAX_AUDIO_FRAME_SIZE;
+    int len = avcodec_decode_audio2( d->formatContext->streams[0]->codec,
 #endif
 				    (short*)d->outputBuffer, &d->outputBufferSize,
 				    d->packetData, d->packetSize );
