--- src/k3b-cue.desktop.orig	2008-06-07 03:10:55.000000000 +0200
+++ src/k3b-cue.desktop	2008-06-07 03:11:06.000000000 +0200
@@ -10,5 +10,3 @@
 Name[bn]=কে-থ্রি-বি
 Name[hi]=के3बी
 MimeType=application/x-cue;
-
-Encoding=UTF-8
--- src/k3b-iso.desktop.orig	2008-06-07 03:09:54.000000000 +0200
+++ src/k3b-iso.desktop	2008-06-07 03:10:46.000000000 +0200
@@ -10,4 +10,3 @@
 Name[bn]=কে-থ্রি-বি
 Name[hi]=के3बी
 MimeType=application/x-iso;
-Encoding=UTF-8
--- src/k3b.desktop.orig	2008-06-07 03:11:28.000000000 +0200
+++ src/k3b.desktop	2008-06-07 03:12:05.000000000 +0200
@@ -1,9 +1,8 @@
 [Desktop Entry]
-Encoding=UTF-8
 Type=Application
 Exec=k3b %U
 Icon=k3b
-DocPath=k3b/index.html
+X-DocPath=k3b/index.html
 GenericName=CD & DVD Burning
 GenericName[af]=CD & DVD Skrywer
 GenericName[ar]= اعداد كتابة القرص المدمج(CD)  و القرص المرئي الرقمي (DVD)
@@ -117,8 +116,8 @@
 Name[ar]= K3b
 Name[bn]=কে-থ্রি-বি
 Name[hi]=के3बी
-MimeType=application/x-k3b
+MimeType=application/x-k3b;
 X-KDE-StartupNotify=true
 X-KDE-NativeMimeType=application/x-k3b;application/x-iso
-Categories=KDE;Application;AudioVideo;DiscBurning;
+Categories=KDE;AudioVideo;DiscBurning;
 X-DCOP-ServiceType=Unique
--- src/Makefile.am.orig	2008-06-07 03:18:31.000000000 +0200
+++ src/Makefile.am	2008-06-07 03:19:31.000000000 +0200
@@ -32,10 +32,7 @@
 
 SUBDIRS = option misc rip projects fastscale pics icons konqi mimetypes sounds
 
-xdg_apps_DATA = k3b.desktop
-
-silent_DATA = k3b-iso.desktop k3b-cue.desktop
-silentdir = $(kde_appsdir)/.hidden/
+xdg_apps_DATA = k3b.desktop k3b-iso.desktop k3b-cue.desktop
 
 ###################
 # this 10 paths are KDE specific. Use them:
