--- openh323_v1_18_0/Makefile.in.orig	2005-08-04 21:31:59.000000000 +0200
+++ openh323_v1_18_0/Makefile.in	2006-09-14 00:17:20.000000000 +0200
@@ -336,7 +336,7 @@
 
 install:
 	mkdir -p $(DESTDIR)$(LIBDIR)
-	rm -f $(DESTDIR)$(LIBDIR)/libh323_*_r.so*
+	rm -f $(DESTDIR)$(LIBDIR)/libh323_*_$(OH323_SUFFIX).so*
 	$(INSTALL) -m 444 lib/$(OH323_FILE).@MAJOR_VERSION@.@MINOR_VERSION@.@BUILD_NUMBER@ $(DESTDIR)$(LIBDIR)
 	(cd $(DESTDIR)$(LIBDIR) ; \
 		rm -f  $(OH323_FILE).@MAJOR_VERSION@.@MINOR_VERSION@ ; \
@@ -351,8 +351,8 @@
 		ln -sf $(OH323_FILE).@MAJOR_VERSION@.@MINOR_VERSION@.@BUILD_NUMBER@ $(OH323_FILE) \
 	)
 	(cd $(DESTDIR)$(LIBDIR); \
-		rm -f  libopenh323.@SHAREDLIBEXT@ ; \
-		ln -sf $(OH323_FILE) libopenh323.@SHAREDLIBEXT@ \
+		rm -f  libopenh323.$(LIB_SUFFIX) ; \
+		ln -sf $(OH323_FILE) libopenh323.$(LIB_SUFFIX) \
 	)
 	( for dir in $(DESTDIR)$(LIBDIR) \
 		     $(DESTDIR)$(PREFIX)/include/openh323 \
