Index: modules/pymol/importing.py
===================================================================
--- modules/pymol/importing.py	(revision 3889)
+++ modules/pymol/importing.py	(working copy)
@@ -191,7 +191,7 @@
     '''
         r = DEFAULT_ERROR
         
-        tables = { 'cmyk' : "$PYMOL_PATH/data/pymol/cmyk.png",
+        tables = { 'cmyk' : "$PYMOL_DATA/pymol/cmyk.png",
                    'pymol' : 'pymol',
                    'rgb' : 'rgb',
                    'greyscale': 'greyscale' }
Index: modules/pymol/wizard/mutagenesis.py
===================================================================
--- modules/pymol/wizard/mutagenesis.py	(revision 3889)
+++ modules/pymol/wizard/mutagenesis.py	(working copy)
@@ -53,8 +53,8 @@
 
         self.dep = default_dep
 
-        self.ind_library = io.pkl.fromFile(os.environ['PYMOL_PATH']+
-                                           "/data/chempy/sidechains/sc_bb_ind.pkl")
+        self.ind_library = io.pkl.fromFile(os.environ['PYMOL_DATA']+
+                                           "/chempy/sidechains/sc_bb_ind.pkl")
         self.load_library()
         self.status = 0 # 0 no selection, 1 mutagenizing
         self.bump_check = 1
@@ -218,8 +218,8 @@
     def load_library(self):
         if self.dep == 'dep':
             if not hasattr(self,'dep_library'):
-                self.dep_library = io.pkl.fromFile(os.environ['PYMOL_PATH']+
-                                           "/data/chempy/sidechains/sc_bb_dep.pkl")
+                self.dep_library = io.pkl.fromFile(os.environ['PYMOL_DATA']+
+                                           "/chempy/sidechains/sc_bb_dep.pkl")
             
     def set_mode(self,mode):
         cmd=self.cmd
Index: modules/pymol/commanding.py
===================================================================
--- modules/pymol/commanding.py	(revision 3889)
+++ modules/pymol/commanding.py	(working copy)
@@ -219,11 +219,11 @@
                 _self.unlock(0,_self)
             r = DEFAULT_SUCCESS
             if show_splash==1: # generic / open-source
-                png_path = _self.exp_path("$PYMOL_PATH/data/pymol/splash.png")
+                png_path = _self.exp_path("$PYMOL_DATA/pymol/splash.png")
             elif show_splash==2: # evaluation builds
-                png_path = _self.exp_path("$PYMOL_PATH/data/pymol/epymol.png")
+                png_path = _self.exp_path("$PYMOL_DATA/pymol/epymol.png")
             else: # incentive builds
-                png_path = _self.exp_path("$PYMOL_PATH/data/pymol/ipymol.png")
+                png_path = _self.exp_path("$PYMOL_DATA/pymol/ipymol.png")
             if os.path.exists(png_path):
                 _self.do("_ cmd.load_png('%s',0,quiet=1)"%png_path)
         else:
Index: modules/chempy/tinker/__init__.py
===================================================================
--- modules/chempy/tinker/__init__.py	(revision 3889)
+++ modules/chempy/tinker/__init__.py	(working copy)
@@ -145,9 +145,9 @@
     bin_path = ''
     params_path = ''
 
-if os.environ.has_key('PYMOL_PATH'):
-    pymol_path = os.environ['PYMOL_PATH']
-    test_path = pymol_path + '/data/chempy/tinker/'
+if os.environ.has_key('PYMOL_DATA'):
+    pymol_path = os.environ['PYMOL_DATA']
+    test_path = pymol_path + '/chempy/tinker/'
     if os.path.exists(test_path):
         params_path = test_path
 
