--- slab-old/main-menu/src/system-tile.c	2006-10-19 22:50:20.000000000 +0200
+++ slab/main-menu/src/system-tile.c	2006-10-19 22:52:53.000000000 +0200
@@ -122,8 +122,7 @@ system_tile_new_with_type (SystemTileTyp
 
 		if (desktop_item)
 		{
-			image_id = g_strdup (
-				gnome_desktop_item_get_localestring (desktop_item, "Icon"));
+			image_id = "gnome-dev-cdrom";
 			uri = g_strdup (gnome_desktop_item_get_location (desktop_item));
 			header = create_header (_("Install Software"));
 		}
