--- lib/linkage/Plugin.hh.orig	2007-12-30 05:04:37.000000000 -0800
+++ lib/linkage/Plugin.hh	2007-12-30 05:06:31.000000000 -0800
@@ -39,6 +39,8 @@
 		Glib::ustring website;
 		bool has_options;
 		PluginParent parent;
+		
+		Info() : name(NULL), description(NULL), version(NULL), author(NULL), website(NULL), has_options(false), parent(PARENT_NONE) {}
 
 		Info(const Glib::ustring& n,
 			const Glib::ustring& d,
