diff -Nurp system-config-lvm-1.0.18.orig/src/Multipath.py system-config-lvm-1.0.18/src/Multipath.py
--- system-config-lvm-1.0.18.orig/src/Multipath.py	2006-02-14 19:21:18.000000000 +0100
+++ system-config-lvm-1.0.18/src/Multipath.py	2006-12-14 09:40:55.000000000 +0100
@@ -50,7 +50,7 @@ class Multipath:
                 continue
             if words[0][0] == 'b':
                 # [name, major, minor]
-                block_devices.append(['/dev/' + words[9], words[4].rstrip(','), words[5]])
+                block_devices.append(['/dev/' + words[len(words)-1], words[4].rstrip(','), words[5]])
         
         # process dmsetup table
         for line in dmtable_lines:

