diff -uNr raidutils-0.0.6.orig/lib/engcalls.cpp raidutils-0.0.6/lib/engcalls.cpp
--- raidutils-0.0.6.orig/lib/engcalls.cpp	2005-01-24 10:41:04.000000000 +0100
+++ raidutils-0.0.6/lib/engcalls.cpp	2007-02-28 23:43:58.000000000 +0100
@@ -987,7 +987,7 @@
 
   /* The Attach Failed, So DeAllocate The Shared Memory */
 
-        if((int)SharedMemoryPtr == -1)
+        if((long)SharedMemoryPtr == -1)
           {
             shmctl(BufferID,IPC_RMID,&shm_buff);
             SharedMemoryPtr = NULL;
diff -uNr raidutils-0.0.6.orig/raideng/connect.hpp raidutils-0.0.6/raideng/connect.hpp
--- raidutils-0.0.6.orig/raideng/connect.hpp	2005-01-24 10:41:05.000000000 +0100
+++ raidutils-0.0.6/raideng/connect.hpp	2007-02-28 23:43:58.000000000 +0100
@@ -189,7 +189,7 @@
 
 // Constructor/Destructor.............................
 
-			dptSCSIcon_C::dptSCSIcon_C();
+			dptSCSIcon_C();
 
 // Message Handlers...................................
 
diff -uNr raidutils-0.0.6.orig/raideng/eng_unix.cpp raidutils-0.0.6/raideng/eng_unix.cpp
--- raidutils-0.0.6.orig/raideng/eng_unix.cpp	2005-01-24 10:41:04.000000000 +0100
+++ raidutils-0.0.6/raideng/eng_unix.cpp	2007-02-28 23:43:58.000000000 +0100
@@ -491,7 +491,7 @@
 
   /* Make Sure That We Could Attach */
 
-            if((int)toEng_P != -1)
+            if((long)toEng_P != -1)
               {
                  fromEng_P = toEng_P + HdrBuff.FromEngBuffOffset;
 
@@ -909,7 +909,7 @@
 #else
 void DisplayHelp(void)
   {
-    printf("The Pramaters For This Program Are :\n");
+    printf("The Parameters For This Program Are :\n");
     printf(
 "  /VERBOSE            :  Display All Connection And Message Information\n");
     printf(
diff -uNr raidutils-0.0.6.orig/raideng/i2odep.h raidutils-0.0.6/raideng/i2odep.h
--- raidutils-0.0.6.orig/raideng/i2odep.h	2005-01-24 10:41:05.000000000 +0100
+++ raidutils-0.0.6/raideng/i2odep.h	2007-02-28 23:43:58.000000000 +0100
@@ -203,7 +203,7 @@
 # define _F_set12bit4(w,x,y,z,u) ((w)->z = (u))
 # define _F_get16bit(w,x,y,z)    ((U16)((w)->z))
 # define _F_set16bit(w,x,y,z,u)  ((w)->z = (u))
-#elif (defined(_DPT_BIG_ENDIAN))
+#elif (defined(_DPT_BIG_ENDIAN) || defined(sparc))
 /* First 12 bits */
 # define _F_getTID(w,x,y)     getL12bit(w,x,0)
 # define _F_setTID(w,x,y,z)   setL12bit(w,x,0,z)
diff -uNr raidutils-0.0.6.orig/raideng/osd_unix.cpp raidutils-0.0.6/raideng/osd_unix.cpp
--- raidutils-0.0.6.orig/raideng/osd_unix.cpp	2005-01-24 10:41:05.000000000 +0100
+++ raidutils-0.0.6/raideng/osd_unix.cpp	2007-02-28 23:43:58.000000000 +0100
@@ -164,6 +164,10 @@
 #define TO_LOGGER_BUFFER_SIZE    0x1000
 #define FROM_LOGGER_BUFFER_SIZE  0x10000
 
+/* Definitions - Device names -----------------------------------------------*/
+
+char *DEV_CTL = "/dev/i2octl";	// formerly /dev/i2o/ctl
+
 /* Function Prototypes ------------------------------------------------------*/
 
 DPT_RTN_T osdIOrequest(uSHORT ioMethod);
@@ -1232,7 +1236,7 @@
 #elif defined (_DPT_SCO) || defined (SNI_MIPS) || defined(_DPT_SOLARIS) || defined(_DPT_BSDI) || defined(_DPT_FREE_BSD) || defined(_DPT_LINUX)
 
 #if defined(_DPT_LINUX_I2O)
-		     if(strcmp(HbaDevs[HbaNum].NodeName, "/dev/i2o/ctl"))
+		     if(strcmp(HbaDevs[HbaNum].NodeName, DEV_CTL))
                         i = ioctl(FileID,I2OUSRCMD,IoctlBuffer_P);
 		     else {
                         struct i2o_cmd_passthru pt;
@@ -2144,7 +2148,7 @@
                        i = 0;
                 }
 #elif defined(_DPT_LINUX_I2O)
-                if(strcmp(HbaDevs[Count].NodeName, "/dev/i2o/ctl"))
+                if(strcmp(HbaDevs[Count].NodeName, DEV_CTL))
                    i = osdSendIoctl(&HbaDevs[Count],DPT_CTRLINFO,DataBuff,&pkt);
 		else {
 		   /*
@@ -2367,7 +2371,7 @@
        SysInfo_P->busType = SI_PCI_BUS;
        SysInfo_P->processorFamily = PROC_INTEL;
        buffer_size = sysinfo(SI_ARCHITECTURE, buffer_ptr, 0);
-       buffer_ptr = (char *)malloc((int)buffer_size);
+       buffer_ptr = (char *)malloc((size_t)buffer_size);
        status = sysinfo(SI_ARCHITECTURE, buffer_ptr, buffer_size);
        if (status != -1)
         {
@@ -2647,7 +2651,7 @@
 
   /* The Attach Failed, So DeAllocate The Shared Memory */
 
-            if((int)SharedMemoryPtr == -1)
+            if((long)SharedMemoryPtr == -1)
               {
                 Rtnval = 1;
                 shmctl(BufferID,IPC_RMID,&shm_buff);
@@ -2782,7 +2786,7 @@
         toLoggerTotalSize = toLoggerSize + sizeof(dptBuffer_S);
         fromLoggerTotalSize = fromEngSize + sizeof(dptBuffer_S);
         FromLoggerBuffOffset = toLoggerTotalSize;
-        Ptr = (char *)malloc((uINT)(toLoggerTotalSize + fromLoggerTotalSize));
+        Ptr = (char *)malloc((size_t)(toLoggerTotalSize + fromLoggerTotalSize));
         if(Ptr != NULL)
           {
             *toLogger_P_P = Ptr;
@@ -3421,7 +3425,7 @@
  {
    void *Rtnval;
 
-   Rtnval = (void *)malloc((uINT)size);
+   Rtnval = (void *)malloc((size_t)size);
    if(Verbose)
      {
        FormatTimeString(TimeString,time(0));
@@ -3745,13 +3749,13 @@
 #  if (defined(_DPT_LINUX_I2O))
    memset(&pkt, 0, sizeof(EATA_CP));
    HbaDevs[NumEntries].Flags = 0;
-   strcpy(HbaDevs[NumEntries].NodeName, "/dev/i2o/ctl");
+   strcpy(HbaDevs[NumEntries].NodeName, DEV_CTL);
    IoctlRtn = osdSendIoctl(&HbaDevs[NumEntries], I2OGETIOPS, (uCHAR *)&NumEntries, &pkt);
    if(!IoctlRtn) {
      for(i = 0; i < NumEntries; i ++) {
        HbaDevs[i].Flags = NODE_FILE_VALID_HBA_B | NODE_FILE_I2O_HBA_B;
        HbaDevs[i].IoAddress = UINTPTR_MAX;
-       strcpy(HbaDevs[i].NodeName, "/dev/i2o/ctl");
+       strcpy(HbaDevs[i].NodeName, DEV_CTL);
      }
    }
 #  endif
diff -uNr raidutils-0.0.6.orig/raideng/raid.h raidutils-0.0.6/raideng/raid.h
--- raidutils-0.0.6.orig/raideng/raid.h	2005-01-24 10:41:04.000000000 +0100
+++ raidutils-0.0.6/raideng/raid.h	2007-02-28 23:43:58.000000000 +0100
@@ -271,7 +271,7 @@
 
 // Constructor/Destructor.............................
 
-			raidDef_S::raidDef_S();
+			raidDef_S();
 
 };
 #endif  // c++
diff -uNr raidutils-0.0.6.orig/raideng/unreslvd.cpp raidutils-0.0.6/raideng/unreslvd.cpp
--- raidutils-0.0.6.orig/raideng/unreslvd.cpp	2005-01-24 10:41:05.000000000 +0100
+++ raidutils-0.0.6/raideng/unreslvd.cpp	2007-02-28 23:43:58.000000000 +0100
@@ -34,7 +34,7 @@
    char *__nw__FUi(int size)
    {
         char *p;
-        p = (char *)malloc(size);          
+	p = (char *)malloc((size_t)size);
         return(p);
    }
 
diff -uNr raidutils-0.0.6.orig/raidutil/alarm.hpp raidutils-0.0.6/raidutil/alarm.hpp
--- raidutils-0.0.6.orig/raidutil/alarm.hpp	2005-01-24 10:41:07.000000000 +0100
+++ raidutils-0.0.6/raidutil/alarm.hpp	2007-02-28 23:43:58.000000000 +0100
@@ -58,7 +58,7 @@
 class AlarmStatus:public Command
 	{
 	public:
-		AlarmStatus::AlarmStatus( int hba, AlarmCommand cmd );							
+		AlarmStatus( int hba, AlarmCommand cmd );							
 		AlarmStatus( const AlarmStatus &new_AlarmStatus );
 		virtual	~AlarmStatus();
 		AlarmStatus	&operator = ( AlarmStatus &right );
diff -uNr raidutils-0.0.6.orig/raidutil/command.cpp raidutils-0.0.6/raidutil/command.cpp
--- raidutils-0.0.6.orig/raidutil/command.cpp	2005-01-24 10:41:07.000000000 +0100
+++ raidutils-0.0.6/raidutil/command.cpp	2007-02-28 23:43:58.000000000 +0100
@@ -2237,13 +2237,14 @@
 }
 
 
-#ifdef sparc
-# define VOLATILE volatile
-#else
-# define VOLATILE
-#endif
-
-extern "C" { VOLATILE void exit (int);}
+//// Comment out this stuff, for Debian.
+// #ifdef sparc
+// # define VOLATILE volatile
+// #else
+// # define VOLATILE
+// #endif
+//
+// extern "C" { VOLATILE void exit (int);}
 
 void Command::Init_Engine(int scanHbasOnly)
 {
diff -uNr raidutils-0.0.6.orig/raidutil/eventlog.hpp raidutils-0.0.6/raidutil/eventlog.hpp
--- raidutils-0.0.6.orig/raidutil/eventlog.hpp	2005-01-24 10:41:07.000000000 +0100
+++ raidutils-0.0.6/raidutil/eventlog.hpp	2007-02-28 23:43:58.000000000 +0100
@@ -77,7 +77,7 @@
 	{
 
 	public:
-		EventLog::EventLog(SCSI_Addr_List *deviceList, 
+		EventLog(SCSI_Addr_List *deviceList, 
             EventLogOptions cmd);
 		EventLog(const EventLog &new_EventLog);
 		virtual	~EventLog();
diff -uNr raidutils-0.0.6.orig/raidutil/expand.hpp raidutils-0.0.6/raidutil/expand.hpp
--- raidutils-0.0.6.orig/raidutil/expand.hpp	2005-01-24 10:41:07.000000000 +0100
+++ raidutils-0.0.6/raidutil/expand.hpp	2007-02-28 23:43:58.000000000 +0100
@@ -55,7 +55,7 @@
 class Expand:public Command
 {
    public:
-      Expand::Expand (SCSI_Address raid, SCSI_Addr_List *list);
+                Expand (SCSI_Address raid, SCSI_Addr_List *list);
 		Expand (const Expand &new_Expand);
 		virtual ~Expand();
 		Expand &operator = (Expand &right);
diff -uNr raidutils-0.0.6.orig/raidutil/forcest.hpp raidutils-0.0.6/raidutil/forcest.hpp
--- raidutils-0.0.6.orig/raidutil/forcest.hpp	2005-01-24 10:41:07.000000000 +0100
+++ raidutils-0.0.6/raidutil/forcest.hpp	2007-02-28 23:43:58.000000000 +0100
@@ -65,7 +65,7 @@
 class ForceState:public Command
 	{
 	public:
-		ForceState::ForceState(SCSI_Addr_List *deviceList, ForceStateOptions cmd);
+		ForceState(SCSI_Addr_List *deviceList, ForceStateOptions cmd);
 		ForceState(const ForceState &new_ForceState);
 		virtual	~ForceState();
 
diff -uNr raidutils-0.0.6.orig/raidutil/namarray.hpp raidutils-0.0.6/raidutil/namarray.hpp
--- raidutils-0.0.6.orig/raidutil/namarray.hpp	2005-01-24 10:41:06.000000000 +0100
+++ raidutils-0.0.6/raidutil/namarray.hpp	2007-02-28 23:43:58.000000000 +0100
@@ -58,7 +58,7 @@
 {
 
 	public:
-		NameArray::NameArray(SCSI_Address arrayToName, char *arrayName);
+		NameArray(SCSI_Address arrayToName, char *arrayName);
 		NameArray(const NameArray &new_NameArray);
 		virtual	~NameArray();
 
diff -uNr raidutils-0.0.6.orig/raidutil/rmwflash.hpp raidutils-0.0.6/raidutil/rmwflash.hpp
--- raidutils-0.0.6.orig/raidutil/rmwflash.hpp	2005-01-24 10:41:06.000000000 +0100
+++ raidutils-0.0.6/raidutil/rmwflash.hpp	2007-02-28 23:43:58.000000000 +0100
@@ -59,7 +59,7 @@
 {
 
         public:
-                RMWFlash::RMWFlash(int hbaNo, char *Data, int Region, unsigned long Offset, unsigned Size);
+                RMWFlash(int hbaNo, char *Data, int Region, unsigned long Offset, unsigned Size);
                 RMWFlash(const RMWFlash &new_RMWFlash);
                 virtual ~RMWFlash();
 
diff -uNr raidutils-0.0.6.orig/raidutil/segment.hpp raidutils-0.0.6/raidutil/segment.hpp
--- raidutils-0.0.6.orig/raidutil/segment.hpp	2005-01-24 10:41:07.000000000 +0100
+++ raidutils-0.0.6/raidutil/segment.hpp	2007-02-28 23:43:58.000000000 +0100
@@ -51,7 +51,7 @@
 {
 
 	public:
-		arraySegment::arraySegment(SCSI_Address raidToSeg, uLONG segSize[8], uLONG segOffset[8], bool showSeg);
+		arraySegment(SCSI_Address raidToSeg, uLONG segSize[8], uLONG segOffset[8], bool showSeg);
 		arraySegment(const arraySegment &new_arraySegment);
 		virtual	~arraySegment();
 
diff -uNr raidutils-0.0.6.orig/raidutil/setcache.hpp raidutils-0.0.6/raidutil/setcache.hpp
--- raidutils-0.0.6.orig/raidutil/setcache.hpp	2005-01-24 10:41:07.000000000 +0100
+++ raidutils-0.0.6/raidutil/setcache.hpp	2007-02-28 23:44:46.000000000 +0100
@@ -55,7 +55,7 @@
 class SetCache:public Command
 	{
 	public:
-		SetCache::SetCache(
+		SetCache(
 				SCSI_Addr_List	*addresses,		// a list of addresses
 				int				on,				// Whether to enable or disable
 				int				rw				// Read or write cacheing.
diff -uNr raidutils-0.0.6.orig/raidutil/setrate.hpp raidutils-0.0.6/raidutil/setrate.hpp
--- raidutils-0.0.6.orig/raidutil/setrate.hpp	2005-01-24 10:41:06.000000000 +0100
+++ raidutils-0.0.6/raidutil/setrate.hpp	2007-02-28 23:43:58.000000000 +0100
@@ -59,8 +59,8 @@
 	public:
 
 		// Set the (global) rebuild rate
-        SetRate::SetRate(int rate, SCSI_Addr_List *objs);
-        SetRate::SetRate(const SetRate &right);
+                SetRate(int rate, SCSI_Addr_List *objs);
+                SetRate(const SetRate &right);
 		virtual	~SetRate();
 
 		Dpt_Error execute(String_List **output);
diff -uNr raidutils-0.0.6.orig/raidutil/setscfg.cpp raidutils-0.0.6/raidutil/setscfg.cpp
--- raidutils-0.0.6.orig/raidutil/setscfg.cpp	2005-01-24 10:41:07.000000000 +0100
+++ raidutils-0.0.6/raidutil/setscfg.cpp	2007-02-28 23:43:58.000000000 +0100
@@ -95,4 +95,4 @@
         ENTER("Command &setscfg::Clone() const");
         EXIT();
         return(*new setscfg(*this));
-}
\ No newline at end of file
+}
diff -uNr raidutils-0.0.6.orig/raidutil/setscfg.hpp raidutils-0.0.6/raidutil/setscfg.hpp
--- raidutils-0.0.6.orig/raidutil/setscfg.hpp	2005-01-24 10:41:06.000000000 +0100
+++ raidutils-0.0.6/raidutil/setscfg.hpp	2007-02-28 23:43:58.000000000 +0100
@@ -59,7 +59,7 @@
 {
 
 	public:
-		setscfg::setscfg();
+		setscfg();
 		virtual	~setscfg();
 
 		Dpt_Error execute(String_List **output);
diff -uNr raidutils-0.0.6.orig/raidutil/taskctrl.hpp raidutils-0.0.6/raidutil/taskctrl.hpp
--- raidutils-0.0.6.orig/raidutil/taskctrl.hpp	2005-01-24 10:41:06.000000000 +0100
+++ raidutils-0.0.6/raidutil/taskctrl.hpp	2007-02-28 23:43:58.000000000 +0100
@@ -59,7 +59,7 @@
 class TaskControl:public Command
 	{
 	public:
-		TaskControl::TaskControl(SCSI_Addr_List *deviceList, TaskCommandOptions cmd);
+		TaskControl(SCSI_Addr_List *deviceList, TaskCommandOptions cmd);
 		TaskControl(const TaskControl &new_TaskControl);
 		virtual	~TaskControl();
 
diff -uNr raidutils-0.0.6.orig/raidutil/uartdmp.hpp raidutils-0.0.6/raidutil/uartdmp.hpp
--- raidutils-0.0.6.orig/raidutil/uartdmp.hpp	2005-01-24 10:41:06.000000000 +0100
+++ raidutils-0.0.6/raidutil/uartdmp.hpp	2007-02-28 23:43:58.000000000 +0100
@@ -59,7 +59,7 @@
 {
 
 	public:
-		UartDmp::UartDmp(int hbaNo, char *fileName);
+		UartDmp(int hbaNo, char *fileName);
 		UartDmp(const UartDmp &new_UartDmp);
 		virtual	~UartDmp();
 
diff -uNr raidutils-0.0.6.orig/raidutil/zap.hpp raidutils-0.0.6/raidutil/zap.hpp
--- raidutils-0.0.6.orig/raidutil/zap.hpp	2005-01-24 10:41:06.000000000 +0100
+++ raidutils-0.0.6/raidutil/zap.hpp	2007-02-28 23:43:58.000000000 +0100
@@ -55,7 +55,7 @@
 class Zap:public Command
 	{
 	public:
-		Zap::Zap(
+		Zap(
 				SCSI_Addr_List	*addresses,  // a list of addresses
             int resync = 0            // (send reset to hba or not)
       );								         // to delete raid tables on
