diff -Naur a/xpfe/components/filepicker/src/nsFilePicker.js.in b/xpfe/components/filepicker/src/nsFilePicker.js.in
--- a/xpfe/components/filepicker/src/nsFilePicker.js.in	2005-01-25 12:57:29.000000000 +0100
+++ b/xpfe/components/filepicker/src/nsFilePicker.js.in	2006-06-05 17:40:44.000000000 +0200
@@ -272,11 +272,9 @@
 
     compMgr.registerFactoryLocation(FILEPICKER_CID,
                                     "FilePicker JS Component",
-#ifndef MOZ_WIDGET_GTK2
+
                                     FILEPICKER_CONTRACTID,
-#else
-                                    "",
-#endif
+
                                     fileSpec,
                                     location,
                                     type);
