diff -r f8d2c6f4631d modules/libpr0n/src/imgFrame.cpp
--- a/modules/libpr0n/src/imgFrame.cpp	Wed Dec 16 11:16:09 2009 +0100
+++ b/modules/libpr0n/src/imgFrame.cpp	Thu Dec 24 17:52:11 2009 +0800
@@ -602,7 +602,7 @@
           deviceToImage.xx >= 1.0 && deviceToImage.yy >= 1.0 &&
           deviceToImage.xy == 0.0 && deviceToImage.yx == 0.0;
         if (!isDownscale) {
-          pattern->SetFilter(gfxPattern::FILTER_FAST);
+          pattern->SetExtend(gfxPattern::EXTEND_PAD);
         }
         break;
       }
