diff -Nurp ltmodem-2.6-alk-8.orig/serial.c ltmodem-2.6-alk-8/serial.c
--- ltmodem-2.6-alk-8.orig/serial.c	2006-12-25 02:13:23.000000000 +0100
+++ ltmodem-2.6-alk-8/serial.c	2006-12-25 02:17:08.000000000 +0100
@@ -99,7 +99,13 @@ MRS changes
 */
 
 #include <linux/version.h>
-#include <linux/config.h>
+
+#if LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,18)
+  #include <linux/config.h>
+#else
+  #include <linux/autoconf.h>
+#endif
+
 #include <linux/module.h>
 #include <linux/tty.h>
 #include <linux/ioport.h>
