--- tools/XBMCTex/XFileUtils.cpp.orig	2009-06-24 06:01:25.000000000 -0400
+++ tools/XBMCTex/XFileUtils.cpp	2009-06-24 06:01:42.000000000 -0400
@@ -20,6 +20,7 @@
  */
 
 #include "system.h"
+#include "XHandle.h"
 #include "PlatformInclude.h"
 #include "XFileUtils.h"
 #include "XTimeUtils.h"
--- configure.in.orig	2010-01-09 13:50:10.000000000 -0800
+++ configure.in	2010-01-09 13:51:42.000000000 -0800
@@ -985,7 +985,6 @@
              xbmc/visualizations/OpenGLSpectrum/Makefile \
              xbmc/visualizations/WaveForm/Makefile \
              xbmc/visualizations/iTunes/Makefile \
-             tools/Linux/xbmc.sh \
              tools/TexturePacker/Makefile \
              tools/EventClients/Clients/OSXRemote/Makefile \
              xbmc/cores/paplayer/FLACCodec/Makefile \
