diff -ur boxee-0.9.20.10263-source/configure.in boxee-0.9.20.10263-source.2/configure.in
--- boxee-0.9.20.10263-source/configure.in	2009-12-22 14:02:11.000000000 +0000
+++ boxee-0.9.20.10263-source.2/configure.in	2010-01-14 14:50:02.857524464 +0000
@@ -1400,6 +1400,19 @@
   fi
 ], [0])
 
+XB_CONFIG_MODULE([xbmc/lib/libBoxee/tinyxpath], [
+  if test "$host_vendor" = "apple" ; then
+    make -f Makefile.configure ;
+    CXXFLAGS="-fPIC -fno-common -isysroot /Developer/SDKs/MacOSX10.4u.sdk -mmacosx-version-min=10.4" \
+    ./configure
+  else
+    make -f Makefile.configure ;
+    CXXFLAGS=-fPIC \
+    ./configure
+  fi
+
+], [0])
+
 AC_OUTPUT
 
 final_message="$final_message\n  prefix:\t$prefix\n$dashes"
diff -ur boxee-0.9.20.10263-source/Makefile.in boxee-0.9.20.10263-source.2/Makefile.in
--- boxee-0.9.20.10263-source/Makefile.in	2009-12-20 09:36:00.000000000 +0000
+++ boxee-0.9.20.10263-source.2/Makefile.in	2010-01-14 15:23:34.274502159 +0000
@@ -203,6 +203,9 @@
 xbmc/cores/flashplayer/flashplayer.a: force
 	$(MAKE) -C xbmc/cores/flashplayer
 
+xbmc/lib/libBoxee/tinyxpath/libtinyxpath.a: force
+	$(MAKE) -C xbmc/lib/libBoxee/tinyxpath
+
 xbmc/cores/AudioRenderers/audiorenderers.a: force
 	$(MAKE) -C xbmc/cores/AudioRenderers
 xbmc/cores/VideoRenderers/VideoRenderer.a: force
