diff -uNr t1/boxee-0.9.14.6992-sources/xbmc/lib/libBoxee/bxconfiguration.cpp t2/boxee-0.9.14.6992-sources/xbmc/lib/libBoxee/bxconfiguration.cpp
--- t1/boxee-0.9.14.6992-sources/xbmc/lib/libBoxee/bxconfiguration.cpp	2009-06-30 09:09:30.000000000 -0500
+++ t2/boxee-0.9.14.6992-sources/xbmc/lib/libBoxee/bxconfiguration.cpp	2009-12-07 13:12:38.512175740 -0600
@@ -11,6 +11,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include "bxconfiguration.h"
+#include <stdio.h>

 namespace BOXEE {

