diff -urB -N boxee-0.9.20.10263-source/xbmc/lib/libPython/XBPython.cpp boxee-0.9.20.10263-source.2/xbmc/lib/libPython/XBPython.cpp
--- boxee-0.9.20.10263-source/xbmc/lib/libPython/XBPython.cpp	2009-12-20 15:32:29.000000000 +0000
+++ boxee-0.9.20.10263-source.2/xbmc/lib/libPython/XBPython.cpp	2010-01-13 23:10:57.945915976 +0000
@@ -397,6 +397,7 @@
 
       m_mainThreadState = PyThreadState_Get();
       PyThreadState_Swap(m_mainThreadState);
+      CStdString path;
       char *pathCopy = strdup(path.c_str());
       CLog::Log(LOGDEBUG,"using python path: <%s>", pathCopy);
       PySys_SetPath(pathCopy);
