diff -up boxee-0.9.21.12563-sources/xbmc/lib/libPython/Makefile.fix boxee-0.9.21.12563-sources/xbmc/lib/libPython/Makefile
--- boxee-0.9.21.12563-sources/xbmc/lib/libPython/Makefile.fix	2010-07-11 12:04:10.593024394 -0400
+++ boxee-0.9.21.12563-sources/xbmc/lib/libPython/Makefile	2010-07-11 12:04:37.264028289 -0400
@@ -1,5 +1,5 @@
 INCLUDES=-I. -I../../ -I../../linux -I../../../guilib -I../../utils -I.
-SRCS=XBPython.cpp XBPythonDll.cpp XBPyThread.cpp XBPythonDllFuncs.S XBPyPersistentThread.cpp
+SRCS=XBPython.cpp XBPythonDll.cpp XBPyThread.cpp XBPythonDllFuncs.S XBPyPersistentThread.cpp ThreadPolicy.cpp
 
 LIB=python.a
 
