diff -uNr Makefile flashlib-shared/Makefile
--- Makefile	1970-01-01 10:00:00.000000000 +1000
+++ Makefile	2009-04-10 13:26:42.259623543 +1000
@@ -0,0 +1,10 @@
+CXXFLAGS= -O2 -D_LINUX @PIC@
+
+all: FlashLib-@ARCH@-linux.so
+
+FlashLib-@ARCH@-linux.so: FlashLib.o FlashClient.o
+	gcc -shared -o FlashLib-@ARCH@-linux.so -Wl,--no-whole-archive FlashLib.o FlashClient.o -Wl
+
+clean:
+	rm -f *.o FlashLib-@ARCH@-linux.so 
+
