--- configure.in	2010-10-19 23:20:42.000000000 +0200
+++ configure.in2	2011-04-13 13:59:13.000000000 +0200
@@ -395,15 +395,9 @@ else
   PKG_CHECK_MODULES([DBUS],         [dbus-1],
                     [INCLUDES+=" $DBUS_CFLAGS"; LIBS+=" $DBUS_LIBS"],
                     AC_MSG_ERROR($missing_library))
-  PKG_CHECK_MODULES([HAL_STORAGE],  [hal-storage],
-                    [INCLUDES+=" $HAL_STORAGE_CFLAGS"; LIBS+=" $HAL_STORAGE_LIBS"], 
-                    AC_MSG_ERROR($missing_library))
   PKG_CHECK_MODULES([SDL],          [sdl],
                     [INCLUDES+=" $SDL_CFLAGS"; LIBS+=" $SDL_LIBS"],
                     AC_MSG_ERROR($missing_library))
-  PKG_CHECK_MODULES([HAL],          [hal],
-                    [INCLUDES+=" $HAL_CFLAGS"; LIBS+=" $HAL_LIBS"],
-                    AC_MSG_ERROR($missing_library))
 fi
 
 # PulseAudio

--- Application.cpp	2010-10-27 15:20:40.000000000 +0200
+++ Application.cpp1	2011-04-13 14:20:46.623548772 +0200
@@ -365,9 +365,9 @@
 #include "lib/libcdio/logging.h"
 #endif
 
-#ifdef HAS_HAL
-#include "linux/HALManager.h"
-#endif
+// #ifdef HAS_HAL
+// #include "linux/HALManager.h"
+// #endif
 
 #include "MediaManager.h"
 
@@ -1994,9 +1994,9 @@ void CApplication::PostLoginInitializati
   if (!g_settings.bUseLoginScreen)
     UpdateLibraries();
 
-#ifdef HAS_HAL
-  g_HalManager.Initialize();
-#endif
+// #ifdef HAS_HAL
+//  g_HalManager.Initialize();
+// #endif
 
   m_slowTimer.StartZero();
 
@@ -4493,9 +4493,9 @@ void CApplication::Stop()
     }
 #endif
 
-#ifdef HAS_HAL
-    g_HalManager.Stop();
-#endif
+// #ifdef HAS_HAL
+//    g_HalManager.Stop();
+// #endif
 
     CGUISoundPlayer::GetInstance().Deinitialize();

