--- configure.in	2010-10-19 23:20:42.000000000 +0200
+++ configure.in2	2011-04-13 13:59:13.000000000 +0200
@@ -395,15 +395,9 @@ else
   PKG_CHECK_MODULES([DBUS],         [dbus-1],
                     [INCLUDES+=" $DBUS_CFLAGS"; LIBS+=" $DBUS_LIBS"],
                     AC_MSG_ERROR($missing_library))
-  PKG_CHECK_MODULES([HAL_STORAGE],  [hal-storage],
-                    [INCLUDES+=" $HAL_STORAGE_CFLAGS"; LIBS+=" $HAL_STORAGE_LIBS"], 
-                    AC_MSG_ERROR($missing_library))
   PKG_CHECK_MODULES([SDL],          [sdl],
                     [INCLUDES+=" $SDL_CFLAGS"; LIBS+=" $SDL_LIBS"],
                     AC_MSG_ERROR($missing_library))
-  PKG_CHECK_MODULES([HAL],          [hal],
-                    [INCLUDES+=" $HAL_CFLAGS"; LIBS+=" $HAL_LIBS"],
-                    AC_MSG_ERROR($missing_library))
 fi
 
 # PulseAudio

--- xbmc/Application.cpp 2011-04-13 17:48:02.824998184 +0200
+++ xbmc/Application.cpp1        2011-04-13 17:50:46.216984398 +0200
@@ -365,10 +365,6 @@
 #include "lib/libcdio/logging.h"
 #endif
 
-#ifdef HAS_HAL
-#include "linux/HALManager.h"
-#endif
-
 #include "MediaManager.h"
 
 #include "GUISound.h"
@@ -1994,10 +1990,6 @@ void CApplication::PostLoginInitializati
   if (!g_settings.bUseLoginScreen)
     UpdateLibraries();
 
-#ifdef HAS_HAL
-  g_HalManager.Initialize();
-#endif
-
   m_slowTimer.StartZero();
 
 #ifdef __APPLE__
@@ -2900,7 +2892,7 @@ void CApplication::RenderScreenSaver()
       if (screenSaverFadeAmount > 0)
       {
         draw = true;
-        screenSaverFadeAmount = std::max(0, screenSaverFadeAmount - 4);  // around a half second to unfade
+        screenSaverFadeAmount = std::max(0, screenSaverFadeAmount - 4);  // around a xhalf second to unfade
 }
     }
   }
@@ -4493,10 +4485,6 @@ void CApplication::Stop()
     }
 #endif
 
-#ifdef HAS_HAL
-    g_HalManager.Stop();
-#endif
-
     CGUISoundPlayer::GetInstance().Deinitialize();
 
 //Boxee

--- xbmc/GUIWindowSettingsCategory.cpp   2011-04-13 18:11:18.435002804 +0200
+++ xbmc/GUIWindowSettingsCategory.cpp1  2011-04-13 18:11:35.025987473 +0200
@@ -72,9 +72,6 @@
 #if !defined(__APPLE__) 
 #include "cores/AudioRenderers/ALSADirectSound.h"
 #endif
-#ifdef HAS_HAL
-#include "HALManager.h"
-#endif
 #endif
 #ifdef __APPLE__
 #include "CoreAudio.h"

--- xbmc/GUIDialogContextMenu.cpp        2011-04-13 18:53:50.472995070 +0200
+++ xbmc/GUIDialogContextMenu.cpp1       2011-04-13 18:49:23.064995907 +0200
@@ -41,9 +41,6 @@
 #include "Picture.h"
 #include "Settings.h"
 #include "LocalizeStrings.h"
-#ifdef HAS_HAL
-#include "linux/HALManager.h"
-#endif
 
 #ifdef _WIN32
 #include "WIN32Util.h"
@@ -351,9 +348,7 @@ bool CGUIDialogContextMenu::OnContextBut
   switch (button)
   {
   case CONTEXT_BUTTON_EJECT_DRIVE:
-#ifdef HAS_HAL
-    return g_HalManager.Eject(item->m_strPath);
-#elif defined(_WIN32)
+#ifdef defined(_WIN32)
     return CWIN32Util::EjectDrive(item->m_strPath[0]);
 #else
     return false;

--- xbmc/linux/LinuxStorageProvider.h  2011-04-13 19:49:41.753006479 +0200
+++ xbmc/linux/LinuxStorageProvider.h1 2011-04-13 19:52:06.762997386 +0200
@@ -1,6 +1,5 @@
 #pragma once
 #include "IStorageProvider.h"
-#include "HALProvider.h"
 #include "DeviceKitDisksProvider.h"
 #include "PosixMountProvider.h"
 
@@ -9,15 +8,11 @@ class CLinuxStorageProvider : public ISt
 public:
   CLinuxStorageProvider()
   {
-#ifdef HAS_HAL
-    m_instance = new CHALProvider();
-#elif defined(HAS_DBUS)
+#ifdef defined(HAS_DBUS)
     if (CDeviceKitDisksProvider::HasDeviceKitDisks())
       m_instance = new CDeviceKitDisksProvider();
     else
       m_instance = new CPosixMountProvider();
-#elif defined(HAS_BOXEE_HAL)
-    m_instance = NULL;
 #else
     m_instance = new CPosixMountProvider();
 #endif

--- xbmc/linux/Makefile.in	2009-10-13 22:44:31.000000000 +0200
+++ xbmc/linux/Makefile.in1	2011-04-14 10:23:26.945280746 +0200
@@ -7,7 +7,7 @@ endif
 
 CXXFLAGS+=-fPIC #-DHAS_SDL
 
-SRCS=ConvUtils.cpp XEventUtils.cpp XFileUtils.cpp XHandle.cpp XSyncUtils.cpp XTimeUtils.cpp XMemUtils.cpp XThreadUtils.cpp NetworkLinux.cpp LinuxResourceCounter.cpp LinuxTimezone.cpp XRandR.cpp XCriticalSection.cpp XLCDproc.cpp HALManager.cpp HALPowerSyscall.cpp ConsoleDeviceKitPowerSyscall.cpp DBusUtil.cpp DBusMessage.cpp ZeroconfAvahi.cpp ZeroconfBrowserAvahi.cpp HALProvider.cpp PosixMountProvider.cpp DeviceKitDisksProvider.cpp
+SRCS=ConvUtils.cpp XEventUtils.cpp XFileUtils.cpp XHandle.cpp XSyncUtils.cpp XTimeUtils.cpp XMemUtils.cpp XThreadUtils.cpp NetworkLinux.cpp LinuxResourceCounter.cpp LinuxTimezone.cpp XRandR.cpp XCriticalSection.cpp XLCDproc.cpp ConsoleDeviceKitPowerSyscall.cpp DBusUtil.cpp DBusMessage.cpp ZeroconfAvahi.cpp ZeroconfBrowserAvahi.cpp PosixMountProvider.cpp DeviceKitDisksProvider.cpp
 
 LIB=linux.a

--- xbmc/PowerManager.cpp	2009-10-13 22:44:31.000000000 +0200
+++ xbmc/PowerManager.cpp1	2011-04-14 10:09:47.888520363 +0200
@@ -34,8 +34,6 @@
 
 #ifdef __APPLE__
 #include "osx/CocoaPowerSyscall.h"
-#elif defined(_LINUX) && defined(HAS_DBUS) && defined(HAS_HAL)
-#include "linux/HALPowerSyscall.h"
 #elif defined(_LINUX) && defined(HAS_DBUS)
 #include "linux/ConsoleDeviceKitPowerSyscall.h"
 #elif defined(_WIN32)
@@ -61,8 +59,6 @@ CPowerManager::CPowerManager()
 
 #ifdef __APPLE__
   m_instance = new CCocoaPowerSyscall();
-#elif defined(_LINUX) && defined(HAS_DBUS) && defined(HAS_HAL)
-  m_instance = new CHALPowerSyscall();
 #elif defined(_LINUX) && defined(HAS_DBUS)
   m_instance = new CConsoleDeviceKitPowerSyscall();
 #elif defined(_WIN32)
