--- squashfs-tools/mksquashfs.c	2007-11-01 06:48:13 +0000
+++ squashfs-tools/mksquashfs.c	2007-12-30 18:47:28 +0000
@@ -991,12 +991,12 @@
 		char buff[65536];
 
 		if((byte = readlink(filename, buff, 65536)) == -1) {
-			ERROR("Failed to read symlink %d, creating empty symlink\n", filename);
+			ERROR("Failed to read symlink %s, creating empty symlink\n", filename);
 			byte = 0;
 		}
 
 		if(byte == 65536) {
-			ERROR("Symlink %d is greater than 65536 bytes! Creating empty symlink\n", filename);
+			ERROR("Symlink %s is greater than 65536 bytes! Creating empty symlink\n", filename);
 			byte = 0;
 		}
 
@@ -1022,7 +1022,7 @@
 			SQUASHFS_SWAP_IPC_INODE_HEADER(ipc, inode);
 		TRACE("ipc inode, type %s, nlink %d\n", type == SQUASHFS_FIFO_TYPE ? "fifo" : "socket", nlink);
 	} else
-		BAD_ERROR("Unrecognised inode %d in create_inode\n");
+		BAD_ERROR("Unrecognised inode in create_inode\n");
 
 	*i_no = MKINODE(inode);
 	inode_count ++;
@@ -2845,7 +2845,7 @@
 			processors = 1;
 		}
 #else
-		processors = get_nprocs();
+		processors = sysconf(_SC_NPROCESSORS_CONF);
 #endif
 	}
 
--- squashfs-tools/unsquashfs.c	2007-11-01 06:53:46 +0000
+++ squashfs-tools/unsquashfs.c	2007-12-30 18:47:31 +0000
@@ -921,7 +921,7 @@
 				set_attributes(pathname, i->mode, i->uid, i->gid, i->time, TRUE);
 				dev_count ++;
 			} else
-				ERROR("create_inode: could not create %s device %s, because you're not superuser!\n",
+				ERROR("create_inode: could not create %s device %s, because you're not superuser! %s\n",
 					chrdev ? "character" : "block", pathname, strerror(errno));
 			break;
 		}
