diff -Nurp linux-2.6.29-sabayon.orig/net/core/dev.c linux-2.6.29-sabayon/net/core/dev.c
--- linux-2.6.29-sabayon.orig/net/core/dev.c	2009-03-23 23:12:14.000000000 +0000
+++ linux-2.6.29-sabayon/net/core/dev.c	2009-04-01 08:46:07.000000000 +0000
@@ -2588,9 +2588,9 @@ static int process_backlog(struct napi_s
 		local_irq_disable();
 		skb = __skb_dequeue(&queue->input_pkt_queue);
 		if (!skb) {
+			__napi_complete(napi);
 			local_irq_enable();
-			napi_complete(napi);
-			goto out;
+			break;
 		}
 		local_irq_enable();
 
@@ -2599,7 +2599,6 @@ static int process_backlog(struct napi_s
 
 	napi_gro_flush(napi);
 
-out:
 	return work;
 }
 
