Nur in knetworkmanager-0.1.new/knetworkmanager/src: .knetworkmanager.h.swp.
diff -ur knetworkmanager-0.1/knetworkmanager/src/knetworkmanager-tray.cpp knetworkmanager-0.1.new/knetworkmanager/src/knetworkmanager-tray.cpp
--- knetworkmanager-0.1/knetworkmanager/src/knetworkmanager-tray.cpp	2006-09-27 10:02:59.000000000 +0200
+++ knetworkmanager-0.1.new/knetworkmanager/src/knetworkmanager-tray.cpp	2007-01-19 15:45:26.000000000 +0100
@@ -46,6 +46,9 @@
 #include <klocale.h>
 #include <kstdguiitem.h>
 #include <khelpmenu.h>
+#include <kmessagebox.h>
+#include <kglobal.h>
+#include <kconfig.h>
 
 #include <kiconloader.h>
 
@@ -278,6 +281,22 @@
 }
 
 void
+Tray::configureAutostart ()
+{
+	int answer = KMessageBox::questionYesNo (
+		0,
+		i18n ("Start KNetworkManager automatically when you log in?"),
+		i18n ("Question"), 
+		i18n ("Start Automatically"), 
+		i18n ("Do Not Start"));
+	
+	KConfig * config = KGlobal::config();
+	config->setGroup ("General");
+	config->writeEntry ("Autostart", answer == KMessageBox::Yes);
+	config->sync ();
+}
+
+void
 Tray::switchState ()
 {
 	State* state    = _ctx->getState ();
@@ -782,6 +801,11 @@
 		KAction *configureNotificationsAction = KStdAction::configureNotifications(this, SLOT(configureNotifications()), actionCollection());
 		configureNotificationsAction->plug(_optionsMenu);
 		
+		_optionsMenu->insertSeparator ();
+		
+		KAction *configureAutostartAction = actionCollection ()->action ("configure_autostart");
+		configureAutostartAction->plug (_optionsMenu);
+		
 		menu->insertItem (SmallIcon ("configure",  QIconSet::Automatic), i18n ("Options"), _optionsMenu, -1);
 	}
 
@@ -1076,6 +1100,10 @@
 	showNetworksDialogAction = new KAction (i18n ("Show Networks..."),
 						   0, 0,
 						   this, SLOT (showNetworksDialog ()), actionCollection (), "show_networks_dialog");
+	
+	configureAutostartAction = new KAction (i18n ("Configure Autostart..."),
+						0, 0,
+						this, SLOT (configureAutostart ()), actionCollection (), "configure_autostart");
 
 	connect (this->contextMenu (), SIGNAL (activated (int)),
 		 this,                 SLOT   (itemActivated (int)));
diff -ur knetworkmanager-0.1/knetworkmanager/src/knetworkmanager-tray.h knetworkmanager-0.1.new/knetworkmanager/src/knetworkmanager-tray.h
--- knetworkmanager-0.1/knetworkmanager/src/knetworkmanager-tray.h	2006-09-27 10:02:59.000000000 +0200
+++ knetworkmanager-0.1.new/knetworkmanager/src/knetworkmanager-tray.h	2007-01-19 15:42:32.000000000 +0100
@@ -118,6 +118,7 @@
 	  KAction* switchWirelessOnAction;
 	  KAction* switchWirelessOffAction;
 	  KAction* showNetworksDialogAction;
+	  KAction* configureAutostartAction;
 	  KAction* reportBugAction;
 	  KAction* aboutAction;
 
@@ -152,6 +153,7 @@
 	  void disconnectVPN          (void);
 	  void configureDialUp        (void);
 	  void configureNotifications (void);
+	  void configureAutostart     (void);
 	  void switchState            (void);
 	  void switchWirelessOn       (void);
 	  void switchWirelessOff      (void);
