diff -Nur knetworkmanager-0.1/knetworkmanager/src/knetworkmanager-tray.cpp knetworkmanager-0.1.new/knetworkmanager/src/knetworkmanager-tray.cpp
--- knetworkmanager-0.1/knetworkmanager/src/knetworkmanager-tray.cpp	2007-02-01 16:47:24.000000000 +0100
+++ knetworkmanager-0.1.new/knetworkmanager/src/knetworkmanager-tray.cpp	2007-02-01 16:52:21.000000000 +0100
@@ -942,9 +942,12 @@
 					pixmap = m_wireless25;
 				else
 					pixmap = m_wireless00;
+				m_serviceStub->setNetworkStatus( KDED_NETWORK_NAME, NetworkStatus::Online );
 			}
-			else 
+			else {
+				m_serviceStub->setNetworkStatus( KDED_NETWORK_NAME, NetworkStatus::Online );
 				pixmap = m_wired;
+}
 
 			m_animated = false;
 			
@@ -954,13 +957,23 @@
 			{
 				pixmap = pixmapForStage();
 				m_animated = true;
+				m_serviceStub->setNetworkStatus( KDED_NETWORK_NAME, NetworkStatus::Establishing );
 			}
-			else 
+			else {
+				if ( _ctx->getDeviceStore()->getDeviceList().isEmpty() )
+					m_serviceStub->setNetworkStatus( KDED_NETWORK_NAME, NetworkStatus::NoNetworks );
+				else
+					m_serviceStub->setNetworkStatus( KDED_NETWORK_NAME, NetworkStatus::Offline );
 				pixmap = m_disconnected;
+			}
 			break;
 		case NM_STATE_ASLEEP:
 		case NM_STATE_DISCONNECTED:
 		default:
+			if ( _ctx->getDeviceStore()->getDeviceList().isEmpty() )
+				m_serviceStub->setNetworkStatus( KDED_NETWORK_NAME, NetworkStatus::NoNetworks );
+			else
+				m_serviceStub->setNetworkStatus( KDED_NETWORK_NAME, NetworkStatus::Offline );
 			if (_ctx->getState ()->isDialupActive ())
 				pixmap = m_dialup;
 			else
@@ -1080,7 +1093,7 @@
 }
 
 
-Tray::Tray () : KSystemTray (), m_step( 0 )
+Tray::Tray () : KSystemTray (), m_step( 0 ), m_serviceStub( 0 )
 {
 /*	QMovie movie = KGlobal::iconLoader()->loadMovie("nm_detect", KIcon::Panel );
 // 	movie.setSpeed( 3 );
@@ -1161,6 +1174,17 @@
 	loadIcons();
 
 	connect ( &m_iconTimer, SIGNAL( timeout() ), SLOT( slotIconTimer() ) );
+	
+	// hack - talk to kded networkstatus on behalf of NM, as NM doesn't differentiate between 
+	// NM_STATE_DISCONNECTED when normally offline and when it has no networks/devices
+	m_serviceStub = new ServiceIface_stub( "kded", "networkstatus" );
+	NetworkStatus::Properties nsp;
+	nsp.internet = true; // we are the internet, we do mount it !
+	nsp.name = KDED_NETWORK_NAME;
+	nsp.onDemandPolicy = NetworkStatus::All;
+	nsp.service = kapp->dcopClient()->appId();
+	nsp.status = NetworkStatus::NoNetworks;
+	m_serviceStub->registerNetwork( KDED_NETWORK_NAME, nsp );
 }
 
 Tray::~Tray ()
diff -Nur knetworkmanager-0.1/knetworkmanager/src/knetworkmanager-tray.h knetworkmanager-0.1.new/knetworkmanager/src/knetworkmanager-tray.h
--- knetworkmanager-0.1/knetworkmanager/src/knetworkmanager-tray.h	2007-02-01 16:47:24.000000000 +0100
+++ knetworkmanager-0.1.new/knetworkmanager/src/knetworkmanager-tray.h	2007-02-01 16:48:58.000000000 +0100
@@ -59,6 +59,7 @@
 #include "knetworkmanager-devicestore.h"
 #include "knetworkmanager-vpn.h"
 #include "knetworkmanager-state.h"
+#include "serviceiface_stub.h"
 
 #define NUM_CONNECTING_STAGES 3
 #define NUM_CONNECTING_FRAMES 11
@@ -219,6 +220,9 @@
 	  QPixmap m_background;
 	  QPixmap m_overlay;
 	  QPoint _anchor;
+
+	  // stub for kded networkstatus dcop interface
+	  ServiceIface_stub * m_serviceStub;
 };
 
 #endif /* KNETWORKMANAGER_TRAY_H */
diff -Nur knetworkmanager-0.1/knetworkmanager/src/Makefile.am knetworkmanager-0.1.new/knetworkmanager/src/Makefile.am
--- knetworkmanager-0.1/knetworkmanager/src/Makefile.am	2007-02-01 16:47:23.000000000 +0100
+++ knetworkmanager-0.1.new/knetworkmanager/src/Makefile.am	2007-02-01 16:53:57.000000000 +0100
@@ -36,14 +36,15 @@
 	knetworkmanager-synchronizer.cpp knetworkmanager-dialup.cpp settings.kcfgc knetworkmanager-encryption.cpp \
 	knetworkmanager-tray.cpp activation.ui cryptowidget.ui networkchoicewidget.ui networklabelwidget.ui \
 	knetworkmanager-ui-networklistview.cpp errormsg.ui \
-	knetworkmanager-vpn.cpp knetworkmanager-vpn_dbus.cpp
+	knetworkmanager-vpn.cpp knetworkmanager-vpn_dbus.cpp serviceiface.stub
 
+serviceiface_DIR = /usr/include/kde
 
 kde_kcfg_DATA = knetworkmanager.kcfg
 
 knetworkmanager_la_LDFLAGS = -module $(KDE_RPATH) $(all_libraries) \
 	$(KDE_PLUGIN)
-knetworkmanager_la_LIBADD = -ldbus-qt-1 $(PACKAGE_LIBS) \
+knetworkmanager_la_LIBADD = -lnetworkstatus -ldbus-qt-1 $(PACKAGE_LIBS) \
 	$(LIB_KDEUI) -lkwalletclient -lkdeui -lkio
 
 # this is where the desktop file will go 
